{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import openai\n",
    "import pymupdf4llm\n",
    "import pymupdf.pro\n",
    "import langchain\n",
    "from langchain.text_splitter import MarkdownTextSplitter\n",
    "from langchain.text_splitter import RecursiveCharacterTextSplitter\n",
    "from langchain_core.documents import Document\n",
    "from langchain_community.vectorstores import FAISS\n",
    "from langchain_community.embeddings import OllamaEmbeddings\n",
    "from langchain_community.document_loaders import PyMuPDFLoader\n",
    "from langchain_core.prompts import ChatPromptTemplate, MessagesPlaceholder\n",
    "from langchain_openai import OpenAIEmbeddings\n",
    "from enum import Enum\n",
    "from dotenv import load_dotenv\n",
    "from openai import OpenAI\n",
    "import os\n",
    "import chromadb\n",
    "from langchain_openai import ChatOpenAI\n",
    "from langchain_chroma import Chroma\n",
    "from uuid import uuid4\n",
    "\n",
    "load_dotenv()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "path = \"./uploads/ISLP.pdf\"\n",
    "model = ChatOpenAI(model=\"gpt-3.5-turbo\", temperature=0.5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "client = OpenAI(api_key=os.getenv(\"DEEPSEEK_API_KEY\"), base_url=\"https://api.deepseek.com\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "loader = PyMuPDFLoader(path)\n",
    "text = loader.load()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "_text = str(' '.join(map(lambda x: x.page_content, text)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "text = Document(page_content=_text)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Document(metadata={}, page_content='(BSFUI\\x01+BNFT\\x01t\\x01%BOJFMB\\x018JUUFO\\x01t\\x015SFWPS\\x01)BTUJF\\n3PCFSU\\x015JCTIJSBOJ\\x01t\\x01+POBUIBO\\x015BZMPS\\n\"O\\x01*OUSPEVDUJPO\\nUP\\x014UBUJTUJDBM\\x01-FBSOJOH\\nXJUI\\x01\"QQMJDBUJPOT\\x01JO\\x011ZUIPO\\nFirst Printing: July 5, 2023 To our parents:\\nAlison and Michael James\\nChiara Nappi and Edward Witten\\nValerie and Patrick Hastie\\nVera and Sami Tibshirani\\nJohn and Brenda Taylor\\nand to our families:\\nMichael, Daniel, and Catherine\\nTessa, Theo, Otto, and Ari\\nSamantha, Timothy, and Lynda\\nCharlie, Ryan, Julie, and Cheryl\\nLee-Ann and Isobel Preface\\nStatistical learning refers to a set of tools for making sense of complex\\ndatasets. In recent years, we have seen a staggering increase in the scale and\\nscope of data collection across virtually all areas of science and industry.\\nAs a result, statistical learning has become a critical toolkit for anyone who\\nwishes to understand data — and as more and more of today’s jobs involve\\ndata, this means that statistical learning is fast becoming a critical toolkit\\nfor everyone.\\nOne of the first books on statistical learning — The Elements of Statisti-\\ncal Learning (ESL, by Hastie, Tibshirani, and Friedman) — was published\\nin 2001, with a second edition in 2009. ESL has become a popular text not\\nonly in statistics but also in related fields. One of the reasons for ESL’s\\npopularity is its relatively accessible style. But ESL is best-suited for indi-\\nviduals with advanced training in the mathematical sciences.\\nAn Introduction to Statistical Learning, With Applications in R (ISLR)\\n— first published in 2013, with a second edition in 2021 — arose from\\nthe clear need for a broader and less technical treatment of the key topics\\nin statistical learning. In addition to a review of linear regression, ISLR\\ncovers many of today’s most important statistical and machine learning\\napproaches, including resampling, sparse methods for classification and re-\\ngression, generalized additive models, tree-based methods, support vector\\nmachines, deep learning, survival analysis, clustering, and multiple testing.\\nSince it was published in 2013, ISLR has become a mainstay of un-\\ndergraduate and graduate classrooms worldwide, as well as an important\\nreference book for data scientists. One of the keys to its success has been\\nthat, beginning with Chapter 2, each chapter contains an R lab illustrating\\nhow to implement the statistical learning methods seen in that chapter,\\nproviding the reader with valuable hands-on experience.\\nHowever, in recent years Python has become an increasingly popular lan-\\nguage for data science, and there has been increasing demand for a Python-\\nvii viii\\nbased alternative to ISLR. Hence, this book, An Introduction to Statistical\\nLearning, With Applications in Python (ISLP), covers the same materials\\nas ISLR but with labs implemented in Python — a feat accomplished by the\\naddition of a new co-author, Jonathan Taylor. Several of the labs make use\\nof the ISLP Python package, which we have written to facilitate carrying out\\nthe statistical learning methods covered in each chapter in Python. These\\nlabs will be useful both for Python novices, as well as experienced users.\\nThe intention behind ISLP (and ISLR) is to concentrate more on the\\napplications of the methods and less on the mathematical details, so it is\\nappropriate for advanced undergraduates or master’s students in statistics\\nor related quantitative fields, or for individuals in other disciplines who\\nwish to use statistical learning tools to analyze their data. It can be used\\nas a textbook for a course spanning two semesters.\\nWe are grateful to these readers for providing valuable comments on the\\nfirst edition of ISLR: Pallavi Basu, Alexandra Chouldechova, Patrick Dana-\\nher, Will Fithian, Luella Fu, Sam Gross, Max Grazier G’Sell, Courtney\\nPaulson, Xinghao Qiao, Elisa Sheng, Noah Simon, Kean Ming Tan, Xin Lu\\nTan. We thank these readers for helpful input on the second edition of ISLR:\\nAlan Agresti, Iain Carmichael, Yiqun Chen, Erin Craig, Daisy Ding, Lucy\\nGao, Ismael Lemhadri, Bryan Martin, Anna Neufeld, Geoff Tims, Carsten\\nVoelkmann, Steve Yadlowsky, and James Zou. We are immensely grateful\\nto Balasubramanian “Naras” Narasimhan for his assistance on both ISLR\\nand ISLP.\\nIt has been an honor and a privilege for us to see the considerable impact\\nthat ISLR has had on the way in which statistical learning is practiced, both\\nin and out of the academic setting. We hope that this new Python edition\\nwill continue to give today’s and tomorrow’s applied statisticians and data\\nscientists the tools they need for success in a data-driven world.\\nIt’s tough to make predictions, especially about the future.\\n-Yogi Berra\\nPreface Contents\\nPreface\\nvii\\n1\\nIntroduction\\n1\\n2\\nStatistical Learning\\n15\\n2.1\\nWhat Is Statistical Learning? . . . . . . . . . . . . . . . . .\\n15\\n2.1.1\\nWhy Estimate f? . . . . . . . . . . . . . . . . . . .\\n17\\n2.1.2\\nHow Do We Estimate f? . . . . . . . . . . . . . . .\\n20\\n2.1.3\\nThe Trade-Off Between Prediction Accuracy\\nand Model Interpretability . . . . . . . . . . . . . .\\n23\\n2.1.4\\nSupervised Versus Unsupervised Learning . . . . .\\n25\\n2.1.5\\nRegression Versus Classification Problems . . . . .\\n27\\n2.2\\nAssessing Model Accuracy\\n. . . . . . . . . . . . . . . . . .\\n27\\n2.2.1\\nMeasuring the Quality of Fit\\n. . . . . . . . . . . .\\n28\\n2.2.2\\nThe Bias-Variance Trade-Off . . . . . . . . . . . . .\\n31\\n2.2.3\\nThe Classification Setting\\n. . . . . . . . . . . . . .\\n34\\n2.3\\nLab: Introduction to Python . . . . . . . . . . . . . . . . .\\n40\\n2.3.1\\nGetting Started . . . . . . . . . . . . . . . . . . . .\\n40\\n2.3.2\\nBasic Commands . . . . . . . . . . . . . . . . . . .\\n40\\n2.3.3\\nIntroduction to Numerical Python\\n. . . . . . . . .\\n42\\n2.3.4\\nGraphics . . . . . . . . . . . . . . . . . . . . . . . .\\n48\\n2.3.5\\nSequences and Slice Notation\\n. . . . . . . . . . . .\\n51\\n2.3.6\\nIndexing Data . . . . . . . . . . . . . . . . . . . . .\\n51\\n2.3.7\\nLoading Data . . . . . . . . . . . . . . . . . . . . .\\n55\\n2.3.8\\nFor Loops . . . . . . . . . . . . . . . . . . . . . . .\\n59\\n2.3.9\\nAdditional Graphical and Numerical Summaries . .\\n61\\n2.4\\nExercises . . . . . . . . . . . . . . . . . . . . . . . . . . . .\\n63\\n3\\nLinear Regression\\n69\\n3.1\\nSimple Linear Regression . . . . . . . . . . . . . . . . . . .\\n70\\n3.1.1\\nEstimating the Coefficients\\n. . . . . . . . . . . . .\\n71\\n3.1.2\\nAssessing the Accuracy of the Coefficient\\nEstimates . . . . . . . . . . . . . . . . . . . . . . .\\n72\\n3.1.3\\nAssessing the Accuracy of the Model . . . . . . . .\\n77\\n3.2\\nMultiple Linear Regression . . . . . . . . . . . . . . . . . .\\n80\\n3.2.1\\nEstimating the Regression Coefficients . . . . . . .\\n81\\nix x\\nContents\\n3.2.2\\nSome Important Questions . . . . . . . . . . . . . .\\n83\\n3.3\\nOther Considerations in the Regression Model . . . . . . .\\n91\\n3.3.1\\nQualitative Predictors\\n. . . . . . . . . . . . . . . .\\n91\\n3.3.2\\nExtensions of the Linear Model . . . . . . . . . . .\\n94\\n3.3.3\\nPotential Problems . . . . . . . . . . . . . . . . . .\\n100\\n3.4\\nThe Marketing Plan . . . . . . . . . . . . . . . . . . . . . .\\n109\\n3.5\\nComparison of Linear Regression with K-Nearest\\nNeighbors . . . . . . . . . . . . . . . . . . . . . . . . . . . .\\n111\\n3.6\\nLab: Linear Regression\\n. . . . . . . . . . . . . . . . . . . .\\n116\\n3.6.1\\nImporting packages . . . . . . . . . . . . . . . . . .\\n116\\n3.6.2\\nSimple Linear Regression . . . . . . . . . . . . . . .\\n117\\n3.6.3\\nMultiple Linear Regression . . . . . . . . . . . . . .\\n122\\n3.6.4\\nMultivariate Goodness of Fit\\n. . . . . . . . . . . .\\n123\\n3.6.5\\nInteraction Terms . . . . . . . . . . . . . . . . . . .\\n124\\n3.6.6\\nNon-linear Transformations of the Predictors\\n. . .\\n125\\n3.6.7\\nQualitative Predictors\\n. . . . . . . . . . . . . . . .\\n126\\n3.7\\nExercises . . . . . . . . . . . . . . . . . . . . . . . . . . . .\\n127\\n4\\nClassification\\n135\\n4.1\\nAn Overview of Classification . . . . . . . . . . . . . . . . .\\n135\\n4.2\\nWhy Not Linear Regression? . . . . . . . . . . . . . . . . .\\n136\\n4.3\\nLogistic Regression\\n. . . . . . . . . . . . . . . . . . . . . .\\n138\\n4.3.1\\nThe Logistic Model . . . . . . . . . . . . . . . . . .\\n139\\n4.3.2\\nEstimating the Regression Coefficients . . . . . . .\\n140\\n4.3.3\\nMaking Predictions . . . . . . . . . . . . . . . . . .\\n141\\n4.3.4\\nMultiple Logistic Regression . . . . . . . . . . . . .\\n142\\n4.3.5\\nMultinomial Logistic Regression . . . . . . . . . . .\\n144\\n4.4\\nGenerative Models for Classification . . . . . . . . . . . . .\\n146\\n4.4.1\\nLinear Discriminant Analysis for p = 1 . . . . . . .\\n147\\n4.4.2\\nLinear Discriminant Analysis for p >1\\n. . . . . . .\\n150\\n4.4.3\\nQuadratic Discriminant Analysis\\n. . . . . . . . . .\\n156\\n4.4.4\\nNaive Bayes . . . . . . . . . . . . . . . . . . . . . .\\n158\\n4.5\\nA Comparison of Classification Methods\\n. . . . . . . . . .\\n161\\n4.5.1\\nAn Analytical Comparison . . . . . . . . . . . . . .\\n161\\n4.5.2\\nAn Empirical Comparison . . . . . . . . . . . . . .\\n164\\n4.6\\nGeneralized Linear Models . . . . . . . . . . . . . . . . . .\\n167\\n4.6.1\\nLinear Regression on the Bikeshare Data . . . . . .\\n167\\n4.6.2\\nPoisson Regression on the Bikeshare Data . . . . .\\n169\\n4.6.3\\nGeneralized Linear Models in Greater Generality .\\n172\\n4.7\\nLab: Logistic Regression, LDA, QDA, and KNN . . . . . .\\n173\\n4.7.1\\nThe Stock Market Data\\n. . . . . . . . . . . . . . .\\n173\\n4.7.2\\nLogistic Regression . . . . . . . . . . . . . . . . . .\\n174\\n4.7.3\\nLinear Discriminant Analysis\\n. . . . . . . . . . . .\\n179\\n4.7.4\\nQuadratic Discriminant Analysis\\n. . . . . . . . . .\\n181\\n4.7.5\\nNaive Bayes . . . . . . . . . . . . . . . . . . . . . .\\n182\\n4.7.6\\nK-Nearest Neighbors . . . . . . . . . . . . . . . . .\\n183\\n4.7.7\\nLinear and Poisson Regression on the Bikeshare Data188\\n4.8\\nExercises . . . . . . . . . . . . . . . . . . . . . . . . . . . .\\n193 Contents\\nxi\\n5\\nResampling Methods\\n201\\n5.1\\nCross-Validation . . . . . . . . . . . . . . . . . . . . . . . .\\n202\\n5.1.1\\nThe Validation Set Approach\\n. . . . . . . . . . . .\\n202\\n5.1.2\\nLeave-One-Out Cross-Validation\\n. . . . . . . . . .\\n204\\n5.1.3\\nk-Fold Cross-Validation\\n. . . . . . . . . . . . . . .\\n206\\n5.1.4\\nBias-Variance Trade-Off for k-Fold\\nCross-Validation\\n. . . . . . . . . . . . . . . . . . .\\n208\\n5.1.5\\nCross-Validation on Classification Problems . . . .\\n209\\n5.2\\nThe Bootstrap . . . . . . . . . . . . . . . . . . . . . . . . .\\n212\\n5.3\\nLab: Cross-Validation and the Bootstrap\\n. . . . . . . . . .\\n215\\n5.3.1\\nThe Validation Set Approach\\n. . . . . . . . . . . .\\n216\\n5.3.2\\nCross-Validation\\n. . . . . . . . . . . . . . . . . . .\\n217\\n5.3.3\\nThe Bootstrap\\n. . . . . . . . . . . . . . . . . . . .\\n220\\n5.4\\nExercises . . . . . . . . . . . . . . . . . . . . . . . . . . . .\\n224\\n6\\nLinear Model Selection and Regularization\\n229\\n6.1\\nSubset Selection . . . . . . . . . . . . . . . . . . . . . . . .\\n231\\n6.1.1\\nBest Subset Selection . . . . . . . . . . . . . . . . .\\n231\\n6.1.2\\nStepwise Selection\\n. . . . . . . . . . . . . . . . . .\\n233\\n6.1.3\\nChoosing the Optimal Model\\n. . . . . . . . . . . .\\n235\\n6.2\\nShrinkage Methods\\n. . . . . . . . . . . . . . . . . . . . . .\\n240\\n6.2.1\\nRidge Regression . . . . . . . . . . . . . . . . . . .\\n240\\n6.2.2\\nThe Lasso . . . . . . . . . . . . . . . . . . . . . . .\\n244\\n6.2.3\\nSelecting the Tuning Parameter . . . . . . . . . . .\\n252\\n6.3\\nDimension Reduction Methods . . . . . . . . . . . . . . . .\\n253\\n6.3.1\\nPrincipal Components Regression . . . . . . . . . .\\n254\\n6.3.2\\nPartial Least Squares . . . . . . . . . . . . . . . . .\\n260\\n6.4\\nConsiderations in High Dimensions\\n. . . . . . . . . . . . .\\n262\\n6.4.1\\nHigh-Dimensional Data . . . . . . . . . . . . . . . .\\n262\\n6.4.2\\nWhat Goes Wrong in High Dimensions? . . . . . .\\n263\\n6.4.3\\nRegression in High Dimensions\\n. . . . . . . . . . .\\n265\\n6.4.4\\nInterpreting Results in High Dimensions . . . . . .\\n266\\n6.5\\nLab: Linear Models and Regularization Methods . . . . . .\\n267\\n6.5.1\\nSubset Selection Methods\\n. . . . . . . . . . . . . .\\n268\\n6.5.2\\nRidge Regression and the Lasso . . . . . . . . . . .\\n273\\n6.5.3\\nPCR and PLS Regression\\n. . . . . . . . . . . . . .\\n280\\n6.6\\nExercises . . . . . . . . . . . . . . . . . . . . . . . . . . . .\\n283\\n7\\nMoving Beyond Linearity\\n289\\n7.1\\nPolynomial Regression . . . . . . . . . . . . . . . . . . . . .\\n290\\n7.2\\nStep Functions . . . . . . . . . . . . . . . . . . . . . . . . .\\n292\\n7.3\\nBasis Functions\\n. . . . . . . . . . . . . . . . . . . . . . . .\\n293\\n7.4\\nRegression Splines . . . . . . . . . . . . . . . . . . . . . . .\\n294\\n7.4.1\\nPiecewise Polynomials . . . . . . . . . . . . . . . .\\n294\\n7.4.2\\nConstraints and Splines\\n. . . . . . . . . . . . . . .\\n296\\n7.4.3\\nThe Spline Basis Representation\\n. . . . . . . . . .\\n296\\n7.4.4\\nChoosing the Number and Locations\\nof the Knots . . . . . . . . . . . . . . . . . . . . . .\\n297\\n7.4.5\\nComparison to Polynomial Regression\\n. . . . . . .\\n299 xii\\nContents\\n7.5\\nSmoothing Splines . . . . . . . . . . . . . . . . . . . . . . .\\n300\\n7.5.1\\nAn Overview of Smoothing Splines . . . . . . . . .\\n300\\n7.5.2\\nChoosing the Smoothing Parameter λ\\n. . . . . . .\\n301\\n7.6\\nLocal Regression . . . . . . . . . . . . . . . . . . . . . . . .\\n303\\n7.7\\nGeneralized Additive Models . . . . . . . . . . . . . . . . .\\n305\\n7.7.1\\nGAMs for Regression Problems . . . . . . . . . . .\\n306\\n7.7.2\\nGAMs for Classification Problems . . . . . . . . . .\\n308\\n7.8\\nLab: Non-Linear Modeling\\n. . . . . . . . . . . . . . . . . .\\n309\\n7.8.1\\nPolynomial Regression and Step Functions . . . . .\\n310\\n7.8.2\\nSplines . . . . . . . . . . . . . . . . . . . . . . . . .\\n315\\n7.8.3\\nSmoothing Splines and GAMs . . . . . . . . . . . .\\n317\\n7.8.4\\nLocal Regression\\n. . . . . . . . . . . . . . . . . . .\\n324\\n7.9\\nExercises . . . . . . . . . . . . . . . . . . . . . . . . . . . .\\n325\\n8\\nTree-Based Methods\\n331\\n8.1\\nThe Basics of Decision Trees . . . . . . . . . . . . . . . . .\\n331\\n8.1.1\\nRegression Trees\\n. . . . . . . . . . . . . . . . . . .\\n331\\n8.1.2\\nClassification Trees . . . . . . . . . . . . . . . . . .\\n337\\n8.1.3\\nTrees Versus Linear Models\\n. . . . . . . . . . . . .\\n341\\n8.1.4\\nAdvantages and Disadvantages of Trees . . . . . . .\\n341\\n8.2\\nBagging, Random Forests, Boosting, and Bayesian Additive\\nRegression Trees . . . . . . . . . . . . . . . . . . . . . . . .\\n343\\n8.2.1\\nBagging . . . . . . . . . . . . . . . . . . . . . . . .\\n343\\n8.2.2\\nRandom Forests . . . . . . . . . . . . . . . . . . . .\\n346\\n8.2.3\\nBoosting . . . . . . . . . . . . . . . . . . . . . . . .\\n347\\n8.2.4\\nBayesian Additive Regression Trees . . . . . . . . .\\n350\\n8.2.5\\nSummary of Tree Ensemble Methods . . . . . . . .\\n353\\n8.3\\nLab: Tree-Based Methods . . . . . . . . . . . . . . . . . . .\\n354\\n8.3.1\\nFitting Classification Trees . . . . . . . . . . . . . .\\n355\\n8.3.2\\nFitting Regression Trees . . . . . . . . . . . . . . .\\n358\\n8.3.3\\nBagging and Random Forests . . . . . . . . . . . .\\n360\\n8.3.4\\nBoosting . . . . . . . . . . . . . . . . . . . . . . . .\\n361\\n8.3.5\\nBayesian Additive Regression Trees . . . . . . . . .\\n362\\n8.4\\nExercises . . . . . . . . . . . . . . . . . . . . . . . . . . . .\\n363\\n9\\nSupport Vector Machines\\n367\\n9.1\\nMaximal Margin Classifier\\n. . . . . . . . . . . . . . . . . .\\n367\\n9.1.1\\nWhat Is a Hyperplane? . . . . . . . . . . . . . . . .\\n368\\n9.1.2\\nClassification Using a Separating Hyperplane\\n. . .\\n368\\n9.1.3\\nThe Maximal Margin Classifier . . . . . . . . . . .\\n370\\n9.1.4\\nConstruction of the Maximal Margin Classifier\\n. .\\n372\\n9.1.5\\nThe Non-separable Case . . . . . . . . . . . . . . .\\n372\\n9.2\\nSupport Vector Classifiers . . . . . . . . . . . . . . . . . . .\\n373\\n9.2.1\\nOverview of the Support Vector Classifier\\n. . . . .\\n373\\n9.2.2\\nDetails of the Support Vector Classifier . . . . . . .\\n374\\n9.3\\nSupport Vector Machines . . . . . . . . . . . . . . . . . . .\\n377\\n9.3.1\\nClassification with Non-Linear Decision\\nBoundaries\\n. . . . . . . . . . . . . . . . . . . . . .\\n378\\n9.3.2\\nThe Support Vector Machine\\n. . . . . . . . . . . .\\n379 Contents\\nxiii\\n9.3.3\\nAn Application to the Heart Disease Data . . . . .\\n382\\n9.4\\nSVMs with More than Two Classes\\n. . . . . . . . . . . . .\\n383\\n9.4.1\\nOne-Versus-One Classification . . . . . . . . . . . .\\n384\\n9.4.2\\nOne-Versus-All Classification\\n. . . . . . . . . . . .\\n384\\n9.5\\nRelationship to Logistic Regression\\n. . . . . . . . . . . . .\\n384\\n9.6\\nLab: Support Vector Machines . . . . . . . . . . . . . . . .\\n387\\n9.6.1\\nSupport Vector Classifier . . . . . . . . . . . . . . .\\n387\\n9.6.2\\nSupport Vector Machine . . . . . . . . . . . . . . .\\n390\\n9.6.3\\nROC Curves . . . . . . . . . . . . . . . . . . . . . .\\n392\\n9.6.4\\nSVM with Multiple Classes\\n. . . . . . . . . . . . .\\n393\\n9.6.5\\nApplication to Gene Expression Data . . . . . . . .\\n394\\n9.7\\nExercises . . . . . . . . . . . . . . . . . . . . . . . . . . . .\\n395\\n10 Deep Learning\\n399\\n10.1\\nSingle Layer Neural Networks\\n. . . . . . . . . . . . . . . .\\n400\\n10.2\\nMultilayer Neural Networks . . . . . . . . . . . . . . . . . .\\n402\\n10.3\\nConvolutional Neural Networks . . . . . . . . . . . . . . . .\\n406\\n10.3.1\\nConvolution Layers . . . . . . . . . . . . . . . . . .\\n407\\n10.3.2\\nPooling Layers\\n. . . . . . . . . . . . . . . . . . . .\\n410\\n10.3.3\\nArchitecture of a Convolutional Neural Network . .\\n410\\n10.3.4\\nData Augmentation . . . . . . . . . . . . . . . . . .\\n411\\n10.3.5\\nResults Using a Pretrained Classifier . . . . . . . .\\n412\\n10.4\\nDocument Classification . . . . . . . . . . . . . . . . . . . .\\n413\\n10.5\\nRecurrent Neural Networks . . . . . . . . . . . . . . . . . .\\n416\\n10.5.1\\nSequential Models for Document Classification\\n. .\\n418\\n10.5.2\\nTime Series Forecasting\\n. . . . . . . . . . . . . . .\\n420\\n10.5.3\\nSummary of RNNs . . . . . . . . . . . . . . . . . .\\n424\\n10.6\\nWhen to Use Deep Learning\\n. . . . . . . . . . . . . . . . .\\n425\\n10.7\\nFitting a Neural Network . . . . . . . . . . . . . . . . . . .\\n427\\n10.7.1\\nBackpropagation\\n. . . . . . . . . . . . . . . . . . .\\n428\\n10.7.2\\nRegularization and Stochastic Gradient Descent . .\\n429\\n10.7.3\\nDropout Learning . . . . . . . . . . . . . . . . . . .\\n431\\n10.7.4\\nNetwork Tuning . . . . . . . . . . . . . . . . . . . .\\n431\\n10.8\\nInterpolation and Double Descent . . . . . . . . . . . . . .\\n432\\n10.9\\nLab: Deep Learning . . . . . . . . . . . . . . . . . . . . . .\\n435\\n10.9.1\\nSingle Layer Network on Hitters Data\\n. . . . . . .\\n437\\n10.9.2\\nMultilayer Network on the MNIST Digit Data . . .\\n444\\n10.9.3\\nConvolutional Neural Networks . . . . . . . . . . .\\n448\\n10.9.4\\nUsing Pretrained CNN Models\\n. . . . . . . . . . .\\n452\\n10.9.5\\nIMDB Document Classification . . . . . . . . . . .\\n454\\n10.9.6\\nRecurrent Neural Networks\\n. . . . . . . . . . . . .\\n458\\n10.10 Exercises . . . . . . . . . . . . . . . . . . . . . . . . . . . .\\n465\\n11 Survival Analysis and Censored Data\\n469\\n11.1\\nSurvival and Censoring Times\\n. . . . . . . . . . . . . . . .\\n470\\n11.2\\nA Closer Look at Censoring . . . . . . . . . . . . . . . . . .\\n470\\n11.3\\nThe Kaplan–Meier Survival Curve . . . . . . . . . . . . . .\\n472\\n11.4\\nThe Log-Rank Test\\n. . . . . . . . . . . . . . . . . . . . . .\\n474\\n11.5\\nRegression Models With a Survival Response . . . . . . . .\\n476 xiv\\nContents\\n11.5.1\\nThe Hazard Function . . . . . . . . . . . . . . . . .\\n476\\n11.5.2\\nProportional Hazards . . . . . . . . . . . . . . . . .\\n478\\n11.5.3\\nExample: Brain Cancer Data\\n. . . . . . . . . . . .\\n482\\n11.5.4\\nExample: Publication Data\\n. . . . . . . . . . . . .\\n482\\n11.6\\nShrinkage for the Cox Model . . . . . . . . . . . . . . . . .\\n484\\n11.7\\nAdditional Topics\\n. . . . . . . . . . . . . . . . . . . . . . .\\n486\\n11.7.1\\nArea Under the Curve for Survival Analysis . . . .\\n486\\n11.7.2\\nChoice of Time Scale . . . . . . . . . . . . . . . . .\\n487\\n11.7.3\\nTime-Dependent Covariates . . . . . . . . . . . . .\\n488\\n11.7.4\\nChecking the Proportional Hazards Assumption . .\\n488\\n11.7.5\\nSurvival Trees . . . . . . . . . . . . . . . . . . . . .\\n488\\n11.8\\nLab: Survival Analysis . . . . . . . . . . . . . . . . . . . . .\\n489\\n11.8.1\\nBrain Cancer Data . . . . . . . . . . . . . . . . . .\\n489\\n11.8.2\\nPublication Data . . . . . . . . . . . . . . . . . . .\\n493\\n11.8.3\\nCall Center Data . . . . . . . . . . . . . . . . . . .\\n494\\n11.9\\nExercises . . . . . . . . . . . . . . . . . . . . . . . . . . . .\\n498\\n12 Unsupervised Learning\\n503\\n12.1\\nThe Challenge of Unsupervised Learning\\n. . . . . . . . . .\\n503\\n12.2\\nPrincipal Components Analysis . . . . . . . . . . . . . . . .\\n504\\n12.2.1\\nWhat Are Principal Components? . . . . . . . . . .\\n505\\n12.2.2\\nAnother Interpretation of Principal Components\\n.\\n508\\n12.2.3\\nThe Proportion of Variance Explained . . . . . . .\\n510\\n12.2.4\\nMore on PCA . . . . . . . . . . . . . . . . . . . . .\\n512\\n12.2.5\\nOther Uses for Principal Components . . . . . . . .\\n515\\n12.3\\nMissing Values and Matrix Completion . . . . . . . . . . .\\n515\\n12.4\\nClustering Methods . . . . . . . . . . . . . . . . . . . . . .\\n520\\n12.4.1\\nK-Means Clustering\\n. . . . . . . . . . . . . . . . .\\n521\\n12.4.2\\nHierarchical Clustering . . . . . . . . . . . . . . . .\\n525\\n12.4.3\\nPractical Issues in Clustering\\n. . . . . . . . . . . .\\n532\\n12.5\\nLab: Unsupervised Learning\\n. . . . . . . . . . . . . . . . .\\n535\\n12.5.1\\nPrincipal Components Analysis . . . . . . . . . . .\\n535\\n12.5.2\\nMatrix Completion . . . . . . . . . . . . . . . . . .\\n539\\n12.5.3\\nClustering . . . . . . . . . . . . . . . . . . . . . . .\\n542\\n12.5.4\\nNCI60 Data Example . . . . . . . . . . . . . . . . .\\n546\\n12.6\\nExercises . . . . . . . . . . . . . . . . . . . . . . . . . . . .\\n552\\n13 Multiple Testing\\n557\\n13.1\\nA Quick Review of Hypothesis Testing\\n. . . . . . . . . . .\\n558\\n13.1.1\\nTesting a Hypothesis . . . . . . . . . . . . . . . . .\\n558\\n13.1.2\\nType I and Type II Errors . . . . . . . . . . . . . .\\n562\\n13.2\\nThe Challenge of Multiple Testing . . . . . . . . . . . . . .\\n563\\n13.3\\nThe Family-Wise Error Rate . . . . . . . . . . . . . . . . .\\n565\\n13.3.1\\nWhat is the Family-Wise Error Rate?\\n. . . . . . .\\n565\\n13.3.2\\nApproaches to Control the Family-Wise Error Rate 567\\n13.3.3\\nTrade-Off Between the FWER and Power\\n. . . . .\\n572\\n13.4\\nThe False Discovery Rate . . . . . . . . . . . . . . . . . . .\\n573\\n13.4.1\\nIntuition for the False Discovery Rate\\n. . . . . . .\\n573\\n13.4.2\\nThe Benjamini–Hochberg Procedure\\n. . . . . . . .\\n575 Contents\\nxv\\n13.5\\nA Re-Sampling Approach to p-Values and False Discovery\\nRates . . . . . . . . . . . . . . . . . . . . . . . . . . . . . .\\n577\\n13.5.1\\nA Re-Sampling Approach to the p-Value . . . . . .\\n578\\n13.5.2\\nA Re-Sampling Approach to the False Discovery Rate579\\n13.5.3\\nWhen Are Re-Sampling Approaches Useful? . . . .\\n581\\n13.6\\nLab: Multiple Testing . . . . . . . . . . . . . . . . . . . . .\\n583\\n13.6.1\\nReview of Hypothesis Tests\\n. . . . . . . . . . . . .\\n583\\n13.6.2\\nFamily-Wise Error Rate . . . . . . . . . . . . . . .\\n585\\n13.6.3\\nFalse Discovery Rate . . . . . . . . . . . . . . . . .\\n588\\n13.6.4\\nA Re-Sampling Approach\\n. . . . . . . . . . . . . .\\n590\\n13.7\\nExercises . . . . . . . . . . . . . . . . . . . . . . . . . . . .\\n593\\nIndex\\n597 1\\nIntroduction\\nAn Overview of Statistical Learning\\nStatistical learning refers to a vast set of tools for understanding data. These\\ntools can be classified as supervised or unsupervised. Broadly speaking,\\nsupervised statistical learning involves building a statistical model for pre-\\ndicting, or estimating, an output based on one or more inputs. Problems of\\nthis nature occur in fields as diverse as business, medicine, astrophysics, and\\npublic policy. With unsupervised statistical learning, there are inputs but\\nno supervising output; nevertheless we can learn relationships and struc-\\nture from such data. To provide an illustration of some applications of\\nstatistical learning, we briefly discuss three real-world data sets that are\\nconsidered in this book.\\nWage Data\\nIn this application (which we refer to as the Wage data set throughout this\\nbook), we examine a number of factors that relate to wages for a group of\\nmen from the Atlantic region of the United States. In particular, we wish\\nto understand the association between an employee’s age and education, as\\nwell as the calendar year, on his wage. Consider, for example, the left-hand\\npanel of Figure 1.1, which displays wage versus age for each of the individu-\\nals in the data set. There is evidence that wage increases with age but then\\ndecreases again after approximately age 60. The blue line, which provides\\nan estimate of the average wage for a given age, makes this trend clearer.\\nGiven an employee’s age, we can use this curve to predict his wage. However,\\nit is also clear from Figure 1.1 that there is a significant amount of vari-\\nability associated with this average value, and so age alone is unlikely to\\nprovide an accurate prediction of a particular man’s wage.\\n© Springer Nature Switzerland AG 2023 \\nG. James et al., An Introduction to Statistical Learning, Springer Texts in Statistics, \\nhttps://doi.org/10.1007/978-3-031-38747-0_1 \\n1 2\\n1. Introduction\\n20\\n40\\n60\\n80\\n50\\n100\\n200\\n300\\nAge\\nWage\\n2003\\n2006\\n2009\\n50\\n100\\n200\\n300\\nYear\\nWage\\n1\\n2\\n3\\n4\\n5\\n50\\n100\\n200\\n300\\nEducation Level\\nWage\\nFIGURE 1.1. Wage data, which contains income survey information for men\\nfrom the central Atlantic region of the United States. Left: wage as a function of\\nage. On average, wage increases with age until about 60 years of age, at which\\npoint it begins to decline. Center: wage as a function of year. There is a slow\\nbut steady increase of approximately $10,000 in the average wage between 2003\\nand 2009. Right: Boxplots displaying wage as a function of education, with 1\\nindicating the lowest level (no high school diploma) and 5 the highest level (an\\nadvanced graduate degree). On average, wage increases with the level of education.\\nWe also have information regarding each employee’s education level and\\nthe year in which the wage was earned. The center and right-hand panels\\nof Figure 1.1, which display wage as a function of both year and education,\\nindicate that both of these factors are associated with wage. Wages increase\\nby approximately $10,000, in a roughly linear (or straight-line) fashion,\\nbetween 2003 and 2009, though this rise is very slight relative to the vari-\\nability in the data. Wages are also typically greater for individuals with\\nhigher education levels: men with the lowest education level (1) tend to\\nhave substantially lower wages than those with the highest education level\\n(5). Clearly, the most accurate prediction of a given man’s wage will be\\nobtained by combining his age, his education, and the year. In Chapter 3,\\nwe discuss linear regression, which can be used to predict wage from this\\ndata set. Ideally, we should predict wage in a way that accounts for the\\nnon-linear relationship between wage and age. In Chapter 7, we discuss a\\nclass of approaches for addressing this problem.\\nStock Market Data\\nThe Wage data involves predicting a continuous or quantitative output value.\\nThis is often referred to as a regression problem. However, in certain cases\\nwe may instead wish to predict a non-numerical value—that is, a categorical\\nor qualitative output. For example, in Chapter 4 we examine a stock market\\ndata set that contains the daily movements in the Standard & Poor’s 500\\n(S&P) stock index over a 5-year period between 2001 and 2005. We refer\\nto this as the Smarket data. The goal is to predict whether the index will\\nincrease or decrease on a given day, using the past 5 days’ percentage\\nchanges in the index. Here the statistical learning problem does not involve\\npredicting a numerical value. Instead it involves predicting whether a given 1. Introduction\\n3\\nDown\\nUp\\n−4\\n−2\\n0\\n2\\n4\\n6\\nYesterday\\nToday’s Direction\\nPercentage change in S&P\\nDown\\nUp\\n−4\\n−2\\n0\\n2\\n4\\n6\\nTwo Days Previous\\nToday’s Direction\\nPercentage change in S&P\\nDown\\nUp\\n−4\\n−2\\n0\\n2\\n4\\n6\\nThree Days Previous\\nToday’s Direction\\nPercentage change in S&P\\nFIGURE 1.2. Left: Boxplots of the previous day’s percentage change in the S&P\\nindex for the days for which the market increased or decreased, obtained from the\\nSmarket data. Center and Right: Same as left panel, but the percentage changes\\nfor 2 and 3 days previous are shown.\\nday’s stock market performance will fall into the Up bucket or the Down\\nbucket. This is known as a classification problem. A model that could\\naccurately predict the direction in which the market will move would be\\nvery useful!\\nThe left-hand panel of Figure 1.2 displays two boxplots of the previous\\nday’s percentage changes in the stock index: one for the 648 days for which\\nthe market increased on the subsequent day, and one for the 602 days for\\nwhich the market decreased. The two plots look almost identical, suggest-\\ning that there is no simple strategy for using yesterday’s movement in the\\nS&P to predict today’s returns. The remaining panels, which display box-\\nplots for the percentage changes 2 and 3 days previous to today, similarly\\nindicate little association between past and present returns. Of course, this\\nlack of pattern is to be expected: in the presence of strong correlations be-\\ntween successive days’ returns, one could adopt a simple trading strategy\\nto generate profits from the market. Nevertheless, in Chapter 4, we explore\\nthese data using several different statistical learning methods. Interestingly,\\nthere are hints of some weak trends in the data that suggest that, at least\\nfor this 5-year period, it is possible to correctly predict the direction of\\nmovement in the market approximately 60% of the time (Figure 1.3).\\nGene Expression Data\\nThe previous two applications illustrate data sets with both input and\\noutput variables. However, another important class of problems involves\\nsituations in which we only observe input variables, with no corresponding\\noutput. For example, in a marketing setting, we might have demographic\\ninformation for a number of current or potential customers. We may wish to\\nunderstand which types of customers are similar to each other by grouping\\nindividuals according to their observed characteristics. This is known as a 4\\n1. Introduction\\nDown\\nUp\\n0.46\\n0.48\\n0.50\\n0.52\\nToday’s Direction\\nPredicted Probability\\nFIGURE 1.3. We fit a quadratic discriminant analysis model to the subset\\nof the Smarket data corresponding to the 2001–2004 time period, and predicted\\nthe probability of a stock market decrease using the 2005 data. On average, the\\npredicted probability of decrease is higher for the days in which the market does\\ndecrease. Based on these results, we are able to correctly predict the direction of\\nmovement in the market 60% of the time.\\nclustering problem. Unlike in the previous examples, here we are not trying\\nto predict an output variable.\\nWe devote Chapter 12 to a discussion of statistical learning methods\\nfor problems in which no natural output variable is available. We consider\\nthe NCI60 data set, which consists of 6,830 gene expression measurements\\nfor each of 64 cancer cell lines. Instead of predicting a particular output\\nvariable, we are interested in determining whether there are groups, or\\nclusters, among the cell lines based on their gene expression measurements.\\nThis is a difficult question to address, in part because there are thousands\\nof gene expression measurements per cell line, making it hard to visualize\\nthe data.\\nThe left-hand panel of Figure 1.4 addresses this problem by represent-\\ning each of the 64 cell lines using just two numbers, Z1 and Z2. These\\nare the first two principal components of the data, which summarize the\\n6,830 expression measurements for each cell line down to two numbers or\\ndimensions. While it is likely that this dimension reduction has resulted in\\nsome loss of information, it is now possible to visually examine the data\\nfor evidence of clustering. Deciding on the number of clusters is often a\\ndifficult problem. But the left-hand panel of Figure 1.4 suggests at least\\nfour groups of cell lines, which we have represented using separate colors.\\nIn this particular data set, it turns out that the cell lines correspond\\nto 14 different types of cancer. (However, this information was not used\\nto create the left-hand panel of Figure 1.4.) The right-hand panel of Fig-\\nure 1.4 is identical to the left-hand panel, except that the 14 cancer types\\nare shown using distinct colored symbols. There is clear evidence that cell\\nlines with the same cancer type tend to be located near each other in this\\ntwo-dimensional representation. In addition, even though the cancer infor-\\nmation was not used to produce the left-hand panel, the clustering obtained\\ndoes bear some resemblance to some of the actual cancer types observed\\nin the right-hand panel. This provides some independent verification of the\\naccuracy of our clustering analysis. 1. Introduction\\n5\\n−40\\n−20\\n0\\n20\\n40\\n60\\n−60\\n−40\\n−20\\n0\\n20\\n−40\\n−20\\n0\\n20\\n40\\n60\\n−60\\n−40\\n−20\\n0\\n20\\nZ1\\nZ1\\nZ2\\nZ2\\nFIGURE 1.4.\\nLeft: Representation of the NCI60 gene expression data set in\\na two-dimensional space, Z1 and Z2. Each point corresponds to one of the 64\\ncell lines. There appear to be four groups of cell lines, which we have represented\\nusing different colors. Right: Same as left panel except that we have represented\\neach of the 14 different types of cancer using a different colored symbol. Cell lines\\ncorresponding to the same cancer type tend to be nearby in the two-dimensional\\nspace.\\nA Brief History of Statistical Learning\\nThough the term statistical learning is fairly new, many of the concepts that\\nunderlie the field were developed long ago. At the beginning of the nine-\\nteenth century, the method of least squares was developed, implementing\\nthe earliest form of what is now known as linear regression. The approach\\nwas first successfully applied to problems in astronomy. Linear regression\\nis used for predicting quantitative values, such as an individual’s salary.\\nIn order to predict qualitative values, such as whether a patient survives\\nor dies, or whether the stock market increases or decreases, linear discrim-\\ninant analysis was proposed in 1936. In the 1940s, various authors put\\nforth an alternative approach, logistic regression. In the early 1970s, the\\nterm generalized linear model was developed to describe an entire class of\\nstatistical learning methods that include both linear and logistic regression\\nas special cases.\\nBy the end of the 1970s, many more techniques for learning from data\\nwere available. However, they were almost exclusively linear methods be-\\ncause fitting non-linear relationships was computationally difficult at the\\ntime. By the 1980s, computing technology had finally improved sufficiently\\nthat non-linear methods were no longer computationally prohibitive. In\\nthe mid 1980s, classification and regression trees were developed, followed\\nshortly by generalized additive models. Neural networks gained popularity\\nin the 1980s, and support vector machines arose in the 1990s.\\nSince that time, statistical learning has emerged as a new subfield in\\nstatistics, focused on supervised and unsupervised modeling and prediction.\\nIn recent years, progress in statistical learning has been marked by the\\nincreasing availability of powerful and relatively user-friendly software, such\\nas the popular and freely available Python system. This has the potential to\\ncontinue the transformation of the field from a set of techniques used and 6\\n1. Introduction\\ndeveloped by statisticians and computer scientists to an essential toolkit\\nfor a much broader community.\\nThis Book\\nThe Elements of Statistical Learning (ESL) by Hastie, Tibshirani, and\\nFriedman was first published in 2001. Since that time, it has become an\\nimportant reference on the fundamentals of statistical machine learning.\\nIts success derives from its comprehensive and detailed treatment of many\\nimportant topics in statistical learning, as well as the fact that (relative to\\nmany upper-level statistics textbooks) it is accessible to a wide audience.\\nHowever, the greatest factor behind the success of ESL has been its topical\\nnature. At the time of its publication, interest in the field of statistical\\nlearning was starting to explode. ESL provided one of the first accessible\\nand comprehensive introductions to the topic.\\nSince ESL was first published, the field of statistical learning has con-\\ntinued to flourish. The field’s expansion has taken two forms. The most\\nobvious growth has involved the development of new and improved statis-\\ntical learning approaches aimed at answering a range of scientific questions\\nacross a number of fields. However, the field of statistical learning has\\nalso expanded its audience. In the 1990s, increases in computational power\\ngenerated a surge of interest in the field from non-statisticians who were\\neager to use cutting-edge statistical tools to analyze their data. Unfortu-\\nnately, the highly technical nature of these approaches meant that the user\\ncommunity remained primarily restricted to experts in statistics, computer\\nscience, and related fields with the training (and time) to understand and\\nimplement them.\\nIn recent years, new and improved software packages have significantly\\neased the implementation burden for many statistical learning methods.\\nAt the same time, there has been growing recognition across a number of\\nfields, from business to health care to genetics to the social sciences and\\nbeyond, that statistical learning is a powerful tool with important practical\\napplications. As a result, the field has moved from one of primarily academic\\ninterest to a mainstream discipline, with an enormous potential audience.\\nThis trend will surely continue with the increasing availability of enormous\\nquantities of data and the software to analyze it.\\nThe purpose of An Introduction to Statistical Learning (ISL) is to facili-\\ntate the transition of statistical learning from an academic to a mainstream\\nfield. ISL is not intended to replace ESL, which is a far more comprehen-\\nsive text both in terms of the number of approaches considered and the\\ndepth to which they are explored. We consider ESL to be an important\\ncompanion for professionals (with graduate degrees in statistics, machine\\nlearning, or related fields) who need to understand the technical details\\nbehind statistical learning approaches. However, the community of users of\\nstatistical learning techniques has expanded to include individuals with a\\nwider range of interests and backgrounds. Therefore, there is a place for a\\nless technical and more accessible version of ESL. 1. Introduction\\n7\\nIn teaching these topics over the years, we have discovered that they are\\nof interest to master’s and PhD students in fields as disparate as business\\nadministration, biology, and computer science, as well as to quantitatively-\\noriented upper-division undergraduates. It is important for this diverse\\ngroup to be able to understand the models, intuitions, and strengths and\\nweaknesses of the various approaches. But for this audience, many of the\\ntechnical details behind statistical learning methods, such as optimiza-\\ntion algorithms and theoretical properties, are not of primary interest.\\nWe believe that these students do not need a deep understanding of these\\naspects in order to become informed users of the various methodologies, and\\nin order to contribute to their chosen fields through the use of statistical\\nlearning tools.\\nISL is based on the following four premises.\\n1. Many statistical learning methods are relevant and useful in a wide\\nrange of academic and non-academic disciplines, beyond just the sta-\\ntistical sciences. We believe that many contemporary statistical learn-\\ning procedures should, and will, become as widely available and used\\nas is currently the case for classical methods such as linear regres-\\nsion. As a result, rather than attempting to consider every possible\\napproach (an impossible task), we have concentrated on presenting\\nthe methods that we believe are most widely applicable.\\n2. Statistical learning should not be viewed as a series of black boxes. No\\nsingle approach will perform well in all possible applications. With-\\nout understanding all of the cogs inside the box, or the interaction\\nbetween those cogs, it is impossible to select the best box. Hence, we\\nhave attempted to carefully describe the model, intuition, assump-\\ntions, and trade-offs behind each of the methods that we consider.\\n3. While it is important to know what job is performed by each cog, it is\\nnot necessary to have the skills to construct the machine inside the\\nbox! Thus, we have minimized discussion of technical details related\\nto fitting procedures and theoretical properties. We assume that the\\nreader is comfortable with basic mathematical concepts, but we do\\nnot assume a graduate degree in the mathematical sciences. For in-\\nstance, we have almost completely avoided the use of matrix algebra,\\nand it is possible to understand the entire book without a detailed\\nknowledge of matrices and vectors.\\n4. We presume that the reader is interested in applying statistical learn-\\ning methods to real-world problems. In order to facilitate this, as well\\nas to motivate the techniques discussed, we have devoted a section\\nwithin each chapter to computer labs. In each lab, we walk the reader\\nthrough a realistic application of the methods considered in that chap-\\nter. When we have taught this material in our courses, we have al-\\nlocated roughly one-third of classroom time to working through the\\nlabs, and we have found them to be extremely useful. Many of the\\nless computationally-oriented students who were initially intimidated\\nby the labs got the hang of things over the course of the quarter or\\nsemester. This book originally appeared (2013, second edition 2021) 8\\n1. Introduction\\nwith computer labs written in the R language. Since then, there has\\nbeen increasing demand for Python implementations of the impor-\\ntant techniques in statistical learning. Consequently, this version has\\nlabs in Python. There are a rapidly growing number of Python pack-\\nages available, and by examination of the imports at the beginning of\\neach lab, readers will see that we have carefully selected and used the\\nmost appropriate. We have also supplied some additional code and\\nfunctionality in our package ISLP. However, the labs in ISL are self-\\ncontained, and can be skipped if the reader wishes to use a different\\nsoftware package or does not wish to apply the methods discussed to\\nreal-world problems.\\nWho Should Read This Book?\\nThis book is intended for anyone who is interested in using modern statis-\\ntical methods for modeling and prediction from data. This group includes\\nscientists, engineers, data analysts, data scientists, and quants, but also less\\ntechnical individuals with degrees in non-quantitative fields such as the so-\\ncial sciences or business. We expect that the reader will have had at least\\none elementary course in statistics. Background in linear regression is also\\nuseful, though not required, since we review the key concepts behind linear\\nregression in Chapter 3. The mathematical level of this book is modest,\\nand a detailed knowledge of matrix operations is not required. This book\\nprovides an introduction to Python. Previous exposure to a programming\\nlanguage, such as MATLAB or R, is useful but not required.\\nThe first edition of this textbook has been used to teach master’s and\\nPhD students in business, economics, computer science, biology, earth sci-\\nences, psychology, and many other areas of the physical and social sciences.\\nIt has also been used to teach advanced undergraduates who have already\\ntaken a course on linear regression. In the context of a more mathemat-\\nically rigorous course in which ESL serves as the primary textbook, ISL\\ncould be used as a supplementary text for teaching computational aspects\\nof the various approaches.\\nNotation and Simple Matrix Algebra\\nChoosing notation for a textbook is always a difficult task. For the most\\npart we adopt the same notational conventions as ESL.\\nWe will use n to represent the number of distinct data points, or observa-\\ntions, in our sample. We will let p denote the number of variables that are\\navailable for use in making predictions. For example, the Wage data set con-\\nsists of 11 variables for 3,000 people, so we have n = 3,000 observations and\\np = 11 variables (such as year, age, race, and more). Note that throughout\\nthis book, we indicate variable names using colored font: Variable Name.\\nIn some examples, p might be quite large, such as on the order of thou-\\nsands or even millions; this situation arises quite often, for example, in the\\nanalysis of modern biological data or web-based advertising data. 1. Introduction\\n9\\nIn general, we will let xij represent the value of the jth variable for the\\nith observation, where i = 1, 2, . . . , n and j = 1, 2, . . . , p. Throughout this\\nbook, i will be used to index the samples or observations (from 1 to n) and\\nj will be used to index the variables (from 1 to p). We let X denote an\\nn × p matrix whose (i, j)th element is xij. That is,\\nX =\\n\\uf8eb\\n\\uf8ec\\n\\uf8ec\\n\\uf8ec\\n\\uf8ed\\nx11\\nx12\\n. . .\\nx1p\\nx21\\nx22\\n. . .\\nx2p\\n...\\n...\\n...\\n...\\nxn1\\nxn2\\n. . .\\nxnp\\n\\uf8f6\\n\\uf8f7\\n\\uf8f7\\n\\uf8f7\\n\\uf8f8.\\nFor readers who are unfamiliar with matrices, it is useful to visualize X as\\na spreadsheet of numbers with n rows and p columns.\\nAt times we will be interested in the rows of X, which we write as\\nx1, x2, . . . , xn. Here xi is a vector of length p, containing the p variable\\nmeasurements for the ith observation. That is,\\nxi =\\n\\uf8eb\\n\\uf8ec\\n\\uf8ec\\n\\uf8ec\\n\\uf8ed\\nxi1\\nxi2\\n...\\nxip\\n\\uf8f6\\n\\uf8f7\\n\\uf8f7\\n\\uf8f7\\n\\uf8f8.\\n(1.1)\\n(Vectors are by default represented as columns.) For example, for the Wage\\ndata, xi is a vector of length 11, consisting of year, age, race, and other\\nvalues for the ith individual. At other times we will instead be interested\\nin the columns of X, which we write as x1, x2, . . . , xp. Each is a vector of\\nlength n. That is,\\nxj =\\n\\uf8eb\\n\\uf8ec\\n\\uf8ec\\n\\uf8ec\\n\\uf8ed\\nx1j\\nx2j\\n...\\nxnj\\n\\uf8f6\\n\\uf8f7\\n\\uf8f7\\n\\uf8f7\\n\\uf8f8.\\nFor example, for the Wage data, x1 contains the n = 3,000 values for year.\\nUsing this notation, the matrix X can be written as\\nX =\\n\\'x1\\nx2\\n· · ·\\nxp\\n(\\n,\\nor\\nX =\\n\\uf8eb\\n\\uf8ec\\n\\uf8ec\\n\\uf8ec\\n\\uf8ed\\nxT\\n1\\nxT\\n2\\n...\\nxT\\nn\\n\\uf8f6\\n\\uf8f7\\n\\uf8f7\\n\\uf8f7\\n\\uf8f8.\\nThe T notation denotes the transpose of a matrix or vector. So, for example,\\nXT =\\n\\uf8eb\\n\\uf8ec\\n\\uf8ec\\n\\uf8ec\\n\\uf8ed\\nx11\\nx21\\n. . .\\nxn1\\nx12\\nx22\\n. . .\\nxn2\\n...\\n...\\n...\\nx1p\\nx2p\\n. . .\\nxnp\\n\\uf8f6\\n\\uf8f7\\n\\uf8f7\\n\\uf8f7\\n\\uf8f8, 10\\n1. Introduction\\nwhile\\nxT\\ni =\\n\\'xi1\\nxi2\\n· · ·\\nxip\\n(\\n.\\nWe use yi to denote the ith observation of the variable on which we\\nwish to make predictions, such as wage. Hence, we write the set of all n\\nobservations in vector form as\\ny =\\n\\uf8eb\\n\\uf8ec\\n\\uf8ec\\n\\uf8ec\\n\\uf8ed\\ny1\\ny2\\n...\\nyn\\n\\uf8f6\\n\\uf8f7\\n\\uf8f7\\n\\uf8f7\\n\\uf8f8.\\nThen our observed data consists of {(x1, y1), (x2, y2), . . . , (xn, yn)}, where\\neach xi is a vector of length p. (If p = 1, then xi is simply a scalar.)\\nIn this text, a vector of length n will always be denoted in lower case\\nbold; e.g.\\na =\\n\\uf8eb\\n\\uf8ec\\n\\uf8ec\\n\\uf8ec\\n\\uf8ed\\na1\\na2\\n...\\nan\\n\\uf8f6\\n\\uf8f7\\n\\uf8f7\\n\\uf8f7\\n\\uf8f8.\\nHowever, vectors that are not of length n (such as feature vectors of length\\np, as in (1.1)) will be denoted in lower case normal font, e.g. a. Scalars will\\nalso be denoted in lower case normal font, e.g. a. In the rare cases in which\\nthese two uses for lower case normal font lead to ambiguity, we will clarify\\nwhich use is intended. Matrices will be denoted using bold capitals, such\\nas A. Random variables will be denoted using capital normal font, e.g. A,\\nregardless of their dimensions.\\nOccasionally we will want to indicate the dimension of a particular object.\\nTo indicate that an object is a scalar, we will use the notation a ∈R. To\\nindicate that it is a vector of length k, we will use a ∈Rk (or a ∈Rn if\\nit is of length n). We will indicate that an object is an r × s matrix using\\nA ∈Rr×s.\\nWe have avoided using matrix algebra whenever possible. However, in\\na few instances it becomes too cumbersome to avoid it entirely. In these\\nrare instances it is important to understand the concept of multiplying\\ntwo matrices. Suppose that A ∈Rr×d and B ∈Rd×s. Then the product\\nof A and B is denoted AB. The (i, j)th element of AB is computed by\\nmultiplying each element of the ith row of A by the corresponding element\\nof the jth column of B. That is, (AB)ij = )d\\nk=1 aikbkj. As an example,\\nconsider\\nA =\\n*1\\n2\\n3\\n4\\n+\\nand\\nB =\\n*5\\n6\\n7\\n8\\n+\\n.\\nThen\\nAB =\\n*1\\n2\\n3\\n4\\n+ *5\\n6\\n7\\n8\\n+\\n=\\n*1 × 5 + 2 × 7\\n1 × 6 + 2 × 8\\n3 × 5 + 4 × 7\\n3 × 6 + 4 × 8\\n+\\n=\\n*19\\n22\\n43\\n50\\n+\\n.\\nNote that this operation produces an r × s matrix. It is only possible to\\ncompute AB if the number of columns of A is the same as the number of\\nrows of B. 1. Introduction\\n11\\nOrganization of This Book\\nChapter 2 introduces the basic terminology and concepts behind statisti-\\ncal learning. This chapter also presents the K-nearest neighbor classifier, a\\nvery simple method that works surprisingly well on many problems. Chap-\\nters 3 and 4 cover classical linear methods for regression and classification.\\nIn particular, Chapter 3 reviews linear regression, the fundamental start-\\ning point for all regression methods. In Chapter 4 we discuss two of the\\nmost important classical classification methods, logistic regression and lin-\\near discriminant analysis.\\nA central problem in all statistical learning situations involves choosing\\nthe best method for a given application. Hence, in Chapter 5 we intro-\\nduce cross-validation and the bootstrap, which can be used to estimate the\\naccuracy of a number of different methods in order to choose the best one.\\nMuch of the recent research in statistical learning has concentrated on\\nnon-linear methods. However, linear methods often have advantages over\\ntheir non-linear competitors in terms of interpretability and sometimes also\\naccuracy. Hence, in Chapter 6 we consider a host of linear methods, both\\nclassical and more modern, which offer potential improvements over stan-\\ndard linear regression. These include stepwise selection, ridge regression,\\nprincipal components regression, and the lasso.\\nThe remaining chapters move into the world of non-linear statistical\\nlearning. We first introduce in Chapter 7 a number of non-linear meth-\\nods that work well for problems with a single input variable. We then\\nshow how these methods can be used to fit non-linear additive models for\\nwhich there is more than one input. In Chapter 8, we investigate tree-based\\nmethods, including bagging, boosting, and random forests. Support vector\\nmachines, a set of approaches for performing both linear and non-linear\\nclassification, are discussed in Chapter 9. We cover deep learning, an ap-\\nproach for non-linear regression and classification that has received a lot\\nof attention in recent years, in Chapter 10. Chapter 11 explores survival\\nanalysis, a regression approach that is specialized to the setting in which\\nthe output variable is censored, i.e. not fully observed.\\nIn Chapter 12, we consider the unsupervised setting in which we have\\ninput variables but no output variable. In particular, we present princi-\\npal components analysis, K-means clustering, and hierarchical clustering.\\nFinally, in Chapter 13 we cover the very important topic of multiple hy-\\npothesis testing.\\nAt the end of each chapter, we present one or more Python lab sections\\nin which we systematically work through applications of the various meth-\\nods discussed in that chapter. These labs demonstrate the strengths and\\nweaknesses of the various approaches, and also provide a useful reference\\nfor the syntax required to implement the various methods. The reader may\\nchoose to work through the labs at their own pace, or the labs may be the\\nfocus of group sessions as part of a classroom environment. Within each\\nPython lab, we present the results that we obtained when we performed the\\nlab at the time of writing this book. However, new versions of Python are\\ncontinuously released, and over time, the packages called in the labs will\\nbe updated. Therefore, in the future, it is possible that the results shown in 12\\n1. Introduction\\nName\\nDescription\\nAuto\\nGas mileage, horsepower, and other information for cars.\\nBikeshare\\nHourly usage of a bike sharing program in Washington, DC.\\nBoston\\nHousing values and other information about Boston census tracts.\\nBrainCancer\\nSurvival times for patients diagnosed with brain cancer.\\nCaravan\\nInformation about individuals offered caravan insurance.\\nCarseats\\nInformation about car seat sales in 400 stores.\\nCollege\\nDemographic characteristics, tuition, and more for USA colleges.\\nCredit\\nInformation about credit card debt for 400 customers.\\nDefault\\nCustomer default records for a credit card company.\\nFund\\nReturns of 2,000 hedge fund managers over 50 months.\\nHitters\\nRecords and salaries for baseball players.\\nKhan\\nGene expression measurements for four cancer types.\\nNCI60\\nGene expression measurements for 64 cancer cell lines.\\nNYSE\\nReturns, volatility, and volume for the New York Stock Exchange.\\nOJ\\nSales information for Citrus Hill and Minute Maid orange juice.\\nPortfolio\\nPast values of financial assets, for use in portfolio allocation.\\nPublication\\nTime to publication for 244 clinical trials.\\nSmarket\\nDaily percentage returns for S&P 500 over a 5-year period.\\nUSArrests\\nCrime statistics per 100,000 residents in 50 states of USA.\\nWage\\nIncome survey data for men in central Atlantic region of USA.\\nWeekly\\n1,089 weekly stock market returns for 21 years.\\nTABLE 1.1. A list of data sets needed to perform the labs and exercises in this\\ntextbook. All data sets are available in the ISLP package, with the exception of\\nUSArrests, which is part of the base R distribution, but accessible from Python.\\nthe lab sections may no longer correspond precisely to the results obtained\\nby the reader who performs the labs. As necessary, we will post updates to\\nthe labs on the book website.\\nWe use the\\nsymbol to denote sections or exercises that contain more\\nchallenging concepts. These can be easily skipped by readers who do not\\nwish to delve as deeply into the material, or who lack the mathematical\\nbackground.\\nData Sets Used in Labs and Exercises\\nIn this textbook, we illustrate statistical learning methods using applica-\\ntions from marketing, finance, biology, and other areas. The ISLP package\\ncontains a number of data sets that are required in order to perform the\\nlabs and exercises associated with this book. One other data set is part of\\nthe base R distribution (the USArrests data), and we show how to access it\\nfrom Python in Section 12.5.1. Table 1.1 contains a summary of the data\\nsets required to perform the labs and exercises. A couple of these data sets\\nare also available as text files on the book website, for use in Chapter 2. 1. Introduction\\n13\\nBook Website\\nThe website for this book is located at\\nwww.statlearning.com\\nIt contains a number of resources, including the Python package associated\\nwith this book, and some additional data sets.\\nAcknowledgements\\nA few of the plots in this book were taken from ESL: Figures 6.7, 8.3,\\nand 12.14. All other plots were produced for the R version of ISL, except\\nfor Figure 13.10 which differs because of the Python software supporting\\nthe plot. 2\\nStatistical Learning\\n2.1\\nWhat Is Statistical Learning?\\nIn order to motivate our study of statistical learning, we begin with a simple\\nexample. Suppose that we are statistical consultants hired by a client to\\ninvestigate the association between advertising and sales of a particular\\nproduct. The Advertising data set consists of the sales of that product\\nin 200 different markets, along with advertising budgets for the product in\\neach of those markets for three different media: TV, radio, and newspaper.\\nThe data are displayed in Figure 2.1. It is not possible for our client to\\ndirectly increase sales of the product. On the other hand, they can control\\nthe advertising expenditure in each of the three media. Therefore, if we\\ndetermine that there is an association between advertising and sales, then\\nwe can instruct our client to adjust advertising budgets, thereby indirectly\\nincreasing sales. In other words, our goal is to develop an accurate model\\nthat can be used to predict sales on the basis of the three media budgets.\\nIn this setting, the advertising budgets are input variables while sales input\\nvariable\\nis an output variable. The input variables are typically denoted using the\\noutput\\nvariable\\nsymbol X, with a subscript to distinguish them. So X1 might be the TV\\nbudget, X2 the radio budget, and X3 the newspaper budget. The inputs\\ngo by different names, such as predictors, independent variables, features, predictor\\nindependent\\nvariable\\nfeature\\nor sometimes just variables. The output variable—in this case, sales—is\\nvariable\\noften called the response or dependent variable, and is typically denoted\\nresponse\\ndependent\\nvariable\\nusing the symbol Y . Throughout this book, we will use all of these terms\\ninterchangeably.\\nMore generally, suppose that we observe a quantitative response Y and p\\ndifferent predictors, X1, X2, . . . , Xp. We assume that there is some\\nrelationship between Y and X = (X1, X2, . . . , Xp), which can be written\\nin the very general form\\nY = f(X) + ϵ.\\n(2.1)\\n© Springer Nature Switzerland AG 2023 \\nG. James et al., An Introduction to Statistical Learning, Springer Texts in Statistics, \\nhttps://doi.org/10.1007/978-3-031-38747-0_2 \\n15 16\\n2. Statistical Learning\\n0\\n50\\n100\\n200\\n300\\n5\\n10\\n15\\n20\\n25\\nTV\\nSales\\n0\\n10\\n20\\n30\\n40\\n50\\n5\\n10\\n15\\n20\\n25\\nRadio\\nSales\\n0\\n20\\n40\\n60\\n80\\n100\\n5\\n10\\n15\\n20\\n25\\nNewspaper\\nSales\\nFIGURE 2.1. The Advertising data set. The plot displays sales, in thousands\\nof units, as a function of TV, radio, and newspaper budgets, in thousands of\\ndollars, for 200 different markets. In each plot we show the simple least squares\\nfit of sales to that variable, as described in Chapter 3. In other words, each blue\\nline represents a simple model that can be used to predict sales using TV, radio,\\nand newspaper, respectively.\\nHere f is some fixed but unknown function of X1, . . . , Xp, and ϵ is a random\\nerror term, which is independent of X and has mean zero. In this formula- error term\\ntion, f represents the systematic information that X provides about Y .\\nsystematic\\n10\\n12\\n14\\n16\\n18\\n20\\n22\\n20\\n30\\n40\\n50\\n60\\n70\\n80\\nYears of Education\\nIncome\\n10\\n12\\n14\\n16\\n18\\n20\\n22\\n20\\n30\\n40\\n50\\n60\\n70\\n80\\nYears of Education\\nIncome\\nFIGURE 2.2. The Income data set. Left: The red dots are the observed values\\nof income (in thousands of dollars) and years of education for 30 individuals.\\nRight: The blue curve represents the true underlying relationship between income\\nand years of education, which is generally unknown (but is known in this case\\nbecause the data were simulated). The black lines represent the error associated\\nwith each observation. Note that some errors are positive (if an observation lies\\nabove the blue curve) and some are negative (if an observation lies below the\\ncurve). Overall, these errors have approximately mean zero.\\nAs another example, consider the left-hand panel of Figure 2.2, a plot of\\nincome versus years of education for 30 individuals in the Income data set.\\nThe plot suggests that one might be able to predict income using years of\\neducation. However, the function f that connects the input variable to the 2.1 What Is Statistical Learning?\\n17\\noutput variable is in general unknown. In this situation one must estimate\\nf based on the observed points. Since Income is a simulated data set, f is\\nknown and is shown by the blue curve in the right-hand panel of Figure 2.2.\\nThe vertical lines represent the error terms ϵ. We note that some of the\\n30 observations lie above the blue curve and some lie below it; overall, the\\nerrors have approximately mean zero.\\nIn general, the function f may involve more than one input variable.\\nIn Figure 2.3 we plot income as a function of years of education and\\nseniority. Here f is a two-dimensional surface that must be estimated\\nbased on the observed data.\\nIn essence, statistical learning refers to a set of approaches for estimating\\nf. In this chapter we outline some of the key theoretical concepts that arise\\nin estimating f, as well as tools for evaluating the estimates obtained.\\n2.1.1\\nWhy Estimate f?\\nThere are two main reasons that we may wish to estimate f: prediction\\nand inference. We discuss each in turn.\\nPrediction\\nIn many situations, a set of inputs X are readily available, but the output\\nY cannot be easily obtained. In this setting, since the error term averages\\nto zero, we can predict Y using\\nˆY = ˆf(X),\\n(2.2)\\nwhere ˆf represents our estimate for f, and ˆY represents the resulting pre-\\ndiction for Y . In this setting, ˆf is often treated as a black box, in the sense\\nthat one is not typically concerned with the exact form of ˆf, provided that\\nit yields accurate predictions for Y .\\nAs an example, suppose that X1, . . . , Xp are characteristics of a patient’s\\nblood sample that can be easily measured in a lab, and Y is a variable\\nencoding the patient’s risk for a severe adverse reaction to a particular\\ndrug. It is natural to seek to predict Y using X, since we can then avoid\\ngiving the drug in question to patients who are at high risk of an adverse\\nreaction—that is, patients for whom the estimate of Y is high.\\nThe accuracy of ˆY as a prediction for Y depends on two quantities,\\nwhich we will call the reducible error and the irreducible error. In general, reducible\\nerror\\nirreducible\\nerror\\nˆf will not be a perfect estimate for f, and this inaccuracy will introduce\\nsome error. This error is reducible because we can potentially improve the\\naccuracy of ˆf by using the most appropriate statistical learning technique to\\nestimate f. However, even if it were possible to form a perfect estimate for\\nf, so that our estimated response took the form ˆY = f(X), our prediction\\nwould still have some error in it! This is because Y is also a function of\\nϵ, which, by definition, cannot be predicted using X. Therefore, variability\\nassociated with ϵ also affects the accuracy of our predictions. This is known\\nas the irreducible error, because no matter how well we estimate f, we\\ncannot reduce the error introduced by ϵ.\\nWhy is the irreducible error larger than zero? The quantity ϵ may con-\\ntain unmeasured variables that are useful in predicting Y : since we don’t 18\\n2. Statistical Learning\\nYears of Education\\nSeniority\\nIncome\\nFIGURE 2.3. The plot displays income as a function of years of education\\nand seniority in the Income data set. The blue surface represents the true\\nunderlying relationship between income and years of education and seniority,\\nwhich is known since the data are simulated. The red dots indicate the observed\\nvalues of these quantities for 30 individuals.\\nmeasure them, f cannot use them for its prediction. The quantity ϵ may\\nalso contain unmeasurable variation. For example, the risk of an adverse\\nreaction might vary for a given patient on a given day, depending on\\nmanufacturing variation in the drug itself or the patient’s general feeling\\nof well-being on that day.\\nConsider a given estimate ˆf and a set of predictors X, which yields the\\nprediction ˆY = ˆf(X). Assume for a moment that both ˆf and X are fixed,\\nso that the only variability comes from ϵ. Then, it is easy to show that\\nE(Y −ˆY )2\\n=\\nE[f(X) + ϵ −ˆf(X)]2\\n=\\n[f(X) −ˆf(X)]2\\n,\\n-.\\n/\\nReducible\\n+\\nVar(ϵ)\\n, -. /\\nIrreducible\\n,\\n(2.3)\\nwhere E(Y −ˆY )2 represents the average, or expected value, of the squared expected\\nvalue\\ndifference between the predicted and actual value of Y , and Var(ϵ) repre-\\nsents the variance associated with the error term ϵ.\\nvariance\\nThe focus of this book is on techniques for estimating f with the aim of\\nminimizing the reducible error. It is important to keep in mind that the\\nirreducible error will always provide an upper bound on the accuracy of\\nour prediction for Y . This bound is almost always unknown in practice.\\nInference\\nWe are often interested in understanding the association between Y and\\nX1, . . . , Xp. In this situation we wish to estimate f, but our goal is not\\nnecessarily to make predictions for Y . Now ˆf cannot be treated as a black\\nbox, because we need to know its exact form. In this setting, one may be\\ninterested in answering the following questions: 2.1 What Is Statistical Learning?\\n19\\n• Which predictors are associated with the response? It is often the case\\nthat only a small fraction of the available predictors are substantially\\nassociated with Y . Identifying the few important predictors among a\\nlarge set of possible variables can be extremely useful, depending on\\nthe application.\\n• What is the relationship between the response and each predictor?\\nSome predictors may have a positive relationship with Y , in the sense\\nthat larger values of the predictor are associated with larger values of\\nY . Other predictors may have the opposite relationship. Depending\\non the complexity of f, the relationship between the response and a\\ngiven predictor may also depend on the values of the other predictors.\\n• Can the relationship between Y and each predictor be adequately sum-\\nmarized using a linear equation, or is the relationship more compli-\\ncated? Historically, most methods for estimating f have taken a linear\\nform. In some situations, such an assumption is reasonable or even de-\\nsirable. But often the true relationship is more complicated, in which\\ncase a linear model may not provide an accurate representation of\\nthe relationship between the input and output variables.\\nIn this book, we will see a number of examples that fall into the prediction\\nsetting, the inference setting, or a combination of the two.\\nFor instance, consider a company that is interested in conducting a\\ndirect-marketing campaign. The goal is to identify individuals who are\\nlikely to respond positively to a mailing, based on observations of demo-\\ngraphic variables measured on each individual. In this case, the demo-\\ngraphic variables serve as predictors, and response to the marketing cam-\\npaign (either positive or negative) serves as the outcome. The company is\\nnot interested in obtaining a deep understanding of the relationships be-\\ntween each individual predictor and the response; instead, the company\\nsimply wants to accurately predict the response using the predictors. This\\nis an example of modeling for prediction.\\nIn contrast, consider the Advertising data illustrated in Figure 2.1. One\\nmay be interested in answering questions such as:\\n– Which media are associated with sales?\\n– Which media generate the biggest boost in sales? or\\n– How large of an increase in sales is associated with a given increase\\nin TV advertising?\\nThis situation falls into the inference paradigm. Another example involves\\nmodeling the brand of a product that a customer might purchase based on\\nvariables such as price, store location, discount levels, competition price,\\nand so forth. In this situation one might really be most interested in the\\nassociation between each variable and the probability of purchase. For in-\\nstance, to what extent is the product’s price associated with sales? This is\\nan example of modeling for inference.\\nFinally, some modeling could be conducted both for prediction and in-\\nference. For example, in a real estate setting, one may seek to relate values 20\\n2. Statistical Learning\\nof homes to inputs such as crime rate, zoning, distance from a river, air\\nquality, schools, income level of community, size of houses, and so forth. In\\nthis case one might be interested in the association between each individ-\\nual input variable and housing price—for instance, how much extra will a\\nhouse be worth if it has a view of the river? This is an inference problem.\\nAlternatively, one may simply be interested in predicting the value of a\\nhome given its characteristics: is this house under- or over-valued? This is\\na prediction problem.\\nDepending on whether our ultimate goal is prediction, inference, or a\\ncombination of the two, different methods for estimating f may be ap-\\npropriate. For example, linear models allow for relatively simple and in- linear model\\nterpretable inference, but may not yield as accurate predictions as some\\nother approaches. In contrast, some of the highly non-linear approaches\\nthat we discuss in the later chapters of this book can potentially provide\\nquite accurate predictions for Y , but this comes at the expense of a less\\ninterpretable model for which inference is more challenging.\\n2.1.2\\nHow Do We Estimate f?\\nThroughout this book, we explore many linear and non-linear approaches\\nfor estimating f. However, these methods generally share certain charac-\\nteristics. We provide an overview of these shared characteristics in this\\nsection. We will always assume that we have observed a set of n different\\ndata points. For example in Figure 2.2 we observed n = 30 data points.\\nThese observations are called the training data because we will use these training\\ndata\\nobservations to train, or teach, our method how to estimate f. Let xij\\nrepresent the value of the jth predictor, or input, for observation i, where\\ni = 1, 2, . . . , n and j = 1, 2, . . . , p. Correspondingly, let yi represent the\\nresponse variable for the ith observation. Then our training data consist of\\n{(x1, y1), (x2, y2), . . . , (xn, yn)} where xi = (xi1, xi2, . . . , xip)T .\\nOur goal is to apply a statistical learning method to the training data\\nin order to estimate the unknown function f. In other words, we want to\\nfind a function ˆf such that Y ≈ˆf(X) for any observation (X, Y ). Broadly\\nspeaking, most statistical learning methods for this task can be character-\\nized as either parametric or non-parametric. We now briefly discuss these parametric\\nnon-\\nparametric\\ntwo types of approaches.\\nParametric Methods\\nParametric methods involve a two-step model-based approach.\\n1. First, we make an assumption about the functional form, or shape,\\nof f. For example, one very simple assumption is that f is linear in\\nX:\\nf(X) = β0 + β1X1 + β2X2 + · · · + βpXp.\\n(2.4)\\nThis is a linear model, which will be discussed extensively in Chap-\\nter 3. Once we have assumed that f is linear, the problem of estimat-\\ning f is greatly simplified. Instead of having to estimate an entirely\\narbitrary p-dimensional function f(X), one only needs to estimate\\nthe p + 1 coefficients β0, β1, . . . , βp. 2.1 What Is Statistical Learning?\\n21\\nYears of Education\\nSeniority\\nIncome\\nFIGURE 2.4.\\nA linear model fit by least squares to the Income data from\\nFigure 2.3. The observations are shown in red, and the yellow plane indicates the\\nleast squares fit to the data.\\n2. After a model has been selected, we need a procedure that uses the\\ntraining data to fit or train the model. In the case of the linear model fit\\ntrain\\n(2.4), we need to estimate the parameters β0, β1, . . . , βp. That is, we\\nwant to find values of these parameters such that\\nY ≈β0 + β1X1 + β2X2 + · · · + βpXp.\\nThe most common approach to fitting the model (2.4) is referred to\\nas (ordinary) least squares, which we discuss in Chapter 3. However, least squares\\nleast squares is one of many possible ways to fit the linear model. In\\nChapter 6, we discuss other approaches for estimating the parameters\\nin (2.4).\\nThe model-based approach just described is referred to as parametric;\\nit reduces the problem of estimating f down to one of estimating a set of\\nparameters. Assuming a parametric form for f simplifies the problem of\\nestimating f because it is generally much easier to estimate a set of pa-\\nrameters, such as β0, β1, . . . , βp in the linear model (2.4), than it is to fit\\nan entirely arbitrary function f. The potential disadvantage of a paramet-\\nric approach is that the model we choose will usually not match the true\\nunknown form of f. If the chosen model is too far from the true f, then\\nour estimate will be poor. We can try to address this problem by choos-\\ning flexible models that can fit many different possible functional forms flexible\\nfor f. But in general, fitting a more flexible model requires estimating a\\ngreater number of parameters. These more complex models can lead to a\\nphenomenon known as overfitting the data, which essentially means they overfitting\\nfollow the errors, or noise, too closely. These issues are discussed through- noise\\nout this book.\\nFigure 2.4 shows an example of the parametric approach applied to the\\nIncome data from Figure 2.3. We have fit a linear model of the form\\nincome ≈β0 + β1 × education + β2 × seniority. 22\\n2. Statistical Learning\\nYears of Education\\nSeniority\\nIncome\\nFIGURE 2.5. A smooth thin-plate spline fit to the Income data from Figure 2.3\\nis shown in yellow; the observations are displayed in red. Splines are discussed in\\nChapter 7.\\nSince we have assumed a linear relationship between the response and the\\ntwo predictors, the entire fitting problem reduces to estimating β0, β1, and\\nβ2, which we do using least squares linear regression. Comparing Figure 2.3\\nto Figure 2.4, we can see that the linear fit given in Figure 2.4 is not quite\\nright: the true f has some curvature that is not captured in the linear fit.\\nHowever, the linear fit still appears to do a reasonable job of capturing the\\npositive relationship between years of education and income, as well as the\\nslightly less positive relationship between seniority and income. It may be\\nthat with such a small number of observations, this is the best we can do.\\nNon-Parametric Methods\\nNon-parametric methods do not make explicit assumptions about the func-\\ntional form of f. Instead they seek an estimate of f that gets as close to the\\ndata points as possible without being too rough or wiggly. Such approaches\\ncan have a major advantage over parametric approaches: by avoiding the\\nassumption of a particular functional form for f, they have the potential\\nto accurately fit a wider range of possible shapes for f. Any parametric\\napproach brings with it the possibility that the functional form used to\\nestimate f is very different from the true f, in which case the resulting\\nmodel will not fit the data well. In contrast, non-parametric approaches\\ncompletely avoid this danger, since essentially no assumption about the\\nform of f is made. But non-parametric approaches do suffer from a major\\ndisadvantage: since they do not reduce the problem of estimating f to a\\nsmall number of parameters, a very large number of observations (far more\\nthan is typically needed for a parametric approach) is required in order to\\nobtain an accurate estimate for f.\\nAn example of a non-parametric approach to fitting the Income data is\\nshown in Figure 2.5. A thin-plate spline is used to estimate f. This ap- thin-plate\\nspline\\nproach does not impose any pre-specified model on f. It instead attempts 2.1 What Is Statistical Learning?\\n23\\nYears of Education\\nSeniority\\nIncome\\nFIGURE 2.6. A rough thin-plate spline fit to the Income data from Figure 2.3.\\nThis fit makes zero errors on the training data.\\nto produce an estimate for f that is as close as possible to the observed\\ndata, subject to the fit—that is, the yellow surface in Figure 2.5—being\\nsmooth. In this case, the non-parametric fit has produced a remarkably ac-\\ncurate estimate of the true f shown in Figure 2.3. In order to fit a thin-plate\\nspline, the data analyst must select a level of smoothness. Figure 2.6 shows\\nthe same thin-plate spline fit using a lower level of smoothness, allowing\\nfor a rougher fit. The resulting estimate fits the observed data perfectly!\\nHowever, the spline fit shown in Figure 2.6 is far more variable than the\\ntrue function f, from Figure 2.3. This is an example of overfitting the\\ndata, which we discussed previously. It is an undesirable situation because\\nthe fit obtained will not yield accurate estimates of the response on new\\nobservations that were not part of the original training data set. We dis-\\ncuss methods for choosing the correct amount of smoothness in Chapter 5.\\nSplines are discussed in Chapter 7.\\nAs we have seen, there are advantages and disadvantages to parametric\\nand non-parametric methods for statistical learning. We explore both types\\nof methods throughout this book.\\n2.1.3\\nThe Trade-Off Between Prediction Accuracy and Model\\nInterpretability\\nOf the many methods that we examine in this book, some are less flexible,\\nor more restrictive, in the sense that they can produce just a relatively\\nsmall range of shapes to estimate f. For example, linear regression is a\\nrelatively inflexible approach, because it can only generate linear functions\\nsuch as the lines shown in Figure 2.1 or the plane shown in Figure 2.4.\\nOther methods, such as the thin plate splines shown in Figures 2.5 and 2.6,\\nare considerably more flexible because they can generate a much wider\\nrange of possible shapes to estimate f. 24\\n2. Statistical Learning\\nFlexibility\\nInterpretability\\nLow\\nHigh\\nLow\\nHigh\\nSubset Selection\\nLasso\\nLeast Squares\\nGeneralized Additive Models\\nTrees\\nBagging, Boosting\\nSupport Vector Machines\\nDeep Learning\\nFIGURE 2.7.\\nA representation of the tradeoff between flexibility and inter-\\npretability, using different statistical learning methods. In general, as the flexibility\\nof a method increases, its interpretability decreases.\\nOne might reasonably ask the following question: why would we ever\\nchoose to use a more restrictive method instead of a very flexible approach?\\nThere are several reasons that we might prefer a more restrictive model.\\nIf we are mainly interested in inference, then restrictive models are much\\nmore interpretable. For instance, when inference is the goal, the linear\\nmodel may be a good choice since it will be quite easy to understand\\nthe relationship between Y and X1, X2, . . . , Xp. In contrast, very flexible\\napproaches, such as the splines discussed in Chapter 7 and displayed in\\nFigures 2.5 and 2.6, and the boosting methods discussed in Chapter 8, can\\nlead to such complicated estimates of f that it is difficult to understand\\nhow any individual predictor is associated with the response.\\nFigure 2.7 provides an illustration of the trade-off between flexibility and\\ninterpretability for some of the methods that we cover in this book. Least\\nsquares linear regression, discussed in Chapter 3, is relatively inflexible but\\nis quite interpretable. The lasso, discussed in Chapter 6, relies upon the lasso\\nlinear model (2.4) but uses an alternative fitting procedure for estimating\\nthe coefficients β0, β1, . . . , βp. The new procedure is more restrictive in es-\\ntimating the coefficients, and sets a number of them to exactly zero. Hence\\nin this sense the lasso is a less flexible approach than linear regression.\\nIt is also more interpretable than linear regression, because in the final\\nmodel the response variable will only be related to a small subset of the\\npredictors—namely, those with nonzero coefficient estimates. Generalized\\nadditive models (GAMs), discussed in Chapter 7, instead extend the lin- generalized\\nadditive\\nmodel\\near model (2.4) to allow for certain non-linear relationships. Consequently,\\nGAMs are more flexible than linear regression. They are also somewhat\\nless interpretable than linear regression, because the relationship between\\neach predictor and the response is now modeled using a curve. Finally,\\nfully non-linear methods such as bagging, boosting, support vector machines bagging\\nboosting\\nwith non-linear kernels, and neural networks (deep learning), discussed in\\nsupport\\nvector\\nmachine\\nChapters 8, 9, and 10, are highly flexible approaches that are harder to\\ninterpret. 2.1 What Is Statistical Learning?\\n25\\nWe have established that when inference is the goal, there are clear ad-\\nvantages to using simple and relatively inflexible statistical learning meth-\\nods. In some settings, however, we are only interested in prediction, and\\nthe interpretability of the predictive model is simply not of interest. For\\ninstance, if we seek to develop an algorithm to predict the price of a\\nstock, our sole requirement for the algorithm is that it predict accurately—\\ninterpretability is not a concern. In this setting, we might expect that it\\nwill be best to use the most flexible model available. Surprisingly, this is\\nnot always the case! We will often obtain more accurate predictions using\\na less flexible method. This phenomenon, which may seem counterintu-\\nitive at first glance, has to do with the potential for overfitting in highly\\nflexible methods. We saw an example of overfitting in Figure 2.6. We will\\ndiscuss this very important concept further in Section 2.2 and throughout\\nthis book.\\n2.1.4\\nSupervised Versus Unsupervised Learning\\nMost statistical learning problems fall into one of two categories: supervised supervised\\nor unsupervised. The examples that we have discussed so far in this chap- unsupervised\\nter all fall into the supervised learning domain. For each observation of the\\npredictor measurement(s) xi, i = 1, . . . , n there is an associated response\\nmeasurement yi. We wish to fit a model that relates the response to the\\npredictors, with the aim of accurately predicting the response for future\\nobservations (prediction) or better understanding the relationship between\\nthe response and the predictors (inference). Many classical statistical learn-\\ning methods such as linear regression and logistic regression (Chapter 4), as logistic\\nregression\\nwell as more modern approaches such as GAM, boosting, and support vec-\\ntor machines, operate in the supervised learning domain. The vast majority\\nof this book is devoted to this setting.\\nBy contrast, unsupervised learning describes the somewhat more chal-\\nlenging situation in which for every observation i = 1, . . . , n, we observe\\na vector of measurements xi but no associated response yi. It is not pos-\\nsible to fit a linear regression model, since there is no response variable\\nto predict. In this setting, we are in some sense working blind; the sit-\\nuation is referred to as unsupervised because we lack a response vari-\\nable that can supervise our analysis. What sort of statistical analysis is\\npossible? We can seek to understand the relationships between the variables\\nor between the observations. One statistical learning tool that we may use\\nin this setting is cluster analysis, or clustering. The goal of cluster analysis cluster\\nanalysis\\nis to ascertain, on the basis of x1, . . . , xn, whether the observations fall into\\nrelatively distinct groups. For example, in a market segmentation study we\\nmight observe multiple characteristics (variables) for potential customers,\\nsuch as zip code, family income, and shopping habits. We might believe\\nthat the customers fall into different groups, such as big spenders versus\\nlow spenders. If the information about each customer’s spending patterns\\nwere available, then a supervised analysis would be possible. However, this\\ninformation is not available—that is, we do not know whether each poten-\\ntial customer is a big spender or not. In this setting, we can try to cluster\\nthe customers on the basis of the variables measured, in order to identify 26\\n2. Statistical Learning\\n0\\n2\\n4\\n6\\n8\\n10\\n12\\n2\\n4\\n6\\n8\\n10\\n12\\n0\\n2\\n4\\n6\\n2\\n4\\n6\\n8\\nX1\\nX1\\nX2\\nX2\\nFIGURE 2.8.\\nA clustering data set involving three groups. Each group is\\nshown using a different colored symbol. Left: The three groups are well-separated.\\nIn this setting, a clustering approach should successfully identify the three groups.\\nRight: There is some overlap among the groups. Now the clustering task is more\\nchallenging.\\ndistinct groups of potential customers. Identifying such groups can be of\\ninterest because it might be that the groups differ with respect to some\\nproperty of interest, such as spending habits.\\nFigure 2.8 provides a simple illustration of the clustering problem. We\\nhave plotted 150 observations with measurements on two variables, X1\\nand X2. Each observation corresponds to one of three distinct groups. For\\nillustrative purposes, we have plotted the members of each group using\\ndifferent colors and symbols. However, in practice the group memberships\\nare unknown, and the goal is to determine the group to which each obser-\\nvation belongs. In the left-hand panel of Figure 2.8, this is a relatively easy\\ntask because the groups are well-separated. By contrast, the right-hand\\npanel illustrates a more challenging setting in which there is some overlap\\nbetween the groups. A clustering method could not be expected to assign\\nall of the overlapping points to their correct group (blue, green, or orange).\\nIn the examples shown in Figure 2.8, there are only two variables, and\\nso one can simply visually inspect the scatterplots of the observations in\\norder to identify clusters. However, in practice, we often encounter data\\nsets that contain many more than two variables. In this case, we cannot\\neasily plot the observations. For instance, if there are p variables in our\\ndata set, then p(p −1)/2 distinct scatterplots can be made, and visual\\ninspection is simply not a viable way to identify clusters. For this reason,\\nautomated clustering methods are important. We discuss clustering and\\nother unsupervised learning approaches in Chapter 12.\\nMany problems fall naturally into the supervised or unsupervised learn-\\ning paradigms. However, sometimes the question of whether an analysis\\nshould be considered supervised or unsupervised is less clear-cut. For in-\\nstance, suppose that we have a set of n observations. For m of the observa-\\ntions, where m < n, we have both predictor measurements and a response 2.2 Assessing Model Accuracy\\n27\\nmeasurement. For the remaining n −m observations, we have predictor\\nmeasurements but no response measurement. Such a scenario can arise if\\nthe predictors can be measured relatively cheaply but the corresponding\\nresponses are much more expensive to collect. We refer to this setting as\\na semi-supervised learning problem. In this setting, we wish to use a sta- semi-\\nsupervised\\nlearning\\ntistical learning method that can incorporate the m observations for which\\nresponse measurements are available as well as the n −m observations for\\nwhich they are not. Although this is an interesting topic, it is beyond the\\nscope of this book.\\n2.1.5\\nRegression Versus Classification Problems\\nVariables can be characterized as either quantitative or qualitative (also quantitative\\nqualitative\\nknown as categorical). Quantitative variables take on numerical values. Ex-\\ncategorical\\namples include a person’s age, height, or income, the value of a house, and\\nthe price of a stock. In contrast, qualitative variables take on values in\\none of K different classes, or categories. Examples of qualitative variables class\\ninclude a person’s marital status (married or not), the brand of product\\npurchased (brand A, B, or C), whether a person defaults on a debt (yes\\nor no), or a cancer diagnosis (Acute Myelogenous Leukemia, Acute Lym-\\nphoblastic Leukemia, or No Leukemia). We tend to refer to problems with\\na quantitative response as regression problems, while those involving a regression\\nqualitative response are often referred to as classification problems. How- classification\\never, the distinction is not always that crisp. Least squares linear regression\\n(Chapter 3) is used with a quantitative response, whereas logistic regression\\n(Chapter 4) is typically used with a qualitative (two-class, or binary) re- binary\\nsponse. Thus, despite its name, logistic regression is a classification method.\\nBut since it estimates class probabilities, it can be thought of as a regres-\\nsion method as well. Some statistical methods, such as K-nearest neighbors\\n(Chapters 2 and 4) and boosting (Chapter 8), can be used in the case of\\neither quantitative or qualitative responses.\\nWe tend to select statistical learning methods on the basis of whether\\nthe response is quantitative or qualitative; i.e. we might use linear regres-\\nsion when quantitative and logistic regression when qualitative. However,\\nwhether the predictors are qualitative or quantitative is generally consid-\\nered less important. Most of the statistical learning methods discussed in\\nthis book can be applied regardless of the predictor variable type, provided\\nthat any qualitative predictors are properly coded before the analysis is\\nperformed. This is discussed in Chapter 3.\\n2.2\\nAssessing Model Accuracy\\nOne of the key aims of this book is to introduce the reader to a wide range\\nof statistical learning methods that extend far beyond the standard linear\\nregression approach. Why is it necessary to introduce so many different\\nstatistical learning approaches, rather than just a single best method? There\\nis no free lunch in statistics: no one method dominates all others over all\\npossible data sets. On a particular data set, one specific method may work 28\\n2. Statistical Learning\\nbest, but some other method may work better on a similar but different\\ndata set. Hence it is an important task to decide for any given set of data\\nwhich method produces the best results. Selecting the best approach can\\nbe one of the most challenging parts of performing statistical learning in\\npractice.\\nIn this section, we discuss some of the most important concepts that\\narise in selecting a statistical learning procedure for a specific data set. As\\nthe book progresses, we will explain how the concepts presented here can\\nbe applied in practice.\\n2.2.1\\nMeasuring the Quality of Fit\\nIn order to evaluate the performance of a statistical learning method on\\na given data set, we need some way to measure how well its predictions\\nactually match the observed data. That is, we need to quantify the extent\\nto which the predicted response value for a given observation is close to\\nthe true response value for that observation. In the regression setting, the\\nmost commonly-used measure is the mean squared error (MSE), given by\\nmean\\nsquared\\nerror\\nMSE = 1\\nn\\nn\\n0\\ni=1\\n(yi −ˆf(xi))2,\\n(2.5)\\nwhere ˆf(xi) is the prediction that ˆf gives for the ith observation. The MSE\\nwill be small if the predicted responses are very close to the true responses,\\nand will be large if for some of the observations, the predicted and true\\nresponses differ substantially.\\nThe MSE in (2.5) is computed using the training data that was used to\\nfit the model, and so should more accurately be referred to as the training\\nMSE. But in general, we do not really care how well the method works training\\nMSE\\non the training data. Rather, we are interested in the accuracy of the pre-\\ndictions that we obtain when we apply our method to previously unseen\\ntest data. Why is this what we care about? Suppose that we are interested test data\\nin developing an algorithm to predict a stock’s price based on previous\\nstock returns. We can train the method using stock returns from the past\\n6 months. But we don’t really care how well our method predicts last week’s\\nstock price. We instead care about how well it will predict tomorrow’s price\\nor next month’s price. On a similar note, suppose that we have clinical\\nmeasurements (e.g. weight, blood pressure, height, age, family history of\\ndisease) for a number of patients, as well as information about whether each\\npatient has diabetes. We can use these patients to train a statistical learn-\\ning method to predict risk of diabetes based on clinical measurements. In\\npractice, we want this method to accurately predict diabetes risk for future\\npatients based on their clinical measurements. We are not very interested\\nin whether or not the method accurately predicts diabetes risk for patients\\nused to train the model, since we already know which of those patients\\nhave diabetes.\\nTo state it more mathematically, suppose that we fit our statistical learn-\\ning method on our training observations {(x1, y1), (x2, y2), . . . , (xn, yn)},\\nand we obtain the estimate ˆf. We can then compute ˆf(x1), ˆf(x2), . . . , ˆf(xn). 2.2 Assessing Model Accuracy\\n29\\n0\\n20\\n40\\n60\\n80\\n100\\n2\\n4\\n6\\n8\\n10\\n12\\nX\\nY\\n2\\n5\\n10\\n20\\n0.0\\n0.5\\n1.0\\n1.5\\n2.0\\n2.5\\nFlexibility\\nMean Squared Error\\nFIGURE 2.9. Left: Data simulated from f, shown in black. Three estimates of\\nf are shown: the linear regression line (orange curve), and two smoothing spline\\nfits (blue and green curves). Right: Training MSE (grey curve), test MSE (red\\ncurve), and minimum possible test MSE over all methods (dashed line). Squares\\nrepresent the training and test MSEs for the three fits shown in the left-hand\\npanel.\\nIf these are approximately equal to y1, y2, . . . , yn, then the training MSE\\ngiven by (2.5) is small. However, we are really not interested in whether\\nˆf(xi) ≈yi; instead, we want to know whether ˆf(x0) is approximately equal\\nto y0, where (x0, y0) is a previously unseen test observation not used to train\\nthe statistical learning method. We want to choose the method that gives\\nthe lowest test MSE, as opposed to the lowest training MSE. In other words, test MSE\\nif we had a large number of test observations, we could compute\\nAve(y0 −ˆf(x0))2,\\n(2.6)\\nthe average squared prediction error for these test observations (x0, y0).\\nWe’d like to select the model for which this quantity is as small as possible.\\nHow can we go about trying to select a method that minimizes the test\\nMSE? In some settings, we may have a test data set available—that is,\\nwe may have access to a set of observations that were not used to train\\nthe statistical learning method. We can then simply evaluate (2.6) on the\\ntest observations, and select the learning method for which the test MSE is\\nsmallest. But what if no test observations are available? In that case, one\\nmight imagine simply selecting a statistical learning method that minimizes\\nthe training MSE (2.5). This seems like it might be a sensible approach,\\nsince the training MSE and the test MSE appear to be closely related.\\nUnfortunately, there is a fundamental problem with this strategy: there\\nis no guarantee that the method with the lowest training MSE will also\\nhave the lowest test MSE. Roughly speaking, the problem is that many\\nstatistical methods specifically estimate coefficients so as to minimize the\\ntraining set MSE. For these methods, the training set MSE can be quite\\nsmall, but the test MSE is often much larger.\\nFigure 2.9 illustrates this phenomenon on a simple example. In the left-\\nhand panel of Figure 2.9, we have generated observations from (2.1) with 30\\n2. Statistical Learning\\nthe true f given by the black curve. The orange, blue and green curves illus-\\ntrate three possible estimates for f obtained using methods with increasing\\nlevels of flexibility. The orange line is the linear regression fit, which is rela-\\ntively inflexible. The blue and green curves were produced using smoothing\\nsplines, discussed in Chapter 7, with different levels of smoothness. It is smoothing\\nspline\\nclear that as the level of flexibility increases, the curves fit the observed\\ndata more closely. The green curve is the most flexible and matches the\\ndata very well; however, we observe that it fits the true f (shown in black)\\npoorly because it is too wiggly. By adjusting the level of flexibility of the\\nsmoothing spline fit, we can produce many different fits to this data.\\nWe now move on to the right-hand panel of Figure 2.9. The grey curve\\ndisplays the average training MSE as a function of flexibility, or more\\nformally the degrees of freedom, for a number of smoothing splines. The degrees of\\nfreedom\\ndegrees of freedom is a quantity that summarizes the flexibility of a curve;\\nit is discussed more fully in Chapter 7. The orange, blue and green squares\\nindicate the MSEs associated with the corresponding curves in the left-\\nhand panel. A more restricted and hence smoother curve has fewer degrees\\nof freedom than a wiggly curve—note that in Figure 2.9, linear regression\\nis at the most restrictive end, with two degrees of freedom. The training\\nMSE declines monotonically as flexibility increases. In this example the\\ntrue f is non-linear, and so the orange linear fit is not flexible enough to\\nestimate f well. The green curve has the lowest training MSE of all three\\nmethods, since it corresponds to the most flexible of the three curves fit in\\nthe left-hand panel.\\nIn this example, we know the true function f, and so we can also com-\\npute the test MSE over a very large test set, as a function of flexibility. (Of\\ncourse, in general f is unknown, so this will not be possible.) The test MSE\\nis displayed using the red curve in the right-hand panel of Figure 2.9. As\\nwith the training MSE, the test MSE initially declines as the level of flex-\\nibility increases. However, at some point the test MSE levels off and then\\nstarts to increase again. Consequently, the orange and green curves both\\nhave high test MSE. The blue curve minimizes the test MSE, which should\\nnot be surprising given that visually it appears to estimate f the best in the\\nleft-hand panel of Figure 2.9. The horizontal dashed line indicates Var(ϵ),\\nthe irreducible error in (2.3), which corresponds to the lowest achievable\\ntest MSE among all possible methods. Hence, the smoothing spline repre-\\nsented by the blue curve is close to optimal.\\nIn the right-hand panel of Figure 2.9, as the flexibility of the statistical\\nlearning method increases, we observe a monotone decrease in the training\\nMSE and a U-shape in the test MSE. This is a fundamental property of\\nstatistical learning that holds regardless of the particular data set at hand\\nand regardless of the statistical method being used. As model flexibility\\nincreases, the training MSE will decrease, but the test MSE may not. When\\na given method yields a small training MSE but a large test MSE, we are\\nsaid to be overfitting the data. This happens because our statistical learning\\nprocedure is working too hard to find patterns in the training data, and\\nmay be picking up some patterns that are just caused by random chance\\nrather than by true properties of the unknown function f. When we overfit\\nthe training data, the test MSE will be very large because the supposed 2.2 Assessing Model Accuracy\\n31\\n0\\n20\\n40\\n60\\n80\\n100\\n2\\n4\\n6\\n8\\n10\\n12\\nX\\nY\\n2\\n5\\n10\\n20\\n0.0\\n0.5\\n1.0\\n1.5\\n2.0\\n2.5\\nFlexibility\\nMean Squared Error\\nFIGURE 2.10.\\nDetails are as in Figure 2.9, using a different true f that is\\nmuch closer to linear. In this setting, linear regression provides a very good fit to\\nthe data.\\npatterns that the method found in the training data simply don’t exist\\nin the test data. Note that regardless of whether or not overfitting has\\noccurred, we almost always expect the training MSE to be smaller than\\nthe test MSE because most statistical learning methods either directly or\\nindirectly seek to minimize the training MSE. Overfitting refers specifically\\nto the case in which a less flexible model would have yielded a smaller\\ntest MSE.\\nFigure 2.10 provides another example in which the true f is approxi-\\nmately linear. Again we observe that the training MSE decreases mono-\\ntonically as the model flexibility increases, and that there is a U-shape in\\nthe test MSE. However, because the truth is close to linear, the test MSE\\nonly decreases slightly before increasing again, so that the orange least\\nsquares fit is substantially better than the highly flexible green curve. Fi-\\nnally, Figure 2.11 displays an example in which f is highly non-linear. The\\ntraining and test MSE curves still exhibit the same general patterns, but\\nnow there is a rapid decrease in both curves before the test MSE starts to\\nincrease slowly.\\nIn practice, one can usually compute the training MSE with relative\\nease, but estimating the test MSE is considerably more difficult because\\nusually no test data are available. As the previous three examples illustrate,\\nthe flexibility level corresponding to the model with the minimal test MSE\\ncan vary considerably among data sets. Throughout this book, we discuss a\\nvariety of approaches that can be used in practice to estimate this minimum\\npoint. One important method is cross-validation (Chapter 5), which is a cross-\\nvalidation\\nmethod for estimating the test MSE using the training data.\\n2.2.2\\nThe Bias-Variance Trade-Off\\nThe U-shape observed in the test MSE curves (Figures 2.9–2.11) turns out\\nto be the result of two competing properties of statistical learning methods. 32\\n2. Statistical Learning\\n0\\n20\\n40\\n60\\n80\\n100\\n−10\\n0\\n10\\n20\\nX\\nY\\n2\\n5\\n10\\n20\\n0\\n5\\n10\\n15\\n20\\nFlexibility\\nMean Squared Error\\nFIGURE 2.11. Details are as in Figure 2.9, using a different f that is far from\\nlinear. In this setting, linear regression provides a very poor fit to the data.\\nThough the mathematical proof is beyond the scope of this book, it is\\npossible to show that the expected test MSE, for a given value x0, can\\nalways be decomposed into the sum of three fundamental quantities: the\\nvariance of ˆf(x0), the squared bias of ˆf(x0) and the variance of the error variance\\nbias\\nterms ϵ. That is,\\nE\\n1\\ny0 −ˆf(x0)\\n22\\n= Var( ˆf(x0)) + [Bias( ˆf(x0))]2 + Var(ϵ).\\n(2.7)\\nHere the notation E\\n1\\ny0 −ˆf(x0)\\n22\\ndefines the expected test MSE at x0, expected\\ntest MSE\\nand refers to the average test MSE that we would obtain if we repeatedly\\nestimated f using a large number of training sets, and tested each at x0. The\\noverall expected test MSE can be computed by averaging E\\n1\\ny0 −ˆf(x0)\\n22\\nover all possible values of x0 in the test set.\\nEquation 2.7 tells us that in order to minimize the expected test error,\\nwe need to select a statistical learning method that simultaneously achieves\\nlow variance and low bias. Note that variance is inherently a nonnegative\\nquantity, and squared bias is also nonnegative. Hence, we see that the\\nexpected test MSE can never lie below Var(ϵ), the irreducible error from\\n(2.3).\\nWhat do we mean by the variance and bias of a statistical learning\\nmethod? Variance refers to the amount by which ˆf would change if we\\nestimated it using a different training data set. Since the training data\\nare used to fit the statistical learning method, different training data sets\\nwill result in a different ˆf. But ideally the estimate for f should not vary\\ntoo much between training sets. However, if a method has high variance\\nthen small changes in the training data can result in large changes in ˆf. In\\ngeneral, more flexible statistical methods have higher variance. Consider the\\ngreen and orange curves in Figure 2.9. The flexible green curve is following\\nthe observations very closely. It has high variance because changing any\\none of these data points may cause the estimate ˆf to change considerably. 2.2 Assessing Model Accuracy\\n33\\n2\\n5\\n10\\n20\\n0.0\\n0.5\\n1.0\\n1.5\\n2.0\\n2.5\\nFlexibility\\n2\\n5\\n10\\n20\\n0.0\\n0.5\\n1.0\\n1.5\\n2.0\\n2.5\\nFlexibility\\n2\\n5\\n10\\n20\\n0\\n5\\n10\\n15\\n20\\nFlexibility\\nMSE\\nBias\\nVar\\nFIGURE 2.12.\\nSquared bias (blue curve), variance (orange curve), Var(ϵ)\\n(dashed line), and test MSE (red curve) for the three data sets in Figures 2.9–2.11.\\nThe vertical dotted line indicates the flexibility level corresponding to the smallest\\ntest MSE.\\nIn contrast, the orange least squares line is relatively inflexible and has low\\nvariance, because moving any single observation will likely cause only a\\nsmall shift in the position of the line.\\nOn the other hand, bias refers to the error that is introduced by approxi-\\nmating a real-life problem, which may be extremely complicated, by a much\\nsimpler model. For example, linear regression assumes that there is a linear\\nrelationship between Y and X1, X2, . . . , Xp. It is unlikely that any real-life\\nproblem truly has such a simple linear relationship, and so performing lin-\\near regression will undoubtedly result in some bias in the estimate of f. In\\nFigure 2.11, the true f is substantially non-linear, so no matter how many\\ntraining observations we are given, it will not be possible to produce an\\naccurate estimate using linear regression. In other words, linear regression\\nresults in high bias in this example. However, in Figure 2.10 the true f\\nis very close to linear, and so given enough data, it should be possible for\\nlinear regression to produce an accurate estimate. Generally, more flexible\\nmethods result in less bias.\\nAs a general rule, as we use more flexible methods, the variance will\\nincrease and the bias will decrease. The relative rate of change of these\\ntwo quantities determines whether the test MSE increases or decreases. As\\nwe increase the flexibility of a class of methods, the bias tends to initially\\ndecrease faster than the variance increases. Consequently, the expected\\ntest MSE declines. However, at some point increasing flexibility has little\\nimpact on the bias but starts to significantly increase the variance. When\\nthis happens the test MSE increases. Note that we observed this pattern\\nof decreasing test MSE followed by increasing test MSE in the right-hand\\npanels of Figures 2.9–2.11.\\nThe three plots in Figure 2.12 illustrate Equation 2.7 for the examples in\\nFigures 2.9–2.11. In each case the blue solid curve represents the squared\\nbias, for different levels of flexibility, while the orange curve corresponds to\\nthe variance. The horizontal dashed line represents Var(ϵ), the irreducible\\nerror. Finally, the red curve, corresponding to the test set MSE, is the sum 34\\n2. Statistical Learning\\nof these three quantities. In all three cases, the variance increases and the\\nbias decreases as the method’s flexibility increases. However, the flexibility\\nlevel corresponding to the optimal test MSE differs considerably among the\\nthree data sets, because the squared bias and variance change at different\\nrates in each of the data sets. In the left-hand panel of Figure 2.12, the\\nbias initially decreases rapidly, resulting in an initial sharp decrease in the\\nexpected test MSE. On the other hand, in the center panel of Figure 2.12\\nthe true f is close to linear, so there is only a small decrease in bias as flex-\\nibility increases, and the test MSE only declines slightly before increasing\\nrapidly as the variance increases. Finally, in the right-hand panel of Fig-\\nure 2.12, as flexibility increases, there is a dramatic decline in bias because\\nthe true f is very non-linear. There is also very little increase in variance\\nas flexibility increases. Consequently, the test MSE declines substantially\\nbefore experiencing a small increase as model flexibility increases.\\nThe relationship between bias, variance, and test set MSE given in Equa-\\ntion 2.7 and displayed in Figure 2.12 is referred to as the bias-variance\\ntrade-off. Good test set performance of a statistical learning method re- bias-variance\\ntrade-off\\nquires low variance as well as low squared bias. This is referred to as a\\ntrade-off because it is easy to obtain a method with extremely low bias but\\nhigh variance (for instance, by drawing a curve that passes through every\\nsingle training observation) or a method with very low variance but high\\nbias (by fitting a horizontal line to the data). The challenge lies in finding\\na method for which both the variance and the squared bias are low. This\\ntrade-off is one of the most important recurring themes in this book.\\nIn a real-life situation in which f is unobserved, it is generally not pos-\\nsible to explicitly compute the test MSE, bias, or variance for a statistical\\nlearning method. Nevertheless, one should always keep the bias-variance\\ntrade-off in mind. In this book we explore methods that are extremely\\nflexible and hence can essentially eliminate bias. However, this does not\\nguarantee that they will outperform a much simpler method such as linear\\nregression. To take an extreme example, suppose that the true f is linear.\\nIn this situation linear regression will have no bias, making it very hard\\nfor a more flexible method to compete. In contrast, if the true f is highly\\nnon-linear and we have an ample number of training observations, then\\nwe may do better using a highly flexible approach, as in Figure 2.11. In\\nChapter 5 we discuss cross-validation, which is a way to estimate the test\\nMSE using the training data.\\n2.2.3\\nThe Classification Setting\\nThus far, our discussion of model accuracy has been focused on the regres-\\nsion setting. But many of the concepts that we have encountered, such\\nas the bias-variance trade-off, transfer over to the classification setting\\nwith only some modifications due to the fact that yi is no longer quan-\\ntitative. Suppose that we seek to estimate f on the basis of training obser-\\nvations {(x1, y1), . . . , (xn, yn)}, where now y1, . . . , yn are qualitative. The\\nmost common approach for quantifying the accuracy of our estimate ˆf is\\nthe training error rate, the proportion of mistakes that are made if we apply error rate 2.2 Assessing Model Accuracy\\n35\\nour estimate ˆf to the training observations:\\n1\\nn\\nn\\n0\\ni=1\\nI(yi ̸= ˆyi).\\n(2.8)\\nHere ˆyi is the predicted class label for the ith observation using ˆf. And\\nI(yi ̸= ˆyi) is an indicator variable that equals 1 if yi ̸= ˆyi and zero if yi = ˆyi. indicator\\nvariable\\nIf I(yi ̸= ˆyi) = 0 then the ith observation was classified correctly by our\\nclassification method; otherwise it was misclassified. Hence Equation 2.8\\ncomputes the fraction of incorrect classifications.\\nEquation 2.8 is referred to as the training error rate because it is com- training\\nerror\\nputed based on the data that was used to train our classifier. As in the\\nregression setting, we are most interested in the error rates that result from\\napplying our classifier to test observations that were not used in training.\\nThe test error rate associated with a set of test observations of the form test error\\n(x0, y0) is given by\\nAve (I(y0 ̸= ˆy0)) ,\\n(2.9)\\nwhere ˆy0 is the predicted class label that results from applying the classifier\\nto the test observation with predictor x0. A good classifier is one for which\\nthe test error (2.9) is smallest.\\nThe Bayes Classifier\\nIt is possible to show (though the proof is outside of the scope of this\\nbook) that the test error rate given in (2.9) is minimized, on average, by a\\nvery simple classifier that assigns each observation to the most likely class,\\ngiven its predictor values. In other words, we should simply assign a test\\nobservation with predictor vector x0 to the class j for which\\nPr(Y = j|X = x0)\\n(2.10)\\nis largest. Note that (2.10) is a conditional probability: it is the probability conditional\\nprobability\\nthat Y = j, given the observed predictor vector x0. This very simple clas-\\nsifier is called the Bayes classifier. In a two-class problem where there are Bayes\\nclassifier\\nonly two possible response values, say class 1 or class 2, the Bayes classifier\\ncorresponds to predicting class one if Pr(Y = 1|X = x0) > 0.5, and class\\ntwo otherwise.\\nFigure 2.13 provides an example using a simulated data set in a two-\\ndimensional space consisting of predictors X1 and X2. The orange and\\nblue circles correspond to training observations that belong to two different\\nclasses. For each value of X1 and X2, there is a different probability of the\\nresponse being orange or blue. Since this is simulated data, we know how\\nthe data were generated and we can calculate the conditional probabilities\\nfor each value of X1 and X2. The orange shaded region reflects the set of\\npoints for which Pr(Y = orange|X) is greater than 50 %, while the blue\\nshaded region indicates the set of points for which the probability is below\\n50 %. The purple dashed line represents the points where the probability\\nis exactly 50 %. This is called the Bayes decision boundary. The Bayes Bayes\\ndecision\\nboundary\\nclassifier’s prediction is determined by the Bayes decision boundary; an\\nobservation that falls on the orange side of the boundary will be assigned 36\\n2. Statistical Learning\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\noo\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\noo\\no\\no\\no\\no\\no\\no\\no o\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\noo o\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\nX1\\nX2\\nFIGURE 2.13. A simulated data set consisting of 100 observations in each of\\ntwo groups, indicated in blue and in orange. The purple dashed line represents\\nthe Bayes decision boundary. The orange background grid indicates the region\\nin which a test observation will be assigned to the orange class, and the blue\\nbackground grid indicates the region in which a test observation will be assigned\\nto the blue class.\\nto the orange class, and similarly an observation on the blue side of the\\nboundary will be assigned to the blue class.\\nThe Bayes classifier produces the lowest possible test error rate, called\\nthe Bayes error rate. Since the Bayes classifier will always choose the class Bayes error\\nrate\\nfor which (2.10) is largest, the error rate will be 1−maxj Pr(Y = j|X = x0)\\nat X = x0. In general, the overall Bayes error rate is given by\\n1 −E\\n*\\nmax\\nj\\nPr(Y = j|X)\\n+\\n,\\n(2.11)\\nwhere the expectation averages the probability over all possible values of\\nX. For our simulated data, the Bayes error rate is 0.133. It is greater than\\nzero, because the classes overlap in the true population, which implies that\\nmaxj Pr(Y = j|X = x0) < 1 for some values of x0. The Bayes error rate is\\nanalogous to the irreducible error, discussed earlier.\\nK-Nearest Neighbors\\nIn theory we would always like to predict qualitative responses using the\\nBayes classifier. But for real data, we do not know the conditional distri-\\nbution of Y given X, and so computing the Bayes classifier is impossi-\\nble. Therefore, the Bayes classifier serves as an unattainable gold standard\\nagainst which to compare other methods. Many approaches attempt to\\nestimate the conditional distribution of Y given X, and then classify a\\ngiven observation to the class with highest estimated probability. One such\\nmethod is the K-nearest neighbors (KNN) classifier. Given a positive in- K-nearest\\nneighbors 2.2 Assessing Model Accuracy\\n37\\nteger K and a test observation x0, the KNN classifier first identifies the\\nK points in the training data that are closest to x0, represented by N0.\\nIt then estimates the conditional probability for class j as the fraction of\\npoints in N0 whose response values equal j:\\nPr(Y = j|X = x0) = 1\\nK\\n0\\ni∈N0\\nI(yi = j).\\n(2.12)\\nFinally, KNN classifies the test observation x0 to the class with the largest\\nprobability from (2.12).\\nFigure 2.14 provides an illustrative example of the KNN approach. In\\nthe left-hand panel, we have plotted a small training data set consisting of\\nsix blue and six orange observations. Our goal is to make a prediction for\\nthe point labeled by the black cross. Suppose that we choose K = 3. Then\\nKNN will first identify the three observations that are closest to the cross.\\nThis neighborhood is shown as a circle. It consists of two blue points and\\none orange point, resulting in estimated probabilities of 2/3 for the blue\\nclass and 1/3 for the orange class. Hence KNN will predict that the black\\ncross belongs to the blue class. In the right-hand panel of Figure 2.14 we\\nhave applied the KNN approach with K = 3 at all of the possible values for\\nX1 and X2, and have drawn in the corresponding KNN decision boundary.\\nDespite the fact that it is a very simple approach, KNN can often pro-\\nduce classifiers that are surprisingly close to the optimal Bayes classifier.\\nFigure 2.15 displays the KNN decision boundary, using K = 10, when ap-\\nplied to the larger simulated data set from Figure 2.13. Notice that even\\nthough the true distribution is not known by the KNN classifier, the KNN\\ndecision boundary is very close to that of the Bayes classifier. The test error\\nrate using KNN is 0.1363, which is close to the Bayes error rate of 0.1304.\\nThe choice of K has a drastic effect on the KNN classifier obtained.\\nFigure 2.16 displays two KNN fits to the simulated data from Figure 2.13,\\nusing K = 1 and K = 100. When K = 1, the decision boundary is overly\\nflexible and finds patterns in the data that don’t correspond to the Bayes\\ndecision boundary. This corresponds to a classifier that has low bias but\\nvery high variance. As K grows, the method becomes less flexible and\\nproduces a decision boundary that is close to linear. This corresponds to\\na low-variance but high-bias classifier. On this simulated data set, neither\\nK = 1 nor K = 100 give good predictions: they have test error rates of\\n0.1695 and 0.1925, respectively.\\nJust as in the regression setting, there is not a strong relationship be-\\ntween the training error rate and the test error rate. With K = 1, the\\nKNN training error rate is 0, but the test error rate may be quite high. In\\ngeneral, as we use more flexible classification methods, the training error\\nrate will decline but the test error rate may not. In Figure 2.17, we have\\nplotted the KNN test and training errors as a function of 1/K. As 1/K in-\\ncreases, the method becomes more flexible. As in the regression setting, the\\ntraining error rate consistently declines as the flexibility increases. However,\\nthe test error exhibits a characteristic U-shape, declining at first (with a\\nminimum at approximately K = 10) before increasing again when the\\nmethod becomes excessively flexible and overfits. 38\\n2. Statistical Learning\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\nFIGURE 2.14.\\nThe KNN approach, using K = 3, is illustrated in a simple\\nsituation with six blue observations and six orange observations. Left: a test\\nobservation at which a predicted class label is desired is shown as a black cross.\\nThe three closest points to the test observation are identified, and it is predicted\\nthat the test observation belongs to the most commonly-occurring class, in this\\ncase blue. Right: The KNN decision boundary for this example is shown in black.\\nThe blue grid indicates the region in which a test observation will be assigned to\\nthe blue class, and the orange grid indicates the region in which it will be assigned\\nto the orange class.\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\noo\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\noo\\no\\no\\no\\no\\no\\no\\no o\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\noo o\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\nX1\\nX2\\nKNN: K=10\\nFIGURE 2.15. The black curve indicates the KNN decision boundary on the\\ndata from Figure 2.13, using K = 10. The Bayes decision boundary is shown as\\na purple dashed line. The KNN and Bayes decision boundaries are very similar.\\nIn both the regression and classification settings, choosing the correct\\nlevel of flexibility is critical to the success of any statistical learning method.\\nThe bias-variance tradeoff, and the resulting U-shape in the test error, can\\nmake this a difficult task. In Chapter 5, we return to this topic and discuss 2.2 Assessing Model Accuracy\\n39\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\noo\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\noo\\no\\no\\no\\no\\no\\no\\no o\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\noo\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\noo\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\noo\\no\\no\\no\\no\\no\\no\\no o\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\noo\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\nKNN: K=1\\nKNN: K=100\\nFIGURE 2.16.\\nA comparison of the KNN decision boundaries (solid black\\ncurves) obtained using K = 1 and K = 100 on the data from Figure 2.13. With\\nK = 1, the decision boundary is overly flexible, while with K = 100 it is not\\nsufficiently flexible. The Bayes decision boundary is shown as a purple dashed\\nline.\\n0.01\\n0.02\\n0.05\\n0.10\\n0.20\\n0.50\\n1.00\\n0.00\\n0.05\\n0.10\\n0.15\\n0.20\\n1/K\\nError Rate\\nTraining Errors\\nTest Errors\\nFIGURE 2.17. The KNN training error rate (blue, 200 observations) and test\\nerror rate (orange, 5,000 observations) on the data from Figure 2.13, as the level\\nof flexibility (assessed using 1/K on the log scale) increases, or equivalently as\\nthe number of neighbors K decreases. The black dashed line indicates the Bayes\\nerror rate. The jumpiness of the curves is due to the small size of the training\\ndata set.\\nvarious methods for estimating test error rates and thereby choosing the\\noptimal level of flexibility for a given statistical learning method. 40\\n2. Statistical Learning\\n2.3\\nLab: Introduction to Python\\n2.3.1\\nGetting Started\\nTo run the labs in this book, you will need two things:\\n1. An installation of Python3, which is the specific version of Python used\\nin the labs.\\n2. Access to Jupyter, a very popular Python interface that runs code\\nthrough a file called a notebook.\\nnotebook\\nYou can download and install Python3 by following the instructions avail-\\nable at anaconda.com.\\nThere are a number of ways to get access to Jupyter. Here are just a few:\\n1. Using Google’s Colaboratory service: colab.research.google.com/.\\n2. Using JupyterHub, available at jupyter.org/hub.\\n3. Using your own jupyter installation. Installation instructions are\\navailable at jupyter.org/install.\\nPlease see the Python resources page on the book website statlearning.com\\nfor up-to-date information about getting Python and Jupyter working on\\nyour computer.\\nYou will need to install the ISLP package, which provides access to the\\ndatasets and custom-built functions that we provide. Inside a macOS or\\nLinux terminal type pip install ISLP; this also installs most other pack-\\nages needed in the labs. The Python resources page has a link to the ISLP\\ndocumentation website.\\nTo run this lab, download the file Ch2-statlearn-lab.ipynb from the\\nPython resources page. Now run the following code at the command line:\\njupyter lab Ch2-statlearn-lab.ipynb.\\nIf you’re using Windows, you can use the start menu to access anaconda,\\nand follow the links. For example, to install ISLP and run this lab, you can\\nrun the same code above in an anaconda shell.\\n2.3.2\\nBasic Commands\\nIn this lab, we will introduce some simple Python commands. For more\\nresources about Python in general, readers may want to consult the tutorial\\nat docs.python.org/3/tutorial/.\\nLike most programming languages, Python uses functions to perform op- function\\nerations. To run a function called fun, we type fun(input1,input2), where\\nthe inputs (or arguments) input1 and input2 tell Python how to run the\\nfunction. A function can have any number of inputs. For example, the argument\\nprint() function outputs a text representation of all of its arguments to print()\\nthe console.\\nIn [1]: print(\\'fit a model with\\', 11, \\'variables\\') 2.3 Lab: Introduction to Python\\n41\\nfit a model with 11 variables\\nThe following command will provide information about the print() func-\\ntion.\\nIn [2]: print?\\nAdding two integers in Python is pretty intuitive.\\nIn [3]: 3 + 5\\nOut[3]: 8\\nIn Python, textual data is handled using strings. For instance, \"hello\" and string\\n\\'hello\\' are strings. We can concatenate them using the addition + symbol.\\nIn [4]: \"hello\" + \" \" + \"world\"\\nOut[4]: \\'hello world\\'\\nA string is actually a type of sequence: this is a generic term for an ordered sequence\\nlist. The three most important types of sequences are lists, tuples, and\\nstrings. We introduce lists now.\\nThe following command instructs Python to join together the numbers 3,\\n4, and 5, and to save them as a list named x. When we type x, it gives us list\\nback the list.\\nIn [5]: x = [3, 4, 5]\\nx\\nOut[5]: [3, 4, 5]\\nNote that we used the brackets [] to construct this list.\\nWe will often want to add two sets of numbers together. It is reasonable\\nto try the following code, though it will not produce the desired results.\\nIn [6]: y = [4, 9, 7]\\nx + y\\nOut[6]: [3, 4, 5, 4, 9, 7]\\nThe result may appear slightly counterintuitive: why did Python not add\\nthe entries of the lists element-by-element? In Python, lists hold arbitrary\\nobjects, and are added using concatenation. In fact, concatenation is the concatenat-\\nion\\nbehavior that we saw earlier when we entered \"hello\" + \" \" + \"world\".\\nThis example reflects the fact that Python is a general-purpose program-\\nming language. Much of Python’s data-specific functionality comes from\\nother packages, notably numpy and pandas. In the next section, we will intro-\\nduce the numpy package. See docs.scipy.org/doc/numpy/user/quickstart.html\\nfor more information about numpy. 42\\n2. Statistical Learning\\n2.3.3\\nIntroduction to Numerical Python\\nAs mentioned earlier, this book makes use of functionality that is contained\\nin the numpy library, or package. A package is a collection of modules that numpy\\npackage\\nare not necessarily included in the base Python distribution. The name numpy\\nis an abbreviation for numerical Python.\\nTo access numpy, we must first import it.\\nimport\\nIn [7]: import numpy as np\\nIn the previous line, we named the numpy module np; an abbreviation for module\\neasier referencing.\\nIn numpy, an array is a generic term for a multidimensional set of numbers. array\\nWe use the np.array() function to define x and y, which are one-dimensional np.array()\\narrays, i.e. vectors.\\nIn [8]: x = np.array([3, 4, 5])\\ny = np.array([4, 9, 7])\\nNote that if you forgot to run the import numpy as np command earlier,\\nthen you will encounter an error in calling the np.array() function in\\nthe previous line. The syntax np.array() indicates that the function be-\\ning called is part of the numpy package, which we have abbreviated as np.\\nSince x and y have been defined using np.array(), we get a sensible result\\nwhen we add them together. Compare this to our results in the previous\\nsection, when we tried to add two lists without using numpy.\\nIn [9]: x + y\\nOut[9]: array([ 7, 13, 12])\\nIn numpy, matrices are typically represented as two-dimensional arrays,\\nand vectors as one-dimensional arrays.1 We can create a two-dimensional\\narray as follows.\\nIn [10]: x = np.array([[1, 2], [3, 4]])\\nx\\nOut[10]: array([[1, 2],\\n[3, 4]])\\nThe object x has several attributes, or associated objects. To access an attribute\\nattribute of x, we type x.attribute, where we replace attribute with the\\nname of the attribute. For instance, we can access the ndim attribute of x ndim\\nas follows.\\nIn [11]: x.ndim\\nOut[11]: 2\\nThe output indicates that x is a two-dimensional array. Similarly, x.dtype\\nis the data type attribute of the object x. This indicates that x is comprised data type\\nof 64-bit integers:\\n1While it is also possible to create matrices using np.matrix(), we will use\\nnp.array() throughout the labs in this book. 2.3 Lab: Introduction to Python\\n43\\nIn [12]: x.dtype\\nOut[12]: dtype(\\'int64\\')\\nWhy is x comprised of integers? This is because we created x by passing\\nin exclusively integers to the np.array() function. If we had passed in any\\ndecimals, then we would have obtained an array of floating point numbers floating\\npoint\\n(i.e. real-valued numbers).\\nIn [13]: np.array([[1, 2], [3.0, 4]]).dtype\\nOut[13]: dtype(\\'float64\\')\\nTyping fun? will cause Python to display documentation associated with\\nthe function fun, if it exists. We can try this for np.array().\\nIn [14]: np.array?\\nThis documentation indicates that we could create a floating point array\\nby passing a dtype argument into np.array().\\ndtype\\nIn [15]: np.array([[1, 2], [3, 4]], float).dtype\\nOut[15]: dtype(\\'float64\\')\\nThe array x is two-dimensional. We can find out the number of rows and\\ncolumns by looking at its shape attribute.\\nshape\\nIn [16]: x.shape\\nOut[16]: (2, 2)\\nA method is a function that is associated with an object. For instance, method\\ngiven an array x, the expression x.sum() sums all of its elements, using the\\nsum() method for arrays. The call x.sum() automatically provides x as the .sum()\\nfirst argument to its sum() method.\\nIn [17]: x = np.array([1, 2, 3, 4])\\nx.sum()\\nOut[17]: 10\\nWe could also sum the elements of x by passing in x as an argument to the\\nnp.sum() function.\\nnp.sum()\\nIn [18]: x = np.array([1, 2, 3, 4])\\nnp.sum(x)\\nOut[18]: 10\\nAs another example, the reshape() method returns a new array with the .reshape()\\nsame elements as x, but a different shape. We do this by passing in a tuple tuple 44\\n2. Statistical Learning\\nin our call to reshape(), in this case (2, 3). This tuple specifies that we\\nwould like to create a two-dimensional array with 2 rows and 3 columns.2\\nIn what follows, the \\\\n character creates a new line.\\nIn [19]: x = np.array([1, 2, 3, 4, 5, 6])\\nprint(\\'beginning x:\\\\n\\', x)\\nx_reshape = x.reshape((2, 3))\\nprint(\\'reshaped x:\\\\n\\', x_reshape)\\nbeginning x:\\n[1 2 3 4 5 6]\\nreshaped x:\\n[[1 2 3]\\n[4 5 6]]\\nThe previous output reveals that numpy arrays are specified as a sequence\\nof rows. This is called row-major ordering, as opposed to column-major\\nordering.\\nPython (and hence numpy) uses 0-based indexing. This means that to\\naccess the top left element of x_reshape, we type in x_reshape[0,0].\\nIn [20]: x_reshape[0, 0]\\nOut[20]: 1\\nSimilarly, x_reshape[1,2] yields the element in the second row and the\\nthird column of x_reshape.\\nIn [21]: x_reshape[1, 2]\\nOut[21]: 6\\nSimilarly, x[2] yields the third entry of x.\\nNow, let’s modify the top left element of x_reshape. To our surprise, we\\ndiscover that the first element of x has been modified as well!\\nIn [22]: print(\\'x before we modify x_reshape:\\\\n\\', x)\\nprint(\\'x_reshape before we modify x_reshape:\\\\n\\', x_reshape)\\nx_reshape[0, 0] = 5\\nprint(\\'x_reshape after we modify its top left element:\\\\n\\',\\nx_reshape)\\nprint(\\'x after we modify top left element of x_reshape:\\\\n\\', x)\\nOut[22]: x before we modify x_reshape:\\n[1 2 3 4 5 6]\\nx_reshape before we modify x_reshape:\\n[[1 2 3]\\n[4 5 6]]\\nx_reshape after we modify its top left element:\\n[[5 2 3]\\n2Like lists, tuples represent a sequence of objects. Why do we need more than one way\\nto create a sequence? There are a few differences between tuples and lists, but perhaps\\nthe most important is that elements of a tuple cannot be modified, whereas elements of\\na list can be. 2.3 Lab: Introduction to Python\\n45\\n[4 5 6]]\\nx after we modify top left element of x_reshape:\\n[5 2 3 4 5 6]\\nModifying x_reshape also modified x because the two objects occupy the\\nsame space in memory.\\nWe just saw that we can modify an element of an array. Can we also mod-\\nify a tuple? It turns out that we cannot — and trying to do so introduces\\nan exception, or error.\\nexception\\nIn [23]: my_tuple = (3, 4, 5)\\nmy_tuple[0] = 2\\nTypeError: \\'tuple\\' object does not support item assignment\\nWe now briefly mention some attributes of arrays that will come in handy.\\nAn array’s shape attribute contains its dimension; this is always a tuple.\\nThe ndim attribute yields the number of dimensions, and T provides its\\ntranspose.\\nIn [24]: x_reshape.shape, x_reshape.ndim, x_reshape.T\\nOut[24]: ((2, 3),\\n2,\\narray([[5, 4],\\n[2, 5],\\n[3, 6]]))\\nNotice that the three individual outputs (2,3), 2, and array([[5, 4],[2,\\n5], [3,6]]) are themselves output as a tuple.\\nWe will often want to apply functions to arrays. For instance, we can\\ncompute the square root of the entries using the np.sqrt() function:\\nnp.sqrt()\\nIn [25]: np.sqrt(x)\\nOut[25]: array([2.24, 1.41, 1.73, 2., 2.24, 2.45])\\nWe can also square the elements:\\nIn [26]: x**2\\nOut[26]: array([25, 4, 9, 16, 25, 36])\\nWe can compute the square roots using the same notation, raising to the\\npower of 1/2 instead of 2.\\nIn [27]: x**0.5\\nOut[27]: array([2.24, 1.41, 1.73, 2., 2.24, 2.45])\\nThroughout this book, we will often want to generate random data. The\\nnp.random.normal() function generates a vector of random normal variables. np.random.\\nnormal()\\nWe can learn more about this function by looking at the help page, via a call\\nto np.random.normal?. The first line of the help page reads normal(loc=0.0,\\nscale=1.0, size=None). This signature line tells us that the function’s ar- signature 46\\n2. Statistical Learning\\nguments are loc, scale, and size. These are keyword arguments, which keyword\\nmeans that when they are passed into the function, they can be referred\\nto by name (in any order).3 By default, this function will generate random\\nnormal variable(s) with mean (loc) 0 and standard deviation (scale) 1; fur-\\nthermore, a single random variable will be generated unless the argument\\nto size is changed.\\nWe now generate 50 independent random variables from a N(0, 1) dis-\\ntribution.\\nIn [28]: x = np.random.normal(size=50)\\nx\\nOut[28]: array([-1.19,\\n0.41,\\n0.9 , -0.44, -0.9 , -0.38,\\n0.13,\\n1.87,\\n-0.35,\\n1.16,\\n0.79, -0.97, -1.21,\\n0.06, -1.62, -0.6 ,\\n-0.77, -2.12,\\n0.38, -1.22, -0.06, -1.97, -1.74, -0.56,\\n1.7 , -0.95,\\n0.56,\\n0.35,\\n0.87,\\n0.88, -1.66, -0.32,\\n-0.3 , -1.36,\\n0.92, -0.31,\\n1.28, -1.94,\\n1.07,\\n0.07,\\n0.79, -0.46,\\n2.19, -0.27, -0.64,\\n0.85,\\n0.13,\\n0.46,\\n-0.09,\\n0.7 ])\\nWe create an array y by adding an independent N(50, 1) random variable\\nto each element of x.\\nIn [29]: y = x + np.random.normal(loc=50, scale=1, size=50)\\nThe np.corrcoef() function computes the correlation matrix between x and np.corrcoef()\\ny. The off-diagonal elements give the correlation between x and y.\\nIn [30]: np.corrcoef(x, y)\\nOut[30]: array([[1.\\n, 0.69],\\n[0.69, 1.\\n]])\\nIf you’re following along in your own Jupyter notebook, then you probably\\nnoticed that you got a different set of results when you ran the past few\\ncommands. In particular, each time we call np.random.normal(), we will get\\na different answer, as shown in the following example.\\nIn [31]: print(np.random.normal(scale=5, size=2))\\nprint(np.random.normal(scale=5, size=2))\\nOut[31]: [4.28 2.59]\\n[4.62 -2.54]\\nIn order to ensure that our code provides exactly the same results each\\ntime it is run, we can set a random seed using the np.random.default_rng() random seed\\nnp.random.\\ndefault_rng()\\nfunction. This function takes an arbitrary, user-specified integer argument.\\nIf we set a random seed before generating random data, then re-running\\nour code will yield the same results. The object rng has essentially all the\\n3Python also uses positional arguments. Positional arguments do not need to use a\\nkeyword. To see an example, type in np.sum?. We see that a is a positional argument,\\ni.e. this function assumes that the first unnamed argument that it receives is the array\\nto be summed. By contrast, axis and dtype are keyword arguments: the position in\\nwhich these arguments are entered into np.sum() does not matter. 2.3 Lab: Introduction to Python\\n47\\nrandom number generating methods found in np.random. Hence, to generate\\nnormal data we use rng.normal().\\nIn [32]: rng = np.random.default_rng(1303)\\nprint(rng.normal(scale=5, size=2))\\nrng2 = np.random.default_rng(1303)\\nprint(rng2.normal(scale=5, size=2))\\nOut[32]: [4.09\\n-1.07 ]\\n[4.09\\n-1.07 ]\\nThroughout the labs in this book, we use np.random.default_rng() when-\\never we perform calculations involving random quantities within numpy. In\\nprinciple, this should enable the reader to exactly reproduce the stated\\nresults. However, as new versions of numpy become available, it is possible\\nthat some small discrepancies may occur between the output in the labs\\nand the output from numpy.\\nThe np.mean(), np.var(), and np.std() functions can be used to compute np.mean()\\nnp.var()\\nnp.std()\\nthe mean, variance, and standard deviation of arrays. These functions are\\nalso available as methods on the arrays.\\nIn [33]: rng = np.random.default_rng(3)\\ny = rng.standard_normal(10)\\nnp.mean(y), y.mean()\\nOut[33]: (-0.11, -0.11)\\nIn [34]: np.var(y), y.var(), np.mean((y - y.mean())**2)\\nOut[34]: (2.72, 2.72, 2.72)\\nNotice that by default np.var() divides by the sample size n rather than\\nn −1; see the ddof argument in np.var?.\\nIn [35]: np.sqrt(np.var(y)), np.std(y)\\nOut[35]: (1.65, 1.65)\\nThe np.mean(), np.var(), and np.std() functions can also be applied to the\\nrows and columns of a matrix. To see this, we construct a 10 × 3 matrix of\\nN(0, 1) random variables, and consider computing its row sums.\\nIn [36]: X = rng.standard_normal((10, 3))\\nX\\nOut[36]: array([[ 0.23, -0.35, -0.28],\\n[-0.67, -1.06, -0.39],\\n[ 0.48, -0.24,\\n0.96],\\n[-0.2 ,\\n0.02,\\n1.55],\\n[ 0.55, -0.51, -0.18],\\n[ 0.54,\\n1.94, -0.27],\\n[-0.24,\\n1.\\n, -0.89],\\n[-0.29,\\n0.88,\\n0.58],\\n[ 0.09,\\n0.67, -2.83],\\n[ 1.02, -0.96, -1.67]]) 48\\n2. Statistical Learning\\nSince arrays are row-major ordered, the first axis, i.e. axis=0, refers to its\\nrows. We pass this argument into the mean() method for the object X.\\n.mean()\\nIn [37]: X.mean(axis=0)\\nOut[37]: array([0.15, 0.14, -0.34])\\nThe following yields the same result.\\nIn [38]: X.mean(0)\\nOut[38]: array([0.15, 0.14, -0.34])\\n2.3.4\\nGraphics\\nIn Python, common practice is to use the library matplotlib for graphics. matplotlib\\nHowever, since Python was not written with data analysis in mind, the no-\\ntion of plotting is not intrinsic to the language. We will use the subplots()\\nfunction from matplotlib.pyplot to create a figure and the axes onto which\\nwe plot our data. For many more examples of how to make plots in Python,\\nreaders are encouraged to visit matplotlib.org/stable/gallery/.\\nIn matplotlib, a plot consists of a figure and one or more axes. You can figure\\naxes\\nthink of the figure as the blank canvas upon which one or more plots will\\nbe displayed: it is the entire plotting window. The axes contain important\\ninformation about each plot, such as its x- and y-axis labels, title, and\\nmore. (Note that in matplotlib, the word axes is not the plural of axis: a\\nplot’s axes contains much more information than just the x-axis and the\\ny-axis.)\\nWe begin by importing the subplots() function from matplotlib. We subplots()\\nuse this function throughout when creating figures. The function returns\\na tuple of length two: a figure object as well as the relevant axes object.\\nWe will typically pass figsize as a keyword argument. Having created our\\naxes, we attempt our first plot using its plot() method. To learn more .plot()\\nabout it, type ax.plot?.\\nIn [39]: from matplotlib.pyplot import subplots\\nfig, ax = subplots(figsize=(8, 8))\\nx = rng.standard_normal(100)\\ny = rng.standard_normal(100)\\nax.plot(x, y);\\nWe pause here to note that we have unpacked the tuple of length two re-\\nturned by subplots() into the two distinct variables fig and ax. Unpacking\\nis typically preferred to the following equivalent but slightly more verbose\\ncode:\\nIn [40]: output = subplots(figsize=(8, 8))\\nfig = output[0]\\nax = output[1]\\nWe see that our earlier cell produced a line plot, which is the default.\\nTo create a scatterplot, we provide an additional argument to ax.plot(),\\nindicating that circles should be displayed. 2.3 Lab: Introduction to Python\\n49\\nIn [41]: fig, ax = subplots(figsize=(8, 8))\\nax.plot(x, y, \\'o\\');\\nDifferent values of this additional argument can be used to produce different\\ncolored lines as well as different linestyles.\\nAs an alternative, we could use the ax.scatter() function to create a .scatter()\\nscatterplot.\\nIn [42]: fig, ax = subplots(figsize=(8, 8))\\nax.scatter(x, y, marker=\\'o\\');\\nNotice that in the code blocks above, we have ended the last line with a\\nsemicolon. This prevents ax.plot(x, y) from printing text to the notebook.\\nHowever, it does not prevent a plot from being produced. If we omit the\\ntrailing semi-colon, then we obtain the following output:\\nIn [43]: fig, ax = subplots(figsize=(8, 8))\\nax.scatter(x, y, marker=\\'o\\')\\nOut[43]: <matplotlib.collections.PathCollection at 0x7fb3d9c8f310 >\\nFigure(432x288)\\nIn what follows, we will use trailing semicolons whenever the text that\\nwould be output is not germane to the discussion at hand.\\nTo label our plot, we make use of the set_xlabel(), set_ylabel(), and .set_xlabel()\\n.set_ylabel()\\nset_title() methods of ax.\\n.set_title()\\nIn [44]: fig, ax = subplots(figsize=(8, 8))\\nax.scatter(x, y, marker=\\'o\\')\\nax.set_xlabel(\"this is the x-axis\")\\nax.set_ylabel(\"this is the y-axis\")\\nax.set_title(\"Plot of X vs Y\");\\nHaving access to the figure object fig itself means that we can go in and\\nchange some aspects and then redisplay it. Here, we change the size from\\n(8, 8) to (12, 3).\\nfig.set_size_inches(12,3)\\nfig\\nOccasionally we will want to create several plots within a figure. This\\ncan be achieved by passing additional arguments to subplots(). Below, we\\ncreate a 2 × 3 grid of plots in a figure of size determined by the figsize\\nargument. In such situations, there is often a relationship between the\\naxes in the plots. For example, all plots may have a common x-axis. The\\nsubplots() function can automatically handle this situation when passed\\nthe keyword argument sharex=True. The axes object below is an array\\npointing to different plots in the figure.\\nIn [45]: fig, axes = subplots(nrows=2,\\nncols=3,\\nfigsize=(15, 5))\\nWe now produce a scatter plot with \\'o\\' in the second column of the first\\nrow and a scatter plot with \\'+\\' in the third column of the second row. 50\\n2. Statistical Learning\\nIn [46]: axes[0,1].plot(x, y, \\'o\\')\\naxes[1,2].scatter(x, y, marker=\\'+\\')\\nfig\\nType subplots? to learn more about subplots().\\nTo save the output of fig, we call its savefig() method. The argument .savefig()\\ndpi is the dots per inch, used to determine how large the figure will be in\\npixels.\\nIn [47]: fig.savefig(\"Figure.png\", dpi=400)\\nfig.savefig(\"Figure.pdf\", dpi=200);\\nWe can continue to modify fig using step-by-step updates; for example,\\nwe can modify the range of the x-axis, re-save the figure, and even re-display\\nit.\\nIn [48]: axes[0,1].set_xlim([-1,1])\\nfig.savefig(\"Figure_updated.jpg\")\\nfig\\nWe now create some more sophisticated plots. The ax.contour() method .contour()\\nproduces a contour plot in order to represent three-dimensional data, similar contour plot\\nto a topographical map. It takes three arguments:\\n• A vector of x values (the first dimension),\\n• A vector of y values (the second dimension), and\\n• A matrix whose elements correspond to the z value (the third dimen-\\nsion) for each pair of (x,y) coordinates.\\nTo create x and y, we’ll use the command\\nnp.linspace(a, b, n), which np.linspace()\\nreturns a vector of n numbers starting at a and ending at b.\\nIn [49]: fig, ax = subplots(figsize=(8, 8))\\nx = np.linspace(-np.pi, np.pi, 50)\\ny = x\\nf = np.multiply.outer(np.cos(y), 1 / (1 + x**2))\\nax.contour(x, y, f);\\nWe can increase the resolution by adding more levels to the image.\\nIn [50]: fig, ax = subplots(figsize=(8, 8))\\nax.contour(x, y, f, levels=45);\\nTo fine-tune the output of the ax.contour() function, take a look at the\\nhelp file by typing ?plt.contour.\\nThe ax.imshow() method is similar to ax.contour(), except that it pro- .imshow()\\nduces a color-coded plot whose colors depend on the z value. This is known\\nas a heatmap, and is sometimes used to plot temperature in weather fore- heatmap\\ncasts.\\nIn [51]: fig, ax = subplots(figsize=(8, 8))\\nax.imshow(f); 2.3 Lab: Introduction to Python\\n51\\n2.3.5\\nSequences and Slice Notation\\nAs seen above, the function np.linspace() can be used to create a sequence\\nof numbers.\\nIn [52]: seq1 = np.linspace(0, 10, 11)\\nseq1\\nOut[52]: array([ 0.,\\n1.,\\n2.,\\n3.,\\n4.,\\n5.,\\n6.,\\n7.,\\n8.,\\n9., 10.])\\nThe function np.arange() returns a sequence of numbers spaced out by np.arange()\\nstep. If step is not specified, then a default value of 1 is used. Let’s create\\na sequence that starts at 0 and ends at 10.\\nIn [53]: seq2 = np.arange(0, 10)\\nseq2\\nOut[53]: array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])\\nWhy isn’t 10 output above? This has to do with slice notation in Python. slice\\nSlice notation is used to index sequences such as lists, tuples and arrays.\\nSuppose we want to retrieve the fourth through sixth (inclusive) entries of\\na string. We obtain a slice of the string using the indexing notation [3:6].\\nIn [54]: \"hello world\"[3:6]\\nOut[54]: \\'lo \\'\\nIn the code block above, the notation 3:6 is shorthand for slice(3,6) when\\nused inside [].\\nIn [55]: \"hello world\"[slice(3,6)]\\nOut[55]: \\'lo \\'\\nYou might have expected slice(3,6) to output the fourth through seventh\\ncharacters in the text string (recalling that Python begins its indexing at\\nzero), but instead it output the fourth through sixth. This also explains\\nwhy the earlier np.arange(0, 10) command output only the integers from\\n0 to 9. See the documentation slice? for useful options in creating slices.\\n2.3.6\\nIndexing Data\\nTo begin, we create a two-dimensional numpy array.\\nIn [56]: A = np.array(np.arange(16)).reshape((4, 4))\\nA\\nOut[56]: array([[ 0,\\n1,\\n2,\\n3],\\n[ 4,\\n5,\\n6,\\n7],\\n[ 8,\\n9, 10, 11],\\n[12, 13, 14, 15]])\\nTyping A[1,2] retrieves the element corresponding to the second row and\\nthird column. (As usual, Python indexes from 0.) 52\\n2. Statistical Learning\\nIn [57]: A[1,2]\\nOut[57]: 6\\nThe first number after the open-bracket symbol [ refers to the row, and\\nthe second number refers to the column.\\nIndexing Rows, Columns, and Submatrices\\nTo select multiple rows at a time, we can pass in a list specifying our\\nselection. For instance, [1,3] will retrieve the second and fourth rows:\\nIn [58]: A[[1,3]]\\nOut[58]: array([[ 4,\\n5,\\n6,\\n7],\\n[12, 13, 14, 15]])\\nTo select the first and third columns, we pass in [0,2] as the second ar-\\ngument in the square brackets. In this case we need to supply the first\\nargument : which selects all rows.\\nIn [59]: A[:,[0,2]]\\nOut[59]: array([[ 0,\\n2],\\n[ 4,\\n6],\\n[ 8, 10],\\n[12, 14]])\\nNow, suppose that we want to select the submatrix made up of the second\\nand fourth rows as well as the first and third columns. This is where index-\\ning gets slightly tricky. It is natural to try to use lists to retrieve the rows\\nand columns:\\nIn [60]: A[[1,3],[0,2]]\\nOut[60]: array([ 4, 14])\\nOops — what happened? We got a one-dimensional array of length two\\nidentical to\\nIn [61]: np.array([A[1,0],A[3,2]])\\nOut[61]: array([ 4, 14])\\nSimilarly, the following code fails to extract the submatrix comprised of\\nthe second and fourth rows and the first, third, and fourth columns:\\nIn [62]: A[[1,3],[0,2,3]]\\nIndexError: shape mismatch: indexing arrays could not be broadcast\\ntogether with shapes (2,) (3,) 2.3 Lab: Introduction to Python\\n53\\nWe can see what has gone wrong here. When supplied with two indexing\\nlists, the numpy interpretation is that these provide pairs of i, j indices for\\na series of entries. That is why the pair of lists must have the same length.\\nHowever, that was not our intent, since we are looking for a submatrix.\\nOne easy way to do this is as follows. We first create a submatrix by\\nsubsetting the rows of A, and then on the fly we make a further submatrix\\nby subsetting its columns.\\nIn [63]: A[[1,3]][:,[0,2]]\\nOut[63]: array([[ 4,\\n6],\\n[12, 14]])\\nThere are more efficient ways of achieving the same result.\\nThe convenience function np.ix_() allows us to extract a submatrix using convenience\\nfunction\\nnp.ix_()\\nlists, by creating an intermediate mesh object.\\nmesh\\nIn [64]: idx = np.ix_([1,3],[0,2,3])\\nA[idx]\\nOut[64]: array([[ 4,\\n6,\\n7],\\n[12, 14, 15]])\\nAlternatively, we can subset matrices efficiently using slices. The slice\\n1:4:2 captures the second and fourth items of a sequence, while the slice\\n0:3:2 captures the first and third items (the third element in a slice se-\\nquence is the step size).\\nIn [65]: A[1:4:2,0:3:2]\\nOut[65]: array([[ 4,\\n6],\\n[12, 14]])\\nWhy are we able to retrieve a submatrix directly using slices but not using\\nlists? Its because they are different Python types, and are treated differently\\nby numpy. Slices can be used to extract objects from arbitrary sequences,\\nsuch as strings, lists, and tuples, while the use of lists for indexing is more\\nlimited.\\nBoolean Indexing\\nIn numpy, a Boolean is a type that equals either True or False (also rep- Boolean\\nresented as 1 and 0, respectively). The next line creates a vector of 0’s,\\nrepresented as Booleans, of length equal to the first dimension of A.\\nIn [66]: keep_rows = np.zeros(A.shape[0], bool)\\nkeep_rows\\nOut[66]: array([False,\\nFalse, False,\\nFalse])\\nWe now set two of the elements to True.\\nIn [67]: keep_rows[[1,3]] = True\\nkeep_rows 54\\n2. Statistical Learning\\nOut[67]: array([False,\\nTrue, False,\\nTrue])\\nNote that the elements of keep_rows, when viewed as integers, are the same\\nas the values of np.array([0,1,0,1]). Below, we use == to verify their equal-\\nity. When applied to two arrays, the == operation is applied elementwise.\\nIn [68]: np.all(keep_rows == np.array([0,1,0,1]))\\nOut[68]: True\\n(Here, the function np.all() has checked whether all entries of an array np.all()\\nare True. A similar function, np.any(), can be used to check whether any np.any()\\nentries of an array are True.)\\nHowever, even though np.array([0,1,0,1]) and keep_rows are equal ac-\\ncording to ==, they index different sets of rows! The former retrieves the\\nfirst, second, first, and second rows of A.\\nIn [69]: A[np.array([0,1,0,1])]\\nOut[69]: array([[0, 1, 2, 3],\\n[4, 5, 6, 7],\\n[0, 1, 2, 3],\\n[4, 5, 6, 7]])\\nBy contrast, keep_rows retrieves only the second and fourth rows of A —\\ni.e. the rows for which the Boolean equals TRUE.\\nIn [70]: A[keep_rows]\\nOut[70]: array([[ 4,\\n5,\\n6,\\n7],\\n[12, 13, 14, 15]])\\nThis example shows that Booleans and integers are treated differently by\\nnumpy.\\nWe again make use of the np.ix_() function to create a mesh containing\\nthe second and fourth rows, and the first, third, and fourth columns. This\\ntime, we apply the function to Booleans, rather than lists.\\nIn [71]: keep_cols = np.zeros(A.shape[1], bool)\\nkeep_cols[[0, 2, 3]] = True\\nidx_bool = np.ix_(keep_rows, keep_cols)\\nA[idx_bool]\\nOut[71]: array([[ 4,\\n6, 7],\\n[12, 14, 15]])\\nWe can also mix a list with an array of Booleans in the arguments to\\nnp.ix_():\\nIn [72]: idx_mixed = np.ix_([1,3], keep_cols)\\nA[idx_mixed]\\nOut[72]: array([[ 4,\\n6, 7],\\n[12, 14, 15]])\\nFor more details on indexing in numpy, readers are referred to the numpy\\ntutorial mentioned earlier. 2.3 Lab: Introduction to Python\\n55\\n2.3.7\\nLoading Data\\nData sets often contain different types of data, and may have names as-\\nsociated with the rows or columns. For these reasons, they typically are\\nbest accommodated using a data frame. We can think of a data frame as data frame\\na sequence of arrays of identical length; these are the columns. Entries in\\nthe different arrays can be combined to form a row. The pandas library can\\nbe used to create and work with data frame objects.\\nReading in a Data Set\\nThe first step of most analyses involves importing a data set into Python.\\nBefore attempting to load a data set, we must make sure that Python knows\\nwhere to find the file containing it. If the file is in the same location as this\\nnotebook file, then we are all set. Otherwise, the command os.chdir() can os.chdir()\\nbe used to change directory. (You will need to call import os before calling\\nos.chdir().)\\nWe will begin by reading in Auto.csv, available on the book website. This\\nis a comma-separated file, and can be read in using pd.read_csv():\\npd.read_csv()\\nIn [73]: import pandas as pd\\nAuto = pd.read_csv(\\'Auto.csv\\')\\nAuto\\nThe book website also has a whitespace-delimited version of this data,\\ncalled Auto.data. This can be read in as follows:\\nIn [74]: Auto = pd.read_csv(\\'Auto.data\\', delim_whitespace=True)\\nBoth Auto.csv and Auto.data are simply text files. Before loading data into\\nPython, it is a good idea to view it using a text editor or other software,\\nsuch as Microsoft Excel.\\nWe now take a look at the column of Auto corresponding to the variable\\nhorsepower:\\nIn [75]: Auto[\\'horsepower\\']\\nOut[75]: 0\\n130.0\\n1\\n165.0\\n2\\n150.0\\n3\\n150.0\\n4\\n140.0\\n...\\n392\\n86.00\\n393\\n52.00\\n394\\n84.00\\n395\\n79.00\\n396\\n82.00\\nName: horsepower , Length: 397, dtype: object\\nWe see that the dtype of this column is object. It turns out that all values\\nof the horsepower column were interpreted as strings when reading in the\\ndata. We can find out why by looking at the unique values.\\nIn [76]: np.unique(Auto[\\'horsepower\\']) 56\\n2. Statistical Learning\\nTo save space, we have omitted the output of the previous code block. We\\nsee the culprit is the value ?, which is being used to encode missing values.\\nTo fix the problem, we must provide pd.read_csv() with an argument\\ncalled na_values. Now, each instance of ? in the file is replaced with the\\nvalue np.nan, which means not a number:\\nIn [77]: Auto = pd.read_csv(\\'Auto.data\\',\\nna_values=[\\'?\\'],\\ndelim_whitespace=True)\\nAuto[\\'horsepower\\'].sum()\\nOut[77]: 40952.0\\nThe Auto.shape attribute tells us that the data has 397 observations, or\\nrows, and nine variables, or columns.\\nIn [78]: Auto.shape\\nOut[78]: (397, 9)\\nThere are various ways to deal with missing data. In this case, since\\nonly five of the rows contain missing observations, we choose to use the\\nAuto.dropna() method to simply remove these rows.\\n.dropna()\\nIn [79]: Auto_new = Auto.dropna()\\nAuto_new.shape\\nOut[79]: (392, 9)\\nBasics of Selecting Rows and Columns\\nWe can use Auto.columns to check the variable names.\\nIn [80]: Auto = Auto_new # overwrite the previous value\\nAuto.columns\\nOut[80]: Index([\\'mpg\\', \\'cylinders\\', \\'displacement\\', \\'horsepower\\',\\n\\'weight\\', \\'acceleration\\', \\'year\\', \\'origin\\', \\'name\\'],\\ndtype=\\'object\\')\\nAccessing the rows and columns of a data frame is similar, but not iden-\\ntical, to accessing the rows and columns of an array. Recall that the first\\nargument to the [] method is always applied to the rows of the array. Sim-\\nilarly, passing in a slice to the [] method creates a data frame whose rows\\nare determined by the slice:\\nIn [81]: Auto[:3]\\nOut[81]:\\nmpg\\ncylinders\\ndisplacement\\nhorsepower\\nweight\\n...\\n0\\n18.0\\n8\\n307.0\\n130.0\\n3504.0\\n...\\n1\\n15.0\\n8\\n350.0\\n165.0\\n3693.0\\n...\\n2\\n18.0\\n8\\n318.0\\n150.0\\n3436.0\\n...\\nSimilarly, an array of Booleans can be used to subset the rows: 2.3 Lab: Introduction to Python\\n57\\nIn [82]: idx_80 = Auto[\\'year\\'] > 80\\nAuto[idx_80]\\nHowever, if we pass in a list of strings to the [] method, then we obtain a\\ndata frame containing the corresponding set of columns.\\nIn [83]: Auto[[\\'mpg\\', \\'horsepower\\']]\\nOut[83]:\\nmpg\\nhorsepower\\n0\\n18.0\\n130.0\\n1\\n15.0\\n165.0\\n2\\n18.0\\n150.0\\n3\\n16.0\\n150.0\\n4\\n17.0\\n140.0\\n...\\n...\\n...\\n392\\n27.0\\n86.0\\n393\\n44.0\\n52.0\\n394\\n32.0\\n84.0\\n395\\n28.0\\n79.0\\n396\\n31.0\\n82.0\\n392 rows x 2 columns\\nSince we did not specify an index column when we loaded our data frame,\\nthe rows are labeled using integers 0 to 396.\\nIn [84]: Auto.index\\nOut[84]: Int64Index([\\n0,\\n1,\\n2,\\n3,\\n4,\\n5,\\n6,\\n7,\\n8,\\n9,\\n...\\n387, 388, 389, 390, 391, 392, 393, 394, 395, 396],\\ndtype=\\'int64\\', length=392)\\nWe can use the set_index() method to re-name the rows using the contents .set_index()\\nof Auto[\\'name\\'].\\nIn [85]: Auto_re = Auto.set_index(\\'name\\')\\nAuto_re\\nOut[85]:\\nmpg\\ncylinders\\ndisplacement\\n...\\nname\\nchevrolet chevelle malibu\\n18.0\\n8\\n307.0\\n...\\nbuick skylark 32\\n15.0\\n8\\n350.0\\n...\\nplymouth satellite\\n18.0\\n8\\n318.0\\n...\\namc rebel sst\\n16.0\\n8\\n304.0\\n...\\nIn [86]: Auto_re.columns\\nOut[86]: Index([\\'mpg\\', \\'cylinders\\', \\'displacement\\', \\'horsepower\\',\\n\\'weight\\', \\'acceleration\\', \\'year\\', \\'origin\\'],\\ndtype=\\'object\\')\\nWe see that the column \\'name\\' is no longer there.\\nNow that the index has been set to name, we can access rows of the data\\nframe by name using the loc[] method of Auto:\\n.loc[] 58\\n2. Statistical Learning\\nIn [87]: rows = [\\'amc rebel sst\\', \\'ford torino\\']\\nAuto_re.loc[rows]\\nOut[87]:\\nmpg\\ncylinders\\ndisplacement horsepower\\n...\\nname\\namc rebel sst\\n16.0\\n8\\n304.0\\n150.0\\n...\\nford torino\\n17.0\\n8\\n302.0\\n140.0\\n...\\nAs an alternative to using the index name, we could retrieve the 4th and\\n5th rows of Auto using the iloc[] method:\\n.iloc[]\\nIn [88]: Auto_re.iloc[[3,4]]\\nWe can also use it to retrieve the 1st, 3rd and and 4th columns of Auto_re:\\nIn [89]: Auto_re.iloc[:,[0,2,3]]\\nWe can extract the 4th and 5th rows, as well as the 1st, 3rd and 4th\\ncolumns, using a single call to iloc[]:\\nIn [90]: Auto_re.iloc[[3,4],[0,2,3]]\\nOut[90]:\\nmpg\\ndisplacement\\nhorsepower\\nname\\namc rebel sst\\n16.0\\n304.0\\n150.0\\nford torino\\n17.0\\n302.0\\n140.0\\nIndex entries need not be unique: there are several cars in the data frame\\nnamed ford galaxie 500.\\nIn [91]: Auto_re.loc[\\'ford galaxie 500\\', [\\'mpg\\', \\'origin\\']]\\nOut[91]:\\nmpg\\norigin\\nname\\nford galaxie 500\\n15.0\\n1\\nford galaxie 500\\n14.0\\n1\\nford galaxie 500\\n14.0\\n1\\nMore on Selecting Rows and Columns\\nSuppose now that we want to create a data frame consisting of the weight\\nand origin of the subset of cars with year greater than 80 — i.e. those\\nbuilt after 1980. To do this, we first create a Boolean array that indexes\\nthe rows. The loc[] method allows for Boolean entries as well as strings:\\nIn [92]: idx_80 = Auto_re[\\'year\\'] > 80\\nAuto_re.loc[idx_80, [\\'weight\\', \\'origin\\']]\\nTo do this more concisely, we can use an anonymous function called a\\nlambda:\\nlambda\\nIn [93]: Auto_re.loc[lambda df: df[\\'year\\'] > 80, [\\'weight\\', \\'origin\\']]\\nThe lambda call creates a function that takes a single argument, here df,\\nand returns df[\\'year\\']>80. Since it is created inside the loc[] method for 2.3 Lab: Introduction to Python\\n59\\nthe dataframe Auto_re, that dataframe will be the argument supplied. As\\nanother example of using a lambda, suppose that we want all cars built after\\n1980 that achieve greater than 30 miles per gallon:\\nIn [94]: Auto_re.loc[lambda df: (df[\\'year\\'] > 80) & (df[\\'mpg\\'] > 30),\\n[\\'weight\\', \\'origin\\']\\n]\\nThe symbol & computes an element-wise and operation. As another ex-\\nample, suppose that we want to retrieve all Ford and Datsun cars with\\ndisplacement less than 300. We check whether each name entry contains\\neither the string ford or datsun using the str.contains() method of the .str.\\ncontains()\\nindex attribute of of the dataframe:\\nIn [95]: Auto_re.loc[lambda df: (df[\\'displacement\\'] < 300)\\n& (df.index.str.contains(\\'ford\\')\\n| df.index.str.contains(\\'datsun\\')),\\n[\\'weight\\', \\'origin\\']\\n]\\nHere, the symbol | computes an element-wise or operation.\\nIn summary, a powerful set of operations is available to index the rows\\nand columns of data frames. For integer based queries, use the iloc[]\\nmethod. For string and Boolean selections, use the loc[] method. For\\nfunctional queries that filter rows, use the loc[] method with a function\\n(typically a lambda) in the rows argument.\\n2.3.8\\nFor Loops\\nA for loop is a standard tool in many languages that repeatedly evaluates for\\nsome chunk of code while varying different values inside the code. For\\nexample, suppose we loop over elements of a list and compute their sum.\\nIn [96]: total = 0\\nfor value in [3,2,19]:\\ntotal += value\\nprint(\\'Total is: {0}\\'.format(total))\\nTotal is: 24\\nThe indented code beneath the line with the for statement is run for each\\nvalue in the sequence specified in the for statement. The loop ends either\\nwhen the cell ends or when code is indented at the same level as the original\\nfor statement. We see that the final line above which prints the total is\\nexecuted only once after the for loop has terminated. Loops can be nested\\nby additional indentation.\\nIn [97]: total = 0\\nfor value in [2,3,19]:\\nfor weight in [3, 2, 1]:\\ntotal += value * weight\\nprint(\\'Total is: {0}\\'.format(total))\\nTotal is: 144 60\\n2. Statistical Learning\\nAbove, we summed over each combination of value and weight. We also\\ntook advantage of the increment notation in Python: the expression a += b increment\\nis equivalent to a = a + b. Besides being a convenient notation, this can\\nsave time in computationally heavy tasks in which the intermediate value\\nof a+b need not be explicitly created.\\nPerhaps a more common task would be to sum over (value, weight)\\npairs. For instance, to compute the average value of a random variable\\nthat takes on possible values 2, 3 or 19 with probability 0.2, 0.3, 0.5 respec-\\ntively we would compute the weighted sum. Tasks such as this can often be\\naccomplished using the zip() function that loops over a sequence of tuples. zip()\\nIn [98]: total = 0\\nfor value, weight in zip([2,3,19],\\n[0.2,0.3,0.5]):\\ntotal += weight * value\\nprint(\\'Weighted average is: {0}\\'.format(total))\\nWeighted average is: 10.8\\nString Formatting\\nIn the code chunk above we also printed a string displaying the total.\\nHowever, the object total is an integer and not a string. Inserting the\\nvalue of something into a string is a common task, made simple using some\\nof the powerful string formatting tools in Python. Many data cleaning tasks\\ninvolve manipulating and programmatically producing strings.\\nFor example we may want to loop over the columns of a data frame\\nand print the percent missing in each column. Let’s create a data frame\\nD with columns in which 20% of the entries are missing i.e. set to np.nan. np.nan\\nWe’ll create the values in D from a normal distribution with mean 0 and\\nvariance 1 using rng.standard_normal() and then overwrite some random\\nentries using rng.choice().\\nIn [99]: rng = np.random.default_rng(1)\\nA = rng.standard_normal((127, 5))\\nM = rng.choice([0, np.nan], p=[0.8,0.2], size=A.shape)\\nA += M\\nD = pd.DataFrame(A, columns=[\\'food\\',\\n\\'bar\\',\\n\\'pickle\\',\\n\\'snack\\',\\n\\'popcorn\\'])\\nD[:3]\\nOut[99]:\\nfood\\nbar\\npickle\\nsnack\\npopcorn\\n0\\n0.345584\\n0.821618 0.330437 -1.303157\\nNaN\\n1\\nNaN -0.536953 0.581118\\n0.364572\\n0.294132\\n2\\nNaN\\n0.546713\\nNaN -0.162910 -0.482119\\nIn [100]: for col in D.columns:\\ntemplate = \\'Column \"{0}\" has {1:.2%} missing values\\'\\nprint(template.format(col,\\nnp.isnan(D[col]).mean())) 2.3 Lab: Introduction to Python\\n61\\nColumn \"food\" has 16.54% missing values\\nColumn \"bar\" has 25.98% missing values\\nColumn \"pickle\" has 29.13% missing values\\nColumn \"snack\" has 21.26% missing values\\nColumn \"popcorn\" has 22.83% missing values\\nWe see that the template.format() method expects two arguments {0} and\\n{1:.2%}, and the latter includes some formatting information. In particular,\\nit specifies that the second argument should be expressed as a percent with\\ntwo decimal digits.\\nThe reference docs.python.org/3/library/string.html includes many help-\\nful and more complex examples.\\n2.3.9\\nAdditional Graphical and Numerical Summaries\\nWe can use the ax.plot() or ax.scatter() functions to display the quan-\\ntitative variables. However, simply typing the variable names will produce\\nan error message, because Python does not know to look in the Auto data\\nset for those variables.\\nIn [101]: fig, ax = subplots(figsize=(8, 8))\\nax.plot(horsepower, mpg, \\'o\\');\\nNameError: name \\'horsepower\\' is not defined\\nWe can address this by accessing the columns directly:\\nIn [102]: fig, ax = subplots(figsize=(8, 8))\\nax.plot(Auto[\\'horsepower\\'], Auto[\\'mpg\\'], \\'o\\');\\nAlternatively, we can use the plot() method with the call Auto.plot(). Us- .plot()\\ning this method, the variables can be accessed by name. The plot methods\\nof a data frame return a familiar object: an axes. We can use it to update\\nthe plot as we did previously:\\nIn [103]: ax = Auto.plot.scatter(\\'horsepower\\', \\'mpg\\');\\nax.set_title(\\'Horsepower vs. MPG\\')\\nIf we want to save the figure that contains a given axes, we can find the\\nrelevant figure by accessing the figure attribute:\\nIn [104]: fig = ax.figure\\nfig.savefig(\\'horsepower_mpg.png\\');\\nWe can further instruct the data frame to plot to a particular axes object.\\nIn this case the corresponding plot() method will return the modified axes\\nwe passed in as an argument. Note that when we request a one-dimensional\\ngrid of plots, the object axes is similarly one-dimensional. We place our\\nscatter plot in the middle plot of a row of three plots within a figure.\\nIn [105]: fig, axes = subplots(ncols=3, figsize=(15, 5))\\nAuto.plot.scatter(\\'horsepower\\', \\'mpg\\', ax=axes[1]);\\nNote also that the columns of a data frame can be accessed as attributes:\\ntry typing in Auto.horsepower. 62\\n2. Statistical Learning\\nWe now consider the cylinders variable. Typing in Auto.cylinders.dtype\\nreveals that it is being treated as a quantitative variable. However, since\\nthere is only a small number of possible values for this variable, we may\\nwish to treat it as qualitative. Below, we replace the cylinders column with\\na categorical version of Auto.cylinders. The function pd.Series() owes its pd.Series()\\nname to the fact that pandas is often used in time series applications.\\nIn [106]: Auto.cylinders = pd.Series(Auto.cylinders, dtype=\\'category\\')\\nAuto.cylinders.dtype\\nNow that cylinders is qualitative, we can display it using the boxplot() .boxplot()\\nmethod.\\nIn [107]: fig, ax = subplots(figsize=(8, 8))\\nAuto.boxplot(\\'mpg\\', by=\\'cylinders\\', ax=ax);\\nThe hist() method can be used to plot a histogram.\\n.hist()\\nIn [108]: fig, ax = subplots(figsize=(8, 8))\\nAuto.hist(\\'mpg\\', ax=ax);\\nThe color of the bars and the number of bins can be changed:\\nIn [109]: fig, ax = subplots(figsize=(8, 8))\\nAuto.hist(\\'mpg\\', color=\\'red\\', bins=12, ax=ax);\\nSee Auto.hist? for more plotting options.\\nWe can use the pd.plotting.scatter_matrix() function to create a scat- pd.plotting.\\nscatter_\\nmatrix()\\nterplot matrix to visualize all of the pairwise relationships between the\\ncolumns in a data frame.\\nIn [110]: pd.plotting.scatter_matrix(Auto);\\nWe can also produce scatterplots for a subset of the variables.\\nIn [111]: pd.plotting.scatter_matrix(Auto[[\\'mpg\\',\\n\\'displacement\\',\\n\\'weight\\']]);\\nThe describe() method produces a numerical summary of each column in .describe()\\na data frame.\\nIn [112]: Auto[[\\'mpg\\', \\'weight\\']].describe()\\nWe can also produce a summary of just a single column.\\nIn [113]: Auto[\\'cylinders\\'].describe()\\nAuto[\\'mpg\\'].describe()\\nTo exit Jupyter, select File / Close and Halt. 2.4 Exercises\\n63\\n2.4\\nExercises\\nConceptual\\n1. For each of parts (a) through (d), indicate whether we would generally\\nexpect the performance of a flexible statistical learning method to be\\nbetter or worse than an inflexible method. Justify your answer.\\n(a) The sample size n is extremely large, and the number of predic-\\ntors p is small.\\n(b) The number of predictors p is extremely large, and the number\\nof observations n is small.\\n(c) The relationship between the predictors and response is highly\\nnon-linear.\\n(d) The variance of the error terms, i.e. σ2 = Var(ϵ), is extremely\\nhigh.\\n2. Explain whether each scenario is a classification or regression prob-\\nlem, and indicate whether we are most interested in inference or pre-\\ndiction. Finally, provide n and p.\\n(a) We collect a set of data on the top 500 firms in the US. For each\\nfirm we record profit, number of employees, industry and the\\nCEO salary. We are interested in understanding which factors\\naffect CEO salary.\\n(b) We are considering launching a new product and wish to know\\nwhether it will be a success or a failure. We collect data on 20\\nsimilar products that were previously launched. For each prod-\\nuct we have recorded whether it was a success or failure, price\\ncharged for the product, marketing budget, competition price,\\nand ten other variables.\\n(c) We are interested in predicting the % change in the USD/Euro\\nexchange rate in relation to the weekly changes in the world\\nstock markets. Hence we collect weekly data for all of 2012. For\\neach week we record the % change in the USD/Euro, the %\\nchange in the US market, the % change in the British market,\\nand the % change in the German market.\\n3. We now revisit the bias-variance decomposition.\\n(a) Provide a sketch of typical (squared) bias, variance, training er-\\nror, test error, and Bayes (or irreducible) error curves, on a sin-\\ngle plot, as we go from less flexible statistical learning methods\\ntowards more flexible approaches. The x-axis should represent\\nthe amount of flexibility in the method, and the y-axis should\\nrepresent the values for each curve. There should be five curves.\\nMake sure to label each one.\\n(b) Explain why each of the five curves has the shape displayed in\\npart (a). 64\\n2. Statistical Learning\\n4. You will now think of some real-life applications for statistical learn-\\ning.\\n(a) Describe three real-life applications in which classification might\\nbe useful. Describe the response, as well as the predictors. Is the\\ngoal of each application inference or prediction? Explain your\\nanswer.\\n(b) Describe three real-life applications in which regression might\\nbe useful. Describe the response, as well as the predictors. Is the\\ngoal of each application inference or prediction? Explain your\\nanswer.\\n(c) Describe three real-life applications in which cluster analysis\\nmight be useful.\\n5. What are the advantages and disadvantages of a very flexible (versus\\na less flexible) approach for regression or classification? Under what\\ncircumstances might a more flexible approach be preferred to a less\\nflexible approach? When might a less flexible approach be preferred?\\n6. Describe the differences between a parametric and a non-parametric\\nstatistical learning approach. What are the advantages of a para-\\nmetric approach to regression or classification (as opposed to a non-\\nparametric approach)? What are its disadvantages?\\n7. The table below provides a training data set containing six observa-\\ntions, three predictors, and one qualitative response variable.\\nObs.\\nX1\\nX2\\nX3\\nY\\n1\\n0\\n3\\n0\\nRed\\n2\\n2\\n0\\n0\\nRed\\n3\\n0\\n1\\n3\\nRed\\n4\\n0\\n1\\n2\\nGreen\\n5\\n−1\\n0\\n1\\nGreen\\n6\\n1\\n1\\n1\\nRed\\nSuppose we wish to use this data set to make a prediction for Y when\\nX1 = X2 = X3 = 0 using K-nearest neighbors.\\n(a) Compute the Euclidean distance between each observation and\\nthe test point, X1 = X2 = X3 = 0.\\n(b) What is our prediction with K = 1? Why?\\n(c) What is our prediction with K = 3? Why?\\n(d) If the Bayes decision boundary in this problem is highly non-\\nlinear, then would we expect the best value for K to be large or\\nsmall? Why? 2.4 Exercises\\n65\\nApplied\\n8. This exercise relates to the College data set, which can be found in\\nthe file College.csv on the book website. It contains a number of\\nvariables for 777 different universities and colleges in the US. The\\nvariables are\\n• Private : Public/private indicator\\n• Apps : Number of applications received\\n• Accept : Number of applicants accepted\\n• Enroll : Number of new students enrolled\\n• Top10perc : New students from top 10 % of high school class\\n• Top25perc : New students from top 25 % of high school class\\n• F.Undergrad : Number of full-time undergraduates\\n• P.Undergrad : Number of part-time undergraduates\\n• Outstate : Out-of-state tuition\\n• Room.Board : Room and board costs\\n• Books : Estimated book costs\\n• Personal : Estimated personal spending\\n• PhD : Percent of faculty with Ph.D.s\\n• Terminal : Percent of faculty with terminal degree\\n• S.F.Ratio : Student/faculty ratio\\n• perc.alumni : Percent of alumni who donate\\n• Expend : Instructional expenditure per student\\n• Grad.Rate : Graduation rate\\nBefore reading the data into Python, it can be viewed in Excel or a\\ntext editor.\\n(a) Use the pd.read_csv() function to read the data into Python. Call\\nthe loaded data college. Make sure that you have the directory\\nset to the correct location for the data.\\n(b) Look at the data used in the notebook by creating and running\\na new cell with just the code college in it. You should notice\\nthat the first column is just the name of each university in a\\ncolumn named something like Unnamed: 0. We don’t really want\\npandas to treat this as data. However, it may be handy to have\\nthese names for later. Try the following commands and similarly\\nlook at the resulting data frames:\\ncollege2 = pd.read_csv(\\'College.csv\\', index_col=0)\\ncollege3 = college.rename({\\'Unnamed: 0\\': \\'College\\'},\\naxis=1)\\ncollege3 = college3.set_index(\\'College\\') 66\\n2. Statistical Learning\\nThis has used the first column in the file as an index for the\\ndata frame. This means that pandas has given each row a name\\ncorresponding to the appropriate university. Now you should see\\nthat the first data column is Private. Note that the names of\\nthe colleges appear on the left of the table. We also introduced\\na new python object above: a dictionary, which is specified by dictionary\\n(key, value) pairs. Keep your modified version of the data with\\nthe following:\\ncollege = college3\\n(c) Use the describe() method of to produce a numerical summary\\nof the variables in the data set.\\n(d) Use the pd.plotting.scatter_matrix() function to produce a\\nscatterplot matrix of the first columns [Top10perc, Apps, Enroll].\\nRecall that you can reference a list C of columns of a data frame\\nA using A[C].\\n(e) Use the boxplot() method of college to produce side-by-side\\nboxplots of Outstate versus Private.\\n(f) Create a new qualitative variable, called Elite, by binning the\\nTop10perc variable into two groups based on whether or not the\\nproportion of students coming from the top 10% of their high\\nschool classes exceeds 50%.\\ncollege[\\'Elite\\'] = pd.cut(college[\\'Top10perc\\'],\\n[0,0.5,1],\\nlabels=[\\'No\\', \\'Yes\\'])\\nUse the value_counts() method of college[\\'Elite\\'] to see how\\nmany elite universities there are. Finally, use the boxplot() method\\nagain to produce side-by-side boxplots of Outstate versus Elite.\\n(g) Use the plot.hist() method of college to produce some his-\\ntograms with differing numbers of bins for a few of the quanti-\\ntative variables. The command plt.subplots(2, 2) may be use-\\nful: it will divide the plot window into four regions so that four\\nplots can be made simultaneously. By changing the arguments\\nyou can divide the screen up in other combinations.\\n(h) Continue exploring the data, and provide a brief summary of\\nwhat you discover.\\n9. This exercise involves the Auto data set studied in the lab. Make sure\\nthat the missing values have been removed from the data.\\n(a) Which of the predictors are quantitative, and which are quali-\\ntative?\\n(b) What is the range of each quantitative predictor? You can an-\\nswer this using the min() and max() methods in numpy.\\n.min()\\n.max()\\n(c) What is the mean and standard deviation of each quantitative\\npredictor? 2.4 Exercises\\n67\\n(d) Now remove the 10th through 85th observations. What is the\\nrange, mean, and standard deviation of each predictor in the\\nsubset of the data that remains?\\n(e) Using the full data set, investigate the predictors graphically,\\nusing scatterplots or other tools of your choice. Create some plots\\nhighlighting the relationships among the predictors. Comment\\non your findings.\\n(f) Suppose that we wish to predict gas mileage (mpg) on the basis\\nof the other variables. Do your plots suggest that any of the\\nother variables might be useful in predicting mpg? Justify your\\nanswer.\\n10. This exercise involves the Boston housing data set.\\n(a) To begin, load in the Boston data set, which is part of the ISLP\\nlibrary.\\n(b) How many rows are in this data set? How many columns? What\\ndo the rows and columns represent?\\n(c) Make some pairwise scatterplots of the predictors (columns) in\\nthis data set. Describe your findings.\\n(d) Are any of the predictors associated with per capita crime rate?\\nIf so, explain the relationship.\\n(e) Do any of the suburbs of Boston appear to have particularly\\nhigh crime rates? Tax rates? Pupil-teacher ratios? Comment on\\nthe range of each predictor.\\n(f) How many of the suburbs in this data set bound the Charles\\nriver?\\n(g) What is the median pupil-teacher ratio among the towns in this\\ndata set?\\n(h) Which suburb of Boston has lowest median value of owner-\\noccupied homes? What are the values of the other predictors\\nfor that suburb, and how do those values compare to the overall\\nranges for those predictors? Comment on your findings.\\n(i) In this data set, how many of the suburbs average more than\\nseven rooms per dwelling? More than eight rooms per dwelling?\\nComment on the suburbs that average more than eight rooms\\nper dwelling. 3\\nLinear Regression\\nThis chapter is about linear regression, a very simple approach for super-\\nvised learning. In particular, linear regression is a useful tool for predicting\\na quantitative response. It has been around for a long time and is the topic\\nof innumerable textbooks. Though it may seem somewhat dull compared to\\nsome of the more modern statistical learning approaches described in later\\nchapters of this book, linear regression is still a useful and widely used sta-\\ntistical learning method. Moreover, it serves as a good jumping-off point for\\nnewer approaches: as we will see in later chapters, many fancy statistical\\nlearning approaches can be seen as generalizations or extensions of linear\\nregression. Consequently, the importance of having a good understanding\\nof linear regression before studying more complex learning methods cannot\\nbe overstated. In this chapter, we review some of the key ideas underlying\\nthe linear regression model, as well as the least squares approach that is\\nmost commonly used to fit this model.\\nRecall the Advertising data from Chapter 2. Figure 2.1 displays sales\\n(in thousands of units) for a particular product as a function of advertis-\\ning budgets (in thousands of dollars) for TV, radio, and newspaper media.\\nSuppose that in our role as statistical consultants we are asked to suggest,\\non the basis of this data, a marketing plan for next year that will result in\\nhigh product sales. What information would be useful in order to provide\\nsuch a recommendation? Here are a few important questions that we might\\nseek to address:\\n1. Is there a relationship between advertising budget and sales?\\nOur first goal should be to determine whether the data provide evi-\\ndence of an association between advertising expenditure and sales. If\\nthe evidence is weak, then one might argue that no money should be\\nspent on advertising!\\n© Springer Nature Switzerland AG 2023 \\nG. James et al., An Introduction to Statistical Learning, Springer Texts in Statistics, \\nhttps://doi.org/10.1007/978-3-031-38747-0_3 \\n69 70\\n3. Linear Regression\\n2. How strong is the relationship between advertising budget and sales?\\nAssuming that there is a relationship between advertising and sales,\\nwe would like to know the strength of this relationship. Does knowl-\\nedge of the advertising budget provide a lot of information about\\nproduct sales?\\n3. Which media are associated with sales?\\nAre all three media—TV, radio, and newspaper—associated with\\nsales, or are just one or two of the media associated? To answer this\\nquestion, we must find a way to separate out the individual contribu-\\ntion of each medium to sales when we have spent money on all three\\nmedia.\\n4. How large is the association between each medium and sales?\\nFor every dollar spent on advertising in a particular medium, by\\nwhat amount will sales increase? How accurately can we predict this\\namount of increase?\\n5. How accurately can we predict future sales?\\nFor any given level of television, radio, or newspaper advertising, what\\nis our prediction for sales, and what is the accuracy of this prediction?\\n6. Is the relationship linear?\\nIf there is approximately a straight-line relationship between advertis-\\ning expenditure in the various media and sales, then linear regression\\nis an appropriate tool. If not, then it may still be possible to trans-\\nform the predictor or the response so that linear regression can be\\nused.\\n7. Is there synergy among the advertising media?\\nPerhaps spending $50,000 on television advertising and $50,000 on ra-\\ndio advertising is associated with higher sales than allocating $100,000\\nto either television or radio individually. In marketing, this is known\\nas a synergy effect, while in statistics it is called an interaction effect. synergy\\ninteraction\\nIt turns out that linear regression can be used to answer each of these\\nquestions. We will first discuss all of these questions in a general context,\\nand then return to them in this specific context in Section 3.4.\\n3.1\\nSimple Linear Regression\\nSimple linear regression lives up to its name: it is a very straightforward simple linear\\nregression\\napproach for predicting a quantitative response Y on the basis of a sin-\\ngle predictor variable X. It assumes that there is approximately a linear\\nrelationship between X and Y . Mathematically, we can write this linear\\nrelationship as\\nY ≈β0 + β1X.\\n(3.1)\\nYou might read “≈” as “is approximately modeled as”. We will sometimes\\ndescribe (3.1) by saying that we are regressing Y on X (or Y onto X). 3.1 Simple Linear Regression\\n71\\nFor example, X may represent TV advertising and Y may represent sales.\\nThen we can regress sales onto TV by fitting the model\\nsales ≈β0 + β1 × TV.\\nIn Equation 3.1, β0 and β1 are two unknown constants that represent\\nthe intercept and slope terms in the linear model. Together, β0 and β1 are intercept\\nslope\\nknown as the model coefficients or parameters.\\nOnce we have used our\\ncoefficient\\nparameter\\ntraining data to produce estimates ˆβ0 and ˆβ1 for the model coefficients, we\\ncan predict future sales on the basis of a particular value of TV advertising\\nby computing\\nˆy = ˆβ0 + ˆβ1x,\\n(3.2)\\nwhere ˆy indicates a prediction of Y on the basis of X = x. Here we use a\\nhat symbol, ˆ , to denote the estimated value for an unknown parameter\\nor coefficient, or to denote the predicted value of the response.\\n3.1.1\\nEstimating the Coefficients\\nIn practice, β0 and β1 are unknown. So before we can use (3.1) to make\\npredictions, we must use data to estimate the coefficients. Let\\n(x1, y1), (x2, y2), . . . , (xn, yn)\\nrepresent n observation pairs, each of which consists of a measurement of\\nX and a measurement of Y . In the Advertising example, this data set con-\\nsists of the TV advertising budget and product sales in n = 200 different\\nmarkets. (Recall that the data are displayed in Figure 2.1.) Our goal is to\\nobtain coefficient estimates ˆβ0 and ˆβ1 such that the linear model (3.1) fits\\nthe available data well—that is, so that yi ≈ˆβ0 + ˆβ1xi for i = 1, . . . , n. In\\nother words, we want to find an intercept ˆβ0 and a slope ˆβ1 such that the\\nresulting line is as close as possible to the n = 200 data points. There are\\na number of ways of measuring closeness. However, by far the most com-\\nmon approach involves minimizing the least squares criterion, and we take least squares\\nthat approach in this chapter. Alternative approaches will be considered in\\nChapter 6.\\nLet ˆyi = ˆβ0 + ˆβ1xi be the prediction for Y based on the ith value of X.\\nThen ei = yi −ˆyi represents the ith residual—this is the difference between residual\\nthe ith observed response value and the ith response value that is predicted\\nby our linear model. We define the residual sum of squares (RSS) as\\nresidual sum\\nof squares\\nRSS = e2\\n1 + e2\\n2 + · · · + e2\\nn,\\nor equivalently as\\nRSS = (y1 −ˆβ0 −ˆβ1x1)2 +(y2 −ˆβ0 −ˆβ1x2)2 +· · ·+(yn −ˆβ0 −ˆβ1xn)2. (3.3)\\nThe least squares approach chooses ˆβ0 and ˆβ1 to minimize the RSS. Using\\nsome calculus, one can show that the minimizers are\\nˆβ1 =\\n)n\\ni=1(xi −¯x)(yi −¯y)\\n)n\\ni=1(xi −¯x)2\\n,\\nˆβ0 = ¯y −ˆβ1¯x,\\n(3.4) 72\\n3. Linear Regression\\n0\\n50\\n100\\n150\\n200\\n250\\n300\\n5\\n10\\n15\\n20\\n25\\nTV\\nSales\\nFIGURE 3.1. For the Advertising data, the least squares fit for the regression\\nof sales onto TV is shown. The fit is found by minimizing the residual sum of\\nsquares. Each grey line segment represents a residual. In this case a linear fit\\ncaptures the essence of the relationship, although it overestimates the trend in the\\nleft of the plot.\\nwhere ¯y ≡1\\nn\\n)n\\ni=1 yi and ¯x ≡1\\nn\\n)n\\ni=1 xi are the sample means. In other\\nwords, (3.4) defines the least squares coefficient estimates for simple linear\\nregression.\\nFigure 3.1 displays the simple linear regression fit to the Advertising\\ndata, where ˆβ0 = 7.03 and ˆβ1 = 0.0475. In other words, according to\\nthis approximation, an additional $1,000 spent on TV advertising is asso-\\nciated with selling approximately 47.5 additional units of the product. In\\nFigure 3.2, we have computed RSS for a number of values of β0 and β1,\\nusing the advertising data with sales as the response and TV as the predic-\\ntor. In each plot, the red dot represents the pair of least squares estimates\\n(ˆβ0, ˆβ1) given by (3.4). These values clearly minimize the RSS.\\n3.1.2\\nAssessing the Accuracy of the Coefficient Estimates\\nRecall from (2.1) that we assume that the true relationship between X and\\nY takes the form Y = f(X) + ϵ for some unknown function f, where ϵ\\nis a mean-zero random error term. If f is to be approximated by a linear\\nfunction, then we can write this relationship as\\nY = β0 + β1X + ϵ.\\n(3.5)\\nHere β0 is the intercept term—that is, the expected value of Y when X = 0,\\nand β1 is the slope—the average increase in Y associated with a one-unit\\nincrease in X. The error term is a catch-all for what we miss with this\\nsimple model: the true relationship is probably not linear, there may be\\nother variables that cause variation in Y , and there may be measurement\\nerror. We typically assume that the error term is independent of X. 3.1 Simple Linear Regression\\n73\\nFIGURE 3.2.\\nContour and three-dimensional plots of the RSS on the\\nAdvertising data, using sales as the response and TV as the predictor. The\\nred dots correspond to the least squares estimates ˆβ0 and ˆβ1, given by (3.4).\\nThe model given by (3.5) defines the population regression line, which population\\nregression\\nline\\nis the best linear approximation to the true relationship between X and\\nY .1 The least squares regression coefficient estimates (3.4) characterize the\\nleast squares line (3.2). The left-hand panel of Figure 3.3 displays these least squares\\nline\\ntwo lines in a simple simulated example. We created 100 random Xs, and\\ngenerated 100 corresponding Y s from the model\\nY = 2 + 3X + ϵ,\\n(3.6)\\nwhere ϵ was generated from a normal distribution with mean zero. The\\nred line in the left-hand panel of Figure 3.3 displays the true relationship,\\nf(X) = 2 + 3X, while the blue line is the least squares estimate based\\non the observed data. The true relationship is generally not known for\\nreal data, but the least squares line can always be computed using the\\ncoefficient estimates given in (3.4). In other words, in real applications,\\nwe have access to a set of observations from which we can compute the\\nleast squares line; however, the population regression line is unobserved.\\nIn the right-hand panel of Figure 3.3 we have generated ten different data\\nsets from the model given by (3.6) and plotted the corresponding ten least\\nsquares lines. Notice that different data sets generated from the same true\\nmodel result in slightly different least squares lines, but the unobserved\\npopulation regression line does not change.\\nAt first glance, the difference between the population regression line and\\nthe least squares line may seem subtle and confusing. We only have one\\ndata set, and so what does it mean that two different lines describe the\\nrelationship between the predictor and the response? Fundamentally, the\\nconcept of these two lines is a natural extension of the standard statistical\\napproach of using information from a sample to estimate characteristics of a\\nlarge population. For example, suppose that we are interested in knowing\\n1The assumption of linearity is often a useful working model. However, despite what\\nmany textbooks might tell us, we seldom believe that the true relationship is linear.\\nβ0\\nβ1\\n 2.11 \\n 2.15 \\n 2.2 \\n 2.3 \\n 2.5 \\n 2.5 \\n 3 \\n 3 \\n5\\n6\\n7\\n8\\n9\\n0.03\\n0.04\\n0.05\\n0.06\\nRSS\\nβ1\\nβ0 74\\n3. Linear Regression\\n−2\\n−1\\n0\\n1\\n2\\n−10\\n−5\\n0\\n5\\n10\\nX\\nY\\n−2\\n−1\\n0\\n1\\n2\\n−10\\n−5\\n0\\n5\\n10\\nX\\nY\\nFIGURE 3.3.\\nA simulated data set. Left: The red line represents the true\\nrelationship, f(X) = 2 + 3X, which is known as the population regression line.\\nThe blue line is the least squares line; it is the least squares estimate for f(X)\\nbased on the observed data, shown in black. Right: The population regression line\\nis again shown in red, and the least squares line in dark blue. In light blue, ten\\nleast squares lines are shown, each computed on the basis of a separate random\\nset of observations. Each least squares line is different, but on average, the least\\nsquares lines are quite close to the population regression line.\\nthe population mean µ of some random variable Y . Unfortunately, µ is\\nunknown, but we do have access to n observations from Y , y1, . . . , yn,\\nwhich we can use to estimate µ. A reasonable estimate is ˆµ = ¯y, where\\n¯y = 1\\nn\\n)n\\ni=1 yi is the sample mean. The sample mean and the population\\nmean are different, but in general the sample mean will provide a good\\nestimate of the population mean. In the same way, the unknown coefficients\\nβ0 and β1 in linear regression define the population regression line. We seek\\nto estimate these unknown coefficients using ˆβ0 and ˆβ1 given in (3.4). These\\ncoefficient estimates define the least squares line.\\nThe analogy between linear regression and estimation of the mean of a\\nrandom variable is an apt one based on the concept of bias. If we use the bias\\nsample mean ˆµ to estimate µ, this estimate is unbiased, in the sense that unbiased\\non average, we expect ˆµ to equal µ. What exactly does this mean? It means\\nthat on the basis of one particular set of observations y1, . . . , yn, ˆµ might\\noverestimate µ, and on the basis of another set of observations, ˆµ might\\nunderestimate µ. But if we could average a huge number of estimates of\\nµ obtained from a huge number of sets of observations, then this average\\nwould exactly equal µ. Hence, an unbiased estimator does not systematically\\nover- or under-estimate the true parameter. The property of unbiasedness\\nholds for the least squares coefficient estimates given by (3.4) as well: if\\nwe estimate β0 and β1 on the basis of a particular data set, then our\\nestimates won’t be exactly equal to β0 and β1. But if we could average\\nthe estimates obtained over a huge number of data sets, then the average\\nof these estimates would be spot on! In fact, we can see from the right-\\nhand panel of Figure 3.3 that the average of many least squares lines, each 3.1 Simple Linear Regression\\n75\\nestimated from a separate data set, is pretty close to the true population\\nregression line.\\nWe continue the analogy with the estimation of the population mean\\nµ of a random variable Y . A natural question is as follows: how accurate\\nis the sample mean ˆµ as an estimate of µ? We have established that the\\naverage of ˆµ’s over many data sets will be very close to µ, but that a\\nsingle estimate ˆµ may be a substantial underestimate or overestimate of µ.\\nHow far off will that single estimate of ˆµ be? In general, we answer this\\nquestion by computing the standard error of ˆµ, written as SE(ˆµ). We have standard\\nerror\\nthe well-known formula\\nVar(ˆµ) = SE(ˆµ)2 = σ2\\nn ,\\n(3.7)\\nwhere σ is the standard deviation of each of the realizations yi of Y .2\\nRoughly speaking, the standard error tells us the average amount that this\\nestimate ˆµ differs from the actual value of µ. Equation 3.7 also tells us how\\nthis deviation shrinks with n—the more observations we have, the smaller\\nthe standard error of ˆµ. In a similar vein, we can wonder how close ˆβ0\\nand ˆβ1 are to the true values β0 and β1. To compute the standard errors\\nassociated with ˆβ0 and ˆβ1, we use the following formulas:\\nSE(ˆβ0)\\n2 = σ2\\n3 1\\nn +\\n¯x2\\n)n\\ni=1(xi −¯x)2\\n4\\n,\\nSE(ˆβ1)\\n2 =\\nσ2\\n)n\\ni=1(xi −¯x)2 , (3.8)\\nwhere σ2 = Var(ϵ). For these formulas to be strictly valid, we need to\\nassume that the errors ϵi for each observation have common variance σ2 and\\nare uncorrelated. This is clearly not true in Figure 3.1, but the formula still\\nturns out to be a good approximation. Notice in the formula that SE(ˆβ1) is\\nsmaller when the xi are more spread out; intuitively we have more leverage\\nto estimate a slope when this is the case. We also see that SE(ˆβ0) would be\\nthe same as SE(ˆµ) if ¯x were zero (in which case ˆβ0 would be equal to ¯y). In\\ngeneral, σ2 is not known, but can be estimated from the data. This estimate\\nof σ is known as the residual standard error, and is given by the formula residual\\nstandard\\nerror\\nRSE =\\n5\\nRSS/(n −2). Strictly speaking, when σ2 is estimated from the\\ndata we should write 6\\nSE(ˆβ1) to indicate that an estimate has been made,\\nbut for simplicity of notation we will drop this extra “hat”.\\nStandard errors can be used to compute confidence intervals. A 95 % confidence\\ninterval\\nconfidence interval is defined as a range of values such that with 95 %\\nprobability, the range will contain the true unknown value of the param-\\neter. The range is defined in terms of lower and upper limits computed\\nfrom the sample of data. A 95% confidence interval has the following prop-\\nerty: if we take repeated samples and construct the confidence interval for\\neach sample, 95% of the intervals will contain the true unknown value of\\nthe parameter. For linear regression, the 95 % confidence interval for β1\\napproximately takes the form\\nˆβ1 ± 2 · SE(ˆβ1).\\n(3.9)\\n2This formula holds provided that the n observations are uncorrelated. 76\\n3. Linear Regression\\nThat is, there is approximately a 95 % chance that the interval\\n7\\nˆβ1 −2 · SE(ˆβ1), ˆβ1 + 2 · SE(ˆβ1)\\n8\\n(3.10)\\nwill contain the true value of β1.3 Similarly, a confidence interval for β0\\napproximately takes the form\\nˆβ0 ± 2 · SE(ˆβ0).\\n(3.11)\\nIn the case of the advertising data, the 95 % confidence interval for β0\\nis [6.130, 7.935] and the 95 % confidence interval for β1 is [0.042, 0.053].\\nTherefore, we can conclude that in the absence of any advertising, sales will,\\non average, fall somewhere between 6,130 and 7,935 units. Furthermore,\\nfor each $1,000 increase in television advertising, there will be an average\\nincrease in sales of between 42 and 53 units.\\nStandard errors can also be used to perform hypothesis tests on the hypothesis\\ntest\\ncoefficients. The most common hypothesis test involves testing the null\\nhypothesis of\\nnull\\nhypothesis\\nH0 : There is no relationship between X and Y\\n(3.12)\\nversus the alternative hypothesis\\nalternative\\nhypothesis\\nHa : There is some relationship between X and Y .\\n(3.13)\\nMathematically, this corresponds to testing\\nH0 : β1 = 0\\nversus\\nHa : β1 ̸= 0,\\nsince if β1 = 0 then the model (3.5) reduces to Y = β0 + ϵ, and X is\\nnot associated with Y . To test the null hypothesis, we need to determine\\nwhether ˆβ1, our estimate for β1, is sufficiently far from zero that we can\\nbe confident that β1 is non-zero. How far is far enough? This of course\\ndepends on the accuracy of ˆβ1—that is, it depends on SE(ˆβ1). If SE(ˆβ1) is\\nsmall, then even relatively small values of ˆβ1 may provide strong evidence\\nthat β1 ̸= 0, and hence that there is a relationship between X and Y . In\\ncontrast, if SE(ˆβ1) is large, then ˆβ1 must be large in absolute value in order\\nfor us to reject the null hypothesis. In practice, we compute a t-statistic, t-statistic\\ngiven by\\nt =\\nˆβ1 −0\\nSE(ˆβ1)\\n,\\n(3.14)\\n3Approximately for several reasons. Equation 3.10 relies on the assumption that the\\nerrors are Gaussian. Also, the factor of 2 in front of the SE(ˆβ1) term will vary slightly\\ndepending on the number of observations n in the linear regression. To be precise, rather\\nthan the number 2, (3.10) should contain the 97.5 % quantile of a t-distribution with\\nn−2 degrees of freedom. Details of how to compute the 95 % confidence interval precisely\\nin R will be provided later in this chapter. 3.1 Simple Linear Regression\\n77\\nCoefficient\\nStd. error\\nt-statistic\\np-value\\nIntercept\\n7.0325\\n0.4578\\n15.36\\n< 0.0001\\nTV\\n0.0475\\n0.0027\\n17.67\\n< 0.0001\\nTABLE 3.1. For the Advertising data, coefficients of the least squares model\\nfor the regression of number of units sold on TV advertising budget. An increase\\nof $1,000 in the TV advertising budget is associated with an increase in sales by\\naround 50 units. (Recall that the sales variable is in thousands of units, and the\\nTV variable is in thousands of dollars.)\\nwhich measures the number of standard deviations that ˆβ1 is away from 0. If\\nthere really is no relationship between X and Y , then we expect that (3.14)\\nwill have a t-distribution with n −2 degrees of freedom. The t-distribution\\nhas a bell shape and for values of n greater than approximately 30 it is\\nquite similar to the standard normal distribution. Consequently, it is a\\nsimple matter to compute the probability of observing any number equal\\nto |t| or larger in absolute value, assuming β1 = 0. We call this probability\\nthe p-value. Roughly speaking, we interpret the p-value as follows: a small p-value\\np-value indicates that it is unlikely to observe such a substantial association\\nbetween the predictor and the response due to chance, in the absence of\\nany real association between the predictor and the response. Hence, if we\\nsee a small p-value, then we can infer that there is an association between\\nthe predictor and the response. We reject the null hypothesis—that is, we\\ndeclare a relationship to exist between X and Y —if the p-value is small\\nenough. Typical p-value cutoffs for rejecting the null hypothesis are 5% or\\n1%, although this topic will be explored in much greater detail in Chap-\\nter 13. When n = 30, these correspond to t-statistics (3.14) of around 2\\nand 2.75, respectively.\\nTable 3.1 provides details of the least squares model for the regression of\\nnumber of units sold on TV advertising budget for the Advertising data.\\nNotice that the coefficients for ˆβ0 and ˆβ1 are very large relative to their\\nstandard errors, so the t-statistics are also large; the probabilities of seeing\\nsuch values if H0 is true are virtually zero. Hence we can conclude that\\nβ0 ̸= 0 and β1 ̸= 0.4\\n3.1.3\\nAssessing the Accuracy of the Model\\nOnce we have rejected the null hypothesis (3.12) in favor of the alternative\\nhypothesis (3.13), it is natural to want to quantify the extent to which the\\nmodel fits the data. The quality of a linear regression fit is typically assessed\\nusing two related quantities: the residual standard error (RSE) and the R2\\nR2\\nstatistic.\\n4In Table 3.1, a small p-value for the intercept indicates that we can reject the null\\nhypothesis that β0 = 0, and a small p-value for TV indicates that we can reject the null\\nhypothesis that β1 = 0. Rejecting the latter null hypothesis allows us to conclude that\\nthere is a relationship between TV and sales. Rejecting the former allows us to conclude\\nthat in the absence of TV expenditure, sales are non-zero. 78\\n3. Linear Regression\\nQuantity\\nValue\\nResidual standard error\\n3.26\\nR2\\n0.612\\nF-statistic\\n312.1\\nTABLE 3.2.\\nFor the Advertising data, more information about the least\\nsquares model for the regression of number of units sold on TV advertising budget.\\nTable 3.2 displays the RSE, the R2 statistic, and the F-statistic (to be\\ndescribed in Section 3.2.2) for the linear regression of number of units sold\\non TV advertising budget.\\nResidual Standard Error\\nRecall from the model (3.5) that associated with each observation is an\\nerror term ϵ. Due to the presence of these error terms, even if we knew the\\ntrue regression line (i.e. even if β0 and β1 were known), we would not be\\nable to perfectly predict Y from X. The RSE is an estimate of the standard\\ndeviation of ϵ. Roughly speaking, it is the average amount that the response\\nwill deviate from the true regression line. It is computed using the formula\\nRSE =\\n9\\n1\\nn −2RSS =\\n:\\n;\\n;\\n<\\n1\\nn −2\\nn\\n0\\ni=1\\n(yi −ˆyi)2.\\n(3.15)\\nNote that RSS was defined in Section 3.1.1, and is given by the formula\\nRSS =\\nn\\n0\\ni=1\\n(yi −ˆyi)2.\\n(3.16)\\nIn the case of the advertising data, we see from the linear regression\\noutput in Table 3.2 that the RSE is 3.26. In other words, actual sales in\\neach market deviate from the true regression line by approximately 3,260\\nunits, on average. Another way to think about this is that even if the\\nmodel were correct and the true values of the unknown coefficients β0\\nand β1 were known exactly, any prediction of sales on the basis of TV\\nadvertising would still be off by about 3,260 units on average. Of course,\\nwhether or not 3,260 units is an acceptable prediction error depends on the\\nproblem context. In the advertising data set, the mean value of sales over\\nall markets is approximately 14,000 units, and so the percentage error is\\n3,260/14,000 = 23 %.\\nThe RSE is considered a measure of the lack of fit of the model (3.5) to\\nthe data. If the predictions obtained using the model are very close to the\\ntrue outcome values—that is, if ˆyi ≈yi for i = 1, . . . , n—then (3.15) will\\nbe small, and we can conclude that the model fits the data very well. On\\nthe other hand, if ˆyi is very far from yi for one or more observations, then\\nthe RSE may be quite large, indicating that the model doesn’t fit the data\\nwell.\\nR2 Statistic\\nThe RSE provides an absolute measure of lack of fit of the model (3.5)\\nto the data. But since it is measured in the units of Y , it is not always 3.1 Simple Linear Regression\\n79\\nclear what constitutes a good RSE. The R2 statistic provides an alternative\\nmeasure of fit. It takes the form of a proportion—the proportion of variance\\nexplained—and so it always takes on a value between 0 and 1, and is\\nindependent of the scale of Y .\\nTo calculate R2, we use the formula\\nR2 = TSS −RSS\\nTSS\\n= 1 −RSS\\nTSS\\n(3.17)\\nwhere TSS = )(yi −¯y)2 is the total sum of squares, and RSS is defined total sum of\\nsquares\\nin (3.16). TSS measures the total variance in the response Y , and can be\\nthought of as the amount of variability inherent in the response before the\\nregression is performed. In contrast, RSS measures the amount of variability\\nthat is left unexplained after performing the regression. Hence, TSS −RSS\\nmeasures the amount of variability in the response that is explained (or\\nremoved) by performing the regression, and R2 measures the proportion of\\nvariability in Y that can be explained using X. An R2 statistic that is close\\nto 1 indicates that a large proportion of the variability in the response is\\nexplained by the regression. A number near 0 indicates that the regression\\ndoes not explain much of the variability in the response; this might occur\\nbecause the linear model is wrong, or the error variance σ2 is high, or both.\\nIn Table 3.2, the R2 was 0.61, and so just under two-thirds of the variability\\nin sales is explained by a linear regression on TV.\\nThe R2 statistic (3.17) has an interpretational advantage over the RSE\\n(3.15), since unlike the RSE, it always lies between 0 and 1. However, it can\\nstill be challenging to determine what is a good R2 value, and in general,\\nthis will depend on the application. For instance, in certain problems in\\nphysics, we may know that the data truly comes from a linear model with\\na small residual error. In this case, we would expect to see an R2 value that\\nis extremely close to 1, and a substantially smaller R2 value might indicate a\\nserious problem with the experiment in which the data were generated. On\\nthe other hand, in typical applications in biology, psychology, marketing,\\nand other domains, the linear model (3.5) is at best an extremely rough\\napproximation to the data, and residual errors due to other unmeasured\\nfactors are often very large. In this setting, we would expect only a very\\nsmall proportion of the variance in the response to be explained by the\\npredictor, and an R2 value well below 0.1 might be more realistic!\\nThe R2 statistic is a measure of the linear relationship between X and\\nY . Recall that correlation, defined as\\ncorrelation\\nCor(X, Y ) =\\n)n\\ni=1(xi −x)(yi −y)\\n5)n\\ni=1(xi −x)25)n\\ni=1(yi −y)2 ,\\n(3.18)\\nis also a measure of the linear relationship between X and Y .5 This sug-\\ngests that we might be able to use r = Cor(X, Y ) instead of R2 in order to\\nassess the fit of the linear model. In fact, it can be shown that in the simple\\n5We note that in fact, the right-hand side of (3.18) is the sample correlation; thus,\\nit would be more correct to write\\n!\\nCor(X, Y ); however, we omit the “hat” for ease of\\nnotation. 80\\n3. Linear Regression\\nSimple regression of sales on radio\\nCoefficient\\nStd. error\\nt-statistic\\np-value\\nIntercept\\n9.312\\n0.563\\n16.54\\n< 0.0001\\nradio\\n0.203\\n0.020\\n9.92\\n< 0.0001\\nSimple regression of sales on newspaper\\nCoefficient\\nStd. error\\nt-statistic\\np-value\\nIntercept\\n12.351\\n0.621\\n19.88\\n< 0.0001\\nnewspaper\\n0.055\\n0.017\\n3.30\\n0.00115\\nTABLE 3.3. More simple linear regression models for the Advertising data.\\nCoefficients of the simple linear regression model for number of units sold on Top:\\nradio advertising budget and Bottom: newspaper advertising budget. A $1,000 in-\\ncrease in spending on radio advertising is associated with an average increase in\\nsales by around 203 units, while the same increase in spending on newspaper ad-\\nvertising is associated with an average increase in sales by around 55 units. (Note\\nthat the sales variable is in thousands of units, and the radio and newspaper\\nvariables are in thousands of dollars.)\\nlinear regression setting, R2 = r2. In other words, the squared correlation\\nand the R2 statistic are identical. However, in the next section we will\\ndiscuss the multiple linear regression problem, in which we use several pre-\\ndictors simultaneously to predict the response. The concept of correlation\\nbetween the predictors and the response does not extend automatically to\\nthis setting, since correlation quantifies the association between a single\\npair of variables rather than between a larger number of variables. We will\\nsee that R2 fills this role.\\n3.2\\nMultiple Linear Regression\\nSimple linear regression is a useful approach for predicting a response on the\\nbasis of a single predictor variable. However, in practice we often have more\\nthan one predictor. For example, in the Advertising data, we have examined\\nthe relationship between sales and TV advertising. We also have data for\\nthe amount of money spent advertising on the radio and in newspapers,\\nand we may want to know whether either of these two media is associated\\nwith sales. How can we extend our analysis of the advertising data in order\\nto accommodate these two additional predictors?\\nOne option is to run three separate simple linear regressions, each of\\nwhich uses a different advertising medium as a predictor. For instance,\\nwe can fit a simple linear regression to predict sales on the basis of the\\namount spent on radio advertisements. Results are shown in Table 3.3 (top\\ntable). We find that a $1,000 increase in spending on radio advertising is\\nassociated with an increase in sales of around 203 units. Table 3.3 (bottom\\ntable) contains the least squares coefficients for a simple linear regression of\\nsales onto newspaper advertising budget. A $1,000 increase in newspaper\\nadvertising budget is associated with an increase in sales of approximately\\n55 units. 3.2 Multiple Linear Regression\\n81\\nHowever, the approach of fitting a separate simple linear regression model\\nfor each predictor is not entirely satisfactory. First of all, it is unclear how to\\nmake a single prediction of sales given the three advertising media budgets,\\nsince each of the budgets is associated with a separate regression equation.\\nSecond, each of the three regression equations ignores the other two media\\nin forming estimates for the regression coefficients. We will see shortly that\\nif the media budgets are correlated with each other in the 200 markets\\nin our data set, then this can lead to very misleading estimates of the\\nassociation between each media budget and sales.\\nInstead of fitting a separate simple linear regression model for each pre-\\ndictor, a better approach is to extend the simple linear regression model\\n(3.5) so that it can directly accommodate multiple predictors. We can do\\nthis by giving each predictor a separate slope coefficient in a single model.\\nIn general, suppose that we have p distinct predictors. Then the multiple\\nlinear regression model takes the form\\nY = β0 + β1X1 + β2X2 + · · · + βpXp + ϵ,\\n(3.19)\\nwhere Xj represents the jth predictor and βj quantifies the association\\nbetween that variable and the response. We interpret βj as the average\\neffect on Y of a one unit increase in Xj, holding all other predictors fixed.\\nIn the advertising example, (3.19) becomes\\nsales = β0 + β1 × TV + β2 × radio + β3 × newspaper + ϵ.\\n(3.20)\\n3.2.1\\nEstimating the Regression Coefficients\\nAs was the case in the simple linear regression setting, the regression coef-\\nficients β0, β1, . . . , βp in (3.19) are unknown, and must be estimated. Given\\nestimates ˆβ0, ˆβ1, . . . , ˆβp, we can make predictions using the formula\\nˆy = ˆβ0 + ˆβ1x1 + ˆβ2x2 + · · · + ˆβpxp.\\n(3.21)\\nThe parameters are estimated using the same least squares approach that\\nwe saw in the context of simple linear regression. We choose β0, β1, . . . , βp\\nto minimize the sum of squared residuals\\nRSS\\n=\\nn\\n0\\ni=1\\n(yi −ˆyi)2\\n=\\nn\\n0\\ni=1\\n(yi −ˆβ0 −ˆβ1xi1 −ˆβ2xi2 −· · · −ˆβpxip)2.\\n(3.22)\\nThe values ˆβ0, ˆβ1, . . . , ˆβp that minimize (3.22) are the multiple least squares\\nregression coefficient estimates. Unlike the simple linear regression esti-\\nmates given in (3.4), the multiple regression coefficient estimates have\\nsomewhat complicated forms that are most easily represented using ma-\\ntrix algebra. For this reason, we do not provide them here. Any statistical\\nsoftware package can be used to compute these coefficient estimates, and\\nlater in this chapter we will show how this can be done in R. Figure 3.4 82\\n3. Linear Regression\\nX1\\nX2\\nY\\nFIGURE 3.4.\\nIn a three-dimensional setting, with two predictors and one\\nresponse, the least squares regression line becomes a plane. The plane is chosen\\nto minimize the sum of the squared vertical distances between each observation\\n(shown in red) and the plane.\\nillustrates an example of the least squares fit to a toy data set with p = 2\\npredictors.\\nTable 3.4 displays the multiple regression coefficient estimates when TV,\\nradio, and newspaper advertising budgets are used to predict product sales\\nusing the Advertising data. We interpret these results as follows: for a given\\namount of TV and newspaper advertising, spending an additional $1,000 on\\nradio advertising is associated with approximately 189 units of additional\\nsales. Comparing these coefficient estimates to those displayed in Tables 3.1\\nand 3.3, we notice that the multiple regression coefficient estimates for\\nTV and radio are pretty similar to the simple linear regression coefficient\\nestimates. However, while the newspaper regression coefficient estimate in\\nTable 3.3 was significantly non-zero, the coefficient estimate for newspaper\\nin the multiple regression model is close to zero, and the corresponding p-\\nvalue is no longer significant, with a value around 0.86. This illustrates that\\nthe simple and multiple regression coefficients can be quite different. This\\ndifference stems from the fact that in the simple regression case, the slope\\nterm represents the average increase in product sales associated with a\\n$1,000 increase in newspaper advertising, ignoring other predictors such as\\nTV and radio. By contrast, in the multiple regression setting, the coefficient\\nfor newspaper represents the average increase in product sales associated\\nwith increasing newspaper spending by $1,000 while holding TV and radio\\nfixed.\\nDoes it make sense for the multiple regression to suggest no relationship\\nbetween sales and newspaper while the simple linear regression implies the 3.2 Multiple Linear Regression\\n83\\nCoefficient\\nStd. error\\nt-statistic\\np-value\\nIntercept\\n2.939\\n0.3119\\n9.42\\n< 0.0001\\nTV\\n0.046\\n0.0014\\n32.81\\n< 0.0001\\nradio\\n0.189\\n0.0086\\n21.89\\n< 0.0001\\nnewspaper\\n−0.001\\n0.0059\\n−0.18\\n0.8599\\nTABLE 3.4. For the Advertising data, least squares coefficient estimates of the\\nmultiple linear regression of number of units sold on TV, radio, and newspaper\\nadvertising budgets.\\nTV\\nradio\\nnewspaper\\nsales\\nTV\\n1.0000\\n0.0548\\n0.0567\\n0.7822\\nradio\\n1.0000\\n0.3541\\n0.5762\\nnewspaper\\n1.0000\\n0.2283\\nsales\\n1.0000\\nTABLE 3.5. Correlation matrix for TV, radio, newspaper, and sales for the\\nAdvertising data.\\nopposite? In fact it does. Consider the correlation matrix for the three\\npredictor variables and response variable, displayed in Table 3.5. Notice\\nthat the correlation between radio and newspaper is 0.35. This indicates\\nthat markets with high newspaper advertising tend to also have high ra-\\ndio advertising. Now suppose that the multiple regression is correct and\\nnewspaper advertising is not associated with sales, but radio advertising\\nis associated with sales. Then in markets where we spend more on radio\\nour sales will tend to be higher, and as our correlation matrix shows, we\\nalso tend to spend more on newspaper advertising in those same mar-\\nkets. Hence, in a simple linear regression which only examines sales versus\\nnewspaper, we will observe that higher values of newspaper tend to be as-\\nsociated with higher values of sales, even though newspaper advertising is\\nnot directly associated with sales. So newspaper advertising is a surrogate\\nfor radio advertising; newspaper gets “credit” for the association between\\nradio on sales.\\nThis slightly counterintuitive result is very common in many real life\\nsituations. Consider an absurd example to illustrate the point. Running\\na regression of shark attacks versus ice cream sales for data collected at\\na given beach community over a period of time would show a positive\\nrelationship, similar to that seen between sales and newspaper. Of course\\nno one has (yet) suggested that ice creams should be banned at beaches\\nto reduce shark attacks. In reality, higher temperatures cause more people\\nto visit the beach, which in turn results in more ice cream sales and more\\nshark attacks. A multiple regression of shark attacks onto ice cream sales\\nand temperature reveals that, as intuition implies, ice cream sales is no\\nlonger a significant predictor after adjusting for temperature.\\n3.2.2\\nSome Important Questions\\nWhen we perform multiple linear regression, we usually are interested in\\nanswering a few important questions. 84\\n3. Linear Regression\\n1. Is at least one of the predictors X1, X2, . . . , Xp useful in predicting\\nthe response?\\n2. Do all the predictors help to explain Y , or is only a subset of the\\npredictors useful?\\n3. How well does the model fit the data?\\n4. Given a set of predictor values, what response value should we predict,\\nand how accurate is our prediction?\\nWe now address each of these questions in turn.\\nOne: Is There a Relationship Between the Response and Predictors?\\nRecall that in the simple linear regression setting, in order to determine\\nwhether there is a relationship between the response and the predictor we\\ncan simply check whether β1 = 0. In the multiple regression setting with p\\npredictors, we need to ask whether all of the regression coefficients are zero,\\ni.e. whether β1 = β2 = · · · = βp = 0. As in the simple linear regression\\nsetting, we use a hypothesis test to answer this question. We test the null\\nhypothesis,\\nH0 : β1 = β2 = · · · = βp = 0\\nversus the alternative\\nHa : at least one βj is non-zero.\\nThis hypothesis test is performed by computing the F-statistic,\\nF-statistic\\nF = (TSS −RSS)/p\\nRSS/(n −p −1),\\n(3.23)\\nwhere, as with simple linear regression, TSS = )(yi −¯y)2 and RSS =\\n)(yi −ˆyi)2. If the linear model assumptions are correct, one can show that\\nE{RSS/(n −p −1)} = σ2\\nand that, provided H0 is true,\\nE{(TSS −RSS)/p} = σ2.\\nHence, when there is no relationship between the response and predictors,\\none would expect the F-statistic to take on a value close to 1. On the other\\nhand, if Ha is true, then E{(TSS −RSS)/p} > σ2, so we expect F to be\\ngreater than 1.\\nThe F-statistic for the multiple linear regression model obtained by re-\\ngressing sales onto radio, TV, and newspaper is shown in Table 3.6. In this\\nexample the F-statistic is 570. Since this is far larger than 1, it provides\\ncompelling evidence against the null hypothesis H0. In other words, the\\nlarge F-statistic suggests that at least one of the advertising media must\\nbe related to sales. However, what if the F-statistic had been closer to\\n1? How large does the F-statistic need to be before we can reject H0 and 3.2 Multiple Linear Regression\\n85\\nQuantity\\nValue\\nResidual standard error\\n1.69\\nR2\\n0.897\\nF-statistic\\n570\\nTABLE 3.6. More information about the least squares model for the regression\\nof number of units sold on TV, newspaper, and radio advertising budgets in the\\nAdvertising data. Other information about this model was displayed in Table 3.4.\\nconclude that there is a relationship? It turns out that the answer depends\\non the values of n and p. When n is large, an F-statistic that is just a\\nlittle larger than 1 might still provide evidence against H0. In contrast,\\na larger F-statistic is needed to reject H0 if n is small. When H0 is true\\nand the errors ϵi have a normal distribution, the F-statistic follows an\\nF-distribution.6 For any given value of n and p, any statistical software\\npackage can be used to compute the p-value associated with the F-statistic\\nusing this distribution. Based on this p-value, we can determine whether\\nor not to reject H0. For the advertising data, the p-value associated with\\nthe F-statistic in Table 3.6 is essentially zero, so we have extremely strong\\nevidence that at least one of the media is associated with increased sales.\\nIn (3.23) we are testing H0 that all the coefficients are zero. Sometimes\\nwe want to test that a particular subset of q of the coefficients are zero.\\nThis corresponds to a null hypothesis\\nH0 :\\nβp−q+1 = βp−q+2 = · · · = βp = 0,\\nwhere for convenience we have put the variables chosen for omission at the\\nend of the list. In this case we fit a second model that uses all the variables\\nexcept those last q. Suppose that the residual sum of squares for that model\\nis RSS0. Then the appropriate F-statistic is\\nF = (RSS0 −RSS)/q\\nRSS/(n −p −1).\\n(3.24)\\nNotice that in Table 3.4, for each individual predictor a t-statistic and\\na p-value were reported. These provide information about whether each\\nindividual predictor is related to the response, after adjusting for the other\\npredictors. It turns out that each of these is exactly equivalent7 to the F-\\ntest that omits that single variable from the model, leaving all the others\\nin—i.e. q=1 in (3.24). So it reports the partial effect of adding that variable\\nto the model. For instance, as we discussed earlier, these p-values indicate\\nthat TV and radio are related to sales, but that there is no evidence that\\nnewspaper is associated with sales, when TV and radio are held fixed.\\nGiven these individual p-values for each variable, why do we need to look\\nat the overall F-statistic? After all, it seems likely that if any one of the\\np-values for the individual variables is very small, then at least one of the\\npredictors is related to the response. However, this logic is flawed, especially\\nwhen the number of predictors p is large.\\n6Even if the errors are not normally-distributed, the F-statistic approximately follows\\nan F-distribution provided that the sample size n is large.\\n7The square of each t-statistic is the corresponding F-statistic. 86\\n3. Linear Regression\\nFor instance, consider an example in which p = 100 and H0 : β1 = β2 =\\n· · · = βp = 0 is true, so no variable is truly associated with the response. In\\nthis situation, about 5 % of the p-values associated with each variable (of\\nthe type shown in Table 3.4) will be below 0.05 by chance. In other words,\\nwe expect to see approximately five small p-values even in the absence of\\nany true association between the predictors and the response.8 In fact, it\\nis likely that we will observe at least one p-value below 0.05 by chance!\\nHence, if we use the individual t-statistics and associated p-values in order\\nto decide whether or not there is any association between the variables and\\nthe response, there is a very high chance that we will incorrectly conclude\\nthat there is a relationship. However, the F-statistic does not suffer from\\nthis problem because it adjusts for the number of predictors. Hence, if H0\\nis true, there is only a 5 % chance that the F-statistic will result in a p-\\nvalue below 0.05, regardless of the number of predictors or the number of\\nobservations.\\nThe approach of using an F-statistic to test for any association between\\nthe predictors and the response works when p is relatively small, and cer-\\ntainly small compared to n. However, sometimes we have a very large num-\\nber of variables. If p > n then there are more coefficients βj to estimate\\nthan observations from which to estimate them. In this case we cannot\\neven fit the multiple linear regression model using least squares, so the F-\\nstatistic cannot be used, and neither can most of the other concepts that\\nwe have seen so far in this chapter. When p is large, some of the approaches\\ndiscussed in the next section, such as forward selection, can be used. This\\nhigh-dimensional setting is discussed in greater detail in Chapter 6.\\nhigh-\\ndimensional\\nTwo: Deciding on Important Variables\\nAs discussed in the previous section, the first step in a multiple regression\\nanalysis is to compute the F-statistic and to examine the associated p-\\nvalue. If we conclude on the basis of that p-value that at least one of the\\npredictors is related to the response, then it is natural to wonder which are\\nthe guilty ones! We could look at the individual p-values as in Table 3.4,\\nbut as discussed (and as further explored in Chapter 13), if p is large we\\nare likely to make some false discoveries.\\nIt is possible that all of the predictors are associated with the response,\\nbut it is more often the case that the response is only associated with\\na subset of the predictors. The task of determining which predictors are\\nassociated with the response, in order to fit a single model involving only\\nthose predictors, is referred to as variable selection. The variable selection variable\\nselection\\nproblem is studied extensively in Chapter 6, and so here we will provide\\nonly a brief outline of some classical approaches.\\nIdeally, we would like to perform variable selection by trying out a lot of\\ndifferent models, each containing a different subset of the predictors. For\\ninstance, if p = 2, then we can consider four models: (1) a model contain-\\ning no variables, (2) a model containing X1 only, (3) a model containing\\n8This is related to the important concept of multiple testing, which is the focus of\\nChapter 13. 3.2 Multiple Linear Regression\\n87\\nX2 only, and (4) a model containing both X1 and X2. We can then se-\\nlect the best model out of all of the models that we have considered. How\\ndo we determine which model is best? Various statistics can be used to\\njudge the quality of a model. These include Mallow’s Cp, Akaike informa- Mallow’s Cp\\ntion criterion (AIC), Bayesian information criterion (BIC), and adjusted\\nAkaike\\ninformation\\ncriterion\\nBayesian\\ninformation\\ncriterion\\nR2. These are discussed in more detail in Chapter 6. We can also deter-\\nadjusted R2\\nmine which model is best by plotting various model outputs, such as the\\nresiduals, in order to search for patterns.\\nUnfortunately, there are a total of 2p models that contain subsets of p\\nvariables. This means that even for moderate p, trying out every possible\\nsubset of the predictors is infeasible. For instance, we saw that if p = 2, then\\nthere are 22 = 4 models to consider. But if p = 30, then we must consider\\n230 = 1,073,741,824 models! This is not practical. Therefore, unless p is very\\nsmall, we cannot consider all 2p models, and instead we need an automated\\nand efficient approach to choose a smaller set of models to consider. There\\nare three classical approaches for this task:\\n• Forward selection. We begin with the null model—a model that con- forward\\nselection\\nnull model\\ntains an intercept but no predictors. We then fit p simple linear re-\\ngressions and add to the null model the variable that results in the\\nlowest RSS. We then add to that model the variable that results\\nin the lowest RSS for the new two-variable model. This approach is\\ncontinued until some stopping rule is satisfied.\\n• Backward selection. We start with all variables in the model, and backward\\nselection\\nremove the variable with the largest p-value—that is, the variable\\nthat is the least statistically significant. The new (p −1)-variable\\nmodel is fit, and the variable with the largest p-value is removed. This\\nprocedure continues until a stopping rule is reached. For instance, we\\nmay stop when all remaining variables have a p-value below some\\nthreshold.\\n• Mixed selection. This is a combination of forward and backward se- mixed\\nselection\\nlection. We start with no variables in the model, and as with forward\\nselection, we add the variable that provides the best fit. We con-\\ntinue to add variables one-by-one. Of course, as we noted with the\\nAdvertising example, the p-values for variables can become larger as\\nnew predictors are added to the model. Hence, if at any point the\\np-value for one of the variables in the model rises above a certain\\nthreshold, then we remove that variable from the model. We con-\\ntinue to perform these forward and backward steps until all variables\\nin the model have a sufficiently low p-value, and all variables outside\\nthe model would have a large p-value if added to the model.\\nBackward selection cannot be used if p > n, while forward selection can\\nalways be used. Forward selection is a greedy approach, and might include\\nvariables early that later become redundant. Mixed selection can remedy\\nthis. 88\\n3. Linear Regression\\nThree: Model Fit\\nTwo of the most common numerical measures of model fit are the RSE and\\nR2, the fraction of variance explained. These quantities are computed and\\ninterpreted in the same fashion as for simple linear regression.\\nRecall that in simple regression, R2 is the square of the correlation of the\\nresponse and the variable. In multiple linear regression, it turns out that it\\nequals Cor(Y, ˆY )2, the square of the correlation between the response and\\nthe fitted linear model; in fact one property of the fitted linear model is\\nthat it maximizes this correlation among all possible linear models.\\nAn R2 value close to 1 indicates that the model explains a large por-\\ntion of the variance in the response variable. As an example, we saw in\\nTable 3.6 that for the Advertising data, the model that uses all three ad-\\nvertising media to predict sales has an R2 of 0.8972. On the other hand,\\nthe model that uses only TV and radio to predict sales has an R2 value\\nof 0.89719. In other words, there is a small increase in R2 if we include\\nnewspaper advertising in the model that already contains TV and radio\\nadvertising, even though we saw earlier that the p-value for newspaper ad-\\nvertising in Table 3.4 is not significant. It turns out that R2 will always\\nincrease when more variables are added to the model, even if those vari-\\nables are only weakly associated with the response. This is due to the fact\\nthat adding another variable always results in a decrease in the residual\\nsum of squares on the training data (though not necessarily the testing\\ndata). Thus, the R2 statistic, which is also computed on the training data,\\nmust increase. The fact that adding newspaper advertising to the model\\ncontaining only TV and radio advertising leads to just a tiny increase in\\nR2 provides additional evidence that newspaper can be dropped from the\\nmodel. Essentially, newspaper provides no real improvement in the model\\nfit to the training samples, and its inclusion will likely lead to poor results\\non independent test samples due to overfitting.\\nBy contrast, the model containing only TV as a predictor had an R2 of\\n0.61 (Table 3.2). Adding radio to the model leads to a substantial improve-\\nment in R2. This implies that a model that uses TV and radio expenditures\\nto predict sales is substantially better than one that uses only TV advertis-\\ning. We could further quantify this improvement by looking at the p-value\\nfor the radio coefficient in a model that contains only TV and radio as\\npredictors.\\nThe model that contains only TV and radio as predictors has an RSE\\nof 1.681, and the model that also contains newspaper as a predictor has\\nan RSE of 1.686 (Table 3.6). In contrast, the model that contains only TV\\nhas an RSE of 3.26 (Table 3.2). This corroborates our previous conclusion\\nthat a model that uses TV and radio expenditures to predict sales is much\\nmore accurate (on the training data) than one that only uses TV spending.\\nFurthermore, given that TV and radio expenditures are used as predictors,\\nthere is no point in also using newspaper spending as a predictor in the\\nmodel. The observant reader may wonder how RSE can increase when\\nnewspaper is added to the model given that RSS must decrease. In general\\nRSE is defined as\\nRSE =\\n9\\n1\\nn −p −1RSS,\\n(3.25) 3.2 Multiple Linear Regression\\n89\\nSales\\nRadio\\nTV\\nFIGURE 3.5. For the Advertising data, a linear regression fit to sales using\\nTV and radio as predictors. From the pattern of the residuals, we can see that\\nthere is a pronounced non-linear relationship in the data. The positive residuals\\n(those visible above the surface), tend to lie along the 45-degree line, where TV\\nand Radio budgets are split evenly. The negative residuals (most not visible), tend\\nto lie away from this line, where budgets are more lopsided.\\nwhich simplifies to (3.15) for a simple linear regression. Thus, models with\\nmore variables can have higher RSE if the decrease in RSS is small relative\\nto the increase in p.\\nIn addition to looking at the RSE and R2 statistics just discussed, it\\ncan be useful to plot the data. Graphical summaries can reveal problems\\nwith a model that are not visible from numerical statistics. For example,\\nFigure 3.5 displays a three-dimensional plot of TV and radio versus sales.\\nWe see that some observations lie above and some observations lie below\\nthe least squares regression plane. In particular, the linear model seems to\\noverestimate sales for instances in which most of the advertising money\\nwas spent exclusively on either TV or radio. It underestimates sales for\\ninstances where the budget was split between the two media. This pro-\\nnounced non-linear pattern suggests a synergy or interaction effect between interaction\\nthe advertising media, whereby combining the media together results in a\\nbigger boost to sales than using any single medium. In Section 3.3.2, we\\nwill discuss extending the linear model to accommodate such synergistic\\neffects through the use of interaction terms.\\nFour: Predictions\\nOnce we have fit the multiple regression model, it is straightforward to\\napply (3.21) in order to predict the response Y on the basis of a set of\\nvalues for the predictors X1, X2, . . . , Xp. However, there are three sorts of\\nuncertainty associated with this prediction.\\n1. The coefficient estimates ˆβ0, ˆβ1, . . . , ˆβp are estimates for β0, β1, . . . , βp.\\nThat is, the least squares plane\\nˆY = ˆβ0 + ˆβ1X1 + · · · + ˆβpXp 90\\n3. Linear Regression\\nis only an estimate for the true population regression plane\\nf(X) = β0 + β1X1 + · · · + βpXp.\\nThe inaccuracy in the coefficient estimates is related to the reducible\\nerror from Chapter 2. We can compute a confidence interval in order\\nto determine how close ˆY will be to f(X).\\n2. Of course, in practice assuming a linear model for f(X) is almost\\nalways an approximation of reality, so there is an additional source of\\npotentially reducible error which we call model bias. So when we use a\\nlinear model, we are in fact estimating the best linear approximation\\nto the true surface. However, here we will ignore this discrepancy,\\nand operate as if the linear model were correct.\\n3. Even if we knew f(X)—that is, even if we knew the true values\\nfor β0, β1, . . . , βp—the response value cannot be predicted perfectly\\nbecause of the random error ϵ in the model (3.20). In Chapter 2, we\\nreferred to this as the irreducible error. How much will Y vary from\\nˆY ? We use prediction intervals to answer this question. Prediction\\nintervals are always wider than confidence intervals, because they\\nincorporate both the error in the estimate for f(X) (the reducible\\nerror) and the uncertainty as to how much an individual point will\\ndiffer from the population regression plane (the irreducible error).\\nWe use a confidence interval to quantify the uncertainty surrounding confidence\\ninterval\\nthe average sales over a large number of cities. For example, given that\\n$100,000 is spent on TV advertising and $20,000 is spent on radio advertising\\nin each city, the 95 % confidence interval is [10,985, 11,528]. We interpret\\nthis to mean that 95 % of intervals of this form will contain the true value of\\nf(X).9 On the other hand, a prediction interval can be used to quantify the prediction\\ninterval\\nuncertainty surrounding sales for a particular city. Given that $100,000 is\\nspent on TV advertising and $20,000 is spent on radio advertising in that city\\nthe 95 % prediction interval is [7,930, 14,580]. We interpret this to mean\\nthat 95 % of intervals of this form will contain the true value of Y for this\\ncity. Note that both intervals are centered at 11,256, but that the prediction\\ninterval is substantially wider than the confidence interval, reflecting the\\nincreased uncertainty about sales for a given city in comparison to the\\naverage sales over many locations.\\n9In other words, if we collect a large number of data sets like the Advertising data\\nset, and we construct a confidence interval for the average sales on the basis of each\\ndata set (given $100,000 in TV and $20,000 in radio advertising), then 95 % of these\\nconfidence intervals will contain the true value of average sales. 3.3 Other Considerations in the Regression Model\\n91\\n3.3\\nOther Considerations in the Regression Model\\n3.3.1\\nQualitative Predictors\\nIn our discussion so far, we have assumed that all variables in our linear\\nregression model are quantitative. But in practice, this is not necessarily\\nthe case; often some predictors are qualitative.\\nFor example, the Credit data set displayed in Figure 3.6 records variables\\nfor a number of credit card holders. The response is balance (average credit\\ncard debt for each individual) and there are several quantitative predictors:\\nage, cards (number of credit cards), education (years of education), income\\n(in thousands of dollars), limit (credit limit), and rating (credit rating).\\nEach panel of Figure 3.6 is a scatterplot for a pair of variables whose iden-\\ntities are given by the corresponding row and column labels. For example,\\nthe scatterplot directly to the right of the word “Balance” depicts balance\\nversus age, while the plot directly to the right of “Age” corresponds to\\nage versus cards. In addition to these quantitative variables, we also have\\nfour qualitative variables: own (house ownership), student (student status),\\nstatus (marital status), and region (East, West or South).\\nPredictors with Only Two Levels\\nSuppose that we wish to investigate differences in credit card balance be-\\ntween those who own a house and those who don’t, ignoring the other vari-\\nables for the moment. If a qualitative predictor (also known as a factor) factor\\nonly has two levels, or possible values, then incorporating it into a regres- level\\nsion model is very simple. We simply create an indicator or dummy variable dummy\\nvariable\\nthat takes on two possible numerical values.10 For example, based on the\\nown variable, we can create a new variable that takes the form\\nxi =\\n=\\n1\\nif ith person owns a house\\n0\\nif ith person does not own a house,\\n(3.26)\\nand use this variable as a predictor in the regression equation. This results\\nin the model\\nyi = β0 + β1xi + ϵi =\\n=\\nβ0 + β1 + ϵi\\nif ith person owns a house\\nβ0 + ϵi\\nif ith person does not.\\n(3.27)\\nNow β0 can be interpreted as the average credit card balance among those\\nwho do not own, β0 + β1 as the average credit card balance among those\\nwho do own their house, and β1 as the average difference in credit card\\nbalance between owners and non-owners.\\nTable 3.7 displays the coefficient estimates and other information asso-\\nciated with the model (3.27). The average credit card debt for non-owners\\nis estimated to be $509.80, whereas owners are estimated to carry $19.73\\nin additional debt for a total of $509.80 + $19.73 = $529.53. However, we\\n10In the machine learning community, the creation of dummy variables to handle\\nqualitative predictors is known as “one-hot encoding”. 92\\n3. Linear Regression\\nBalance\\n20\\n40\\n60\\n80\\n100\\n5\\n10\\n15\\n20\\n2000\\n8000\\n14000\\n0\\n500\\n1500\\n20\\n40\\n60\\n80\\n100\\nAge\\nCards\\n2\\n4\\n6\\n8\\n5\\n10\\n15\\n20\\nEducation\\nIncome\\n50\\n100\\n150\\n2000\\n8000\\n14000\\nLimit\\n0\\n500\\n1500\\n2\\n4\\n6\\n8\\n50\\n100\\n150\\n200\\n600\\n1000\\n200\\n600\\n1000\\nRating\\nFIGURE 3.6. The Credit data set contains information about balance, age,\\ncards, education, income, limit, and rating for a number of potential cus-\\ntomers.\\nnotice that the p-value for the dummy variable is very high. This indicates\\nthat there is no statistical evidence of a difference in average credit card\\nbalance based on house ownership.\\nThe decision to code owners as 1 and non-owners as 0 in (3.27) is ar-\\nbitrary, and has no effect on the regression fit, but does alter the inter-\\npretation of the coefficients. If we had coded non-owners as 1 and own-\\ners as 0, then the estimates for β0 and β1 would have been 529.53 and\\n−19.73, respectively, leading once again to a prediction of credit card debt\\nof $529.53 −$19.73 = $509.80 for non-owners and a prediction of $529.53\\nfor owners. Alternatively, instead of a 0/1 coding scheme, we could create\\na dummy variable\\nxi =\\n=\\n1\\nif ith person owns a house\\n−1\\nif ith person does not own a house\\nand use this variable in the regression equation. This results in the model\\nyi = β0+β1xi+ϵi =\\n=\\nβ0 + β1 + ϵi\\nif ith person owns a house\\nβ0 −β1 + ϵi\\nif ith person does not own a house. 3.3 Other Considerations in the Regression Model\\n93\\nCoefficient\\nStd. error\\nt-statistic\\np-value\\nIntercept\\n509.80\\n33.13\\n15.389\\n< 0.0001\\nown[Yes]\\n19.73\\n46.05\\n0.429\\n0.6690\\nTABLE 3.7. Least squares coefficient estimates associated with the regression\\nof balance onto own in the Credit data set. The linear model is given in (3.27).\\nThat is, ownership is encoded as a dummy variable, as in (3.26).\\nNow β0 can be interpreted as the overall average credit card balance (ig-\\nnoring the house ownership effect), and β1 is the amount by which house\\nowners and non-owners have credit card balances that are above and below\\nthe average, respectively.11 In this example, the estimate for β0 is $519.665,\\nhalfway between the non-owner and owner averages of $509.80 and $529.53.\\nThe estimate for β1 is $9.865, which is half of $19.73, the average difference\\nbetween owners and non-owners. It is important to note that the final pre-\\ndictions for the credit balances of owners and non-owners will be identical\\nregardless of the coding scheme used. The only difference is in the way that\\nthe coefficients are interpreted.\\nQualitative Predictors with More than Two Levels\\nWhen a qualitative predictor has more than two levels, a single dummy\\nvariable cannot represent all possible values. In this situation, we can create\\nadditional dummy variables. For example, for the region variable we create\\ntwo dummy variables. The first could be\\nxi1 =\\n=\\n1\\nif ith person is from the South\\n0\\nif ith person is not from the South,\\n(3.28)\\nand the second could be\\nxi2 =\\n=\\n1\\nif ith person is from the West\\n0\\nif ith person is not from the West.\\n(3.29)\\nThen both of these variables can be used in the regression equation, in\\norder to obtain the model\\nyi = β0+β1xi1+β2xi2+ϵi =\\n\\uf8f1\\n\\uf8f4\\n\\uf8f2\\n\\uf8f4\\n\\uf8f3\\nβ0+β1+ϵi\\nif ith person is from the South\\nβ0+β2+ϵi\\nif ith person is from the West\\nβ0+ϵi\\nif ith person is from the East.\\n(3.30)\\nNow β0 can be interpreted as the average credit card balance for individuals\\nfrom the East, β1 can be interpreted as the difference in the average balance\\nbetween people from the South versus the East, and β2 can be interpreted\\nas the difference in the average balance between those from the West versus\\nthe East. There will always be one fewer dummy variable than the number\\nof levels. The level with no dummy variable—East in this example—is\\nknown as the baseline.\\nbaseline\\n11Technically β0 is half the sum of the average debt for house owners and the average\\ndebt for non-house owners. Hence, β0 is exactly equal to the overall average only if the\\ntwo groups have an equal number of members. 94\\n3. Linear Regression\\nCoefficient\\nStd. error\\nt-statistic\\np-value\\nIntercept\\n531.00\\n46.32\\n11.464\\n< 0.0001\\nregion[South]\\n−12.50\\n56.68\\n−0.221\\n0.8260\\nregion[West]\\n−18.69\\n65.02\\n−0.287\\n0.7740\\nTABLE 3.8. Least squares coefficient estimates associated with the regression of\\nbalance onto region in the Credit data set. The linear model is given in (3.30).\\nThat is, region is encoded via two dummy variables (3.28) and (3.29).\\nFrom Table 3.8, we see that the estimated balance for the baseline, East,\\nis $531.00. It is estimated that those in the South will have $18.69 less\\ndebt than those in the East, and that those in the West will have $12.50\\nless debt than those in the East. However, the p-values associated with the\\ncoefficient estimates for the two dummy variables are very large, suggesting\\nno statistical evidence of a real difference in average credit card balance\\nbetween South and East or between West and East.12 Once again, the\\nlevel selected as the baseline category is arbitrary, and the final predictions\\nfor each group will be the same regardless of this choice. However, the\\ncoefficients and their p-values do depend on the choice of dummy variable\\ncoding. Rather than rely on the individual coefficients, we can use an F-test\\nto test H0 : β1 = β2 = 0; this does not depend on the coding. This F-test\\nhas a p-value of 0.96, indicating that we cannot reject the null hypothesis\\nthat there is no relationship between balance and region.\\nUsing this dummy variable approach presents no difficulties when in-\\ncorporating both quantitative and qualitative predictors. For example, to\\nregress balance on both a quantitative variable such as income and a qual-\\nitative variable such as student, we must simply create a dummy variable\\nfor student and then fit a multiple regression model using income and the\\ndummy variable as predictors for credit card balance.\\nThere are many different ways of coding qualitative variables besides\\nthe dummy variable approach taken here. All of these approaches lead to\\nequivalent model fits, but the coefficients are different and have different\\ninterpretations, and are designed to measure particular contrasts. This topic contrast\\nis beyond the scope of the book.\\n3.3.2\\nExtensions of the Linear Model\\nThe standard linear regression model (3.19) provides interpretable results\\nand works quite well on many real-world problems. However, it makes sev-\\neral highly restrictive assumptions that are often violated in practice. Two\\nof the most important assumptions state that the relationship between the\\npredictors and response are additive and linear. The additivity assumption additive\\nlinear\\nmeans that the association between a predictor Xj and the response Y does\\nnot depend on the values of the other predictors. The linearity assumption\\nstates that the change in the response Y associated with a one-unit change\\nin Xj is constant, regardless of the value of Xj. In later chapters of this\\nbook, we examine a number of sophisticated methods that relax these two\\n12There could still in theory be a difference between South and West, although the\\ndata here does not suggest any difference. 3.3 Other Considerations in the Regression Model\\n95\\nassumptions. Here, we briefly examine some common classical approaches\\nfor extending the linear model.\\nRemoving the Additive Assumption\\nIn our previous analysis of the Advertising data, we concluded that both TV\\nand radio seem to be associated with sales. The linear models that formed\\nthe basis for this conclusion assumed that the effect on sales of increasing\\none advertising medium is independent of the amount spent on the other\\nmedia. For example, the linear model (3.20) states that the average increase\\nin sales associated with a one-unit increase in TV is always β1, regardless\\nof the amount spent on radio.\\nHowever, this simple model may be incorrect. Suppose that spending\\nmoney on radio advertising actually increases the effectiveness of TV ad-\\nvertising, so that the slope term for TV should increase as radio increases.\\nIn this situation, given a fixed budget of $100,000, spending half on radio\\nand half on TV may increase sales more than allocating the entire amount\\nto either TV or to radio. In marketing, this is known as a synergy effect,\\nand in statistics it is referred to as an interaction effect. Figure 3.5 sug-\\ngests that such an effect may be present in the advertising data. Notice\\nthat when levels of either TV or radio are low, then the true sales are lower\\nthan predicted by the linear model. But when advertising is split between\\nthe two media, then the model tends to underestimate sales.\\nConsider the standard linear regression model with two variables,\\nY = β0 + β1X1 + β2X2 + ϵ.\\nAccording to this model, a one-unit increase in X1 is associated with an\\naverage increase in Y of β1 units. Notice that the presence of X2 does\\nnot alter this statement—that is, regardless of the value of X2, a one-\\nunit increase in X1 is associated with a β1-unit increase in Y . One way of\\nextending this model is to include a third predictor, called an interaction\\nterm, which is constructed by computing the product of X1 and X2. This\\nresults in the model\\nY = β0 + β1X1 + β2X2 + β3X1X2 + ϵ.\\n(3.31)\\nHow does inclusion of this interaction term relax the additive assumption?\\nNotice that (3.31) can be rewritten as\\nY\\n=\\nβ0 + (β1 + β3X2)X1 + β2X2 + ϵ\\n(3.32)\\n=\\nβ0 + ˜β1X1 + β2X2 + ϵ\\nwhere ˜β1 = β1 + β3X2. Since ˜β1 is now a function of X2, the association\\nbetween X1 and Y is no longer constant: a change in the value of X2 will\\nchange the association between X1 and Y . A similar argument shows that\\na change in the value of X1 changes the association between X2 and Y .\\nFor example, suppose that we are interested in studying the productiv-\\nity of a factory. We wish to predict the number of units produced on the\\nbasis of the number of production lines and the total number of workers.\\nIt seems likely that the effect of increasing the number of production lines 96\\n3. Linear Regression\\nCoefficient\\nStd. error\\nt-statistic\\np-value\\nIntercept\\n6.7502\\n0.248\\n27.23\\n< 0.0001\\nTV\\n0.0191\\n0.002\\n12.70\\n< 0.0001\\nradio\\n0.0289\\n0.009\\n3.24\\n0.0014\\nTV×radio\\n0.0011\\n0.000\\n20.73\\n< 0.0001\\nTABLE 3.9. For the Advertising data, least squares coefficient estimates asso-\\nciated with the regression of sales onto TV and radio, with an interaction term,\\nas in (3.33).\\nwill depend on the number of workers, since if no workers are available\\nto operate the lines, then increasing the number of lines will not increase\\nproduction. This suggests that it would be appropriate to include an inter-\\naction term between lines and workers in a linear model to predict units.\\nSuppose that when we fit the model, we obtain\\nunits\\n≈\\n1.2 + 3.4 × lines + 0.22 × workers + 1.4 × (lines × workers)\\n=\\n1.2 + (3.4 + 1.4 × workers) × lines + 0.22 × workers.\\nIn other words, adding an additional line will increase the number of units\\nproduced by 3.4 + 1.4 × workers. Hence the more workers we have, the\\nstronger will be the effect of lines.\\nWe now return to the Advertising example. A linear model that uses\\nradio, TV, and an interaction between the two to predict sales takes the\\nform\\nsales\\n=\\nβ0 + β1 × TV + β2 × radio + β3 × (radio × TV) + ϵ\\n=\\nβ0 + (β1 + β3 × radio) × TV + β2 × radio + ϵ.\\n(3.33)\\nWe can interpret β3 as the increase in the effectiveness of TV advertising\\nassociated with a one-unit increase in radio advertising (or vice-versa). The\\ncoefficients that result from fitting the model (3.33) are given in Table 3.9.\\nThe results in Table 3.9 strongly suggest that the model that includes the\\ninteraction term is superior to the model that contains only main effects. main effect\\nThe p-value for the interaction term, TV×radio, is extremely low, indicating\\nthat there is strong evidence for Ha : β3 ̸= 0. In other words, it is clear that\\nthe true relationship is not additive. The R2 for the model (3.33) is 96.8 %,\\ncompared to only 89.7 % for the model that predicts sales using TV and\\nradio without an interaction term. This means that (96.8 −89.7)/(100 −\\n89.7) = 69 % of the variability in sales that remains after fitting the ad-\\nditive model has been explained by the interaction term. The coefficient\\nestimates in Table 3.9 suggest that an increase in TV advertising of $1,000 is\\nassociated with increased sales of (ˆβ1+ ˆβ3×radio)×1,000 = 19+1.1×radio\\nunits. And an increase in radio advertising of $1,000 will be associated with\\nan increase in sales of (ˆβ2 + ˆβ3 × TV) × 1,000 = 29 + 1.1 × TV units.\\nIn this example, the p-values associated with TV, radio, and the interac-\\ntion term all are statistically significant (Table 3.9), and so it is obvious\\nthat all three variables should be included in the model. However, it is\\nsometimes the case that an interaction term has a very small p-value, but\\nthe associated main effects (in this case, TV and radio) do not. The hier-\\narchical principle states that if we include an interaction in a model, we hierarchical\\nprinciple 3.3 Other Considerations in the Regression Model\\n97\\nshould also include the main effects, even if the p-values associated with\\ntheir coefficients are not significant. In other words, if the interaction be-\\ntween X1 and X2 seems important, then we should include both X1 and\\nX2 in the model even if their coefficient estimates have large p-values. The\\nrationale for this principle is that if X1 × X2 is related to the response,\\nthen whether or not the coefficients of X1 or X2 are exactly zero is of lit-\\ntle interest. Also X1 × X2 is typically correlated with X1 and X2, and so\\nleaving them out tends to alter the meaning of the interaction.\\nIn the previous example, we considered an interaction between TV and\\nradio, both of which are quantitative variables. However, the concept of\\ninteractions applies just as well to qualitative variables, or to a combination\\nof quantitative and qualitative variables. In fact, an interaction between\\na qualitative variable and a quantitative variable has a particularly nice\\ninterpretation. Consider the Credit data set from Section 3.3.1, and suppose\\nthat we wish to predict balance using the income (quantitative) and student\\n(qualitative) variables. In the absence of an interaction term, the model\\ntakes the form\\nbalancei\\n≈\\nβ0 + β1 × incomei +\\n=\\nβ2\\nif ith person is a student\\n0\\nif ith person is not a student\\n=\\nβ1 × incomei +\\n=\\nβ0 + β2\\nif ith person is a student\\nβ0\\nif ith person is not a student.\\n(3.34)\\nNotice that this amounts to fitting two parallel lines to the data, one for\\nstudents and one for non-students. The lines for students and non-students\\nhave different intercepts, β0 + β2 versus β0, but the same slope, β1. This\\nis illustrated in the left-hand panel of Figure 3.7. The fact that the lines\\nare parallel means that the average effect on balance of a one-unit increase\\nin income does not depend on whether or not the individual is a student.\\nThis represents a potentially serious limitation of the model, since in fact a\\nchange in income may have a very different effect on the credit card balance\\nof a student versus a non-student.\\nThis limitation can be addressed by adding an interaction variable, cre-\\nated by multiplying income with the dummy variable for student. Our\\nmodel now becomes\\nbalancei\\n≈\\nβ0 + β1 × incomei +\\n=\\nβ2 + β3 × incomei\\nif student\\n0\\nif not student\\n=\\n=\\n(β0 + β2) + (β1 + β3) × incomei\\nif student\\nβ0 + β1 × incomei\\nif not student.\\n(3.35)\\nOnce again, we have two different regression lines for the students and\\nthe non-students. But now those regression lines have different intercepts,\\nβ0+β2 versus β0, as well as different slopes, β1+β3 versus β1. This allows for\\nthe possibility that changes in income may affect the credit card balances\\nof students and non-students differently. The right-hand panel of Figure 3.7 98\\n3. Linear Regression\\n0\\n50\\n100\\n150\\n200\\n600\\n1000\\n1400\\nIncome\\nBalance\\n0\\n50\\n100\\n150\\n200\\n600\\n1000\\n1400\\nIncome\\nBalance\\nstudent\\nnon−student\\nFIGURE 3.7. For the Credit data, the least squares lines are shown for pre-\\ndiction of balance from income for students and non-students. Left: The model\\n(3.34) was fit. There is no interaction between income and student. Right: The\\nmodel (3.35) was fit. There is an interaction term between income and student.\\nshows the estimated relationships between income and balance for students\\nand non-students in the model (3.35). We note that the slope for students\\nis lower than the slope for non-students. This suggests that increases in\\nincome are associated with smaller increases in credit card balance among\\nstudents as compared to non-students.\\nNon-linear Relationships\\nAs discussed previously, the linear regression model (3.19) assumes a linear\\nrelationship between the response and predictors. But in some cases, the\\ntrue relationship between the response and the predictors may be non-\\nlinear. Here we present a very simple way to directly extend the linear model\\nto accommodate non-linear relationships, using polynomial regression. In polynomial\\nregression\\nlater chapters, we will present more complex approaches for performing\\nnon-linear fits in more general settings.\\nConsider Figure 3.8, in which the mpg (gas mileage in miles per gallon)\\nversus horsepower is shown for a number of cars in the Auto data set. The\\norange line represents the linear regression fit. There is a pronounced rela-\\ntionship between mpg and horsepower, but it seems clear that this relation-\\nship is in fact non-linear: the data suggest a curved relationship. A simple\\napproach for incorporating non-linear associations in a linear model is to\\ninclude transformed versions of the predictors. For example, the points in\\nFigure 3.8 seem to have a quadratic shape, suggesting that a model of the quadratic\\nform\\nmpg = β0 + β1 × horsepower + β2 × horsepower2 + ϵ\\n(3.36)\\nmay provide a better fit. Equation 3.36 involves predicting mpg using a\\nnon-linear function of horsepower. But it is still a linear model! That is,\\n(3.36) is simply a multiple linear regression model with X1 = horsepower\\nand X2 = horsepower2. So we can use standard linear regression software to\\nestimate β0, β1, and β2 in order to produce a non-linear fit. The blue curve\\nin Figure 3.8 shows the resulting quadratic fit to the data. The quadratic 3.3 Other Considerations in the Regression Model\\n99\\n50\\n100\\n150\\n200\\n10\\n20\\n30\\n40\\n50\\nHorsepower\\nMiles per gallon\\nLinear\\nDegree 2\\nDegree 5\\nFIGURE 3.8. The Auto data set. For a number of cars, mpg and horsepower are\\nshown. The linear regression fit is shown in orange. The linear regression fit for a\\nmodel that includes horsepower2 is shown as a blue curve. The linear regression\\nfit for a model that includes all polynomials of horsepower up to fifth-degree is\\nshown in green.\\nCoefficient\\nStd. error\\nt-statistic\\np-value\\nIntercept\\n56.9001\\n1.8004\\n31.6\\n< 0.0001\\nhorsepower\\n−0.4662\\n0.0311\\n−15.0\\n< 0.0001\\nhorsepower2\\n0.0012\\n0.0001\\n10.1\\n< 0.0001\\nTABLE 3.10. For the Auto data set, least squares coefficient estimates associated\\nwith the regression of mpg onto horsepower and horsepower2.\\nfit appears to be substantially better than the fit obtained when just the\\nlinear term is included. The R2 of the quadratic fit is 0.688, compared to\\n0.606 for the linear fit, and the p-value in Table 3.10 for the quadratic term\\nis highly significant.\\nIf including horsepower2 led to such a big improvement in the model, why\\nnot include horsepower3, horsepower4, or even horsepower5? The green curve\\nin Figure 3.8 displays the fit that results from including all polynomials up\\nto fifth degree in the model (3.36). The resulting fit seems unnecessarily\\nwiggly—that is, it is unclear that including the additional terms really has\\nled to a better fit to the data.\\nThe approach that we have just described for extending the linear model\\nto accommodate non-linear relationships is known as polynomial regres-\\nsion, since we have included polynomial functions of the predictors in the\\nregression model. We further explore this approach and other non-linear\\nextensions of the linear model in Chapter 7. 100\\n3. Linear Regression\\n3.3.3\\nPotential Problems\\nWhen we fit a linear regression model to a particular data set, many prob-\\nlems may occur. Most common among these are the following:\\n1. Non-linearity of the response-predictor relationships.\\n2. Correlation of error terms.\\n3. Non-constant variance of error terms.\\n4. Outliers.\\n5. High-leverage points.\\n6. Collinearity.\\nIn practice, identifying and overcoming these problems is as much an\\nart as a science. Many pages in countless books have been written on this\\ntopic. Since the linear regression model is not our primary focus here, we\\nwill provide only a brief summary of some key points.\\n1. Non-linearity of the Data\\n5\\n10\\n15\\n20\\n25\\n30\\n−15\\n−10\\n−5\\n0\\n5\\n10\\n15\\n20\\nFitted values\\nResiduals\\nResidual Plot for Linear Fit\\n323\\n330\\n334\\n15\\n20\\n25\\n30\\n35\\n−15\\n−10\\n−5\\n0\\n5\\n10\\n15\\nFitted values\\nResiduals\\nResidual Plot for Quadratic Fit\\n334\\n323\\n155\\nFIGURE 3.9. Plots of residuals versus predicted (or fitted) values for the Auto\\ndata set. In each plot, the red line is a smooth fit to the residuals, intended to make\\nit easier to identify a trend. Left: A linear regression of mpg on horsepower. A\\nstrong pattern in the residuals indicates non-linearity in the data. Right: A linear\\nregression of mpg on horsepower and horsepower2. There is little pattern in the\\nresiduals.\\nThe linear regression model assumes that there is a straight-line rela-\\ntionship between the predictors and the response. If the true relationship\\nis far from linear, then virtually all of the conclusions that we draw from\\nthe fit are suspect. In addition, the prediction accuracy of the model can\\nbe significantly reduced.\\nResidual plots are a useful graphical tool for identifying non-linearity. residual plot\\nGiven a simple linear regression model, we can plot the residuals, ei = 3.3 Other Considerations in the Regression Model\\n101\\nyi −ˆyi, versus the predictor xi. In the case of a multiple regression model,\\nsince there are multiple predictors, we instead plot the residuals versus\\nthe predicted (or fitted) values ˆyi. Ideally, the residual plot will show no fitted\\ndiscernible pattern. The presence of a pattern may indicate a problem with\\nsome aspect of the linear model.\\nThe left panel of Figure 3.9 displays a residual plot from the linear re-\\ngression of mpg onto horsepower on the Auto data set that was illustrated in\\nFigure 3.8. The red line is a smooth fit to the residuals, which is displayed in\\norder to make it easier to identify any trends. The residuals exhibit a clear\\nU-shape, which provides a strong indication of non-linearity in the data.\\nIn contrast, the right-hand panel of Figure 3.9 displays the residual plot\\nthat results from the model (3.36), which contains a quadratic term. There\\nappears to be little pattern in the residuals, suggesting that the quadratic\\nterm improves the fit to the data.\\nIf the residual plot indicates that there are non-linear associations in the\\ndata, then a simple approach is to use non-linear transformations of the\\npredictors, such as log X,\\n√\\nX, and X2, in the regression model. In the\\nlater chapters of this book, we will discuss other more advanced non-linear\\napproaches for addressing this issue.\\n2. Correlation of Error Terms\\nAn important assumption of the linear regression model is that the error\\nterms, ϵ1, ϵ2, . . . , ϵn, are uncorrelated. What does this mean? For instance,\\nif the errors are uncorrelated, then the fact that ϵi is positive provides\\nlittle or no information about the sign of ϵi+1. The standard errors that\\nare computed for the estimated regression coefficients or the fitted values\\nare based on the assumption of uncorrelated error terms. If in fact there is\\ncorrelation among the error terms, then the estimated standard errors will\\ntend to underestimate the true standard errors. As a result, confidence and\\nprediction intervals will be narrower than they should be. For example,\\na 95 % confidence interval may in reality have a much lower probability\\nthan 0.95 of containing the true value of the parameter. In addition, p-\\nvalues associated with the model will be lower than they should be; this\\ncould cause us to erroneously conclude that a parameter is statistically\\nsignificant. In short, if the error terms are correlated, we may have an\\nunwarranted sense of confidence in our model.\\nAs an extreme example, suppose we accidentally doubled our data, lead-\\ning to observations and error terms identical in pairs. If we ignored this, our\\nstandard error calculations would be as if we had a sample of size 2n, when\\nin fact we have only n samples. Our estimated parameters would be the\\nsame for the 2n samples as for the n samples, but the confidence intervals\\nwould be narrower by a factor of\\n√\\n2!\\nWhy might correlations among the error terms occur? Such correlations\\nfrequently occur in the context of time series data, which consists of ob- time series\\nservations for which measurements are obtained at discrete points in time.\\nIn many cases, observations that are obtained at adjacent time points will\\nhave positively correlated errors. In order to determine if this is the case for\\na given data set, we can plot the residuals from our model as a function of\\ntime. If the errors are uncorrelated, then there should be no discernible pat- 102\\n3. Linear Regression\\n0\\n20\\n40\\n60\\n80\\n100\\n−3\\n−1\\n0\\n1\\n2\\n3\\nρ=0.0\\nResidual\\n0\\n20\\n40\\n60\\n80\\n100\\n−4\\n−2\\n0\\n1\\n2\\nρ=0.5\\nResidual\\n0\\n20\\n40\\n60\\n80\\n100\\n−1.5\\n−0.5\\n0.5\\n1.5\\nρ=0.9\\nResidual\\nObservation\\nFIGURE 3.10. Plots of residuals from simulated time series data sets generated\\nwith differing levels of correlation ρ between error terms for adjacent time points.\\ntern. On the other hand, if the error terms are positively correlated, then\\nwe may see tracking in the residuals—that is, adjacent residuals may have tracking\\nsimilar values. Figure 3.10 provides an illustration. In the top panel, we see\\nthe residuals from a linear regression fit to data generated with uncorre-\\nlated errors. There is no evidence of a time-related trend in the residuals.\\nIn contrast, the residuals in the bottom panel are from a data set in which\\nadjacent errors had a correlation of 0.9. Now there is a clear pattern in the\\nresiduals—adjacent residuals tend to take on similar values. Finally, the\\ncenter panel illustrates a more moderate case in which the residuals had a\\ncorrelation of 0.5. There is still evidence of tracking, but the pattern is less\\nclear.\\nMany methods have been developed to properly take account of corre-\\nlations in the error terms in time series data. Correlation among the error\\nterms can also occur outside of time series data. For instance, consider a\\nstudy in which individuals’ heights are predicted from their weights. The\\nassumption of uncorrelated errors could be violated if some of the indi-\\nviduals in the study are members of the same family, eat the same diet,\\nor have been exposed to the same environmental factors. In general, the\\nassumption of uncorrelated errors is extremely important for linear regres-\\nsion as well as for other statistical methods, and good experimental design\\nis crucial in order to mitigate the risk of such correlations. 3.3 Other Considerations in the Regression Model\\n103\\n10\\n15\\n20\\n25\\n30\\n−10\\n−5\\n0\\n5\\n10\\n15\\nFitted values\\nResiduals\\nResponse Y\\n998\\n975\\n845\\n2.4\\n2.6\\n2.8\\n3.0\\n3.2\\n3.4\\n−0.8\\n−0.6\\n−0.4\\n−0.2\\n0.0\\n0.2\\n0.4\\nFitted values\\nResiduals\\nResponse log(Y)\\n437\\n671\\n605\\nFIGURE 3.11. Residual plots. In each plot, the red line is a smooth fit to the\\nresiduals, intended to make it easier to identify a trend. The blue lines track the\\nouter quantiles of the residuals, and emphasize patterns. Left: The funnel shape\\nindicates heteroscedasticity. Right: The response has been log transformed, and\\nthere is now no evidence of heteroscedasticity.\\n3. Non-constant Variance of Error Terms\\nAnother important assumption of the linear regression model is that the\\nerror terms have a constant variance, Var(ϵi) = σ2. The standard errors,\\nconfidence intervals, and hypothesis tests associated with the linear model\\nrely upon this assumption.\\nUnfortunately, it is often the case that the variances of the error terms are\\nnon-constant. For instance, the variances of the error terms may increase\\nwith the value of the response. One can identify non-constant variances in\\nthe errors, or heteroscedasticity, from the presence of a funnel shape in hetero-\\nscedasticity\\nthe residual plot. An example is shown in the left-hand panel of Figure 3.11,\\nin which the magnitude of the residuals tends to increase with the fitted\\nvalues. When faced with this problem, one possible solution is to trans-\\nform the response Y using a concave function such as log Y or\\n√\\nY . Such\\na transformation results in a greater amount of shrinkage of the larger re-\\nsponses, leading to a reduction in heteroscedasticity. The right-hand panel\\nof Figure 3.11 displays the residual plot after transforming the response\\nusing log Y . The residuals now appear to have constant variance, though\\nthere is some evidence of a slight non-linear relationship in the data.\\nSometimes we have a good idea of the variance of each response. For\\nexample, the ith response could be an average of ni raw observations. If\\neach of these raw observations is uncorrelated with variance σ2, then their\\naverage has variance σ2\\ni = σ2/ni. In this case a simple remedy is to fit our\\nmodel by weighted least squares, with weights proportional to the inverse weighted\\nleast squares\\nvariances—i.e. wi = ni in this case. Most linear regression software allows\\nfor observation weights.\\n4. Outliers\\nAn outlier is a point for which yi is far from the value predicted by the outlier 104\\n3. Linear Regression\\n−2\\n−1\\n0\\n1\\n2\\n−4\\n−2\\n0\\n2\\n4\\n6\\n20\\n−2\\n0\\n2\\n4\\n6\\n−1\\n0\\n1\\n2\\n3\\n4\\nFitted Values\\nResiduals\\n20\\n−2\\n0\\n2\\n4\\n6\\n0\\n2\\n4\\n6\\nFitted Values\\nStudentized Residuals\\n20\\nX\\nY\\nFIGURE 3.12. Left: The least squares regression line is shown in red, and the\\nregression line after removing the outlier is shown in blue. Center: The residual\\nplot clearly identifies the outlier. Right: The outlier has a studentized residual of\\n6; typically we expect values between −3 and 3.\\nmodel. Outliers can arise for a variety of reasons, such as incorrect recording\\nof an observation during data collection.\\nThe red point (observation 20) in the left-hand panel of Figure 3.12\\nillustrates a typical outlier. The red solid line is the least squares regression\\nfit, while the blue dashed line is the least squares fit after removal of the\\noutlier. In this case, removing the outlier has little effect on the least squares\\nline: it leads to almost no change in the slope, and a miniscule reduction\\nin the intercept. It is typical for an outlier that does not have an unusual\\npredictor value to have little effect on the least squares fit. However, even\\nif an outlier does not have much effect on the least squares fit, it can cause\\nother problems. For instance, in this example, the RSE is 1.09 when the\\noutlier is included in the regression, but it is only 0.77 when the outlier\\nis removed. Since the RSE is used to compute all confidence intervals and\\np-values, such a dramatic increase caused by a single data point can have\\nimplications for the interpretation of the fit. Similarly, inclusion of the\\noutlier causes the R2 to decline from 0.892 to 0.805.\\nResidual plots can be used to identify outliers. In this example, the out-\\nlier is clearly visible in the residual plot illustrated in the center panel of\\nFigure 3.12. But in practice, it can be difficult to decide how large a resid-\\nual needs to be before we consider the point to be an outlier. To address\\nthis problem, instead of plotting the residuals, we can plot the studentized\\nresiduals, computed by dividing each residual ei by its estimated standard studentized\\nresidual\\nerror. Observations whose studentized residuals are greater than 3 in abso-\\nlute value are possible outliers. In the right-hand panel of Figure 3.12, the\\noutlier’s studentized residual exceeds 6, while all other observations have\\nstudentized residuals between −2 and 2.\\nIf we believe that an outlier has occurred due to an error in data collec-\\ntion or recording, then one solution is to simply remove the observation.\\nHowever, care should be taken, since an outlier may instead indicate a\\ndeficiency with the model, such as a missing predictor.\\n5. High Leverage Points\\nWe just saw that outliers are observations for which the response yi is\\nunusual given the predictor xi. In contrast, observations with high leverage high\\nleverage\\nhave an unusual value for xi. For example, observation 41 in the left-hand 3.3 Other Considerations in the Regression Model\\n105\\n−2\\n−1\\n0\\n1\\n2\\n3\\n4\\n0\\n5\\n10\\n20\\n41\\n−2\\n−1\\n0\\n1\\n2\\n−2\\n−1\\n0\\n1\\n2\\n0.00\\n0.05\\n0.10\\n0.15\\n0.20\\n0.25\\n−1\\n0\\n1\\n2\\n3\\n4\\n5\\nLeverage\\nStudentized Residuals\\n20\\n41\\nX\\nY\\nX1\\nX2\\nFIGURE 3.13. Left: Observation 41 is a high leverage point, while 20 is not.\\nThe red line is the fit to all the data, and the blue line is the fit with observation\\n41 removed. Center: The red observation is not unusual in terms of its X1 value\\nor its X2 value, but still falls outside the bulk of the data, and hence has high\\nleverage. Right: Observation 41 has a high leverage and a high residual.\\npanel of Figure 3.13 has high leverage, in that the predictor value for this\\nobservation is large relative to the other observations. (Note that the data\\ndisplayed in Figure 3.13 are the same as the data displayed in Figure 3.12,\\nbut with the addition of a single high leverage observation.) The red solid\\nline is the least squares fit to the data, while the blue dashed line is the\\nfit produced when observation 41 is removed. Comparing the left-hand\\npanels of Figures 3.12 and 3.13, we observe that removing the high leverage\\nobservation has a much more substantial impact on the least squares line\\nthan removing the outlier. In fact, high leverage observations tend to have\\na sizable impact on the estimated regression line. It is cause for concern if\\nthe least squares line is heavily affected by just a couple of observations,\\nbecause any problems with these points may invalidate the entire fit. For\\nthis reason, it is important to identify high leverage observations.\\nIn a simple linear regression, high leverage observations are fairly easy to\\nidentify, since we can simply look for observations for which the predictor\\nvalue is outside of the normal range of the observations. But in a multiple\\nlinear regression with many predictors, it is possible to have an observation\\nthat is well within the range of each individual predictor’s values, but that\\nis unusual in terms of the full set of predictors. An example is shown in\\nthe center panel of Figure 3.13, for a data set with two predictors, X1 and\\nX2. Most of the observations’ predictor values fall within the blue dashed\\nellipse, but the red observation is well outside of this range. But neither its\\nvalue for X1 nor its value for X2 is unusual. So if we examine just X1 or\\njust X2, we will fail to notice this high leverage point. This problem is more\\npronounced in multiple regression settings with more than two predictors,\\nbecause then there is no simple way to plot all dimensions of the data\\nsimultaneously.\\nIn order to quantify an observation’s leverage, we compute the leverage\\nstatistic. A large value of this statistic indicates an observation with high leverage\\nstatistic\\nleverage. For a simple linear regression,\\nhi = 1\\nn +\\n(xi −¯x)2\\n)n\\ni′=1(xi′ −¯x)2 .\\n(3.37) 106\\n3. Linear Regression\\n2000\\n4000\\n6000\\n8000\\n12000\\n30\\n40\\n50\\n60\\n70\\n80\\nLimit\\nAge\\n2000\\n4000\\n6000\\n8000\\n12000\\n200\\n400\\n600\\n800\\nLimit\\nRating\\nFIGURE 3.14. Scatterplots of the observations from the Credit data set. Left:\\nA plot of age versus limit. These two variables are not collinear. Right: A plot\\nof rating versus limit. There is high collinearity.\\nIt is clear from this equation that hi increases with the distance of xi from ¯x.\\nThere is a simple extension of hi to the case of multiple predictors, though\\nwe do not provide the formula here. The leverage statistic hi is always\\nbetween 1/n and 1, and the average leverage for all the observations is\\nalways equal to (p + 1)/n. So if a given observation has a leverage statistic\\nthat greatly exceeds (p+1)/n, then we may suspect that the corresponding\\npoint has high leverage.\\nThe right-hand panel of Figure 3.13 provides a plot of the studentized\\nresiduals versus hi for the data in the left-hand panel of Figure 3.13. Ob-\\nservation 41 stands out as having a very high leverage statistic as well as a\\nhigh studentized residual. In other words, it is an outlier as well as a high\\nleverage observation. This is a particularly dangerous combination! This\\nplot also reveals the reason that observation 20 had relatively little effect\\non the least squares fit in Figure 3.12: it has low leverage.\\n6. Collinearity\\nCollinearity refers to the situation in which two or more predictor variables collinearity\\nare closely related to one another. The concept of collinearity is illustrated\\nin Figure 3.14 using the Credit data set. In the left-hand panel of Fig-\\nure 3.14, the two predictors limit and age appear to have no obvious rela-\\ntionship. In contrast, in the right-hand panel of Figure 3.14, the predictors\\nlimit and rating are very highly correlated with each other, and we say\\nthat they are collinear. The presence of collinearity can pose problems in\\nthe regression context, since it can be difficult to separate out the indi-\\nvidual effects of collinear variables on the response. In other words, since\\nlimit and rating tend to increase or decrease together, it can be difficult to\\ndetermine how each one separately is associated with the response, balance.\\nFigure 3.15 illustrates some of the difficulties that can result from collinear-\\nity. The left-hand panel of Figure 3.15 is a contour plot of the RSS (3.22)\\nassociated with different possible coefficient estimates for the regression\\nof balance on limit and age. Each ellipse represents a set of coefficients 3.3 Other Considerations in the Regression Model\\n107\\n 21.25 \\n 21.5 \\n 21.8 \\n0.16\\n0.17\\n0.18\\n0.19\\n−5\\n−4\\n−3\\n−2\\n−1\\n0\\n 21.5 \\n 21.8 \\n−0.1\\n0.0\\n0.1\\n0.2\\n0\\n1\\n2\\n3\\n4\\n5\\nβLimit\\nβLimit\\nβAge\\nβRating\\nFIGURE 3.15. Contour plots for the RSS values as a function of the parameters\\nβ for various regressions involving the Credit data set. In each plot, the black\\ndots represent the coefficient values corresponding to the minimum RSS. Left:\\nA contour plot of RSS for the regression of balance onto age and limit. The\\nminimum value is well defined. Right: A contour plot of RSS for the regression\\nof balance onto rating and limit. Because of the collinearity, there are many\\npairs (βLimit, βRating) with a similar value for RSS.\\nthat correspond to the same RSS, with ellipses nearest to the center tak-\\ning on the lowest values of RSS. The black dots and associated dashed\\nlines represent the coefficient estimates that result in the smallest possible\\nRSS—in other words, these are the least squares estimates. The axes for\\nlimit and age have been scaled so that the plot includes possible coeffi-\\ncient estimates that are up to four standard errors on either side of the\\nleast squares estimates. Thus the plot includes all plausible values for the\\ncoefficients. For example, we see that the true limit coefficient is almost\\ncertainly somewhere between 0.15 and 0.20.\\nIn contrast, the right-hand panel of Figure 3.15 displays contour plots\\nof the RSS associated with possible coefficient estimates for the regression\\nof balance onto limit and rating, which we know to be highly collinear.\\nNow the contours run along a narrow valley; there is a broad range of\\nvalues for the coefficient estimates that result in equal values for RSS.\\nHence a small change in the data could cause the pair of coefficient values\\nthat yield the smallest RSS—that is, the least squares estimates—to move\\nanywhere along this valley. This results in a great deal of uncertainty in the\\ncoefficient estimates. Notice that the scale for the limit coefficient now runs\\nfrom roughly −0.2 to 0.2; this is an eight-fold increase over the plausible\\nrange of the limit coefficient in the regression with age. Interestingly, even\\nthough the limit and rating coefficients now have much more individual\\nuncertainty, they will almost certainly lie somewhere in this contour valley.\\nFor example, we would not expect the true value of the limit and rating\\ncoefficients to be −0.1 and 1 respectively, even though such a value is\\nplausible for each coefficient individually.\\nSince collinearity reduces the accuracy of the estimates of the regression\\ncoefficients, it causes the standard error for ˆβj to grow. Recall that the\\nt-statistic for each predictor is calculated by dividing ˆβj by its standard 108\\n3. Linear Regression\\nCoefficient\\nStd. error\\nt-statistic\\np-value\\nIntercept\\n−173.411\\n43.828\\n−3.957\\n< 0.0001\\nModel 1\\nage\\n−2.292\\n0.672\\n−3.407\\n0.0007\\nlimit\\n0.173\\n0.005\\n34.496\\n< 0.0001\\nIntercept\\n−377.537\\n45.254\\n−8.343\\n< 0.0001\\nModel 2\\nrating\\n2.202\\n0.952\\n2.312\\n0.0213\\nlimit\\n0.025\\n0.064\\n0.384\\n0.7012\\nTABLE 3.11. The results for two multiple regression models involving the Credit\\ndata set are shown. Model 1 is a regression of balance on age and limit, and\\nModel 2 a regression of balance on rating and limit. The standard error of\\nˆβlimit increases 12-fold in the second regression, due to collinearity.\\nerror. Consequently, collinearity results in a decline in the t-statistic. As a\\nresult, in the presence of collinearity, we may fail to reject H0 : βj = 0. This\\nmeans that the power of the hypothesis test—the probability of correctly power\\ndetecting a non-zero coefficient—is reduced by collinearity.\\nTable 3.11 compares the coefficient estimates obtained from two separate\\nmultiple regression models. The first is a regression of balance on age and\\nlimit, and the second is a regression of balance on rating and limit. In the\\nfirst regression, both age and limit are highly significant with very small p-\\nvalues. In the second, the collinearity between limit and rating has caused\\nthe standard error for the limit coefficient estimate to increase by a factor\\nof 12 and the p-value to increase to 0.701. In other words, the importance\\nof the limit variable has been masked due to the presence of collinearity.\\nTo avoid such a situation, it is desirable to identify and address potential\\ncollinearity problems while fitting the model.\\nA simple way to detect collinearity is to look at the correlation matrix\\nof the predictors. An element of this matrix that is large in absolute value\\nindicates a pair of highly correlated variables, and therefore a collinearity\\nproblem in the data. Unfortunately, not all collinearity problems can be\\ndetected by inspection of the correlation matrix: it is possible for collinear-\\nity to exist between three or more variables even if no pair of variables\\nhas a particularly high correlation. We call this situation multicollinearity. multi-\\ncollinearity\\nInstead of inspecting the correlation matrix, a better way to assess multi-\\ncollinearity is to compute the variance inflation factor (VIF). The VIF is variance\\ninflation\\nfactor\\nthe ratio of the variance of ˆβj when fitting the full model divided by the\\nvariance of ˆβj if fit on its own. The smallest possible value for VIF is 1,\\nwhich indicates the complete absence of collinearity. Typically in practice\\nthere is a small amount of collinearity among the predictors. As a rule of\\nthumb, a VIF value that exceeds 5 or 10 indicates a problematic amount of\\ncollinearity. The VIF for each variable can be computed using the formula\\nVIF(ˆβj) =\\n1\\n1 −R2\\nXj|X−j\\n,\\nwhere R2\\nXj|X−j is the R2 from a regression of Xj onto all of the other\\npredictors. If R2\\nXj|X−j is close to one, then collinearity is present, and so\\nthe VIF will be large. 3.4 The Marketing Plan\\n109\\nIn the Credit data, a regression of balance on age, rating, and limit\\nindicates that the predictors have VIF values of 1.01, 160.67, and 160.59.\\nAs we suspected, there is considerable collinearity in the data!\\nWhen faced with the problem of collinearity, there are two simple solu-\\ntions. The first is to drop one of the problematic variables from the regres-\\nsion. This can usually be done without much compromise to the regression\\nfit, since the presence of collinearity implies that the information that this\\nvariable provides about the response is redundant in the presence of the\\nother variables. For instance, if we regress balance onto age and limit,\\nwithout the rating predictor, then the resulting VIF values are close to\\nthe minimum possible value of 1, and the R2 drops from 0.754 to 0.75.\\nSo dropping rating from the set of predictors has effectively solved the\\ncollinearity problem without compromising the fit. The second solution is\\nto combine the collinear variables together into a single predictor. For in-\\nstance, we might take the average of standardized versions of limit and\\nrating in order to create a new variable that measures credit worthiness.\\n3.4\\nThe Marketing Plan\\nWe now briefly return to the seven questions about the Advertising data\\nthat we set out to answer at the beginning of this chapter.\\n1. Is there a relationship between sales and advertising budget?\\nThis question can be answered by fitting a multiple regression model\\nof sales onto TV, radio, and newspaper, as in (3.20), and testing the\\nhypothesis H0 : βTV = βradio = βnewspaper = 0. In Section 3.2.2,\\nwe showed that the F-statistic can be used to determine whether\\nor not we should reject this null hypothesis. In this case the p-value\\ncorresponding to the F-statistic in Table 3.6 is very low, indicating\\nclear evidence of a relationship between advertising and sales.\\n2. How strong is the relationship?\\nWe discussed two measures of model accuracy in Section 3.1.3. First,\\nthe RSE estimates the standard deviation of the response from the\\npopulation regression line. For the Advertising data, the RSE is 1.69\\nunits while the mean value for the response is 14.022, indicating a\\npercentage error of roughly 12 %. Second, the R2 statistic records\\nthe percentage of variability in the response that is explained by\\nthe predictors. The predictors explain almost 90 % of the variance in\\nsales. The RSE and R2 statistics are displayed in Table 3.6.\\n3. Which media are associated with sales?\\nTo answer this question, we can examine the p-values associated with\\neach predictor’s t-statistic (Section 3.1.2). In the multiple linear re-\\ngression displayed in Table 3.4, the p-values for TV and radio are low,\\nbut the p-value for newspaper is not. This suggests that only TV and\\nradio are related to sales. In Chapter 6 we explore this question in\\ngreater detail. 110\\n3. Linear Regression\\n4. How large is the association between each medium and sales?\\nWe saw in Section 3.1.2 that the standard error of ˆβj can be used to\\nconstruct confidence intervals for βj. For the Advertising data, we\\ncan use the results in Table 3.4 to compute the 95 % confidence inter-\\nvals for the coefficients in a multiple regression model using all three\\nmedia budgets as predictors. The confidence intervals are as follows:\\n(0.043, 0.049) for TV, (0.172, 0.206) for radio, and (−0.013, 0.011) for\\nnewspaper. The confidence intervals for TV and radio are narrow and\\nfar from zero, providing evidence that these media are related to\\nsales. But the interval for newspaper includes zero, indicating that\\nthe variable is not statistically significant given the values of TV and\\nradio.\\nWe saw in Section 3.3.3 that collinearity can result in very wide stan-\\ndard errors. Could collinearity be the reason that the confidence in-\\nterval associated with newspaper is so wide? The VIF scores are 1.005,\\n1.145, and 1.145 for TV, radio, and newspaper, suggesting no evidence\\nof collinearity.\\nIn order to assess the association of each medium individually on\\nsales, we can perform three separate simple linear regressions. Re-\\nsults are shown in Tables 3.1 and 3.3. There is evidence of an ex-\\ntremely strong association between TV and sales and between radio\\nand sales. There is evidence of a mild association between newspaper\\nand sales, when the values of TV and radio are ignored.\\n5. How accurately can we predict future sales?\\nThe response can be predicted using (3.21). The accuracy associ-\\nated with this estimate depends on whether we wish to predict an\\nindividual response, Y = f(X) + ϵ, or the average response, f(X)\\n(Section 3.2.2). If the former, we use a prediction interval, and if\\nthe latter, we use a confidence interval. Prediction intervals will al-\\nways be wider than confidence intervals because they account for the\\nuncertainty associated with ϵ, the irreducible error.\\n6. Is the relationship linear?\\nIn Section 3.3.3, we saw that residual plots can be used in order to\\nidentify non-linearity. If the relationships are linear, then the residual\\nplots should display no pattern. In the case of the Advertising data,\\nwe observe a non-linear effect in Figure 3.5, though this effect could\\nalso be observed in a residual plot. In Section 3.3.2, we discussed the\\ninclusion of transformations of the predictors in the linear regression\\nmodel in order to accommodate non-linear relationships.\\n7. Is there synergy among the advertising media?\\nThe standard linear regression model assumes an additive relation-\\nship between the predictors and the response. An additive model\\nis easy to interpret because the association between each predictor\\nand the response is unrelated to the values of the other predictors.\\nHowever, the additive assumption may be unrealistic for certain data\\nsets. In Section 3.3.2, we showed how to include an interaction term 111\\nin the regression model in order to accommodate non-additive rela-\\ntionships. A small p-value associated with the interaction term indi-\\ncates the presence of such relationships. Figure 3.5 suggested that the\\nAdvertising data may not be additive. Including an interaction term\\nin the model results in a substantial increase in R2, from around 90 %\\nto almost 97 %.\\n3.5\\nComparison of Linear Regression\\nwith K-Nearest Neighbors\\nAs discussed in Chapter 2, linear regression is an example of a parametric\\napproach because it assumes a linear functional form for f(X). Parametric\\nmethods have several advantages. They are often easy to fit, because one\\nneed estimate only a small number of coefficients. In the case of linear re-\\ngression, the coefficients have simple interpretations, and tests of statistical\\nsignificance can be easily performed. But parametric methods do have a\\ndisadvantage: by construction, they make strong assumptions about the\\nform of f(X). If the specified functional form is far from the truth, and\\nprediction accuracy is our goal, then the parametric method will perform\\npoorly. For instance, if we assume a linear relationship between X and Y\\nbut the true relationship is far from linear, then the resulting model will\\nprovide a poor fit to the data, and any conclusions drawn from it will be\\nsuspect.\\nIn contrast, non-parametric methods do not explicitly assume a para-\\nmetric form for f(X), and thereby provide an alternative and more flexi-\\nble approach for performing regression. We discuss various non-parametric\\nmethods in this book. Here we consider one of the simplest and best-known\\nnon-parametric methods, K-nearest neighbors regression (KNN regression). K-nearest\\nneighbors\\nregression\\nThe KNN regression method is closely related to the KNN classifier dis-\\ncussed in Chapter 2. Given a value for K and a prediction point x0, KNN\\nregression first identifies the K training observations that are closest to\\nx0, represented by N0. It then estimates f(x0) using the average of all the\\ntraining responses in N0. In other words,\\nˆf(x0) = 1\\nK\\n0\\nxi∈N0\\nyi.\\nFigure 3.16 illustrates two KNN fits on a data set with p = 2 predictors. The\\nfit with K = 1 is shown in the left-hand panel, while the right-hand panel\\ncorresponds to K = 9. We see that when K = 1, the KNN fit perfectly\\ninterpolates the training observations, and consequently takes the form\\nof a step function. When K = 9, the KNN fit still is a step function, but\\naveraging over nine observations results in much smaller regions of constant\\nprediction, and consequently a smoother fit. In general, the optimal value\\nfor K will depend on the bias-variance tradeoff, which we introduced in\\nChapter 2. A small value for K provides the most flexible fit, which will\\nhave low bias but high variance. This variance is due to the fact that the\\nprediction in a given region is entirely dependent on just one observation.\\n3.5 Comparison of Linear Regression with K-Nearest Neighbors 112\\n3. Linear Regression\\ny\\ny\\nx1\\nx1\\nx2\\nx2\\ny\\ny\\ny\\nFIGURE 3.16.\\nPlots of ˆf(X) using KNN regression on a two-dimensional\\ndata set with 64 observations (orange dots). Left: K = 1 results in a rough step\\nfunction fit. Right: K = 9 produces a much smoother fit.\\nIn contrast, larger values of K provide a smoother and less variable fit; the\\nprediction in a region is an average of several points, and so changing one\\nobservation has a smaller effect. However, the smoothing may cause bias by\\nmasking some of the structure in f(X). In Chapter 5, we introduce several\\napproaches for estimating test error rates. These methods can be used to\\nidentify the optimal value of K in KNN regression.\\nIn what setting will a parametric approach such as least squares linear re-\\ngression outperform a non-parametric approach such as KNN regression?\\nThe answer is simple: the parametric approach will outperform the non-\\nparametric approach if the parametric form that has been selected is close\\nto the true form of f. Figure 3.17 provides an example with data generated\\nfrom a one-dimensional linear regression model. The black solid lines rep-\\nresent f(X), while the blue curves correspond to the KNN fits using K = 1\\nand K = 9. In this case, the K = 1 predictions are far too variable, while\\nthe smoother K = 9 fit is much closer to f(X). However, since the true\\nrelationship is linear, it is hard for a non-parametric approach to compete\\nwith linear regression: a non-parametric approach incurs a cost in variance\\nthat is not offset by a reduction in bias. The blue dashed line in the left-\\nhand panel of Figure 3.18 represents the linear regression fit to the same\\ndata. It is almost perfect. The right-hand panel of Figure 3.18 reveals that\\nlinear regression outperforms KNN for this data. The green solid line, plot-\\nted as a function of 1/K, represents the test set mean squared error (MSE)\\nfor KNN. The KNN errors are well above the black dashed line, which is\\nthe test MSE for linear regression. When the value of K is large, then KNN\\nperforms only a little worse than least squares regression in terms of MSE.\\nIt performs far worse when K is small.\\nIn practice, the true relationship between X and Y is rarely exactly lin-\\near. Figure 3.19 examines the relative performances of least squares regres-\\nsion and KNN under increasing levels of non-linearity in the relationship\\nbetween X and Y . In the top row, the true relationship is nearly linear.\\nIn this case we see that the test MSE for linear regression is still superior 3.5 Comparison of Linear Regression with K-Nearest Neighbors\\n113\\n−1.0\\n−0.5\\n0.0\\n0.5\\n1.0\\n1\\n2\\n3\\n4\\n−1.0\\n−0.5\\n0.0\\n0.5\\n1.0\\n1\\n2\\n3\\n4\\ny\\ny\\nx\\nx\\nFIGURE 3.17. Plots of ˆf(X) using KNN regression on a one-dimensional data\\nset with 50 observations. The true relationship is given by the black solid line.\\nLeft: The blue curve corresponds to K = 1 and interpolates (i.e. passes directly\\nthrough) the training data. Right: The blue curve corresponds to K = 9, and\\nrepresents a smoother fit.\\n−1.0\\n−0.5\\n0.0\\n0.5\\n1.0\\n1\\n2\\n3\\n4\\n0.2\\n0.5\\n1.0\\n0.00\\n0.05\\n0.10\\n0.15\\nMean Squared Error\\ny\\nx\\n1/K\\nFIGURE 3.18. The same data set shown in Figure 3.17 is investigated further.\\nLeft: The blue dashed line is the least squares fit to the data. Since f(X) is in\\nfact linear (displayed as the black line), the least squares regression line provides\\na very good estimate of f(X). Right: The dashed horizontal line represents the\\nleast squares test set MSE, while the green solid line corresponds to the MSE\\nfor KNN as a function of 1/K (on the log scale). Linear regression achieves a\\nlower test MSE than does KNN regression, since f(X) is in fact linear. For KNN\\nregression, the best results occur with a very large value of K, corresponding to a\\nsmall value of 1/K. 114\\n3. Linear Regression\\n−1.0\\n−0.5\\n0.0\\n0.5\\n1.0\\n0.5\\n1.0\\n1.5\\n2.0\\n2.5\\n3.0\\n3.5\\n0.2\\n0.5\\n1.0\\n0.00\\n0.02\\n0.04\\n0.06\\n0.08\\nMean Squared Error\\n−1.0\\n−0.5\\n0.0\\n0.5\\n1.0\\n1.0\\n1.5\\n2.0\\n2.5\\n3.0\\n3.5\\n0.2\\n0.5\\n1.0\\n0.00\\n0.05\\n0.10\\n0.15\\nMean Squared Error\\ny\\ny\\nx\\nx\\n1/K\\n1/K\\nFIGURE 3.19.\\nTop Left: In a setting with a slightly non-linear relationship\\nbetween X and Y (solid black line), the KNN fits with K = 1 (blue) and K = 9\\n(red) are displayed. Top Right: For the slightly non-linear data, the test set MSE\\nfor least squares regression (horizontal black) and KNN with various values of\\n1/K (green) are displayed. Bottom Left and Bottom Right: As in the top panel,\\nbut with a strongly non-linear relationship between X and Y .\\nto that of KNN for low values of K. However, for K ≥4, KNN out-\\nperforms linear regression. The second row illustrates a more substantial\\ndeviation from linearity. In this situation, KNN substantially outperforms\\nlinear regression for all values of K. Note that as the extent of non-linearity\\nincreases, there is little change in the test set MSE for the non-parametric\\nKNN method, but there is a large increase in the test set MSE of linear\\nregression.\\nFigures 3.18 and 3.19 display situations in which KNN performs slightly\\nworse than linear regression when the relationship is linear, but much bet-\\nter than linear regression for nonlinear situations. In a real life situation\\nin which the true relationship is unknown, one might suspect that KNN\\nshould be favored over linear regression because it will at worst be slightly\\ninferior to linear regression if the true relationship is linear, and may give\\nsubstantially better results if the true relationship is non-linear. But in re-\\nality, even when the true relationship is highly non-linear, KNN may still\\nprovide inferior results to linear regression. In particular, both Figures 3.18 3.5 Comparison of Linear Regression with K-Nearest Neighbors\\n115\\n0.2\\n0.5\\n1.0\\n0.0\\n0.2\\n0.4\\n0.6\\n0.8\\n1.0\\np=1\\n0.2\\n0.5\\n1.0\\n0.0\\n0.2\\n0.4\\n0.6\\n0.8\\n1.0\\np=2\\n0.2\\n0.5\\n1.0\\n0.0\\n0.2\\n0.4\\n0.6\\n0.8\\n1.0\\np=3\\n0.2\\n0.5\\n1.0\\n0.0\\n0.2\\n0.4\\n0.6\\n0.8\\n1.0\\np=4\\n0.2\\n0.5\\n1.0\\n0.0\\n0.2\\n0.4\\n0.6\\n0.8\\n1.0\\np=10\\n0.2\\n0.5\\n1.0\\n0.0\\n0.2\\n0.4\\n0.6\\n0.8\\n1.0\\np=20\\nMean Squared Error\\n1/K\\nFIGURE 3.20. Test MSE for linear regression (black dashed lines) and KNN\\n(green curves) as the number of variables p increases. The true function is non-\\nlinear in the first variable, as in the lower panel in Figure 3.19, and does not\\ndepend on the additional variables. The performance of linear regression deteri-\\norates slowly in the presence of these additional noise variables, whereas KNN’s\\nperformance degrades much more quickly as p increases.\\nand 3.19 illustrate settings with p = 1 predictor. But in higher dimensions,\\nKNN often performs worse than linear regression.\\nFigure 3.20 considers the same strongly non-linear situation as in the\\nsecond row of Figure 3.19, except that we have added additional noise\\npredictors that are not associated with the response. When p = 1 or p = 2,\\nKNN outperforms linear regression. But for p = 3 the results are mixed,\\nand for p ≥4 linear regression is superior to KNN. In fact, the increase in\\ndimension has only caused a small deterioration in the linear regression test\\nset MSE, but it has caused more than a ten-fold increase in the MSE for\\nKNN. This decrease in performance as the dimension increases is a common\\nproblem for KNN, and results from the fact that in higher dimensions\\nthere is effectively a reduction in sample size. In this data set there are\\n50 training observations; when p = 1, this provides enough information to\\naccurately estimate f(X). However, spreading 50 observations over p = 20\\ndimensions results in a phenomenon in which a given observation has no\\nnearby neighbors—this is the so-called curse of dimensionality. That is, curse of di-\\nmensionality\\nthe K observations that are nearest to a given test observation x0 may be\\nvery far away from x0 in p-dimensional space when p is large, leading to a\\nvery poor prediction of f(x0) and hence a poor KNN fit. As a general rule,\\nparametric methods will tend to outperform non-parametric approaches\\nwhen there is a small number of observations per predictor.\\nEven when the dimension is small, we might prefer linear regression to\\nKNN from an interpretability standpoint. If the test MSE of KNN is only\\nslightly lower than that of linear regression, we might be willing to forego\\na little bit of prediction accuracy for the sake of a simple model that can\\nbe described in terms of just a few coefficients, and for which p-values are\\navailable. 116\\n3. Linear Regression\\n3.6\\nLab: Linear Regression\\n3.6.1\\nImporting packages\\nWe import our standard libraries at this top level.\\nIn [1]: import numpy as np\\nimport pandas as pd\\nfrom matplotlib.pyplot import subplots\\nNew imports\\nThroughout this lab we will introduce new functions and libraries. However,\\nwe will import them here to emphasize these are the new code objects in\\nthis lab. Keeping imports near the top of a notebook makes the code more\\nreadable, since scanning the first few lines tells us what libraries are used.\\nIn [2]: import statsmodels.api as sm\\nWe will provide relevant details about the functions below as they are\\nneeded.\\nBesides importing whole modules, it is also possible to import only a\\nfew items from a given module. This will help keep the namespace clean. namespace\\nWe will use a few specific objects from the statsmodels package which we statsmodels\\nimport here.\\nIn [3]: from statsmodels.stats.outliers_influence \\\\\\nimport variance_inflation_factor as VIF\\nfrom statsmodels.stats.anova import anova_lm\\nAs one of the import statements above is quite a long line, we inserted a\\nline break \\\\ to ease readability.\\nWe will also use some functions written for the labs in this book in the\\nISLP package.\\nIn [4]: from ISLP import load_data\\nfrom ISLP.models import (ModelSpec as MS,\\nsummarize,\\npoly)\\nInspecting Objects and Namespaces\\nThe function dir() provides a list of objects in a namespace.\\ndir()\\nIn [5]: dir()\\nOut[5]: [\\'In\\',\\n\\'MS\\',\\n\\'_\\',\\n\\'__\\',\\n\\'___\\',\\n\\'__builtin__\\',\\n\\'__builtins__\\',\\n... 3.6 Lab: Linear Regression\\n117\\n\\'poly\\',\\n\\'quit\\',\\n\\'sm\\',\\n\\'summarize\\']\\nThis shows you everything that Python can find at the top level. There\\nare certain objects like __builtins__ that contain references to built-in\\nfunctions like print().\\nEvery python object has its own notion of namespace, also accessible\\nwith dir(). This will include both the attributes of the object as well as\\nany methods associated with it. For instance, we see \\'sum\\' in the listing\\nfor an array.\\nIn [6]: A = np.array([3,5,11])\\ndir(A)\\nOut[6]:\\n...\\n\\'strides\\',\\n\\'sum\\',\\n\\'swapaxes\\',\\n...\\nThis indicates that the object A.sum exists. In this case it is a method that\\ncan be used to compute the sum of the array A as can be seen by typing\\nA.sum?.\\nIn [7]: A.sum()\\nOut[7]: 19\\n3.6.2\\nSimple Linear Regression\\nIn this section we will construct model matrices (also called design matri-\\nces) using the ModelSpec() transform from ISLP.models.\\nWe will use the Boston housing data set, which is contained in the ISLP\\npackage. The Boston dataset records medv (median house value) for 506\\nneighborhoods around Boston. We will build a regression model to pre-\\ndict medv using 13 predictors such as rmvar (average number of rooms per\\nhouse), age (proportion of owner-occupied units built prior to 1940), and\\nlstat (percent of households with low socioeconomic status). We will use\\nstatsmodels for this task, a Python package that implements several com-\\nmonly used regression methods.\\nWe have included a simple loading function load_data() in the ISLP pack- load_data()\\nage:\\nIn [8]: Boston = load_data(\"Boston\")\\nBoston.columns\\nOut[8]: Index([\\'crim\\', \\'zn\\', \\'indus\\', \\'chas\\', \\'nox\\', \\'rm\\', \\'age\\', \\'dis\\',\\n\\'rad\\', \\'tax\\', \\'ptratio\\', \\'black\\', \\'lstat\\', \\'medv\\'],\\ndtype=\\'object\\') 118\\n3. Linear Regression\\nType Boston? to find out more about these data.\\nWe start by using the sm.OLS() function to fit a simple linear regression sm.OLS()\\nmodel. Our response will be medv and lstat will be the single predictor.\\nFor this model, we can create the model matrix by hand.\\nIn [9]: X = pd.DataFrame({\\'intercept\\': np.ones(Boston.shape[0]),\\n\\'lstat\\': Boston[\\'lstat\\']})\\nX[:4]\\nOut[9]:\\nintercept\\nlstat\\n0\\n1.0\\n4.98\\n1\\n1.0\\n9.14\\n2\\n1.0\\n4.03\\n3\\n1.0\\n2.94\\nWe extract the response, and fit the model.\\nIn [10]: y = Boston[\\'medv\\']\\nmodel = sm.OLS(y, X)\\nresults = model.fit()\\nNote that sm.OLS() does not fit the model; it specifies the model, and then\\nmodel.fit() does the actual fitting.\\nOur ISLP function summarize() produces a simple table of the parame- summarize()\\nter estimates, their standard errors, t-statistics and p-values. The function\\ntakes a single argument, such as the object results returned here by the\\nfit method, and returns such a summary.\\nIn [11]: summarize(results)\\nOut[11]:\\ncoef\\nstd err\\nt\\nP>|t|\\nintercept\\n34.5538\\n0.563\\n61.415\\n0.0\\nlstat\\n-0.9500\\n0.039 -24.528\\n0.0\\nBefore we describe other methods for working with fitted models, we\\noutline a more useful and general framework for constructing a model ma-\\ntrix X.\\nUsing Transformations: Fit and Transform\\nOur model above has a single predictor, and constructing X was straight-\\nforward. In practice we often fit models with more than one predictor,\\ntypically selected from an array or data frame. We may wish to introduce\\ntransformations to the variables before fitting the model, specify interac-\\ntions between variables, and expand some particular variables into sets of\\nvariables (e.g. polynomials). The sklearn package has a particular notion sklearn\\nfor this type of task: a transform. A transform is an object that is created\\nwith some parameters as arguments. The object has two main methods:\\nfit() and transform().\\n.fit()\\n.transform()\\nWe provide a general approach for specifying models and constructing\\nthe model matrix through the transform ModelSpec() in the ISLP library. ModelSpec()\\nModelSpec() (renamed MS() in the preamble) creates a transform object,\\nand then a pair of methods transform() and fit() are used to construct a\\ncorresponding model matrix. 3.6 Lab: Linear Regression\\n119\\nWe first describe this process for our simple regression model using a\\nsingle predictor lstat in the Boston data frame, but will use it repeatedly\\nin more complex tasks in this and other labs in this book. In our case the\\ntransform is created by the expression design = MS([\\'lstat\\']).\\nThe fit() method takes the original array and may do some initial com-\\nputations on it, as specified in the transform object. For example, it may\\ncompute means and standard deviations for centering and scaling. The\\ntransform() method applies the fitted transformation to the array of data,\\nand produces the model matrix.\\nIn [12]: design = MS([\\'lstat\\'])\\ndesign = design.fit(Boston)\\nX = design.transform(Boston)\\nX[:4]\\nOut[12]:\\nintercept\\nlstat\\n0\\n1.0\\n4.98\\n1\\n1.0\\n9.14\\n2\\n1.0\\n4.03\\n3\\n1.0\\n2.94\\nIn this simple case, the fit() method does very little; it simply checks that\\nthe variable \\'lstat\\' specified in design exists in Boston. Then transform()\\nconstructs the model matrix with two columns: an intercept and the vari-\\nable lstat.\\nThese two operations can be combined with the fit_transform() method. .fit_\\ntransform()\\nIn [13]: design = MS([\\'lstat\\'])\\nX = design.fit_transform(Boston)\\nX[:4]\\nOut[13]:\\nintercept\\nlstat\\n0\\n1.0\\n4.98\\n1\\n1.0\\n9.14\\n2\\n1.0\\n4.03\\n3\\n1.0\\n2.94\\nNote that, as in the previous code chunk when the two steps were done\\nseparately, the design object is changed as a result of the fit() operation.\\nThe power of this pipeline will become clearer when we fit more complex\\nmodels that involve interactions and transformations.\\nLet’s return to our fitted regression model. The object results has several\\nmethods that can be used for inference. We already presented a function\\nsummarize() for showing the essentials of the fit. For a full and somewhat\\nexhaustive summary of the fit, we can use the summary() method (output\\nnot shown).\\nIn [14]: results.summary()\\nThe fitted coefficients can also be retrieved as the params attribute of\\nresults.\\nIn [15]: results.params 120\\n3. Linear Regression\\nOut[15]: intercept\\n34.553841\\nlstat\\n-0.950049\\ndtype: float64\\nThe get_prediction() method can be used to obtain predictions, and .get_\\nprediction()\\nproduce confidence intervals and prediction intervals for the prediction of\\nmedv for given values of lstat.\\nWe first create a new data frame, in this case containing only the vari-\\nable lstat, with the values for this variable at which we wish to make\\npredictions. We then use the transform() method of design to create the\\ncorresponding model matrix.\\nIn [16]: new_df = pd.DataFrame({\\'lstat\\':[5, 10, 15]})\\nnewX = design.transform(new_df)\\nnewX\\nOut[16]:\\nintercept\\nlstat\\n0\\n1.0\\n5\\n1\\n1.0\\n10\\n2\\n1.0\\n15\\nNext we compute the predictions at newX, and view them by extracting\\nthe predicted_mean attribute.\\nIn [17]: new_predictions = results.get_prediction(newX);\\nnew_predictions.predicted_mean\\nOut[17]: array([29.80359411, 25.05334734, 20.30310057])\\nWe can produce confidence intervals for the predicted values.\\nIn [18]: new_predictions.conf_int(alpha=0.05)\\nOut[18]: array([[29.00741194, 30.59977628],\\n[24.47413202, 25.63256267],\\n[19.73158815, 20.87461299]])\\nPrediction intervals are computing by setting obs=True:\\nIn [19]: new_predictions.conf_int(obs=True, alpha=0.05)\\nOut[19]: array([[17.56567478, 42.04151344],\\n[12.82762635, 37.27906833],\\n[ 8.0777421 , 32.52845905]])\\nFor instance, the 95% confidence interval associated with an lstat value of\\n10 is (24.47, 25.63), and the 95% prediction interval is (12.82, 37.28). As\\nexpected, the confidence and prediction intervals are centered around the\\nsame point (a predicted value of 25.05 for medv when lstat equals 10), but\\nthe latter are substantially wider.\\nNext we will plot medv and lstat using DataFrame.plot.scatter(), and .plot.\\nscatter()\\nwish to add the regression line to the resulting plot. 3.6 Lab: Linear Regression\\n121\\nDefining Functions\\nWhile there is a function within the ISLP package that adds a line to an\\nexisting plot, we take this opportunity to define our first function to do so. def\\nIn [20]: def abline(ax, b, m):\\n\"Add a line with slope m and intercept b to ax\"\\nxlim = ax.get_xlim()\\nylim = [m * xlim[0] + b, m * xlim[1] + b]\\nax.plot(xlim, ylim)\\nA few things are illustrated above. First we see the syntax for defining a\\nfunction: def funcname(...). The function has arguments ax, b, m where\\nax is an axis object for an exisiting plot, b is the intercept and m is the slope\\nof the desired line. Other plotting options can be passed on to ax.plot by\\nincluding additional optional arguments as follows:\\nIn [21]: def abline(ax, b, m, *args, **kwargs):\\n\"Add a line with slope m and intercept b to ax\"\\nxlim = ax.get_xlim()\\nylim = [m * xlim[0] + b, m * xlim[1] + b]\\nax.plot(xlim, ylim, *args, **kwargs)\\nThe addition of *args allows any number of non-named arguments to\\nabline, while *kwargs allows any number of named arguments (such as\\nlinewidth=3) to abline. In our function, we pass these arguments verbatim\\nto ax.plot above. Readers interested in learning more about functions are\\nreferred to the section on defining functions in docs.python.org/tutorial.\\nLet’s use our new function to add this regression line to a plot of medv\\nvs. lstat.\\nIn [22]: ax = Boston.plot.scatter(\\'lstat\\', \\'medv\\')\\nabline(ax,\\nresults.params[0],\\nresults.params[1],\\n\\'r--\\',\\nlinewidth=3)\\nThus, the final call to ax.plot() is ax.plot(xlim, ylim, \\'r--\\', linewidth=3).\\nWe have used the argument \\'r--\\' to produce a red dashed line, and added\\nan argument to make it of width 3. There is some evidence for non-linearity\\nin the relationship between lstat and medv. We will explore this issue later\\nin this lab.\\nAs mentioned above, there is an existing function to add a line to a plot\\n— ax.axline() — but knowing how to write such functions empowers us\\nto create more expressive displays.\\nNext we examine some diagnostic plots, several of which were discussed\\nin Section 3.3.3. We can find the fitted values and residuals of the fit as\\nattributes of the results object. Various influence measures describing the\\nregression model are computed with the get_influence() method. As we .get_\\ninfluence()\\nwill not use the fig component returned as the first value from subplots(),\\nwe simply capture the second returned value in ax below.\\nIn [23]: ax = subplots(figsize=(8,8))[1] 122\\n3. Linear Regression\\nax.scatter(results.fittedvalues , results.resid)\\nax.set_xlabel(\\'Fitted value\\')\\nax.set_ylabel(\\'Residual\\')\\nax.axhline(0, c=\\'k\\', ls=\\'--\\');\\nWe add a horizontal line at 0 for reference using the ax.axhline() method, .axhline()\\nindicating it should be black (c=\\'k\\') and have a dashed linestyle (ls=\\'--\\').\\nOn the basis of the residual plot (not shown), there is some evidence\\nof non-linearity. Leverage statistics can be computed for any number of\\npredictors using the hat_matrix_diag attribute of the value returned by the\\nget_influence() method.\\nIn [24]: infl = results.get_influence()\\nax = subplots(figsize=(8,8))[1]\\nax.scatter(np.arange(X.shape[0]), infl.hat_matrix_diag)\\nax.set_xlabel(\\'Index\\')\\nax.set_ylabel(\\'Leverage\\')\\nnp.argmax(infl.hat_matrix_diag)\\nOut[24]: 374\\nThe np.argmax() function identifies the index of the largest element of an np.argmax()\\narray, optionally computed over an axis of the array. In this case, we maxi-\\nmized over the entire array to determine which observation has the largest\\nleverage statistic.\\n3.6.3\\nMultiple Linear Regression\\nIn order to fit a multiple linear regression model using least squares, we\\nagain use the ModelSpec() transform to construct the required model matrix\\nand response. The arguments to ModelSpec() can be quite general, but in\\nthis case a list of column names suffice. We consider a fit here with the two\\nvariables lstat and age.\\nIn [25]: X = MS([\\'lstat\\', \\'age\\']).fit_transform(Boston)\\nmodel1 = sm.OLS(y, X)\\nresults1 = model1.fit()\\nsummarize(results1)\\nOut[25]:\\ncoef\\nstd err\\nt\\nP>|t|\\nintercept\\n33.2228\\n0.731\\n45.458\\n0.000\\nlstat\\n-1.0321\\n0.048\\n-21.416\\n0.000\\nage\\n0.0345\\n0.012\\n2.826\\n0.005\\nNotice how we have compacted the first line into a succinct expression\\ndescribing the construction of X.\\nThe Boston data set contains 12 variables, and so it would be cumbersome\\nto have to type all of these in order to perform a regression using all of the\\npredictors. Instead, we can use the following short-hand:\\n.columns.\\ndrop()\\nIn [26]: terms = Boston.columns.drop(\\'medv\\')\\nterms 3.6 Lab: Linear Regression\\n123\\nOut[26]: Index([\\'crim\\', \\'zn\\', \\'indus\\', \\'chas\\', \\'nox\\', \\'rm\\', \\'age\\', \\'dis\\',\\n\\'rad\\', \\'tax\\', \\'ptratio\\', \\'lstat\\'],\\ndtype=\\'object\\')\\nWe can now fit the model with all the variables in terms using the same\\nmodel matrix builder.\\nIn [27]: X = MS(terms).fit_transform(Boston)\\nmodel = sm.OLS(y, X)\\nresults = model.fit()\\nsummarize(results)\\nOut[27]:\\ncoef\\nstd err\\nt\\nP>|t|\\nintercept\\n41.6173\\n4.936\\n8.431\\n0.000\\ncrim\\n-0.1214\\n0.033\\n-3.678\\n0.000\\nzn\\n0.0470\\n0.014\\n3.384\\n0.001\\nindus\\n0.0135\\n0.062\\n0.217\\n0.829\\nchas\\n2.8400\\n0.870\\n3.264\\n0.001\\nnox\\n-18.7580\\n3.851\\n-4.870\\n0.000\\nrm\\n3.6581\\n0.420\\n8.705\\n0.000\\nage\\n0.0036\\n0.013\\n0.271\\n0.787\\ndis\\n-1.4908\\n0.202\\n-7.394\\n0.000\\nrad\\n0.2894\\n0.067\\n4.325\\n0.000\\ntax\\n-0.0127\\n0.004\\n-3.337\\n0.001\\nptratio\\n-0.9375\\n0.132\\n-7.091\\n0.000\\nlstat\\n-0.5520\\n0.051\\n-10.897 0.000\\nWhat if we would like to perform a regression using all of the variables but\\none? For example, in the above regression output, age has a high p-value.\\nSo we may wish to run a regression excluding this predictor. The following\\nsyntax results in a regression using all predictors except age (output not\\nshown).\\nIn [28]: minus_age = Boston.columns.drop([\\'medv\\', \\'age\\'])\\nXma = MS(minus_age).fit_transform(Boston)\\nmodel1 = sm.OLS(y, Xma)\\nsummarize(model1.fit())\\n3.6.4\\nMultivariate Goodness of Fit\\nWe can access the individual components of results by name (dir(results)\\nshows us what is available). Hence results.rsquared gives us the R2, and\\nnp.sqrt(results.scale) gives us the RSE.\\nVariance inflation factors (section 3.3.3) are sometimes useful to assess\\nthe effect of collinearity in the model matrix of a regression model. We will\\ncompute the VIFs in our multiple regression fit, and use the opportunity\\nto introduce the idea of list comprehension.\\nlist compre-\\nhension\\nList Comprehension\\nOften we encounter a sequence of objects which we would like to transform\\nfor some other task. Below, we compute the VIF for each feature in our X\\nmatrix and produce a data frame whose index agrees with the columns of\\nX. The notion of list comprehension can often make such a task easier. 124\\n3. Linear Regression\\nList comprehensions are simple and powerful ways to form lists of Python\\nobjects. The language also supports dictionary and generator comprehen-\\nsion, though these are beyond our scope here. Let’s look at an example.\\nWe compute the VIF for each of the variables in the model matrix X, using\\nthe function variance_inflation_factor().\\nvariance_\\ninflation_\\nfactor()\\nIn [29]: vals = [VIF(X, i)\\nfor i in range(1, X.shape[1])]\\nvif = pd.DataFrame({\\'vif\\':vals},\\nindex=X.columns[1:])\\nvif\\nOut[29]:\\nvif\\ncrim\\n1.767\\nzn\\n2.298\\nindus\\n3.987\\nchas\\n1.071\\nnox\\n4.369\\nrm\\n1.913\\nage\\n3.088\\ndis\\n3.954\\nrad\\n7.445\\ntax\\n9.002\\nptratio 1.797\\nlstat\\n2.871\\nThe function VIF() takes two arguments: a dataframe or array, and a vari-\\nable column index. In the code above we call VIF() on the fly for all columns\\nin X. We have excluded column 0 above (the intercept), which is not of in-\\nterest. In this case the VIFs are not that exciting.\\nThe object vals above could have been constructed with the following\\nfor loop:\\nIn [30]: vals = []\\nfor i in range(1, X.values.shape[1]):\\nvals.append(VIF(X.values, i))\\nList comprehension allows us to perform such repetitive operations in a\\nmore straightforward way.\\n3.6.5\\nInteraction Terms\\nIt is easy to include interaction terms in a linear model using ModelSpec().\\nIncluding a tuple (\"lstat\",\"age\") tells the model matrix builder to include\\nan interaction term between lstat and age.\\nIn [31]: X = MS([\\'lstat\\',\\n\\'age\\',\\n(\\'lstat\\', \\'age\\')]).fit_transform(Boston)\\nmodel2 = sm.OLS(y, X)\\nsummarize(model2.fit())\\nOut[31]:\\ncoef\\nstd err\\nt\\nP>|t|\\nintercept\\n36.0885\\n1.470\\n24.553\\n0.000\\nlstat\\n-1.3921\\n0.167\\n-8.313\\n0.000 3.6 Lab: Linear Regression\\n125\\nage\\n-0.0007\\n0.020\\n-0.036\\n0.971\\nlstat:age\\n0.0042\\n0.002\\n2.244\\n0.025\\n3.6.6\\nNon-linear Transformations of the Predictors\\nThe model matrix builder can include terms beyond just column names and\\ninteractions. For instance, the poly() function supplied in ISLP specifies poly()\\nthat columns representing polynomial functions of its first argument are\\nadded to the model matrix.\\nIn [32]: X = MS([poly(\\'lstat\\', degree=2), \\'age\\']).fit_transform(Boston)\\nmodel3 = sm.OLS(y, X)\\nresults3 = model3.fit()\\nsummarize(results3)\\nOut[32]:\\ncoef\\nstd err\\nt\\nP>|t|\\nintercept\\n17.7151\\n0.781\\n22.681\\n0.000\\npoly(lstat, degree=2)[0]\\n-179.2279\\n6.733\\n-26.620\\n0.000\\npoly(lstat, degree=2)[1]\\n72.9908\\n5.482\\n13.315\\n0.000\\nage\\n0.0703\\n0.011\\n6.471\\n0.000\\nThe effectively zero p-value associated with the quadratic term (i.e. the\\nthird row above) suggests that it leads to an improved model.\\nBy default, poly() creates a basis matrix for inclusion in the model ma-\\ntrix whose columns are orthogonal polynomials, which are designed for sta- orthogonal\\npolynomial\\nble least squares computations.13 Alternatively, had we included an argu-\\nment raw=True in the above call to poly(), the basis matrix would consist\\nsimply of lstat and lstat**2. Since either of these bases represent quadratic\\npolynomials, the fitted values would not change in this case, just the poly-\\nnomial coefficients. Also by default, the columns created by poly() do not\\ninclude an intercept column as that is automatically added by MS().\\nWe use the anova_lm() function to further quantify the extent to which anova_lm()\\nthe quadratic fit is superior to the linear fit.\\nIn [33]: anova_lm(results1, results3)\\nOut[33]:\\ndf_resid\\nssr\\ndf_diff\\nss_diff\\nF\\nPr(>F)\\n0\\n503.0\\n19168.13\\n0.0\\nNaN\\nNaN\\nNaN\\n1\\n502.0\\n14165.61\\n1.0\\n5002.52\\n177.28\\n7.47e-35\\nHere results1 represents the linear submodel containing predictors lstat\\nand age, while results3 corresponds to the larger model above with a\\nquadratic term in lstat. The anova_lm() function performs a hypothesis\\ntest comparing the two models. The null hypothesis is that the quadratic\\nterm in the bigger model is not needed, and the alternative hypothesis is\\nthat the bigger model is superior. Here the F-statistic is 177.28 and the\\nassociated p-value is zero. In this case the F-statistic is the square of the\\nt-statistic for the quadratic term in the linear model summary for results3\\n— a consequence of the fact that these nested models differ by one degree of\\n13Actually, poly() is a wrapper for the workhorse and standalone function Poly()\\nthat does the work in building the model matrix. 126\\n3. Linear Regression\\nfreedom. This provides very clear evidence that the quadratic polynomial\\nin lstat improves the linear model. This is not surprising, since earlier we\\nsaw evidence for non-linearity in the relationship between medv and lstat.\\nThe function anova_lm() can take more than two nested models as input,\\nin which case it compares every successive pair of models. That also explains\\nwhy their are NaNs in the first row above, since there is no previous model\\nwith which to compare the first.\\nIn [34]: ax = subplots(figsize=(8,8))[1]\\nax.scatter(results3.fittedvalues , results3.resid)\\nax.set_xlabel(\\'Fitted value\\')\\nax.set_ylabel(\\'Residual\\')\\nax.axhline(0, c=\\'k\\', ls=\\'--\\')\\nWe see that when the quadratic term is included in the model, there is\\nlittle discernible pattern in the residuals. In order to create a cubic or\\nhigher-degree polynomial fit, we can simply change the degree argument to\\npoly().\\n3.6.7\\nQualitative Predictors\\nHere we use the Carseats data, which is included in the ISLP package. We\\nwill attempt to predict Sales (child car seat sales) in 400 locations based\\non a number of predictors.\\nIn [35]: Carseats = load_data(\\'Carseats\\')\\nCarseats.columns\\nOut[35]: Index([\\'Sales\\', \\'CompPrice\\', \\'Income\\', \\'Advertising\\',\\n\\'Population\\', \\'Price\\', \\'ShelveLoc\\', \\'Age\\', \\'Education\\',\\n\\'Urban\\', \\'US\\'],\\ndtype=\\'object\\')\\nThe Carseats data includes qualitative predictors such as ShelveLoc, an\\nindicator of the quality of the shelving location — that is, the space within\\na store in which the car seat is displayed. The predictor ShelveLoc takes on\\nthree possible values, Bad, Medium, and Good. Given a qualitative variable\\nsuch as ShelveLoc, ModelSpec() generates dummy variables automatically.\\nThese variables are often referred to as a one-hot encoding of the categorical one-hot\\nencoding\\nfeature. Their columns sum to one, so to avoid collinearity with an inter-\\ncept, the first column is dropped. Below we see the column ShelveLoc[Bad]\\nhas been dropped, since Bad is the first level of ShelveLoc. Below we fit a\\nmultiple regression model that includes some interaction terms.\\nIn [36]: allvars = list(Carseats.columns.drop(\\'Sales\\'))\\ny = Carseats[\\'Sales\\']\\nfinal = allvars + [(\\'Income\\', \\'Advertising\\'),\\n(\\'Price\\', \\'Age\\')]\\nX = MS(final).fit_transform(Carseats)\\nmodel = sm.OLS(y, X)\\nsummarize(model.fit())\\nOut[36]:\\ncoef\\nstd err\\nt\\nP>|t|\\nintercept\\n6.5756\\n1.009\\n6.519\\n0.000 3.7 Exercises\\n127\\nCompPrice\\n0.0929\\n0.004\\n22.567\\n0.000\\nIncome\\n0.0109\\n0.003\\n4.183\\n0.000\\nAdvertising\\n0.0702\\n0.023\\n3.107\\n0.002\\nPopulation\\n0.0002\\n0.000\\n0.433\\n0.665\\nPrice\\n-0.1008\\n0.007\\n-13.549\\n0.000\\nShelveLoc[Good]\\n4.8487\\n0.153\\n31.724\\n0.000\\nShelveLoc[Medium]\\n1.9533\\n0.126\\n15.531\\n0.000\\nAge\\n-0.0579\\n0.016\\n-3.633\\n0.000\\nEducation\\n-0.0209\\n0.020\\n-1.063\\n0.288\\nUrban[Yes]\\n0.1402\\n0.112\\n1.247\\n0.213\\nUS[Yes]\\n-0.1576\\n0.149\\n-1.058\\n0.291\\nIncome:Advertising\\n0.0008\\n0.000\\n2.698\\n0.007\\nPrice:Age\\n0.0001\\n0.000\\n0.801\\n0.424\\nIn the first line above, we made allvars a list, so that we could add the\\ninteraction terms two lines down. Our model-matrix builder has created a\\nShelveLoc[Good] dummy variable that takes on a value of 1 if the shelving\\nlocation is good, and 0 otherwise. It has also created a ShelveLoc[Medium]\\ndummy variable that equals 1 if the shelving location is medium, and 0 oth-\\nerwise. A bad shelving location corresponds to a zero for each of the two\\ndummy variables. The fact that the coefficient for ShelveLoc[Good] in the\\nregression output is positive indicates that a good shelving location is asso-\\nciated with high sales (relative to a bad location). And ShelveLoc[Medium]\\nhas a smaller positive coefficient, indicating that a medium shelving loca-\\ntion leads to higher sales than a bad shelving location, but lower sales than\\na good shelving location.\\n3.7\\nExercises\\nConceptual\\n1. Describe the null hypotheses to which the p-values given in Table 3.4\\ncorrespond. Explain what conclusions you can draw based on these\\np-values. Your explanation should be phrased in terms of sales, TV,\\nradio, and newspaper, rather than in terms of the coefficients of the\\nlinear model.\\n2. Carefully explain the differences between the KNN classifier and KNN\\nregression methods.\\n3. Suppose we have a data set with five predictors, X1 = GPA, X2 =\\nIQ, X3 = Level (1 for College and 0 for High School), X4 = Interac-\\ntion between GPA and IQ, and X5 = Interaction between GPA and\\nLevel. The response is starting salary after graduation (in thousands\\nof dollars). Suppose we use least squares to fit the model, and get\\nˆβ0 = 50, ˆβ1 = 20, ˆβ2 = 0.07, ˆβ3 = 35, ˆβ4 = 0.01, ˆβ5 = −10.\\n(a) Which answer is correct, and why?\\ni. For a fixed value of IQ and GPA, high school graduates earn\\nmore, on average, than college graduates.\\nii. For a fixed value of IQ and GPA, college graduates earn\\nmore, on average, than high school graduates. 128\\n3. Linear Regression\\niii. For a fixed value of IQ and GPA, high school graduates earn\\nmore, on average, than college graduates provided that the\\nGPA is high enough.\\niv. For a fixed value of IQ and GPA, college graduates earn\\nmore, on average, than high school graduates provided that\\nthe GPA is high enough.\\n(b) Predict the salary of a college graduate with IQ of 110 and a\\nGPA of 4.0.\\n(c) True or false: Since the coefficient for the GPA/IQ interaction\\nterm is very small, there is very little evidence of an interaction\\neffect. Justify your answer.\\n4. I collect a set of data (n = 100 observations) containing a single\\npredictor and a quantitative response. I then fit a linear regression\\nmodel to the data, as well as a separate cubic regression, i.e. Y =\\nβ0 + β1X + β2X2 + β3X3 + ϵ.\\n(a) Suppose that the true relationship between X and Y is linear,\\ni.e. Y = β0 + β1X + ϵ. Consider the training residual sum of\\nsquares (RSS) for the linear regression, and also the training\\nRSS for the cubic regression. Would we expect one to be lower\\nthan the other, would we expect them to be the same, or is there\\nnot enough information to tell? Justify your answer.\\n(b) Answer (a) using test rather than training RSS.\\n(c) Suppose that the true relationship between X and Y is not linear,\\nbut we don’t know how far it is from linear. Consider the training\\nRSS for the linear regression, and also the training RSS for the\\ncubic regression. Would we expect one to be lower than the\\nother, would we expect them to be the same, or is there not\\nenough information to tell? Justify your answer.\\n(d) Answer (c) using test rather than training RSS.\\n5. Consider the fitted values that result from performing linear regres-\\nsion without an intercept. In this setting, the ith fitted value takes\\nthe form\\nˆyi = xi ˆβ,\\nwhere\\nˆβ =\\n> n\\n0\\ni=1\\nxiyi\\n?\\n/\\n> n\\n0\\ni′=1\\nx2\\ni′\\n?\\n.\\n(3.38)\\nShow that we can write\\nˆyi =\\nn\\n0\\ni′=1\\nai′yi′.\\nWhat is ai′?\\nNote: We interpret this result by saying that the fitted values from\\nlinear regression are linear combinations of the response values. 3.7 Exercises\\n129\\n6. Using (3.4), argue that in the case of simple linear regression, the\\nleast squares line always passes through the point (¯x, ¯y).\\n7. It is claimed in the text that in the case of simple linear regression\\nof Y onto X, the R2 statistic (3.17) is equal to the square of the\\ncorrelation between X and Y (3.18). Prove that this is the case. For\\nsimplicity, you may assume that ¯x = ¯y = 0.\\nApplied\\n8. This question involves the use of simple linear regression on the Auto\\ndata set.\\n(a) Use the sm.OLS() function to perform a simple linear regression\\nwith mpg as the response and horsepower as the predictor. Use\\nthe summarize() function to print the results. Comment on the\\noutput. For example:\\ni. Is there a relationship between the predictor and the re-\\nsponse?\\nii. How strong is the relationship between the predictor and\\nthe response?\\niii. Is the relationship between the predictor and the response\\npositive or negative?\\niv. What is the predicted mpg associated with a horsepower of\\n98? What are the associated 95 % confidence and prediction\\nintervals?\\n(b) Plot the response and the predictor in a new set of axes ax. Use\\nthe ax.axline() method or the abline() function defined in the\\nlab to display the least squares regression line.\\n(c) Produce some of diagnostic plots of the least squares regression\\nfit as described in the lab. Comment on any problems you see\\nwith the fit.\\n9. This question involves the use of multiple linear regression on the\\nAuto data set.\\n(a) Produce a scatterplot matrix which includes all of the variables\\nin the data set.\\n(b) Compute the matrix of correlations between the variables using\\nthe DataFrame.corr() method.\\n.corr()\\n(c) Use the sm.OLS() function to perform a multiple linear regression\\nwith mpg as the response and all other variables except name as\\nthe predictors. Use the summarize() function to print the results.\\nComment on the output. For instance:\\ni. Is there a relationship between the predictors and the re-\\nsponse? Use the anova_lm() function from statsmodels to\\nanswer this question. 130\\n3. Linear Regression\\nii. Which predictors appear to have a statistically significant\\nrelationship to the response?\\niii. What does the coefficient for the year variable suggest?\\n(d) Produce some of diagnostic plots of the linear regression fit as\\ndescribed in the lab. Comment on any problems you see with the\\nfit. Do the residual plots suggest any unusually large outliers?\\nDoes the leverage plot identify any observations with unusually\\nhigh leverage?\\n(e) Fit some models with interactions as described in the lab. Do\\nany interactions appear to be statistically significant?\\n(f) Try a few different transformations of the variables, such as\\nlog(X),\\n√\\nX, X2. Comment on your findings.\\n10. This question should be answered using the Carseats data set.\\n(a) Fit a multiple regression model to predict Sales using Price,\\nUrban, and US.\\n(b) Provide an interpretation of each coefficient in the model. Be\\ncareful—some of the variables in the model are qualitative!\\n(c) Write out the model in equation form, being careful to handle\\nthe qualitative variables properly.\\n(d) For which of the predictors can you reject the null hypothesis\\nH0 : βj = 0?\\n(e) On the basis of your response to the previous question, fit a\\nsmaller model that only uses the predictors for which there is\\nevidence of association with the outcome.\\n(f) How well do the models in (a) and (e) fit the data?\\n(g) Using the model from (e), obtain 95 % confidence intervals for\\nthe coefficient(s).\\n(h) Is there evidence of outliers or high leverage observations in the\\nmodel from (e)?\\n11. In this problem we will investigate the t-statistic for the null hypoth-\\nesis H0 : β = 0 in simple linear regression without an intercept. To\\nbegin, we generate a predictor x and a response y as follows.\\nrng = np.random.default_rng(1)\\nx = rng.normal(size=100)\\ny = 2 * x + rng.normal(size=100)\\n(a) Perform a simple linear regression of y onto x, without an in-\\ntercept. Report the coefficient estimate ˆβ, the standard error of\\nthis coefficient estimate, and the t-statistic and p-value associ-\\nated with the null hypothesis H0 : β = 0. Comment on these\\nresults. (You can perform regression without an intercept using\\nthe keywords argument intercept=False to ModelSpec().) 3.7 Exercises\\n131\\n(b) Now perform a simple linear regression of x onto y without an\\nintercept, and report the coefficient estimate, its standard error,\\nand the corresponding t-statistic and p-values associated with\\nthe null hypothesis H0 : β = 0. Comment on these results.\\n(c) What is the relationship between the results obtained in (a) and\\n(b)?\\n(d) For the regression of Y onto X without an intercept, the t-\\nstatistic for H0 : β = 0 takes the form ˆβ/SE(ˆβ), where ˆβ is\\ngiven by (3.38), and where\\nSE(ˆβ) =\\n@ )n\\ni=1(yi −xi ˆβ)2\\n(n −1) )n\\ni′=1 x2\\ni′ .\\n(These formulas are slightly different from those given in Sec-\\ntions 3.1.1 and 3.1.2, since here we are performing regression\\nwithout an intercept.) Show algebraically, and confirm numeri-\\ncally in R, that the t-statistic can be written as\\n(√n −1) )n\\ni=1 xiyi\\n5\\n()n\\ni=1 x2\\ni )()n\\ni′=1 y2\\ni′) −()n\\ni′=1 xi′yi′)2 .\\n(e) Using the results from (d), argue that the t-statistic for the re-\\ngression of y onto x is the same as the t-statistic for the regression\\nof x onto y.\\n(f) In R, show that when regression is performed with an intercept,\\nthe t-statistic for H0 : β1 = 0 is the same for the regression of y\\nonto x as it is for the regression of x onto y.\\n12. This problem involves simple linear regression without an intercept.\\n(a) Recall that the coefficient estimate ˆβ for the linear regression of\\nY onto X without an intercept is given by (3.38). Under what\\ncircumstance is the coefficient estimate for the regression of X\\nonto Y the same as the coefficient estimate for the regression of\\nY onto X?\\n(b) Generate an example in Python with n = 100 observations in\\nwhich the coefficient estimate for the regression of X onto Y\\nis different from the coefficient estimate for the regression of Y\\nonto X.\\n(c) Generate an example in Python with n = 100 observations in\\nwhich the coefficient estimate for the regression of X onto Y is\\nthe same as the coefficient estimate for the regression of Y onto\\nX.\\n13. In this exercise you will create some simulated data and will fit simple\\nlinear regression models to it. Make sure to use the default random\\nnumber generator with seed set to 1 prior to starting part (a) to\\nensure consistent results. 132\\n3. Linear Regression\\n(a) Using the normal() method of your random number generator,\\ncreate a vector, x, containing 100 observations drawn from a\\nN(0, 1) distribution. This represents a feature, X.\\n(b) Using the normal() method, create a vector, eps, containing 100\\nobservations drawn from a N(0, 0.25) distribution—a normal\\ndistribution with mean zero and variance 0.25.\\n(c) Using x and eps, generate a vector y according to the model\\nY = −1 + 0.5X + ϵ.\\n(3.39)\\nWhat is the length of the vector y? What are the values of β0\\nand β1 in this linear model?\\n(d) Create a scatterplot displaying the relationship between x and\\ny. Comment on what you observe.\\n(e) Fit a least squares linear model to predict y using x. Comment\\non the model obtained. How do ˆβ0 and ˆβ1 compare to β0 and\\nβ1?\\n(f) Display the least squares line on the scatterplot obtained in (d).\\nDraw the population regression line on the plot, in a different\\ncolor. Use the legend() method of the axes to create an appro-\\npriate legend.\\n(g) Now fit a polynomial regression model that predicts y using x\\nand x2. Is there evidence that the quadratic term improves the\\nmodel fit? Explain your answer.\\n(h) Repeat (a)–(f) after modifying the data generation process in\\nsuch a way that there is less noise in the data. The model (3.39)\\nshould remain the same. You can do this by decreasing the vari-\\nance of the normal distribution used to generate the error term\\nϵ in (b). Describe your results.\\n(i) Repeat (a)–(f) after modifying the data generation process in\\nsuch a way that there is more noise in the data. The model\\n(3.39) should remain the same. You can do this by increasing\\nthe variance of the normal distribution used to generate the\\nerror term ϵ in (b). Describe your results.\\n(j) What are the confidence intervals for β0 and β1 based on the\\noriginal data set, the noisier data set, and the less noisy data\\nset? Comment on your results.\\n14. This problem focuses on the collinearity problem.\\n(a) Perform the following commands in Python:\\nrng = np.random.default_rng(10)\\nx1 = rng.uniform(0, 1, size=100)\\nx2 = 0.5 * x1 + rng.normal(size=100) / 10\\ny = 2 + 2 * x1 + 0.3 * x2 + rng.normal(size=100)\\nThe last line corresponds to creating a linear model in which y is\\na function of x1 and x2. Write out the form of the linear model.\\nWhat are the regression coefficients? 3.7 Exercises\\n133\\n(b) What is the correlation between x1 and x2? Create a scatterplot\\ndisplaying the relationship between the variables.\\n(c) Using this data, fit a least squares regression to predict y using\\nx1 and x2. Describe the results obtained. What are ˆβ0, ˆβ1, and\\nˆβ2? How do these relate to the true β0, β1, and β2? Can you\\nreject the null hypothesis H0 : β1 = 0? How about the null\\nhypothesis H0 : β2 = 0?\\n(d) Now fit a least squares regression to predict y using only x1.\\nComment on your results. Can you reject the null hypothesis\\nH0 : β1 = 0?\\n(e) Now fit a least squares regression to predict y using only x2.\\nComment on your results. Can you reject the null hypothesis\\nH0 : β1 = 0?\\n(f) Do the results obtained in (c)–(e) contradict each other? Explain\\nyour answer.\\n(g) Suppose we obtain one additional observation, which was unfor-\\ntunately mismeasured. We use the function np.concatenate() to np.conca-\\ntenate()\\nadd this additional observation to each of x1, x2 and y.\\nx1 = np.concatenate([x1, [0.1]])\\nx2 = np.concatenate([x2, [0.8]])\\ny = np.concatenate([y, [6]])\\nRe-fit the linear models from (c) to (e) using this new data. What\\neffect does this new observation have on the each of the models?\\nIn each model, is this observation an outlier? A high-leverage\\npoint? Both? Explain your answers.\\n15. This problem involves the Boston data set, which we saw in the lab\\nfor this chapter. We will now try to predict per capita crime rate\\nusing the other variables in this data set. In other words, per capita\\ncrime rate is the response, and the other variables are the predictors.\\n(a) For each predictor, fit a simple linear regression model to predict\\nthe response. Describe your results. In which of the models is\\nthere a statistically significant association between the predictor\\nand the response? Create some plots to back up your assertions.\\n(b) Fit a multiple regression model to predict the response using\\nall of the predictors. Describe your results. For which predictors\\ncan we reject the null hypothesis H0 : βj = 0?\\n(c) How do your results from (a) compare to your results from (b)?\\nCreate a plot displaying the univariate regression coefficients\\nfrom (a) on the x-axis, and the multiple regression coefficients\\nfrom (b) on the y-axis. That is, each predictor is displayed as a\\nsingle point in the plot. Its coefficient in a simple linear regres-\\nsion model is shown on the x-axis, and its coefficient estimate\\nin the multiple linear regression model is shown on the y-axis. 134\\n3. Linear Regression\\n(d) Is there evidence of non-linear association between any of the\\npredictors and the response? To answer this question, for each\\npredictor X, fit a model of the form\\nY = β0 + β1X + β2X2 + β3X3 + ϵ. 4\\nClassification\\nThe linear regression model discussed in Chapter 3 assumes that the re-\\nsponse variable Y is quantitative. But in many situations, the response\\nvariable is instead qualitative. For example, eye color is qualitative. Of- qualitative\\nten qualitative variables are referred to as categorical; we will use these\\nterms interchangeably. In this chapter, we study approaches for predicting\\nqualitative responses, a process that is known as classification. Predicting classification\\na qualitative response for an observation can be referred to as classifying\\nthat observation, since it involves assigning the observation to a category,\\nor class. On the other hand, often the methods used for classification first\\npredict the probability that the observation belongs to each of the cate-\\ngories of a qualitative variable, as the basis for making the classification.\\nIn this sense they also behave like regression methods.\\nThere are many possible classification techniques, or classifiers, that one classifier\\nmight use to predict a qualitative response. We touched on some of these\\nin Sections 2.1.5 and 2.2.3. In this chapter we discuss some widely-used\\nclassifiers: logistic regression, linear discriminant analysis, quadratic dis- logistic\\nregression\\nlinear\\ndiscriminant\\nanalysis\\ncriminant analysis, naive Bayes, and K-nearest neighbors. The discussion\\nquadratic\\ndiscriminant\\nanalysis\\nnaive Bayes\\nK-nearest\\nneighbors\\nof logistic regression is used as a jumping-off point for a discussion of gen-\\neralized linear models, and in particular, Poisson regression. We discuss\\ngeneralized\\nlinear\\nmodels\\nPoisson\\nregression\\nmore computer-intensive classification methods in later chapters: these in-\\nclude generalized additive models (Chapter 7); trees, random forests, and\\nboosting (Chapter 8); and support vector machines (Chapter 9).\\n4.1\\nAn Overview of Classification\\nClassification problems occur often, perhaps even more so than regression\\nproblems. Some examples include:\\n© Springer Nature Switzerland AG 2023 \\nG. James et al., An Introduction to Statistical Learning, Springer Texts in Statistics, \\nhttps://doi.org/10.1007/978-3-031-38747-0_4 \\n135 136\\n4. Classification\\n1. A person arrives at the emergency room with a set of symptoms\\nthat could possibly be attributed to one of three medical conditions.\\nWhich of the three conditions does the individual have?\\n2. An online banking service must be able to determine whether or not\\na transaction being performed on the site is fraudulent, on the basis\\nof the user’s IP address, past transaction history, and so forth.\\n3. On the basis of DNA sequence data for a number of patients with\\nand without a given disease, a biologist would like to figure out which\\nDNA mutations are deleterious (disease-causing) and which are not.\\nJust as in the regression setting, in the classification setting we have a\\nset of training observations (x1, y1), . . . , (xn, yn) that we can use to build\\na classifier. We want our classifier to perform well not only on the training\\ndata, but also on test observations that were not used to train the classifier.\\nIn this chapter, we will illustrate the concept of classification using the\\nsimulated Default data set. We are interested in predicting whether an\\nindividual will default on his or her credit card payment, on the basis of\\nannual income and monthly credit card balance. The data set is displayed\\nin Figure 4.1. In the left-hand panel of Figure 4.1, we have plotted annual\\nincome and monthly credit card balance for a subset of 10, 000 individuals.\\nThe individuals who defaulted in a given month are shown in orange, and\\nthose who did not in blue. (The overall default rate is about 3 %, so we\\nhave plotted only a fraction of the individuals who did not default.) It\\nappears that individuals who defaulted tended to have higher credit card\\nbalances than those who did not. In the center and right-hand panels of\\nFigure 4.1, two pairs of boxplots are shown. The first shows the distribution\\nof balance split by the binary default variable; the second is a similar plot\\nfor income. In this chapter, we learn how to build a model to predict default\\n(Y ) for any given value of balance (X1) and income (X2). Since Y is not\\nquantitative, the simple linear regression model of Chapter 3 is not a good\\nchoice: we will elaborate on this further in Section 4.2.\\nIt is worth noting that Figure 4.1 displays a very pronounced relation-\\nship between the predictor balance and the response default. In most real\\napplications, the relationship between the predictor and the response will\\nnot be nearly so strong. However, for the sake of illustrating the classifica-\\ntion procedures discussed in this chapter, we use an example in which the\\nrelationship between the predictor and the response is somewhat exagger-\\nated.\\n4.2\\nWhy Not Linear Regression?\\nWe have stated that linear regression is not appropriate in the case of a\\nqualitative response. Why not?\\nSuppose that we are trying to predict the medical condition of a patient\\nin the emergency room on the basis of her symptoms. In this simplified\\nexample, there are three possible diagnoses: stroke, drug overdose, and 4.2 Why Not Linear Regression?\\n137\\n0\\n500\\n1000\\n1500\\n2000\\n2500\\n0\\n20000\\n40000\\n60000\\nBalance\\nIncome\\nNo\\nYes\\n0\\n500\\n1000\\n1500\\n2000\\n2500\\nDefault\\nBalance\\nNo\\nYes\\n0\\n20000\\n40000\\n60000\\nDefault\\nIncome\\nFIGURE 4.1. The Default data set. Left: The annual incomes and monthly\\ncredit card balances of a number of individuals. The individuals who defaulted on\\ntheir credit card payments are shown in orange, and those who did not are shown\\nin blue. Center: Boxplots of balance as a function of default status. Right:\\nBoxplots of income as a function of default status.\\nepileptic seizure. We could consider encoding these values as a quantita-\\ntive response variable, Y , as follows:\\nY =\\n\\uf8f1\\n\\uf8f4\\n\\uf8f2\\n\\uf8f4\\n\\uf8f3\\n1\\nif stroke;\\n2\\nif drug overdose;\\n3\\nif epileptic seizure.\\nUsing this coding, least squares could be used to fit a linear regression model\\nto predict Y on the basis of a set of predictors X1, . . . , Xp. Unfortunately,\\nthis coding implies an ordering on the outcomes, putting drug overdose in\\nbetween stroke and epileptic seizure, and insisting that the difference\\nbetween stroke and drug overdose is the same as the difference between\\ndrug overdose and epileptic seizure. In practice there is no particular\\nreason that this needs to be the case. For instance, one could choose an\\nequally reasonable coding,\\nY =\\n\\uf8f1\\n\\uf8f4\\n\\uf8f2\\n\\uf8f4\\n\\uf8f3\\n1\\nif epileptic seizure;\\n2\\nif stroke;\\n3\\nif drug overdose,\\nwhich would imply a totally different relationship among the three condi-\\ntions. Each of these codings would produce fundamentally different linear\\nmodels that would ultimately lead to different sets of predictions on test\\nobservations.\\nIf the response variable’s values did take on a natural ordering, such as\\nmild, moderate, and severe, and we felt the gap between mild and moderate\\nwas similar to the gap between moderate and severe, then a 1, 2, 3 coding\\nwould be reasonable. Unfortunately, in general there is no natural way to 138\\n4. Classification\\nconvert a qualitative response variable with more than two levels into a\\nquantitative response that is ready for linear regression.\\nFor a binary (two level) qualitative response, the situation is better. For binary\\ninstance, perhaps there are only two possibilities for the patient’s medical\\ncondition: stroke and drug overdose. We could then potentially use the\\ndummy variable approach from Section 3.3.1 to code the response as follows:\\nY =\\n=\\n0\\nif stroke;\\n1\\nif drug overdose.\\nWe could then fit a linear regression to this binary response, and predict\\ndrug overdose if ˆY > 0.5 and stroke otherwise. In the binary case it is not\\nhard to show that even if we flip the above coding, linear regression will\\nproduce the same final predictions.\\nFor a binary response with a 0/1 coding as above, regression by least\\nsquares is not completely unreasonable: it can be shown that the X ˆβ ob-\\ntained using linear regression is in fact an estimate of Pr(drug overdose|X)\\nin this special case. However, if we use linear regression, some of our es-\\ntimates might be outside the [0, 1] interval (see Figure 4.2), making them\\nhard to interpret as probabilities! Nevertheless, the predictions provide an\\nordering and can be interpreted as crude probability estimates. Curiously,\\nit turns out that the classifications that we get if we use linear regression\\nto predict a binary response will be the same as for the linear discriminant\\nanalysis (LDA) procedure we discuss in Section 4.4.\\nTo summarize, there are at least two reasons not to perform classifica-\\ntion using a regression method: (a) a regression method cannot accommo-\\ndate a qualitative response with more than two classes; (b) a regression\\nmethod will not provide meaningful estimates of Pr(Y |X), even with just\\ntwo classes. Thus, it is preferable to use a classification method that is\\ntruly suited for qualitative response values. In the next section, we present\\nlogistic regression, which is well-suited for the case of a binary qualita-\\ntive response; in later sections we will cover classification methods that are\\nappropriate when the qualitative response has two or more classes.\\n4.3\\nLogistic Regression\\nConsider again the Default data set, where the response default falls into\\none of two categories, Yes or No. Rather than modeling this response Y\\ndirectly, logistic regression models the probability that Y belongs to a par-\\nticular category.\\nFor the Default data, logistic regression models the probability of default.\\nFor example, the probability of default given balance can be written as\\nPr(default = Yes|balance).\\nThe values of Pr(default = Yes|balance), which we abbreviate p(balance),\\nwill range between 0 and 1. Then for any given value of balance, a prediction\\ncan be made for default. For example, one might predict default = Yes 4.3 Logistic Regression\\n139\\n0\\n500\\n1000\\n1500\\n2000\\n2500\\n0.0\\n0.2\\n0.4\\n0.6\\n0.8\\n1.0\\nBalance\\nProbability of Default\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n| |\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| | ||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n||\\n|\\n| |\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|||\\n|\\n| |\\n|||\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|||\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| | ||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n| |\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| ||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n| |\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n| ||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n| |\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|| |\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|| |\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n| |\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| | |\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|||\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n| ||\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n||\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| ||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n| ||\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| ||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n| |\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n||\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| | |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n| |\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| ||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|| |\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| ||\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n| | |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n||\\n||\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| ||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| ||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n0\\n500\\n1000\\n1500\\n2000\\n2500\\n0.0\\n0.2\\n0.4\\n0.6\\n0.8\\n1.0\\nBalance\\nProbability of Default\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n| |\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| | ||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n||\\n|\\n| |\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|||\\n|\\n| |\\n|||\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|||\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| | ||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n| |\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| ||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n| |\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n| ||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n| |\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|| |\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|| |\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n| |\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| | |\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|||\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n| ||\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n||\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| ||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n| ||\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| ||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n| |\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n||\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| | |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n| |\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| ||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|| |\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| ||\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n| | |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n||\\n||\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| ||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| ||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\nFIGURE 4.2. Classification using the Default data. Left: Estimated probability\\nof default using linear regression. Some estimated probabilities are negative! The\\norange ticks indicate the 0/1 values coded for default(No or Yes). Right: Predicted\\nprobabilities of default using logistic regression. All probabilities lie between 0\\nand 1.\\nfor any individual for whom p(balance) > 0.5. Alternatively, if a company\\nwishes to be conservative in predicting individuals who are at risk for de-\\nfault, then they may choose to use a lower threshold, such as p(balance) >\\n0.1.\\n4.3.1\\nThe Logistic Model\\nHow should we model the relationship between p(X) = Pr(Y = 1|X) and\\nX? (For convenience we are using the generic 0/1 coding for the response.)\\nIn Section 4.2 we considered using a linear regression model to represent\\nthese probabilities:\\np(X) = β0 + β1X.\\n(4.1)\\nIf we use this approach to predict default=Yes using balance, then we\\nobtain the model shown in the left-hand panel of Figure 4.2. Here we see\\nthe problem with this approach: for balances close to zero we predict a\\nnegative probability of default; if we were to predict for very large balances,\\nwe would get values bigger than 1. These predictions are not sensible, since\\nof course the true probability of default, regardless of credit card balance,\\nmust fall between 0 and 1. This problem is not unique to the credit default\\ndata. Any time a straight line is fit to a binary response that is coded as\\n0 or 1, in principle we can always predict p(X) < 0 for some values of X\\nand p(X) > 1 for others (unless the range of X is limited).\\nTo avoid this problem, we must model p(X) using a function that gives\\noutputs between 0 and 1 for all values of X. Many functions meet this\\ndescription. In logistic regression, we use the logistic function,\\nlogistic\\nfunction\\np(X) =\\neβ0+β1X\\n1 + eβ0+β1X .\\n(4.2)\\nTo fit the model (4.2), we use a method called maximum likelihood, which maximum\\nlikelihood\\nwe discuss in the next section. The right-hand panel of Figure 4.2 illustrates\\nthe fit of the logistic regression model to the Default data. Notice that for 140\\n4. Classification\\nlow balances we now predict the probability of default as close to, but never\\nbelow, zero. Likewise, for high balances we predict a default probability\\nclose to, but never above, one. The logistic function will always produce\\nan S-shaped curve of this form, and so regardless of the value of X, we\\nwill obtain a sensible prediction. We also see that the logistic model is\\nbetter able to capture the range of probabilities than is the linear regression\\nmodel in the left-hand plot. The average fitted probability in both cases is\\n0.0333 (averaged over the training data), which is the same as the overall\\nproportion of defaulters in the data set.\\nAfter a bit of manipulation of (4.2), we find that\\np(X)\\n1 −p(X) = eβ0+β1X.\\n(4.3)\\nThe quantity p(X)/[1−p(X)] is called the odds, and can take on any value odds\\nbetween 0 and ∞. Values of the odds close to 0 and ∞indicate very low\\nand very high probabilities of default, respectively. For example, on average\\n1 in 5 people with an odds of 1/4 will default, since p(X) = 0.2 implies an\\nodds of\\n0.2\\n1−0.2 = 1/4. Likewise, on average nine out of every ten people with\\nan odds of 9 will default, since p(X) = 0.9 implies an odds of\\n0.9\\n1−0.9 = 9.\\nOdds are traditionally used instead of probabilities in horse-racing, since\\nthey relate more naturally to the correct betting strategy.\\nBy taking the logarithm of both sides of (4.3), we arrive at\\nlog\\n*\\np(X)\\n1 −p(X)\\n+\\n= β0 + β1X.\\n(4.4)\\nThe left-hand side is called the log odds or logit. We see that the logistic log odds\\nlogit\\nregression model (4.2) has a logit that is linear in X.\\nRecall from Chapter 3 that in a linear regression model, β1 gives the\\naverage change in Y associated with a one-unit increase in X. By contrast,\\nin a logistic regression model, increasing X by one unit changes the log\\nodds by β1 (4.4). Equivalently, it multiplies the odds by eβ1 (4.3). However,\\nbecause the relationship between p(X) and X in (4.2) is not a straight line,\\nβ1 does not correspond to the change in p(X) associated with a one-unit\\nincrease in X. The amount that p(X) changes due to a one-unit change in\\nX depends on the current value of X. But regardless of the value of X, if\\nβ1 is positive then increasing X will be associated with increasing p(X),\\nand if β1 is negative then increasing X will be associated with decreasing\\np(X). The fact that there is not a straight-line relationship between p(X)\\nand X, and the fact that the rate of change in p(X) per unit change in X\\ndepends on the current value of X, can also be seen by inspection of the\\nright-hand panel of Figure 4.2.\\n4.3.2\\nEstimating the Regression Coefficients\\nThe coefficients β0 and β1 in (4.2) are unknown, and must be estimated\\nbased on the available training data. In Chapter 3, we used the least squares\\napproach to estimate the unknown linear regression coefficients. Although\\nwe could use (non-linear) least squares to fit the model (4.4), the more\\ngeneral method of maximum likelihood is preferred, since it has better sta-\\ntistical properties. The basic intuition behind using maximum likelihood 4.3 Logistic Regression\\n141\\nto fit a logistic regression model is as follows: we seek estimates for β0 and\\nβ1 such that the predicted probability ˆp(xi) of default for each individual,\\nusing (4.2), corresponds as closely as possible to the individual’s observed\\ndefault status. In other words, we try to find ˆβ0 and ˆβ1 such that plugging\\nthese estimates into the model for p(X), given in (4.2), yields a number\\nclose to one for all individuals who defaulted, and a number close to zero\\nfor all individuals who did not. This intuition can be formalized using a\\nmathematical equation called a likelihood function:\\nlikelihood\\nfunction\\nℓ(β0, β1) =\\nE\\ni:yi=1\\np(xi)\\nE\\ni′:yi′=0\\n(1 −p(xi′)).\\n(4.5)\\nThe estimates ˆβ0 and ˆβ1 are chosen to maximize this likelihood function.\\nMaximum likelihood is a very general approach that is used to fit many\\nof the non-linear models that we examine throughout this book. In the\\nlinear regression setting, the least squares approach is in fact a special case\\nof maximum likelihood. The mathematical details of maximum likelihood\\nare beyond the scope of this book. However, in general, logistic regression\\nand other models can be easily fit using statistical software such as R, and\\nso we do not need to concern ourselves with the details of the maximum\\nlikelihood fitting procedure.\\nTable 4.1 shows the coefficient estimates and related information that\\nresult from fitting a logistic regression model on the Default data in order\\nto predict the probability of default=Yes using balance. We see that ˆβ1 =\\n0.0055; this indicates that an increase in balance is associated with an\\nincrease in the probability of default. To be precise, a one-unit increase in\\nbalance is associated with an increase in the log odds of default by 0.0055\\nunits.\\nMany aspects of the logistic regression output shown in Table 4.1 are\\nsimilar to the linear regression output of Chapter 3. For example, we can\\nmeasure the accuracy of the coefficient estimates by computing their stan-\\ndard errors. The z-statistic in Table 4.1 plays the same role as the t-statistic\\nin the linear regression output, for example in Table 3.1 on page 77. For\\ninstance, the z-statistic associated with β1 is equal to ˆβ1/SE(ˆβ1), and so a\\nlarge (absolute) value of the z-statistic indicates evidence against the null\\nhypothesis H0 : β1 = 0. This null hypothesis implies that p(X) =\\neβ0\\n1+eβ0 : in\\nother words, that the probability of default does not depend on balance.\\nSince the p-value associated with balance in Table 4.1 is tiny, we can reject\\nH0. In other words, we conclude that there is indeed an association between\\nbalance and probability of default. The estimated intercept in Table 4.1\\nis typically not of interest; its main purpose is to adjust the average fitted\\nprobabilities to the proportion of ones in the data (in this case, the overall\\ndefault rate).\\n4.3.3\\nMaking Predictions\\nOnce the coefficients have been estimated, we can compute the probability\\nof default for any given credit card balance. For example, using the coeffi-\\ncient estimates given in Table 4.1, we predict that the default probability 142\\n4. Classification\\nCoefficient\\nStd. error\\nz-statistic\\np-value\\nIntercept\\n−10.6513\\n0.3612\\n−29.5\\n<0.0001\\nbalance\\n0.0055\\n0.0002\\n24.9\\n<0.0001\\nTABLE 4.1. For the Default data, estimated coefficients of the logistic regres-\\nsion model that predicts the probability of default using balance. A one-unit\\nincrease in balance is associated with an increase in the log odds of default by\\n0.0055 units.\\nCoefficient\\nStd. error\\nz-statistic\\np-value\\nIntercept\\n−3.5041\\n0.0707\\n−49.55\\n<0.0001\\nstudent[Yes]\\n0.4049\\n0.1150\\n3.52\\n0.0004\\nTABLE 4.2. For the Default data, estimated coefficients of the logistic regression\\nmodel that predicts the probability of default using student status. Student status\\nis encoded as a dummy variable, with a value of 1 for a student and a value of 0\\nfor a non-student, and represented by the variable student[Yes] in the table.\\nfor an individual with a balance of $1, 000 is\\nˆp(X) =\\ne ˆβ0+ ˆβ1X\\n1 + e ˆβ0+ ˆβ1X =\\ne−10.6513+0.0055×1,000\\n1 + e−10.6513+0.0055×1,000 = 0.00576,\\nwhich is below 1 %. In contrast, the predicted probability of default for an\\nindividual with a balance of $2, 000 is much higher, and equals 0.586 or\\n58.6 %.\\nOne can use qualitative predictors with the logistic regression model us-\\ning the dummy variable approach from Section 3.3.1. As an example, the\\nDefault data set contains the qualitative variable student. To fit a model\\nthat uses student status as a predictor variable, we simply create a dummy\\nvariable that takes on a value of 1 for students and 0 for non-students. The\\nlogistic regression model that results from predicting probability of default\\nfrom student status can be seen in Table 4.2. The coefficient associated\\nwith the dummy variable is positive, and the associated p-value is statisti-\\ncally significant. This indicates that students tend to have higher default\\nprobabilities than non-students:\\n6\\nPr(default=Yes|student=Yes) =\\ne−3.5041+0.4049×1\\n1 + e−3.5041+0.4049×1 = 0.0431,\\n6\\nPr(default=Yes|student=No) =\\ne−3.5041+0.4049×0\\n1 + e−3.5041+0.4049×0 = 0.0292.\\n4.3.4\\nMultiple Logistic Regression\\nWe now consider the problem of predicting a binary response using multiple\\npredictors. By analogy with the extension from simple to multiple linear\\nregression in Chapter 3, we can generalize (4.4) as follows:\\nlog\\n*\\np(X)\\n1 −p(X)\\n+\\n= β0 + β1X1 + · · · + βpXp,\\n(4.6)\\nwhere X = (X1, . . . , Xp) are p predictors. Equation 4.6 can be rewritten as\\np(X) =\\neβ0+β1X1+···+βpXp\\n1 + eβ0+β1X1+···+βpXp .\\n(4.7) 4.3 Logistic Regression\\n143\\nCoefficient\\nStd. error\\nz-statistic\\np-value\\nIntercept\\n−10.8690\\n0.4923\\n−22.08\\n<0.0001\\nbalance\\n0.0057\\n0.0002\\n24.74\\n<0.0001\\nincome\\n0.0030\\n0.0082\\n0.37\\n0.7115\\nstudent[Yes]\\n−0.6468\\n0.2362\\n−2.74\\n0.0062\\nTABLE 4.3. For the Default data, estimated coefficients of the logistic regression\\nmodel that predicts the probability of default using balance, income, and student\\nstatus. Student status is encoded as a dummy variable student[Yes], with a value\\nof 1 for a student and a value of 0 for a non-student. In fitting this model, income\\nwas measured in thousands of dollars.\\nJust as in Section 4.3.2, we use the maximum likelihood method to estimate\\nβ0, β1, . . . , βp.\\nTable 4.3 shows the coefficient estimates for a logistic regression model\\nthat uses balance, income (in thousands of dollars), and student status to\\npredict probability of default. There is a surprising result here. The p-\\nvalues associated with balance and the dummy variable for student status\\nare very small, indicating that each of these variables is associated with\\nthe probability of default. However, the coefficient for the dummy variable\\nis negative, indicating that students are less likely to default than non-\\nstudents. In contrast, the coefficient for the dummy variable is positive in\\nTable 4.2. How is it possible for student status to be associated with an\\nincrease in probability of default in Table 4.2 and a decrease in probability\\nof default in Table 4.3? The left-hand panel of Figure 4.3 provides a graph-\\nical illustration of this apparent paradox. The orange and blue solid lines\\nshow the average default rates for students and non-students, respectively,\\nas a function of credit card balance. The negative coefficient for student in\\nthe multiple logistic regression indicates that for a fixed value of balance\\nand income, a student is less likely to default than a non-student. Indeed,\\nwe observe from the left-hand panel of Figure 4.3 that the student default\\nrate is at or below that of the non-student default rate for every value of\\nbalance. But the horizontal broken lines near the base of the plot, which\\nshow the default rates for students and non-students averaged over all val-\\nues of balance and income, suggest the opposite effect: the overall student\\ndefault rate is higher than the non-student default rate. Consequently, there\\nis a positive coefficient for student in the single variable logistic regression\\noutput shown in Table 4.2.\\nThe right-hand panel of Figure 4.3 provides an explanation for this dis-\\ncrepancy. The variables student and balance are correlated. Students tend\\nto hold higher levels of debt, which is in turn associated with higher prob-\\nability of default. In other words, students are more likely to have large\\ncredit card balances, which, as we know from the left-hand panel of Fig-\\nure 4.3, tend to be associated with high default rates. Thus, even though\\nan individual student with a given credit card balance will tend to have a\\nlower probability of default than a non-student with the same credit card\\nbalance, the fact that students on the whole tend to have higher credit card\\nbalances means that overall, students tend to default at a higher rate than\\nnon-students. This is an important distinction for a credit card company\\nthat is trying to determine to whom they should offer credit. A student is\\nriskier than a non-student if no information about the student’s credit card 144\\n4. Classification\\n500\\n1000\\n1500\\n2000\\n0.0\\n0.2\\n0.4\\n0.6\\n0.8\\nCredit Card Balance\\nDefault Rate\\nNo\\nYes\\n0\\n500\\n1000\\n1500\\n2000\\n2500\\nStudent Status\\nCredit Card Balance\\nFIGURE 4.3. Confounding in the Default data. Left: Default rates are shown\\nfor students (orange) and non-students (blue). The solid lines display default rate\\nas a function of balance, while the horizontal broken lines display the overall\\ndefault rates. Right: Boxplots of balance for students (orange) and non-students\\n(blue) are shown.\\nbalance is available. However, that student is less risky than a non-student\\nwith the same credit card balance!\\nThis simple example illustrates the dangers and subtleties associated\\nwith performing regressions involving only a single predictor when other\\npredictors may also be relevant. As in the linear regression setting, the\\nresults obtained using one predictor may be quite different from those ob-\\ntained using multiple predictors, especially when there is correlation among\\nthe predictors. In general, the phenomenon seen in Figure 4.3 is known as\\nconfounding.\\nconfounding\\nBy substituting estimates for the regression coefficients from Table 4.3\\ninto (4.7), we can make predictions. For example, a student with a credit\\ncard balance of $1, 500 and an income of $40, 000 has an estimated proba-\\nbility of default of\\nˆp(X) =\\ne−10.869+0.00574×1,500+0.003×40−0.6468×1\\n1 + e−10.869+0.00574×1,500+0.003×40−0.6468×1 = 0.058.\\n(4.8)\\nA non-student with the same balance and income has an estimated prob-\\nability of default of\\nˆp(X) =\\ne−10.869+0.00574×1,500+0.003×40−0.6468×0\\n1 + e−10.869+0.00574×1,500+0.003×40−0.6468×0 = 0.105.\\n(4.9)\\n(Here we multiply the income coefficient estimate from Table 4.3 by 40,\\nrather than by 40,000, because in that table the model was fit with income\\nmeasured in units of $1, 000.)\\n4.3.5\\nMultinomial Logistic Regression\\nWe sometimes wish to classify a response variable that has more than two\\nclasses. For example, in Section 4.2 we had three categories of medical con-\\ndition in the emergency room: stroke, drug overdose, epileptic seizure.\\nHowever, the logistic regression approach that we have seen in this section\\nonly allows for K = 2 classes for the response variable. 4.3 Logistic Regression\\n145\\nIt turns out that it is possible to extend the two-class logistic regression\\napproach to the setting of K > 2 classes. This extension is sometimes\\nknown as multinomial logistic regression. To do this, we first select a single multinomial\\nlogistic\\nregression\\nclass to serve as the baseline; without loss of generality, we select the Kth\\nclass for this role. Then we replace the model (4.7) with the model\\nPr(Y = k|X = x) =\\neβk0+βk1x1+···+βkpxp\\n1 + )K−1\\nl=1 eβl0+βl1x1+···+βlpxp\\n(4.10)\\nfor k = 1, . . . , K−1, and\\nPr(Y = K|X = x) =\\n1\\n1 + )K−1\\nl=1 eβl0+βl1x1+···+βlpxp .\\n(4.11)\\nIt is not hard to show that for k = 1, . . . , K−1,\\nlog\\n* Pr(Y = k|X = x)\\nPr(Y = K|X = x)\\n+\\n= βk0 + βk1x1 + · · · + βkpxp.\\n(4.12)\\nNotice that (4.12) is quite similar to (4.6). Equation 4.12 indicates that once\\nagain, the log odds between any pair of classes is linear in the features.\\nIt turns out that in (4.10)–(4.12), the decision to treat the Kth class as\\nthe baseline is unimportant. For example, when classifying emergency room\\nvisits into stroke, drug overdose, and epileptic seizure, suppose that we\\nfit two multinomial logistic regression models: one treating stroke as the\\nbaseline, another treating drug overdose as the baseline. The coefficient\\nestimates will differ between the two fitted models due to the differing\\nchoice of baseline, but the fitted values (predictions), the log odds between\\nany pair of classes, and the other key model outputs will remain the same.\\nNonetheless, interpretation of the coefficients in a multinomial logistic\\nregression model must be done with care, since it is tied to the choice\\nof baseline. For example, if we set epileptic seizure to be the baseline,\\nthen we can interpret βstroke0 as the log odds of stroke versus epileptic\\nseizure, given that x1 = · · · = xp = 0. Furthermore, a one-unit increase\\nin Xj is associated with a βstrokej increase in the log odds of stroke over\\nepileptic seizure. Stated another way, if Xj increases by one unit, then\\nPr(Y = stroke|X = x)\\nPr(Y = epileptic seizure|X = x)\\nincreases by eβstrokej.\\nWe now briefly present an alternative coding for multinomial logistic\\nregression, known as the softmax coding. The softmax coding is equivalent softmax\\nto the coding just described in the sense that the fitted values, log odds\\nbetween any pair of classes, and other key model outputs will remain the\\nsame, regardless of coding. But the softmax coding is used extensively in\\nsome areas of the machine learning literature (and will appear again in\\nChapter 10), so it is worth being aware of it. In the softmax coding, rather\\nthan selecting a baseline class, we treat all K classes symmetrically, and\\nassume that for k = 1, . . . , K,\\nPr(Y = k|X = x) =\\neβk0+βk1x1+···+βkpxp\\n)K\\nl=1 eβl0+βl1x1+···+βlpxp .\\n(4.13) 146\\n4. Classification\\nThus, rather than estimating coefficients for K −1 classes, we actually\\nestimate coefficients for all K classes. It is not hard to see that as a result\\nof (4.13), the log odds ratio between the kth and k′th classes equals\\nlog\\n% Pr(Y = k|X = x)\\nPr(Y = k′|X = x)\\n&\\n= (βk0 −βk′0) + (βk1 −βk′1)x1 + · · · + (βkp −βk′p)xp.\\n(4.14)\\n4.4\\nGenerative Models for Classification\\nLogistic regression involves directly modeling Pr(Y = k|X = x) using the\\nlogistic function, given by (4.7) for the case of two response classes. In\\nstatistical jargon, we model the conditional distribution of the response Y ,\\ngiven the predictor(s) X. We now consider an alternative and less direct\\napproach to estimating these probabilities. In this new approach, we model\\nthe distribution of the predictors X separately in each of the response\\nclasses (i.e. for each value of Y ). We then use Bayes’ theorem to flip these\\naround into estimates for Pr(Y = k|X = x). When the distribution of X\\nwithin each class is assumed to be normal, it turns out that the model is\\nvery similar in form to logistic regression.\\nWhy do we need another method, when we have logistic regression?\\nThere are several reasons:\\n• When there is substantial separation between the two classes, the\\nparameter estimates for the logistic regression model are surprisingly\\nunstable. The methods that we consider in this section do not suffer\\nfrom this problem.\\n• If the distribution of the predictors X is approximately normal in\\neach of the classes and the sample size is small, then the approaches\\nin this section may be more accurate than logistic regression.\\n• The methods in this section can be naturally extended to the case\\nof more than two response classes. (In the case of more than two\\nresponse classes, we can also use multinomial logistic regression from\\nSection 4.3.5.)\\nSuppose that we wish to classify an observation into one of K classes,\\nwhere K ≥2. In other words, the qualitative response variable Y can take\\non K possible distinct and unordered values. Let πk represent the overall\\nor prior probability that a randomly chosen observation comes from the prior\\nkth class. Let fk(X) ≡Pr(X|Y = k)1 denote the density function of X\\ndensity\\nfunction\\nfor an observation that comes from the kth class. In other words, fk(x) is\\nrelatively large if there is a high probability that an observation in the kth\\nclass has X ≈x, and fk(x) is small if it is very unlikely that an observation\\nin the kth class has X ≈x. Then Bayes’ theorem states that\\nBayes’\\ntheorem\\n1Technically, this definition is only correct if X is a qualitative random variable. If\\nX is quantitative, then fk(x)dx corresponds to the probability of X falling in a small\\nregion dx around x. 4.4 Generative Models for Classification\\n147\\nPr(Y = k|X = x) =\\nπkfk(x)\\n)K\\nl=1 πlfl(x)\\n.\\n(4.15)\\nIn accordance with our earlier notation, we will use the abbreviation pk(x) =\\nPr(Y = k|X = x); this is the posterior probability that an observation posterior\\nX = x belongs to the kth class. That is, it is the probability that the\\nobservation belongs to the kth class, given the predictor value for that\\nobservation.\\nEquation 4.15 suggests that instead of directly computing the posterior\\nprobability pk(x) as in Section 4.3.1, we can simply plug in estimates of πk\\nand fk(x) into (4.15). In general, estimating πk is easy if we have a random\\nsample from the population: we simply compute the fraction of the training\\nobservations that belong to the kth class. However, estimating the density\\nfunction fk(x) is much more challenging. As we will see, to estimate fk(x),\\nwe will typically have to make some simplifying assumptions.\\nWe know from Chapter 2 that the Bayes classifier, which classifies an\\nobservation x to the class for which pk(x) is largest, has the lowest possible\\nerror rate out of all classifiers. (Of course, this is only true if all of the\\nterms in (4.15) are correctly specified.) Therefore, if we can find a way to\\nestimate fk(x), then we can plug it into (4.15) in order to approximate the\\nBayes classifier.\\nIn the following sections, we discuss three classifiers that use different\\nestimates of fk(x) in (4.15) to approximate the Bayes classifier: linear dis-\\ncriminant analysis, quadratic discriminant analysis, and naive Bayes.\\n4.4.1\\nLinear Discriminant Analysis for p = 1\\nFor now, assume that p = 1—that is, we have only one predictor. We would\\nlike to obtain an estimate for fk(x) that we can plug into (4.15) in order to\\nestimate pk(x). We will then classify an observation to the class for which\\npk(x) is greatest. To estimate fk(x), we will first make some assumptions\\nabout its form.\\nIn particular, we assume that fk(x) is normal or Gaussian. In the one- normal\\nGaussian\\ndimensional setting, the normal density takes the form\\nfk(x) =\\n1\\n√\\n2πσk\\nexp\\n*\\n−1\\n2σ2\\nk\\n(x −µk)2\\n+\\n,\\n(4.16)\\nwhere µk and σ2\\nk are the mean and variance parameters for the kth class.\\nFor now, let us further assume that σ2\\n1 = · · · = σ2\\nK: that is, there is a shared\\nvariance term across all K classes, which for simplicity we can denote by\\nσ2. Plugging (4.16) into (4.15), we find that\\npk(x) =\\nπk\\n1\\n√\\n2πσ exp\\n\\'\\n−\\n1\\n2σ2 (x −µk)2(\\n)K\\nl=1 πl\\n1\\n√\\n2πσ exp\\n\\'\\n−\\n1\\n2σ2 (x −µl)2(.\\n(4.17)\\n(Note that in (4.17), πk denotes the prior probability that an observation\\nbelongs to the kth class, not to be confused with π ≈3.14159, the math-\\nematical constant.) The Bayes classifier2 involves assigning an observation\\n2Recall that the Bayes classifier assigns an observation to the class for which pk(x)\\nis largest. This is different from Bayes’ theorem in (4.15), which allows us to manipulate\\nconditional distributions. 148\\n4. Classification\\n−4\\n−2\\n0\\n2\\n4\\n−3\\n−2\\n−1\\n0\\n1\\n2\\n3\\n4\\n0\\n1\\n2\\n3\\n4\\n5\\nFIGURE 4.4. Left: Two one-dimensional normal density functions are shown.\\nThe dashed vertical line represents the Bayes decision boundary. Right: 20 obser-\\nvations were drawn from each of the two classes, and are shown as histograms.\\nThe Bayes decision boundary is again shown as a dashed vertical line. The solid\\nvertical line represents the LDA decision boundary estimated from the training\\ndata.\\nX = x to the class for which (4.17) is largest. Taking the log of (4.17) and\\nrearranging the terms, it is not hard to show3 that this is equivalent to\\nassigning the observation to the class for which\\nδk(x) = x · µk\\nσ2 −µ2\\nk\\n2σ2 + log(πk)\\n(4.18)\\nis largest. For instance, if K = 2 and π1 = π2, then the Bayes classifier\\nassigns an observation to class 1 if 2x (µ1 −µ2) > µ2\\n1 −µ2\\n2, and to class\\n2 otherwise. The Bayes decision boundary is the point for which δ1(x) =\\nδ2(x); one can show that this amounts to\\nx =\\nµ2\\n1 −µ2\\n2\\n2(µ1 −µ2) = µ1 + µ2\\n2\\n.\\n(4.19)\\nAn example is shown in the left-hand panel of Figure 4.4. The two normal\\ndensity functions that are displayed, f1(x) and f2(x), represent two distinct\\nclasses. The mean and variance parameters for the two density functions\\nare µ1 = −1.25, µ2 = 1.25, and σ2\\n1 = σ2\\n2 = 1. The two densities overlap,\\nand so given that X = x, there is some uncertainty about the class to which\\nthe observation belongs. If we assume that an observation is equally likely\\nto come from either class—that is, π1 = π2 = 0.5—then by inspection of\\n(4.19), we see that the Bayes classifier assigns the observation to class 1\\nif x < 0 and class 2 otherwise. Note that in this case, we can compute\\nthe Bayes classifier because we know that X is drawn from a Gaussian\\ndistribution within each class, and we know all of the parameters involved.\\nIn a real-life situation, we are not able to calculate the Bayes classifier.\\nIn practice, even if we are quite certain of our assumption that X is\\ndrawn from a Gaussian distribution within each class, to apply the Bayes\\nclassifier we still have to estimate the parameters µ1, . . . , µK, π1, . . . , πK,\\nand σ2. The linear discriminant analysis (LDA) method approximates the linear\\ndiscriminant\\nanalysis\\nBayes classifier by plugging estimates for πk, µk, and σ2 into (4.18). In\\n3See Exercise 2 at the end of this chapter. 4.4 Generative Models for Classification\\n149\\nparticular, the following estimates are used:\\nˆµk\\n=\\n1\\nnk\\n0\\ni:yi=k\\nxi\\nˆσ2\\n=\\n1\\nn −K\\nK\\n0\\nk=1\\n0\\ni:yi=k\\n(xi −ˆµk)2\\n(4.20)\\nwhere n is the total number of training observations, and nk is the number\\nof training observations in the kth class. The estimate for µk is simply the\\naverage of all the training observations from the kth class, while ˆσ2 can\\nbe seen as a weighted average of the sample variances for each of the K\\nclasses. Sometimes we have knowledge of the class membership probabili-\\nties π1, . . . , πK, which can be used directly. In the absence of any additional\\ninformation, LDA estimates πk using the proportion of the training obser-\\nvations that belong to the kth class. In other words,\\nˆπk = nk/n.\\n(4.21)\\nThe LDA classifier plugs the estimates given in (4.20) and (4.21) into (4.18),\\nand assigns an observation X = x to the class for which\\nˆδk(x) = x · ˆµk\\nˆσ2 −ˆµ2\\nk\\n2ˆσ2 + log(ˆπk)\\n(4.22)\\nis largest. The word linear in the classifier’s name stems from the fact\\nthat the discriminant functions ˆδk(x) in (4.22) are linear functions of x (as discriminant\\nfunction\\nopposed to a more complex function of x).\\nThe right-hand panel of Figure 4.4 displays a histogram of a random\\nsample of 20 observations from each class. To implement LDA, we began\\nby estimating πk, µk, and σ2 using (4.20) and (4.21). We then computed the\\ndecision boundary, shown as a black solid line, that results from assigning\\nan observation to the class for which (4.22) is largest. All points to the left\\nof this line will be assigned to the green class, while points to the right of\\nthis line are assigned to the purple class. In this case, since n1 = n2 = 20,\\nwe have ˆπ1 = ˆπ2. As a result, the decision boundary corresponds to the\\nmidpoint between the sample means for the two classes, (ˆµ1 + ˆµ2)/2. The\\nfigure indicates that the LDA decision boundary is slightly to the left of\\nthe optimal Bayes decision boundary, which instead equals (µ1 + µ2)/2 =\\n0. How well does the LDA classifier perform on this data? Since this is\\nsimulated data, we can generate a large number of test observations in order\\nto compute the Bayes error rate and the LDA test error rate. These are\\n10.6 % and 11.1 %, respectively. In other words, the LDA classifier’s error\\nrate is only 0.5 % above the smallest possible error rate! This indicates that\\nLDA is performing pretty well on this data set.\\nTo reiterate, the LDA classifier results from assuming that the obser-\\nvations within each class come from a normal distribution with a class-\\nspecific mean and a common variance σ2, and plugging estimates for these\\nparameters into the Bayes classifier. In Section 4.4.3, we will consider a less\\nstringent set of assumptions, by allowing the observations in the kth class\\nto have a class-specific variance, σ2\\nk. 150\\n4. Classification\\nx1\\nx1\\nx2\\nx2\\nFIGURE 4.5.\\nTwo multivariate Gaussian density functions are shown, with\\np = 2. Left: The two predictors are uncorrelated. Right: The two variables have\\na correlation of 0.7.\\n4.4.2\\nLinear Discriminant Analysis for p >1\\nWe now extend the LDA classifier to the case of multiple predictors. To\\ndo this, we will assume that X = (X1, X2, . . . , Xp) is drawn from a multi-\\nvariate Gaussian (or multivariate normal) distribution, with a class-specific multivariate\\nGaussian\\nmean vector and a common covariance matrix. We begin with a brief review\\nof this distribution.\\nThe multivariate Gaussian distribution assumes that each individual pre-\\ndictor follows a one-dimensional normal distribution, as in (4.16), with some\\ncorrelation between each pair of predictors. Two examples of multivariate\\nGaussian distributions with p = 2 are shown in Figure 4.5. The height of\\nthe surface at any particular point represents the probability that both X1\\nand X2 fall in a small region around that point. In either panel, if the sur-\\nface is cut along the X1 axis or along the X2 axis, the resulting cross-section\\nwill have the shape of a one-dimensional normal distribution. The left-hand\\npanel of Figure 4.5 illustrates an example in which Var(X1) = Var(X2) and\\nCor(X1, X2) = 0; this surface has a characteristic bell shape. However, the\\nbell shape will be distorted if the predictors are correlated or have unequal\\nvariances, as is illustrated in the right-hand panel of Figure 4.5. In this\\nsituation, the base of the bell will have an elliptical, rather than circular,\\nshape. To indicate that a p-dimensional random variable X has a multi-\\nvariate Gaussian distribution, we write X ∼N(µ, Σ). Here E(X) = µ is\\nthe mean of X (a vector with p components), and Cov(X) = Σ is the\\np × p covariance matrix of X. Formally, the multivariate Gaussian density\\nis defined as\\nf(x) =\\n1\\n(2π)p/2|Σ|1/2 exp\\n*\\n−1\\n2(x −µ)T Σ−1(x −µ)\\n+\\n.\\n(4.23)\\nIn the case of p > 1 predictors, the LDA classifier assumes that the\\nobservations in the kth class are drawn from a multivariate Gaussian dis-\\ntribution N(µk, Σ), where µk is a class-specific mean vector, and Σ is a\\ncovariance matrix that is common to all K classes. Plugging the density\\nfunction for the kth class, fk(X = x), into (4.15) and performing a little\\nbit of algebra reveals that the Bayes classifier assigns an observation X = x 4.4 Generative Models for Classification\\n151\\n−4\\n−2\\n0\\n2\\n4\\n−4\\n−2\\n0\\n2\\n4\\n−4\\n−2\\n0\\n2\\n4\\n−4\\n−2\\n0\\n2\\n4\\nX1\\nX1\\nX2\\nX2\\nFIGURE 4.6. An example with three classes. The observations from each class\\nare drawn from a multivariate Gaussian distribution with p = 2, with a class-spe-\\ncific mean vector and a common covariance matrix. Left: Ellipses that contain\\n95 % of the probability for each of the three classes are shown. The dashed lines\\nare the Bayes decision boundaries. Right: 20 observations were generated from\\neach class, and the corresponding LDA decision boundaries are indicated using\\nsolid black lines. The Bayes decision boundaries are once again shown as dashed\\nlines.\\nto the class for which\\nδk(x) = xT Σ−1µk −1\\n2µT\\nk Σ−1µk + log πk\\n(4.24)\\nis largest. This is the vector/matrix version of (4.18).\\nAn example is shown in the left-hand panel of Figure 4.6. Three equally-\\nsized Gaussian classes are shown with class-specific mean vectors and a\\ncommon covariance matrix. The three ellipses represent regions that con-\\ntain 95 % of the probability for each of the three classes. The dashed lines\\nare the Bayes decision boundaries. In other words, they represent the set\\nof values x for which δk(x) = δℓ(x); i.e.\\nxT Σ−1µk −1\\n2µT\\nk Σ−1µk = xT Σ−1µl −1\\n2µT\\nl Σ−1µl\\n(4.25)\\nfor k ̸= l. (The log πk term from (4.24) has disappeared because each of\\nthe three classes has the same number of training observations; i.e. πk is\\nthe same for each class.) Note that there are three lines representing the\\nBayes decision boundaries because there are three pairs of classes among\\nthe three classes. That is, one Bayes decision boundary separates class 1\\nfrom class 2, one separates class 1 from class 3, and one separates class 2\\nfrom class 3. These three Bayes decision boundaries divide the predictor\\nspace into three regions. The Bayes classifier will classify an observation\\naccording to the region in which it is located.\\nOnce again, we need to estimate the unknown parameters µ1, . . . , µK,\\nπ1, . . . , πK, and Σ; the formulas are similar to those used in the one-\\ndimensional case, given in (4.20). To assign a new observation X = x,\\nLDA plugs these estimates into (4.24) to obtain quantities ˆδk(x), and clas-\\nsifies to the class for which ˆδk(x) is largest. Note that in (4.24) δk(x) is\\na linear function of x; that is, the LDA decision rule depends on x only 152\\n4. Classification\\nTrue default status\\nNo\\nYes\\nTotal\\nPredicted\\nNo\\n9644\\n252\\n9896\\ndefault status\\nYes\\n23\\n81\\n104\\nTotal\\n9667\\n333\\n10000\\nTABLE 4.4.\\nA confusion matrix compares the LDA predictions to the true\\ndefault statuses for the 10,000 training observations in the Default data set.\\nElements on the diagonal of the matrix represent individuals whose default statuses\\nwere correctly predicted, while off-diagonal elements represent individuals that\\nwere misclassified. LDA made incorrect predictions for 23 individuals who did\\nnot default and for 252 individuals who did default.\\nthrough a linear combination of its elements. As previously discussed, this\\nis the reason for the word linear in LDA.\\nIn the right-hand panel of Figure 4.6, 20 observations drawn from each of\\nthe three classes are displayed, and the resulting LDA decision boundaries\\nare shown as solid black lines. Overall, the LDA decision boundaries are\\npretty close to the Bayes decision boundaries, shown again as dashed lines.\\nThe test error rates for the Bayes and LDA classifiers are 0.0746 and 0.0770,\\nrespectively. This indicates that LDA is performing well on this data.\\nWe can perform LDA on the Default data in order to predict whether\\nor not an individual will default on the basis of credit card balance and\\nstudent status.4 The LDA model fit to the 10,000 training samples results\\nin a training error rate of 2.75 %. This sounds like a low error rate, but two\\ncaveats must be noted.\\n• First of all, training error rates will usually be lower than test error\\nrates, which are the real quantity of interest. In other words, we\\nmight expect this classifier to perform worse if we use it to predict\\nwhether or not a new set of individuals will default. The reason is\\nthat we specifically adjust the parameters of our model to do well on\\nthe training data. The higher the ratio of parameters p to number\\nof samples n, the more we expect this overfitting to play a role. For overfitting\\nthese data we don’t expect this to be a problem, since p = 2 and\\nn = 10, 000.\\n• Second, since only 3.33 % of the individuals in the training sample\\ndefaulted, a simple but useless classifier that always predicts that\\nan individual will not default, regardless of his or her credit card\\nbalance and student status, will result in an error rate of 3.33 %. In\\nother words, the trivial null classifier will achieve an error rate that null\\nis only a bit higher than the LDA training set error rate.\\nIn practice, a binary classifier such as this one can make two types of\\nerrors: it can incorrectly assign an individual who defaults to the no default\\ncategory, or it can incorrectly assign an individual who does not default to\\n4The careful reader will notice that student status is qualitative — thus, the normality\\nassumption made by LDA is clearly violated in this example! However, LDA is often\\nremarkably robust to model violations, as this example shows. Naive Bayes, discussed in\\nSection 4.4.4, provides an alternative to LDA that does not assume normally distributed\\npredictors. 4.4 Generative Models for Classification\\n153\\nthe default category. It is often of interest to determine which of these two\\ntypes of errors are being made. A confusion matrix, shown for the Default confusion\\nmatrix\\ndata in Table 4.4, is a convenient way to display this information. The\\ntable reveals that LDA predicted that a total of 104 people would default.\\nOf these people, 81 actually defaulted and 23 did not. Hence only 23 out\\nof 9,667 of the individuals who did not default were incorrectly labeled.\\nThis looks like a pretty low error rate! However, of the 333 individuals who\\ndefaulted, 252 (or 75.7 %) were missed by LDA. So while the overall error\\nrate is low, the error rate among individuals who defaulted is very high.\\nFrom the perspective of a credit card company that is trying to identify\\nhigh-risk individuals, an error rate of 252/333 = 75.7 % among individuals\\nwho default may well be unacceptable.\\nClass-specific performance is also important in medicine and biology,\\nwhere the terms sensitivity and specificity characterize the performance of sensitivity\\nspecificity\\na classifier or screening test. In this case the sensitivity is the percent-\\nage of true defaulters that are identified; it equals 24.3 %. The specificity\\nis the percentage of non-defaulters that are correctly identified; it equals\\n(1 −23/9667) = 99.8 %.\\nWhy does LDA do such a poor job of classifying the customers who de-\\nfault? In other words, why does it have such low sensitivity? As we have\\nseen, LDA is trying to approximate the Bayes classifier, which has the low-\\nest total error rate out of all classifiers. That is, the Bayes classifier will\\nyield the smallest possible total number of misclassified observations, re-\\ngardless of the class from which the errors stem. Some misclassifications will\\nresult from incorrectly assigning a customer who does not default to the\\ndefault class, and others will result from incorrectly assigning a customer\\nwho defaults to the non-default class. In contrast, a credit card company\\nmight particularly wish to avoid incorrectly classifying an individual who\\nwill default, whereas incorrectly classifying an individual who will not de-\\nfault, though still to be avoided, is less problematic. We will now see that it\\nis possible to modify LDA in order to develop a classifier that better meets\\nthe credit card company’s needs.\\nThe Bayes classifier works by assigning an observation to the class for\\nwhich the posterior probability pk(X) is greatest. In the two-class case, this\\namounts to assigning an observation to the default class if\\nPr(default = Yes|X = x) > 0.5.\\n(4.26)\\nThus, the Bayes classifier, and by extension LDA, uses a threshold of 50 %\\nfor the posterior probability of default in order to assign an observation\\nto the default class. However, if we are concerned about incorrectly pre-\\ndicting the default status for individuals who default, then we can consider\\nlowering this threshold. For instance, we might label any customer with a\\nposterior probability of default above 20 % to the default class. In other\\nwords, instead of assigning an observation to the default class if (4.26)\\nholds, we could instead assign an observation to this class if\\nPr(default = Yes|X = x) > 0.2.\\n(4.27)\\nThe error rates that result from taking this approach are shown in Table 4.5.\\nNow LDA predicts that 430 individuals will default. Of the 333 individuals\\nwho default, LDA correctly predicts all but 138, or 41.4 %. This is a vast 154\\n4. Classification\\nTrue default status\\nNo\\nYes\\nTotal\\nPredicted\\nNo\\n9432\\n138\\n9570\\ndefault status\\nYes\\n235\\n195\\n430\\nTotal\\n9667\\n333\\n10000\\nTABLE 4.5.\\nA confusion matrix compares the LDA predictions to the true\\ndefault statuses for the 10,000 training observations in the Default data set,\\nusing a modified threshold value that predicts default for any individuals whose\\nposterior default probability exceeds 20 %.\\n0.0\\n0.1\\n0.2\\n0.3\\n0.4\\n0.5\\n0.0\\n0.2\\n0.4\\n0.6\\nThreshold\\nError Rate\\nFIGURE 4.7. For the Default data set, error rates are shown as a function of\\nthe threshold value for the posterior probability that is used to perform the assign-\\nment. The black solid line displays the overall error rate. The blue dashed line\\nrepresents the fraction of defaulting customers that are incorrectly classified, and\\nthe orange dotted line indicates the fraction of errors among the non-defaulting\\ncustomers.\\nimprovement over the error rate of 75.7 % that resulted from using the\\nthreshold of 50 %. However, this improvement comes at a cost: now 235\\nindividuals who do not default are incorrectly classified. As a result, the\\noverall error rate has increased slightly to 3.73 %. But a credit card company\\nmay consider this slight increase in the total error rate to be a small price to\\npay for more accurate identification of individuals who do indeed default.\\nFigure 4.7 illustrates the trade-off that results from modifying the thresh-\\nold value for the posterior probability of default. Various error rates are\\nshown as a function of the threshold value. Using a threshold of 0.5, as in\\n(4.26), minimizes the overall error rate, shown as a black solid line. This\\nis to be expected, since the Bayes classifier uses a threshold of 0.5 and is\\nknown to have the lowest overall error rate. But when a threshold of 0.5 is\\nused, the error rate among the individuals who default is quite high (blue\\ndashed line). As the threshold is reduced, the error rate among individuals\\nwho default decreases steadily, but the error rate among the individuals\\nwho do not default increases. How can we decide which threshold value is\\nbest? Such a decision must be based on domain knowledge, such as detailed\\ninformation about the costs associated with default.\\nThe ROC curve is a popular graphic for simultaneously displaying the ROC curve\\ntwo types of errors for all possible thresholds. The name “ROC” is historic,\\nand comes from communications theory. It is an acronym for receiver op-\\nerating characteristics. Figure 4.8 displays the ROC curve for the LDA\\nclassifier on the training data. The overall performance of a classifier, sum- 4.4 Generative Models for Classification\\n155\\nROC Curve\\nFalse positive rate\\nTrue positive rate\\n0.0\\n0.2\\n0.4\\n0.6\\n0.8\\n1.0\\n0.0\\n0.2\\n0.4\\n0.6\\n0.8\\n1.0\\nFIGURE 4.8. A ROC curve for the LDA classifier on the Default data. It\\ntraces out two types of error as we vary the threshold value for the posterior\\nprobability of default. The actual thresholds are not shown. The true positive rate\\nis the sensitivity: the fraction of defaulters that are correctly identified, using\\na given threshold value. The false positive rate is 1-specificity: the fraction of\\nnon-defaulters that we classify incorrectly as defaulters, using that same threshold\\nvalue. The ideal ROC curve hugs the top left corner, indicating a high true positive\\nrate and a low false positive rate. The dotted line represents the “no information”\\nclassifier; this is what we would expect if student status and credit card balance\\nare not associated with probability of default.\\nmarized over all possible thresholds, is given by the area under the (ROC)\\ncurve (AUC). An ideal ROC curve will hug the top left corner, so the larger area under\\nthe (ROC)\\ncurve\\nthe AUC the better the classifier. For this data the AUC is 0.95, which is\\nclose to the maximum of 1.0, so would be considered very good. We expect\\na classifier that performs no better than chance to have an AUC of 0.5\\n(when evaluated on an independent test set not used in model training).\\nROC curves are useful for comparing different classifiers, since they take\\ninto account all possible thresholds. It turns out that the ROC curve for\\nthe logistic regression model of Section 4.3.4 fit to these data is virtually\\nindistinguishable from this one for the LDA model, so we do not display it\\nhere.\\nAs we have seen above, varying the classifier threshold changes its true\\npositive and false positive rate. These are also called the sensitivity and one sensitivity\\nminus the specificity of our classifier. Since there is an almost bewildering specificity\\narray of terms used in this context, we now give a summary. Table 4.6\\nshows the possible results when applying a classifier (or diagnostic test)\\nto a population. To make the connection with the epidemiology literature,\\nwe think of “+” as the “disease” that we are trying to detect, and “−” as\\nthe “non-disease” state. To make the connection to the classical hypothesis\\ntesting literature, we think of “−” as the null hypothesis and “+” as the 156\\n4. Classification\\nTrue class\\n−or Null\\n+ or Non-null\\nTotal\\nPredicted\\n−or Null\\nTrue Neg. (TN)\\nFalse Neg. (FN)\\nN∗\\nclass\\n+ or Non-null\\nFalse Pos. (FP)\\nTrue Pos. (TP)\\nP∗\\nTotal\\nN\\nP\\nTABLE 4.6. Possible results when applying a classifier or diagnostic test to a\\npopulation.\\nName\\nDefinition\\nSynonyms\\nFalse Pos. rate\\nFP/N\\nType I error, 1−Specificity\\nTrue Pos. rate\\nTP/P\\n1−Type II error, power, sensitivity, recall\\nPos. Pred. value\\nTP/P∗\\nPrecision, 1−false discovery proportion\\nNeg. Pred. value\\nTN/N∗\\nTABLE 4.7. Important measures for classification and diagnostic testing, derived\\nfrom quantities in Table 4.6.\\nalternative (non-null) hypothesis. In the context of the Default data, “+”\\nindicates an individual who defaults, and “−” indicates one who does not.\\nTable 4.7 lists many of the popular performance measures that are used in\\nthis context. The denominators for the false positive and true positive rates\\nare the actual population counts in each class. In contrast, the denominators\\nfor the positive predictive value and the negative predictive value are the\\ntotal predicted counts for each class.\\n4.4.3\\nQuadratic Discriminant Analysis\\nAs we have discussed, LDA assumes that the observations within each class\\nare drawn from a multivariate Gaussian distribution with a class-specific\\nmean vector and a covariance matrix that is common to all K classes.\\nQuadratic discriminant analysis (QDA) provides an alternative approach. quadratic\\ndiscriminant\\nanalysis\\nLike LDA, the QDA classifier results from assuming that the observations\\nfrom each class are drawn from a Gaussian distribution, and plugging es-\\ntimates for the parameters into Bayes’ theorem in order to perform pre-\\ndiction. However, unlike LDA, QDA assumes that each class has its own\\ncovariance matrix. That is, it assumes that an observation from the kth\\nclass is of the form X ∼N(µk, Σk), where Σk is a covariance matrix for\\nthe kth class. Under this assumption, the Bayes classifier assigns an obser-\\nvation X = x to the class for which\\nδk(x)\\n=\\n−1\\n2(x −µk)T Σ−1\\nk (x −µk) −1\\n2 log |Σk| + log πk\\n=\\n−1\\n2xT Σ−1\\nk x + xT Σ−1\\nk µk −1\\n2µT\\nk Σ−1\\nk µk −1\\n2 log |Σk| + log πk\\n(4.28)\\nis largest. So the QDA classifier involves plugging estimates for Σk, µk,\\nand πk into (4.28), and then assigning an observation X = x to the class\\nfor which this quantity is largest. Unlike in (4.24), the quantity x appears\\nas a quadratic function in (4.28). This is where QDA gets its name.\\nWhy does it matter whether or not we assume that the K classes share a\\ncommon covariance matrix? In other words, why would one prefer LDA to 4.4 Generative Models for Classification\\n157\\n−4\\n−2\\n0\\n2\\n4\\n−4\\n−3\\n−2\\n−1\\n0\\n1\\n2\\n−4\\n−2\\n0\\n2\\n4\\n−4\\n−3\\n−2\\n−1\\n0\\n1\\n2\\nX1\\nX1\\nX2\\nX2\\nFIGURE 4.9. Left: The Bayes (purple dashed), LDA (black dotted), and QDA\\n(green solid) decision boundaries for a two-class problem with Σ1 = Σ2. The\\nshading indicates the QDA decision rule. Since the Bayes decision boundary is\\nlinear, it is more accurately approximated by LDA than by QDA. Right: Details\\nare as given in the left-hand panel, except that Σ1 ̸= Σ2. Since the Bayes decision\\nboundary is non-linear, it is more accurately approximated by QDA than by LDA.\\nQDA, or vice-versa? The answer lies in the bias-variance trade-off. When\\nthere are p predictors, then estimating a covariance matrix requires esti-\\nmating p(p+1)/2 parameters. QDA estimates a separate covariance matrix\\nfor each class, for a total of Kp(p+1)/2 parameters. With 50 predictors this\\nis some multiple of 1,275, which is a lot of parameters. By instead assum-\\ning that the K classes share a common covariance matrix, the LDA model\\nbecomes linear in x, which means there are Kp linear coefficients to esti-\\nmate. Consequently, LDA is a much less flexible classifier than QDA, and\\nso has substantially lower variance. This can potentially lead to improved\\nprediction performance. But there is a trade-off: if LDA’s assumption that\\nthe K classes share a common covariance matrix is badly off, then LDA\\ncan suffer from high bias. Roughly speaking, LDA tends to be a better bet\\nthan QDA if there are relatively few training observations and so reducing\\nvariance is crucial. In contrast, QDA is recommended if the training set is\\nvery large, so that the variance of the classifier is not a major concern, or if\\nthe assumption of a common covariance matrix for the K classes is clearly\\nuntenable.\\nFigure 4.9 illustrates the performances of LDA and QDA in two scenarios.\\nIn the left-hand panel, the two Gaussian classes have a common correla-\\ntion of 0.7 between X1 and X2. As a result, the Bayes decision boundary\\nis linear and is accurately approximated by the LDA decision boundary.\\nThe QDA decision boundary is inferior, because it suffers from higher vari-\\nance without a corresponding decrease in bias. In contrast, the right-hand\\npanel displays a situation in which the orange class has a correlation of 0.7\\nbetween the variables and the blue class has a correlation of −0.7. Now\\nthe Bayes decision boundary is quadratic, and so QDA more accurately\\napproximates this boundary than does LDA. 158\\n4. Classification\\n4.4.4\\nNaive Bayes\\nIn previous sections, we used Bayes’ theorem (4.15) to develop the LDA\\nand QDA classifiers. Here, we use Bayes’ theorem to motivate the popular\\nnaive Bayes classifier.\\nnaive Bayes\\nRecall that Bayes’ theorem (4.15) provides an expression for the pos-\\nterior probability pk(x) = Pr(Y = k|X = x) in terms of π1, . . . , πK and\\nf1(x), . . . , fK(x). To use (4.15) in practice, we need estimates for π1, . . . , πK\\nand f1(x), . . . , fK(x). As we saw in previous sections, estimating the prior\\nprobabilities π1, . . . , πK is typically straightforward: for instance, we can\\nestimate ˆπk as the proportion of training observations belonging to the kth\\nclass, for k = 1, . . . , K.\\nHowever, estimating f1(x), . . . , fK(x) is more subtle. Recall that fk(x)\\nis the p-dimensional density function for an observation in the kth class,\\nfor k = 1, . . . , K. In general, estimating a p-dimensional density function is\\nchallenging. In LDA, we make a very strong assumption that greatly sim-\\nplifies the task: we assume that fk is the density function for a multivariate\\nnormal random variable with class-specific mean µk, and shared covariance\\nmatrix Σ. By contrast, in QDA, we assume that fk is the density function\\nfor a multivariate normal random variable with class-specific mean µk, and\\nclass-specific covariance matrix Σk. By making these very strong assump-\\ntions, we are able to replace the very challenging problem of estimating K\\np-dimensional density functions with the much simpler problem of estimat-\\ning K p-dimensional mean vectors and one (in the case of LDA) or K (in\\nthe case of QDA) (p × p)-dimensional covariance matrices.\\nThe naive Bayes classifier takes a different tack for estimating f1(x), . . . ,\\nfK(x). Instead of assuming that these functions belong to a particular\\nfamily of distributions (e.g. multivariate normal), we instead make a single\\nassumption:\\nWithin the kth class, the p predictors are independent.\\nStated mathematically, this assumption means that for k = 1, . . . , K,\\nfk(x) = fk1(x1) × fk2(x2) × · · · × fkp(xp),\\n(4.29)\\nwhere fkj is the density function of the jth predictor among observations\\nin the kth class.\\nWhy is this assumption so powerful? Essentially, estimating a p-dimen-\\nsional density function is challenging because we must consider not only\\nthe marginal distribution of each predictor — that is, the distribution of marginal\\ndistribution\\neach predictor on its own — but also the joint distribution of the predictors\\njoint\\ndistribution\\n— that is, the association between the different predictors. In the case of\\na multivariate normal distribution, the association between the different\\npredictors is summarized by the off-diagonal elements of the covariance\\nmatrix. However, in general, this association can be very hard to charac-\\nterize, and exceedingly challenging to estimate. But by assuming that the\\np covariates are independent within each class, we completely eliminate the\\nneed to worry about the association between the p predictors, because we\\nhave simply assumed that there is no association between the predictors!\\nDo we really believe the naive Bayes assumption that the p covariates\\nare independent within each class? In most settings, we do not. But even\\nthough this modeling assumption is made for convenience, it often leads to 4.4 Generative Models for Classification\\n159\\npretty decent results, especially in settings where n is not large enough rela-\\ntive to p for us to effectively estimate the joint distribution of the predictors\\nwithin each class. In fact, since estimating a joint distribution requires such\\na huge amount of data, naive Bayes is a good choice in a wide range of set-\\ntings. Essentially, the naive Bayes assumption introduces some bias, but\\nreduces variance, leading to a classifier that works quite well in practice as\\na result of the bias-variance trade-off.\\nOnce we have made the naive Bayes assumption, we can plug (4.29) into\\n(4.15) to obtain an expression for the posterior probability,\\nPr(Y = k|X = x) =\\nπk × fk1(x1) × fk2(x2) × · · · × fkp(xp)\\n)K\\nl=1 πl × fl1(x1) × fl2(x2) × · · · × flp(xp)\\n(4.30)\\nfor k = 1, . . . , K.\\nTo estimate the one-dimensional density function fkj using training data\\nx1j, . . . , xnj, we have a few options.\\n• If Xj is quantitative, then we can assume that Xj|Y = k ∼N(µjk, σ2\\njk).\\nIn other words, we assume that within each class, the jth predictor is\\ndrawn from a (univariate) normal distribution. While this may sound\\na bit like QDA, there is one key difference, in that here we are assum-\\ning that the predictors are independent; this amounts to QDA with\\nan additional assumption that the class-specific covariance matrix is\\ndiagonal.\\n• If Xj is quantitative, then another option is to use a non-parametric\\nestimate for fkj. A very simple way to do this is by making a his-\\ntogram for the observations of the jth predictor within each class.\\nThen we can estimate fkj(xj) as the fraction of the training obser-\\nvations in the kth class that belong to the same histogram bin as\\nxj. Alternatively, we can use a kernel density estimator, which is kernel\\ndensity\\nestimator\\nessentially a smoothed version of a histogram.\\n• If Xj is qualitative, then we can simply count the proportion of train-\\ning observations for the jth predictor corresponding to each class. For\\ninstance, suppose that Xj ∈{1, 2, 3}, and we have 100 observations\\nin the kth class. Suppose that the jth predictor takes on values of 1,\\n2, and 3 in 32, 55, and 13 of those observations, respectively. Then\\nwe can estimate fkj as\\nˆfkj(xj) =\\n\\uf8f1\\n\\uf8f4\\n\\uf8f2\\n\\uf8f4\\n\\uf8f3\\n0.32\\nif xj = 1\\n0.55\\nif xj = 2\\n0.13\\nif xj = 3.\\nWe now consider the naive Bayes classifier in a toy example with p = 3\\npredictors and K = 2 classes. The first two predictors are quantitative,\\nand the third predictor is qualitative with three levels. Suppose further\\nthat ˆπ1 = ˆπ2 = 0.5. The estimated density functions ˆfkj for k = 1, 2\\nand j = 1, 2, 3 are displayed in Figure 4.10. Now suppose that we wish\\nto classify a new observation, x∗= (0.4, 1.5, 1)T . It turns out that in this 160\\n4. Classification\\nDensity estimates for class k=1\\nˆf11\\nˆf12\\nˆf13\\n−4\\n−2\\n0\\n2\\n4\\n−2\\n0\\n2\\n4\\n1\\n2\\n3\\nDensity estimates for class k=2\\nˆf21\\nˆf22\\nˆf23\\n−4\\n−2\\n0\\n2\\n4\\n−2\\n0\\n2\\n4\\n1\\n2\\n3\\nFIGURE 4.10. In the toy example in Section 4.4.4, we generate data with p = 3\\npredictors and K = 2 classes. The first two predictors are quantitative, and the\\nthird predictor is qualitative with three levels. In each class, the estimated density\\nfor each of the three predictors is displayed. If the prior probabilities for the two\\nclasses are equal, then the observation x∗= (0.4, 1.5, 1)T has a 94.4% posterior\\nprobability of belonging to the first class.\\nTrue default status\\nNo\\nYes\\nTotal\\nPredicted\\nNo\\n9621\\n244\\n9865\\ndefault status\\nYes\\n46\\n89\\n135\\nTotal\\n9667\\n333\\n10000\\nTABLE 4.8.\\nComparison of the naive Bayes predictions to the true default\\nstatus for the 10, 000 training observations in the Default data set, when we\\npredict default for any observation for which P(Y = default|X = x) > 0.5.\\nexample, ˆf11(0.4) = 0.368, ˆf12(1.5) = 0.484, ˆf13(1) = 0.226, and ˆf21(0.4) =\\n0.030, ˆf22(1.5) = 0.130, ˆf23(1) = 0.616. Plugging these estimates into (4.30)\\nresults in posterior probability estimates of Pr(Y = 1|X = x∗) = 0.944 and\\nPr(Y = 2|X = x∗) = 0.056.\\nTable 4.8 provides the confusion matrix resulting from applying the naive\\nBayes classifier to the Default data set, where we predict a default if the\\nposterior probability of a default — that is, P(Y = default|X = x) — ex-\\nceeds 0.5. Comparing this to the results for LDA in Table 4.4, our findings\\nare mixed. While LDA has a slightly lower overall error rate, naive Bayes 4.5 A Comparison of Classification Methods\\n161\\nTrue default status\\nNo\\nYes\\nTotal\\nPredicted\\nNo\\n9339\\n130\\n9469\\ndefault status\\nYes\\n328\\n203\\n531\\nTotal\\n9667\\n333\\n10000\\nTABLE 4.9.\\nComparison of the naive Bayes predictions to the true default\\nstatus for the 10, 000 training observations in the Default data set, when we\\npredict default for any observation for which P(Y = default|X = x) > 0.2.\\ncorrectly predicts a higher fraction of the true defaulters. In this implemen-\\ntation of naive Bayes, we have assumed that each quantitative predictor is\\ndrawn from a Gaussian distribution (and, of course, that within each class,\\neach predictor is independent).\\nJust as with LDA, we can easily adjust the probability threshold for\\npredicting a default. For example, Table 4.9 provides the confusion matrix\\nresulting from predicting a default if P(Y = default|X = x) > 0.2. Again,\\nthe results are mixed relative to LDA with the same threshold (Table 4.5).\\nNaive Bayes has a higher error rate, but correctly predicts almost two-thirds\\nof the true defaults.\\nIn this example, it should not be too surprising that naive Bayes does\\nnot convincingly outperform LDA: this data set has n = 10,000 and p = 2,\\nand so the reduction in variance resulting from the naive Bayes assumption\\nis not necessarily worthwhile. We expect to see a greater pay-off to using\\nnaive Bayes relative to LDA or QDA in instances where p is larger or n is\\nsmaller, so that reducing the variance is very important.\\n4.5\\nA Comparison of Classification Methods\\n4.5.1\\nAn Analytical Comparison\\nWe now perform an analytical (or mathematical) comparison of LDA, QDA,\\nnaive Bayes, and logistic regression. We consider these approaches in a\\nsetting with K classes, so that we assign an observation to the class that\\nmaximizes Pr(Y = k|X = x). Equivalently, we can set K as the baseline\\nclass and assign an observation to the class that maximizes\\nlog\\n* Pr(Y = k|X = x)\\nPr(Y = K|X = x)\\n+\\n(4.31)\\nfor k = 1, . . . , K. Examining the specific form of (4.31) for each method\\nprovides a clear understanding of their similarities and differences.\\nFirst, for LDA, we can make use of Bayes’ theorem (4.15) as well as\\nthe assumption that the predictors within each class are drawn from a\\nmultivariate normal density (4.23) with class-specific mean and shared co- 162\\n4. Classification\\nvariance matrix in order to show that\\nlog\\n* Pr(Y = k|X = x)\\nPr(Y = K|X = x)\\n+\\n=\\nlog\\n* πkfk(x)\\nπKfK(x)\\n+\\n=\\nlog\\n>\\nπk exp\\n\\'\\n−1\\n2(x −µk)T Σ−1(x −µk)\\n(\\nπK exp\\n\\'\\n−1\\n2(x −µK)T Σ−1(x −µK)\\n(\\n?\\n=\\nlog\\n* πk\\nπK\\n+\\n−1\\n2(x −µk)T Σ−1(x −µk)\\n+1\\n2(x −µK)T Σ−1(x −µK)\\n=\\nlog\\n* πk\\nπK\\n+\\n−1\\n2(µk + µK)T Σ−1(µk −µK)\\n+ xT Σ−1(µk −µK)\\n=\\nak +\\np\\n0\\nj=1\\nbkjxj,\\n(4.32)\\nwhere ak = log\\n1\\nπk\\nπK\\n2\\n−1\\n2(µk + µK)T Σ−1(µk −µK) and bkj is the jth\\ncomponent of Σ−1(µk −µK). Hence LDA, like logistic regression, assumes\\nthat the log odds of the posterior probabilities is linear in x.\\nUsing similar calculations, in the QDA setting (4.31) becomes\\nlog\\n* Pr(Y = k|X = x)\\nPr(Y = K|X = x)\\n+\\n= ak +\\np\\n0\\nj=1\\nbkjxj +\\np\\n0\\nj=1\\np\\n0\\nl=1\\nckjlxjxl,\\n(4.33)\\nwhere ak, bkj, and ckjl are functions of πk, πK, µk, µK, Σk and ΣK. Again,\\nas the name suggests, QDA assumes that the log odds of the posterior\\nprobabilities is quadratic in x.\\nFinally, we examine (4.31) in the naive Bayes setting. Recall that in\\nthis setting, fk(x) is modeled as a product of p one-dimensional functions\\nfkj(xj) for j = 1, . . . , p. Hence,\\nlog\\n* Pr(Y = k|X = x)\\nPr(Y = K|X = x)\\n+\\n=\\nlog\\n* πkfk(x)\\nπKfK(x)\\n+\\n=\\nlog\\n>\\nπk\\nFp\\nj=1 fkj(xj)\\nπK\\nFp\\nj=1 fKj(xj)\\n?\\n=\\nlog\\n* πk\\nπK\\n+\\n+\\np\\n0\\nj=1\\nlog\\n* fkj(xj)\\nfKj(xj)\\n+\\n=\\nak +\\np\\n0\\nj=1\\ngkj(xj),\\n(4.34)\\nwhere ak = log\\n1\\nπk\\nπK\\n2\\nand gkj(xj) = log\\n1\\nfkj(xj)\\nfKj(xj)\\n2\\n. Hence, the right-hand\\nside of (4.34) takes the form of a generalized additive model, a topic that is\\ndiscussed further in Chapter 7. 4.5 A Comparison of Classification Methods\\n163\\nInspection of (4.32), (4.33), and (4.34) yields the following observations\\nabout LDA, QDA, and naive Bayes:\\n• LDA is a special case of QDA with ckjl = 0 for all j = 1, . . . , p,\\nl = 1, . . . , p, and k = 1, . . . , K. (Of course, this is not surprising, since\\nLDA is simply a restricted version of QDA with Σ1 = · · · = ΣK = Σ.)\\n• Any classifier with a linear decision boundary is a special case of naive\\nBayes with gkj(xj) = bkjxj. In particular, this means that LDA is\\na special case of naive Bayes! This is not at all obvious from the\\ndescriptions of LDA and naive Bayes earlier in this chapter, since\\neach method makes very different assumptions: LDA assumes that\\nthe features are normally distributed with a common within-class\\ncovariance matrix, and naive Bayes instead assumes independence of\\nthe features.\\n• If we model fkj(xj) in the naive Bayes classifier using a one-dimensio-\\nnal Gaussian distribution N(µkj, σ2\\nj ), then we end up with gkj(xj) =\\nbkjxj where bkj = (µkj−µKj)/σ2\\nj . In this case, naive Bayes is actually\\na special case of LDA with Σ restricted to be a diagonal matrix with\\njth diagonal element equal to σ2\\nj .\\n• Neither QDA nor naive Bayes is a special case of the other. Naive\\nBayes can produce a more flexible fit, since any choice can be made\\nfor gkj(xj). However, it is restricted to a purely additive fit, in the\\nsense that in (4.34), a function of xj is added to a function of xl, for\\nj ̸= l; however, these terms are never multiplied. By contrast, QDA\\nincludes multiplicative terms of the form ckjlxjxl. Therefore, QDA\\nhas the potential to be more accurate in settings where interactions\\namong the predictors are important in discriminating between classes.\\nNone of these methods uniformly dominates the others: in any setting, the\\nchoice of method will depend on the true distribution of the predictors in\\neach of the K classes, as well as other considerations, such as the values of\\nn and p. The latter ties into the bias-variance trade-off.\\nHow does logistic regression tie into this story? Recall from (4.12) that\\nmultinomial logistic regression takes the form\\nlog\\n* Pr(Y = k|X = x)\\nPr(Y = K|X = x)\\n+\\n= βk0 +\\np\\n0\\nj=1\\nβkjxj.\\nThis is identical to the linear form of LDA (4.32): in both cases,\\nlog\\n1\\nPr(Y =k|X=x)\\nPr(Y =K|X=x)\\n2\\nis a linear function of the predictors. In LDA, the co-\\nefficients in this linear function are functions of estimates for πk, πK, µk,\\nµK, and Σ obtained by assuming that X1, . . . , Xp follow a normal distri-\\nbution within each class. By contrast, in logistic regression, the coefficients\\nare chosen to maximize the likelihood function (4.5). Thus, we expect LDA\\nto outperform logistic regression when the normality assumption (approxi-\\nmately) holds, and we expect logistic regression to perform better when it\\ndoes not. 164\\n4. Classification\\nWe close with a brief discussion of K-nearest neighbors (KNN), intro-\\nduced in Chapter 2. Recall that KNN takes a completely different approach\\nfrom the classifiers seen in this chapter. In order to make a prediction for\\nan observation X = x, the training observations that are closest to x are\\nidentified. Then X is assigned to the class to which the plurality of these\\nobservations belong. Hence KNN is a completely non-parametric approach:\\nno assumptions are made about the shape of the decision boundary. We\\nmake the following observations about KNN:\\n• Because KNN is completely non-parametric, we can expect this ap-\\nproach to dominate LDA and logistic regression when the decision\\nboundary is highly non-linear, provided that n is very large and p is\\nsmall.\\n• In order to provide accurate classification, KNN requires a lot of ob-\\nservations relative to the number of predictors—that is, n much larger\\nthan p. This has to do with the fact that KNN is non-parametric, and\\nthus tends to reduce the bias while incurring a lot of variance.\\n• In settings where the decision boundary is non-linear but n is only\\nmodest, or p is not very small, then QDA may be preferred to KNN.\\nThis is because QDA can provide a non-linear decision boundary\\nwhile taking advantage of a parametric form, which means that it\\nrequires a smaller sample size for accurate classification, relative to\\nKNN.\\n• Unlike logistic regression, KNN does not tell us which predictors are\\nimportant: we don’t get a table of coefficients as in Table 4.3.\\n4.5.2\\nAn Empirical Comparison\\nWe now compare the empirical (practical) performance of logistic regres-\\nsion, LDA, QDA, naive Bayes, and KNN. We generated data from six dif-\\nferent scenarios, each of which involves a binary (two-class) classification\\nproblem. In three of the scenarios, the Bayes decision boundary is linear,\\nand in the remaining scenarios it is non-linear. For each scenario, we pro-\\nduced 100 random training data sets. On each of these training sets, we\\nfit each method to the data and computed the resulting test error rate on\\na large test set. Results for the linear scenarios are shown in Figure 4.11,\\nand the results for the non-linear scenarios are in Figure 4.12. The KNN\\nmethod requires selection of K, the number of neighbors (not to be con-\\nfused with the number of classes in earlier sections of this chapter). We\\nperformed KNN with two values of K: K = 1, and a value of K that was\\nchosen automatically using an approach called cross-validation, which we\\ndiscuss further in Chapter 5. We applied naive Bayes assuming univariate\\nGaussian densities for the features within each class (and, of course — since\\nthis is the key characteristic of naive Bayes — assuming independence of\\nthe features).\\nIn each of the six scenarios, there were p = 2 quantitative predictors.\\nThe scenarios were as follows: 4.5 A Comparison of Classification Methods\\n165\\nKNN−1\\nKNN−CV\\nLDA\\nLogistic\\nNBayes\\nQDA\\nSCENARIO 1\\n0.25\\n0.30\\n0.35\\n0.40\\n0.45\\nKNN−1\\nKNN−CV\\nLDA\\nLogistic\\nNBayes\\nQDA\\nSCENARIO 2\\n0.15\\n0.20\\n0.25\\n0.30\\nKNN−1\\nKNN−CV\\nLDA\\nLogistic\\nNBayes\\nQDA\\nSCENARIO 3\\n0.20\\n0.25\\n0.30\\n0.35\\n0.40\\n0.45\\n0.50\\nFIGURE 4.11. Boxplots of the test error rates for each of the linear scenarios\\ndescribed in the main text.\\nScenario 1: There were 20 training observations in each of two classes. The\\nobservations within each class were uncorrelated random normal variables\\nwith a different mean in each class. The left-hand panel of Figure 4.11 shows\\nthat LDA performed well in this setting, as one would expect since this is\\nthe model assumed by LDA. Logistic regression also performed quite well,\\nsince it assumes a linear decision boundary. KNN performed poorly because\\nit paid a price in terms of variance that was not offset by a reduction in bias.\\nQDA also performed worse than LDA, since it fit a more flexible classifier\\nthan necessary. The performance of naive Bayes was slightly better than\\nQDA, because the naive Bayes assumption of independent predictors is\\ncorrect.\\nScenario 2: Details are as in Scenario 1, except that within each class, the\\ntwo predictors had a correlation of −0.5. The center panel of Figure 4.11\\nindicates that the performance of most methods is similar to the previ-\\nous scenario. The notable exception is naive Bayes, which performs very\\npoorly here, since the naive Bayes assumption of independent predictors is\\nviolated.\\nScenario 3: As in the previous scenario, there is substantial negative cor-\\nrelation between the predictors within each class. However, this time we\\ngenerated X1 and X2 from the t-distribution, with 50 observations per class. t-\\ndistribution\\nThe t-distribution has a similar shape to the normal distribution, but it\\nhas a tendency to yield more extreme points—that is, more points that are\\nfar from the mean. In this setting, the decision boundary was still linear,\\nand so fit into the logistic regression framework. The set-up violated the\\nassumptions of LDA, since the observations were not drawn from a normal\\ndistribution. The right-hand panel of Figure 4.11 shows that logistic regres-\\nsion outperformed LDA, though both methods were superior to the other\\napproaches. In particular, the QDA results deteriorated considerably as a\\nconsequence of non-normality. Naive Bayes performed very poorly because\\nthe independence assumption is violated.\\nScenario 4: The data were generated from a normal distribution, with a\\ncorrelation of 0.5 between the predictors in the first class, and correlation of\\n−0.5 between the predictors in the second class. This setup corresponded to\\nthe QDA assumption, and resulted in quadratic decision boundaries. The\\nleft-hand panel of Figure 4.12 shows that QDA outperformed all of the 166\\n4. Classification\\nKNN−1\\nKNN−CV\\nLDA\\nLogistic\\nNBayes\\nQDA\\nSCENARIO 4\\n0.30\\n0.35\\n0.40\\nKNN−1\\nKNN−CV\\nLDA\\nLogistic\\nNBayes\\nQDA\\nSCENARIO 5\\n0.18\\n0.20\\n0.22\\n0.24\\n0.26\\n0.28\\n0.30\\n0.32\\nKNN−1\\nKNN−CV\\nLDA\\nLogistic\\nNBayes\\nQDA\\nSCENARIO 6\\n0.15\\n0.20\\n0.25\\n0.30\\n0.35\\n0.40\\n0.45\\nFIGURE 4.12.\\nBoxplots of the test error rates for each of the non-linear\\nscenarios described in the main text.\\nother approaches. The naive Bayes assumption of independent predictors\\nis violated, so naive Bayes performs poorly.\\nScenario 5: The data were generated from a normal distribution with un-\\ncorrelated predictors. Then the responses were sampled from the logistic\\nfunction applied to a complicated non-linear function of the predictors. The\\ncenter panel of Figure 4.12 shows that both QDA and naive Bayes gave\\nslightly better results than the linear methods, while the much more flexi-\\nble KNN-CV method gave the best results. But KNN with K = 1 gave the\\nworst results out of all methods. This highlights the fact that even when the\\ndata exhibits a complex non-linear relationship, a non-parametric method\\nsuch as KNN can still give poor results if the level of smoothness is not\\nchosen correctly.\\nScenario 6: The observations were generated from a normal distribution\\nwith a different diagonal covariance matrix for each class. However, the\\nsample size was very small: just n = 6 in each class. Naive Bayes performed\\nvery well, because its assumptions are met. LDA and logistic regression\\nperformed poorly because the true decision boundary is non-linear, due to\\nthe unequal covariance matrices. QDA performed a bit worse than naive\\nBayes, because given the very small sample size, the former incurred too\\nmuch variance in estimating the correlation between the predictors within\\neach class. KNN’s performance also suffered due to the very small sample\\nsize.\\nThese six examples illustrate that no one method will dominate the oth-\\ners in every situation. When the true decision boundaries are linear, then\\nthe LDA and logistic regression approaches will tend to perform well. When\\nthe boundaries are moderately non-linear, QDA or naive Bayes may give\\nbetter results. Finally, for much more complicated decision boundaries, a\\nnon-parametric approach such as KNN can be superior. But the level of\\nsmoothness for a non-parametric approach must be chosen carefully. In the\\nnext chapter we examine a number of approaches for choosing the correct\\nlevel of smoothness and, in general, for selecting the best overall method.\\nFinally, recall from Chapter 3 that in the regression setting we can accom-\\nmodate a non-linear relationship between the predictors and the response\\nby performing regression using transformations of the predictors. A similar\\napproach could be taken in the classification setting. For instance, we could 4.6 Generalized Linear Models\\n167\\nCoefficient Std. error t-statistic p-value\\nIntercept\\n73.60\\n5.13\\n14.34\\n0.00\\nworkingday\\n1.27\\n1.78\\n0.71\\n0.48\\ntemp\\n157.21\\n10.26\\n15.32\\n0.00\\nweathersit[cloudy/misty]\\n-12.89\\n1.96\\n-6.56\\n0.00\\nweathersit[light rain/snow]\\n-66.49\\n2.97\\n-22.43\\n0.00\\nweathersit[heavy rain/snow]\\n-109.75\\n76.67\\n-1.43\\n0.15\\nTABLE 4.10. Results for a least squares linear model fit to predict bikers in\\nthe Bikeshare data. The predictors mnth and hr are omitted from this table due\\nto space constraints, and can be seen in Figure 4.13. For the qualitative variable\\nweathersit, the baseline level corresponds to clear skies.\\ncreate a more flexible version of logistic regression by including X2, X3,\\nand even X4 as predictors. This may or may not improve logistic regres-\\nsion’s performance, depending on whether the increase in variance due to\\nthe added flexibility is offset by a sufficiently large reduction in bias. We\\ncould do the same for LDA. If we added all possible quadratic terms and\\ncross-products to LDA, the form of the model would be the same as the\\nQDA model, although the parameter estimates would be different. This\\ndevice allows us to move somewhere between an LDA and a QDA model.\\n4.6\\nGeneralized Linear Models\\nIn Chapter 3, we assumed that the response Y is quantitative, and ex-\\nplored the use of least squares linear regression to predict Y . Thus far in\\nthis chapter, we have instead assumed that Y is qualitative. However, we\\nmay sometimes be faced with situations in which Y is neither qualitative\\nnor quantitative, and so neither linear regression from Chapter 3 nor the\\nclassification approaches covered in this chapter is applicable.\\nAs a concrete example, we consider the Bikeshare data set. The response\\nis bikers, the number of hourly users of a bike sharing program in Wash-\\nington, DC. This response value is neither qualitative nor quantitative:\\ninstead, it takes on non-negative integer values, or counts. We will consider counts\\npredicting bikers using the covariates mnth (month of the year), hr (hour\\nof the day, from 0 to 23), workingday (an indicator variable that equals 1 if\\nit is neither a weekend nor a holiday), temp (the normalized temperature,\\nin Celsius), and weathersit (a qualitative variable that takes on one of four\\npossible values: clear; misty or cloudy; light rain or light snow; or heavy\\nrain or heavy snow.)\\nIn the analyses that follow, we will treat mnth, hr, and weathersit as\\nqualitative variables.\\n4.6.1\\nLinear Regression on the Bikeshare Data\\nTo begin, we consider predicting bikers using linear regression. The results\\nare shown in Table 4.10.\\nWe see, for example, that a progression of weather from clear to cloudy\\nresults in, on average, 12.89 fewer bikers per hour; however, if the weather\\nprogresses further to rain or snow, then this further results in 53.60 fewer\\nbikers per hour. Figure 4.13 displays the coefficients associated with mnth 168\\n4. Classification\\nFIGURE 4.13. A least squares linear regression model was fit to predict bikers\\nin the Bikeshare data set. Left: The coefficients associated with the month of the\\nyear. Bike usage is highest in the spring and fall, and lowest in the winter. Right:\\nThe coefficients associated with the hour of the day. Bike usage is highest during\\npeak commute times, and lowest overnight.\\nand the coefficients associated with hr. We see that bike usage is highest in\\nthe spring and fall, and lowest during the winter months. Furthermore, bike\\nusage is greatest around rush hour (9 AM and 6 PM), and lowest overnight.\\nThus, at first glance, fitting a linear regression model to the Bikeshare data\\nset seems to provide reasonable and intuitive results.\\nBut upon more careful inspection, some issues become apparent. For\\nexample, 9.6% of the fitted values in the Bikeshare data set are negative:\\nthat is, the linear regression model predicts a negative number of users\\nduring 9.6% of the hours in the data set. This calls into question our ability\\nto perform meaningful predictions on the data, and it also raises concerns\\nabout the accuracy of the coefficient estimates, confidence intervals, and\\nother outputs of the regression model.\\nFurthermore, it is reasonable to suspect that when the expected value\\nof bikers is small, the variance of bikers should be small as well. For\\ninstance, at 2 AM during a heavy December snow storm, we expect that\\nextremely few people will use a bike, and moreover that there should be\\nlittle variance associated with the number of users during those conditions.\\nThis is borne out in the data: between 1 AM and 4 AM, in December,\\nJanuary, and February, when it is raining, there are 5.05 users, on average,\\nwith a standard deviation of 3.73. By contrast, between 7 AM and 10 AM,\\nin April, May, and June, when skies are clear, there are 243.59 users, on\\naverage, with a standard deviation of 131.7. The mean-variance relationship\\nis displayed in the left-hand panel of Figure 4.14. This is a major violation\\nof the assumptions of a linear model, which state that Y = )p\\nj=1 Xjβj +ϵ,\\nwhere ϵ is a mean-zero error term with variance σ2 that is constant, and\\nnot a function of the covariates. Therefore, the heteroscedasticity of the\\ndata calls into question the suitability of a linear regression model.\\nFinally, the response bikers is integer-valued. But under a linear model,\\nY = β0 + )p\\nj=1 Xjβj + ϵ, where ϵ is a continuous-valued error term. This\\nmeans that in a linear model, the response Y is necessarily continuous-\\nvalued (quantitative). Thus, the integer nature of the response bikers sug-\\ngests that a linear regression model is not entirely satisfactory for this data\\nset.\\n●\\n●\\n●\\n●\\n●\\n●\\n●\\n●\\n●\\n●\\n●\\n●\\n−40\\n−20\\n0\\n20\\nMonth\\nCoefficient\\nJ\\nF\\nM\\nA\\nM\\nJ\\nJ\\nA\\nS\\nO\\nN\\nD\\n●\\n●●●●\\n●\\n●\\n●\\n●\\n●\\n●\\n●\\n●●●●\\n●\\n●\\n●\\n●\\n●\\n●\\n●\\n●\\n5\\n10\\n15\\n20\\n−100\\n0\\n50\\n100\\n200\\nHour\\nCoefficient 4.6 Generalized Linear Models\\n169\\n5\\n10\\n15\\n20\\n0\\n100\\n200\\n300\\n400\\n500\\n600\\nHour\\nNumber of Bikers\\n5\\n10\\n15\\n20\\n0\\n1\\n2\\n3\\n4\\n5\\n6\\nHour\\nLog(Number of Bikers)\\nFIGURE 4.14. Left: On the Bikeshare dataset, the number of bikers is dis-\\nplayed on the y-axis, and the hour of the day is displayed on the x-axis. Jitter\\nwas applied for ease of visualization. For the most part, as the mean number of\\nbikers increases, so does the variance in the number of bikers. A smoothing spline\\nfit is shown in green. Right: The log of the number of bikers is now displayed on\\nthe y-axis.\\nSome of the problems that arise when fitting a linear regression model\\nto the Bikeshare data can be overcome by transforming the response; for\\ninstance, we can fit the model\\nlog(Y ) =\\np\\n0\\nj=1\\nXjβj + ϵ.\\nTransforming the response avoids the possibility of negative predictions,\\nand it overcomes much of the heteroscedasticity in the untransformed data,\\nas is shown in the right-hand panel of Figure 4.14. However, it is not quite\\na satisfactory solution, since predictions and inference are made in terms of\\nthe log of the response, rather than the response. This leads to challenges\\nin interpretation, e.g. “a one-unit increase in Xj is associated with an\\nincrease in the mean of the log of Y by an amount βj”. Furthermore, a\\nlog transformation of the response cannot be applied in settings where the\\nresponse can take on a value of 0. Thus, while fitting a linear model to\\na transformation of the response may be an adequate approach for some\\ncount-valued data sets, it often leaves something to be desired. We will see\\nin the next section that a Poisson regression model provides a much more\\nnatural and elegant approach for this task.\\n4.6.2\\nPoisson Regression on the Bikeshare Data\\nTo overcome the inadequacies of linear regression for analyzing the Bikeshare\\ndata set, we will make use of an alternative approach, called Poisson\\nregression. Before we can talk about Poisson regression, we must first in- Poisson\\nregression\\ntroduce the Poisson distribution.\\nPoisson\\ndistribution\\nSuppose that a random variable Y takes on nonnegative integer values,\\ni.e. Y ∈{0, 1, 2, . . .}. If Y follows the Poisson distribution, then\\nPr(Y = k) = e−λλk\\nk!\\nfor k = 0, 1, 2, . . . .\\n(4.35) 170\\n4. Classification\\nHere, λ > 0 is the expected value of Y , i.e. E(Y ). It turns out that λ also\\nequals the variance of Y , i.e. λ = E(Y ) = Var(Y ). This means that if Y\\nfollows the Poisson distribution, then the larger the mean of Y , the larger\\nits variance. (In (4.35), the notation k!, pronounced “k factorial”, is defined\\nas k! = k × (k −1) × (k −2) × . . . × 3 × 2 × 1.)\\nThe Poisson distribution is typically used to model counts; this is a nat-\\nural choice for a number of reasons, including the fact that counts, like\\nthe Poisson distribution, take on nonnegative integer values. To see how\\nwe might use the Poisson distribution in practice, let Y denote the num-\\nber of users of the bike sharing program during a particular hour of the\\nday, under a particular set of weather conditions, and during a particu-\\nlar month of the year. We might model Y as a Poisson distribution with\\nmean E(Y ) = λ = 5. This means that the probability of no users dur-\\ning this particular hour is Pr(Y = 0) =\\ne−550\\n0!\\n= e−5 = 0.0067 (where\\n0! = 1 by convention). The probability that there is exactly one user\\nis Pr(Y = 1) =\\ne−551\\n1!\\n= 5e−5 = 0.034, the probability of two users is\\nPr(Y = 2) = e−552\\n2!\\n= 0.084, and so on.\\nOf course, in reality, we expect the mean number of users of the bike\\nsharing program, λ = E(Y ), to vary as a function of the hour of the day,\\nthe month of the year, the weather conditions, and so forth. So rather\\nthan modeling the number of bikers, Y , as a Poisson distribution with a\\nfixed mean value like λ = 5, we would like to allow the mean to vary as a\\nfunction of the covariates. In particular, we consider the following model\\nfor the mean λ = E(Y ), which we now write as λ(X1, . . . , Xp) to emphasize\\nthat it is a function of the covariates X1, . . . , Xp:\\nlog(λ(X1, . . . , Xp)) = β0 + β1X1 + · · · + βpXp\\n(4.36)\\nor equivalently\\nλ(X1, . . . , Xp) = eβ0+β1X1+···+βpXp.\\n(4.37)\\nHere, β0, β1, . . . , βp are parameters to be estimated. Together, (4.35) and\\n(4.36) define the Poisson regression model. Notice that in (4.36), we take\\nthe log of λ(X1, . . . , Xp) to be linear in X1, . . . , Xp, rather than having\\nλ(X1, . . . , Xp) itself be linear in X1, . . . , Xp; this ensures that λ(X1, . . . , Xp)\\ntakes on nonnegative values for all values of the covariates.\\nTo estimate the coefficients β0, β1, . . . , βp, we use the same maximum\\nlikelihood approach that we adopted for logistic regression in Section 4.3.2.\\nSpecifically, given n independent observations from the Poisson regression\\nmodel, the likelihood takes the form\\nℓ(β0, β1, . . . , βp) =\\nn\\nE\\ni=1\\ne−λ(xi)λ(xi)yi\\nyi!\\n,\\n(4.38)\\nwhere λ(xi) = eβ0+β1xi1+···+βpxip, due to (4.37). We estimate the coef-\\nficients that maximize the likelihood ℓ(β0, β1, . . . , βp), i.e. that make the\\nobserved data as likely as possible.\\nWe now fit a Poisson regression model to the Bikeshare data set. The\\nresults are shown in Table 4.11 and Figure 4.15. Qualitatively, the results\\nare similar to those from linear regression in Section 4.6.1. We again see\\nthat bike usage is highest in the spring and fall and during rush hour, 4.6 Generalized Linear Models\\n171\\nCoefficient Std. error z-statistic p-value\\nIntercept\\n4.12\\n0.01\\n683.96\\n0.00\\nworkingday\\n0.01\\n0.00\\n7.5\\n0.00\\ntemp\\n0.79\\n0.01\\n68.43\\n0.00\\nweathersit[cloudy/misty]\\n-0.08\\n0.00\\n-34.53\\n0.00\\nweathersit[light rain/snow]\\n-0.58\\n0.00\\n-141.91\\n0.00\\nweathersit[heavy rain/snow]\\n-0.93\\n0.17\\n-5.55\\n0.00\\nTABLE 4.11. Results for a Poisson regression model fit to predict bikers in\\nthe Bikeshare data. The predictors mnth and hr are omitted from this table due\\nto space constraints, and can be seen in Figure 4.15. For the qualitative variable\\nweathersit, the baseline corresponds to clear skies.\\nFIGURE 4.15.\\nA Poisson regression model was fit to predict bikers in the\\nBikeshare data set. Left: The coefficients associated with the month of the year.\\nBike usage is highest in the spring and fall, and lowest in the winter. Right: The\\ncoefficients associated with the hour of the day. Bike usage is highest during peak\\ncommute times, and lowest overnight.\\nand lowest during the winter and in the early morning hours. Moreover,\\nbike usage increases as the temperature increases, and decreases as the\\nweather worsens. Interestingly, the coefficient associated with workingday\\nis statistically significant under the Poisson regression model, but not under\\nthe linear regression model.\\nSome important distinctions between the Poisson regression model and\\nthe linear regression model are as follows:\\n• Interpretation: To interpret the coefficients in the Poisson regression\\nmodel, we must pay close attention to (4.37), which states that an\\nincrease in Xj by one unit is associated with a change in E(Y ) = λ\\nby a factor of exp(βj). For example, a change in weather from clear\\nto cloudy skies is associated with a change in mean bike usage by a\\nfactor of exp(−0.08) = 0.923, i.e. on average, only 92.3% as many\\npeople will use bikes when it is cloudy relative to when it is clear.\\nIf the weather worsens further and it begins to rain, then the mean\\nbike usage will further change by a factor of exp(−0.5) = 0.607, i.e.\\non average only 60.7% as many people will use bikes when it is rainy\\nrelative to when it is cloudy.\\n●\\n●\\n●\\n●\\n●\\n●\\n●\\n●\\n●\\n●\\n●\\n●\\n−0.6\\n−0.4\\n−0.2\\n0.0\\n0.2\\nMonth\\nCoefficient\\nJ\\nF\\nM\\nA\\nM\\nJ\\nJ\\nA\\nS\\nO\\nN\\nD\\n●\\n●\\n●\\n●\\n●\\n●\\n●\\n●\\n●\\n●\\n●\\n●\\n●●●●\\n●\\n●●\\n●\\n●\\n●\\n●\\n●\\n5\\n10\\n15\\n20\\n−2\\n−1\\n0\\n1\\nHour\\nCoefficient 172\\n4. Classification\\n• Mean-variance relationship: As mentioned earlier, under the Poisson\\nmodel, λ = E(Y ) = Var(Y ). Thus, by modeling bike usage with a\\nPoisson regression, we implicitly assume that mean bike usage in a\\ngiven hour equals the variance of bike usage during that hour. By\\ncontrast, under a linear regression model, the variance of bike usage\\nalways takes on a constant value. Recall from Figure 4.14 that in the\\nBikeshare data, when biking conditions are favorable, both the mean\\nand the variance in bike usage are much higher than when conditions\\nare unfavorable. Thus, the Poisson regression model is able to handle\\nthe mean-variance relationship seen in the Bikeshare data in a way\\nthat the linear regression model is not.5\\noverdispersion\\n• nonnegative fitted values: There are no negative predictions using the\\nPoisson regression model. This is because the Poisson model itself\\nonly allows for nonnegative values; see (4.35). By contrast, when we\\nfit a linear regression model to the Bikeshare data set, almost 10% of\\nthe predictions were negative.\\n4.6.3\\nGeneralized Linear Models in Greater Generality\\nWe have now discussed three types of regression models: linear, logistic and\\nPoisson. These approaches share some common characteristics:\\n1. Each approach uses predictors X1, . . . , Xp to predict a response Y .\\nWe assume that, conditional on X1, . . . , Xp, Y belongs to a certain\\nfamily of distributions. For linear regression, we typically assume that\\nY follows a Gaussian or normal distribution. For logistic regression,\\nwe assume that Y follows a Bernoulli distribution. Finally, for Poisson\\nregression, we assume that Y follows a Poisson distribution.\\n2. Each approach models the mean of Y as a function of the predictors.\\nIn linear regression, the mean of Y takes the form\\nE(Y |X1, . . . , Xp) = β0 + β1X1 + · · · + βpXp,\\n(4.39)\\ni.e. it is a linear function of the predictors. For logistic regression, the\\nmean instead takes the form\\nE(Y |X1, . . . , Xp)\\n=\\nPr(Y = 1|X1, . . . , Xp)\\n=\\neβ0+β1X1+···+βpXp\\n1 + eβ0+β1X1+···+βpXp ,\\n(4.40)\\nwhile for Poisson regression it takes the form\\nE(Y |X1, . . . , Xp) = λ(X1, . . . , Xp) = eβ0+β1X1+···+βpXp.\\n(4.41)\\nEquations (4.39)–(4.41) can be expressed using a link function, η, which link function\\n5In fact, the variance in the Bikeshare data appears to be much higher than the\\nmean, a situation referred to as overdispersion. This causes the Z-values to be inflated\\nin Table 4.11. A more careful analysis should account for this overdispersion to obtain\\nmore accurate Z-values, and there are a variety of methods for doing this. But they are\\nbeyond the scope of this book. 4.7 Lab: Logistic Regression, LDA, QDA, and KNN\\n173\\napplies a transformation to E(Y |X1, . . . , Xp) so that the transformed mean\\nis a linear function of the predictors. That is,\\nη(E(Y |X1, . . . , Xp)) = β0 + β1X1 + · · · + βpXp.\\n(4.42)\\nThe link functions for linear, logistic and Poisson regression are η(µ) = µ,\\nη(µ) = log(µ/(1 −µ)), and η(µ) = log(µ), respectively.\\nThe Gaussian, Bernoulli and Poisson distributions are all members of a\\nwider class of distributions, known as the exponential family. Other well- exponential\\nfamily\\nknown members of this family are the exponential distribution, the Gamma\\nexponential\\nGamma\\ndistribution, and the negative binomial distribution. In general, we can per-\\nnegative\\nbinomial\\nform a regression by modeling the response Y as coming from a particular\\nmember of the exponential family, and then transforming the mean of the\\nresponse so that the transformed mean is a linear function of the predictors\\nvia (4.42). Any regression approach that follows this very general recipe is\\nknown as a generalized linear model (GLM). Thus, linear regression, logistic generalized\\nlinear model\\nregression, and Poisson regression are three examples of GLMs. Other ex-\\namples not covered here include Gamma regression and negative binomial\\nregression.\\n4.7\\nLab: Logistic Regression, LDA, QDA, and\\nKNN\\n4.7.1\\nThe Stock Market Data\\nIn this lab we will examine the Smarket data, which is part of the ISLP\\nlibrary. This data set consists of percentage returns for the S&P 500 stock\\nindex over 1,250 days, from the beginning of 2001 until the end of 2005.\\nFor each date, we have recorded the percentage returns for each of the five\\nprevious trading days, Lag1 through Lag5. We have also recorded Volume\\n(the number of shares traded on the previous day, in billions), Today (the\\npercentage return on the date in question) and Direction (whether the\\nmarket was Up or Down on this date).\\nWe start by importing our libraries at this top level; these are all imports\\nwe have seen in previous labs.\\nIn [1]: import numpy as np\\nimport pandas as pd\\nfrom matplotlib.pyplot import subplots\\nimport statsmodels.api as sm\\nfrom ISLP import load_data\\nfrom ISLP.models import (ModelSpec as MS,\\nsummarize)\\nWe also collect together the new imports needed for this lab.\\nIn [2]: from ISLP import confusion_table\\nfrom ISLP.models import contrast\\nfrom sklearn.discriminant_analysis import \\\\\\n(LinearDiscriminantAnalysis as LDA,\\nQuadraticDiscriminantAnalysis as QDA)\\nfrom sklearn.naive_bayes import GaussianNB\\nfrom sklearn.neighbors import KNeighborsClassifier\\nfrom sklearn.preprocessing import StandardScaler 174\\n4. Classification\\nfrom sklearn.model_selection import train_test_split\\nfrom sklearn.linear_model import LogisticRegression\\nNow we are ready to load the Smarket data.\\nIn [3]: Smarket = load_data(\\'Smarket\\')\\nSmarket\\nThis gives a truncated listing of the data, which we do not show here. We\\ncan see what the variable names are.\\nIn [4]: Smarket.columns\\nOut[4]: Index([\\'Year\\', \\'Lag1\\', \\'Lag2\\', \\'Lag3\\', \\'Lag4\\', \\'Lag5\\', \\'Volume\\',\\n\\'Today\\', \\'Direction\\'],\\ndtype=\\'object\\')\\nWe compute the correlation matrix using the corr() method for data .corr()\\nframes, which produces a matrix that contains all of the pairwise corre-\\nlations among the variables. (We suppress the output here.) The pandas\\nlibrary does not report a correlation for the Direction variable because it\\nis qualitative.\\nIn [5]: Smarket.corr()\\nAs one would expect, the correlations between the lagged return variables\\nand today’s return are close to zero. The only substantial correlation is\\nbetween Year and Volume. By plotting the data we see that Volume is in-\\ncreasing over time. In other words, the average number of shares traded\\ndaily increased from 2001 to 2005.\\nIn [6]: Smarket.plot(y=\\'Volume\\');\\n4.7.2\\nLogistic Regression\\nNext, we will fit a logistic regression model in order to predict Direction\\nusing Lag1 through Lag5 and Volume. The sm.GLM() function fits gener- sm.GLM()\\nalized linear models, a class of models that includes logistic regression. generalized\\nlinear model\\nAlternatively, the function sm.Logit() fits a logistic regression model di-\\nrectly. The syntax of sm.GLM() is similar to that of sm.OLS(), except that\\nwe must pass in the argument family=sm.families.Binomial() in order to\\ntell statsmodels to run a logistic regression rather than some other type of\\ngeneralized linear model.\\nIn [7]: allvars = Smarket.columns.drop([\\'Today\\', \\'Direction\\', \\'Year\\'])\\ndesign = MS(allvars)\\nX = design.fit_transform(Smarket)\\ny = Smarket.Direction == \\'Up\\'\\nglm = sm.GLM(y,\\nX,\\nfamily=sm.families.Binomial())\\nresults = glm.fit()\\nsummarize(results) 4.7 Lab: Logistic Regression, LDA, QDA, and KNN\\n175\\nOut[7]:\\ncoef\\nstd err\\nz\\nP>|z|\\nintercept -0.1260\\n0.241\\n-0.523\\n0.601\\nLag1\\n-0.0731\\n0.050\\n-1.457\\n0.145\\nLag2\\n-0.0423\\n0.050\\n-0.845\\n0.398\\nLag3\\n0.0111\\n0.050\\n0.222\\n0.824\\nLag4\\n0.0094\\n0.050\\n0.187\\n0.851\\nLag5\\n0.0103\\n0.050\\n0.208\\n0.835\\nVolume\\n0.1354\\n0.158\\n0.855\\n0.392\\nThe smallest p-value here is associated with Lag1. The negative coefficient\\nfor this predictor suggests that if the market had a positive return yesterday,\\nthen it is less likely to go up today. However, at a value of 0.15, the p-value\\nis still relatively large, and so there is no clear evidence of a real association\\nbetween Lag1 and Direction.\\nWe use the params attribute of results in order to access just the coeffi-\\ncients for this fitted model.\\nIn [8]: results.params\\nOut[8]: intercept\\n-0.126000\\nLag1\\n-0.073074\\nLag2\\n-0.042301\\nLag3\\n0.011085\\nLag4\\n0.009359\\nLag5\\n0.010313\\nVolume\\n0.135441\\ndtype: float64\\nLikewise we can use the pvalues attribute to access the p-values for the\\ncoefficients (not shown).\\nIn [9]: results.pvalues\\nThe predict() method of results can be used to predict the probability\\nthat the market will go up, given values of the predictors. This method\\nreturns predictions on the probability scale. If no data set is supplied to\\nthe predict() function, then the probabilities are computed for the train-\\ning data that was used to fit the logistic regression model. As with linear\\nregression, one can pass an optional exog argument consistent with a design\\nmatrix if desired. Here we have printed only the first ten probabilities.\\nIn [10]: probs = results.predict()\\nprobs[:10]\\nOut[10]: array([0.5070841, 0.4814679, 0.4811388, 0.5152223, 0.5107812,\\n0.5069565, 0.4926509, 0.5092292, 0.5176135, 0.4888378])\\nIn order to make a prediction as to whether the market will go up or\\ndown on a particular day, we must convert these predicted probabilities\\ninto class labels, Up or Down. The following two commands create a vector\\nof class predictions based on whether the predicted probability of a market\\nincrease is greater than or less than 0.5.\\nIn [11]: labels = np.array([\\'Down\\']*1250)\\nlabels[probs >0.5] = \"Up\" 176\\n4. Classification\\nThe confusion_table() function from the ISLP package summarizes these confusion_\\ntable()\\npredictions, showing how many observations were correctly or incorrectly\\nclassified. Our function, which is adapted from a similar function in the\\nmodule sklearn.metrics, transposes the resulting matrix and includes row\\nand column labels. The confusion_table() function takes as first argument\\nthe predicted labels, and second argument the true labels.\\nIn [12]: confusion_table(labels, Smarket.Direction)\\nOut[12]:\\nTruth\\nDown\\nUp\\nPredicted\\nDown\\n145\\n141\\nUp\\n457\\n507\\nThe diagonal elements of the confusion matrix indicate correct predictions,\\nwhile the off-diagonals represent incorrect predictions. Hence our model\\ncorrectly predicted that the market would go up on 507 days and that\\nit would go down on 145 days, for a total of 507 + 145 = 652 correct\\npredictions. The np.mean() function can be used to compute the fraction of\\ndays for which the prediction was correct. In this case, logistic regression\\ncorrectly predicted the movement of the market 52.2% of the time.\\nIn [13]: (507+145)/1250, np.mean(labels == Smarket.Direction)\\nOut[13]: (0.5216, 0.5216)\\nAt first glance, it appears that the logistic regression model is working\\na little better than random guessing. However, this result is misleading\\nbecause we trained and tested the model on the same set of 1,250 obser-\\nvations. In other words, 100 −52.2 = 47.8% is the training error rate. As\\nwe have seen previously, the training error rate is often overly optimistic\\n— it tends to underestimate the test error rate. In order to better assess\\nthe accuracy of the logistic regression model in this setting, we can fit the\\nmodel using part of the data, and then examine how well it predicts the\\nheld out data. This will yield a more realistic error rate, in the sense that in\\npractice we will be interested in our model’s performance not on the data\\nthat we used to fit the model, but rather on days in the future for which\\nthe market’s movements are unknown.\\nTo implement this strategy, we first create a Boolean vector correspond-\\ning to the observations from 2001 through 2004. We then use this vector\\nto create a held out data set of observations from 2005.\\nIn [14]: train = (Smarket.Year < 2005)\\nSmarket_train = Smarket.loc[train]\\nSmarket_test = Smarket.loc[∼train]\\nSmarket_test.shape\\nOut[14]: (252, 9)\\nThe object train is a vector of 1,250 elements, corresponding to the\\nobservations in our data set. The elements of the vector that correspond\\nto observations that occurred before 2005 are set to True, whereas those\\nthat correspond to observations in 2005 are set to False. Hence train is a\\nboolean array, since its elements are True and False. Boolean arrays can be\\nused to obtain a subset of the rows or columns of a data frame using the 4.7 Lab: Logistic Regression, LDA, QDA, and KNN\\n177\\nloc method. For instance, the command Smarket.loc[train] would pick\\nout a submatrix of the stock market data set, corresponding only to the\\ndates before 2005, since those are the ones for which the elements of train\\nare True. The ∼symbol can be used to negate all of the elements of a\\nBoolean vector. That is, ∼train is a vector similar to train, except that\\nthe elements that are True in train get swapped to False in ∼train, and\\nvice versa. Therefore, Smarket.loc[∼train] yields a subset of the rows of\\nthe data frame of the stock market data containing only the observations\\nfor which train is False. The output above indicates that there are 252\\nsuch observations.\\nWe now fit a logistic regression model using only the subset of the ob-\\nservations that correspond to dates before 2005. We then obtain predicted\\nprobabilities of the stock market going up for each of the days in our test\\nset — that is, for the days in 2005.\\nIn [15]: X_train, X_test = X.loc[train], X.loc[∼train]\\ny_train, y_test = y.loc[train], y.loc[∼train]\\nglm_train = sm.GLM(y_train,\\nX_train,\\nfamily=sm.families.Binomial())\\nresults = glm_train.fit()\\nprobs = results.predict(exog=X_test)\\nNotice that we have trained and tested our model on two completely\\nseparate data sets: training was performed using only the dates before\\n2005, and testing was performed using only the dates in 2005.\\nFinally, we compare the predictions for 2005 to the actual movements of\\nthe market over that time period. We will first store the test and training\\nlabels (recall y_test is binary).\\nIn [16]: D = Smarket.Direction\\nL_train, L_test = D.loc[train], D.loc[∼train]\\nNow we threshold the fitted probability at 50% to form our predicted labels.\\nIn [17]: labels = np.array([\\'Down\\']*252)\\nlabels[probs >0.5] = \\'Up\\'\\nconfusion_table(labels, L_test)\\nOut[17]:\\nTruth\\nDown\\nUp\\nPredicted\\nDown\\n77\\n97\\nUp\\n34\\n44\\nThe test accuracy is about 48% while the error rate is about 52%\\nIn [18]: np.mean(labels == L_test), np.mean(labels != L_test)\\nOut[18]: (0.4802, 0.5198)\\nThe != notation means not equal to, and so the last command computes !=\\nthe test set error rate. The results are rather disappointing: the test error\\nrate is 52%, which is worse than random guessing! Of course this result\\nis not all that surprising, given that one would not generally expect to be\\nable to use previous days’ returns to predict future market performance.\\n(After all, if it were possible to do so, then the authors of this book would\\nbe out striking it rich rather than writing a statistics textbook.) 178\\n4. Classification\\nWe recall that the logistic regression model had very underwhelming p-\\nvalues associated with all of the predictors, and that the smallest p-value,\\nthough not very small, corresponded to Lag1. Perhaps by removing the\\nvariables that appear not to be helpful in predicting Direction, we can\\nobtain a more effective model. After all, using predictors that have no\\nrelationship with the response tends to cause a deterioration in the test\\nerror rate (since such predictors cause an increase in variance without a\\ncorresponding decrease in bias), and so removing such predictors may in\\nturn yield an improvement. Below we refit the logistic regression using just\\nLag1 and Lag2, which seemed to have the highest predictive power in the\\noriginal logistic regression model.\\nIn [19]: model = MS([\\'Lag1\\', \\'Lag2\\']).fit(Smarket)\\nX = model.transform(Smarket)\\nX_train, X_test = X.loc[train], X.loc[∼train]\\nglm_train = sm.GLM(y_train,\\nX_train,\\nfamily=sm.families.Binomial())\\nresults = glm_train.fit()\\nprobs = results.predict(exog=X_test)\\nlabels = np.array([\\'Down\\']*252)\\nlabels[probs >0.5] = \\'Up\\'\\nconfusion_table(labels, L_test)\\nOut[19]:\\nTruth\\nDown\\nUp\\nPredicted\\nDown\\n35\\n35\\nUp\\n76\\n106\\nLet’s evaluate the overall accuracy as well as the accuracy within the\\ndays when logistic regression predicts an increase.\\nIn [20]: (35+106)/252,106/(106+76)\\nOut[20]: (0.5595, 0.5824)\\nNow the results appear to be a little better: 56% of the daily movements\\nhave been correctly predicted. It is worth noting that in this case, a much\\nsimpler strategy of predicting that the market will increase every day will\\nalso be correct 56% of the time! Hence, in terms of overall error rate, the\\nlogistic regression method is no better than the naive approach. However,\\nthe confusion matrix shows that on days when logistic regression predicts\\nan increase in the market, it has a 58% accuracy rate. This suggests a\\npossible trading strategy of buying on days when the model predicts an in-\\ncreasing market, and avoiding trades on days when a decrease is predicted.\\nOf course one would need to investigate more carefully whether this small\\nimprovement was real or just due to random chance.\\nSuppose that we want to predict the returns associated with particular\\nvalues of Lag1 and Lag2. In particular, we want to predict Direction on a\\nday when Lag1 and Lag2 equal 1.2 and 1.1, respectively, and on a day when\\nthey equal 1.5 and −0.8. We do this using the predict() function.\\nIn [21]: newdata = pd.DataFrame({\\'Lag1\\':[1.2, 1.5],\\n\\'Lag2\\':[1.1, -0.8]}); 4.7 Lab: Logistic Regression, LDA, QDA, and KNN\\n179\\nnewX = model.transform(newdata)\\nresults.predict(newX)\\nOut[21]: 0\\n0.4791\\n1\\n0.4961\\ndtype: float64\\n4.7.3\\nLinear Discriminant Analysis\\nWe begin by performing LDA on the Smarket data, using the function\\nLinearDiscriminantAnalysis(), which we have abbreviated LDA(). We fit Linear\\nDiscriminant\\nAnalysis()\\nthe model using only the observations before 2005.\\nIn [22]: lda = LDA(store_covariance=True)\\nSince the LDA estimator automatically adds an intercept, we should re-\\nmove the column corresponding to the intercept in both X_train and X_test.\\nWe can also directly use the labels rather than the Boolean vectors y_train.\\nIn [23]: X_train, X_test = [M.drop(columns=[\\'intercept\\'])\\nfor M in [X_train, X_test]]\\nlda.fit(X_train, L_train)\\nOut[23]: LinearDiscriminantAnalysis(store_covariance=True)\\nHere we have used the list comprehensions introduced in Section 3.6.4.\\nLooking at our first line above, we see that the right-hand side is a list of\\nlength two. This is because the code for M in [X_train, X_test] iterates\\nover a list of length two. While here we loop over a list, the list comprehen-\\nsion method works when looping over any iterable object. We then apply\\nthe drop() method to each element in the iteration, collecting the result .drop()\\nin a list. The left-hand side tells Python to unpack this list of length two,\\nassigning its elements to the variables X_train and X_test. Of course, this\\noverwrites the previous values of X_train and X_test.\\nHaving fit the model, we can extract the means in the two classes with\\nthe means_ attribute. These are the average of each predictor within each\\nclass, and are used by LDA as estimates of µk. These suggest that there is\\na tendency for the previous 2 days’ returns to be negative on days when\\nthe market increases, and a tendency for the previous days’ returns to be\\npositive on days when the market declines.\\nIn [24]: lda.means_\\nOut[24]: array([[ 0.04,\\n0.03],\\n[-0.04, -0.03]])\\nThe estimated prior probabilities are stored in the priors_ attribute. The\\npackage sklearn typically uses this trailing _ to denote a quantity estimated\\nwhen using the fit() method. We can be sure of which entry corresponds\\nto which label by looking at the classes_ attribute.\\nIn [25]: lda.classes_\\nOut[25]: array([\\'Down\\', \\'Up\\'], dtype=\\'<U4\\') 180\\n4. Classification\\nThe LDA output indicates that ˆπDown = 0.492 and ˆπUp = 0.508.\\nIn [26]: lda.priors_\\nOut[26]: array([0.492, 0.508])\\nThe linear discriminant vectors can be found in the scalings_ attribute:\\nIn [27]: lda.scalings_\\nOut[27]: array([[-0.642],\\n[-0.513]])\\nThese values provide the linear combination of Lag1 and Lag2 that are used\\nto form the LDA decision rule. In other words, these are the multipliers of\\nthe elements of X = x in (4.24). If −0.64 × Lag1 −0.51 × Lag2 is large,\\nthen the LDA classifier will predict a market increase, and if it is small,\\nthen the LDA classifier will predict a market decline.\\nIn [28]: lda_pred = lda.predict(X_test)\\nAs we observed in our comparison of classification methods (Section 4.5),\\nthe LDA and logistic regression predictions are almost identical.\\nIn [29]: confusion_table(lda_pred, L_test)\\nOut[29]:\\nTruth\\nDown\\nUp\\nPredicted\\nDown\\n35\\n35\\nUp\\n76\\n106\\nWe can also estimate the probability of each class for each point in a\\ntraining set. Applying a 50% threshold to the posterior probabilities of be-\\ning in class one allows us to recreate the predictions contained in lda_pred.\\nIn [30]: lda_prob = lda.predict_proba(X_test)\\nnp.all(\\nnp.where(lda_prob[:,1] >= 0.5, \\'Up\\',\\'Down\\') == lda_pred\\n)\\nOut[30]: True\\nAbove, we used the np.where() function that creates an array with value np.where()\\n\\'Up\\' for indices where the second column of lda_prob (the estimated poste-\\nrior probability of \\'Up\\') is greater than 0.5. For problems with more than\\ntwo classes the labels are chosen as the class whose posterior probability is\\nhighest:\\nIn [31]: np.all(\\n[lda.classes_[i] for i in np.argmax(lda_prob, 1)] ==\\nlda_pred\\n)\\nOut[31]: True\\nIf we wanted to use a posterior probability threshold other than 50% in\\norder to make predictions, then we could easily do so. For instance, suppose\\nthat we wish to predict a market decrease only if we are very certain that the 4.7 Lab: Logistic Regression, LDA, QDA, and KNN\\n181\\nmarket will indeed decrease on that day — say, if the posterior probability\\nis at least 90%. We know that the first column of lda_prob corresponds to\\nthe label Down after having checked the classes_ attribute, hence we use\\nthe column index 0 rather than 1 as we did above.\\nIn [32]: np.sum(lda_prob[:,0] > 0.9)\\nOut[32]: 0\\nNo days in 2005 meet that threshold! In fact, the greatest posterior prob-\\nability of decrease in all of 2005 was 52.02%.\\nThe LDA classifier above is the first classifier from the sklearn library. We\\nwill use several other objects from this library. The objects follow a common\\nstructure that simplifies tasks such as cross-validation, which we will see in\\nChapter 5. Specifically, the methods first create a generic classifier without\\nreferring to any data. This classifier is then fit to data with the fit()\\nmethod and predictions are always produced with the predict() method.\\nThis pattern of first instantiating the classifier, followed by fitting it, and\\nthen producing predictions is an explicit design choice of sklearn. This\\nuniformity makes it possible to cleanly copy the classifier so that it can be\\nfit on different data; e.g. different training sets arising in cross-validation.\\nThis standard pattern also allows for a predictable formation of workflows.\\n4.7.4\\nQuadratic Discriminant Analysis\\nWe will now fit a QDA model to the Smarket data. QDA is implemented\\nvia QuadraticDiscriminantAnalysis() in the sklearn package, which we ab- Quadratic\\nDiscriminant\\nAnalysis()\\nbreviate to QDA(). The syntax is very similar to LDA().\\nIn [33]: qda = QDA(store_covariance=True)\\nqda.fit(X_train, L_train)\\nOut[33]: QuadraticDiscriminantAnalysis(store_covariance=True)\\nThe QDA() function will again compute means_ and priors_.\\nIn [34]: qda.means_, qda.priors_\\nOut[34]: (array([[ 0.04279022,\\n0.03389409],\\n[-0.03954635, -0.03132544]]),\\narray([0.49198397, 0.50801603]))\\nThe QDA() classifier will estimate one covariance per class. Here is the\\nestimated covariance in the first class:\\nIn [35]: qda.covariance_[0]\\nOut[35]: array([[ 1.50662277, -0.03924806],\\n[-0.03924806,\\n1.53559498]])\\nThe output contains the group means. But it does not contain the coef-\\nficients of the linear discriminants, because the QDA classifier involves a\\nquadratic, rather than a linear, function of the predictors. The predict()\\nfunction works in exactly the same fashion as for LDA. 182\\n4. Classification\\nIn [36]: qda_pred = qda.predict(X_test)\\nconfusion_table(qda_pred, L_test)\\nOut[36]:\\nTruth\\nDown\\nUp\\nPredicted\\nDown\\n30\\n20\\nUp\\n81\\n121\\nInterestingly, the QDA predictions are accurate almost 60% of the time,\\neven though the 2005 data was not used to fit the model.\\nIn [37]: np.mean(qda_pred == L_test)\\nOut[37]: 0.599\\nThis level of accuracy is quite impressive for stock market data, which\\nis known to be quite hard to model accurately. This suggests that the\\nquadratic form assumed by QDA may capture the true relationship more\\naccurately than the linear forms assumed by LDA and logistic regression.\\nHowever, we recommend evaluating this method’s performance on a larger\\ntest set before betting that this approach will consistently beat the market!\\n4.7.5\\nNaive Bayes\\nNext, we fit a naive Bayes model to the Smarket data. The syntax is similar\\nto that of LDA() and QDA(). By default, this implementation GaussianNB() of GaussianNB()\\nthe naive Bayes classifier models each quantitative feature using a Gaussian\\ndistribution. However, a kernel density method can also be used to estimate\\nthe distributions.\\nIn [38]: NB = GaussianNB()\\nNB.fit(X_train, L_train)\\nOut[38]: GaussianNB()\\nThe classes are stored as classes_.\\nIn [39]: NB.classes_\\nOut[39]: array([\\'Down\\', \\'Up\\'], dtype=\\'<U4\\')\\nThe class prior probabilities are stored in the class_prior_ attribute.\\nIn [40]: NB.class_prior_\\nOut[40]: array([0.49, 0.51])\\nThe parameters of the features can be found in the theta_ and var_\\nattributes. The number of rows is equal to the number of classes, while the\\nnumber of columns is equal to the number of features. We see below that\\nthe mean for feature Lag1 in the Down class is 0.043.\\nIn [41]: NB.theta_\\nOut[41]: array([[ 0.043,\\n0.034],\\n[-0.040, -0.031]]) 4.7 Lab: Logistic Regression, LDA, QDA, and KNN\\n183\\nIts variance is 1.503.\\nIn [42]: NB.var_\\nOut[42]: array([[1.503, 1.532],\\n[1.514, 1.487]])\\nHow do we know the names of these attributes? We use NB? (or ?NB).\\nWe can easily verify the mean computation:\\nIn [43]: X_train[L_train == \\'Down\\'].mean()\\nOut[43]: Lag1\\n0.042790\\nLag2\\n0.033894\\ndtype: float64\\nSimilarly for the variance:\\nIn [44]: X_train[L_train == \\'Down\\'].var(ddof=0)\\nOut[44]: Lag1\\n1.503554\\nLag2\\n1.532467\\ndtype: float64\\nThe GaussianNB() function calculates variances using the 1/n formula.6\\nSince NB() is a classifier in the sklearn library, making predictions uses the\\nsame syntax as for LDA() and QDA() above.\\nIn [45]: nb_labels = NB.predict(X_test)\\nconfusion_table(nb_labels, L_test)\\nOut[45]:\\nTruth\\nDown\\nUp\\nPredicted\\nDown\\n29\\n20\\nUp\\n82\\n121\\nNaive Bayes performs well on these data, with accurate predictions over\\n59% of the time. This is slightly worse than QDA, but much better than\\nLDA.\\nAs for LDA, the predict_proba() method estimates the probability that\\neach observation belongs to a particular class.\\nIn [46]: NB.predict_proba(X_test)[:5]\\nOut[46]: array([[0.4873, 0.5127],\\n[0.4762, 0.5238],\\n[0.4653, 0.5347],\\n[0.4748, 0.5252],\\n[0.4902, 0.5098]])\\n4.7.6\\nK-Nearest Neighbors\\nWe will now perform KNN using the KNeighborsClassifier() function. This KNeighbors\\nClassifier()\\n6There are two formulas for computing the sample variance of n observations\\nx1, . . . , xn:\\n1\\nn\\n!n\\ni=1(xi −¯x)2 and\\n1\\nn−1\\n!n\\ni=1(xi −¯x)2 where ¯x is the sample mean.\\nIn most cases the distinction is not important. 184\\n4. Classification\\nfunction works similarly to the other model-fitting functions that we have\\nencountered thus far.\\nAs is the case for LDA and QDA, we fit the classifier using the fit\\nmethod. New predictions are formed using the predict method of the object\\nreturned by fit().\\nIn [47]: knn1 = KNeighborsClassifier(n_neighbors=1)\\nknn1.fit(X_train, L_train)\\nknn1_pred = knn1.predict(X_test)\\nconfusion_table(knn1_pred, L_test)\\nOut[47]:\\nTruth\\nDown\\nUp\\nPredicted\\nDown\\n43\\n58\\nUp\\n68\\n83\\nThe results using K = 1 are not very good, since only 50% of the observa-\\ntions are correctly predicted. Of course, it may be that K = 1 results in an\\noverly-flexible fit to the data.\\nIn [48]: (83+43)/252, np.mean(knn1_pred == L_test)\\nOut[48]: (0.5, 0.5)\\nWe repeat the analysis below using K = 3.\\nIn [49]: knn3 = KNeighborsClassifier(n_neighbors=3)\\nknn3_pred = knn3.fit(X_train, L_train).predict(X_test)\\nnp.mean(knn3_pred == L_test)\\nOut[49]: 0.532\\nThe results have improved slightly. But increasing K further provides no\\nfurther improvements. It appears that for these data, and this train/test\\nsplit, QDA gives the best results of the methods that we have examined so\\nfar.\\nKNN does not perform well on the Smarket data, but it often does provide\\nimpressive results. As an example we will apply the KNN approach to the\\nCaravan data set, which is part of the ISLP library. This data set includes 85\\npredictors that measure demographic characteristics for 5,822 individuals.\\nThe response variable is Purchase, which indicates whether or not a given\\nindividual purchases a caravan insurance policy. In this data set, only 6%\\nof people purchased caravan insurance.\\nIn [50]: Caravan = load_data(\\'Caravan\\')\\nPurchase = Caravan.Purchase\\nPurchase.value_counts()\\nOut[50]: No\\n5474\\nYes\\n348\\nName: Purchase, dtype: int64\\nThe method value_counts() takes a pd.Series or pd.DataFrame and re-\\nturns a pd.Series with the corresponding counts for each unique element.\\nIn this case Purchase has only Yes and No values and returns how many\\nvalues of each there are. 4.7 Lab: Logistic Regression, LDA, QDA, and KNN\\n185\\nIn [51]: 348 / 5822\\nOut[51]: 0.0598\\nOur features will include all columns except Purchase.\\nIn [52]: feature_df = Caravan.drop(columns=[\\'Purchase\\'])\\nBecause the KNN classifier predicts the class of a given test observa-\\ntion by identifying the observations that are nearest to it, the scale of\\nthe variables matters. Any variables that are on a large scale will have a\\nmuch larger effect on the distance between the observations, and hence on\\nthe KNN classifier, than variables that are on a small scale. For instance,\\nimagine a data set that contains two variables, salary and age (measured\\nin dollars and years, respectively). As far as KNN is concerned, a difference\\nof 1,000 USD in salary is enormous compared to a difference of 50 years in\\nage. Consequently, salary will drive the KNN classification results, and age\\nwill have almost no effect. This is contrary to our intuition that a salary\\ndifference of 1,000 USD is quite small compared to an age difference of 50\\nyears. Furthermore, the importance of scale to the KNN classifier leads to\\nanother issue: if we measured salary in Japanese yen, or if we measured age\\nin minutes, then we’d get quite different classification results from what we\\nget if these two variables are measured in dollars and years.\\nA good way to handle this problem is to standardize the data so that all standardize\\nvariables are given a mean of zero and a standard deviation of one. Then\\nall variables will be on a comparable scale. This is accomplished using the\\nStandardScaler() transformation.\\nStandard\\nScaler()\\nIn [53]: scaler = StandardScaler(with_mean=True,\\nwith_std=True,\\ncopy=True)\\nThe argument with_mean indicates whether or not we should subtract the\\nmean, while with_std indicates whether or not we should scale the columns\\nto have standard deviation of 1 or not. Finally, the argument copy=True in-\\ndicates that we will always copy data, rather than trying to do calculations\\nin place where possible.\\nThis transformation can be fit and then applied to arbitrary data. In the\\nfirst line below, the parameters for the scaling are computed and stored\\nin scaler, while the second line actually constructs the standardized set of\\nfeatures.\\nIn [54]: scaler.fit(feature_df)\\nX_std = scaler.transform(feature_df)\\nNow every column of feature_std below has a standard deviation of one\\nand a mean of zero.\\nIn [55]: feature_std = pd.DataFrame(\\nX_std,\\ncolumns=feature_df.columns);\\nfeature_std.std()\\nOut[55]: MOSTYPE\\n1.000086\\nMAANTHUI\\n1.000086 186\\n4. Classification\\nMGEMOMV\\n1.000086\\nMGEMLEEF\\n1.000086\\nMOSHOOFD\\n1.000086\\n...\\nAZEILPL\\n1.000086\\nAPLEZIER\\n1.000086\\nAFIETS\\n1.000086\\nAINBOED\\n1.000086\\nABYSTAND\\n1.000086\\nLength: 85, dtype: float64\\nNotice that the standard deviations are not quite 1 here; this is again due\\nto some procedures using the 1/n convention for variances (in this case\\nscaler()), while others use 1/(n −1) (the std() method). See the footnote .std()\\non page 183. In this case it does not matter, as long as the variables are all\\non the same scale.\\nUsing the function train_test_split() we now split the observations into train_test_\\nsplit()\\na test set, containing 1000 observations, and a training set containing the\\nremaining observations. The argument random_state=0 ensures that we get\\nthe same split each time we rerun the code.\\nIn [56]: (X_train,\\nX_test,\\ny_train,\\ny_test) = train_test_split(feature_std ,\\nPurchase,\\ntest_size=1000,\\nrandom_state=0)\\n?train_test_split reveals that the non-keyword arguments can be lists,\\narrays, pandas dataframes etc that all have the same length (shape[0]) and\\nhence are indexable. In this case they are the dataframe feature_std and indexable\\nthe response variable Purchase. We fit a KNN model on the training data\\nusing K = 1, and evaluate its performance on the test data.\\nIn [57]: knn1 = KNeighborsClassifier(n_neighbors=1)\\nknn1_pred = knn1.fit(X_train, y_train).predict(X_test)\\nnp.mean(y_test != knn1_pred), np.mean(y_test != \"No\")\\nOut[57]: (0.111, 0.067)\\nThe KNN error rate on the 1,000 test observations is about 11%. At first\\nglance, this may appear to be fairly good. However, since just over 6% of\\ncustomers purchased insurance, we could get the error rate down to almost\\n6% by always predicting No regardless of the values of the predictors! This\\nis known as the null rate.\\nnull rate\\nSuppose that there is some non-trivial cost to trying to sell insurance\\nto a given individual. For instance, perhaps a salesperson must visit each\\npotential customer. If the company tries to sell insurance to a random\\nselection of customers, then the success rate will be only 6%, which may\\nbe far too low given the costs involved. Instead, the company would like\\nto try to sell insurance only to customers who are likely to buy it. So the\\noverall error rate is not of interest. Instead, the fraction of individuals that\\nare correctly predicted to buy insurance is of interest.\\nIn [58]: confusion_table(knn1_pred, y_test) 4.7 Lab: Logistic Regression, LDA, QDA, and KNN\\n187\\nOut[58]:\\nTruth\\nNo\\nYes\\nPredicted\\nNo\\n880\\n58\\nYes\\n53\\n9\\nIt turns out that KNN with K = 1 does far better than random guessing\\namong the customers that are predicted to buy insurance. Among 62 such\\ncustomers, 9, or 14.5%, actually do purchase insurance. This is double the\\nrate that one would obtain from random guessing.\\nIn [59]: 9/(53+9)\\nOut[59]: 0.145\\nTuning Parameters\\nThe number of neighbors in KNN is referred to as a tuning parameter, also tuning\\nparameter\\nreferred to as a hyperparameter. We do not know a priori what value to\\nhyper-\\nparameter\\nuse. It is therefore of interest to see how the classifier performs on test\\ndata as we vary these parameters. This can be achieved with a for loop,\\ndescribed in Section 2.3.8. Here we use a for loop to look at the accuracy of\\nour classifier in the group predicted to purchase insurance as we vary the\\nnumber of neighbors from 1 to 5:\\nIn [60]: for K in range(1,6):\\nknn = KNeighborsClassifier(n_neighbors=K)\\nknn_pred = knn.fit(X_train, y_train).predict(X_test)\\nC = confusion_table(knn_pred, y_test)\\ntempl = (\\'K={0:d}: # predicted to rent: {1:>2},\\' +\\n\\'\\n# who did rent {2:d}, accuracy {3:.1%}\\')\\npred = C.loc[\\'Yes\\'].sum()\\ndid_rent = C.loc[\\'Yes\\',\\'Yes\\']\\nprint(templ.format(\\nK,\\npred,\\ndid_rent,\\ndid_rent / pred))\\nK=1: # predicted to rent: 62,# who did rent 9, accuracy 14.5%\\nK=2: # predicted to rent: 6,# who did rent 1, accuracy 16.7%\\nK=3: # predicted to rent: 20,# who did rent 3, accuracy 15.0%\\nK=4: # predicted to rent: 3,# who did rent 0, accuracy 0.0%\\nK=5: # predicted to rent: 7,# who did rent 1, accuracy 14.3%\\nWe see some variability — the numbers for K=4 are very different from the\\nrest.\\nComparison to Logistic Regression\\nAs a comparison, we can also fit a logistic regression model to the data.\\nThis can also be done with sklearn, though by default it fits something\\nlike the ridge regression version of logistic regression, which we introduce\\nin Chapter 6. This can be modified by appropriately setting the argument\\nC below. Its default value is 1 but by setting it to a very large number,\\nthe algorithm converges to the same solution as the usual (unregularized)\\nlogistic regression estimator discussed above. 188\\n4. Classification\\nUnlike the statsmodels package, sklearn focuses less on inference and\\nmore on classification. Hence, the summary methods seen in statsmodels\\nand our simplified version seen with summarize are not generally available\\nfor the classifiers in sklearn.\\nIn [61]: logit = LogisticRegression(C=1e10, solver=\\'liblinear\\')\\nlogit.fit(X_train, y_train)\\nlogit_pred = logit.predict_proba(X_test)\\nlogit_labels = np.where(logit_pred[:,1] > 5, \\'Yes\\', \\'No\\')\\nconfusion_table(logit_labels , y_test)\\nOut[61]:\\nTruth\\nNo\\nYes\\nPredicted\\nNo\\n933\\n67\\nYes\\n0\\n0\\nWe used the argument solver=\\'liblinear\\' above to avoid a warning with\\nthe default solver which would indicate that the algorithm does not con-\\nverge.\\nIf we use 0.5 as the predicted probability cut-off for the classifier, then\\nwe have a problem: none of the test observations are predicted to purchase\\ninsurance. However, we are not required to use a cut-off of 0.5. If we instead\\npredict a purchase any time the predicted probability of purchase exceeds\\n0.25, we get much better results: we predict that 29 people will purchase\\ninsurance, and we are correct for about 31% of these people. This is almost\\nfive times better than random guessing!\\nIn [62]: logit_labels = np.where(logit_pred[:,1]>0.25, \\'Yes\\', \\'No\\')\\nconfusion_table(logit_labels , y_test)\\nOut[62]:\\nTruth\\nNo\\nYes\\nPredicted\\nNo\\n913\\n58\\nYes\\n20\\n9\\nIn [63]: 9/(20+9)\\nOut[63]: 0.310\\n4.7.7\\nLinear and Poisson Regression on the Bikeshare Data\\nHere we fit linear and Poisson regression models to the Bikeshare data, as\\ndescribed in Section 4.6. The response bikers measures the number of bike\\nrentals per hour in Washington, DC in the period 2010–2012.\\nIn [64]: Bike = load_data(\\'Bikeshare\\')\\nLet’s have a peek at the dimensions and names of the variables in this\\ndataframe.\\nIn [65]: Bike.shape, Bike.columns 4.7 Lab: Logistic Regression, LDA, QDA, and KNN\\n189\\nOut[65]: ((8645, 15),\\nIndex([\\'season\\', \\'mnth\\', \\'day\\', \\'hr\\', \\'holiday\\', \\'weekday\\',\\n\\'workingday\\', \\'weathersit\\', \\'temp\\', \\'atemp\\', \\'hum\\',\\n\\'windspeed\\', \\'casual\\', \\'registered\\', \\'bikers\\'],\\ndtype=\\'object\\'))\\nLinear Regression\\nWe begin by fitting a linear regression model to the data.\\nIn [66]: X = MS([\\'mnth\\',\\n\\'hr\\',\\n\\'workingday\\',\\n\\'temp\\',\\n\\'weathersit\\']).fit_transform(Bike)\\nY = Bike[\\'bikers\\']\\nM_lm = sm.OLS(Y, X).fit()\\nsummarize(M_lm)\\nOut[66]:\\ncoef\\nstd err\\nt\\nP>|t|\\nintercept\\n-68.6317\\n5.307 -12.932\\n0.000\\nmnth[Feb]\\n6.8452\\n4.287\\n1.597\\n0.110\\nmnth[March]\\n16.5514\\n4.301\\n3.848\\n0.000\\nmnth[April]\\n41.4249\\n4.972\\n8.331\\n0.000\\nmnth[May]\\n72.5571\\n5.641\\n12.862\\n0.000\\nmnth[June]\\n67.8187\\n6.544\\n10.364\\n0.000\\nmnth[July]\\n45.3245\\n7.081\\n6.401\\n0.000\\nmnth[Aug]\\n53.2430\\n6.640\\n8.019\\n0.000\\nmnth[Sept]\\n66.6783\\n5.925\\n11.254\\n0.000\\nmnth[Oct]\\n75.8343\\n4.950\\n15.319\\n0.000\\nmnth[Nov]\\n60.3100\\n4.610\\n13.083\\n0.000\\nmnth[Dec]\\n46.4577\\n4.271\\n10.878\\n0.000\\nhr[1]\\n-14.5793\\n5.699\\n-2.558\\n0.011\\nhr[2]\\n-21.5791\\n5.733\\n-3.764\\n0.000\\nhr[3]\\n-31.1408\\n5.778\\n-5.389\\n0.000\\n.....\\n.......\\n.....\\n.....\\n.....\\nThere are 24 levels in hr and 40 rows in all, so we have truncated the\\nsummary. In M_lm, the first levels hr[0] and mnth[Jan] are treated as the\\nbaseline values, and so no coefficient estimates are provided for them: im-\\nplicitly, their coefficient estimates are zero, and all other levels are measured\\nrelative to these baselines. For example, the Feb coefficient of 6.845 signi-\\nfies that, holding all other variables constant, there are on average about\\n7 more riders in February than in January. Similarly there are about 16.5\\nmore riders in March than in January.\\nThe results seen in Section 4.6.1 used a slightly different coding of the\\nvariables hr and mnth, as follows:\\nIn [67]: hr_encode = contrast(\\'hr\\', \\'sum\\')\\nmnth_encode = contrast(\\'mnth\\', \\'sum\\')\\nRefitting again:\\nIn [68]: X2 = MS([mnth_encode ,\\nhr_encode,\\n\\'workingday\\',\\n\\'temp\\', 190\\n4. Classification\\n\\'weathersit\\']).fit_transform(Bike)\\nM2_lm = sm.OLS(Y, X2).fit()\\nS2 = summarize(M2_lm)\\nS2\\nOut[68]:\\ncoef\\nstd err\\nt\\nP>|t|\\nintercept\\n73.5974\\n5.132\\n14.340\\n0.000\\nmnth[Jan]\\n-46.0871\\n4.085 -11.281\\n0.000\\nmnth[Feb]\\n-39.2419\\n3.539 -11.088\\n0.000\\nmnth[March]\\n-29.5357\\n3.155\\n-9.361\\n0.000\\nmnth[April]\\n-4.6622\\n2.741\\n-1.701\\n0.089\\nmnth[May]\\n26.4700\\n2.851\\n9.285\\n0.000\\nmnth[June]\\n21.7317\\n3.465\\n6.272\\n0.000\\nmnth[July]\\n-0.7626\\n3.908\\n-0.195\\n0.845\\nmnth[Aug]\\n7.1560\\n3.535\\n2.024\\n0.043\\nmnth[Sept]\\n20.5912\\n3.046\\n6.761\\n0.000\\nmnth[Oct]\\n29.7472\\n2.700\\n11.019\\n0.000\\nmnth[Nov]\\n14.2229\\n2.860\\n4.972\\n0.000\\nhr[0]\\n-96.1420\\n3.955 -24.307\\n0.000\\nhr[1]\\n-110.7213\\n3.966 -27.916\\n0.000\\nhr[2]\\n-117.7212\\n4.016 -29.310\\n0.000\\n.....\\n.......\\n.....\\n......\\n.....\\nWhat is the difference between the two codings? In M2_lm, a coefficient esti-\\nmate is reported for all but level 23 of hr and level Dec of mnth. Importantly,\\nin M2_lm, the (unreported) coefficient estimate for the last level of mnth is\\nnot zero: instead, it equals the negative of the sum of the coefficient esti-\\nmates for all of the other levels. Similarly, in M2_lm, the coefficient estimate\\nfor the last level of hr is the negative of the sum of the coefficient estimates\\nfor all of the other levels. This means that the coefficients of hr and mnth\\nin M2_lm will always sum to zero, and can be interpreted as the difference\\nfrom the mean level. For example, the coefficient for January of −46.087\\nindicates that, holding all other variables constant, there are typically 46\\nfewer riders in January relative to the yearly average.\\nIt is important to realize that the choice of coding really does not matter,\\nprovided that we interpret the model output correctly in light of the coding\\nused. For example, we see that the predictions from the linear model are\\nthe same regardless of coding:\\nIn [69]: np.sum((M_lm.fittedvalues - M2_lm.fittedvalues)**2)\\nOut[69]: 1.53e-20\\nThe sum of squared differences is zero. We can also see this using the\\nnp.allclose() function:\\nnp.allclose()\\nIn [70]: np.allclose(M_lm.fittedvalues , M2_lm.fittedvalues)\\nOut[70]: True\\nTo reproduce the left-hand side of Figure 4.13 we must first obtain the\\ncoefficient estimates associated with mnth. The coefficients for January\\nthrough November can be obtained directly from the M2_lm object. The\\ncoefficient for December must be explicitly computed as the negative sum\\nof all the other months. We first extract all the coefficients for month from\\nthe coefficients of M2_lm. 4.7 Lab: Logistic Regression, LDA, QDA, and KNN\\n191\\nIn [71]: coef_month = S2[S2.index.str.contains(\\'mnth\\')][\\'coef\\']\\ncoef_month\\nOut[71]: mnth[Jan]\\n-46.0871\\nmnth[Feb]\\n-39.2419\\nmnth[March]\\n-29.5357\\nmnth[April]\\n-4.6622\\nmnth[May]\\n26.4700\\nmnth[June]\\n21.7317\\nmnth[July]\\n-0.7626\\nmnth[Aug]\\n7.1560\\nmnth[Sept]\\n20.5912\\nmnth[Oct]\\n29.7472\\nmnth[Nov]\\n14.2229\\nName: coef, dtype: float64\\nNext, we append Dec as the negative of the sum of all other months.\\nIn [72]: months = Bike[\\'mnth\\'].dtype.categories\\ncoef_month = pd.concat([\\ncoef_month,\\npd.Series([-coef_month.sum()],\\nindex=[\\'mnth[Dec]\\'\\n])\\n])\\ncoef_month\\nOut[72]: mnth[Jan]\\n-46.0871\\nmnth[Feb]\\n-39.2419\\nmnth[March]\\n-29.5357\\nmnth[April]\\n-4.6622\\nmnth[May]\\n26.4700\\nmnth[June]\\n21.7317\\nmnth[July]\\n-0.7626\\nmnth[Aug]\\n7.1560\\nmnth[Sept]\\n20.5912\\nmnth[Oct]\\n29.7472\\nmnth[Nov]\\n14.2229\\nmnth[Dec]\\n0.3705\\nName: coef, dtype: float64\\nFinally, to make the plot neater, we’ll just use the first letter of each month,\\nwhich is the 6th entry of each of the labels in the index.\\nIn [73]: fig_month, ax_month = subplots(figsize=(8,8))\\nx_month = np.arange(coef_month.shape[0])\\nax_month.plot(x_month, coef_month , marker=\\'o\\', ms=10)\\nax_month.set_xticks(x_month)\\nax_month.set_xticklabels([l[5] for l in coef_month.index], fontsize\\n=20)\\nax_month.set_xlabel(\\'Month\\', fontsize=20)\\nax_month.set_ylabel(\\'Coefficient\\', fontsize=20);\\nReproducing the right-hand plot in Figure 4.13 follows a similar process.\\nIn [74]: coef_hr = S2[S2.index.str.contains(\\'hr\\')][\\'coef\\']\\ncoef_hr = coef_hr.reindex([\\'hr[{0}]\\'.format(h) for h in range(23)])\\ncoef_hr = pd.concat([coef_hr, 192\\n4. Classification\\npd.Series([-coef_hr.sum()], index=[\\'hr[23]\\'])\\n])\\nWe now make the hour plot.\\nIn [75]: fig_hr, ax_hr = subplots(figsize=(8,8))\\nx_hr = np.arange(coef_hr.shape[0])\\nax_hr.plot(x_hr, coef_hr, marker=\\'o\\', ms=10)\\nax_hr.set_xticks(x_hr[::2])\\nax_hr.set_xticklabels(range(24)[::2], fontsize=20)\\nax_hr.set_xlabel(\\'Hour\\', fontsize=20)\\nax_hr.set_ylabel(\\'Coefficient\\', fontsize=20);\\nPoisson Regression\\nNow we fit instead a Poisson regression model to the Bikeshare data. Very\\nlittle changes, except that we now use the function sm.GLM() with the Pois-\\nson family specified:\\nIn [76]: M_pois = sm.GLM(Y, X2, family=sm.families.Poisson()).fit()\\nWe can plot the coefficients associated with mnth and hr, in order to\\nreproduce Figure 4.15. We first complete these coefficients as before.\\nIn [77]: S_pois = summarize(M_pois)\\ncoef_month = S_pois[S_pois.index.str.contains(\\'mnth\\')][\\'coef\\']\\ncoef_month = pd.concat([coef_month,\\npd.Series([-coef_month.sum()],\\nindex=[\\'mnth[Dec]\\'])])\\ncoef_hr = S_pois[S_pois.index.str.contains(\\'hr\\')][\\'coef\\']\\ncoef_hr = pd.concat([coef_hr,\\npd.Series([-coef_hr.sum()],\\nindex=[\\'hr[23]\\'])])\\nThe plotting is as before.\\nIn [78]: fig_pois, (ax_month, ax_hr) = subplots(1, 2, figsize=(16,8))\\nax_month.plot(x_month, coef_month , marker=\\'o\\', ms=10)\\nax_month.set_xticks(x_month)\\nax_month.set_xticklabels([l[5] for l in coef_month.index], fontsize\\n=20)\\nax_month.set_xlabel(\\'Month\\', fontsize=20)\\nax_month.set_ylabel(\\'Coefficient\\', fontsize=20)\\nax_hr.plot(x_hr, coef_hr, marker=\\'o\\', ms=10)\\nax_hr.set_xticklabels(range(24)[::2], fontsize=20)\\nax_hr.set_xlabel(\\'Hour\\', fontsize=20)\\nax_hr.set_ylabel(\\'Coefficient\\', fontsize=20);\\nWe compare the fitted values of the two models. The fitted values are stored\\nin the fittedvalues attribute returned by the fit() method for both the\\nlinear regression and the Poisson fits. The linear predictors are stored as\\nthe attribute lin_pred.\\nIn [79]: fig, ax = subplots(figsize=(8, 8))\\nax.scatter(M2_lm.fittedvalues ,\\nM_pois.fittedvalues ,\\ns=20)\\nax.set_xlabel(\\'Linear Regression Fit\\', fontsize=20) 4.8 Exercises\\n193\\nax.set_ylabel(\\'Poisson Regression Fit\\', fontsize=20)\\nax.axline([0,0], c=\\'black\\', linewidth=3,\\nlinestyle=\\'--\\', slope=1);\\nThe predictions from the Poisson regression model are correlated with\\nthose from the linear model; however, the former are non-negative. As a\\nresult the Poisson regression predictions tend to be larger than those from\\nthe linear model for either very low or very high levels of ridership.\\nIn this section, we fit Poisson regression models using the sm.GLM() func-\\ntion with the argument family=sm.families.Poisson(). Earlier in this lab\\nwe used the sm.GLM() function with family=sm.families.Binomial() to per-\\nform logistic regression. Other choices for the family argument can be used\\nto fit other types of GLMs. For instance, family=sm.families.Gamma() fits\\na Gamma regression model.\\n4.8\\nExercises\\nConceptual\\n1. Using a little bit of algebra, prove that (4.2) is equivalent to (4.3). In\\nother words, the logistic function representation and logit represen-\\ntation for the logistic regression model are equivalent.\\n2. It was stated in the text that classifying an observation to the class\\nfor which (4.17) is largest is equivalent to classifying an observation\\nto the class for which (4.18) is largest. Prove that this is the case. In\\nother words, under the assumption that the observations in the kth\\nclass are drawn from a N(µk, σ2) distribution, the Bayes classifier\\nassigns an observation to the class for which the discriminant function\\nis maximized.\\n3. This problem relates to the QDA model, in which the observations\\nwithin each class are drawn from a normal distribution with a class-\\nspecific mean vector and a class specific covariance matrix. We con-\\nsider the simple case where p = 1; i.e. there is only one feature.\\nSuppose that we have K classes, and that if an observation belongs\\nto the kth class then X comes from a one-dimensional normal dis-\\ntribution, X ∼N(µk, σ2\\nk). Recall that the density function for the\\none-dimensional normal distribution is given in (4.16). Prove that in\\nthis case, the Bayes classifier is not linear. Argue that it is in fact\\nquadratic.\\nHint: For this problem, you should follow the arguments laid out in\\nSection 4.4.1, but without making the assumption that σ2\\n1 = · · · = σ2\\nK.\\n4. When the number of features p is large, there tends to be a deteri-\\noration in the performance of KNN and other local approaches that\\nperform prediction using only observations that are near the test ob-\\nservation for which a prediction must be made. This phenomenon is\\nknown as the curse of dimensionality, and it ties into the fact that curse of di-\\nmensionality\\nnon-parametric approaches often perform poorly when p is large. We\\nwill now investigate this curse. 194\\n4. Classification\\n(a) Suppose that we have a set of observations, each with measure-\\nments on p = 1 feature, X. We assume that X is uniformly\\n(evenly) distributed on [0, 1]. Associated with each observation\\nis a response value. Suppose that we wish to predict a test obser-\\nvation’s response using only observations that are within 10 % of\\nthe range of X closest to that test observation. For instance, in\\norder to predict the response for a test observation with X = 0.6,\\nwe will use observations in the range [0.55, 0.65]. On average,\\nwhat fraction of the available observations will we use to make\\nthe prediction?\\n(b) Now suppose that we have a set of observations, each with\\nmeasurements on p = 2 features, X1 and X2. We assume that\\n(X1, X2) are uniformly distributed on [0, 1] × [0, 1]. We wish to\\npredict a test observation’s response using only observations that\\nare within 10 % of the range of X1 and within 10 % of the range\\nof X2 closest to that test observation. For instance, in order to\\npredict the response for a test observation with X1 = 0.6 and\\nX2 = 0.35, we will use observations in the range [0.55, 0.65] for\\nX1 and in the range [0.3, 0.4] for X2. On average, what fraction\\nof the available observations will we use to make the prediction?\\n(c) Now suppose that we have a set of observations on p = 100 fea-\\ntures. Again the observations are uniformly distributed on each\\nfeature, and again each feature ranges in value from 0 to 1. We\\nwish to predict a test observation’s response using observations\\nwithin the 10 % of each feature’s range that is closest to that test\\nobservation. What fraction of the available observations will we\\nuse to make the prediction?\\n(d) Using your answers to parts (a)–(c), argue that a drawback of\\nKNN when p is large is that there are very few training obser-\\nvations “near” any given test observation.\\n(e) Now suppose that we wish to make a prediction for a test obser-\\nvation by creating a p-dimensional hypercube centered around\\nthe test observation that contains, on average, 10 % of the train-\\ning observations. For p = 1, 2, and 100, what is the length of each\\nside of the hypercube? Comment on your answer.\\nNote: A hypercube is a generalization of a cube to an arbitrary\\nnumber of dimensions. When p = 1, a hypercube is simply a line\\nsegment, when p = 2 it is a square, and when p = 100 it is a\\n100-dimensional cube.\\n5. We now examine the differences between LDA and QDA.\\n(a) If the Bayes decision boundary is linear, do we expect LDA or\\nQDA to perform better on the training set? On the test set?\\n(b) If the Bayes decision boundary is non-linear, do we expect LDA\\nor QDA to perform better on the training set? On the test set? 4.8 Exercises\\n195\\n(c) In general, as the sample size n increases, do we expect the test\\nprediction accuracy of QDA relative to LDA to improve, decline,\\nor be unchanged? Why?\\n(d) True or False: Even if the Bayes decision boundary for a given\\nproblem is linear, we will probably achieve a superior test er-\\nror rate using QDA rather than LDA because QDA is flexible\\nenough to model a linear decision boundary. Justify your an-\\nswer.\\n6. Suppose we collect data for a group of students in a statistics class\\nwith variables X1 = hours studied, X2 = undergrad GPA, and Y =\\nreceive an A. We fit a logistic regression and produce estimated\\ncoefficient, ˆβ0 = −6, ˆβ1 = 0.05, ˆβ2 = 1.\\n(a) Estimate the probability that a student who studies for 40 h and\\nhas an undergrad GPA of 3.5 gets an A in the class.\\n(b) How many hours would the student in part (a) need to study to\\nhave a 50 % chance of getting an A in the class?\\n7. Suppose that we wish to predict whether a given stock will issue a\\ndividend this year (“Yes” or “No”) based on X, last year’s percent\\nprofit. We examine a large number of companies and discover that the\\nmean value of X for companies that issued a dividend was ¯X = 10,\\nwhile the mean for those that didn’t was ¯X = 0. In addition, the\\nvariance of X for these two sets of companies was ˆσ2 = 36. Finally,\\n80 % of companies issued dividends. Assuming that X follows a nor-\\nmal distribution, predict the probability that a company will issue\\na dividend this year given that its percentage profit was X = 4 last\\nyear.\\nHint: Recall that the density function for a normal random variable\\nis f(x) =\\n1\\n√\\n2πσ2 e−(x−µ)2/2σ2. You will need to use Bayes’ theorem.\\n8. Suppose that we take a data set, divide it into equally-sized training\\nand test sets, and then try out two different classification procedures.\\nFirst we use logistic regression and get an error rate of 20 % on the\\ntraining data and 30 % on the test data. Next we use 1-nearest neigh-\\nbors (i.e. K = 1) and get an average error rate (averaged over both\\ntest and training data sets) of 18 %. Based on these results, which\\nmethod should we prefer to use for classification of new observations?\\nWhy?\\n9. This problem has to do with odds.\\n(a) On average, what fraction of people with an odds of 0.37 of\\ndefaulting on their credit card payment will in fact default?\\n(b) Suppose that an individual has a 16 % chance of defaulting on\\nher credit card payment. What are the odds that she will de-\\nfault? 196\\n4. Classification\\n10. Equation 4.32 derived an expression for log\\n1\\nPr(Y =k|X=x)\\nPr(Y =K|X=x)\\n2\\nin the\\nsetting where p > 1, so that the mean for the kth class, µk, is a p-\\ndimensional vector, and the shared covariance Σ is a p × p matrix.\\nHowever, in the setting with p = 1, (4.32) takes a simpler form, since\\nthe means µ1, . . . , µK and the variance σ2 are scalars. In this simpler\\nsetting, repeat the calculation in (4.32), and provide expressions for\\nak and bkj in terms of πk, πK, µk, µK, and σ2.\\n11. Work out the detailed forms of ak, bkj, and bkjl in (4.33). Your answer\\nshould involve πk, πK, µk, µK, Σk, and ΣK.\\n12. Suppose that you wish to classify an observation X ∈R into apples\\nand oranges. You fit a logistic regression model and find that\\n6\\nPr(Y = orange|X = x) =\\nexp(ˆβ0 + ˆβ1x)\\n1 + exp(ˆβ0 + ˆβ1x)\\n.\\nYour friend fits a logistic regression model to the same data using the\\nsoftmax formulation in (4.13), and finds that\\n6\\nPr(Y = orange|X = x) =\\nexp(ˆαorange0 + ˆαorange1x)\\nexp(ˆαorange0 + ˆαorange1x) + exp(ˆαapple0 + ˆαapple1x).\\n(a) What is the log odds of orange versus apple in your model?\\n(b) What is the log odds of orange versus apple in your friend’s\\nmodel?\\n(c) Suppose that in your model, ˆβ0 = 2 and ˆβ1 = −1. What are\\nthe coefficient estimates in your friend’s model? Be as specific\\nas possible.\\n(d) Now suppose that you and your friend fit the same two models\\non a different data set. This time, your friend gets the coefficient\\nestimates ˆαorange0 = 1.2, ˆαorange1 = −2, ˆαorange0 = 3, ˆαorange1 =\\n0.6. What are the coefficient estimates in your model?\\n(e) Finally, suppose you apply both models from (d) to a data set\\nwith 2,000 test observations. What fraction of the time do you\\nexpect the predicted class labels from your model to agree with\\nthose from your friend’s model? Explain your answer.\\nApplied\\n13. This question should be answered using the Weekly data set, which\\nis part of the ISLP package. This data is similar in nature to the\\nSmarket data from this chapter’s lab, except that it contains 1, 089\\nweekly returns for 21 years, from the beginning of 1990 to the end of\\n2010.\\n(a) Produce some numerical and graphical summaries of the Weekly\\ndata. Do there appear to be any patterns? 4.8 Exercises\\n197\\n(b) Use the full data set to perform a logistic regression with\\nDirection as the response and the five lag variables plus Volume\\nas predictors. Use the summary function to print the results. Do\\nany of the predictors appear to be statistically significant? If so,\\nwhich ones?\\n(c) Compute the confusion matrix and overall fraction of correct\\npredictions. Explain what the confusion matrix is telling you\\nabout the types of mistakes made by logistic regression.\\n(d) Now fit the logistic regression model using a training data period\\nfrom 1990 to 2008, with Lag2 as the only predictor. Compute the\\nconfusion matrix and the overall fraction of correct predictions\\nfor the held out data (that is, the data from 2009 and 2010).\\n(e) Repeat (d) using LDA.\\n(f) Repeat (d) using QDA.\\n(g) Repeat (d) using KNN with K = 1.\\n(h) Repeat (d) using naive Bayes.\\n(i) Which of these methods appears to provide the best results on\\nthis data?\\n(j) Experiment with different combinations of predictors, includ-\\ning possible transformations and interactions, for each of the\\nmethods. Report the variables, method, and associated confu-\\nsion matrix that appears to provide the best results on the held\\nout data. Note that you should also experiment with values for\\nK in the KNN classifier.\\n14. In this problem, you will develop a model to predict whether a given\\ncar gets high or low gas mileage based on the Auto data set.\\n(a) Create a binary variable, mpg01, that contains a 1 if mpg contains\\na value above its median, and a 0 if mpg contains a value below\\nits median. You can compute the median using the median()\\nmethod of the data frame. Note you may find it helpful to add\\na column mpg01 to the data frame by assignment. Assuming you\\nhave stored the data frame as Auto, this can be done as follows:\\nAuto[\\'mpg01\\'] = mpg01\\n(b) Explore the data graphically in order to investigate the associ-\\nation between mpg01 and the other features. Which of the other\\nfeatures seem most likely to be useful in predicting mpg01? Scat-\\nterplots and boxplots may be useful tools to answer this ques-\\ntion. Describe your findings.\\n(c) Split the data into a training set and a test set.\\n(d) Perform LDA on the training data in order to predict mpg01\\nusing the variables that seemed most associated with mpg01 in\\n(b). What is the test error of the model obtained? 198\\n4. Classification\\n(e) Perform QDA on the training data in order to predict mpg01\\nusing the variables that seemed most associated with mpg01 in\\n(b). What is the test error of the model obtained?\\n(f) Perform logistic regression on the training data in order to pre-\\ndict mpg01 using the variables that seemed most associated with\\nmpg01 in (b). What is the test error of the model obtained?\\n(g) Perform naive Bayes on the training data in order to predict\\nmpg01 using the variables that seemed most associated with mpg01\\nin (b). What is the test error of the model obtained?\\n(h) Perform KNN on the training data, with several values of K, in\\norder to predict mpg01. Use only the variables that seemed most\\nassociated with mpg01 in (b). What test errors do you obtain?\\nWhich value of K seems to perform the best on this data set?\\n15. This problem involves writing functions.\\n(a) Write a function, Power(), that prints out the result of raising 2\\nto the 3rd power. In other words, your function should compute\\n23 and print out the results.\\nHint: Recall that x**a raises x to the power a. Use the print()\\nfunction to display the result.\\n(b) Create a new function, Power2(), that allows you to pass any\\ntwo numbers, x and a, and prints out the value of x**a. You can\\ndo this by beginning your function with the line\\ndef Power2(x, a):\\nYou should be able to call your function by entering, for instance,\\nPower2(3, 8)\\non the command line. This should output the value of 38, namely,\\n6, 561.\\n(c) Using the Power2() function that you just wrote, compute 103,\\n817, and 1313.\\n(d) Now create a new function, Power3(), that actually returns the\\nresult x**a as a Python object, rather than simply printing it\\nto the screen. That is, if you store the value x**a in an object\\ncalled result within your function, then you can simply return return\\nthis result, using the following line:\\nreturn result\\nNote that the line above should be the last line in your function,\\nand it should be indented 4 spaces.\\n(e) Now using the Power3() function, create a plot of f(x) = x2.\\nThe x-axis should display a range of integers from 1 to 10, and\\nthe y-axis should display x2. Label the axes appropriately, and\\nuse an appropriate title for the figure. Consider displaying either\\nthe x-axis, the y-axis, or both on the log-scale. You can do this\\nby using the ax.set_xscale() and ax.set_yscale() methods of .set_xscale()\\n.set_yscale()\\nthe axes you are plotting to. 4.8 Exercises\\n199\\n(f) Create a function, PlotPower(), that allows you to create a plot\\nof x against x**a for a fixed a and a sequence of values of x. For\\ninstance, if you call\\nPlotPower(np.arange(1, 11), 3)\\nthen a plot should be created with an x-axis taking on values\\n1, 2, . . . , 10, and a y-axis taking on values 13, 23, . . . , 103.\\n16. Using the Boston data set, fit classification models in order to predict\\nwhether a given suburb has a crime rate above or below the median.\\nExplore logistic regression, LDA, naive Bayes, and KNN models using\\nvarious subsets of the predictors. Describe your findings.\\nHint: You will have to create the response variable yourself, using the\\nvariables that are contained in the Boston data set. 5\\nResampling Methods\\nResampling methods are an indispensable tool in modern statistics. They\\ninvolve repeatedly drawing samples from a training set and refitting a model\\nof interest on each sample in order to obtain additional information about\\nthe fitted model. For example, in order to estimate the variability of a linear\\nregression fit, we can repeatedly draw different samples from the training\\ndata, fit a linear regression to each new sample, and then examine the\\nextent to which the resulting fits differ. Such an approach may allow us to\\nobtain information that would not be available from fitting the model only\\nonce using the original training sample.\\nResampling approaches can be computationally expensive, because they\\ninvolve fitting the same statistical method multiple times using different\\nsubsets of the training data. However, due to recent advances in computing\\npower, the computational requirements of resampling methods generally\\nare not prohibitive. In this chapter, we discuss two of the most commonly\\nused resampling methods, cross-validation and the bootstrap. Both methods\\nare important tools in the practical application of many statistical learning\\nprocedures. For example, cross-validation can be used to estimate the test\\nerror associated with a given statistical learning method in order to evaluate\\nits performance, or to select the appropriate level of flexibility. The process\\nof evaluating a model’s performance is known as model assessment, whereas model\\nassessment\\nthe process of selecting the proper level of flexibility for a model is known as\\nmodel selection. The bootstrap is used in several contexts, most commonly model\\nselection\\nto provide a measure of accuracy of a parameter estimate or of a given\\nstatistical learning method.\\n© Springer Nature Switzerland AG 2023 \\nG. James et al., An Introduction to Statistical Learning, Springer Texts in Statistics, \\nhttps://doi.org/10.1007/978-3-031-38747-0_5 \\n201 202\\n5. Resampling Methods\\n5.1\\nCross-Validation\\nIn Chapter 2 we discuss the distinction between the test error rate and the\\ntraining error rate. The test error is the average error that results from using\\na statistical learning method to predict the response on a new observation—\\nthat is, a measurement that was not used in training the method. Given\\na data set, the use of a particular statistical learning method is warranted\\nif it results in a low test error. The test error can be easily calculated if a\\ndesignated test set is available. Unfortunately, this is usually not the case.\\nIn contrast, the training error can be easily calculated by applying the\\nstatistical learning method to the observations used in its training. But as\\nwe saw in Chapter 2, the training error rate often is quite different from the\\ntest error rate, and in particular the former can dramatically underestimate\\nthe latter.\\nIn the absence of a very large designated test set that can be used to\\ndirectly estimate the test error rate, a number of techniques can be used\\nto estimate this quantity using the available training data. Some methods\\nmake a mathematical adjustment to the training error rate in order to\\nestimate the test error rate. Such approaches are discussed in Chapter 6.\\nIn this section, we instead consider a class of methods that estimate the\\ntest error rate by holding out a subset of the training observations from the\\nfitting process, and then applying the statistical learning method to those\\nheld out observations.\\nIn Sections 5.1.1–5.1.4, for simplicity we assume that we are interested\\nin performing regression with a quantitative response. In Section 5.1.5 we\\nconsider the case of classification with a qualitative response. As we will\\nsee, the key concepts remain the same regardless of whether the response\\nis quantitative or qualitative.\\n5.1.1\\nThe Validation Set Approach\\nSuppose that we would like to estimate the test error associated with fit-\\nting a particular statistical learning method on a set of observations. The\\nvalidation set approach, displayed in Figure 5.1, is a very simple strategy validation\\nset approach\\nfor this task. It involves randomly dividing the available set of observa-\\ntions into two parts, a training set and a validation set or hold-out set. The validation\\nset\\nhold-out set\\nmodel is fit on the training set, and the fitted model is used to predict the\\nresponses for the observations in the validation set. The resulting validation\\nset error rate—typically assessed using MSE in the case of a quantitative\\nresponse—provides an estimate of the test error rate.\\nWe illustrate the validation set approach on the Auto data set. Recall from\\nChapter 3 that there appears to be a non-linear relationship between mpg\\nand horsepower, and that a model that predicts mpg using horsepower and\\nhorsepower2 gives better results than a model that uses only a linear term.\\nIt is natural to wonder whether a cubic or higher-order fit might provide\\neven better results. We answer this question in Chapter 3 by looking at\\nthe p-values associated with a cubic term and higher-order polynomial\\nterms in a linear regression. But we could also answer this question using\\nthe validation method. We randomly split the 392 observations into two 5.1 Cross-Validation\\n203\\n!\"#\"$\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"%\"\\n&\"\"##\"\"!$\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\\'!\"\\nFIGURE 5.1. A schematic display of the validation set approach. A set of n\\nobservations are randomly split into a training set (shown in blue, containing\\nobservations 7, 22, and 13, among others) and a validation set (shown in beige,\\nand containing observation 91, among others). The statistical learning method is\\nfit on the training set, and its performance is evaluated on the validation set.\\nsets, a training set containing 196 of the data points, and a validation set\\ncontaining the remaining 196 observations. The validation set error rates\\nthat result from fitting various regression models on the training sample\\nand evaluating their performance on the validation sample, using MSE\\nas a measure of validation set error, are shown in the left-hand panel of\\nFigure 5.2. The validation set MSE for the quadratic fit is considerably\\nsmaller than for the linear fit. However, the validation set MSE for the cubic\\nfit is actually slightly larger than for the quadratic fit. This implies that\\nincluding a cubic term in the regression does not lead to better prediction\\nthan simply using a quadratic term.\\nRecall that in order to create the left-hand panel of Figure 5.2, we ran-\\ndomly divided the data set into two parts, a training set and a validation\\nset. If we repeat the process of randomly splitting the sample set into two\\nparts, we will get a somewhat different estimate for the test MSE. As an\\nillustration, the right-hand panel of Figure 5.2 displays ten different vali-\\ndation set MSE curves from the Auto data set, produced using ten different\\nrandom splits of the observations into training and validation sets. All ten\\ncurves indicate that the model with a quadratic term has a dramatically\\nsmaller validation set MSE than the model with only a linear term. Fur-\\nthermore, all ten curves indicate that there is not much benefit in including\\ncubic or higher-order polynomial terms in the model. But it is worth noting\\nthat each of the ten curves results in a different test MSE estimate for each\\nof the ten regression models considered. And there is no consensus among\\nthe curves as to which model results in the smallest validation set MSE.\\nBased on the variability among these curves, all that we can conclude with\\nany confidence is that the linear fit is not adequate for this data.\\nThe validation set approach is conceptually simple and is easy to imple-\\nment. But it has two potential drawbacks:\\n1. As is shown in the right-hand panel of Figure 5.2, the validation esti-\\nmate of the test error rate can be highly variable, depending on pre-\\ncisely which observations are included in the training set and which\\nobservations are included in the validation set.\\n2. In the validation approach, only a subset of the observations—those\\nthat are included in the training set rather than in the validation\\nset—are used to fit the model. Since statistical methods tend to per-\\nform worse when trained on fewer observations, this suggests that the 204\\n5. Resampling Methods\\n2\\n4\\n6\\n8\\n10\\n16\\n18\\n20\\n22\\n24\\n26\\n28\\nDegree of Polynomial\\nMean Squared Error\\n2\\n4\\n6\\n8\\n10\\n16\\n18\\n20\\n22\\n24\\n26\\n28\\nDegree of Polynomial\\nMean Squared Error\\nFIGURE 5.2.\\nThe validation set approach was used on the Auto data set in\\norder to estimate the test error that results from predicting mpg using polynomial\\nfunctions of horsepower. Left: Validation error estimates for a single split into\\ntraining and validation data sets. Right: The validation method was repeated ten\\ntimes, each time using a different random split of the observations into a training\\nset and a validation set. This illustrates the variability in the estimated test MSE\\nthat results from this approach.\\nvalidation set error rate may tend to overestimate the test error rate\\nfor the model fit on the entire data set.\\nIn the coming subsections, we will present cross-validation, a refinement of\\nthe validation set approach that addresses these two issues.\\n5.1.2\\nLeave-One-Out Cross-Validation\\nLeave-one-out cross-validation (LOOCV) is closely related to the validation leave-one-\\nout\\ncross-\\nvalidation\\nset approach of Section 5.1.1, but it attempts to address that method’s\\ndrawbacks.\\nLike the validation set approach, LOOCV involves splitting the set of\\nobservations into two parts. However, instead of creating two subsets of\\ncomparable size, a single observation (x1, y1) is used for the validation\\nset, and the remaining observations {(x2, y2), . . . , (xn, yn)} make up the\\ntraining set. The statistical learning method is fit on the n −1 training\\nobservations, and a prediction ˆy1 is made for the excluded observation,\\nusing its value x1. Since (x1, y1) was not used in the fitting process, MSE1 =\\n(y1 −ˆy1)2 provides an approximately unbiased estimate for the test error.\\nBut even though MSE1 is unbiased for the test error, it is a poor estimate\\nbecause it is highly variable, since it is based upon a single observation\\n(x1, y1).\\nWe can repeat the procedure by selecting (x2, y2) for the validation\\ndata, training the statistical learning procedure on the n −1 observations\\n{(x1, y1), (x3, y3), . . . , (xn, yn)}, and computing MSE2 = (y2−ˆy2)2. Repeat-\\ning this approach n times produces n squared errors, MSE1, . . . , MSEn.\\nThe LOOCV estimate for the test MSE is the average of these n test error\\nestimates:\\nCV(n) = 1\\nn\\nn\\n0\\ni=1\\nMSEi.\\n(5.1) 5.1 Cross-Validation\\n205\\n!\"#\"$\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"%\"\\n!\"#\"$\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"%\"\\n!\"#\"$\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"%\"\\n!\"#\"$\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"%\"\\n!\"#\"$\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"%\"\\n&\"\\n&\"\\n&\"\\nFIGURE 5.3. A schematic display of LOOCV. A set of n data points is repeat-\\nedly split into a training set (shown in blue) containing all but one observation,\\nand a validation set that contains only that observation (shown in beige). The test\\nerror is then estimated by averaging the n resulting MSEs. The first training set\\ncontains all but observation 1, the second training set contains all but observation\\n2, and so forth.\\nA schematic of the LOOCV approach is illustrated in Figure 5.3.\\nLOOCV has a couple of major advantages over the validation set ap-\\nproach. First, it has far less bias. In LOOCV, we repeatedly fit the sta-\\ntistical learning method using training sets that contain n −1 observa-\\ntions, almost as many as are in the entire data set. This is in contrast to\\nthe validation set approach, in which the training set is typically around\\nhalf the size of the original data set. Consequently, the LOOCV approach\\ntends not to overestimate the test error rate as much as the validation\\nset approach does. Second, in contrast to the validation approach which\\nwill yield different results when applied repeatedly due to randomness in\\nthe training/validation set splits, performing LOOCV multiple times will\\nalways yield the same results: there is no randomness in the training/vali-\\ndation set splits.\\nWe used LOOCV on the Auto data set in order to obtain an estimate\\nof the test set MSE that results from fitting a linear regression model to\\npredict mpg using polynomial functions of horsepower. The results are shown\\nin the left-hand panel of Figure 5.4.\\nLOOCV has the potential to be expensive to implement, since the model\\nhas to be fit n times. This can be very time consuming if n is large, and if\\neach individual model is slow to fit. With least squares linear or polynomial\\nregression, an amazing shortcut makes the cost of LOOCV the same as that\\nof a single model fit! The following formula holds:\\nCV(n) = 1\\nn\\nn\\n0\\ni=1\\n*yi −ˆyi\\n1 −hi\\n+2\\n,\\n(5.2) 206\\n5. Resampling Methods\\n2\\n4\\n6\\n8\\n10\\n16\\n18\\n20\\n22\\n24\\n26\\n28\\nLOOCV\\nDegree of Polynomial\\nMean Squared Error\\n2\\n4\\n6\\n8\\n10\\n16\\n18\\n20\\n22\\n24\\n26\\n28\\n10−fold CV\\nDegree of Polynomial\\nMean Squared Error\\nFIGURE 5.4.\\nCross-validation was used on the Auto data set in order to\\nestimate the test error that results from predicting mpg using polynomial functions\\nof horsepower. Left: The LOOCV error curve. Right: 10-fold CV was run nine\\nseparate times, each with a different random split of the data into ten parts. The\\nfigure shows the nine slightly different CV error curves.\\nwhere ˆyi is the ith fitted value from the original least squares fit, and hi is\\nthe leverage defined in (3.37) on page 105.1 This is like the ordinary MSE,\\nexcept the ith residual is divided by 1 −hi. The leverage lies between 1/n\\nand 1, and reflects the amount that an observation influences its own fit.\\nHence the residuals for high-leverage points are inflated in this formula by\\nexactly the right amount for this equality to hold.\\nLOOCV is a very general method, and can be used with any kind of\\npredictive modeling. For example we could use it with logistic regression\\nor linear discriminant analysis, or any of the methods discussed in later\\nchapters. The magic formula (5.2) does not hold in general, in which case\\nthe model has to be refit n times.\\n5.1.3\\nk-Fold Cross-Validation\\nAn alternative to LOOCV is k-fold CV. This approach involves randomly k-fold CV\\ndividing the set of observations into k groups, or folds, of approximately\\nequal size. The first fold is treated as a validation set, and the method\\nis fit on the remaining k −1 folds. The mean squared error, MSE1, is\\nthen computed on the observations in the held-out fold. This procedure is\\nrepeated k times; each time, a different group of observations is treated\\nas a validation set. This process results in k estimates of the test error,\\nMSE1, MSE2, . . . , MSEk. The k-fold CV estimate is computed by averaging\\nthese values,\\nCV(k) = 1\\nk\\nk\\n0\\ni=1\\nMSEi.\\n(5.3)\\nFigure 5.5 illustrates the k-fold CV approach.\\n1In the case of multiple linear regression, the leverage takes a slightly more compli-\\ncated form than (3.37), but (5.2) still holds. 5.1 Cross-Validation\\n207\\n!\"#\"$\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"%\"\\n!!\"&\\'\"(\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\")&\"\\n!!\"&\\'\"(\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\")&\"\\n!!\"&\\'\"(\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\")&\"\\n!!\"&\\'\"(\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\")&\"\\n!!\"&\\'\"(\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\")&\"\\n!!\"&\\'\"(\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\")&\"\\nFIGURE 5.5.\\nA schematic display of 5-fold CV. A set of n observations is\\nrandomly split into five non-overlapping groups. Each of these fifths acts as a\\nvalidation set (shown in beige), and the remainder as a training set (shown in\\nblue). The test error is estimated by averaging the five resulting MSE estimates.\\nIt is not hard to see that LOOCV is a special case of k-fold CV in which k\\nis set to equal n. In practice, one typically performs k-fold CV using k = 5\\nor k = 10. What is the advantage of using k = 5 or k = 10 rather than\\nk = n? The most obvious advantage is computational. LOOCV requires\\nfitting the statistical learning method n times. This has the potential to be\\ncomputationally expensive (except for linear models fit by least squares,\\nin which case formula (5.2) can be used). But cross-validation is a very\\ngeneral approach that can be applied to almost any statistical learning\\nmethod. Some statistical learning methods have computationally intensive\\nfitting procedures, and so performing LOOCV may pose computational\\nproblems, especially if n is extremely large. In contrast, performing 10-fold\\nCV requires fitting the learning procedure only ten times, which may be\\nmuch more feasible. As we see in Section 5.1.4, there also can be other\\nnon-computational advantages to performing 5-fold or 10-fold CV, which\\ninvolve the bias-variance trade-off.\\nThe right-hand panel of Figure 5.4 displays nine different 10-fold CV\\nestimates for the Auto data set, each resulting from a different random split\\nof the observations into ten folds. As we can see from the figure, there is\\nsome variability in the CV estimates as a result of the variability in how\\nthe observations are divided into ten folds. But this variability is typically\\nmuch lower than the variability in the test error estimates that results from\\nthe validation set approach (right-hand panel of Figure 5.2).\\nWhen we examine real data, we do not know the true test MSE, and\\nso it is difficult to determine the accuracy of the cross-validation estimate.\\nHowever, if we examine simulated data, then we can compute the true\\ntest MSE, and can thereby evaluate the accuracy of our cross-validation\\nresults. In Figure 5.6, we plot the cross-validation estimates and true test\\nerror rates that result from applying smoothing splines to the simulated\\ndata sets illustrated in Figures 2.9–2.11 of Chapter 2. The true test MSE\\nis displayed in blue. The black dashed and orange solid lines respectively\\nshow the estimated LOOCV and 10-fold CV estimates. In all three plots,\\nthe two cross-validation estimates are very similar. In the right-hand panel 208\\n5. Resampling Methods\\n2\\n5\\n10\\n20\\n0.0\\n0.5\\n1.0\\n1.5\\n2.0\\n2.5\\n3.0\\nFlexibility\\nMean Squared Error\\n2\\n5\\n10\\n20\\n0.0\\n0.5\\n1.0\\n1.5\\n2.0\\n2.5\\n3.0\\nFlexibility\\nMean Squared Error\\n2\\n5\\n10\\n20\\n0\\n5\\n10\\n15\\n20\\nFlexibility\\nMean Squared Error\\nFIGURE 5.6.\\nTrue and estimated test MSE for the simulated data sets in\\nFigures 2.9 (left), 2.10 (center), and 2.11 (right). The true test MSE is shown\\nin blue, the LOOCV estimate is shown as a black dashed line, and the 10-fold\\nCV estimate is shown in orange. The crosses indicate the minimum of each of\\nthe MSE curves.\\nof Figure 5.6, the true test MSE and the cross-validation curves are almost\\nidentical. In the center panel of Figure 5.6, the two sets of curves are similar\\nat the lower degrees of flexibility, while the CV curves overestimate the test\\nset MSE for higher degrees of flexibility. In the left-hand panel of Figure 5.6,\\nthe CV curves have the correct general shape, but they underestimate the\\ntrue test MSE.\\nWhen we perform cross-validation, our goal might be to determine how\\nwell a given statistical learning procedure can be expected to perform on\\nindependent data; in this case, the actual estimate of the test MSE is\\nof interest. But at other times we are interested only in the location of\\nthe minimum point in the estimated test MSE curve. This is because we\\nmight be performing cross-validation on a number of statistical learning\\nmethods, or on a single method using different levels of flexibility, in order\\nto identify the method that results in the lowest test error. For this purpose,\\nthe location of the minimum point in the estimated test MSE curve is\\nimportant, but the actual value of the estimated test MSE is not. We find\\nin Figure 5.6 that despite the fact that they sometimes underestimate the\\ntrue test MSE, all of the CV curves come close to identifying the correct\\nlevel of flexibility—that is, the flexibility level corresponding to the smallest\\ntest MSE.\\n5.1.4\\nBias-Variance Trade-Off for k-Fold Cross-Validation\\nWe mentioned in Section 5.1.3 that k-fold CV with k < n has a compu-\\ntational advantage to LOOCV. But putting computational issues aside,\\na less obvious but potentially more important advantage of k-fold CV is\\nthat it often gives more accurate estimates of the test error rate than does\\nLOOCV. This has to do with a bias-variance trade-off.\\nIt was mentioned in Section 5.1.1 that the validation set approach can\\nlead to overestimates of the test error rate, since in this approach the\\ntraining set used to fit the statistical learning method contains only half\\nthe observations of the entire data set. Using this logic, it is not hard to see 5.1 Cross-Validation\\n209\\nthat LOOCV will give approximately unbiased estimates of the test error,\\nsince each training set contains n−1 observations, which is almost as many\\nas the number of observations in the full data set. And performing k-fold\\nCV for, say, k = 5 or k = 10 will lead to an intermediate level of bias,\\nsince each training set contains approximately (k −1)n/k observations—\\nfewer than in the LOOCV approach, but substantially more than in the\\nvalidation set approach. Therefore, from the perspective of bias reduction,\\nit is clear that LOOCV is to be preferred to k-fold CV.\\nHowever, we know that bias is not the only source for concern in an esti-\\nmating procedure; we must also consider the procedure’s variance. It turns\\nout that LOOCV has higher variance than does k-fold CV with k < n. Why\\nis this the case? When we perform LOOCV, we are in effect averaging the\\noutputs of n fitted models, each of which is trained on an almost identical\\nset of observations; therefore, these outputs are highly (positively) corre-\\nlated with each other. In contrast, when we perform k-fold CV with k < n,\\nwe are averaging the outputs of k fitted models that are somewhat less\\ncorrelated with each other, since the overlap between the training sets in\\neach model is smaller. Since the mean of many highly correlated quantities\\nhas higher variance than does the mean of many quantities that are not\\nas highly correlated, the test error estimate resulting from LOOCV tends\\nto have higher variance than does the test error estimate resulting from\\nk-fold CV.\\nTo summarize, there is a bias-variance trade-off associated with the\\nchoice of k in k-fold cross-validation. Typically, given these considerations,\\none performs k-fold cross-validation using k = 5 or k = 10, as these values\\nhave been shown empirically to yield test error rate estimates that suffer\\nneither from excessively high bias nor from very high variance.\\n5.1.5\\nCross-Validation on Classification Problems\\nIn this chapter so far, we have illustrated the use of cross-validation in the\\nregression setting where the outcome Y is quantitative, and so have used\\nMSE to quantify test error. But cross-validation can also be a very useful\\napproach in the classification setting when Y is qualitative. In this setting,\\ncross-validation works just as described earlier in this chapter, except that\\nrather than using MSE to quantify test error, we instead use the number\\nof misclassified observations. For instance, in the classification setting, the\\nLOOCV error rate takes the form\\nCV(n) = 1\\nn\\nn\\n0\\ni=1\\nErri,\\n(5.4)\\nwhere Erri = I(yi ̸= ˆyi). The k-fold CV error rate and validation set error\\nrates are defined analogously.\\nAs an example, we fit various logistic regression models on the two-\\ndimensional classification data displayed in Figure 2.13. In the top-left\\npanel of Figure 5.7, the black solid line shows the estimated decision bound-\\nary resulting from fitting a standard logistic regression model to this data\\nset. Since this is simulated data, we can compute the true test error rate,\\nwhich takes a value of 0.201 and so is substantially larger than the Bayes 210\\n5. Resampling Methods\\nDegree=1\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\noo\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\noo\\no\\no\\no o\\no\\no\\no o\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\noo o\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no o\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\nDegree=2\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\noo\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\noo\\no\\no\\no o\\no\\no\\no o\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\noo o\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no o\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\nDegree=3\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\noo\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\noo\\no\\no\\no o\\no\\no\\no o\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\noo o\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no o\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\nDegree=4\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\noo\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\noo\\no\\no\\no o\\no\\no\\no o\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\noo o\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no o\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\no\\nFIGURE 5.7. Logistic regression fits on the two-dimensional classification data\\ndisplayed in Figure 2.13. The Bayes decision boundary is represented using a\\npurple dashed line. Estimated decision boundaries from linear, quadratic, cubic\\nand quartic (degrees 1–4) logistic regressions are displayed in black. The test error\\nrates for the four logistic regression fits are respectively 0.201, 0.197, 0.160, and\\n0.162, while the Bayes error rate is 0.133.\\nerror rate of 0.133. Clearly logistic regression does not have enough flexi-\\nbility to model the Bayes decision boundary in this setting. We can easily\\nextend logistic regression to obtain a non-linear decision boundary by using\\npolynomial functions of the predictors, as we did in the regression setting in\\nSection 3.3.2. For example, we can fit a quadratic logistic regression model,\\ngiven by\\nlog\\n*\\np\\n1 −p\\n+\\n= β0 + β1X1 + β2X2\\n1 + β3X2 + β4X2\\n2.\\n(5.5)\\nThe top-right panel of Figure 5.7 displays the resulting decision boundary,\\nwhich is now curved. However, the test error rate has improved only slightly,\\nto 0.197. A much larger improvement is apparent in the bottom-left panel 5.1 Cross-Validation\\n211\\n2\\n4\\n6\\n8\\n10\\n0.12\\n0.14\\n0.16\\n0.18\\n0.20\\nOrder of Polynomials Used\\nError Rate\\n0.01\\n0.02\\n0.05\\n0.10\\n0.20\\n0.50\\n1.00\\n0.12\\n0.14\\n0.16\\n0.18\\n0.20\\n1/K\\nError Rate\\nFIGURE 5.8. Test error (brown), training error (blue), and 10-fold CV error\\n(black) on the two-dimensional classification data displayed in Figure 5.7. Left:\\nLogistic regression using polynomial functions of the predictors. The order of\\nthe polynomials used is displayed on the x-axis. Right: The KNN classifier with\\ndifferent values of K, the number of neighbors used in the KNN classifier.\\nof Figure 5.7, in which we have fit a logistic regression model involving\\ncubic polynomials of the predictors. Now the test error rate has decreased\\nto 0.160. Going to a quartic polynomial (bottom-right) slightly increases\\nthe test error.\\nIn practice, for real data, the Bayes decision boundary and the test er-\\nror rates are unknown. So how might we decide between the four logistic\\nregression models displayed in Figure 5.7? We can use cross-validation in\\norder to make this decision. The left-hand panel of Figure 5.8 displays in\\nblack the 10-fold CV error rates that result from fitting ten logistic regres-\\nsion models to the data, using polynomial functions of the predictors up\\nto tenth order. The true test errors are shown in brown, and the training\\nerrors are shown in blue. As we have seen previously, the training error\\ntends to decrease as the flexibility of the fit increases. (The figure indicates\\nthat though the training error rate doesn’t quite decrease monotonically,\\nit tends to decrease on the whole as the model complexity increases.) In\\ncontrast, the test error displays a characteristic U-shape. The 10-fold CV\\nerror rate provides a pretty good approximation to the test error rate.\\nWhile it somewhat underestimates the error rate, it reaches a minimum\\nwhen fourth-order polynomials are used, which is very close to the min-\\nimum of the test curve, which occurs when third-order polynomials are\\nused. In fact, using fourth-order polynomials would likely lead to good test\\nset performance, as the true test error rate is approximately the same for\\nthird, fourth, fifth, and sixth-order polynomials.\\nThe right-hand panel of Figure 5.8 displays the same three curves us-\\ning the KNN approach for classification, as a function of the value of K\\n(which in this context indicates the number of neighbors used in the KNN\\nclassifier, rather than the number of CV folds used). Again the training\\nerror rate declines as the method becomes more flexible, and so we see that\\nthe training error rate cannot be used to select the optimal value for K.\\nThough the cross-validation error curve slightly underestimates the test\\nerror rate, it takes on a minimum very close to the best value for K. 212\\n5. Resampling Methods\\n5.2\\nThe Bootstrap\\nThe bootstrap is a widely applicable and extremely powerful statistical tool bootstrap\\nthat can be used to quantify the uncertainty associated with a given esti-\\nmator or statistical learning method. As a simple example, the bootstrap\\ncan be used to estimate the standard errors of the coefficients from a linear\\nregression fit. In the specific case of linear regression, this is not particularly\\nuseful, since we saw in Chapter 3 that standard statistical software such as\\nR outputs such standard errors automatically. However, the power of the\\nbootstrap lies in the fact that it can be easily applied to a wide range of\\nstatistical learning methods, including some for which a measure of vari-\\nability is otherwise difficult to obtain and is not automatically output by\\nstatistical software.\\nIn this section we illustrate the bootstrap on a toy example in which we\\nwish to determine the best investment allocation under a simple model.\\nIn Section 5.3 we explore the use of the bootstrap to assess the variability\\nassociated with the regression coefficients in a linear model fit.\\nSuppose that we wish to invest a fixed sum of money in two financial\\nassets that yield returns of X and Y , respectively, where X and Y are\\nrandom quantities. We will invest a fraction α of our money in X, and will\\ninvest the remaining 1 −α in Y . Since there is variability associated with\\nthe returns on these two assets, we wish to choose α to minimize the total\\nrisk, or variance, of our investment. In other words, we want to minimize\\nVar(αX + (1 −α)Y ). One can show that the value that minimizes the risk\\nis given by\\nα =\\nσ2\\nY −σXY\\nσ2\\nX + σ2\\nY −2σXY\\n,\\n(5.6)\\nwhere σ2\\nX = Var(X), σ2\\nY = Var(Y ), and σXY = Cov(X, Y ).\\nIn reality, the quantities σ2\\nX, σ2\\nY , and σXY are unknown. We can compute\\nestimates for these quantities, ˆσ2\\nX, ˆσ2\\nY , and ˆσXY , using a data set that\\ncontains past measurements for X and Y . We can then estimate the value\\nof α that minimizes the variance of our investment using\\nˆα =\\nˆσ2\\nY −ˆσXY\\nˆσ2\\nX + ˆσ2\\nY −2ˆσXY\\n.\\n(5.7)\\nFigure 5.9 illustrates this approach for estimating α on a simulated data\\nset. In each panel, we simulated 100 pairs of returns for the investments\\nX and Y . We used these returns to estimate σ2\\nX, σ2\\nY , and σXY , which we\\nthen substituted into (5.7) in order to obtain estimates for α. The value of\\nˆα resulting from each simulated data set ranges from 0.532 to 0.657.\\nIt is natural to wish to quantify the accuracy of our estimate of α. To\\nestimate the standard deviation of ˆα, we repeated the process of simu-\\nlating 100 paired observations of X and Y , and estimating α using (5.7),\\n1,000 times. We thereby obtained 1,000 estimates for α, which we can call\\nˆα1, ˆα2, . . . , ˆα1,000. The left-hand panel of Figure 5.10 displays a histogram\\nof the resulting estimates. For these simulations the parameters were set to\\nσ2\\nX = 1, σ2\\nY = 1.25, and σXY = 0.5, and so we know that the true value of\\nα is 0.6. We indicated this value using a solid vertical line on the histogram. 5.2 The Bootstrap\\n213\\n−2\\n−1\\n0\\n1\\n2\\n−2\\n−1\\n0\\n1\\n2\\nX\\nY\\n−2\\n−1\\n0\\n1\\n2\\n−2\\n−1\\n0\\n1\\n2\\nX\\nY\\n−3\\n−2\\n−1\\n0\\n1\\n2\\n−3\\n−2\\n−1\\n0\\n1\\n2\\nX\\nY\\n−2\\n−1\\n0\\n1\\n2\\n3\\n−3\\n−2\\n−1\\n0\\n1\\n2\\nX\\nY\\nFIGURE 5.9.\\nEach panel displays 100 simulated returns for investments\\nX and Y . From left to right and top to bottom, the resulting estimates for α\\nare 0.576, 0.532, 0.657, and 0.651.\\nThe mean over all 1,000 estimates for α is\\n¯α =\\n1\\n1000\\n1000\\n0\\nr=1\\nˆαr = 0.5996,\\nvery close to α = 0.6, and the standard deviation of the estimates is\\n:\\n;\\n;\\n<\\n1\\n1000 −1\\n1000\\n0\\nr=1\\n(ˆαr −¯α)2 = 0.083.\\nThis gives us a very good idea of the accuracy of ˆα: SE(ˆα) ≈0.083. So\\nroughly speaking, for a random sample from the population, we would\\nexpect ˆα to differ from α by approximately 0.08, on average.\\nIn practice, however, the procedure for estimating SE(ˆα) outlined above\\ncannot be applied, because for real data we cannot generate new samples\\nfrom the original population. However, the bootstrap approach allows us\\nto use a computer to emulate the process of obtaining new sample sets,\\nso that we can estimate the variability of ˆα without generating additional\\nsamples. Rather than repeatedly obtaining independent data sets from the\\npopulation, we instead obtain distinct data sets by repeatedly sampling\\nobservations from the original data set.\\nThis approach is illustrated in Figure 5.11 on a simple data set, which\\nwe call Z, that contains only n = 3 observations. We randomly select n\\nobservations from the data set in order to produce a bootstrap data set, 214\\n5. Resampling Methods\\n0.4\\n0.5\\n0.6\\n0.7\\n0.8\\n0.9\\n0\\n50\\n100\\n150\\n200\\n0.3\\n0.4\\n0.5\\n0.6\\n0.7\\n0.8\\n0.9\\n0\\n50\\n100\\n150\\n200\\nTrue\\nBootstrap\\n0.3\\n0.4\\n0.5\\n0.6\\n0.7\\n0.8\\n0.9\\nα\\nα\\nα\\nFIGURE 5.10. Left: A histogram of the estimates of α obtained by generating\\n1,000 simulated data sets from the true population. Center: A histogram of the\\nestimates of α obtained from 1,000 bootstrap samples from a single data set.\\nRight: The estimates of α displayed in the left and center panels are shown as\\nboxplots. In each panel, the pink line indicates the true value of α.\\nZ∗1. The sampling is performed with replacement, which means that the with\\nreplacement\\nsame observation can occur more than once in the bootstrap data set. In\\nthis example, Z∗1 contains the third observation twice, the first observation\\nonce, and no instances of the second observation. Note that if an observation\\nis contained in Z∗1, then both its X and Y values are included. We can use\\nZ∗1 to produce a new bootstrap estimate for α, which we call ˆα∗1. This\\nprocedure is repeated B times for some large value of B, in order to produce\\nB different bootstrap data sets, Z∗1, Z∗2, . . . , Z∗B, and B corresponding α\\nestimates, ˆα∗1, ˆα∗2, . . . , ˆα∗B. We can compute the standard error of these\\nbootstrap estimates using the formula\\nSEB(ˆα) =\\n:\\n;\\n;\\n<\\n1\\nB −1\\nB\\n0\\nr=1\\n>\\nˆα∗r −1\\nB\\nB\\n0\\nr′=1\\nˆα∗r′\\n?2\\n.\\n(5.8)\\nThis serves as an estimate of the standard error of ˆα estimated from the\\noriginal data set.\\nThe bootstrap approach is illustrated in the center panel of Figure 5.10,\\nwhich displays a histogram of 1,000 bootstrap estimates of α, each com-\\nputed using a distinct bootstrap data set. This panel was constructed on\\nthe basis of a single data set, and hence could be created using real data.\\nNote that the histogram looks very similar to the left-hand panel, which\\ndisplays the idealized histogram of the estimates of α obtained by generat-\\ning 1,000 simulated data sets from the true population. In particular the\\nbootstrap estimate SE(ˆα) from (5.8) is 0.087, very close to the estimate of\\n0.083 obtained using 1,000 simulated data sets. The right-hand panel dis-\\nplays the information in the center and left panels in a different way, via\\nboxplots of the estimates for α obtained by generating 1,000 simulated data\\nsets from the true population and using the bootstrap approach. Again, the\\nboxplots have similar spreads, indicating that the bootstrap approach can\\nbe used to effectively estimate the variability associated with ˆα. 5.3 Lab: Cross-Validation and the Bootstrap\\n215\\n2.8 \\n5.3 \\n3 \\n1.1 \\n2.1 \\n2 \\n2.4 \\n4.3 \\n1 \\nY \\nX \\nObs \\n2.8 \\n5.3 \\n3 \\n2.4 \\n4.3 \\n1 \\n2.8 \\n5.3 \\n3 \\nY \\nX \\nObs \\n2.4 \\n4.3 \\n1 \\n2.8 \\n5.3 \\n3 \\n1.1 \\n2.1 \\n2 \\nY \\nX \\nObs \\n2.4 \\n4.3 \\n1 \\n1.1 \\n2.1 \\n2 \\n1.1 \\n2.1 \\n2 \\nY \\nX \\nObs \\nOriginal Data (Z) \\n1\\n*\\nZ\\n2\\n*\\nZ\\nZ *B\\n1\\n*ˆ\\n2\\n*ˆ\\nˆ*B\\n!!\\n!!\\n!!\\n!!\\n!\\n!!\\n!!\\n!!\\n!!\\n!!\\n!!\\n!!\\n!!\\nFIGURE 5.11.\\nA graphical illustration of the bootstrap approach on a small\\nsample containing n = 3 observations. Each bootstrap data set contains n obser-\\nvations, sampled with replacement from the original data set. Each bootstrap data\\nset is used to obtain an estimate of α.\\n5.3\\nLab: Cross-Validation and the Bootstrap\\nIn this lab, we explore the resampling techniques covered in this chap-\\nter. Some of the commands in this lab may take a while to run on your\\ncomputer.\\nWe again begin by placing most of our imports at this top level.\\nIn [1]: import numpy as np\\nimport statsmodels.api as sm\\nfrom ISLP import load_data\\nfrom ISLP.models import (ModelSpec as MS,\\nsummarize,\\npoly)\\nfrom sklearn.model_selection import train_test_split\\nThere are several new imports needed for this lab.\\nIn [2]: from functools import partial\\nfrom sklearn.model_selection import \\\\\\n(cross_validate ,\\nKFold,\\nShuffleSplit)\\nfrom sklearn.base import clone\\nfrom ISLP.models import sklearn_sm\\n\\x01\\n\\x01\\n\\x01 216\\n5. Resampling Methods\\n5.3.1\\nThe Validation Set Approach\\nWe explore the use of the validation set approach in order to estimate the\\ntest error rates that result from fitting various linear models on the Auto\\ndata set.\\nWe use the function train_test_split() to split the data into training train_test_\\nsplit()\\nand validation sets. As there are 392 observations, we split into two equal\\nsets of size 196 using the argument test_size=196. It is generally a good\\nidea to set a random seed when performing operations like this that contain\\nan element of randomness, so that the results obtained can be reproduced\\nprecisely at a later time. We set the random seed of the splitter with the\\nargument random_state=0.\\nIn [3]: Auto = load_data(\\'Auto\\')\\nAuto_train, Auto_valid = train_test_split(Auto,\\ntest_size=196,\\nrandom_state=0)\\nNow we can fit a linear regression using only the observations corre-\\nsponding to the training set Auto_train.\\nIn [4]: hp_mm = MS([\\'horsepower\\'])\\nX_train = hp_mm.fit_transform(Auto_train)\\ny_train = Auto_train[\\'mpg\\']\\nmodel = sm.OLS(y_train, X_train)\\nresults = model.fit()\\nWe now use the predict() method of results evaluated on the model ma-\\ntrix for this model created using the validation data set. We also calculate\\nthe validation MSE of our model.\\nIn [5]: X_valid = hp_mm.transform(Auto_valid)\\ny_valid = Auto_valid[\\'mpg\\']\\nvalid_pred = results.predict(X_valid)\\nnp.mean((y_valid - valid_pred)**2)\\nOut[5]: 23.6166\\nHence our estimate for the validation MSE of the linear regression fit is\\n23.62.\\nWe can also estimate the validation error for higher-degree polynomial\\nregressions. We first provide a function evalMSE() that takes a model string\\nas well as a training and test set and returns the MSE on the test set.\\nIn [6]: def evalMSE(terms,\\nresponse,\\ntrain,\\ntest):\\nmm = MS(terms)\\nX_train = mm.fit_transform(train)\\ny_train = train[response]\\nX_test = mm.transform(test)\\ny_test = test[response] 5.3 Lab: Cross-Validation and the Bootstrap\\n217\\nresults = sm.OLS(y_train, X_train).fit()\\ntest_pred = results.predict(X_test)\\nreturn np.mean((y_test - test_pred)**2)\\nLet’s use this function to estimate the validation MSE using linear,\\nquadratic and cubic fits. We use the enumerate() function here, which gives enumerate()\\nboth the values and indices of objects as one iterates over a for loop.\\nIn [7]: MSE = np.zeros(3)\\nfor idx, degree in enumerate(range(1, 4)):\\nMSE[idx] = evalMSE([poly(\\'horsepower\\', degree)],\\n\\'mpg\\',\\nAuto_train ,\\nAuto_valid)\\nMSE\\nOut[7]: array([23.62, 18.76, 18.80])\\nThese error rates are 23.62, 18.76, and 18.80, respectively. If we choose a\\ndifferent training/validation split instead, then we can expect somewhat\\ndifferent errors on the validation set.\\nIn [8]: Auto_train, Auto_valid = train_test_split(Auto,\\ntest_size=196,\\nrandom_state=3)\\nMSE = np.zeros(3)\\nfor idx, degree in enumerate(range(1, 4)):\\nMSE[idx] = evalMSE([poly(\\'horsepower\\', degree)],\\n\\'mpg\\',\\nAuto_train ,\\nAuto_valid)\\nMSE\\nOut[8]: array([20.76, 16.95, 16.97])\\nUsing this split of the observations into a training set and a validation\\nset, we find that the validation set error rates for the models with linear,\\nquadratic, and cubic terms are 20.76, 16.95, and 16.97, respectively.\\nThese results are consistent with our previous findings: a model that\\npredicts mpg using a quadratic function of horsepower performs better than\\na model that involves only a linear function of horsepower, and there is no\\nevidence of an improvement in using a cubic function of horsepower.\\n5.3.2\\nCross-Validation\\nIn theory, the cross-validation estimate can be computed for any general-\\nized linear model. In practice, however, the simplest way to cross-validate\\nin Python is to use sklearn, which has a different interface or API than\\nstatsmodels, the code we have been using to fit GLMs.\\nThis is a problem which often confronts data scientists: “I have a function\\nto do task A, and need to feed it into something that performs task B, so\\nthat I can compute B(A(D)), where D is my data.” When A and B don’t\\nnaturally speak to each other, this requires the use of a wrapper. In the ISLP wrapper 218\\n5. Resampling Methods\\npackage, we provide a wrapper, sklearn_sm(), that enables us to easily use sklearn_sm()\\nthe cross-validation tools of sklearn with models fit by statsmodels.\\nThe class sklearn_sm() has as its first argument a model from statsmodels.\\nIt can take two additional optional arguments: model_str which can be used\\nto specify a formula, and model_args which should be a dictionary of addi-\\ntional arguments used when fitting the model. For example, to fit a logistic\\nregression model we have to specify a family argument. This is passed as\\nmodel_args={\\'family\\':sm.families.Binomial()}.\\nHere is our wrapper in action:\\nIn [9]: hp_model = sklearn_sm(sm.OLS,\\nMS([\\'horsepower\\']))\\nX, Y = Auto.drop(columns=[\\'mpg\\']), Auto[\\'mpg\\']\\ncv_results = cross_validate(hp_model,\\nX,\\nY,\\ncv=Auto.shape[0])\\ncv_err = np.mean(cv_results[\\'test_score\\'])\\ncv_err\\nOut[9]: 24.2315\\nThe arguments to cross_validate() are as follows: an object with the ap-\\npropriate fit(), predict(), and score() methods, an array of features X and\\na response Y. We also included an additional argument cv to cross_validate();\\nspecifying an integer K results in K-fold cross-validation. We have provided\\na value corresponding to the total number of observations, which results\\nin leave-one-out cross-validation (LOOCV). The cross_validate() func- cross_\\nvalidate()\\ntion produces a dictionary with several components; we simply want the\\ncross-validated test score here (MSE), which is estimated to be 24.23.\\nWe can repeat this procedure for increasingly complex polynomial fits.\\nTo automate the process, we again use a for loop which iteratively fits\\npolynomial regressions of degree 1 to 5, computes the associated cross-\\nvalidation error, and stores it in the ith element of the vector cv_error.\\nThe variable d in the for loop corresponds to the degree of the polynomial.\\nWe begin by initializing the vector. This command may take a couple of\\nseconds to run.\\nIn [10]: cv_error = np.zeros(5)\\nH = np.array(Auto[\\'horsepower\\'])\\nM = sklearn_sm(sm.OLS)\\nfor i, d in enumerate(range(1,6)):\\nX = np.power.outer(H, np.arange(d+1))\\nM_CV = cross_validate(M,\\nX,\\nY,\\ncv=Auto.shape[0])\\ncv_error[i] = np.mean(M_CV[\\'test_score\\'])\\ncv_error\\nOut[10]: array([24.2315, 19.2482, 19.3350, 19.4244, 19.0332])\\nAs in Figure 5.4, we see a sharp drop in the estimated test MSE between\\nthe linear and quadratic fits, but then no clear improvement from using\\nhigher-degree polynomials. 5.3 Lab: Cross-Validation and the Bootstrap\\n219\\nAbove we introduced the outer() method of the np.power() function. .outer()\\nnp.power()\\nThe outer() method is applied to an operation that has two arguments,\\nsuch as add(), min(), or power(). It has two arrays as arguments, and then\\nforms a larger array where the operation is applied to each pair of elements\\nof the two arrays.\\nIn [11]: A = np.array([3, 5, 9])\\nB = np.array([2, 4])\\nnp.add.outer(A, B)\\nOut[11]: array([[ 5,\\n7],\\n[ 7,\\n9],\\n[11, 13]])\\nIn the CV example above, we used K = n, but of course we can also use\\nK < n. The code is very similar to the above (and is significantly faster).\\nHere we use KFold() to partition the data into K = 10 random groups. We KFold()\\nuse random_state to set a random seed and initialize a vector cv_error in\\nwhich we will store the CV errors corresponding to the polynomial fits of\\ndegrees one to five.\\nIn [12]: cv_error = np.zeros(5)\\ncv = KFold(n_splits=10,\\nshuffle=True,\\nrandom_state=0) # use same splits for each degree\\nfor i, d in enumerate(range(1,6)):\\nX = np.power.outer(H, np.arange(d+1))\\nM_CV = cross_validate(M,\\nX,\\nY,\\ncv=cv)\\ncv_error[i] = np.mean(M_CV[\\'test_score\\'])\\ncv_error\\nOut[12]: array([24.2077, 19.1853, 19.2763, 19.4785, 19.1372])\\nNotice that the computation time is much shorter than that of LOOCV.\\n(In principle, the computation time for LOOCV for a least squares linear\\nmodel should be faster than for K-fold CV, due to the availability of the\\nformula (5.2) for LOOCV; however, the generic cross_validate() function\\ndoes not make use of this formula.) We still see little evidence that using\\ncubic or higher-degree polynomial terms leads to a lower test error than\\nsimply using a quadratic fit.\\nThe cross_validate() function is flexible and can take different splitting\\nmechanisms as an argument. For instance, one can use the ShuffleSplit() Shuffle\\nSplit()\\nfuntion to implement the validation set approach just as easily as K-fold\\ncross-validation.\\nIn [13]: validation = ShuffleSplit(n_splits=1,\\ntest_size=196,\\nrandom_state=0)\\nresults = cross_validate(hp_model,\\nAuto.drop([\\'mpg\\'], axis=1),\\nAuto[\\'mpg\\'],\\ncv=validation);\\nresults[\\'test_score\\'] 220\\n5. Resampling Methods\\nOut[13]: array([23.6166])\\nOne can estimate the variability in the test error by running the following:\\nIn [14]: validation = ShuffleSplit(n_splits=10,\\ntest_size=196,\\nrandom_state=0)\\nresults = cross_validate(hp_model,\\nAuto.drop([\\'mpg\\'], axis=1),\\nAuto[\\'mpg\\'],\\ncv=validation)\\nresults[\\'test_score\\'].mean(), results[\\'test_score\\'].std()\\nOut[14]: (23.8022, 1.4218)\\nNote that this standard deviation is not a valid estimate of the sam-\\npling variability of the mean test score or the individual scores, since the\\nrandomly-selected training samples overlap and hence introduce correla-\\ntions. But it does give an idea of the Monte Carlo variation incurred by\\npicking different random folds.\\n5.3.3\\nThe Bootstrap\\nWe illustrate the use of the bootstrap in the simple example of Section 5.2,\\nas well as on an example involving estimating the accuracy of the linear\\nregression model on the Auto data set.\\nEstimating the Accuracy of a Statistic of Interest\\nOne of the great advantages of the bootstrap approach is that it can be\\napplied in almost all situations. No complicated mathematical calculations\\nare required. While there are several implementations of the bootstrap in\\nPython, its use for estimating standard error is simple enough that we write\\nour own function below for the case when our data is stored in a dataframe.\\nTo illustrate the bootstrap, we start with a simple example. The Portfolio\\ndata set in the ISLP package is described in Section 5.2. The goal is to es-\\ntimate the sampling variance of the parameter α given in formula (5.7).\\nWe will create a function alpha_func(), which takes as input a dataframe D\\nassumed to have columns X and Y, as well as a vector idx indicating which\\nobservations should be used to estimate α. The function then outputs the\\nestimate for α based on the selected observations.\\nIn [15]: Portfolio = load_data(\\'Portfolio\\')\\ndef alpha_func(D, idx):\\ncov_ = np.cov(D[[\\'X\\',\\'Y\\']].loc[idx], rowvar=False)\\nreturn ((cov_[1,1] - cov_[0,1]) /\\n(cov_[0,0]+cov_[1,1]-2*cov_[0,1]))\\nThis function returns an estimate for α based on applying the minimum\\nvariance formula (5.7) to the observations indexed by the argument idx. For\\ninstance, the following command estimates α using all 100 observations.\\nIn [16]: alpha_func(Portfolio, range(100)) 5.3 Lab: Cross-Validation and the Bootstrap\\n221\\nOut[16]: 0.5758\\nNext we randomly select 100 observations from range(100), with replace-\\nment. This is equivalent to constructing a new bootstrap data set and\\nrecomputing ˆα based on the new data set.\\nIn [17]: rng = np.random.default_rng(0)\\nalpha_func(Portfolio,\\nrng.choice(100,\\n100,\\nreplace=True))\\nOut[17]: 0.6074\\nThis process can be generalized to create a simple function boot_SE() for\\ncomputing the bootstrap standard error for arbitrary functions that take\\nonly a data frame as an argument.\\nIn [18]: def boot_SE(func,\\nD,\\nn=None,\\nB=1000,\\nseed=0):\\nrng = np.random.default_rng(seed)\\nfirst_, second_ = 0, 0\\nn = n or D.shape[0]\\nfor _ in range(B):\\nidx = rng.choice(D.index,\\nn,\\nreplace=True)\\nvalue = func(D, idx)\\nfirst_ += value\\nsecond_ += value**2\\nreturn np.sqrt(second_ / B - (first_ / B)**2)\\nNotice the use of _ as a loop variable in for _ in range(B). This is often\\nused if the value of the counter is unimportant and simply makes sure the\\nloop is executed B times.\\nLet’s use our function to evaluate the accuracy of our estimate of α using\\nB = 1,000 bootstrap replications.\\nIn [19]: alpha_SE = boot_SE(alpha_func,\\nPortfolio,\\nB=1000,\\nseed=0)\\nalpha_SE\\nOut[19]: 0.0912\\nThe final output shows that the bootstrap estimate for SE(ˆα) is 0.0912.\\nEstimating the Accuracy of a Linear Regression Model\\nThe bootstrap approach can be used to assess the variability of the coef-\\nficient estimates and predictions from a statistical learning method. Here\\nwe use the bootstrap approach in order to assess the variability of the 222\\n5. Resampling Methods\\nestimates for β0 and β1, the intercept and slope terms for the linear regres-\\nsion model that uses horsepower to predict mpg in the Auto data set. We\\nwill compare the estimates obtained using the bootstrap to those obtained\\nusing the formulas for SE(ˆβ0) and SE(ˆβ1) described in Section 3.1.2.\\nTo use our boot_SE() function, we must write a function (its first argu-\\nment) that takes a data frame D and indices idx as its only arguments. But\\nhere we want to bootstrap a specific regression model, specified by a model\\nformula and data. We show how to do this in a few simple steps.\\nWe start by writing a generic function boot_OLS() for bootstrapping a\\nregression model that takes a formula to define the corresponding regres-\\nsion. We use the clone() function to make a copy of the formula that can clone()\\nbe refit to the new dataframe. This means that any derived features such\\nas those defined by poly() (which we will see shortly), will be re-fit on the\\nresampled data frame.\\nIn [20]: def boot_OLS(model_matrix , response, D, idx):\\nD_ = D.loc[idx]\\nY_ = D_[response]\\nX_ = clone(model_matrix).fit_transform(D_)\\nreturn sm.OLS(Y_, X_).fit().params\\nThis is not quite what is needed as the first argument to boot_SE(). The first\\ntwo arguments which specify the model will not change in the bootstrap\\nprocess, and we would like to freeze them. The function partial() from the partial()\\nfunctools module does precisely this: it takes a function as an argument,\\nand freezes some of its arguments, starting from the left. We use it to freeze\\nthe first two model-formula arguments of boot_OLS().\\nIn [21]: hp_func = partial(boot_OLS, MS([\\'horsepower\\']), \\'mpg\\')\\nTyping hp_func? will show that it has two arguments D and idx — it is a\\nversion of boot_OLS() with the first two arguments frozen — and hence is\\nideal as the first argument for boot_SE().\\nThe hp_func() function can now be used in order to create bootstrap\\nestimates for the intercept and slope terms by randomly sampling from\\namong the observations with replacement. We first demonstrate its utility\\non 10 bootstrap samples.\\nIn [22]: rng = np.random.default_rng(0)\\nnp.array([hp_func(Auto,\\nrng.choice(392,\\n392,\\nreplace=True)) for _ in range(10)])\\nOut[22]: array([[39.8806, -0.1568],\\n[38.733 , -0.147 ],\\n[38.3173, -0.1444],\\n[39.9145, -0.1578],\\n[39.4335, -0.1507],\\n[40.3663, -0.1591],\\n[39.6233, -0.1545],\\n[39.0581, -0.1495],\\n[38.6669, -0.1452],\\n[39.6428, -0.1556]]) 5.3 Lab: Cross-Validation and the Bootstrap\\n223\\nNext, we use the boot_SE()\\nfunction to compute the standard errors of\\n1,000 bootstrap estimates for the intercept and slope terms.\\nIn [23]: hp_se = boot_SE(hp_func,\\nAuto,\\nB=1000,\\nseed=10)\\nhp_se\\nOut[23]: intercept\\n0.8488\\nhorsepower\\n0.0074\\ndtype: float64\\nThis indicates that the bootstrap estimate for SE(ˆβ0) is 0.85, and that\\nthe bootstrap estimate for SE(ˆβ1) is 0.0074. As discussed in Section 3.1.2,\\nstandard formulas can be used to compute the standard errors for the\\nregression coefficients in a linear model. These can be obtained using the\\nsummarize() function from ISLP.sm.\\nIn [24]: hp_model.fit(Auto, Auto[\\'mpg\\'])\\nmodel_se = summarize(hp_model.results_)[\\'std err\\']\\nmodel_se\\nOut[24]: intercept\\n0.717\\nhorsepower\\n0.006\\nName: std err, dtype: float64\\nThe standard error estimates for ˆβ0 and ˆβ1 obtained using the formulas\\nfrom Section 3.1.2 are 0.717 for the intercept and 0.006 for the slope. Inter-\\nestingly, these are somewhat different from the estimates obtained using\\nthe bootstrap. Does this indicate a problem with the bootstrap? In fact,\\nit suggests the opposite. Recall that the standard formulas given in Equa-\\ntion 3.8 on page 75 rely on certain assumptions. For example, they depend\\non the unknown parameter σ2, the noise variance. We then estimate σ2\\nusing the RSS. Now although the formula for the standard errors do not\\nrely on the linear model being correct, the estimate for σ2 does. We see in\\nFigure 3.8 on page 99 that there is a non-linear relationship in the data,\\nand so the residuals from a linear fit will be inflated, and so will ˆσ2. Sec-\\nondly, the standard formulas assume (somewhat unrealistically) that the\\nxi are fixed, and all the variability comes from the variation in the errors\\nϵi. The bootstrap approach does not rely on any of these assumptions, and\\nso it is likely giving a more accurate estimate of the standard errors of ˆβ0\\nand ˆβ1 than the results from sm.OLS.\\nBelow we compute the bootstrap standard error estimates and the stan-\\ndard linear regression estimates that result from fitting the quadratic model\\nto the data. Since this model provides a good fit to the data (Figure 3.8),\\nthere is now a better correspondence between the bootstrap estimates and\\nthe standard estimates of SE(ˆβ0), SE(ˆβ1) and SE(ˆβ2).\\nIn [25]: quad_model = MS([poly(\\'horsepower\\', 2, raw=True)])\\nquad_func = partial(boot_OLS,\\nquad_model,\\n\\'mpg\\')\\nboot_SE(quad_func, Auto, B=1000) 224\\n5. Resampling Methods\\nOut[25]: intercept\\n2.067840\\npoly(horsepower , 2, raw=True)[0]\\n0.033019\\npoly(horsepower , 2, raw=True)[1]\\n0.000120\\ndtype: float64\\nWe compare the results to the standard errors computed using sm.OLS().\\nIn [26]: M = sm.OLS(Auto[\\'mpg\\'],\\nquad_model.fit_transform(Auto))\\nsummarize(M.fit())[\\'std err\\']\\nOut[26]: intercept\\n1.800\\npoly(horsepower , 2, raw=True)[0]\\n0.031\\npoly(horsepower , 2, raw=True)[1]\\n0.000\\nName: std err, dtype: float64\\n5.4\\nExercises\\nConceptual\\n1. Using basic statistical properties of the variance, as well as single-\\nvariable calculus, derive (5.6). In other words, prove that α given by\\n(5.6) does indeed minimize Var(αX + (1 −α)Y ).\\n2. We will now derive the probability that a given observation is part\\nof a bootstrap sample. Suppose that we obtain a bootstrap sample\\nfrom a set of n observations.\\n(a) What is the probability that the first bootstrap observation is\\nnot the jth observation from the original sample? Justify your\\nanswer.\\n(b) What is the probability that the second bootstrap observation\\nis not the jth observation from the original sample?\\n(c) Argue that the probability that the jth observation is not in the\\nbootstrap sample is (1 −1/n)n.\\n(d) When n = 5, what is the probability that the jth observation is\\nin the bootstrap sample?\\n(e) When n = 100, what is the probability that the jth observation\\nis in the bootstrap sample?\\n(f) When n = 10, 000, what is the probability that the jth observa-\\ntion is in the bootstrap sample?\\n(g) Create a plot that displays, for each integer value of n from 1\\nto 100, 000, the probability that the jth observation is in the\\nbootstrap sample. Comment on what you observe.\\n(h) We will now investigate numerically the probability that a boot-\\nstrap sample of size n = 100 contains the jth observation. Here\\nj = 4. We first create an array store with values that will subse-\\nquently be overwritten using the function np.empty(). We then np.empty() 5.4 Exercises\\n225\\nrepeatedly create bootstrap samples, and each time we record\\nwhether or not the fifth observation is contained in the bootstrap\\nsample.\\nrng = np.random.default_rng(10)\\nstore = np.empty(10000)\\nfor i in range(10000):\\nstore[i] = np.sum(rng.choice(100, replace=True) == 4)\\n> 0\\nnp.mean(store)\\nComment on the results obtained.\\n3. We now review k-fold cross-validation.\\n(a) Explain how k-fold cross-validation is implemented.\\n(b) What are the advantages and disadvantages of k-fold cross-\\nvalidation relative to:\\ni. The validation set approach?\\nii. LOOCV?\\n4. Suppose that we use some statistical learning method to make a pre-\\ndiction for the response Y for a particular value of the predictor X.\\nCarefully describe how we might estimate the standard deviation of\\nour prediction.\\nApplied\\n5. In Chapter 4, we used logistic regression to predict the probability of\\ndefault using income and balance on the Default data set. We will\\nnow estimate the test error of this logistic regression model using the\\nvalidation set approach. Do not forget to set a random seed before\\nbeginning your analysis.\\n(a) Fit a logistic regression model that uses income and balance to\\npredict default.\\n(b) Using the validation set approach, estimate the test error of this\\nmodel. In order to do this, you must perform the following steps:\\ni. Split the sample set into a training set and a validation set.\\nii. Fit a multiple logistic regression model using only the train-\\ning observations.\\niii. Obtain a prediction of default status for each individual in\\nthe validation set by computing the posterior probability of\\ndefault for that individual, and classifying the individual to\\nthe default category if the posterior probability is greater\\nthan 0.5.\\niv. Compute the validation set error, which is the fraction of\\nthe observations in the validation set that are misclassified.\\n(c) Repeat the process in (b) three times, using three different splits\\nof the observations into a training set and a validation set. Com-\\nment on the results obtained. 226\\n5. Resampling Methods\\n(d) Now consider a logistic regression model that predicts the prob-\\nability of default using income, balance, and a dummy variable\\nfor student. Estimate the test error for this model using the val-\\nidation set approach. Comment on whether or not including a\\ndummy variable for student leads to a reduction in the test error\\nrate.\\n6. We continue to consider the use of a logistic regression model to\\npredict the probability of default using income and balance on the\\nDefault data set. In particular, we will now compute estimates for the\\nstandard errors of the income and balance logistic regression coeffi-\\ncients in two different ways: (1) using the bootstrap, and (2) using the\\nstandard formula for computing the standard errors in the sm.GLM()\\nfunction. Do not forget to set a random seed before beginning your\\nanalysis.\\n(a) Using the summarize() and sm.GLM() functions, determine the\\nestimated standard errors for the coefficients associated with\\nincome and balance in a multiple logistic regression model that\\nuses both predictors.\\n(b) Write a function, boot_fn(), that takes as input the Default data\\nset as well as an index of the observations, and that outputs\\nthe coefficient estimates for income and balance in the multiple\\nlogistic regression model.\\n(c) Following the bootstrap example in the lab, use your boot_fn()\\nfunction to estimate the standard errors of the logistic regression\\ncoefficients for income and balance.\\n(d) Comment on the estimated standard errors obtained using the\\nsm.GLM() function and using the bootstrap.\\n7. In Sections 5.1.2 and 5.1.3, we saw that the cross_validate() function\\ncan be used in order to compute the LOOCV test error estimate.\\nAlternatively, one could compute those quantities using just sm.GLM()\\nand the predict() method of the fitted model within a for loop. You\\nwill now take this approach in order to compute the LOOCV error\\nfor a simple logistic regression model on the Weekly data set. Recall\\nthat in the context of classification problems, the LOOCV error is\\ngiven in (5.4).\\n(a) Fit a logistic regression model that predicts Direction using Lag1\\nand Lag2.\\n(b) Fit a logistic regression model that predicts Direction using Lag1\\nand Lag2 using all but the first observation.\\n(c) Use the model from (b) to predict the direction of the first obser-\\nvation. You can do this by predicting that the first observation\\nwill go up if P(Direction = \"Up\"|Lag1, Lag2) > 0.5. Was this\\nobservation correctly classified?\\n(d) Write a for loop from i = 1 to i = n, where n is the number of\\nobservations in the data set, that performs each of the following\\nsteps: 5.4 Exercises\\n227\\ni. Fit a logistic regression model using all but the ith obser-\\nvation to predict Direction using Lag1 and Lag2.\\nii. Compute the posterior probability of the market moving up\\nfor the ith observation.\\niii. Use the posterior probability for the ith observation in order\\nto predict whether or not the market moves up.\\niv. Determine whether or not an error was made in predicting\\nthe direction for the ith observation. If an error was made,\\nthen indicate this as a 1, and otherwise indicate it as a 0.\\n(e) Take the average of the n numbers obtained in (d)iv in order to\\nobtain the LOOCV estimate for the test error. Comment on the\\nresults.\\n8. We will now perform cross-validation on a simulated data set.\\n(a) Generate a simulated data set as follows:\\nrng = np.random.default_rng(1)\\nx = rng.normal(size=100)\\ny = x - 2 * x**2 + rng.normal(size=100)\\nIn this data set, what is n and what is p? Write out the model\\nused to generate the data in equation form.\\n(b) Create a scatterplot of X against Y . Comment on what you find.\\n(c) Set a random seed, and then compute the LOOCV errors that\\nresult from fitting the following four models using least squares:\\ni. Y = β0 + β1X + ϵ\\nii. Y = β0 + β1X + β2X2 + ϵ\\niii. Y = β0 + β1X + β2X2 + β3X3 + ϵ\\niv. Y = β0 + β1X + β2X2 + β3X3 + β4X4 + ϵ.\\nNote you may find it helpful to use the data.frame() function\\nto create a single data set containing both X and Y .\\n(d) Repeat (c) using another random seed, and report your results.\\nAre your results the same as what you got in (c)? Why?\\n(e) Which of the models in (c) had the smallest LOOCV error? Is\\nthis what you expected? Explain your answer.\\n(f) Comment on the statistical significance of the coefficient esti-\\nmates that results from fitting each of the models in (c) using\\nleast squares. Do these results agree with the conclusions drawn\\nbased on the cross-validation results?\\n9. We will now consider the Boston housing data set, from the ISLP\\nlibrary.\\n(a) Based on this data set, provide an estimate for the population\\nmean of medv. Call this estimate ˆµ. 228\\n5. Resampling Methods\\n(b) Provide an estimate of the standard error of ˆµ. Interpret this\\nresult.\\nHint: We can compute the standard error of the sample mean by\\ndividing the sample standard deviation by the square root of the\\nnumber of observations.\\n(c) Now estimate the standard error of ˆµ using the bootstrap. How\\ndoes this compare to your answer from (b)?\\n(d) Based on your bootstrap estimate from (c), provide a 95 % con-\\nfidence interval for the mean of medv. Compare it to the results\\nobtained by using Boston[\\'medv\\'].std() and the two standard\\nerror rule (3.9).\\nHint: You can approximate a 95 % confidence interval using the\\nformula [ˆµ −2SE(ˆµ), ˆµ + 2SE(ˆµ)].\\n(e) Based on this data set, provide an estimate, ˆµmed, for the median\\nvalue of medv in the population.\\n(f) We now would like to estimate the standard error of ˆµmed. Unfor-\\ntunately, there is no simple formula for computing the standard\\nerror of the median. Instead, estimate the standard error of the\\nmedian using the bootstrap. Comment on your findings.\\n(g) Based on this data set, provide an estimate for the tenth per-\\ncentile of medv in Boston census tracts. Call this quantity ˆµ0.1.\\n(You can use the np.percentile() function.)\\nnp.\\npercentile()\\n(h) Use the bootstrap to estimate the standard error of ˆµ0.1. Com-\\nment on your findings. 6\\nLinear Model Selection\\nand Regularization\\nIn the regression setting, the standard linear model\\nY = β0 + β1X1 + · · · + βpXp + ϵ\\n(6.1)\\nis commonly used to describe the relationship between a response Y and\\na set of variables X1, X2, . . . , Xp. We have seen in Chapter 3 that one\\ntypically fits this model using least squares.\\nIn the chapters that follow, we consider some approaches for extending\\nthe linear model framework. In Chapter 7 we generalize (6.1) in order to\\naccommodate non-linear, but still additive, relationships, while in Chap-\\nters 8 and 10 we consider even more general non-linear models. However,\\nthe linear model has distinct advantages in terms of inference and, on real-\\nworld problems, is often surprisingly competitive in relation to non-linear\\nmethods. Hence, before moving to the non-linear world, we discuss in this\\nchapter some ways in which the simple linear model can be improved, by re-\\nplacing plain least squares fitting with some alternative fitting procedures.\\nWhy might we want to use another fitting procedure instead of least\\nsquares? As we will see, alternative fitting procedures can yield better pre-\\ndiction accuracy and model interpretability.\\n• Prediction Accuracy: Provided that the true relationship between the\\nresponse and the predictors is approximately linear, the least squares\\nestimates will have low bias. If n ≫p—that is, if n, the number of\\nobservations, is much larger than p, the number of variables—then the\\nleast squares estimates tend to also have low variance, and hence will\\nperform well on test observations. However, if n is not much larger\\nthan p, then there can be a lot of variability in the least squares fit,\\nresulting in overfitting and consequently poor predictions on future\\nobservations not used in model training. And if p > n, then there is no\\nlonger a unique least squares coefficient estimate: there are infinitely\\n© Springer Nature Switzerland AG 2023 \\nG. James et al., An Introduction to Statistical Learning, Springer Texts in Statistics, \\nhttps://doi.org/10.1007/978-3-031-38747-0_6 \\n229 230\\n6. Linear Model Selection and Regularization\\nmany solutions. Each of these least squares solutions gives zero error\\non the training data, but typically very poor test set performance\\ndue to extremely high variance.1 By constraining or shrinking the\\nestimated coefficients, we can often substantially reduce the variance\\nat the cost of a negligible increase in bias. This can lead to substantial\\nimprovements in the accuracy with which we can predict the response\\nfor observations not used in model training.\\n• Model Interpretability: It is often the case that some or many of the\\nvariables used in a multiple regression model are in fact not associ-\\nated with the response. Including such irrelevant variables leads to\\nunnecessary complexity in the resulting model. By removing these\\nvariables—that is, by setting the corresponding coefficient estimates\\nto zero—we can obtain a model that is more easily interpreted. Now\\nleast squares is extremely unlikely to yield any coefficient estimates\\nthat are exactly zero. In this chapter, we see some approaches for au-\\ntomatically performing feature selection or variable selection—that is, feature\\nselection\\nvariable\\nselection\\nfor excluding irrelevant variables from a multiple regression model.\\nThere are many alternatives, both classical and modern, to using least\\nsquares to fit (6.1). In this chapter, we discuss three important classes of\\nmethods.\\n• Subset Selection. This approach involves identifying a subset of the p\\npredictors that we believe to be related to the response. We then fit\\na model using least squares on the reduced set of variables.\\n• Shrinkage. This approach involves fitting a model involving all p pre-\\ndictors. However, the estimated coefficients are shrunken towards zero\\nrelative to the least squares estimates. This shrinkage (also known as\\nregularization) has the effect of reducing variance. Depending on what\\ntype of shrinkage is performed, some of the coefficients may be esti-\\nmated to be exactly zero. Hence, shrinkage methods can also perform\\nvariable selection.\\n• Dimension Reduction. This approach involves projecting the p predic-\\ntors into an M-dimensional subspace, where M < p. This is achieved\\nby computing M different linear combinations, or projections, of the\\nvariables. Then these M projections are used as predictors to fit a\\nlinear regression model by least squares.\\nIn the following sections we describe each of these approaches in greater de-\\ntail, along with their advantages and disadvantages. Although this chapter\\ndescribes extensions and modifications to the linear model for regression\\nseen in Chapter 3, the same concepts apply to other methods, such as the\\nclassification models seen in Chapter 4.\\n1When p ≫n, the least squares solution that has the smallest sum of squared coeffi-\\ncients can sometimes perform quite well. See Section 10.8 for a more detailed discussion. 6.1 Subset Selection\\n231\\n6.1\\nSubset Selection\\nIn this section we consider some methods for selecting subsets of predictors.\\nThese include best subset and stepwise model selection procedures.\\n6.1.1\\nBest Subset Selection\\nTo perform best subset selection, we fit a separate least squares regression best subset\\nselection\\nfor each possible combination of the p predictors. That is, we fit all p models\\nthat contain exactly one predictor, all\\n\\'p\\n2\\n(\\n= p(p−1)/2 models that contain\\nexactly two predictors, and so forth. We then look at all of the resulting\\nmodels, with the goal of identifying the one that is best.\\nThe problem of selecting the best model from among the 2p possibilities\\nconsidered by best subset selection is not trivial. This is usually broken up\\ninto two stages, as described in Algorithm 6.1.\\nAlgorithm 6.1 Best subset selection\\n1. Let M0 denote the null model, which contains no predictors. This\\nmodel simply predicts the sample mean for each observation.\\n2. For k = 1, 2, . . . p:\\n(a) Fit all\\n\\'p\\nk\\n(\\nmodels that contain exactly k predictors.\\n(b) Pick the best among these\\n\\'p\\nk\\n(\\nmodels, and call it Mk. Here best\\nis defined as having the smallest RSS, or equivalently largest R2.\\n3. Select a single best model from among M0, . . . , Mp using using the\\nprediction error on a validation set, Cp (AIC), BIC, or adjusted R2.\\nOr use the cross-validation method.\\nIn Algorithm 6.1, Step 2 identifies the best model (on the training data)\\nfor each subset size, in order to reduce the problem from one of 2p possible\\nmodels to one of p + 1 possible models. In Figure 6.1, these models form\\nthe lower frontier depicted in red.\\nNow in order to select a single best model, we must simply choose among\\nthese p + 1 options. This task must be performed with care, because the\\nRSS of these p + 1 models decreases monotonically, and the R2 increases\\nmonotonically, as the number of features included in the models increases.\\nTherefore, if we use these statistics to select the best model, then we will\\nalways end up with a model involving all of the variables. The problem is\\nthat a low RSS or a high R2 indicates a model with a low training error,\\nwhereas we wish to choose a model that has a low test error. (As shown in\\nChapter 2 in Figures 2.9–2.11, training error tends to be quite a bit smaller\\nthan test error, and a low training error by no means guarantees a low test\\nerror.) Therefore, in Step 3, we use the error on a validation set, Cp, BIC, or\\nadjusted R2 in order to select among M0, M1, . . . , Mp. If cross-validation\\nis used to select the best model, then Step 2 is repeated on each training\\nfold, and the validation errors are averaged to select the best value of k. 232\\n6. Linear Model Selection and Regularization\\n2\\n4\\n6\\n8\\n10\\n2e+07\\n4e+07\\n6e+07\\n8e+07\\nNumber of Predictors\\nResidual Sum of Squares\\n2\\n4\\n6\\n8\\n10\\n0.0\\n0.2\\n0.4\\n0.6\\n0.8\\n1.0\\nNumber of Predictors\\nR2\\nFIGURE 6.1. For each possible model containing a subset of the ten predictors\\nin the Credit data set, the RSS and R2 are displayed. The red frontier tracks the\\nbest model for a given number of predictors, according to RSS and R2. Though\\nthe data set contains only ten predictors, the x-axis ranges from 1 to 11, since one\\nof the variables is categorical and takes on three values, leading to the creation of\\ntwo dummy variables.\\nThen the model Mk fit on the full training set is delivered for the chosen\\nk. These approaches are discussed in Section 6.1.3.\\nAn application of best subset selection is shown in Figure 6.1. Each\\nplotted point corresponds to a least squares regression model fit using a\\ndifferent subset of the 10 predictors in the Credit data set, discussed in\\nChapter 3. Here the variable region is a three-level qualitative variable,\\nand so is represented by two dummy variables, which are selected sepa-\\nrately in this case. Hence, there are a total of 11 possible variables which\\ncan be included in the model. We have plotted the RSS and R2 statistics\\nfor each model, as a function of the number of variables. The red curves\\nconnect the best models for each model size, according to RSS or R2. The\\nfigure shows that, as expected, these quantities improve as the number of\\nvariables increases; however, from the three-variable model on, there is little\\nimprovement in RSS and R2 as a result of including additional predictors.\\nAlthough we have presented best subset selection here for least squares\\nregression, the same ideas apply to other types of models, such as logistic\\nregression. In the case of logistic regression, instead of ordering models by\\nRSS in Step 2 of Algorithm 6.1, we instead use the deviance, a measure deviance\\nthat plays the role of RSS for a broader class of models. The deviance is\\nnegative two times the maximized log-likelihood; the smaller the deviance,\\nthe better the fit.\\nWhile best subset selection is a simple and conceptually appealing ap-\\nproach, it suffers from computational limitations. The number of possible\\nmodels that must be considered grows rapidly as p increases. In general,\\nthere are 2p models that involve subsets of p predictors. So if p = 10,\\nthen there are approximately 1,000 possible models to be considered, and if\\np = 20, then there are over one million possibilities! Consequently, best sub-\\nset selection becomes computationally infeasible for values of p greater than 6.1 Subset Selection\\n233\\nAlgorithm 6.2 Forward stepwise selection\\n1. Let M0 denote the null model, which contains no predictors.\\n2. For k = 0, . . . , p −1:\\n(a) Consider all p −k models that augment the predictors in Mk\\nwith one additional predictor.\\n(b) Choose the best among these p −k models, and call it Mk+1.\\nHere best is defined as having smallest RSS or highest R2.\\n3. Select a single best model from among M0, . . . , Mp using the pre-\\ndiction error on a validation set, Cp (AIC), BIC, or adjusted R2. Or\\nuse the cross-validation method.\\naround 40, even with extremely fast modern computers. There are compu-\\ntational shortcuts—so called branch-and-bound techniques—for eliminat-\\ning some choices, but these have their limitations as p gets large. They also\\nonly work for least squares linear regression. We present computationally\\nefficient alternatives to best subset selection next.\\n6.1.2\\nStepwise Selection\\nFor computational reasons, best subset selection cannot be applied with\\nvery large p. Best subset selection may also suffer from statistical problems\\nwhen p is large. The larger the search space, the higher the chance of finding\\nmodels that look good on the training data, even though they might not\\nhave any predictive power on future data. Thus an enormous search space\\ncan lead to overfitting and high variance of the coefficient estimates.\\nFor both of these reasons, stepwise methods, which explore a far more\\nrestricted set of models, are attractive alternatives to best subset selection.\\nForward Stepwise Selection\\nForward stepwise selection is a computationally efficient alternative to best forward\\nstepwise\\nselection\\nsubset selection. While the best subset selection procedure considers all\\n2p possible models containing subsets of the p predictors, forward step-\\nwise considers a much smaller set of models. Forward stepwise selection\\nbegins with a model containing no predictors, and then adds predictors\\nto the model, one-at-a-time, until all of the predictors are in the model.\\nIn particular, at each step the variable that gives the greatest additional\\nimprovement to the fit is added to the model. More formally, the forward\\nstepwise selection procedure is given in Algorithm 6.2.\\nUnlike best subset selection, which involved fitting 2p models, forward\\nstepwise selection involves fitting one null model, along with p −k models\\nin the kth iteration, for k = 0, . . . , p −1. This amounts to a total of 1 +\\n)p−1\\nk=0(p−k) = 1+p(p+1)/2 models. This is a substantial difference: when 234\\n6. Linear Model Selection and Regularization\\n# Variables\\nBest subset\\nForward stepwise\\nOne\\nrating\\nrating\\nTwo\\nrating, income\\nrating, income\\nThree\\nrating, income, student\\nrating, income, student\\nFour\\ncards, income\\nrating, income,\\nstudent, limit\\nstudent, limit\\nTABLE 6.1. The first four selected models for best subset selection and forward\\nstepwise selection on the Credit data set. The first three models are identical but\\nthe fourth models differ.\\np = 20, best subset selection requires fitting 1,048,576 models, whereas\\nforward stepwise selection requires fitting only 211 models.2\\nIn Step 2(b) of Algorithm 6.2, we must identify the best model from\\namong those p−k that augment Mk with one additional predictor. We can\\ndo this by simply choosing the model with the lowest RSS or the highest\\nR2. However, in Step 3, we must identify the best model among a set of\\nmodels with different numbers of variables. This is more challenging, and\\nis discussed in Section 6.1.3.\\nForward stepwise selection’s computational advantage over best subset\\nselection is clear. Though forward stepwise tends to do well in practice,\\nit is not guaranteed to find the best possible model out of all 2p mod-\\nels containing subsets of the p predictors. For instance, suppose that in a\\ngiven data set with p = 3 predictors, the best possible one-variable model\\ncontains X1, and the best possible two-variable model instead contains X2\\nand X3. Then forward stepwise selection will fail to select the best possible\\ntwo-variable model, because M1 will contain X1, so M2 must also contain\\nX1 together with one additional variable.\\nTable 6.1, which shows the first four selected models for best subset\\nand forward stepwise selection on the Credit data set, illustrates this phe-\\nnomenon. Both best subset selection and forward stepwise selection choose\\nrating for the best one-variable model and then include income and student\\nfor the two- and three-variable models. However, best subset selection re-\\nplaces rating by cards in the four-variable model, while forward stepwise\\nselection must maintain rating in its four-variable model. In this example,\\nFigure 6.1 indicates that there is not much difference between the three-\\nand four-variable models in terms of RSS, so either of the four-variable\\nmodels will likely be adequate.\\nForward stepwise selection can be applied even in the high-dimensional\\nsetting where n < p; however, in this case, it is possible to construct sub-\\nmodels M0, . . . , Mn−1 only, since each submodel is fit using least squares,\\nwhich will not yield a unique solution if p ≥n.\\nBackward Stepwise Selection\\nLike forward stepwise selection, backward stepwise selection provides an backward\\nstepwise\\nselection\\n2Though forward stepwise selection considers p(p + 1)/2 + 1 models, it performs a\\nguided search over model space, and so the effective model space considered contains\\nsubstantially more than p(p + 1)/2 + 1 models. 6.1 Subset Selection\\n235\\nefficient alternative to best subset selection. However, unlike forward step-\\nwise selection, it begins with the full least squares model containing all p\\npredictors, and then iteratively removes the least useful predictor, one-at-\\na-time. Details are given in Algorithm 6.3.\\nAlgorithm 6.3 Backward stepwise selection\\n1. Let Mp denote the full model, which contains all p predictors.\\n2. For k = p, p −1, . . . , 1:\\n(a) Consider all k models that contain all but one of the predictors\\nin Mk, for a total of k −1 predictors.\\n(b) Choose the best among these k models, and call it Mk−1. Here\\nbest is defined as having smallest RSS or highest R2.\\n3. Select a single best model from among M0, . . . , Mp using the pre-\\ndiction error on a validation set, Cp (AIC), BIC, or adjusted R2. Or\\nuse the cross-validation method.\\nLike forward stepwise selection, the backward selection approach searches\\nthrough only 1+p(p+1)/2 models, and so can be applied in settings where\\np is too large to apply best subset selection.3 Also like forward stepwise\\nselection, backward stepwise selection is not guaranteed to yield the best\\nmodel containing a subset of the p predictors.\\nBackward selection requires that the number of samples n is larger than\\nthe number of variables p (so that the full model can be fit). In contrast,\\nforward stepwise can be used even when n < p, and so is the only viable\\nsubset method when p is very large.\\nHybrid Approaches\\nThe best subset, forward stepwise, and backward stepwise selection ap-\\nproaches generally give similar but not identical models. As another al-\\nternative, hybrid versions of forward and backward stepwise selection are\\navailable, in which variables are added to the model sequentially, in analogy\\nto forward selection. However, after adding each new variable, the method\\nmay also remove any variables that no longer provide an improvement in\\nthe model fit. Such an approach attempts to more closely mimic best sub-\\nset selection while retaining the computational advantages of forward and\\nbackward stepwise selection.\\n6.1.3\\nChoosing the Optimal Model\\nBest subset selection, forward selection, and backward selection result in\\nthe creation of a set of models, each of which contains a subset of the p\\n3Like forward stepwise selection, backward stepwise selection performs a guided search\\nover model space, and so effectively considers substantially more than 1 + p(p + 1)/2\\nmodels. 236\\n6. Linear Model Selection and Regularization\\npredictors. To apply these methods, we need a way to determine which of\\nthese models is best. As we discussed in Section 6.1.1, the model containing\\nall of the predictors will always have the smallest RSS and the largest R2,\\nsince these quantities are related to the training error. Instead, we wish to\\nchoose a model with a low test error. As is evident here, and as we show\\nin Chapter 2, the training error can be a poor estimate of the test error.\\nTherefore, RSS and R2 are not suitable for selecting the best model among\\na collection of models with different numbers of predictors.\\nIn order to select the best model with respect to test error, we need to\\nestimate this test error. There are two common approaches:\\n1. We can indirectly estimate test error by making an adjustment to the\\ntraining error to account for the bias due to overfitting.\\n2. We can directly estimate the test error, using either a validation set\\napproach or a cross-validation approach, as discussed in Chapter 5.\\nWe consider both of these approaches below.\\nCp, AIC, BIC, and Adjusted R2\\nWe show in Chapter 2 that the training set MSE is generally an under-\\nestimate of the test MSE. (Recall that MSE = RSS/n.) This is because\\nwhen we fit a model to the training data using least squares, we specifi-\\ncally estimate the regression coefficients such that the training RSS (but\\nnot the test RSS) is as small as possible. In particular, the training error\\nwill decrease as more variables are included in the model, but the test error\\nmay not. Therefore, training set RSS and training set R2 cannot be used\\nto select from among a set of models with different numbers of variables.\\nHowever, a number of techniques for adjusting the training error for the\\nmodel size are available. These approaches can be used to select among a set\\nof models with different numbers of variables. We now consider four such\\napproaches: Cp, Akaike information criterion (AIC), Bayesian information Cp\\nAkaike\\ninformation\\ncriterion\\ncriterion (BIC), and adjusted R2. Figure 6.2 displays Cp, BIC, and adjusted\\nBayesian\\ninformation\\ncriterion\\nadjusted R2\\nR2 for the best model of each size produced by best subset selection on the\\nCredit data set.\\nFor a fitted least squares model containing d predictors, the Cp estimate\\nof test MSE is computed using the equation\\nCp = 1\\nn\\n\\'\\nRSS + 2dˆσ2(\\n,\\n(6.2)\\nwhere ˆσ2 is an estimate of the variance of the error ϵ associated with each\\nresponse measurement in (6.1).4 Typically ˆσ2 is estimated using the full\\nmodel containing all predictors. Essentially, the Cp statistic adds a penalty\\nof 2dˆσ2 to the training RSS in order to adjust for the fact that the training\\nerror tends to underestimate the test error. Clearly, the penalty increases as\\nthe number of predictors in the model increases; this is intended to adjust\\n4Mallow’s Cp is sometimes defined as C′\\np = RSS/ˆσ2 + 2d −n. This is equivalent to\\nthe definition given above in the sense that Cp = 1\\nn ˆσ2(C′\\np + n), and so the model with\\nsmallest Cp also has smallest C′\\np. 6.1 Subset Selection\\n237\\n2\\n4\\n6\\n8\\n10\\n10000\\n15000\\n20000\\n25000\\n30000\\nNumber of Predictors\\nCp\\n2\\n4\\n6\\n8\\n10\\n10000\\n15000\\n20000\\n25000\\n30000\\nNumber of Predictors\\nBIC\\n2\\n4\\n6\\n8\\n10\\n0.86\\n0.88\\n0.90\\n0.92\\n0.94\\n0.96\\nNumber of Predictors\\nAdjusted R2 \\nFIGURE 6.2. Cp, BIC, and adjusted R2 are shown for the best models of each\\nsize for the Credit data set (the lower frontier in Figure 6.1). Cp and BIC are\\nestimates of test MSE. In the middle plot we see that the BIC estimate of test\\nerror shows an increase after four variables are selected. The other two plots are\\nrather flat after four variables are included.\\nfor the corresponding decrease in training RSS. Though it is beyond the\\nscope of this book, one can show that if ˆσ2 is an unbiased estimate of σ2 in\\n(6.2), then Cp is an unbiased estimate of test MSE. As a consequence, the\\nCp statistic tends to take on a small value for models with a low test error,\\nso when determining which of a set of models is best, we choose the model\\nwith the lowest Cp value. In Figure 6.2, Cp selects the six-variable model\\ncontaining the predictors income, limit, rating, cards, age and student.\\nThe AIC criterion is defined for a large class of models fit by maximum\\nlikelihood. In the case of the model (6.1) with Gaussian errors, maximum\\nlikelihood and least squares are the same thing. In this case AIC is given by\\nAIC = 1\\nn\\n\\'\\nRSS + 2dˆσ2(\\n,\\nwhere, for simplicity, we have omitted irrelevant constants.5 Hence for least\\nsquares models, Cp and AIC are proportional to each other, and so only\\nCp is displayed in Figure 6.2.\\nBIC is derived from a Bayesian point of view, but ends up looking similar\\nto Cp (and AIC) as well. For the least squares model with d predictors, the\\nBIC is, up to irrelevant constants, given by\\nBIC = 1\\nn\\n\\'\\nRSS + log(n)dˆσ2(\\n.\\n(6.3)\\nLike Cp, the BIC will tend to take on a small value for a model with a\\nlow test error, and so generally we select the model that has the lowest\\nBIC value. Notice that BIC replaces the 2dˆσ2 used by Cp with a log(n)dˆσ2\\nterm, where n is the number of observations. Since log n > 2 for any n > 7,\\n5There are two formulas for AIC for least squares regression. The formula that we\\nprovide here requires an expression for σ2, which we obtain using the full model con-\\ntaining all predictors. The second formula is appropriate when σ2 is unknown and we do\\nnot want to explicitly estimate it; that formula has a log(RSS) term instead of an RSS\\nterm. Detailed derivations of these two formulas are outside of the scope of this book. 238\\n6. Linear Model Selection and Regularization\\nthe BIC statistic generally places a heavier penalty on models with many\\nvariables, and hence results in the selection of smaller models than Cp.\\nIn Figure 6.2, we see that this is indeed the case for the Credit data set;\\nBIC chooses a model that contains only the four predictors income, limit,\\ncards, and student. In this case the curves are very flat and so there does\\nnot appear to be much difference in accuracy between the four-variable and\\nsix-variable models.\\nThe adjusted R2 statistic is another popular approach for selecting among\\na set of models that contain different numbers of variables. Recall from\\nChapter 3 that the usual R2 is defined as 1 −RSS/TSS, where TSS =\\n)(yi −y)2 is the total sum of squares for the response. Since RSS always\\ndecreases as more variables are added to the model, the R2 always increases\\nas more variables are added. For a least squares model with d variables,\\nthe adjusted R2 statistic is calculated as\\nAdjusted R2 = 1 −RSS/(n −d −1)\\nTSS/(n −1)\\n.\\n(6.4)\\nUnlike Cp, AIC, and BIC, for which a small value indicates a model with\\na low test error, a large value of adjusted R2 indicates a model with a\\nsmall test error. Maximizing the adjusted R2 is equivalent to minimizing\\nRSS\\nn−d−1. While RSS always decreases as the number of variables in the model\\nincreases,\\nRSS\\nn−d−1 may increase or decrease, due to the presence of d in the\\ndenominator.\\nThe intuition behind the adjusted R2 is that once all of the correct\\nvariables have been included in the model, adding additional noise variables\\nwill lead to only a very small decrease in RSS. Since adding noise variables\\nleads to an increase in d, such variables will lead to an increase in\\nRSS\\nn−d−1,\\nand consequently a decrease in the adjusted R2. Therefore, in theory, the\\nmodel with the largest adjusted R2 will have only correct variables and\\nno noise variables. Unlike the R2 statistic, the adjusted R2 statistic pays\\na price for the inclusion of unnecessary variables in the model. Figure 6.2\\ndisplays the adjusted R2 for the Credit data set. Using this statistic results\\nin the selection of a model that contains seven variables, adding own to the\\nmodel selected by Cp and AIC.\\nCp, AIC, and BIC all have rigorous theoretical justifications that are\\nbeyond the scope of this book. These justifications rely on asymptotic ar-\\nguments (scenarios where the sample size n is very large). Despite its pop-\\nularity, and even though it is quite intuitive, the adjusted R2 is not as well\\nmotivated in statistical theory as AIC, BIC, and Cp. All of these measures\\nare simple to use and compute. Here we have presented their formulas in\\nthe case of a linear model fit using least squares; however, AIC and BIC\\ncan also be defined for more general types of models.\\nValidation and Cross-Validation\\nAs an alternative to the approaches just discussed, we can directly esti-\\nmate the test error using the validation set and cross-validation methods\\ndiscussed in Chapter 5. We can compute the validation set error or the\\ncross-validation error for each model under consideration, and then select 6.1 Subset Selection\\n239\\n2\\n4\\n6\\n8\\n10\\n100\\n120\\n140\\n160\\n180\\n200\\n220\\nNumber of Predictors\\nSquare Root of BIC\\n2\\n4\\n6\\n8\\n10\\n100\\n120\\n140\\n160\\n180\\n200\\n220\\nNumber of Predictors\\nValidation Set Error\\n2\\n4\\n6\\n8\\n10\\n100\\n120\\n140\\n160\\n180\\n200\\n220\\nNumber of Predictors\\nCross−Validation Error\\nFIGURE 6.3. For the Credit data set, three quantities are displayed for the\\nbest model containing d predictors, for d ranging from 1 to 11. The overall best\\nmodel, based on each of these quantities, is shown as a blue cross. Left: Square\\nroot of BIC. Center: Validation set errors. Right: Cross-validation errors.\\nthe model for which the resulting estimated test error is smallest. This pro-\\ncedure has an advantage relative to AIC, BIC, Cp, and adjusted R2, in that\\nit provides a direct estimate of the test error, and makes fewer assumptions\\nabout the true underlying model. It can also be used in a wider range of\\nmodel selection tasks, even in cases where it is hard to pinpoint the model\\ndegrees of freedom (e.g. the number of predictors in the model) or hard\\nto estimate the error variance σ2. Note that when cross-validation is used,\\nthe sequence of models Mk in Algorithms 6.1–6.3 is determined separately\\nfor each training fold, and the validation errors are averaged over all folds\\nfor each model size k. This means, for example with best-subset regression,\\nthat Mk, the best subset of size k, can differ across the folds. Once the\\nbest size k is chosen, we find the best model of that size on the full data\\nset.\\nIn the past, performing cross-validation was computationally prohibitive\\nfor many problems with large p and/or large n, and so AIC, BIC, Cp,\\nand adjusted R2 were more attractive approaches for choosing among a\\nset of models. However, nowadays with fast computers, the computations\\nrequired to perform cross-validation are hardly ever an issue. Thus, cross-\\nvalidation is a very attractive approach for selecting from among a number\\nof models under consideration.\\nFigure 6.3 displays, as a function of d, the BIC, validation set errors, and\\ncross-validation errors on the Credit data, for the best d-variable model.\\nThe validation errors were calculated by randomly selecting three-quarters\\nof the observations as the training set, and the remainder as the valida-\\ntion set. The cross-validation errors were computed using k = 10 folds.\\nIn this case, the validation and cross-validation methods both result in a\\nsix-variable model. However, all three approaches suggest that the four-,\\nfive-, and six-variable models are roughly equivalent in terms of their test\\nerrors.\\nIn fact, the estimated test error curves displayed in the center and right-\\nhand panels of Figure 6.3 are quite flat. While a three-variable model clearly\\nhas lower estimated test error than a two-variable model, the estimated test\\nerrors of the 3- to 11-variable models are quite similar. Furthermore, if we 240\\n6. Linear Model Selection and Regularization\\nrepeated the validation set approach using a different split of the data into\\na training set and a validation set, or if we repeated cross-validation using\\na different set of cross-validation folds, then the precise model with the\\nlowest estimated test error would surely change. In this setting, we can\\nselect a model using the one-standard-error rule. We first calculate the one-\\nstandard-\\nerror\\nrule\\nstandard error of the estimated test MSE for each model size, and then\\nselect the smallest model for which the estimated test error is within one\\nstandard error of the lowest point on the curve. The rationale here is that\\nif a set of models appear to be more or less equally good, then we might\\nas well choose the simplest model—that is, the model with the smallest\\nnumber of predictors. In this case, applying the one-standard-error rule\\nto the validation set or cross-validation approach leads to selection of the\\nthree-variable model.\\n6.2\\nShrinkage Methods\\nThe subset selection methods described in Section 6.1 involve using least\\nsquares to fit a linear model that contains a subset of the predictors. As an\\nalternative, we can fit a model containing all p predictors using a technique\\nthat constrains or regularizes the coefficient estimates, or equivalently, that\\nshrinks the coefficient estimates towards zero. It may not be immediately\\nobvious why such a constraint should improve the fit, but it turns out that\\nshrinking the coefficient estimates can significantly reduce their variance.\\nThe two best-known techniques for shrinking the regression coefficients\\ntowards zero are ridge regression and the lasso.\\n6.2.1\\nRidge Regression\\nRecall from Chapter 3 that the least squares fitting procedure estimates\\nβ0, β1, . . . , βp using the values that minimize\\nRSS =\\nn\\n0\\ni=1\\n\\uf8eb\\n\\uf8edyi −β0 −\\np\\n0\\nj=1\\nβjxij\\n\\uf8f6\\n\\uf8f8\\n2\\n.\\nRidge regression is very similar to least squares, except that the coefficients ridge\\nregression\\nare estimated by minimizing a slightly different quantity. In particular, the\\nridge regression coefficient estimates ˆβR are the values that minimize\\nn\\n0\\ni=1\\n\\uf8eb\\n\\uf8edyi −β0 −\\np\\n0\\nj=1\\nβjxij\\n\\uf8f6\\n\\uf8f8\\n2\\n+ λ\\np\\n0\\nj=1\\nβ2\\nj = RSS + λ\\np\\n0\\nj=1\\nβ2\\nj ,\\n(6.5)\\nwhere λ ≥0 is a tuning parameter, to be determined separately. Equa- tuning\\nparameter\\ntion 6.5 trades off two different criteria. As with least squares, ridge regres-\\nsion seeks coefficient estimates that fit the data well, by making the RSS\\nsmall. However, the second term, λ )\\nj β2\\nj , called a shrinkage penalty, is shrinkage\\npenalty\\nsmall when β1, . . . , βp are close to zero, and so it has the effect of shrinking\\nthe estimates of βj towards zero. The tuning parameter λ serves to control 6.2 Shrinkage Methods\\n241\\n1e−02\\n1e+00\\n1e+02\\n1e+04\\n−300\\n−100\\n0\\n100\\n200\\n300\\n400\\nStandardized Coefficients\\nIncome\\nLimit\\nRating\\nStudent\\n0.0\\n0.2\\n0.4\\n0.6\\n0.8\\n1.0\\n−300\\n−100\\n0\\n100\\n200\\n300\\n400\\nStandardized Coefficients\\nλ\\n∥ˆβR\\nλ ∥2/∥ˆβ∥2\\nFIGURE 6.4. The standardized ridge regression coefficients are displayed for\\nthe Credit data set, as a function of λ and ∥ˆβR\\nλ ∥2/∥ˆβ∥2.\\nthe relative impact of these two terms on the regression coefficient esti-\\nmates. When λ = 0, the penalty term has no effect, and ridge regression\\nwill produce the least squares estimates. However, as λ →∞, the impact of\\nthe shrinkage penalty grows, and the ridge regression coefficient estimates\\nwill approach zero. Unlike least squares, which generates only one set of co-\\nefficient estimates, ridge regression will produce a different set of coefficient\\nestimates, ˆβR\\nλ , for each value of λ. Selecting a good value for λ is critical;\\nwe defer this discussion to Section 6.2.3, where we use cross-validation.\\nNote that in (6.5), the shrinkage penalty is applied to β1, . . . , βp, but\\nnot to the intercept β0. We want to shrink the estimated association of\\neach variable with the response; however, we do not want to shrink the\\nintercept, which is simply a measure of the mean value of the response\\nwhen xi1 = xi2 = . . . = xip = 0. If we assume that the variables—that is,\\nthe columns of the data matrix X—have been centered to have mean zero\\nbefore ridge regression is performed, then the estimated intercept will take\\nthe form ˆβ0 = ¯y = )n\\ni=1 yi/n.\\nAn Application to the Credit Data\\nIn Figure 6.4, the ridge regression coefficient estimates for the Credit data\\nset are displayed. In the left-hand panel, each curve corresponds to the\\nridge regression coefficient estimate for one of the ten variables, plotted\\nas a function of λ. For example, the black solid line represents the ridge\\nregression estimate for the income coefficient, as λ is varied. At the extreme\\nleft-hand side of the plot, λ is essentially zero, and so the corresponding\\nridge coefficient estimates are the same as the usual least squares esti-\\nmates. But as λ increases, the ridge coefficient estimates shrink towards\\nzero. When λ is extremely large, then all of the ridge coefficient estimates\\nare basically zero; this corresponds to the null model that contains no pre-\\ndictors. In this plot, the income, limit, rating, and student variables are\\ndisplayed in distinct colors, since these variables tend to have by far the\\nlargest coefficient estimates. While the ridge coefficient estimates tend to\\ndecrease in aggregate as λ increases, individual coefficients, such as rating\\nand income, may occasionally increase as λ increases. 242\\n6. Linear Model Selection and Regularization\\nThe right-hand panel of Figure 6.4 displays the same ridge coefficient\\nestimates as the left-hand panel, but instead of displaying λ on the x-axis,\\nwe now display ∥ˆβR\\nλ ∥2/∥ˆβ∥2, where ˆβ denotes the vector of least squares\\ncoefficient estimates. The notation ∥β∥2 denotes the ℓ2 norm (pronounced ℓ2 norm\\n“ell 2”) of a vector, and is defined as ∥β∥2 =\\nG)p\\nj=1 βj\\n2. It measures the\\ndistance of β from zero. As λ increases, the ℓ2 norm of ˆβR\\nλ will always\\ndecrease, and so will ∥ˆβR\\nλ ∥2/∥ˆβ∥2. The latter quantity ranges from 1 (when\\nλ = 0, in which case the ridge regression coefficient estimate is the same\\nas the least squares estimate, and so their ℓ2 norms are the same) to 0\\n(when λ = ∞, in which case the ridge regression coefficient estimate is a\\nvector of zeros, with ℓ2 norm equal to zero). Therefore, we can think of the\\nx-axis in the right-hand panel of Figure 6.4 as the amount that the ridge\\nregression coefficient estimates have been shrunken towards zero; a small\\nvalue indicates that they have been shrunken very close to zero.\\nThe standard least squares coefficient estimates discussed in Chapter 3\\nare scale equivariant: multiplying Xj by a constant c simply leads to a scale\\nequivariant\\nscaling of the least squares coefficient estimates by a factor of 1/c. In other\\nwords, regardless of how the jth predictor is scaled, Xj ˆβj will remain the\\nsame. In contrast, the ridge regression coefficient estimates can change sub-\\nstantially when multiplying a given predictor by a constant. For instance,\\nconsider the income variable, which is measured in dollars. One could rea-\\nsonably have measured income in thousands of dollars, which would result\\nin a reduction in the observed values of income by a factor of 1,000. Now due\\nto the sum of squared coefficients term in the ridge regression formulation\\n(6.5), such a change in scale will not simply cause the ridge regression co-\\nefficient estimate for income to change by a factor of 1,000. In other words,\\nXj ˆβR\\nj,λ will depend not only on the value of λ, but also on the scaling of the\\njth predictor. In fact, the value of Xj ˆβR\\nj,λ may even depend on the scaling\\nof the other predictors! Therefore, it is best to apply ridge regression after\\nstandardizing the predictors, using the formula\\n˜xij =\\nxij\\nG\\n1\\nn\\n)n\\ni=1(xij −xj)2\\n,\\n(6.6)\\nso that they are all on the same scale. In (6.6), the denominator is the\\nestimated standard deviation of the jth predictor. Consequently, all of the\\nstandardized predictors will have a standard deviation of one. As a re-\\nsult the final fit will not depend on the scale on which the predictors are\\nmeasured. In Figure 6.4, the y-axis displays the standardized ridge regres-\\nsion coefficient estimates—that is, the coefficient estimates that result from\\nperforming ridge regression using standardized predictors.\\nWhy Does Ridge Regression Improve Over Least Squares?\\nRidge regression’s advantage over least squares is rooted in the bias-variance\\ntrade-off. As λ increases, the flexibility of the ridge regression fit decreases,\\nleading to decreased variance but increased bias. This is illustrated in the\\nleft-hand panel of Figure 6.5, using a simulated data set containing p = 45\\npredictors and n = 50 observations. The green curve in the left-hand panel 6.2 Shrinkage Methods\\n243\\n1e−01\\n1e+01\\n1e+03\\n0\\n10\\n20\\n30\\n40\\n50\\n60\\nMean Squared Error\\n0.0\\n0.2\\n0.4\\n0.6\\n0.8\\n1.0\\n0\\n10\\n20\\n30\\n40\\n50\\n60\\nMean Squared Error\\nλ\\n∥ˆβR\\nλ ∥2/∥ˆβ∥2\\nFIGURE 6.5.\\nSquared bias (black), variance (green), and test mean squared\\nerror (purple) for the ridge regression predictions on a simulated data set, as a\\nfunction of λ and ∥ˆβR\\nλ ∥2/∥ˆβ∥2. The horizontal dashed lines indicate the minimum\\npossible MSE. The purple crosses indicate the ridge regression models for which\\nthe MSE is smallest.\\nof Figure 6.5 displays the variance of the ridge regression predictions as a\\nfunction of λ. At the least squares coefficient estimates, which correspond\\nto ridge regression with λ = 0, the variance is high but there is no bias. But\\nas λ increases, the shrinkage of the ridge coefficient estimates leads to a\\nsubstantial reduction in the variance of the predictions, at the expense of a\\nslight increase in bias. Recall that the test mean squared error (MSE), plot-\\nted in purple, is closely related to the variance plus the squared bias. For\\nvalues of λ up to about 10, the variance decreases rapidly, with very little\\nincrease in bias, plotted in black. Consequently, the MSE drops consider-\\nably as λ increases from 0 to 10. Beyond this point, the decrease in variance\\ndue to increasing λ slows, and the shrinkage on the coefficients causes them\\nto be significantly underestimated, resulting in a large increase in the bias.\\nThe minimum MSE is achieved at approximately λ = 30. Interestingly,\\nbecause of its high variance, the MSE associated with the least squares\\nfit, when λ = 0, is almost as high as that of the null model for which all\\ncoefficient estimates are zero, when λ = ∞. However, for an intermediate\\nvalue of λ, the MSE is considerably lower.\\nThe right-hand panel of Figure 6.5 displays the same curves as the left-\\nhand panel, this time plotted against the ℓ2 norm of the ridge regression\\ncoefficient estimates divided by the ℓ2 norm of the least squares estimates.\\nNow as we move from left to right, the fits become more flexible, and so\\nthe bias decreases and the variance increases.\\nIn general, in situations where the relationship between the response\\nand the predictors is close to linear, the least squares estimates will have\\nlow bias but may have high variance. This means that a small change in\\nthe training data can cause a large change in the least squares coefficient\\nestimates. In particular, when the number of variables p is almost as large\\nas the number of observations n, as in the example in Figure 6.5, the\\nleast squares estimates will be extremely variable. And if p > n, then the\\nleast squares estimates do not even have a unique solution, whereas ridge\\nregression can still perform well by trading off a small increase in bias for a 244\\n6. Linear Model Selection and Regularization\\nlarge decrease in variance. Hence, ridge regression works best in situations\\nwhere the least squares estimates have high variance.\\nRidge regression also has substantial computational advantages over best\\nsubset selection, which requires searching through 2p models. As we dis-\\ncussed previously, even for moderate values of p, such a search can be\\ncomputationally infeasible. In contrast, for any fixed value of λ, ridge re-\\ngression only fits a single model, and the model-fitting procedure can be\\nperformed quite quickly. In fact, one can show that the computations re-\\nquired to solve (6.5), simultaneously for all values of λ, are almost identical\\nto those for fitting a model using least squares.\\n6.2.2\\nThe Lasso\\nRidge regression does have one obvious disadvantage. Unlike best subset,\\nforward stepwise, and backward stepwise selection, which will generally\\nselect models that involve just a subset of the variables, ridge regression\\nwill include all p predictors in the final model. The penalty λ ) β2\\nj in (6.5)\\nwill shrink all of the coefficients towards zero, but it will not set any of them\\nexactly to zero (unless λ = ∞). This may not be a problem for prediction\\naccuracy, but it can create a challenge in model interpretation in settings in\\nwhich the number of variables p is quite large. For example, in the Credit\\ndata set, it appears that the most important variables are income, limit,\\nrating, and student. So we might wish to build a model including just\\nthese predictors. However, ridge regression will always generate a model\\ninvolving all ten predictors. Increasing the value of λ will tend to reduce\\nthe magnitudes of the coefficients, but will not result in exclusion of any of\\nthe variables.\\nThe lasso is a relatively recent alternative to ridge regression that over- lasso\\ncomes this disadvantage. The lasso coefficients, ˆβL\\nλ , minimize the quantity\\nn\\n0\\ni=1\\n\\uf8eb\\n\\uf8edyi −β0 −\\np\\n0\\nj=1\\nβjxij\\n\\uf8f6\\n\\uf8f8\\n2\\n+ λ\\np\\n0\\nj=1\\n|βj| = RSS + λ\\np\\n0\\nj=1\\n|βj|.\\n(6.7)\\nComparing (6.7) to (6.5), we see that the lasso and ridge regression have\\nsimilar formulations. The only difference is that the β2\\nj term in the ridge\\nregression penalty (6.5) has been replaced by |βj| in the lasso penalty (6.7).\\nIn statistical parlance, the lasso uses an ℓ1 (pronounced “ell 1”) penalty\\ninstead of an ℓ2 penalty. The ℓ1 norm of a coefficient vector β is given by\\n∥β∥1 = ) |βj|.\\nAs with ridge regression, the lasso shrinks the coefficient estimates to-\\nwards zero. However, in the case of the lasso, the ℓ1 penalty has the effect\\nof forcing some of the coefficient estimates to be exactly equal to zero when\\nthe tuning parameter λ is sufficiently large. Hence, much like best subset se-\\nlection, the lasso performs variable selection. As a result, models generated\\nfrom the lasso are generally much easier to interpret than those produced\\nby ridge regression. We say that the lasso yields sparse models—that is, sparse\\nmodels that involve only a subset of the variables. As in ridge regression,\\nselecting a good value of λ for the lasso is critical; we defer this discussion\\nto Section 6.2.3, where we use cross-validation. 6.2 Shrinkage Methods\\n245\\n20\\n50\\n100\\n200\\n500\\n2000\\n5000\\n−200\\n0\\n100\\n200\\n300\\n400\\nStandardized Coefficients\\n0.0\\n0.2\\n0.4\\n0.6\\n0.8\\n1.0\\n−300\\n−100\\n0\\n100\\n200\\n300\\n400\\nStandardized Coefficients\\nIncome\\nLimit\\nRating\\nStudent\\nλ\\n∥ˆβL\\nλ ∥1/∥ˆβ∥1\\nFIGURE 6.6. The standardized lasso coefficients on the Credit data set are\\nshown as a function of λ and ∥ˆβL\\nλ ∥1/∥ˆβ∥1.\\nAs an example, consider the coefficient plots in Figure 6.6, which are gen-\\nerated from applying the lasso to the Credit data set. When λ = 0, then\\nthe lasso simply gives the least squares fit, and when λ becomes sufficiently\\nlarge, the lasso gives the null model in which all coefficient estimates equal\\nzero. However, in between these two extremes, the ridge regression and\\nlasso models are quite different from each other. Moving from left to right\\nin the right-hand panel of Figure 6.6, we observe that at first the lasso re-\\nsults in a model that contains only the rating predictor. Then student and\\nlimit enter the model almost simultaneously, shortly followed by income.\\nEventually, the remaining variables enter the model. Hence, depending on\\nthe value of λ, the lasso can produce a model involving any number of vari-\\nables. In contrast, ridge regression will always include all of the variables in\\nthe model, although the magnitude of the coefficient estimates will depend\\non λ.\\nAnother Formulation for Ridge Regression and the Lasso\\nOne can show that the lasso and ridge regression coefficient estimates solve\\nthe problems\\nminimize\\nβ\\n\\uf8f1\\n\\uf8f4\\n\\uf8f2\\n\\uf8f4\\n\\uf8f3\\nn\\n0\\ni=1\\n\\uf8eb\\n\\uf8edyi −β0 −\\np\\n0\\nj=1\\nβjxij\\n\\uf8f6\\n\\uf8f8\\n2\\uf8fc\\n\\uf8f4\\n\\uf8fd\\n\\uf8f4\\n\\uf8fe\\nsubject to\\np\\n0\\nj=1\\n|βj| ≤s\\n(6.8)\\nand\\nminimize\\nβ\\n\\uf8f1\\n\\uf8f4\\n\\uf8f2\\n\\uf8f4\\n\\uf8f3\\nn\\n0\\ni=1\\n\\uf8eb\\n\\uf8edyi −β0 −\\np\\n0\\nj=1\\nβjxij\\n\\uf8f6\\n\\uf8f8\\n2\\uf8fc\\n\\uf8f4\\n\\uf8fd\\n\\uf8f4\\n\\uf8fe\\nsubject to\\np\\n0\\nj=1\\nβ2\\nj ≤s,\\n(6.9)\\nrespectively. In other words, for every value of λ, there is some s such that\\nthe Equations (6.7) and (6.8) will give the same lasso coefficient estimates.\\nSimilarly, for every value of λ there is a corresponding s such that Equa-\\ntions (6.5) and (6.9) will give the same ridge regression coefficient estimates. 246\\n6. Linear Model Selection and Regularization\\nWhen p = 2, then (6.8) indicates that the lasso coefficient estimates have\\nthe smallest RSS out of all points that lie within the diamond defined by\\n|β1| + |β2| ≤s. Similarly, the ridge regression estimates have the smallest\\nRSS out of all points that lie within the circle defined by β2\\n1 + β2\\n2 ≤s.\\nWe can think of (6.8) as follows. When we perform the lasso we are trying\\nto find the set of coefficient estimates that lead to the smallest RSS, subject\\nto the constraint that there is a budget s for how large )p\\nj=1 |βj| can be.\\nWhen s is extremely large, then this budget is not very restrictive, and so\\nthe coefficient estimates can be large. In fact, if s is large enough that the\\nleast squares solution falls within the budget, then (6.8) will simply yield\\nthe least squares solution. In contrast, if s is small, then )p\\nj=1 |βj| must be\\nsmall in order to avoid violating the budget. Similarly, (6.9) indicates that\\nwhen we perform ridge regression, we seek a set of coefficient estimates\\nsuch that the RSS is as small as possible, subject to the requirement that\\n)p\\nj=1 β2\\nj not exceed the budget s.\\nThe formulations (6.8) and (6.9) reveal a close connection between the\\nlasso, ridge regression, and best subset selection. Consider the problem\\nminimize\\nβ\\n\\uf8f1\\n\\uf8f4\\n\\uf8f2\\n\\uf8f4\\n\\uf8f3\\nn\\n0\\ni=1\\n\\uf8eb\\n\\uf8edyi −β0 −\\np\\n0\\nj=1\\nβjxij\\n\\uf8f6\\n\\uf8f8\\n2\\uf8fc\\n\\uf8f4\\n\\uf8fd\\n\\uf8f4\\n\\uf8fe\\nsubject to\\np\\n0\\nj=1\\nI(βj ̸= 0) ≤s.\\n(6.10)\\nHere I(βj ̸= 0) is an indicator variable: it takes on a value of 1 if βj ̸= 0, and\\nequals zero otherwise. Then (6.10) amounts to finding a set of coefficient\\nestimates such that RSS is as small as possible, subject to the constraint\\nthat no more than s coefficients can be nonzero. The problem (6.10) is\\nequivalent to best subset selection. Unfortunately, solving (6.10) is com-\\nputationally infeasible when p is large, since it requires considering all\\n\\'p\\ns\\n(\\nmodels containing s predictors. Therefore, we can interpret ridge regression\\nand the lasso as computationally feasible alternatives to best subset selec-\\ntion that replace the intractable form of the budget in (6.10) with forms\\nthat are much easier to solve. Of course, the lasso is much more closely\\nrelated to best subset selection, since the lasso performs feature selection\\nfor s sufficiently small in (6.8), while ridge regression does not.\\nThe Variable Selection Property of the Lasso\\nWhy is it that the lasso, unlike ridge regression, results in coefficient esti-\\nmates that are exactly equal to zero? The formulations (6.8) and (6.9) can\\nbe used to shed light on the issue. Figure 6.7 illustrates the situation. The\\nleast squares solution is marked as ˆβ, while the blue diamond and circle\\nrepresent the lasso and ridge regression constraints in (6.8) and (6.9), re-\\nspectively. If s is sufficiently large, then the constraint regions will contain\\nˆβ, and so the ridge regression and lasso estimates will be the same as the\\nleast squares estimates. (Such a large value of s corresponds to λ = 0 in\\n(6.5) and (6.7).) However, in Figure 6.7 the least squares estimates lie out-\\nside of the diamond and the circle, and so the least squares estimates are\\nnot the same as the lasso and ridge regression estimates.\\nEach of the ellipses centered around ˆβ represents a contour: this means contour\\nthat all of the points on a particular ellipse have the same RSS value. As 6.2 Shrinkage Methods\\n247\\nFIGURE 6.7.\\nContours of the error and constraint functions for the lasso\\n(left) and ridge regression (right). The solid blue areas are the constraint regions,\\n|β1|+|β2| ≤s and β2\\n1 +β2\\n2 ≤s, while the red ellipses are the contours of the RSS.\\nthe ellipses expand away from the least squares coefficient estimates, the\\nRSS increases. Equations (6.8) and (6.9) indicate that the lasso and ridge\\nregression coefficient estimates are given by the first point at which an\\nellipse contacts the constraint region. Since ridge regression has a circular\\nconstraint with no sharp points, this intersection will not generally occur on\\nan axis, and so the ridge regression coefficient estimates will be exclusively\\nnon-zero. However, the lasso constraint has corners at each of the axes, and\\nso the ellipse will often intersect the constraint region at an axis. When this\\noccurs, one of the coefficients will equal zero. In higher dimensions, many of\\nthe coefficient estimates may equal zero simultaneously. In Figure 6.7, the\\nintersection occurs at β1 = 0, and so the resulting model will only include\\nβ2.\\nIn Figure 6.7, we considered the simple case of p = 2. When p = 3,\\nthen the constraint region for ridge regression becomes a sphere, and the\\nconstraint region for the lasso becomes a polyhedron. When p > 3, the\\nconstraint for ridge regression becomes a hypersphere, and the constraint\\nfor the lasso becomes a polytope. However, the key ideas depicted in Fig-\\nure 6.7 still hold. In particular, the lasso leads to feature selection when\\np > 2 due to the sharp corners of the polyhedron or polytope.\\nComparing the Lasso and Ridge Regression\\nIt is clear that the lasso has a major advantage over ridge regression, in\\nthat it produces simpler and more interpretable models that involve only a\\nsubset of the predictors. However, which method leads to better prediction\\naccuracy? Figure 6.8 displays the variance, squared bias, and test MSE of\\nthe lasso applied to the same simulated data as in Figure 6.5. Clearly the\\nlasso leads to qualitatively similar behavior to ridge regression, in that as λ\\nincreases, the variance decreases and the bias increases. In the right-hand 248\\n6. Linear Model Selection and Regularization\\n0.02\\n0.10\\n0.50\\n2.00\\n10.00\\n50.00\\n0\\n10\\n20\\n30\\n40\\n50\\n60\\nMean Squared Error\\n0.0\\n0.2\\n0.4\\n0.6\\n0.8\\n1.0\\n0\\n10\\n20\\n30\\n40\\n50\\n60\\nR2 on Training Data\\nMean Squared Error\\nλ\\nFIGURE 6.8.\\nLeft: Plots of squared bias (black), variance (green), and test\\nMSE (purple) for the lasso on a simulated data set. Right: Comparison of squared\\nbias, variance, and test MSE between lasso (solid) and ridge (dotted). Both are\\nplotted against their R2 on the training data, as a common form of indexing. The\\ncrosses in both plots indicate the lasso model for which the MSE is smallest.\\npanel of Figure 6.8, the dotted lines represent the ridge regression fits.\\nHere we plot both against their R2 on the training data. This is another\\nuseful way to index models, and can be used to compare models with\\ndifferent types of regularization, as is the case here. In this example, the\\nlasso and ridge regression result in almost identical biases. However, the\\nvariance of ridge regression is slightly lower than the variance of the lasso.\\nConsequently, the minimum MSE of ridge regression is slightly smaller than\\nthat of the lasso.\\nHowever, the data in Figure 6.8 were generated in such a way that all 45\\npredictors were related to the response—that is, none of the true coefficients\\nβ1, . . . , β45 equaled zero. The lasso implicitly assumes that a number of the\\ncoefficients truly equal zero. Consequently, it is not surprising that ridge\\nregression outperforms the lasso in terms of prediction error in this setting.\\nFigure 6.9 illustrates a similar situation, except that now the response is a\\nfunction of only 2 out of 45 predictors. Now the lasso tends to outperform\\nridge regression in terms of bias, variance, and MSE.\\nThese two examples illustrate that neither ridge regression nor the lasso\\nwill universally dominate the other. In general, one might expect the lasso\\nto perform better in a setting where a relatively small number of predictors\\nhave substantial coefficients, and the remaining predictors have coefficients\\nthat are very small or that equal zero. Ridge regression will perform better\\nwhen the response is a function of many predictors, all with coefficients of\\nroughly equal size. However, the number of predictors that is related to the\\nresponse is never known a priori for real data sets. A technique such as\\ncross-validation can be used in order to determine which approach is better\\non a particular data set.\\nAs with ridge regression, when the least squares estimates have exces-\\nsively high variance, the lasso solution can yield a reduction in variance\\nat the expense of a small increase in bias, and consequently can gener-\\nate more accurate predictions. Unlike ridge regression, the lasso performs\\nvariable selection, and hence results in models that are easier to interpret. 6.2 Shrinkage Methods\\n249\\n0.02\\n0.10\\n0.50\\n2.00\\n10.00\\n50.00\\n0\\n20\\n40\\n60\\n80\\n100\\nMean Squared Error\\n0.4\\n0.5\\n0.6\\n0.7\\n0.8\\n0.9\\n1.0\\n0\\n20\\n40\\n60\\n80\\n100\\nR2 on Training Data\\nMean Squared Error\\nλ\\nFIGURE 6.9.\\nLeft: Plots of squared bias (black), variance (green), and test\\nMSE (purple) for the lasso. The simulated data is similar to that in Figure 6.8,\\nexcept that now only two predictors are related to the response. Right: Comparison\\nof squared bias, variance, and test MSE between lasso (solid) and ridge (dotted).\\nBoth are plotted against their R2 on the training data, as a common form of\\nindexing. The crosses in both plots indicate the lasso model for which the MSE is\\nsmallest.\\nThere are very efficient algorithms for fitting both ridge and lasso models;\\nin both cases the entire coefficient paths can be computed with about the\\nsame amount of work as a single least squares fit. We will explore this\\nfurther in the lab at the end of this chapter.\\nA Simple Special Case for Ridge Regression and the Lasso\\nIn order to obtain a better intuition about the behavior of ridge regression\\nand the lasso, consider a simple special case with n = p, and X a diag-\\nonal matrix with 1’s on the diagonal and 0’s in all off-diagonal elements.\\nTo simplify the problem further, assume also that we are performing regres-\\nsion without an intercept. With these assumptions, the usual least squares\\nproblem simplifies to finding β1, . . . , βp that minimize\\np\\n0\\nj=1\\n(yj −βj)2.\\n(6.11)\\nIn this case, the least squares solution is given by\\nˆβj = yj.\\nAnd in this setting, ridge regression amounts to finding β1, . . . , βp such that\\np\\n0\\nj=1\\n(yj −βj)2 + λ\\np\\n0\\nj=1\\nβ2\\nj\\n(6.12)\\nis minimized, and the lasso amounts to finding the coefficients such that\\np\\n0\\nj=1\\n(yj −βj)2 + λ\\np\\n0\\nj=1\\n|βj|\\n(6.13) 250\\n6. Linear Model Selection and Regularization\\n−1.5\\n−0.5\\n0.0\\n0.5\\n1.0\\n1.5\\n−1.5\\n−0.5\\n0.5\\n1.5\\nCoefficient Estimate\\nRidge\\nLeast Squares\\n−1.5\\n−0.5\\n0.0\\n0.5\\n1.0\\n1.5\\n−1.5\\n−0.5\\n0.5\\n1.5\\nCoefficient Estimate\\nLasso\\nLeast Squares\\nyj\\nyj\\nFIGURE 6.10. The ridge regression and lasso coefficient estimates for a simple\\nsetting with n = p and X a diagonal matrix with 1’s on the diagonal. Left: The\\nridge regression coefficient estimates are shrunken proportionally towards zero,\\nrelative to the least squares estimates. Right: The lasso coefficient estimates are\\nsoft-thresholded towards zero.\\nis minimized. One can show that in this setting, the ridge regression esti-\\nmates take the form\\nˆβR\\nj = yj/(1 + λ),\\n(6.14)\\nand the lasso estimates take the form\\nˆβL\\nj =\\n\\uf8f1\\n\\uf8f4\\n\\uf8f2\\n\\uf8f4\\n\\uf8f3\\nyj −λ/2\\nif yj > λ/2;\\nyj + λ/2\\nif yj < −λ/2;\\n0\\nif |yj| ≤λ/2.\\n(6.15)\\nFigure 6.10 displays the situation. We can see that ridge regression and\\nthe lasso perform two very different types of shrinkage. In ridge regression,\\neach least squares coefficient estimate is shrunken by the same proportion.\\nIn contrast, the lasso shrinks each least squares coefficient towards zero by\\na constant amount, λ/2; the least squares coefficients that are less than\\nλ/2 in absolute value are shrunken entirely to zero. The type of shrink-\\nage performed by the lasso in this simple setting (6.15) is known as soft-\\nthresholding. The fact that some lasso coefficients are shrunken entirely to soft-\\nthresholding\\nzero explains why the lasso performs feature selection.\\nIn the case of a more general data matrix X, the story is a little more\\ncomplicated than what is depicted in Figure 6.10, but the main ideas still\\nhold approximately: ridge regression more or less shrinks every dimension\\nof the data by the same proportion, whereas the lasso more or less shrinks\\nall coefficients toward zero by a similar amount, and sufficiently small co-\\nefficients are shrunken all the way to zero.\\nBayesian Interpretation of Ridge Regression and the Lasso\\nWe now show that one can view ridge regression and the lasso through\\na Bayesian lens. A Bayesian viewpoint for regression assumes that the\\ncoefficient vector β has some prior distribution, say p(β), where β =\\n(β0, β1, . . . , βp)T . The likelihood of the data can be written as f(Y |X, β), 6.2 Shrinkage Methods\\n251\\n−3\\n−2\\n−1\\n0\\n1\\n2\\n3\\n0.0\\n0.1\\n0.2\\n0.3\\n0.4\\n0.5\\n0.6\\n0.7\\n−3\\n−2\\n−1\\n0\\n1\\n2\\n3\\n0.0\\n0.1\\n0.2\\n0.3\\n0.4\\n0.5\\n0.6\\n0.7\\nβj\\nβj\\ng(βj)\\ng(βj)\\nFIGURE 6.11. Left: Ridge regression is the posterior mode for β under a Gaus-\\nsian prior. Right: The lasso is the posterior mode for β under a double-exponential\\nprior.\\nwhere X = (X1, . . . , Xp). Multiplying the prior distribution by the likeli-\\nhood gives us (up to a proportionality constant) the posterior distribution, posterior\\ndistribution\\nwhich takes the form\\np(β|X, Y ) ∝f(Y |X, β)p(β|X) = f(Y |X, β)p(β),\\nwhere the proportionality above follows from Bayes’ theorem, and the\\nequality above follows from the assumption that X is fixed.\\nWe assume the usual linear model,\\nY = β0 + X1β1 + · · · + Xpβp + ϵ,\\nand suppose that the errors are independent and drawn from a normal dis-\\ntribution. Furthermore, assume that p(β) = Fp\\nj=1 g(βj), for some density\\nfunction g. It turns out that ridge regression and the lasso follow naturally\\nfrom two special cases of g:\\n• If g is a Gaussian distribution with mean zero and standard deviation\\na function of λ, then it follows that the posterior mode for β—that posterior\\nmode\\nis, the most likely value for β, given the data—is given by the ridge\\nregression solution. (In fact, the ridge regression solution is also the\\nposterior mean.)\\n• If g is a double-exponential (Laplace) distribution with mean zero\\nand scale parameter a function of λ, then it follows that the posterior\\nmode for β is the lasso solution. (However, the lasso solution is not\\nthe posterior mean, and in fact, the posterior mean does not yield a\\nsparse coefficient vector.)\\nThe Gaussian and double-exponential priors are displayed in Figure 6.11.\\nTherefore, from a Bayesian viewpoint, ridge regression and the lasso follow\\ndirectly from assuming the usual linear model with normal errors, together\\nwith a simple prior distribution for β. Notice that the lasso prior is steeply\\npeaked at zero, while the Gaussian is flatter and fatter at zero. Hence, the\\nlasso expects a priori that many of the coefficients are (exactly) zero, while\\nridge assumes the coefficients are randomly distributed about zero. 252\\n6. Linear Model Selection and Regularization\\n5e−03\\n5e−02\\n5e−01\\n5e+00\\n25.0\\n25.2\\n25.4\\n25.6\\nCross−Validation Error\\n5e−03\\n5e−02\\n5e−01\\n5e+00\\n−300\\n−100\\n0\\n100\\n300\\nStandardized Coefficients\\nλ\\nλ\\nFIGURE 6.12.\\nLeft: Cross-validation errors that result from applying ridge\\nregression to the Credit data set with various values of λ. Right: The coefficient\\nestimates as a function of λ. The vertical dashed lines indicate the value of λ\\nselected by cross-validation.\\n6.2.3\\nSelecting the Tuning Parameter\\nJust as the subset selection approaches considered in Section 6.1 require\\na method to determine which of the models under consideration is best,\\nimplementing ridge regression and the lasso requires a method for selecting\\na value for the tuning parameter λ in (6.5) and (6.7), or equivalently, the\\nvalue of the constraint s in (6.9) and (6.8). Cross-validation provides a sim-\\nple way to tackle this problem. We choose a grid of λ values, and compute\\nthe cross-validation error for each value of λ, as described in Chapter 5. We\\nthen select the tuning parameter value for which the cross-validation error\\nis smallest. Finally, the model is re-fit using all of the available observations\\nand the selected value of the tuning parameter.\\nFigure 6.12 displays the choice of λ that results from performing leave-\\none-out cross-validation on the ridge regression fits from the Credit data\\nset. The dashed vertical lines indicate the selected value of λ. In this case\\nthe value is relatively small, indicating that the optimal fit only involves a\\nsmall amount of shrinkage relative to the least squares solution. In addition,\\nthe dip is not very pronounced, so there is rather a wide range of values\\nthat would give a very similar error. In a case like this we might simply use\\nthe least squares solution.\\nFigure 6.13 provides an illustration of ten-fold cross-validation applied to\\nthe lasso fits on the sparse simulated data from Figure 6.9. The left-hand\\npanel of Figure 6.13 displays the cross-validation error, while the right-hand\\npanel displays the coefficient estimates. The vertical dashed lines indicate\\nthe point at which the cross-validation error is smallest. The two colored\\nlines in the right-hand panel of Figure 6.13 represent the two predictors\\nthat are related to the response, while the grey lines represent the unre-\\nlated predictors; these are often referred to as signal and noise variables, signal\\nrespectively. Not only has the lasso correctly given much larger coeffi-\\ncient estimates to the two signal predictors, but also the minimum cross-\\nvalidation error corresponds to a set of coefficient estimates for which only\\nthe signal variables are non-zero. Hence cross-validation together with the\\nlasso has correctly identified the two signal variables in the model, even\\nthough this is a challenging setting, with p = 45 variables and only n = 50 6.3 Dimension Reduction Methods\\n253\\n0.0\\n0.2\\n0.4\\n0.6\\n0.8\\n1.0\\n0\\n200\\n600\\n1000\\n1400\\nCross−Validation Error\\n0.0\\n0.2\\n0.4\\n0.6\\n0.8\\n1.0\\n−5\\n0\\n5\\n10\\n15\\nStandardized Coefficients\\n∥ˆβL\\nλ ∥1/∥ˆβ∥1\\n∥ˆβL\\nλ ∥1/∥ˆβ∥1\\nFIGURE 6.13.\\nLeft: Ten-fold cross-validation MSE for the lasso, applied to\\nthe sparse simulated data set from Figure 6.9. Right: The corresponding lasso\\ncoefficient estimates are displayed. The two signal variables are shown in color,\\nand the noise variables are in gray. The vertical dashed lines indicate the lasso\\nfit for which the cross-validation error is smallest.\\nobservations. In contrast, the least squares solution—displayed on the far\\nright of the right-hand panel of Figure 6.13—assigns a large coefficient\\nestimate to only one of the two signal variables.\\n6.3\\nDimension Reduction Methods\\nThe methods that we have discussed so far in this chapter have controlled\\nvariance in two different ways, either by using a subset of the original vari-\\nables, or by shrinking their coefficients toward zero. All of these methods\\nare defined using the original predictors, X1, X2, . . . , Xp. We now explore\\na class of approaches that transform the predictors and then fit a least\\nsquares model using the transformed variables. We will refer to these tech-\\nniques as dimension reduction methods.\\ndimension\\nreduction\\nLet Z1, Z2, . . . , ZM represent M < p linear combinations of our original\\nlinear\\ncombination\\np predictors. That is,\\nZm =\\np\\n0\\nj=1\\nφjmXj\\n(6.16)\\nfor some constants φ1m, φ2m . . . , φpm, m = 1, . . . , M. We can then fit the\\nlinear regression model\\nyi = θ0 +\\nM\\n0\\nm=1\\nθmzim + ϵi,\\ni = 1, . . . , n,\\n(6.17)\\nusing least squares. Note that in (6.17), the regression coefficients are given\\nby θ0, θ1, . . . , θM. If the constants φ1m, φ2m, . . . , φpm are chosen wisely, then\\nsuch dimension reduction approaches can often outperform least squares\\nregression. In other words, fitting (6.17) using least squares can lead to\\nbetter results than fitting (6.1) using least squares.\\nThe term dimension reduction comes from the fact that this approach\\nreduces the problem of estimating the p+1 coefficients β0, β1, . . . , βp to the 254\\n6. Linear Model Selection and Regularization\\n10\\n20\\n30\\n40\\n50\\n60\\n70\\n0\\n5\\n10\\n15\\n20\\n25\\n30\\n35\\nPopulation\\nAd Spending\\nFIGURE 6.14. The population size (pop) and ad spending (ad) for 100 different\\ncities are shown as purple circles. The green solid line indicates the first principal\\ncomponent, and the blue dashed line indicates the second principal component.\\nsimpler problem of estimating the M + 1 coefficients θ0, θ1, . . . , θM, where\\nM < p. In other words, the dimension of the problem has been reduced\\nfrom p + 1 to M + 1.\\nNotice that from (6.16),\\nM\\n0\\nm=1\\nθmzim =\\nM\\n0\\nm=1\\nθm\\np\\n0\\nj=1\\nφjmxij =\\np\\n0\\nj=1\\nM\\n0\\nm=1\\nθmφjmxij =\\np\\n0\\nj=1\\nβjxij,\\nwhere\\nβj =\\nM\\n0\\nm=1\\nθmφjm.\\n(6.18)\\nHence (6.17) can be thought of as a special case of the original linear\\nregression model given by (6.1). Dimension reduction serves to constrain\\nthe estimated βj coefficients, since now they must take the form (6.18).\\nThis constraint on the form of the coefficients has the potential to bias the\\ncoefficient estimates. However, in situations where p is large relative to n,\\nselecting a value of M ≪p can significantly reduce the variance of the fitted\\ncoefficients. If M = p, and all the Zm are linearly independent, then (6.18)\\nposes no constraints. In this case, no dimension reduction occurs, and so\\nfitting (6.17) is equivalent to performing least squares on the original p\\npredictors.\\nAll dimension reduction methods work in two steps. First, the trans-\\nformed predictors Z1, Z2, . . . , ZM are obtained. Second, the model is fit\\nusing these M predictors. However, the choice of Z1, Z2, . . . , ZM, or equiv-\\nalently, the selection of the φjm’s, can be achieved in different ways. In this\\nchapter, we will consider two approaches for this task: principal components\\nand partial least squares.\\n6.3.1\\nPrincipal Components Regression\\nPrincipal components analysis (PCA) is a popular approach for deriving principal\\ncomponents\\nanalysis 6.3 Dimension Reduction Methods\\n255\\na low-dimensional set of features from a large set of variables. PCA is\\ndiscussed in greater detail as a tool for unsupervised learning in Chapter 12.\\nHere we describe its use as a dimension reduction technique for regression.\\nAn Overview of Principal Components Analysis\\nPCA is a technique for reducing the dimension of an n × p data matrix X.\\nThe first principal component direction of the data is that along which the\\nobservations vary the most. For instance, consider Figure 6.14, which shows\\npopulation size (pop) in tens of thousands of people, and ad spending for a\\nparticular company (ad) in thousands of dollars, for 100 cities.6 The green\\nsolid line represents the first principal component direction of the data. We\\ncan see by eye that this is the direction along which there is the greatest\\nvariability in the data. That is, if we projected the 100 observations onto\\nthis line (as shown in the left-hand panel of Figure 6.15), then the resulting\\nprojected observations would have the largest possible variance; projecting\\nthe observations onto any other line would yield projected observations\\nwith lower variance. Projecting a point onto a line simply involves finding\\nthe location on the line which is closest to the point.\\nThe first principal component is displayed graphically in Figure 6.14, but\\nhow can it be summarized mathematically? It is given by the formula\\nZ1 = 0.839 × (pop −pop) + 0.544 × (ad −ad).\\n(6.19)\\nHere φ11 = 0.839 and φ21 = 0.544 are the principal component loadings,\\nwhich define the direction referred to above. In (6.19), pop indicates the\\nmean of all pop values in this data set, and ad indicates the mean of all ad-\\nvertising spending. The idea is that out of every possible linear combination\\nof pop and ad such that φ2\\n11 + φ2\\n21 = 1, this particular linear combination\\nyields the highest variance: i.e. this is the linear combination for which\\nVar(φ11 × (pop −pop) + φ21 × (ad −ad)) is maximized. It is necessary to\\nconsider only linear combinations of the form φ2\\n11+φ2\\n21 = 1, since otherwise\\nwe could increase φ11 and φ21 arbitrarily in order to blow up the variance.\\nIn (6.19), the two loadings are both positive and have similar size, and so\\nZ1 is almost an average of the two variables.\\nSince n = 100, pop and ad are vectors of length 100, and so is Z1 in\\n(6.19). For instance,\\nzi1 = 0.839 × (popi −pop) + 0.544 × (adi −ad).\\n(6.20)\\nThe values of z11, . . . , zn1 are known as the principal component scores, and\\ncan be seen in the right-hand panel of Figure 6.15.\\nThere is also another interpretation of PCA: the first principal compo-\\nnent vector defines the line that is as close as possible to the data. For\\ninstance, in Figure 6.14, the first principal component line minimizes the\\nsum of the squared perpendicular distances between each point and the\\nline. These distances are plotted as dashed line segments in the left-hand\\n6This dataset is distinct from the Advertising data discussed in Chapter 3. 256\\n6. Linear Model Selection and Regularization\\n20\\n30\\n40\\n50\\n5\\n10\\n15\\n20\\n25\\n30\\nPopulation\\nAd Spending\\n−20\\n−10\\n0\\n10\\n20\\n−10\\n−5\\n0\\n5\\n10\\n1st Principal Component\\n2nd Principal Component\\nFIGURE 6.15. A subset of the advertising data. The mean pop and ad budgets\\nare indicated with a blue circle. Left: The first principal component direction is\\nshown in green. It is the dimension along which the data vary the most, and it also\\ndefines the line that is closest to all n of the observations. The distances from each\\nobservation to the principal component are represented using the black dashed line\\nsegments. The blue dot represents (pop, ad). Right: The left-hand panel has been\\nrotated so that the first principal component direction coincides with the x-axis.\\npanel of Figure 6.15, in which the crosses represent the projection of each\\npoint onto the first principal component line. The first principal component\\nhas been chosen so that the projected observations are as close as possible\\nto the original observations.\\nIn the right-hand panel of Figure 6.15, the left-hand panel has been\\nrotated so that the first principal component direction coincides with the\\nx-axis. It is possible to show that the first principal component score for\\nthe ith observation, given in (6.20), is the distance in the x-direction of the\\nith cross from zero. So for example, the point in the bottom-left corner of\\nthe left-hand panel of Figure 6.15 has a large negative principal component\\nscore, zi1 = −26.1, while the point in the top-right corner has a large\\npositive score, zi1 = 18.7. These scores can be computed directly using\\n(6.20).\\nWe can think of the values of the principal component Z1 as single-\\nnumber summaries of the joint pop and ad budgets for each location. In\\nthis example, if zi1 = 0.839 × (popi −pop) + 0.544 × (adi −ad) < 0,\\nthen this indicates a city with below-average population size and below-\\naverage ad spending. A positive score suggests the opposite. How well can a\\nsingle number represent both pop and ad? In this case, Figure 6.14 indicates\\nthat pop and ad have approximately a linear relationship, and so we might\\nexpect that a single-number summary will work well. Figure 6.16 displays\\nzi1 versus both pop and ad.7 The plots show a strong relationship between\\nthe first principal component and the two features. In other words, the first\\nprincipal component appears to capture most of the information contained\\nin the pop and ad predictors.\\nSo far we have concentrated on the first principal component. In gen-\\neral, one can construct up to p distinct principal components. The second\\n7The principal components were calculated after first standardizing both pop and ad,\\na common approach. Hence, the x-axes on Figures 6.15 and 6.16 are not on the same\\nscale. 6.3 Dimension Reduction Methods\\n257\\n−3\\n−2\\n−1\\n0\\n1\\n2\\n3\\n20\\n30\\n40\\n50\\n60\\n1st Principal Component\\nPopulation\\n−3\\n−2\\n−1\\n0\\n1\\n2\\n3\\n5\\n10\\n15\\n20\\n25\\n30\\n1st Principal Component\\nAd Spending\\nFIGURE 6.16. Plots of the first principal component scores zi1 versus pop and\\nad. The relationships are strong.\\nprincipal component Z2 is a linear combination of the variables that is un-\\ncorrelated with Z1, and has largest variance subject to this constraint. The\\nsecond principal component direction is illustrated as a dashed blue line in\\nFigure 6.14. It turns out that the zero correlation condition of Z1 with Z2\\nis equivalent to the condition that the direction must be perpendicular, or perpen-\\ndicular\\northogonal, to the first principal component direction. The second principal\\northogonal\\ncomponent is given by the formula\\nZ2 = 0.544 × (pop −pop) −0.839 × (ad −ad).\\nSince the advertising data has two predictors, the first two principal com-\\nponents contain all of the information that is in pop and ad. However, by\\nconstruction, the first component will contain the most information. Con-\\nsider, for example, the much larger variability of zi1 (the x-axis) versus\\nzi2 (the y-axis) in the right-hand panel of Figure 6.15. The fact that the\\nsecond principal component scores are much closer to zero indicates that\\nthis component captures far less information. As another illustration, Fig-\\nure 6.17 displays zi2 versus pop and ad. There is little relationship between\\nthe second principal component and these two predictors, again suggesting\\nthat in this case, one only needs the first principal component in order to\\naccurately represent the pop and ad budgets.\\nWith two-dimensional data, such as in our advertising example, we can\\nconstruct at most two principal components. However, if we had other\\npredictors, such as population age, income level, education, and so forth,\\nthen additional components could be constructed. They would successively\\nmaximize variance, subject to the constraint of being uncorrelated with the\\npreceding components.\\nThe Principal Components Regression Approach\\nThe principal components regression (PCR) approach involves construct- principal\\ncomponents\\nregression\\ning the first M principal components, Z1, . . . , ZM, and then using these\\ncomponents as the predictors in a linear regression model that is fit us-\\ning least squares. The key idea is that often a small number of principal\\ncomponents suffice to explain most of the variability in the data, as well\\nas the relationship with the response. In other words, we assume that the 258\\n6. Linear Model Selection and Regularization\\n−1.0\\n−0.5\\n0.0\\n0.5\\n1.0\\n20\\n30\\n40\\n50\\n60\\n2nd Principal Component\\nPopulation\\n−1.0\\n−0.5\\n0.0\\n0.5\\n1.0\\n5\\n10\\n15\\n20\\n25\\n30\\n2nd Principal Component\\nAd Spending\\nFIGURE 6.17. Plots of the second principal component scores zi2 versus pop\\nand ad. The relationships are weak.\\n0\\n10\\n20\\n30\\n40\\n0\\n10\\n20\\n30\\n40\\n50\\n60\\n70\\nNumber of Components\\nMean Squared Error\\n0\\n10\\n20\\n30\\n40\\n0\\n50\\n100\\n150\\nNumber of Components\\nMean Squared Error\\nSquared Bias\\nTest MSE\\nVariance\\nFIGURE 6.18. PCR was applied to two simulated data sets. In each panel, the\\nhorizontal dashed line represents the irreducible error. Left: Simulated data from\\nFigure 6.8. Right: Simulated data from Figure 6.9.\\ndirections in which X1, . . . , Xp show the most variation are the directions\\nthat are associated with Y . While this assumption is not guaranteed to be\\ntrue, it often turns out to be a reasonable enough approximation to give\\ngood results.\\nIf the assumption underlying PCR holds, then fitting a least squares\\nmodel to Z1, . . . , ZM will lead to better results than fitting a least squares\\nmodel to X1, . . . , Xp, since most or all of the information in the data that\\nrelates to the response is contained in Z1, . . . , ZM, and by estimating only\\nM ≪p coefficients we can mitigate overfitting. In the advertising data, the\\nfirst principal component explains most of the variance in both pop and ad,\\nso a principal component regression that uses this single variable to predict\\nsome response of interest, such as sales, will likely perform quite well.\\nFigure 6.18 displays the PCR fits on the simulated data sets from Fig-\\nures 6.8 and 6.9. Recall that both data sets were generated using n = 50\\nobservations and p = 45 predictors. However, while the response in the\\nfirst data set was a function of all the predictors, the response in the sec-\\nond data set was generated using only two of the predictors. The curves are\\nplotted as a function of M, the number of principal components used as\\npredictors in the regression model. As more principal components are used 6.3 Dimension Reduction Methods\\n259\\n0\\n10\\n20\\n30\\n40\\n0\\n10\\n20\\n30\\n40\\n50\\n60\\n70\\nPCR\\nNumber of Components\\nMean Squared Error\\nSquared Bias\\nTest MSE\\nVariance\\n0.0\\n0.2\\n0.4\\n0.6\\n0.8\\n1.0\\n0\\n10\\n20\\n30\\n40\\n50\\n60\\n70\\nRidge Regression and Lasso\\nShrinkage Factor\\nMean Squared Error\\nFIGURE 6.19. PCR, ridge regression, and the lasso were applied to a simu-\\nlated data set in which the first five principal components of X contain all the\\ninformation about the response Y . In each panel, the irreducible error Var(ϵ) is\\nshown as a horizontal dashed line. Left: Results for PCR. Right: Results for lasso\\n(solid) and ridge regression (dotted). The x-axis displays the shrinkage factor\\nof the coefficient estimates, defined as the ℓ2 norm of the shrunken coefficient\\nestimates divided by the ℓ2 norm of the least squares estimate.\\nin the regression model, the bias decreases, but the variance increases. This\\nresults in a typical U-shape for the mean squared error. When M = p = 45,\\nthen PCR amounts simply to a least squares fit using all of the original\\npredictors. The figure indicates that performing PCR with an appropriate\\nchoice of M can result in a substantial improvement over least squares, es-\\npecially in the left-hand panel. However, by examining the ridge regression\\nand lasso results in Figures 6.5, 6.8, and 6.9, we see that PCR does not\\nperform as well as the two shrinkage methods in this example.\\nThe relatively worse performance of PCR in Figure 6.18 is a consequence\\nof the fact that the data were generated in such a way that many princi-\\npal components are required in order to adequately model the response.\\nIn contrast, PCR will tend to do well in cases when the first few principal\\ncomponents are sufficient to capture most of the variation in the predictors\\nas well as the relationship with the response. The left-hand panel of Fig-\\nure 6.19 illustrates the results from another simulated data set designed to\\nbe more favorable to PCR. Here the response was generated in such a way\\nthat it depends exclusively on the first five principal components. Now the\\nbias drops to zero rapidly as M, the number of principal components used\\nin PCR, increases. The mean squared error displays a clear minimum at\\nM = 5. The right-hand panel of Figure 6.19 displays the results on these\\ndata using ridge regression and the lasso. All three methods offer a signif-\\nicant improvement over least squares. However, PCR and ridge regression\\nslightly outperform the lasso.\\nWe note that even though PCR provides a simple way to perform re-\\ngression using M < p predictors, it is not a feature selection method. This\\nis because each of the M principal components used in the regression is a\\nlinear combination of all p of the original features. For instance, in (6.19),\\nZ1 was a linear combination of both pop and ad. Therefore, while PCR of-\\nten performs quite well in many practical settings, it does not result in the 260\\n6. Linear Model Selection and Regularization\\n2\\n4\\n6\\n8\\n10\\n−300\\n−100\\n0\\n100\\n200\\n300\\n400\\nNumber of Components\\nStandardized Coefficients\\nIncome\\nLimit\\nRating\\nStudent\\n2\\n4\\n6\\n8\\n10\\n20000\\n40000\\n60000\\n80000\\nNumber of Components\\nCross−Validation MSE\\nFIGURE 6.20. Left: PCR standardized coefficient estimates on the Credit data\\nset for different values of M. Right: The ten-fold cross-validation MSE obtained\\nusing PCR, as a function of M.\\ndevelopment of a model that relies upon a small set of the original features.\\nIn this sense, PCR is more closely related to ridge regression than to the\\nlasso. In fact, one can show that PCR and ridge regression are very closely\\nrelated. One can even think of ridge regression as a continuous version of\\nPCR!8\\nIn PCR, the number of principal components, M, is typically chosen by\\ncross-validation. The results of applying PCR to the Credit data set are\\nshown in Figure 6.20; the right-hand panel displays the cross-validation er-\\nrors obtained, as a function of M. On these data, the lowest cross-validation\\nerror occurs when there are M = 10 components; this corresponds to al-\\nmost no dimension reduction at all, since PCR with M = 11 is equivalent\\nto simply performing least squares.\\nWhen performing PCR, we generally recommend standardizing each pre-\\ndictor, using (6.6), prior to generating the principal components. This stan-\\ndardization ensures that all variables are on the same scale. In the absence\\nof standardization, the high-variance variables will tend to play a larger\\nrole in the principal components obtained, and the scale on which the vari-\\nables are measured will ultimately have an effect on the final PCR model.\\nHowever, if the variables are all measured in the same units (say, kilograms,\\nor inches), then one might choose not to standardize them.\\n6.3.2\\nPartial Least Squares\\nThe PCR approach that we just described involves identifying linear combi-\\nnations, or directions, that best represent the predictors X1, . . . , Xp. These\\ndirections are identified in an unsupervised way, since the response Y is\\nnot used to help determine the principal component directions. That is,\\nthe response does not supervise the identification of the principal compo-\\nnents. Consequently, PCR suffers from a drawback: there is no guarantee\\n8More details can be found in Section 3.5 of The Elements of Statistical Learning by\\nHastie, Tibshirani, and Friedman. 6.3 Dimension Reduction Methods\\n261\\n20\\n30\\n40\\n50\\n60\\n5\\n10\\n15\\n20\\n25\\n30\\nPopulation\\nAd Spending\\nFIGURE 6.21.\\nFor the advertising data, the first PLS direction (solid line)\\nand first PCR direction (dotted line) are shown.\\nthat the directions that best explain the predictors will also be the best\\ndirections to use for predicting the response. Unsupervised methods are\\ndiscussed further in Chapter 12.\\nWe now present partial least squares (PLS), a supervised alternative to partial least\\nsquares\\nPCR. Like PCR, PLS is a dimension reduction method, which first identifies\\na new set of features Z1, . . . , ZM that are linear combinations of the original\\nfeatures, and then fits a linear model via least squares using these M new\\nfeatures. But unlike PCR, PLS identifies these new features in a supervised\\nway—that is, it makes use of the response Y in order to identify new\\nfeatures that not only approximate the old features well, but also that are\\nrelated to the response. Roughly speaking, the PLS approach attempts to\\nfind directions that help explain both the response and the predictors.\\nWe now describe how the first PLS direction is computed. After stan-\\ndardizing the p predictors, PLS computes the first direction Z1 by setting\\neach φj1 in (6.16) equal to the coefficient from the simple linear regression\\nof Y onto Xj. One can show that this coefficient is proportional to the cor-\\nrelation between Y and Xj. Hence, in computing Z1 = )p\\nj=1 φj1Xj, PLS\\nplaces the highest weight on the variables that are most strongly related to\\nthe response.\\nFigure 6.21 displays an example of PLS on a synthetic dataset with Sales\\nin each of 100 regions as the response, and two predictors; Population Size\\nand Advertising Spending. The solid green line indicates the first PLS di-\\nrection, while the dotted line shows the first principal component direction.\\nPLS has chosen a direction that has less change in the ad dimension per\\nunit change in the pop dimension, relative to PCA. This suggests that pop\\nis more highly correlated with the response than is ad. The PLS direction\\ndoes not fit the predictors as closely as does PCA, but it does a better job\\nexplaining the response.\\nTo identify the second PLS direction we first adjust each of the variables\\nfor Z1, by regressing each variable on Z1 and taking residuals. These resid-\\nuals can be interpreted as the remaining information that has not been\\nexplained by the first PLS direction. We then compute Z2 using this or- 262\\n6. Linear Model Selection and Regularization\\nthogonalized data in exactly the same fashion as Z1 was computed based\\non the original data. This iterative approach can be repeated M times to\\nidentify multiple PLS components Z1, . . . , ZM. Finally, at the end of this\\nprocedure, we use least squares to fit a linear model to predict Y using\\nZ1, . . . , ZM in exactly the same fashion as for PCR.\\nAs with PCR, the number M of partial least squares directions used in\\nPLS is a tuning parameter that is typically chosen by cross-validation. We\\ngenerally standardize the predictors and response before performing PLS.\\nPLS is popular in the field of chemometrics, where many variables arise\\nfrom digitized spectrometry signals. In practice it often performs no better\\nthan ridge regression or PCR. While the supervised dimension reduction\\nof PLS can reduce bias, it also has the potential to increase variance, so\\nthat the overall benefit of PLS relative to PCR is a wash.\\n6.4\\nConsiderations in High Dimensions\\n6.4.1\\nHigh-Dimensional Data\\nMost traditional statistical techniques for regression and classification are\\nintended for the low-dimensional setting in which n, the number of ob- low-\\ndimensional\\nservations, is much greater than p, the number of features. This is due in\\npart to the fact that throughout most of the field’s history, the bulk of sci-\\nentific problems requiring the use of statistics have been low-dimensional.\\nFor instance, consider the task of developing a model to predict a patient’s\\nblood pressure on the basis of his or her age, sex, and body mass index\\n(BMI). There are three predictors, or four if an intercept is included in the\\nmodel, and perhaps several thousand patients for whom blood pressure\\nand age, sex, and BMI are available. Hence n ≫p, and so the problem is\\nlow-dimensional. (By dimension here we are referring to the size of p.)\\nIn the past 20 years, new technologies have changed the way that data\\nare collected in fields as diverse as finance, marketing, and medicine. It is\\nnow commonplace to collect an almost unlimited number of feature mea-\\nsurements (p very large). While p can be extremely large, the number of\\nobservations n is often limited due to cost, sample availability, or other\\nconsiderations. Two examples are as follows:\\n1. Rather than predicting blood pressure on the basis of just age, sex,\\nand BMI, one might also collect measurements for half a million sin-\\ngle nucleotide polymorphisms (SNPs; these are individual DNA mu-\\ntations that are relatively common in the population) for inclusion in\\nthe predictive model. Then n ≈200 and p ≈500,000.\\n2. A marketing analyst interested in understanding people’s online shop-\\nping patterns could treat as features all of the search terms entered\\nby users of a search engine. This is sometimes known as the “bag-of-\\nwords” model. The same researcher might have access to the search\\nhistories of only a few hundred or a few thousand search engine users\\nwho have consented to share their information with the researcher.\\nFor a given user, each of the p search terms is scored present (0) or 6.4 Considerations in High Dimensions\\n263\\nabsent (1), creating a large binary feature vector. Then n ≈1,000\\nand p is much larger.\\nData sets containing more features than observations are often referred\\nto as high-dimensional. Classical approaches such as least squares linear high-\\ndimensional\\nregression are not appropriate in this setting. Many of the issues that arise\\nin the analysis of high-dimensional data were discussed earlier in this book,\\nsince they apply also when n > p: these include the role of the bias-variance\\ntrade-off and the danger of overfitting. Though these issues are always rele-\\nvant, they can become particularly important when the number of features\\nis very large relative to the number of observations.\\nWe have defined the high-dimensional setting as the case where the num-\\nber of features p is larger than the number of observations n. But the con-\\nsiderations that we will now discuss certainly also apply if p is slightly\\nsmaller than n, and are best always kept in mind when performing super-\\nvised learning.\\n6.4.2\\nWhat Goes Wrong in High Dimensions?\\nIn order to illustrate the need for extra care and specialized techniques\\nfor regression and classification when p > n, we begin by examining what\\ncan go wrong if we apply a statistical technique not intended for the high-\\ndimensional setting. For this purpose, we examine least squares regression.\\nBut the same concepts apply to logistic regression, linear discriminant anal-\\nysis, and other classical statistical approaches.\\nWhen the number of features p is as large as, or larger than, the number\\nof observations n, least squares as described in Chapter 3 cannot (or rather,\\nshould not) be performed. The reason is simple: regardless of whether or\\nnot there truly is a relationship between the features and the response,\\nleast squares will yield a set of coefficient estimates that result in a perfect\\nfit to the data, such that the residuals are zero.\\nAn example is shown in Figure 6.22 with p = 1 feature (plus an intercept)\\nin two cases: when there are 20 observations, and when there are only\\ntwo observations. When there are 20 observations, n > p and the least\\nsquares regression line does not perfectly fit the data; instead, the regression\\nline seeks to approximate the 20 observations as well as possible. On the\\nother hand, when there are only two observations, then regardless of the\\nvalues of those observations, the regression line will fit the data exactly.\\nThis is problematic because this perfect fit will almost certainly lead to\\noverfitting of the data. In other words, though it is possible to perfectly fit\\nthe training data in the high-dimensional setting, the resulting linear model\\nwill perform extremely poorly on an independent test set, and therefore\\ndoes not constitute a useful model. In fact, we can see that this happened\\nin Figure 6.22: the least squares line obtained in the right-hand panel will\\nperform very poorly on a test set comprised of the observations in the left-\\nhand panel. The problem is simple: when p > n or p ≈n, a simple least\\nsquares regression line is too flexible and hence overfits the data.\\nFigure 6.23 further illustrates the risk of carelessly applying least squares\\nwhen the number of features p is large. Data were simulated with n = 20\\nobservations, and regression was performed with between 1 and 20 features, 264\\n6. Linear Model Selection and Regularization\\n−1.5\\n−1.0\\n−0.5\\n0.0\\n0.5\\n1.0\\n−10\\n−5\\n0\\n5\\n10\\n−1.5\\n−1.0\\n−0.5\\n0.0\\n0.5\\n1.0\\n−10\\n−5\\n0\\n5\\n10\\nX\\nX\\nY\\nY\\nFIGURE 6.22. Left: Least squares regression in the low-dimensional setting.\\nRight: Least squares regression with n = 2 observations and two parameters to be\\nestimated (an intercept and a coefficient).\\n5\\n10\\n15\\n0.2\\n0.4\\n0.6\\n0.8\\n1.0\\nNumber of Variables\\nR2\\n5\\n10\\n15\\n0.0\\n0.2\\n0.4\\n0.6\\n0.8\\nNumber of Variables\\nTraining MSE\\n5\\n10\\n15\\n1\\n5\\n50\\n500\\nNumber of Variables\\nTest MSE\\nFIGURE 6.23.\\nOn a simulated example with n = 20 training observations,\\nfeatures that are completely unrelated to the outcome are added to the model.\\nLeft: The R2 increases to 1 as more features are included. Center: The training\\nset MSE decreases to 0 as more features are included. Right: The test set MSE\\nincreases as more features are included.\\neach of which was completely unrelated to the response. As shown in the\\nfigure, the model R2 increases to 1 as the number of features included in the\\nmodel increases, and correspondingly the training set MSE decreases to 0\\nas the number of features increases, even though the features are completely\\nunrelated to the response. On the other hand, the MSE on an independent\\ntest set becomes extremely large as the number of features included in the\\nmodel increases, because including the additional predictors leads to a vast\\nincrease in the variance of the coefficient estimates. Looking at the test\\nset MSE, it is clear that the best model contains at most a few variables.\\nHowever, someone who carelessly examines only the R2 or the training set\\nMSE might erroneously conclude that the model with the greatest number\\nof variables is best. This indicates the importance of applying extra care\\nwhen analyzing data sets with a large number of variables, and of always\\nevaluating model performance on an independent test set. 6.4 Considerations in High Dimensions\\n265\\n1\\n16\\n21\\n0\\n1\\n2\\n3\\n4\\n5\\n1\\n28\\n51\\n0\\n1\\n2\\n3\\n4\\n5\\n1\\n70\\n111\\n0\\n1\\n2\\n3\\n4\\n5\\np = 20\\np = 50\\np = 2000\\nDegrees of Freedom\\nDegrees of Freedom\\nDegrees of Freedom\\nFIGURE 6.24. The lasso was performed with n = 100 observations and three\\nvalues of p, the number of features. Of the p features, 20 were associated with\\nthe response. The boxplots show the test MSEs that result using three different\\nvalues of the tuning parameter λ in (6.7). For ease of interpretation, rather than\\nreporting λ, the degrees of freedom are reported; for the lasso this turns out\\nto be simply the number of estimated non-zero coefficients. When p = 20, the\\nlowest test MSE was obtained with the smallest amount of regularization. When\\np = 50, the lowest test MSE was achieved when there is a substantial amount\\nof regularization. When p = 2,000 the lasso performed poorly regardless of the\\namount of regularization, due to the fact that only 20 of the 2,000 features truly\\nare associated with the outcome.\\nIn Section 6.1.3, we saw a number of approaches for adjusting the training\\nset RSS or R2 in order to account for the number of variables used to fit\\na least squares model. Unfortunately, the Cp, AIC, and BIC approaches\\nare not appropriate in the high-dimensional setting, because estimating ˆσ2\\nis problematic. (For instance, the formula for ˆσ2 from Chapter 3 yields an\\nestimate ˆσ2 = 0 in this setting.) Similarly, problems arise in the application\\nof adjusted R2 in the high-dimensional setting, since one can easily obtain\\na model with an adjusted R2 value of 1. Clearly, alternative approaches\\nthat are better-suited to the high-dimensional setting are required.\\n6.4.3\\nRegression in High Dimensions\\nIt turns out that many of the methods seen in this chapter for fitting\\nless flexible least squares models, such as forward stepwise selection, ridge\\nregression, the lasso, and principal components regression, are particularly\\nuseful for performing regression in the high-dimensional setting. Essentially,\\nthese approaches avoid overfitting by using a less flexible fitting approach\\nthan least squares.\\nFigure 6.24 illustrates the performance of the lasso in a simple simulated\\nexample. There are p = 20, 50, or 2,000 features, of which 20 are truly\\nassociated with the outcome. The lasso was performed on n = 100 training\\nobservations, and the mean squared error was evaluated on an independent\\ntest set. As the number of features increases, the test set error increases.\\nWhen p = 20, the lowest validation set error was achieved when λ in 266\\n6. Linear Model Selection and Regularization\\n(6.7) was small; however, when p was larger then the lowest validation\\nset error was achieved using a larger value of λ. In each boxplot, rather\\nthan reporting the values of λ used, the degrees of freedom of the resulting\\nlasso solution is displayed; this is simply the number of non-zero coefficient\\nestimates in the lasso solution, and is a measure of the flexibility of the\\nlasso fit. Figure 6.24 highlights three important points: (1) regularization\\nor shrinkage plays a key role in high-dimensional problems, (2) appropriate\\ntuning parameter selection is crucial for good predictive performance, and\\n(3) the test error tends to increase as the dimensionality of the problem\\n(i.e. the number of features or predictors) increases, unless the additional\\nfeatures are truly associated with the response.\\nThe third point above is in fact a key principle in the analysis of high-\\ndimensional data, which is known as the curse of dimensionality. One might curse of di-\\nmensionality\\nthink that as the number of features used to fit a model increases, the\\nquality of the fitted model will increase as well. However, comparing the\\nleft-hand and right-hand panels in Figure 6.24, we see that this is not\\nnecessarily the case: in this example, the test set MSE almost doubles as\\np increases from 20 to 2,000. In general, adding additional signal features\\nthat are truly associated with the response will improve the fitted model,\\nin the sense of leading to a reduction in test set error. However, adding\\nnoise features that are not truly associated with the response will lead\\nto a deterioration in the fitted model, and consequently an increased test\\nset error. This is because noise features increase the dimensionality of the\\nproblem, exacerbating the risk of overfitting (since noise features may be\\nassigned nonzero coefficients due to chance associations with the response\\non the training set) without any potential upside in terms of improved test\\nset error. Thus, we see that new technologies that allow for the collection\\nof measurements for thousands or millions of features are a double-edged\\nsword: they can lead to improved predictive models if these features are in\\nfact relevant to the problem at hand, but will lead to worse results if the\\nfeatures are not relevant. Even if they are relevant, the variance incurred\\nin fitting their coefficients may outweigh the reduction in bias that they\\nbring.\\n6.4.4\\nInterpreting Results in High Dimensions\\nWhen we perform the lasso, ridge regression, or other regression proce-\\ndures in the high-dimensional setting, we must be quite cautious in the way\\nthat we report the results obtained. In Chapter 3, we learned about multi-\\ncollinearity, the concept that the variables in a regression might be corre-\\nlated with each other. In the high-dimensional setting, the multicollinearity\\nproblem is extreme: any variable in the model can be written as a linear\\ncombination of all of the other variables in the model. Essentially, this\\nmeans that we can never know exactly which variables (if any) truly are\\npredictive of the outcome, and we can never identify the best coefficients\\nfor use in the regression. At most, we can hope to assign large regression\\ncoefficients to variables that are correlated with the variables that truly are\\npredictive of the outcome. 6.5 Lab: Linear Models and Regularization Methods\\n267\\nFor instance, suppose that we are trying to predict blood pressure on the\\nbasis of half a million SNPs, and that forward stepwise selection indicates\\nthat 17 of those SNPs lead to a good predictive model on the training data.\\nIt would be incorrect to conclude that these 17 SNPs predict blood pressure\\nmore effectively than the other SNPs not included in the model. There are\\nlikely to be many sets of 17 SNPs that would predict blood pressure just\\nas well as the selected model. If we were to obtain an independent data set\\nand perform forward stepwise selection on that data set, we would likely\\nobtain a model containing a different, and perhaps even non-overlapping,\\nset of SNPs. This does not detract from the value of the model obtained—\\nfor instance, the model might turn out to be very effective in predicting\\nblood pressure on an independent set of patients, and might be clinically\\nuseful for physicians. But we must be careful not to overstate the results\\nobtained, and to make it clear that what we have identified is simply one\\nof many possible models for predicting blood pressure, and that it must be\\nfurther validated on independent data sets.\\nIt is also important to be particularly careful in reporting errors and mea-\\nsures of model fit in the high-dimensional setting. We have seen that when\\np > n, it is easy to obtain a useless model that has zero residuals. There-\\nfore, one should never use sum of squared errors, p-values, R2 statistics, or\\nother traditional measures of model fit on the training data as evidence of\\na good model fit in the high-dimensional setting. For instance, as we saw\\nin Figure 6.23, one can easily obtain a model with R2 = 1 when p > n.\\nReporting this fact might mislead others into thinking that a statistically\\nvalid and useful model has been obtained, whereas in fact this provides\\nabsolutely no evidence of a compelling model. It is important to instead\\nreport results on an independent test set, or cross-validation errors. For\\ninstance, the MSE or R2 on an independent test set is a valid measure of\\nmodel fit, but the MSE on the training set certainly is not.\\n6.5\\nLab: Linear Models and Regularization\\nMethods\\nIn this lab we implement many of the techniques discussed in this chapter.\\nWe import some of our libraries at this top level.\\nIn [1]: import numpy as np\\nimport pandas as pd\\nfrom matplotlib.pyplot import subplots\\nfrom statsmodels.api import OLS\\nimport sklearn.model_selection as skm\\nimport sklearn.linear_model as skl\\nfrom sklearn.preprocessing import StandardScaler\\nfrom ISLP import load_data\\nfrom ISLP.models import ModelSpec as MS\\nfrom functools import partial\\nWe again collect the new imports needed for this lab.\\nIn [2]: from sklearn.pipeline import Pipeline\\nfrom sklearn.decomposition import PCA 268\\n6. Linear Model Selection and Regularization\\nfrom sklearn.cross_decomposition import PLSRegression\\nfrom ISLP.models import \\\\\\n(Stepwise,\\nsklearn_selected ,\\nsklearn_selection_path)\\n!pip install l0bnb\\nfrom l0bnb import fit_path\\nWe have installed the package l0bnb on the fly. Note the escaped !pip\\ninstall — this is run as a separate system command.\\n6.5.1\\nSubset Selection Methods\\nHere we implement methods that reduce the number of parameters in a\\nmodel by restricting the model to a subset of the input variables.\\nForward Selection\\nWe will apply the forward-selection approach to the Hitters data. We wish\\nto predict a baseball player’s Salary on the basis of various statistics asso-\\nciated with performance in the previous year.\\nFirst of all, we note that the Salary variable is missing for some of the\\nplayers. The np.isnan() function can be used to identify the missing ob- np.isnan()\\nservations. It returns an array of the same shape as the input vector, with\\na True for any elements that are missing, and a False for non-missing el-\\nements. The sum() method can then be used to count all of the missing sum()\\nelements.\\nIn [3]: Hitters = load_data(\\'Hitters\\')\\nnp.isnan(Hitters[\\'Salary\\']).sum()\\nOut[3]: 59\\nWe see that Salary is missing for 59 players. The dropna() method of data\\nframes removes all of the rows that have missing values in any variable (by\\ndefault — see Hitters.dropna?).\\nIn [4]: Hitters = Hitters.dropna();\\nHitters.shape\\nOut[4]: (263, 20)\\nWe first choose the best model using forward selection based on Cp (6.2).\\nThis score is not built in as a metric to sklearn. We therefore define a\\nfunction to compute it ourselves, and use it as a scorer. By default, sklearn\\ntries to maximize a score, hence our scoring function computes the negative\\nCp statistic.\\nIn [5]: def nCp(sigma2, estimator, X, Y):\\n\"Negative Cp statistic\"\\nn, p = X.shape\\nYhat = estimator.predict(X)\\nRSS = np.sum((Y - Yhat)**2)\\nreturn -(RSS + 2 * p * sigma2) / n 6.5 Lab: Linear Models and Regularization Methods\\n269\\nWe need to estimate the residual variance σ2, which is the first argument\\nin our scoring function above. We will fit the biggest model, using all the\\nvariables, and estimate σ2 based on its MSE.\\nIn [6]: design = MS(Hitters.columns.drop(\\'Salary\\')).fit(Hitters)\\nY = np.array(Hitters[\\'Salary\\'])\\nX = design.transform(Hitters)\\nsigma2 = OLS(Y,X).fit().scale\\nThe function sklearn_selected() expects a scorer with just three argu-\\nments — the last three in the definition of nCp() above. We use the function\\npartial() first seen in Section 5.3.3 to freeze the first argument with our\\nestimate of σ2.\\nIn [7]: neg_Cp = partial(nCp, sigma2)\\nWe can now use neg_Cp() as a scorer for model selection.\\nAlong with a score we need to specify the search strategy. This is done\\nthrough the object Stepwise() in the ISLP.models package. The method\\nStepwise.first_peak() runs forward stepwise until any further additions\\nto the model do not result in an improvement in the evaluation score.\\nSimilarly, the method Stepwise.fixed_steps() runs a fixed number of steps\\nof stepwise search.\\nIn [8]: strategy = Stepwise.first_peak(design,\\ndirection=\\'forward\\',\\nmax_terms=len(design.terms))\\nWe now fit a linear regression model with Salary as outcome using for-\\nward selection. To do so, we use the function sklearn_selected() from the sklearn_\\nselected()\\nISLP.models package. This takes a model from statsmodels along with a\\nsearch strategy and selects a model with its fit method. Without specify-\\ning a scoring argument, the score defaults to MSE, and so all 19 variables\\nwill be selected (output not shown).\\nIn [9]: hitters_MSE = sklearn_selected(OLS,\\nstrategy)\\nhitters_MSE.fit(Hitters, Y)\\nhitters_MSE.selected_state_\\nUsing neg_Cp results in a smaller model, as expected, with just 10 vari-\\nables selected.\\nIn [10]: hitters_Cp = sklearn_selected(OLS,\\nstrategy,\\nscoring=neg_Cp)\\nhitters_Cp.fit(Hitters, Y)\\nhitters_Cp.selected_state_\\nOut[10]: (\\'Assists\\',\\n\\'AtBat\\',\\n\\'CAtBat\\',\\n\\'CRBI\\',\\n\\'CRuns\\',\\n\\'CWalks\\',\\n\\'Division\\', 270\\n6. Linear Model Selection and Regularization\\n\\'Hits\\',\\n\\'PutOuts\\',\\n\\'Walks\\')\\nChoosing Among Models Using the Validation Set Approach and\\nCross-Validation\\nAs an alternative to using Cp, we might try cross-validation to select a\\nmodel in forward selection. For this, we need a method that stores the\\nfull path of models found in forward selection, and allows predictions for\\neach of these. This can be done with the sklearn_selection_path() estima- sklearn_\\nselection_\\npath()\\ntor from ISLP.models. The function cross_val_predict() from ISLP.models\\ncross_val_\\npredict()\\ncomputes the cross-validated predictions for each of the models along the\\npath, which we can use to evaluate the cross-validated MSE along the path.\\nHere we define a strategy that fits the full forward selection path. While\\nthere are various parameter choices for sklearn_selection_path(), we use\\nthe defaults here, which selects the model at each step based on the biggest\\nreduction in RSS.\\nIn [11]: strategy = Stepwise.fixed_steps(design,\\nlen(design.terms),\\ndirection=\\'forward\\')\\nfull_path = sklearn_selection_path(OLS, strategy)\\nWe now fit the full forward-selection path on the Hitters data and com-\\npute the fitted values.\\nIn [12]: full_path.fit(Hitters, Y)\\nYhat_in = full_path.predict(Hitters)\\nYhat_in.shape\\nOut[12]: (263, 20)\\nThis gives us an array of fitted values — 20 steps in all, including the\\nfitted mean for the null model — which we can use to evaluate in-sample\\nMSE. As expected, the in-sample MSE improves each step we take, indicat-\\ning we must use either the validation or cross-validation approach to select\\nthe number of steps. We fix the y-axis to range from 50,000 to 250,000\\nto compare to the cross-validation and validation set MSE below, as well\\nas other methods such as ridge regression, lasso and principal components\\nregression.\\nIn [13]: mse_fig, ax = subplots(figsize=(8,8))\\ninsample_mse = ((Yhat_in - Y[:,None])**2).mean(0)\\nn_steps = insample_mse.shape[0]\\nax.plot(np.arange(n_steps),\\ninsample_mse ,\\n\\'k\\', # color black\\nlabel=\\'In-sample\\')\\nax.set_ylabel(\\'MSE\\',\\nfontsize=20)\\nax.set_xlabel(\\'# steps of forward stepwise\\',\\nfontsize=20)\\nax.set_xticks(np.arange(n_steps)[::2])\\nax.legend() 6.5 Lab: Linear Models and Regularization Methods\\n271\\nax.set_ylim([50000,250000]);\\nNotice the expression None in Y[:,None] above. This adds an axis (dimen-\\nsion) to the one-dimensional array Y, which allows it to be recycled when\\nsubtracted from the two-dimensional Yhat_in.\\nWe are now ready to use cross-validation to estimate test error along\\nthe model path. We must use only the training observations to perform\\nall aspects of model-fitting — including variable selection. Therefore, the\\ndetermination of which model of a given size is best must be made using\\nonly the training observations in each training fold. This point is subtle but\\nimportant. If the full data set is used to select the best subset at each step,\\nthen the validation set errors and cross-validation errors that we obtain\\nwill not be accurate estimates of the test error.\\nWe now compute the cross-validated predicted values using 5-fold cross-\\nvalidation.\\nIn [14]: K = 5\\nkfold = skm.KFold(K,\\nrandom_state=0,\\nshuffle=True)\\nYhat_cv = skm.cross_val_predict(full_path,\\nHitters,\\nY,\\ncv=kfold)\\nYhat_cv.shape\\nskm.KFold()\\nskm.cross_\\nval_predict()\\nOut[14]: (263, 20)\\nThe prediction matrix Yhat_cv is the same shape as Yhat_in; the difference\\nis that the predictions in each row, corresponding to a particular sample\\nindex, were made from models fit on a training fold that did not include\\nthat row.\\nAt each model along the path, we compute the MSE in each of the cross-\\nvalidation folds. These we will average to get the mean MSE, and can also\\nuse the individual values to compute a crude estimate of the standard error\\nof the mean.9 Hence we must know the test indices for each cross-validation\\nsplit. This can be found by using the split() method of kfold. Because we\\nfixed the random state above, whenever we split any array with the same\\nnumber of rows as Y we recover the same training and test indices, though\\nwe simply ignore the training indices below.\\nIn [15]: cv_mse = []\\nfor train_idx, test_idx in kfold.split(Y):\\nerrors = (Yhat_cv[test_idx] - Y[test_idx,None])**2\\ncv_mse.append(errors.mean(0)) # column means\\ncv_mse = np.array(cv_mse).T\\ncv_mse.shape\\nOut[15]: (20, 5)\\n9The estimate is crude because the five error estimates are based on overlapping\\ntraining sets, and hence are not independent. 272\\n6. Linear Model Selection and Regularization\\nWe now add the cross-validation error estimates to our MSE plot. We\\ninclude the mean error across the five folds, and the estimate of the standard\\nerror of the mean.\\nIn [16]: ax.errorbar(np.arange(n_steps),\\ncv_mse.mean(1),\\ncv_mse.std(1) / np.sqrt(K),\\nlabel=\\'Cross-validated\\',\\nc=\\'r\\') # color red\\nax.set_ylim([50000,250000])\\nax.legend()\\nmse_fig\\nTo repeat the above using the validation set approach, we simply change\\nour cv argument to a validation set: one random split of the data into a\\ntest and training. We choose a test size of 20%, similar to the size of each\\ntest set in 5-fold cross-validation.\\nskm.Shuffle\\nSplit()\\nIn [17]: validation = skm.ShuffleSplit(n_splits=1,\\ntest_size=0.2,\\nrandom_state=0)\\nfor train_idx, test_idx in validation.split(Y):\\nfull_path.fit(Hitters.iloc[train_idx],\\nY[train_idx])\\nYhat_val = full_path.predict(Hitters.iloc[test_idx])\\nerrors = (Yhat_val - Y[test_idx,None])**2\\nvalidation_mse = errors.mean(0)\\nAs for the in-sample MSE case, the validation set approach does not provide\\nstandard errors.\\nIn [18]: ax.plot(np.arange(n_steps),\\nvalidation_mse ,\\n\\'b--\\', # color blue, broken line\\nlabel=\\'Validation\\')\\nax.set_xticks(np.arange(n_steps)[::2])\\nax.set_ylim([50000,250000])\\nax.legend()\\nmse_fig\\nBest Subset Selection\\nForward stepwise is a greedy selection procedure; at each step it augments\\nthe current set by including one additional variable. We now apply best\\nsubset selection to the Hitters data, which for every subset size, searches\\nfor the best set of predictors.\\nWe will use a package called l0bnb to perform best subset selection.\\nInstead of constraining the subset to be a given size, this package produces a\\npath of solutions using the subset size as a penalty rather than a constraint.\\nAlthough the distinction is subtle, the difference comes when we cross-\\nvalidate.\\nIn [19]: D = design.fit_transform(Hitters)\\nD = D.drop(\\'intercept\\', axis=1)\\nX = np.asarray(D) 6.5 Lab: Linear Models and Regularization Methods\\n273\\nHere we excluded the first column corresponding to the intercept, as l0bnb\\nwill fit the intercept separately. We can find a path using the fit_path()\\nfunction.\\nIn [20]: path = fit_path(X,\\nY,\\nmax_nonzeros=X.shape[1])\\nThe function fit_path() returns a list whose values include the fitted\\ncoefficients as B, an intercept as B0, as well as a few other attributes related\\nto the particular path algorithm used. Such details are beyond the scope\\nof this book.\\nIn [21]: path[3]\\nOut[21]: {\\'B\\': array([0.\\n, 3.254844, 0.\\n, 0.\\n, 0.\\n,\\n0.\\n, 0.\\n, 0.\\n, 0.\\n, 0.\\n,\\n0.\\n, 0.677753 , 0.\\n, 0.\\n, 0.\\n,\\n0.\\n, 0.\\n, 0.\\n, 0.\\n]),\\n\\'B0\\': -38.98216739555494,\\n\\'lambda_0\\': 0.011416248027450194,\\n\\'M\\': 0.5829861733382011,\\n\\'Time_exceeded\\': False}\\nIn the example above, we see that at the fourth step in the path, we have\\ntwo nonzero coefficients in \\'B\\', corresponding to the value 0.114 for the\\npenalty parameter lambda_0. We could make predictions using this sequence\\nof fits on a validation set as a function of lambda_0, or with more work using\\ncross-validation.\\n6.5.2\\nRidge Regression and the Lasso\\nWe will use the sklearn.linear_model package (for which we use skl as\\nshorthand below) to fit ridge and lasso regularized linear models on the\\nHitters data. We start with the model matrix X (without an intercept)\\nthat we computed in the previous section on best subset regression.\\nRidge Regression\\nWe will use the function skl.ElasticNet() to fit both ridge and the lasso. To skl.Elastic\\nNet()\\nfit a path of ridge regressions models, we use skl.ElasticNet.path(), which\\nskl.Elastic\\nNet.path()\\ncan fit both ridge and lasso, as well as a hybrid mixture; ridge regression\\ncorresponds to l1_ratio=0. It is good practice to standardize the columns\\nof X in these applications, if the variables are measured in different units.\\nSince skl.ElasticNet() does no normalization, we have to take care of that\\nourselves. Since we standardize first, in order to find coefficient estimates\\non the original scale, we must unstandardize the coefficient estimates. The\\nparameter λ in (6.5) and (6.7) is called alphas in sklearn. In order to be\\nconsistent with the rest of this chapter, we use lambdas rather than alphas\\nin what follows. 10\\n10At the time of publication, ridge fits like the one in code chunk [22] issue unwarranted\\nconvergence warning messages; we expect these to disappear as this package matures. 274\\n6. Linear Model Selection and Regularization\\nIn [22]: Xs = X - X.mean(0)[None,:]\\nX_scale = X.std(0)\\nXs = Xs / X_scale[None,:]\\nlambdas = 10**np.linspace(8, -2, 100) / Y.std()\\nsoln_array = skl.ElasticNet.path(Xs,\\nY,\\nl1_ratio=0.,\\nalphas=lambdas)[1]\\nsoln_array.shape\\nOut[22]: (19, 100)\\nHere we extract the array of coefficients corresponding to the solutions\\nalong the regularization path. By default the skl.ElasticNet.path method\\nfits a path along an automatically selected range of λ values, except for\\nthe case when l1_ratio=0, which results in ridge regression (as is the case\\nhere).11 So here we have chosen to implement the function over a grid of\\nvalues ranging from λ = 108 to λ = 10−2 scaled by the standard deviation\\nof y, essentially covering the full range of scenarios from the null model\\ncontaining only the intercept, to the least squares fit.\\nAssociated with each value of λ is a vector of ridge regression coefficients,\\nthat can be accessed by a column of soln_array. In this case, soln_array is\\na 19 × 100 matrix, with 19 rows (one for each predictor) and 100 columns\\n(one for each value of λ).\\nWe transpose this matrix and turn it into a data frame to facilitate\\nviewing and plotting.\\nIn [23]: soln_path = pd.DataFrame(soln_array.T,\\ncolumns=D.columns,\\nindex=-np.log(lambdas))\\nsoln_path.index.name = \\'negative log(lambda)\\'\\nsoln_path\\nOut[23]:\\nAtBat\\nHits\\nHmRun\\nRuns\\n...\\nnegative\\nlog(lambda)\\n-12.310855\\n0.000800\\n0.000889\\n0.000695\\n0.000851\\n...\\n-12.078271\\n0.001010\\n0.001122\\n0.000878\\n0.001074\\n...\\n-11.845686\\n0.001274\\n0.001416\\n0.001107\\n0.001355\\n...\\n-11.613102\\n0.001608\\n0.001787\\n0.001397\\n0.001710\\n...\\n-11.380518\\n0.002029\\n0.002255\\n0.001763\\n0.002158\\n...\\n...\\n...\\n...\\n...\\n...\\n...\\n100 rows × 19 columns\\nWe plot the paths to get a sense of how the coefficients vary with λ. To\\ncontrol the location of the legend we first set legend to False in the plot\\nmethod, adding it afterward with the legend() method of ax.\\nIn [24]: path_fig, ax = subplots(figsize=(8,8))\\nsoln_path.plot(ax=ax, legend=False)\\nax.set_xlabel(\\'$-\\\\log(\\\\lambda)$\\', fontsize=20)\\n11The reason is rather technical; for all models except ridge, we can find the smallest\\nvalue of λ for which all coefficients are zero. For ridge this value is ∞. 6.5 Lab: Linear Models and Regularization Methods\\n275\\nax.set_ylabel(\\'Standardized coefficients\\', fontsize=20)\\nax.legend(loc=\\'upper left\\');\\n(We have used latex formatting in the horizontal label, in order to format\\nthe Greek λ appropriately.) We expect the coefficient estimates to be much\\nsmaller, in terms of ℓ2 norm, when a large value of λ is used, as compared\\nto when a small value of λ is used. (Recall that the ℓ2 norm is the square\\nroot of the sum of squared coefficient values.) We display the coefficients\\nat the 40th step, where λ is 25.535.\\nIn [25]: beta_hat = soln_path.loc[soln_path.index[39]]\\nlambdas[39], beta_hat\\nOut[25]: (25.535,\\nAtBat\\n5.433750\\nHits\\n6.223582\\nHmRun\\n4.585498\\nRuns\\n5.880855\\nRBI\\n6.195921\\nWalks\\n6.277975\\nYears\\n5.299767\\n...\\n...\\nLet’s compute the ℓ2 norm of the standardized coefficients.\\nIn [26]: np.linalg.norm(beta_hat)\\nOut[26]: 24.17\\nIn contrast, here is the ℓ2 norm when λ is 2.44e-01. Note the much larger\\nℓ2 norm of the coefficients associated with this smaller value of λ.\\nIn [27]: beta_hat = soln_path.loc[soln_path.index[59]]\\nlambdas[59], np.linalg.norm(beta_hat)\\nOut[27]: (0.2437, 160.4237)\\nAbove we normalized X upfront, and fit the ridge model using Xs. The\\nPipeline() object in sklearn provides a clear way to separate feature nor-\\nmalization from the fitting of the ridge model itself.\\nIn [28]: ridge = skl.ElasticNet(alpha=lambdas[59], l1_ratio=0)\\nscaler = StandardScaler(with_mean=True,\\nwith_std=True)\\npipe = Pipeline(steps=[(\\'scaler\\', scaler), (\\'ridge\\', ridge)])\\npipe.fit(X, Y)\\nWe show that it gives the same ℓ2 norm as in our previous fit on the\\nstandardized data.\\nIn [29]: np.linalg.norm(ridge.coef_)\\nOut[29]: 160.4237\\nNotice that the operation pipe.fit(X, Y) above has changed the ridge\\nobject, and in particular has added attributes such as coef_ that were not\\nthere before. 276\\n6. Linear Model Selection and Regularization\\nEstimating Test Error of Ridge Regression\\nChoosing an a priori value of λ for ridge regression is difficult if not im-\\npossible. We will want to use the validation method or cross-validation to\\nselect the tuning parameter. The reader may not be surprised that the\\nPipeline() approach can be used in skm.cross_validate() with either a\\nvalidation method (i.e. validation) or k-fold cross-validation.\\nWe fix the random state of the splitter so that the results obtained will\\nbe reproducible.\\nIn [30]: validation = skm.ShuffleSplit(n_splits=1,\\ntest_size=0.5,\\nrandom_state=0)\\nridge.alpha = 0.01\\nresults = skm.cross_validate(ridge,\\nX,\\nY,\\nscoring=\\'neg_mean_squared_error\\',\\ncv=validation)\\n-results[\\'test_score\\']\\nOut[30]: array([134214.0])\\nThe test MSE is 1.342e+05. Note that if we had instead simply fit a model\\nwith just an intercept, we would have predicted each test observation using\\nthe mean of the training observations. We can get the same result by fitting\\na ridge regression model with a very large value of λ. Note that 1e10 means\\n1010.\\nIn [31]: ridge.alpha = 1e10\\nresults = skm.cross_validate(ridge,\\nX,\\nY,\\nscoring=\\'neg_mean_squared_error\\',\\ncv=validation)\\n-results[\\'test_score\\']\\nOut[31]: array([231788.32])\\nObviously choosing λ = 0.01 is arbitrary, so we will use cross-validation or\\nthe validation-set approach to choose the tuning parameter λ. The object\\nGridSearchCV() allows exhaustive grid search to choose such a parameter.\\nGrid\\nSearchCV()\\nWe first use the validation set method to choose λ.\\nIn [32]: param_grid = {\\'ridge__alpha\\': lambdas}\\ngrid = skm.GridSearchCV(pipe,\\nparam_grid ,\\ncv=validation,\\nscoring=\\'neg_mean_squared_error\\')\\ngrid.fit(X, Y)\\ngrid.best_params_[\\'ridge__alpha\\']\\ngrid.best_estimator_\\nOut[32]: Pipeline(steps=[(\\'scaler\\', StandardScaler()),\\n(\\'ridge\\', ElasticNet(alpha=0.005899, l1_ratio=0))]) 6.5 Lab: Linear Models and Regularization Methods\\n277\\nAlternatively, we can use 5-fold cross-validation.\\nIn [33]: grid = skm.GridSearchCV(pipe,\\nparam_grid,\\ncv=kfold,\\nscoring=\\'neg_mean_squared_error\\')\\ngrid.fit(X, Y)\\ngrid.best_params_[\\'ridge__alpha\\']\\ngrid.best_estimator_\\nRecall we set up the kfold object for 5-fold cross-validation on page 271.\\nWe now plot the cross-validated MSE as a function of −log(λ), which has\\nshrinkage decreasing from left to right.\\nIn [34]: ridge_fig, ax = subplots(figsize=(8,8))\\nax.errorbar(-np.log(lambdas),\\n-grid.cv_results_[\\'mean_test_score\\'],\\nyerr=grid.cv_results_[\\'std_test_score\\'] / np.sqrt(K))\\nax.set_ylim([50000,250000])\\nax.set_xlabel(\\'$-\\\\log(\\\\lambda)$\\', fontsize=20)\\nax.set_ylabel(\\'Cross-validated MSE\\', fontsize=20);\\nOne can cross-validate different metrics to choose a parameter. The de-\\nfault metric for skl.ElasticNet() is test R2. Let’s compare R2 to MSE for\\ncross-validation here.\\nIn [35]: grid_r2 = skm.GridSearchCV(pipe,\\nparam_grid ,\\ncv=kfold)\\ngrid_r2.fit(X, Y)\\nFinally, let’s plot the results for cross-validated R2.\\nIn [36]: r2_fig, ax = subplots(figsize=(8,8))\\nax.errorbar(-np.log(lambdas),\\ngrid_r2.cv_results_[\\'mean_test_score\\'],\\nyerr=grid_r2.cv_results_[\\'std_test_score\\'] / np.sqrt(K)\\n)\\nax.set_xlabel(\\'$-\\\\log(\\\\lambda)$\\', fontsize=20)\\nax.set_ylabel(\\'Cross-validated $R^2$\\', fontsize=20);\\nFast Cross-Validation for Solution Paths\\nThe ridge, lasso, and elastic net can be efficiently fit along a sequence of\\nλ values, creating what is known as a solution path or regularization path.\\nHence there is specialized code to fit such paths, and to choose a suitable\\nvalue of λ using cross-validation. Even with identical splits the results will\\nnot agree exactly with our grid above because the standardization of each\\nfeature in grid is carried out on each fold, while in pipeCV below it is carried\\nout only once. Nevertheless, the results are similar as the normalization is\\nrelatively stable across folds.\\nIn [37]: ridgeCV = skl.ElasticNetCV(alphas=lambdas,\\nl1_ratio=0,\\ncv=kfold)\\npipeCV = Pipeline(steps=[(\\'scaler\\', scaler), 278\\n6. Linear Model Selection and Regularization\\n(\\'ridge\\', ridgeCV)])\\npipeCV.fit(X, Y)\\nLet’s produce a plot again of the cross-validation error to see that it is\\nsimilar to using skm.GridSearchCV.\\nIn [38]: tuned_ridge = pipeCV.named_steps[\\'ridge\\']\\nridgeCV_fig , ax = subplots(figsize=(8,8))\\nax.errorbar(-np.log(lambdas),\\ntuned_ridge.mse_path_.mean(1),\\nyerr=tuned_ridge.mse_path_.std(1) / np.sqrt(K))\\nax.axvline(-np.log(tuned_ridge.alpha_), c=\\'k\\', ls=\\'--\\')\\nax.set_ylim([50000,250000])\\nax.set_xlabel(\\'$-\\\\log(\\\\lambda)$\\', fontsize=20)\\nax.set_ylabel(\\'Cross-validated MSE\\', fontsize=20);\\nWe see that the value of λ that results in the smallest cross-validation\\nerror is 1.19e-02, available as the value tuned_ridge.alpha_. What is the\\ntest MSE associated with this value of λ?\\nIn [39]: np.min(tuned_ridge.mse_path_.mean(1))\\nOut[39]: 115526.71\\nThis represents a further improvement over the test MSE that we got using\\nλ = 4. Finally, tuned_ridge.coef_ has the coefficients fit on the entire data\\nset at this value of λ.\\nIn [40]: tuned_ridge.coef_\\nOut[40]: array([-222.80877051,\\n238.77246614,\\n3.21103754,\\n-2.93050845,\\n3.64888723,\\n108.90953869,\\n-50.81896152, -105.15731984,\\n122.00714801,\\n57.1859509 ,\\n210.35170348,\\n118.05683748,\\n-150.21959435,\\n30.36634231,\\n-61.62459095,\\n77.73832472,\\n40.07350744,\\n-25.02151514,\\n-13.68429544])\\nAs expected, none of the coefficients are zero—ridge regression does not\\nperform variable selection!\\nEvaluating Test Error of Cross-Validated Ridge\\nChoosing λ using cross-validation provides a single regression estimator,\\nsimilar to fitting a linear regression model as we saw in Chapter 3. It is\\ntherefore reasonable to estimate what its test error is. We run into a prob-\\nlem here in that cross-validation will have touched all of its data in choosing\\nλ, hence we have no further data to estimate test error. A compromise is\\nto do an initial split of the data into two disjoint sets: a training set and a\\ntest set. We then fit a cross-validation tuned ridge regression on the train-\\ning set, and evaluate its performance on the test set. We might call this\\ncross-validation nested within the validation set approach. A priori there\\nis no reason to use half of the data for each of the two sets in validation.\\nBelow, we use 75% for training and 25% for test, with the estimator being\\nridge regression tuned using 5-fold cross-validation. This can be achieved\\nin code as follows: 6.5 Lab: Linear Models and Regularization Methods\\n279\\nIn [41]: outer_valid = skm.ShuffleSplit(n_splits=1,\\ntest_size=0.25,\\nrandom_state=1)\\ninner_cv = skm.KFold(n_splits=5,\\nshuffle=True,\\nrandom_state=2)\\nridgeCV = skl.ElasticNetCV(alphas=lambdas,\\nl1_ratio=0,\\ncv=inner_cv)\\npipeCV = Pipeline(steps=[(\\'scaler\\', scaler),\\n(\\'ridge\\', ridgeCV)]);\\nIn [42]: results = skm.cross_validate(pipeCV,\\nX,\\nY,\\ncv=outer_valid ,\\nscoring=\\'neg_mean_squared_error\\')\\n-results[\\'test_score\\']\\nOut[42]: array([132393.84])\\nThe Lasso\\nWe saw that ridge regression with a wise choice of λ can outperform least\\nsquares as well as the null model on the Hitters data set. We now ask\\nwhether the lasso can yield either a more accurate or a more interpretable\\nmodel than ridge regression. In order to fit a lasso model, we once again\\nuse the ElasticNetCV() function; however, this time we use the argument\\nl1_ratio=1. Other than that change, we proceed just as we did in fitting a\\nridge model.\\nIn [43]: lassoCV = skl.ElasticNetCV(n_alphas=100,\\nl1_ratio=1,\\ncv=kfold)\\npipeCV = Pipeline(steps=[(\\'scaler\\', scaler),\\n(\\'lasso\\', lassoCV)])\\npipeCV.fit(X, Y)\\ntuned_lasso = pipeCV.named_steps[\\'lasso\\']\\ntuned_lasso.alpha_\\nOut[43]: 3.147\\nIn [44]: lambdas, soln_array = skl.Lasso.path(Xs,\\nY,\\nl1_ratio=1,\\nn_alphas=100)[:2]\\nsoln_path = pd.DataFrame(soln_array.T,\\ncolumns=D.columns,\\nindex=-np.log(lambdas))\\nWe can see from the coefficient plot of the standardized coefficients that\\ndepending on the choice of tuning parameter, some of the coefficients will\\nbe exactly equal to zero. 280\\n6. Linear Model Selection and Regularization\\nIn [45]: path_fig, ax = subplots(figsize=(8,8))\\nsoln_path.plot(ax=ax, legend=False)\\nax.legend(loc=\\'upper left\\')\\nax.set_xlabel(\\'$-\\\\log(\\\\lambda)$\\', fontsize=20)\\nax.set_ylabel(\\'Standardized coefficiients\\', fontsize=20);\\nThe smallest cross-validated error is lower than the test set MSE of the null\\nmodel and of least squares, and very similar to the test MSE of 115526.71\\nof ridge regression (page 278) with λ chosen by cross-validation.\\nIn [46]: np.min(tuned_lasso.mse_path_.mean(1))\\nOut[46]: 114690.73\\nLet’s again produce a plot of the cross-validation error.\\nIn [47]: lassoCV_fig , ax = subplots(figsize=(8,8))\\nax.errorbar(-np.log(tuned_lasso.alphas_),\\ntuned_lasso.mse_path_.mean(1),\\nyerr=tuned_lasso.mse_path_.std(1) / np.sqrt(K))\\nax.axvline(-np.log(tuned_lasso.alpha_), c=\\'k\\', ls=\\'--\\')\\nax.set_ylim([50000,250000])\\nax.set_xlabel(\\'$-\\\\log(\\\\lambda)$\\', fontsize=20)\\nax.set_ylabel(\\'Cross-validated MSE\\', fontsize=20);\\nHowever, the lasso has a substantial advantage over ridge regression in\\nthat the resulting coefficient estimates are sparse. Here we see that 6 of the\\n19 coefficient estimates are exactly zero. So the lasso model with λ chosen\\nby cross-validation contains only 13 variables.\\nIn [48]: tuned_lasso.coef_\\nOut[48]: array([-210.01008773,\\n243.4550306 ,\\n0.\\n,\\n0.\\n,\\n0.\\n,\\n97.69397357,\\n-41.52283116,\\n-0.\\n,\\n0.\\n,\\n39.62298193,\\n205.75273856,\\n124.55456561,\\n-126.29986768,\\n15.70262427,\\n-59.50157967,\\n75.24590036,\\n21.62698014,\\n-12.04423675,\\n-0.\\n])\\nAs in ridge regression, we could evaluate the test error of cross-validated\\nlasso by first splitting into test and training sets and internally running\\ncross-validation on the training set. We leave this as an exercise.\\n6.5.3\\nPCR and PLS Regression\\nPrincipal Components Regression\\nPrincipal components regression (PCR) can be performed using PCA() from PCA()\\nthe sklearn.decomposition module. We now apply PCR to the Hitters\\ndata, in order to predict Salary. Again, ensure that the missing values\\nhave been removed from the data, as described in Section 6.5.1.\\nWe use LinearRegression() to fit the regression model here. Note that Linear\\nRegression()\\nit fits an intercept by default, unlike the OLS() function seen earlier in\\nSection 6.5.1. 6.5 Lab: Linear Models and Regularization Methods\\n281\\nIn [49]: pca = PCA(n_components=2)\\nlinreg = skl.LinearRegression()\\npipe = Pipeline([(\\'pca\\', pca),\\n(\\'linreg\\', linreg)])\\npipe.fit(X, Y)\\npipe.named_steps[\\'linreg\\'].coef_\\nOut[49]: array([0.09846131, 0.4758765 ])\\nWhen performing PCA, the results vary depending on whether the data\\nhas been standardized or not. As in the earlier examples, this can be ac-\\ncomplished by including an additional step in the pipeline.\\nIn [50]: pipe = Pipeline([(\\'scaler\\', scaler),\\n(\\'pca\\', pca),\\n(\\'linreg\\', linreg)])\\npipe.fit(X, Y)\\npipe.named_steps[\\'linreg\\'].coef_\\nOut[50]: array([106.36859204, -21.60350456])\\nWe can of course use CV to choose the number of components, by\\nusing skm.GridSearchCV, in this case fixing the parameters to vary the\\nn_components.\\nIn [51]: param_grid = {\\'pca__n_components\\': range(1, 20)}\\ngrid = skm.GridSearchCV(pipe,\\nparam_grid ,\\ncv=kfold,\\nscoring=\\'neg_mean_squared_error\\')\\ngrid.fit(X, Y)\\nLet’s plot the results as we have for other methods.\\nIn [52]: pcr_fig, ax = subplots(figsize=(8,8))\\nn_comp = param_grid[\\'pca__n_components\\']\\nax.errorbar(n_comp,\\n-grid.cv_results_[\\'mean_test_score\\'],\\ngrid.cv_results_[\\'std_test_score\\'] / np.sqrt(K))\\nax.set_ylabel(\\'Cross-validated MSE\\', fontsize=20)\\nax.set_xlabel(\\'# principal components\\', fontsize=20)\\nax.set_xticks(n_comp[::2])\\nax.set_ylim([50000,250000]);\\nWe see that the smallest cross-validation error occurs when 17 compo-\\nnents are used. However, from the plot we also see that the cross-validation\\nerror is roughly the same when only one component is included in the\\nmodel. This suggests that a model that uses just a small number of com-\\nponents might suffice.\\nThe CV score is provided for each possible number of components from\\n1 to 19 inclusive. The PCA() method complains if we try to fit an intercept\\nonly with n_components=0 so we also compute the MSE for just the null\\nmodel with these splits.\\nIn [53]: Xn = np.zeros((X.shape[0], 1))\\ncv_null = skm.cross_validate(linreg, 282\\n6. Linear Model Selection and Regularization\\nXn,\\nY,\\ncv=kfold,\\nscoring=\\'neg_mean_squared_error\\')\\n-cv_null[\\'test_score\\'].mean()\\nOut[53]: 204139.31\\nThe explained_variance_ratio_ attribute of our PCA object provides the\\npercentage of variance explained in the predictors and in the response using\\ndifferent numbers of components. This concept is discussed in greater detail\\nin Section 12.2.\\nIn [54]: pipe.named_steps[\\'pca\\'].explained_variance_ratio_\\nOut[54]: array([0.3831424 , 0.21841076])\\nBriefly, we can think of this as the amount of information about the predic-\\ntors that is captured using M principal components. For example, setting\\nM = 1 only captures 38.31% of the variance, while M = 2 captures an ad-\\nditional 21.84%, for a total of 60.15% of the variance. By M = 6 it increases\\nto 88.63%. Beyond this the increments continue to diminish, until we use\\nall M = p = 19 components, which captures all 100% of the variance.\\nPartial Least Squares\\nPartial least squares (PLS) is implemented in the PLSRegression() function. PLS\\nRegression()\\nIn [55]: pls = PLSRegression(n_components=2,\\nscale=True)\\npls.fit(X, Y)\\nAs was the case in PCR, we will want to use CV to choose the number\\nof components.\\nIn [56]: param_grid = {\\'n_components\\':range(1, 20)}\\ngrid = skm.GridSearchCV(pls,\\nparam_grid,\\ncv=kfold,\\nscoring=\\'neg_mean_squared_error\\')\\ngrid.fit(X, Y)\\nAs for our other methods, we plot the MSE.\\nIn [57]: pls_fig, ax = subplots(figsize=(8,8))\\nn_comp = param_grid[\\'n_components\\']\\nax.errorbar(n_comp,\\n-grid.cv_results_[\\'mean_test_score\\'],\\ngrid.cv_results_[\\'std_test_score\\'] / np.sqrt(K))\\nax.set_ylabel(\\'Cross-validated MSE\\', fontsize=20)\\nax.set_xlabel(\\'# principal components\\', fontsize=20)\\nax.set_xticks(n_comp[::2])\\nax.set_ylim([50000,250000]);\\nCV error is minimized at 12, though there is little noticable difference\\nbetween this point and a much lower number like 2 or 3 components. 6.6 Exercises\\n283\\n6.6\\nExercises\\nConceptual\\n1. We perform best subset, forward stepwise, and backward stepwise\\nselection on a single data set. For each approach, we obtain p + 1\\nmodels, containing 0, 1, 2, . . . , p predictors. Explain your answers:\\n(a) Which of the three models with k predictors has the smallest\\ntraining RSS?\\n(b) Which of the three models with k predictors has the smallest\\ntest RSS?\\n(c) True or False:\\ni. The predictors in the k-variable model identified by forward\\nstepwise are a subset of the predictors in the (k+1)-variable\\nmodel identified by forward stepwise selection.\\nii. The predictors in the k-variable model identified by back-\\nward stepwise are a subset of the predictors in the (k + 1)-\\nvariable model identified by backward stepwise selection.\\niii. The predictors in the k-variable model identified by back-\\nward stepwise are a subset of the predictors in the (k + 1)-\\nvariable model identified by forward stepwise selection.\\niv. The predictors in the k-variable model identified by forward\\nstepwise are a subset of the predictors in the (k+1)-variable\\nmodel identified by backward stepwise selection.\\nv. The predictors in the k-variable model identified by best\\nsubset are a subset of the predictors in the (k + 1)-variable\\nmodel identified by best subset selection.\\n2. For parts (a) through (c), indicate which of i. through iv. is correct.\\nJustify your answer.\\n(a) The lasso, relative to least squares, is:\\ni. More flexible and hence will give improved prediction ac-\\ncuracy when its increase in bias is less than its decrease in\\nvariance.\\nii. More flexible and hence will give improved prediction accu-\\nracy when its increase in variance is less than its decrease\\nin bias.\\niii. Less flexible and hence will give improved prediction accu-\\nracy when its increase in bias is less than its decrease in\\nvariance.\\niv. Less flexible and hence will give improved prediction accu-\\nracy when its increase in variance is less than its decrease\\nin bias.\\n(b) Repeat (a) for ridge regression relative to least squares.\\n(c) Repeat (a) for non-linear methods relative to least squares. 284\\n6. Linear Model Selection and Regularization\\n3. Suppose we estimate the regression coefficients in a linear regression\\nmodel by minimizing\\nn\\n0\\ni=1\\n\\uf8eb\\n\\uf8edyi −β0 −\\np\\n0\\nj=1\\nβjxij\\n\\uf8f6\\n\\uf8f8\\n2\\nsubject to\\np\\n0\\nj=1\\n|βj| ≤s\\nfor a particular value of s. For parts (a) through (e), indicate which\\nof i. through v. is correct. Justify your answer.\\n(a) As we increase s from 0, the training RSS will:\\ni. Increase initially, and then eventually start decreasing in an\\ninverted U shape.\\nii. Decrease initially, and then eventually start increasing in a\\nU shape.\\niii. Steadily increase.\\niv. Steadily decrease.\\nv. Remain constant.\\n(b) Repeat (a) for test RSS.\\n(c) Repeat (a) for variance.\\n(d) Repeat (a) for (squared) bias.\\n(e) Repeat (a) for the irreducible error.\\n4. Suppose we estimate the regression coefficients in a linear regression\\nmodel by minimizing\\nn\\n0\\ni=1\\n\\uf8eb\\n\\uf8edyi −β0 −\\np\\n0\\nj=1\\nβjxij\\n\\uf8f6\\n\\uf8f8\\n2\\n+ λ\\np\\n0\\nj=1\\nβ2\\nj\\nfor a particular value of λ. For parts (a) through (e), indicate which\\nof i. through v. is correct. Justify your answer.\\n(a) As we increase λ from 0, the training RSS will:\\ni. Increase initially, and then eventually start decreasing in an\\ninverted U shape.\\nii. Decrease initially, and then eventually start increasing in a\\nU shape.\\niii. Steadily increase.\\niv. Steadily decrease.\\nv. Remain constant.\\n(b) Repeat (a) for test RSS.\\n(c) Repeat (a) for variance.\\n(d) Repeat (a) for (squared) bias.\\n(e) Repeat (a) for the irreducible error. 6.6 Exercises\\n285\\n5. It is well-known that ridge regression tends to give similar coefficient\\nvalues to correlated variables, whereas the lasso may give quite dif-\\nferent coefficient values to correlated variables. We will now explore\\nthis property in a very simple setting.\\nSuppose that n = 2, p = 2, x11 = x12, x21 = x22. Furthermore,\\nsuppose that y1 +y2 = 0 and x11 +x21 = 0 and x12 +x22 = 0, so that\\nthe estimate for the intercept in a least squares, ridge regression, or\\nlasso model is zero: ˆβ0 = 0.\\n(a) Write out the ridge regression optimization problem in this set-\\nting.\\n(b) Argue that in this setting, the ridge coefficient estimates satisfy\\nˆβ1 = ˆβ2.\\n(c) Write out the lasso optimization problem in this setting.\\n(d) Argue that in this setting, the lasso coefficients ˆβ1 and ˆβ2 are\\nnot unique—in other words, there are many possible solutions\\nto the optimization problem in (c). Describe these solutions.\\n6. We will now explore (6.12) and (6.13) further.\\n(a) Consider (6.12) with p = 1. For some choice of y1 and λ > 0,\\nplot (6.12) as a function of β1. Your plot should confirm that\\n(6.12) is solved by (6.14).\\n(b) Consider (6.13) with p = 1. For some choice of y1 and λ > 0,\\nplot (6.13) as a function of β1. Your plot should confirm that\\n(6.13) is solved by (6.15).\\n7. We will now derive the Bayesian connection to the lasso and ridge\\nregression discussed in Section 6.2.2.\\n(a) Suppose that yi = β0+)p\\nj=1 xijβj +ϵi where ϵ1, . . . , ϵn are inde-\\npendent and identically distributed from a N(0, σ2) distribution.\\nWrite out the likelihood for the data.\\n(b) Assume the following prior for β: β1, . . . , βp are independent\\nand identically distributed according to a double-exponential\\ndistribution with mean 0 and common scale parameter b: i.e.\\np(β) =\\n1\\n2b exp(−|β|/b). Write out the posterior for β in this\\nsetting.\\n(c) Argue that the lasso estimate is the mode for β under this pos-\\nterior distribution.\\n(d) Now assume the following prior for β: β1, . . . , βp are independent\\nand identically distributed according to a normal distribution\\nwith mean zero and variance c. Write out the posterior for β in\\nthis setting.\\n(e) Argue that the ridge regression estimate is both the mode and\\nthe mean for β under this posterior distribution. 286\\n6. Linear Model Selection and Regularization\\nApplied\\n8. In this exercise, we will generate simulated data, and will then use\\nthis data to perform forward and backward stepwise selection.\\n(a) Create a random number generator and use its normal() method\\nto generate a predictor X of length n = 100, as well as a noise\\nvector ϵ of length n = 100.\\n(b) Generate a response vector Y of length n = 100 according to\\nthe model\\nY = β0 + β1X + β2X2 + β3X3 + ϵ,\\nwhere β0, β1, β2, and β3 are constants of your choice.\\n(c) Use forward stepwise selection in order to select a model contain-\\ning the predictors X, X2, . . . , X10. What is the model obtained\\naccording to Cp? Report the coefficients of the model obtained.\\n(d) Repeat (c), using backwards stepwise selection. How does your\\nanswer compare to the results in (c)?\\n(e) Now fit a lasso model to the simulated data, again using X, X2,\\n. . . , X10 as predictors. Use cross-validation to select the optimal\\nvalue of λ. Create plots of the cross-validation error as a function\\nof λ. Report the resulting coefficient estimates, and discuss the\\nresults obtained.\\n(f) Now generate a response vector Y according to the model\\nY = β0 + β7X7 + ϵ,\\nand perform forward stepwise selection and the lasso. Discuss\\nthe results obtained.\\n9. In this exercise, we will predict the number of applications received\\nusing the other variables in the College data set.\\n(a) Split the data set into a training set and a test set.\\n(b) Fit a linear model using least squares on the training set, and\\nreport the test error obtained.\\n(c) Fit a ridge regression model on the training set, with λ chosen\\nby cross-validation. Report the test error obtained.\\n(d) Fit a lasso model on the training set, with λ chosen by cross-\\nvalidation. Report the test error obtained, along with the num-\\nber of non-zero coefficient estimates.\\n(e) Fit a PCR model on the training set, with M chosen by cross-\\nvalidation. Report the test error obtained, along with the value\\nof M selected by cross-validation.\\n(f) Fit a PLS model on the training set, with M chosen by cross-\\nvalidation. Report the test error obtained, along with the value\\nof M selected by cross-validation. 6.6 Exercises\\n287\\n(g) Comment on the results obtained. How accurately can we pre-\\ndict the number of college applications received? Is there much\\ndifference among the test errors resulting from these five ap-\\nproaches?\\n10. We have seen that as the number of features used in a model increases,\\nthe training error will necessarily decrease, but the test error may not.\\nWe will now explore this in a simulated data set.\\n(a) Generate a data set with p = 20 features, n = 1,000 observa-\\ntions, and an associated quantitative response vector generated\\naccording to the model\\nY = Xβ + ϵ,\\nwhere β has some elements that are exactly equal to zero.\\n(b) Split your data set into a training set containing 100 observations\\nand a test set containing 900 observations.\\n(c) Perform best subset selection on the training set, and plot the\\ntraining set MSE associated with the best model of each size.\\n(d) Plot the test set MSE associated with the best model of each\\nsize.\\n(e) For which model size does the test set MSE take on its minimum\\nvalue? Comment on your results. If it takes on its minimum value\\nfor a model containing only an intercept or a model containing\\nall of the features, then play around with the way that you are\\ngenerating the data in (a) until you come up with a scenario in\\nwhich the test set MSE is minimized for an intermediate model\\nsize.\\n(f) How does the model at which the test set MSE is minimized\\ncompare to the true model used to generate the data? Comment\\non the coefficient values.\\n(g) Create a plot displaying\\nG)p\\nj=1(βj −ˆβr\\nj )2 for a range of values\\nof r, where ˆβr\\nj is the jth coefficient estimate for the best model\\ncontaining r coefficients. Comment on what you observe. How\\ndoes this compare to the test MSE plot from (d)?\\n11. We will now try to predict per capita crime rate in the Boston data\\nset.\\n(a) Try out some of the regression methods explored in this chapter,\\nsuch as best subset selection, the lasso, ridge regression, and\\nPCR. Present and discuss results for the approaches that you\\nconsider.\\n(b) Propose a model (or set of models) that seem to perform well on\\nthis data set, and justify your answer. Make sure that you are\\nevaluating model performance using validation set error, cross-\\nvalidation, or some other reasonable alternative, as opposed to\\nusing training error. 288\\n6. Linear Model Selection and Regularization\\n(c) Does your chosen model involve all of the features in the data\\nset? Why or why not? 7\\nMoving Beyond Linearity\\nSo far in this book, we have mostly focused on linear models. Linear models\\nare relatively simple to describe and implement, and have advantages over\\nother approaches in terms of interpretation and inference. However, stan-\\ndard linear regression can have significant limitations in terms of predic-\\ntive power. This is because the linearity assumption is almost always an\\napproximation, and sometimes a poor one. In Chapter 6 we see that we can\\nimprove upon least squares using ridge regression, the lasso, principal com-\\nponents regression, and other techniques. In that setting, the improvement\\nis obtained by reducing the complexity of the linear model, and hence the\\nvariance of the estimates. But we are still using a linear model, which can\\nonly be improved so far! In this chapter we relax the linearity assumption\\nwhile still attempting to maintain as much interpretability as possible. We\\ndo this by examining very simple extensions of linear models like polyno-\\nmial regression and step functions, as well as more sophisticated approaches\\nsuch as splines, local regression, and generalized additive models.\\n• Polynomial regression extends the linear model by adding extra pre-\\ndictors, obtained by raising each of the original predictors to a power.\\nFor example, a cubic regression uses three variables, X, X2, and X3,\\nas predictors. This approach provides a simple way to provide a non-\\nlinear fit to data.\\n• Step functions cut the range of a variable into K distinct regions in\\norder to produce a qualitative variable. This has the effect of fitting\\na piecewise constant function.\\n• Regression splines are more flexible than polynomials and step func-\\ntions, and in fact are an extension of the two. They involve dividing\\nthe range of X into K distinct regions. Within each region, a poly-\\nnomial function is fit to the data. However, these polynomials are\\n© Springer Nature Switzerland AG 2023 \\nG. James et al., An Introduction to Statistical Learning, Springer Texts in Statistics, \\nhttps://doi.org/10.1007/978-3-031-38747-0_7\\n289 290\\n7. Moving Beyond Linearity\\nconstrained so that they join smoothly at the region boundaries, or\\nknots. Provided that the interval is divided into enough regions, this\\ncan produce an extremely flexible fit.\\n• Smoothing splines are similar to regression splines, but arise in a\\nslightly different situation. Smoothing splines result from minimizing\\na residual sum of squares criterion subject to a smoothness penalty.\\n• Local regression is similar to splines, but differs in an important way.\\nThe regions are allowed to overlap, and indeed they do so in a very\\nsmooth way.\\n• Generalized additive models allow us to extend the methods above to\\ndeal with multiple predictors.\\nIn Sections 7.1–7.6, we present a number of approaches for modeling the\\nrelationship between a response Y and a single predictor X in a flexible\\nway. In Section 7.7, we show that these approaches can be seamlessly in-\\ntegrated in order to model a response Y as a function of several predictors\\nX1, . . . , Xp.\\n7.1\\nPolynomial Regression\\nHistorically, the standard way to extend linear regression to settings in\\nwhich the relationship between the predictors and the response is non-\\nlinear has been to replace the standard linear model\\nyi = β0 + β1xi + ϵi\\nwith a polynomial function\\nyi = β0 + β1xi + β2x2\\ni + β3x3\\ni + · · · + βdxd\\ni + ϵi,\\n(7.1)\\nwhere ϵi is the error term. This approach is known as polynomial regression, polynomial\\nregression\\nand in fact we saw an example of this method in Section 3.3.2. For large\\nenough degree d, a polynomial regression allows us to produce an extremely\\nnon-linear curve. Notice that the coefficients in (7.1) can be easily estimated\\nusing least squares linear regression because this is just a standard linear\\nmodel with predictors xi, x2\\ni , x3\\ni , . . . , xd\\ni . Generally speaking, it is unusual\\nto use d greater than 3 or 4 because for large values of d, the polynomial\\ncurve can become overly flexible and can take on some very strange shapes.\\nThis is especially true near the boundary of the X variable.\\nThe left-hand panel in Figure 7.1 is a plot of wage against age for the\\nWage data set, which contains income and demographic information for\\nmales who reside in the central Atlantic region of the United States. We\\nsee the results of fitting a degree-4 polynomial using least squares (solid\\nblue curve). Even though this is a linear regression model like any other,\\nthe individual coefficients are not of particular interest. Instead, we look at\\nthe entire fitted function across a grid of 63 values for age from 18 to 80 in\\norder to understand the relationship between age and wage. 7.1 Polynomial Regression\\n291\\n20\\n30\\n40\\n50\\n60\\n70\\n80\\n50\\n100\\n150\\n200\\n250\\n300\\nAge\\nWage\\nDegree−4 Polynomial\\n20\\n30\\n40\\n50\\n60\\n70\\n80\\n0.00\\n0.05\\n0.10\\n0.15\\n0.20\\nAge\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n||\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|||\\n|\\n| |\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n| |\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|||\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| ||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n| |\\n| |\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| ||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n| |\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n| |\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|||\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n| | |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n||\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n| ||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n||\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n||\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| | |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n||\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n| |\\n|\\n|\\n|\\n|\\n| | |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|| |\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n||\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n||\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| ||\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n||\\n|\\n|\\n| |\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n||\\n|\\nPr(Wage>250 | Age)\\nFIGURE 7.1. The Wage data. Left: The solid blue curve is a degree-4 polynomial\\nof wage (in thousands of dollars) as a function of age, fit by least squares. The\\ndashed curves indicate an estimated 95 % confidence interval. Right: We model the\\nbinary event wage>250 using logistic regression, again with a degree-4 polynomial.\\nThe fitted posterior probability of wage exceeding $250,000 is shown in blue, along\\nwith an estimated 95 % confidence interval.\\nIn Figure 7.1, a pair of dashed curves accompanies the fit; these are (2×)\\nstandard error curves. Let’s see how these arise. Suppose we have computed\\nthe fit at a particular value of age, x0:\\nˆf(x0) = ˆβ0 + ˆβ1x0 + ˆβ2x2\\n0 + ˆβ3x3\\n0 + ˆβ4x4\\n0.\\n(7.2)\\nWhat is the variance of the fit, i.e. Var ˆf(x0)? Least squares returns variance\\nestimates for each of the fitted coefficients ˆβj, as well as the covariances\\nbetween pairs of coefficient estimates. We can use these to compute the\\nestimated variance of ˆf(x0).1 The estimated pointwise standard error of\\nˆf(x0) is the square-root of this variance. This computation is repeated\\nat each reference point x0, and we plot the fitted curve, as well as twice\\nthe standard error on either side of the fitted curve. We plot twice the\\nstandard error because, for normally distributed error terms, this quantity\\ncorresponds to an approximate 95 % confidence interval.\\nIt seems like the wages in Figure 7.1 are from two distinct populations:\\nthere appears to be a high earners group earning more than $250,000 per\\nannum, as well as a low earners group. We can treat wage as a binary\\nvariable by splitting it into these two groups. Logistic regression can then\\nbe used to predict this binary response, using polynomial functions of age\\n1If ˆC is the 5 × 5 covariance matrix of the ˆβj, and if ℓT\\n0 = (1, x0, x2\\n0, x3\\n0, x4\\n0), then\\nVar[ ˆf(x0)] = ℓT\\n0 ˆCℓ0. 292\\n7. Moving Beyond Linearity\\nas predictors. In other words, we fit the model\\nPr(yi > 250|xi) =\\nexp(β0 + β1xi + β2x2\\ni + · · · + βdxd\\ni )\\n1 + exp(β0 + β1xi + β2x2\\ni + · · · + βdxd\\ni ).\\n(7.3)\\nThe result is shown in the right-hand panel of Figure 7.1. The gray marks\\non the top and bottom of the panel indicate the ages of the high earners\\nand the low earners. The solid blue curve indicates the fitted probabilities\\nof being a high earner, as a function of age. The estimated 95 % confidence\\ninterval is shown as well. We see that here the confidence intervals are fairly\\nwide, especially on the right-hand side. Although the sample size for this\\ndata set is substantial (n = 3,000), there are only 79 high earners, which\\nresults in a high variance in the estimated coefficients and consequently\\nwide confidence intervals.\\n7.2\\nStep Functions\\nUsing polynomial functions of the features as predictors in a linear model\\nimposes a global structure on the non-linear function of X. We can instead\\nuse step functions in order to avoid imposing such a global structure. Here step\\nfunction\\nwe break the range of X into bins, and fit a different constant in each bin.\\nThis amounts to converting a continuous variable into an ordered categorical\\nvariable.\\nordered\\ncategorical\\nvariable\\nIn greater detail, we create cutpoints c1, c2, . . . , cK in the range of X,\\nand then construct K + 1 new variables\\nC0(X)\\n=\\nI(X < c1),\\nC1(X)\\n=\\nI(c1 ≤X < c2),\\nC2(X)\\n=\\nI(c2 ≤X < c3),\\n...\\nCK−1(X)\\n=\\nI(cK−1 ≤X < cK),\\nCK(X)\\n=\\nI(cK ≤X),\\n(7.4)\\nwhere I(·) is an indicator function that returns a 1 if the condition is true, indicator\\nfunction\\nand returns a 0 otherwise. For example, I(cK ≤X) equals 1 if cK ≤X, and\\nequals 0 otherwise. These are sometimes called dummy variables. Notice\\nthat for any value of X, C0(X) + C1(X) + · · · + CK(X) = 1, since X must\\nbe in exactly one of the K + 1 intervals. We then use least squares to fit a\\nlinear model using C1(X), C2(X), . . . , CK(X) as predictors2:\\nyi = β0 + β1C1(xi) + β2C2(xi) + · · · + βKCK(xi) + ϵi.\\n(7.5)\\nFor a given value of X, at most one of C1, C2, . . . , CK can be non-zero.\\nNote that when X < c1, all of the predictors in (7.5) are zero, so β0 can\\n2We exclude C0(X) as a predictor in (7.5) because it is redundant with the intercept.\\nThis is similar to the fact that we need only two dummy variables to code a qualitative\\nvariable with three levels, provided that the model will contain an intercept. The decision\\nto exclude C0(X) instead of some other Ck(X) in (7.5) is arbitrary. Alternatively, we\\ncould include C0(X), C1(X), . . . , CK(X), and exclude the intercept. 7.3 Basis Functions\\n293\\n20\\n30\\n40\\n50\\n60\\n70\\n80\\n50\\n100\\n150\\n200\\n250\\n300\\nAge\\nWage\\nPiecewise Constant\\n20\\n30\\n40\\n50\\n60\\n70\\n80\\n0.00\\n0.05\\n0.10\\n0.15\\n0.20\\nAge\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n| ||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n||\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n||\\n|||\\n|\\n| |\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n| |\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|||\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| ||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| ||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n| |\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n| |\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|||\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n| | |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n||\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n| ||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n| |\\n|\\n|\\n|\\n|\\n| | |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n||\\n|\\n| |\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| ||\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n||\\n|\\n|| |\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n||\\n|\\nPr(Wage>250 | Age)\\nFIGURE 7.2. The Wage data. Left: The solid curve displays the fitted value from\\na least squares regression of wage (in thousands of dollars) using step functions\\nof age. The dashed curves indicate an estimated 95 % confidence interval. Right:\\nWe model the binary event wage>250 using logistic regression, again using step\\nfunctions of age. The fitted posterior probability of wage exceeding $250,000 is\\nshown, along with an estimated 95 % confidence interval.\\nbe interpreted as the mean value of Y for X < c1. By comparison, (7.5)\\npredicts a response of β0+βj for cj ≤X < cj+1, so βj represents the average\\nincrease in the response for X in cj ≤X < cj+1 relative to X < c1.\\nAn example of fitting step functions to the Wage data from Figure 7.1 is\\nshown in the left-hand panel of Figure 7.2. We also fit the logistic regression\\nmodel\\nPr(yi > 250|xi) =\\nexp(β0 + β1C1(xi) + · · · + βKCK(xi))\\n1 + exp(β0 + β1C1(xi) + · · · + βKCK(xi))\\n(7.6)\\nin order to predict the probability that an individual is a high earner on the\\nbasis of age. The right-hand panel of Figure 7.2 displays the fitted posterior\\nprobabilities obtained using this approach.\\nUnfortunately, unless there are natural breakpoints in the predictors,\\npiecewise-constant functions can miss the action. For example, in the left-\\nhand panel of Figure 7.2, the first bin clearly misses the increasing trend\\nof wage with age. Nevertheless, step function approaches are very popular\\nin biostatistics and epidemiology, among other disciplines. For example,\\n5-year age groups are often used to define the bins.\\n7.3\\nBasis Functions\\nPolynomial and piecewise-constant regression models are in fact special\\ncases of a basis function approach. The idea is to have at hand a fam- basis\\nfunction 294\\n7. Moving Beyond Linearity\\nily of functions or transformations that can be applied to a variable X:\\nb1(X), b2(X), . . . , bK(X). Instead of fitting a linear model in X, we fit the\\nmodel\\nyi = β0 + β1b1(xi) + β2b2(xi) + β3b3(xi) + · · · + βKbK(xi) + ϵi.\\n(7.7)\\nNote that the basis functions b1(·), b2(·), . . . , bK(·) are fixed and known.\\n(In other words, we choose the functions ahead of time.) For polynomial\\nregression, the basis functions are bj(xi) = xj\\ni, and for piecewise constant\\nfunctions they are bj(xi) = I(cj ≤xi < cj+1). We can think of (7.7) as\\na standard linear model with predictors b1(xi), b2(xi), . . . , bK(xi). Hence,\\nwe can use least squares to estimate the unknown regression coefficients\\nin (7.7). Importantly, this means that all of the inference tools for linear\\nmodels that are discussed in Chapter 3, such as standard errors for the\\ncoefficient estimates and F-statistics for the model’s overall significance,\\nare available in this setting.\\nThus far we have considered the use of polynomial functions and piece-\\nwise constant functions for our basis functions; however, many alternatives\\nare possible. For instance, we can use wavelets or Fourier series to construct\\nbasis functions. In the next section, we investigate a very common choice\\nfor a basis function: regression splines.\\nregression\\nspline\\n7.4\\nRegression Splines\\nNow we discuss a flexible class of basis functions that extends upon the\\npolynomial regression and piecewise constant regression approaches that\\nwe have just seen.\\n7.4.1\\nPiecewise Polynomials\\nInstead of fitting a high-degree polynomial over the entire range of X, piece-\\nwise polynomial regression involves fitting separate low-degree polynomials piecewise\\npolynomial\\nregression\\nover different regions of X. For example, a piecewise cubic polynomial\\nworks by fitting a cubic regression model of the form\\nyi = β0 + β1xi + β2x2\\ni + β3x3\\ni + ϵi,\\n(7.8)\\nwhere the coefficients β0, β1, β2, and β3 differ in different parts of the range\\nof X. The points where the coefficients change are called knots.\\nknot\\nFor example, a piecewise cubic with no knots is just a standard cubic\\npolynomial, as in (7.1) with d = 3. A piecewise cubic polynomial with a\\nsingle knot at a point c takes the form\\nyi =\\n=\\nβ01 + β11xi + β21x2\\ni + β31x3\\ni + ϵi\\nif xi < c\\nβ02 + β12xi + β22x2\\ni + β32x3\\ni + ϵi\\nif xi ≥c.\\nIn other words, we fit two different polynomial functions to the data, one\\non the subset of the observations with xi < c, and one on the subset of\\nthe observations with xi ≥c. The first polynomial function has coefficients 7.4 Regression Splines\\n295\\n20\\n30\\n40\\n50\\n60\\n70\\n50\\n100\\n150\\n200\\n250\\nAge\\nWage\\nPiecewise Cubic\\n20\\n30\\n40\\n50\\n60\\n70\\n50\\n100\\n150\\n200\\n250\\nAge\\nWage\\nContinuous Piecewise Cubic\\n20\\n30\\n40\\n50\\n60\\n70\\n50\\n100\\n150\\n200\\n250\\nAge\\nWage\\nCubic Spline\\n20\\n30\\n40\\n50\\n60\\n70\\n50\\n100\\n150\\n200\\n250\\nAge\\nWage\\nLinear Spline\\nFIGURE 7.3. Various piecewise polynomials are fit to a subset of the Wage\\ndata, with a knot at age=50. Top Left: The cubic polynomials are unconstrained.\\nTop Right: The cubic polynomials are constrained to be continuous at age=50.\\nBottom Left: The cubic polynomials are constrained to be continuous, and to have\\ncontinuous first and second derivatives. Bottom Right: A linear spline is shown,\\nwhich is constrained to be continuous.\\nβ01, β11, β21, and β31, and the second has coefficients β02, β12, β22, and β32.\\nEach of these polynomial functions can be fit using least squares applied\\nto simple functions of the original predictor.\\nUsing more knots leads to a more flexible piecewise polynomial. In gen-\\neral, if we place K different knots throughout the range of X, then we\\nwill end up fitting K + 1 different cubic polynomials. Note that we do not\\nneed to use a cubic polynomial. For example, we can instead fit piecewise\\nlinear functions. In fact, our piecewise constant functions of Section 7.2 are\\npiecewise polynomials of degree 0!\\nThe top left panel of Figure 7.3 shows a piecewise cubic polynomial fit to\\na subset of the Wage data, with a single knot at age=50. We immediately see\\na problem: the function is discontinuous and looks ridiculous! Since each\\npolynomial has four parameters, we are using a total of eight degrees of\\nfreedom in fitting this piecewise polynomial model.\\ndegrees of\\nfreedom 296\\n7. Moving Beyond Linearity\\n7.4.2\\nConstraints and Splines\\nThe top left panel of Figure 7.3 looks wrong because the fitted curve is just\\ntoo flexible. To remedy this problem, we can fit a piecewise polynomial\\nunder the constraint that the fitted curve must be continuous. In other\\nwords, there cannot be a jump when age=50. The top right plot in Figure 7.3\\nshows the resulting fit. This looks better than the top left plot, but the V-\\nshaped join looks unnatural.\\nIn the lower left plot, we have added two additional constraints: now both\\nthe first and second derivatives of the piecewise polynomials are continuous derivative\\nat age=50. In other words, we are requiring that the piecewise polynomial\\nbe not only continuous when age=50, but also very smooth. Each constraint\\nthat we impose on the piecewise cubic polynomials effectively frees up one\\ndegree of freedom, by reducing the complexity of the resulting piecewise\\npolynomial fit. So in the top left plot, we are using eight degrees of free-\\ndom, but in the bottom left plot we imposed three constraints (continuity,\\ncontinuity of the first derivative, and continuity of the second derivative)\\nand so are left with five degrees of freedom. The curve in the bottom left\\nplot is called a cubic spline.3 In general, a cubic spline with K knots uses cubic spline\\na total of 4 + K degrees of freedom.\\nIn Figure 7.3, the lower right plot is a linear spline, which is continuous linear spline\\nat age=50. The general definition of a degree-d spline is that it is a piecewise\\ndegree-d polynomial, with continuity in derivatives up to degree d −1 at\\neach knot. Therefore, a linear spline is obtained by fitting a line in each\\nregion of the predictor space defined by the knots, requiring continuity at\\neach knot.\\nIn Figure 7.3, there is a single knot at age=50. Of course, we could add\\nmore knots, and impose continuity at each.\\n7.4.3\\nThe Spline Basis Representation\\nThe regression splines that we just saw in the previous section may have\\nseemed somewhat complex: how can we fit a piecewise degree-d polynomial\\nunder the constraint that it (and possibly its first d −1 derivatives) be\\ncontinuous? It turns out that we can use the basis model (7.7) to represent\\na regression spline. A cubic spline with K knots can be modeled as\\nyi = β0 + β1b1(xi) + β2b2(xi) + · · · + βK+3bK+3(xi) + ϵi,\\n(7.9)\\nfor an appropriate choice of basis functions b1, b2, . . . , bK+3. The model\\n(7.9) can then be fit using least squares.\\nJust as there were several ways to represent polynomials, there are also\\nmany equivalent ways to represent cubic splines using different choices of\\nbasis functions in (7.9). The most direct way to represent a cubic spline\\nusing (7.9) is to start off with a basis for a cubic polynomial—namely,\\nx, x2, and x3—and then add one truncated power basis function per knot. truncated\\npower basis\\n3Cubic splines are popular because most human eyes cannot detect the discontinuity\\nat the knots. 7.4 Regression Splines\\n297\\n20\\n30\\n40\\n50\\n60\\n70\\n50\\n100\\n150\\n200\\n250\\nAge\\nWage\\nNatural Cubic Spline\\nCubic Spline\\nFIGURE 7.4. A cubic spline and a natural cubic spline, with three knots, fit to\\na subset of the Wage data. The dashed lines denote the knot locations.\\nA truncated power basis function is defined as\\nh(x, ξ) = (x −ξ)3\\n+ =\\nK\\n(x −ξ)3\\nif x > ξ\\n0\\notherwise,\\n(7.10)\\nwhere ξ is the knot. One can show that adding a term of the form β4h(x, ξ)\\nto the model (7.8) for a cubic polynomial will lead to a discontinuity in\\nonly the third derivative at ξ; the function will remain continuous, with\\ncontinuous first and second derivatives, at each of the knots.\\nIn other words, in order to fit a cubic spline to a data set with K knots, we\\nperform least squares regression with an intercept and 3 + K predictors, of\\nthe form X, X2, X3, h(X, ξ1), h(X, ξ2), . . . , h(X, ξK), where ξ1, . . . , ξK are\\nthe knots. This amounts to estimating a total of K + 4 regression coeffi-\\ncients; for this reason, fitting a cubic spline with K knots uses K+4 degrees\\nof freedom.\\nUnfortunately, splines can have high variance at the outer range of the\\npredictors—that is, when X takes on either a very small or very large\\nvalue. Figure 7.4 shows a fit to the Wage data with three knots. We see that\\nthe confidence bands in the boundary region appear fairly wild. A natu-\\nral spline is a regression spline with additional boundary constraints: the natural\\nspline\\nfunction is required to be linear at the boundary (in the region where X is\\nsmaller than the smallest knot, or larger than the largest knot). This addi-\\ntional constraint means that natural splines generally produce more stable\\nestimates at the boundaries. In Figure 7.4, a natural cubic spline is also\\ndisplayed as a red line. Note that the corresponding confidence intervals\\nare narrower.\\n7.4.4\\nChoosing the Number and Locations of the Knots\\nWhen we fit a spline, where should we place the knots? The regression\\nspline is most flexible in regions that contain a lot of knots, because in\\nthose regions the polynomial coefficients can change rapidly. Hence, one 298\\n7. Moving Beyond Linearity\\n20\\n30\\n40\\n50\\n60\\n70\\n80\\n50\\n100\\n150\\n200\\n250\\n300\\nAge\\nWage\\nNatural Cubic Spline\\n20\\n30\\n40\\n50\\n60\\n70\\n80\\n0.00\\n0.05\\n0.10\\n0.15\\n0.20\\nAge\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n| |\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n| |\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|||\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| ||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n||\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| ||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n| |\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n| |\\n|\\n|\\n||\\n|\\n|\\n|\\n||\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|||\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n| | |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n||| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n| |\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n||\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n| |\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n||\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n| |\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n| ||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|| |\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n||\\n|\\n|\\n| |\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n| |\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n|\\n||\\n|\\n|\\n||\\n|\\nPr(Wage>250 | Age)\\nFIGURE 7.5. A natural cubic spline function with four degrees of freedom is\\nfit to the Wage data. Left: A spline is fit to wage (in thousands of dollars) as\\na function of age. Right: Logistic regression is used to model the binary event\\nwage>250 as a function of age. The fitted posterior probability of wage exceeding\\n$250,000 is shown. The dashed lines denote the knot locations.\\noption is to place more knots in places where we feel the function might\\nvary most rapidly, and to place fewer knots where it seems more stable.\\nWhile this option can work well, in practice it is common to place knots in\\na uniform fashion. One way to do this is to specify the desired degrees of\\nfreedom, and then have the software automatically place the corresponding\\nnumber of knots at uniform quantiles of the data.\\nFigure 7.5 shows an example on the Wage data. As in Figure 7.4, we\\nhave fit a natural cubic spline with three knots, except this time the knot\\nlocations were chosen automatically as the 25th, 50th, and 75th percentiles\\nof age. This was specified by requesting four degrees of freedom. The ar-\\ngument by which four degrees of freedom leads to three interior knots is\\nsomewhat technical.4\\nHow many knots should we use, or equivalently how many degrees of\\nfreedom should our spline contain? One option is to try out different num-\\nbers of knots and see which produces the best looking curve. A somewhat\\nmore objective approach is to use cross-validation, as discussed in Chap-\\nters 5 and 6. With this method, we remove a portion of the data (say 10 %),\\nfit a spline with a certain number of knots to the remaining data, and then\\nuse the spline to make predictions for the held-out portion. We repeat this\\nprocess multiple times until each observation has been left out once, and\\n4There are actually five knots, including the two boundary knots. A cubic spline with\\nfive knots has nine degrees of freedom. But natural cubic splines have two additional\\nnatural constraints at each boundary to enforce linearity, resulting in 9 −4 = 5 degrees\\nof freedom. Since this includes a constant, which is absorbed in the intercept, we count\\nit as four degrees of freedom. 7.4 Regression Splines\\n299\\n2\\n4\\n6\\n8\\n10\\n1600\\n1620\\n1640\\n1660\\n1680\\nDegrees of Freedom of Natural Spline\\nMean Squared Error\\n2\\n4\\n6\\n8\\n10\\n1600\\n1620\\n1640\\n1660\\n1680\\nDegrees of Freedom of Cubic Spline\\nMean Squared Error\\nFIGURE 7.6. Ten-fold cross-validated mean squared errors for selecting the\\ndegrees of freedom when fitting splines to the Wage data. The response is wage\\nand the predictor age. Left: A natural cubic spline. Right: A cubic spline.\\nthen compute the overall cross-validated RSS. This procedure can be re-\\npeated for different numbers of knots K. Then the value of K giving the\\nsmallest RSS is chosen.\\nFigure 7.6 shows ten-fold cross-validated mean squared errors for splines\\nwith various degrees of freedom fit to the Wage data. The left-hand panel\\ncorresponds to a natural cubic spline and the right-hand panel to a cu-\\nbic spline. The two methods produce almost identical results, with clear\\nevidence that a one-degree fit (a linear regression) is not adequate. Both\\ncurves flatten out quickly, and it seems that three degrees of freedom for\\nthe natural spline and four degrees of freedom for the cubic spline are quite\\nadequate.\\nIn Section 7.7 we fit additive spline models simultaneously on several\\nvariables at a time. This could potentially require the selection of degrees\\nof freedom for each variable. In cases like this we typically adopt a more\\npragmatic approach and set the degrees of freedom to a fixed number, say\\nfour, for all terms.\\n7.4.5\\nComparison to Polynomial Regression\\nFigure 7.7 compares a natural cubic spline with 15 degrees of freedom to a\\ndegree-15 polynomial on the Wage data set. The extra flexibility in the poly-\\nnomial produces undesirable results at the boundaries, while the natural\\ncubic spline still provides a reasonable fit to the data. Regression splines\\noften give superior results to polynomial regression. This is because unlike\\npolynomials, which must use a high degree (exponent in the highest mono-\\nmial term, e.g. X15) to produce flexible fits, splines introduce flexibility\\nby increasing the number of knots but keeping the degree fixed. Generally,\\nthis approach produces more stable estimates. Splines also allow us to place\\nmore knots, and hence flexibility, over regions where the function f seems\\nto be changing rapidly, and fewer knots where f appears more stable. 300\\n7. Moving Beyond Linearity\\n20\\n30\\n40\\n50\\n60\\n70\\n80\\n50\\n100\\n150\\n200\\n250\\n300\\nAge\\nWage\\nNatural Cubic Spline\\nPolynomial\\nFIGURE 7.7. On the Wage data set, a natural cubic spline with 15 degrees\\nof freedom is compared to a degree-15 polynomial. Polynomials can show wild\\nbehavior, especially near the tails.\\n7.5\\nSmoothing Splines\\nIn the last section we discussed regression splines, which we create by spec-\\nifying a set of knots, producing a sequence of basis functions, and then\\nusing least squares to estimate the spline coefficients. We now introduce a\\nsomewhat different approach that also produces a spline.\\n7.5.1\\nAn Overview of Smoothing Splines\\nIn fitting a smooth curve to a set of data, what we really want to do is\\nfind some function, say g(x), that fits the observed data well: that is, we\\nwant RSS = )n\\ni=1(yi −g(xi))2 to be small. However, there is a problem\\nwith this approach. If we don’t put any constraints on g(xi), then we can\\nalways make RSS zero simply by choosing g such that it interpolates all\\nof the yi. Such a function would woefully overfit the data—it would be far\\ntoo flexible. What we really want is a function g that makes RSS small,\\nbut that is also smooth.\\nHow might we ensure that g is smooth? There are a number of ways to\\ndo this. A natural approach is to find the function g that minimizes\\nn\\n0\\ni=1\\n(yi −g(xi))2 + λ\\nL\\ng′′(t)2dt\\n(7.11)\\nwhere λ is a nonnegative tuning parameter. The function g that minimizes\\n(7.11) is known as a smoothing spline.\\nsmoothing\\nspline\\nWhat does (7.11) mean? Equation 7.11 takes the “Loss+Penalty” for-\\nmulation that we encounter in the context of ridge regression and the lasso\\nin Chapter 6. The term )n\\ni=1(yi −g(xi))2 is a loss function that encour- loss function\\nages g to fit the data well, and the term λ\\nM\\ng′′(t)2dt is a penalty term\\nthat penalizes the variability in g. The notation g′′(t) indicates the second\\nderivative of the function g. The first derivative g′(t) measures the slope 7.5 Smoothing Splines\\n301\\nof a function at t, and the second derivative corresponds to the amount by\\nwhich the slope is changing. Hence, broadly speaking, the second derivative\\nof a function is a measure of its roughness: it is large in absolute value if\\ng(t) is very wiggly near t, and it is close to zero otherwise. (The second\\nderivative of a straight line is zero; note that a line is perfectly smooth.)\\nThe\\nM\\nnotation is an integral, which we can think of as a summation over\\nthe range of t. In other words,\\nM\\ng′′(t)2dt is simply a measure of the total\\nchange in the function g′(t), over its entire range. If g is very smooth, then\\ng′(t) will be close to constant and\\nM\\ng′′(t)2dt will take on a small value.\\nConversely, if g is jumpy and variable then g′(t) will vary significantly and\\nM\\ng′′(t)2dt will take on a large value. Therefore, in (7.11), λ\\nM\\ng′′(t)2dt en-\\ncourages g to be smooth. The larger the value of λ, the smoother g will be.\\nWhen λ = 0, then the penalty term in (7.11) has no effect, and so the\\nfunction g will be very jumpy and will exactly interpolate the training\\nobservations. When λ →∞, g will be perfectly smooth—it will just be\\na straight line that passes as closely as possible to the training points.\\nIn fact, in this case, g will be the linear least squares line, since the loss\\nfunction in (7.11) amounts to minimizing the residual sum of squares. For\\nan intermediate value of λ, g will approximate the training observations\\nbut will be somewhat smooth. We see that λ controls the bias-variance\\ntrade-off of the smoothing spline.\\nThe function g(x) that minimizes (7.11) can be shown to have some spe-\\ncial properties: it is a piecewise cubic polynomial with knots at the unique\\nvalues of x1, . . . , xn, and continuous first and second derivatives at each\\nknot. Furthermore, it is linear in the region outside of the extreme knots.\\nIn other words, the function g(x) that minimizes (7.11) is a natural cubic\\nspline with knots at x1, . . . , xn! However, it is not the same natural cubic\\nspline that one would get if one applied the basis function approach de-\\nscribed in Section 7.4.3 with knots at x1, . . . , xn—rather, it is a shrunken\\nversion of such a natural cubic spline, where the value of the tuning pa-\\nrameter λ in (7.11) controls the level of shrinkage.\\n7.5.2\\nChoosing the Smoothing Parameter λ\\nWe have seen that a smoothing spline is simply a natural cubic spline\\nwith knots at every unique value of xi. It might seem that a smoothing\\nspline will have far too many degrees of freedom, since a knot at each data\\npoint allows a great deal of flexibility. But the tuning parameter λ controls\\nthe roughness of the smoothing spline, and hence the effective degrees of\\nfreedom. It is possible to show that as λ increases from 0 to ∞, the effective effective\\ndegrees of\\nfreedom\\ndegrees of freedom, which we write dfλ, decrease from n to 2.\\nIn the context of smoothing splines, why do we discuss effective degrees\\nof freedom instead of degrees of freedom? Usually degrees of freedom refer\\nto the number of free parameters, such as the number of coefficients fit in a\\npolynomial or cubic spline. Although a smoothing spline has n parameters\\nand hence n nominal degrees of freedom, these n parameters are heavily\\nconstrained or shrunk down. Hence dfλ is a measure of the flexibility of the\\nsmoothing spline—the higher it is, the more flexible (and the lower-bias but\\nhigher-variance) the smoothing spline. The definition of effective degrees of 302\\n7. Moving Beyond Linearity\\nfreedom is somewhat technical. We can write\\nˆgλ = Sλy,\\n(7.12)\\nwhere ˆgλ is the solution to (7.11) for a particular choice of λ—that is, it\\nis an n-vector containing the fitted values of the smoothing spline at the\\ntraining points x1, . . . , xn. Equation 7.12 indicates that the vector of fitted\\nvalues when applying a smoothing spline to the data can be written as a\\nn × n matrix Sλ (for which there is a formula) times the response vector\\ny. Then the effective degrees of freedom is defined to be\\ndfλ =\\nn\\n0\\ni=1\\n{Sλ}ii,\\n(7.13)\\nthe sum of the diagonal elements of the matrix Sλ.\\nIn fitting a smoothing spline, we do not need to select the number or\\nlocation of the knots—there will be a knot at each training observation,\\nx1, . . . , xn. Instead, we have another problem: we need to choose the value\\nof λ. It should come as no surprise that one possible solution to this problem\\nis cross-validation. In other words, we can find the value of λ that makes\\nthe cross-validated RSS as small as possible. It turns out that the leave-\\none-out cross-validation error (LOOCV) can be computed very efficiently\\nfor smoothing splines, with essentially the same cost as computing a single\\nfit, using the following formula:\\nRSScv(λ) =\\nn\\n0\\ni=1\\n(yi −ˆg(−i)\\nλ\\n(xi))2 =\\nn\\n0\\ni=1\\n3yi −ˆgλ(xi)\\n1 −{Sλ}ii\\n42\\n.\\nThe notation ˆg(−i)\\nλ\\n(xi) indicates the fitted value for this smoothing spline\\nevaluated at xi, where the fit uses all of the training observations except\\nfor the ith observation (xi, yi). In contrast, ˆgλ(xi) indicates the smoothing\\nspline function fit to all of the training observations and evaluated at xi.\\nThis remarkable formula says that we can compute each of these leave-\\none-out fits using only ˆgλ, the original fit to all of the data!5 We have\\na very similar formula (5.2) on page 205 in Chapter 5 for least squares\\nlinear regression. Using (5.2), we can very quickly perform LOOCV for\\nthe regression splines discussed earlier in this chapter, as well as for least\\nsquares regression using arbitrary basis functions.\\nFigure 7.8 shows the results from fitting a smoothing spline to the Wage\\ndata. The red curve indicates the fit obtained from pre-specifying that we\\nwould like a smoothing spline with 16 effective degrees of freedom. The blue\\ncurve is the smoothing spline obtained when λ is chosen using LOOCV; in\\nthis case, the value of λ chosen results in 6.8 effective degrees of freedom\\n(computed using (7.13)). For this data, there is little discernible difference\\nbetween the two smoothing splines, beyond the fact that the one with 16\\ndegrees of freedom seems slightly wigglier. Since there is little difference\\nbetween the two fits, the smoothing spline fit with 6.8 degrees of freedom\\n5The exact formulas for computing ˆg(xi) and Sλ are very technical; however, efficient\\nalgorithms are available for computing these quantities. 7.6 Local Regression\\n303\\n20\\n30\\n40\\n50\\n60\\n70\\n80\\n0\\n50\\n100\\n200\\n300\\nAge\\nWage\\nSmoothing Spline\\n16 Degrees of Freedom\\n6.8 Degrees of Freedom (LOOCV)\\nFIGURE 7.8. Smoothing spline fits to the Wage data. The red curve results\\nfrom specifying 16 effective degrees of freedom. For the blue curve, λ was found\\nautomatically by leave-one-out cross-validation, which resulted in 6.8 effective\\ndegrees of freedom.\\nis preferable, since in general simpler models are better unless the data\\nprovides evidence in support of a more complex model.\\n7.6\\nLocal Regression\\nLocal regression is a different approach for fitting flexible non-linear func- local\\nregression\\ntions, which involves computing the fit at a target point x0 using only the\\nnearby training observations. Figure 7.9 illustrates the idea on some simu-\\nlated data, with one target point near 0.4, and another near the boundary\\nat 0.05. In this figure the blue line represents the function f(x) from which\\nthe data were generated, and the light orange line corresponds to the local\\nregression estimate ˆf(x). Local regression is described in Algorithm 7.1.\\nNote that in Step 3 of Algorithm 7.1, the weights Ki0 will differ for each\\nvalue of x0. In other words, in order to obtain the local regression fit at a\\nnew point, we need to fit a new weighted least squares regression model by\\nminimizing (7.14) for a new set of weights. Local regression is sometimes\\nreferred to as a memory-based procedure, because like nearest-neighbors, we\\nneed all the training data each time we wish to compute a prediction. We\\nwill avoid getting into the technical details of local regression here—there\\nare books written on the topic.\\nIn order to perform local regression, there are a number of choices to\\nbe made, such as how to define the weighting function K, and whether\\nto fit a linear, constant, or quadratic regression in Step 3. (Equation 7.14\\ncorresponds to a linear regression.) While all of these choices make some\\ndifference, the most important choice is the span s, which is the proportion\\nof points used to compute the local regression at x0, as defined in Step 1\\nabove. The span plays a role like that of the tuning parameter λ in smooth- 304\\n7. Moving Beyond Linearity\\n0.0\\n0.2\\n0.4\\n0.6\\n0.8\\n1.0\\n−1.0\\n−0.5\\n0.0\\n0.5\\n1.0\\n1.5\\nO\\nO\\nO\\nO\\nO\\nOO\\nO\\nO\\nO\\nO\\nO\\nO\\nO\\nO\\nOOO\\nO\\nO\\nO\\nO\\nO\\nO\\nO\\nO\\nOO\\nO\\nO\\nOO\\nO\\nO\\nO\\nO\\nO\\nO\\nOO\\nO\\nO\\nO\\nO\\nO\\nO\\nO\\nO\\nO\\nO\\nOO\\nO\\nO\\nO\\nO\\nO\\nOO\\nO\\nO\\nO\\nO\\nOO\\nO\\nO\\nOO\\nO\\nO\\nO\\nOO\\nO\\nO\\nO\\nO\\nO\\nO\\nO\\nOO\\nO\\nO\\nO\\nOO\\nO\\nO\\nO\\nO\\nOO\\nO\\nO\\nO\\nO\\nO\\nO\\nO\\nO\\nO\\nO\\nO\\nOO\\nO\\nO\\nO\\nO\\nO\\nO\\nO\\nO\\nOOO\\nO\\nO\\nO\\nO\\n0.0\\n0.2\\n0.4\\n0.6\\n0.8\\n1.0\\n−1.0\\n−0.5\\n0.0\\n0.5\\n1.0\\n1.5\\nO\\nO\\nO\\nO\\nO\\nOO\\nO\\nO\\nO\\nO\\nO\\nO\\nO\\nO\\nOOO\\nO\\nO\\nO\\nO\\nO\\nO\\nO\\nO\\nOO\\nO\\nO\\nOO\\nO\\nO\\nO\\nO\\nO\\nO\\nOO\\nO\\nO\\nO\\nO\\nO\\nO\\nO\\nO\\nO\\nO\\nOO\\nO\\nO\\nO\\nO\\nO\\nOO\\nO\\nO\\nO\\nO\\nOO\\nO\\nO\\nOO\\nO\\nO\\nO\\nOO\\nO\\nO\\nO\\nO\\nO\\nO\\nO\\nOO\\nO\\nO\\nO\\nOO\\nO\\nO\\nO\\nO\\nOO\\nO\\nO\\nO\\nO\\nO\\nO\\nO\\nO\\nO\\nO\\nO\\nOO\\nO\\nO\\nOO\\nO\\nO\\nO\\nO\\nO\\nO\\nOO\\nO\\nO\\nO\\nO\\nO\\nO\\nO\\nO\\nO\\nO\\nOO\\nO\\nO\\nO\\nO\\nO\\nOO\\nO\\nO\\nO\\nO\\nOO\\nO\\nO\\nOO\\nO\\nO\\nLocal Regression\\nFIGURE 7.9. Local regression illustrated on some simulated data, where the\\nblue curve represents f(x) from which the data were generated, and the light\\norange curve corresponds to the local regression estimate ˆf(x). The orange colored\\npoints are local to the target point x0, represented by the orange vertical line. The\\nyellow bell-shape superimposed on the plot indicates weights assigned to each\\npoint, decreasing to zero with distance from the target point. The fit ˆf(x0) at x0\\nis obtained by fitting a weighted linear regression (orange line segment), and using\\nthe fitted value at x0 (orange solid dot) as the estimate ˆf(x0).\\ning splines: it controls the flexibility of the non-linear fit. The smaller the\\nvalue of s, the more local and wiggly will be our fit; alternatively, a very\\nlarge value of s will lead to a global fit to the data using all of the train-\\ning observations. We can again use cross-validation to choose s, or we can\\nspecify it directly. Figure 7.10 displays local linear regression fits on the\\nWage data, using two values of s: 0.7 and 0.2. As expected, the fit obtained\\nusing s = 0.7 is smoother than that obtained using s = 0.2.\\nThe idea of local regression can be generalized in many different ways.\\nIn a setting with multiple features X1, X2, . . . , Xp, one very useful general-\\nization involves fitting a multiple linear regression model that is global in\\nsome variables, but local in another, such as time. Such varying coefficient\\nmodels are a useful way of adapting a model to the most recently gathered varying\\ncoefficient\\nmodel\\ndata. Local regression also generalizes very naturally when we want to fit\\nmodels that are local in a pair of variables X1 and X2, rather than one.\\nWe can simply use two-dimensional neighborhoods, and fit bivariate linear\\nregression models using the observations that are near each target point\\nin two-dimensional space. Theoretically the same approach can be imple-\\nmented in higher dimensions, using linear regressions fit to p-dimensional\\nneighborhoods. However, local regression can perform poorly if p is much\\nlarger than about 3 or 4 because there will generally be very few training\\nobservations close to x0. Nearest-neighbors regression, discussed in Chap-\\nter 3, suffers from a similar problem in high dimensions. 7.7 Generalized Additive Models\\n305\\nAlgorithm 7.1 Local Regression At X = x0\\n1. Gather the fraction s = k/n of training points whose xi are closest\\nto x0.\\n2. Assign a weight Ki0 = K(xi, x0) to each point in this neighborhood,\\nso that the point furthest from x0 has weight zero, and the closest\\nhas the highest weight. All but these k nearest neighbors get weight\\nzero.\\n3. Fit a weighted least squares regression of the yi on the xi using the\\naforementioned weights, by finding ˆβ0 and ˆβ1 that minimize\\nn\\n0\\ni=1\\nKi0(yi −β0 −β1xi)2.\\n(7.14)\\n4. The fitted value at x0 is given by ˆf(x0) = ˆβ0 + ˆβ1x0.\\n20\\n30\\n40\\n50\\n60\\n70\\n80\\n0\\n50\\n100\\n200\\n300\\nAge\\nWage\\nLocal Linear Regression\\nSpan is 0.2  (16.4 Degrees of Freedom)\\nSpan is 0.7  (5.3 Degrees of Freedom)\\nFIGURE 7.10. Local linear fits to the Wage data. The span specifies the fraction\\nof the data used to compute the fit at each target point.\\n7.7\\nGeneralized Additive Models\\nIn Sections 7.1–7.6, we present a number of approaches for flexibly predict-\\ning a response Y on the basis of a single predictor X. These approaches can\\nbe seen as extensions of simple linear regression. Here we explore the prob-\\nlem of flexibly predicting Y on the basis of several predictors, X1, . . . , Xp.\\nThis amounts to an extension of multiple linear regression.\\nGeneralized additive models (GAMs) provide a general framework for generalized\\nadditive\\nmodel\\nextending a standard linear model by allowing non-linear functions of each\\nof the variables, while maintaining additivity. Just like linear models, GAMs\\nadditivity\\ncan be applied with both quantitative and qualitative responses. We first 306\\n7. Moving Beyond Linearity\\n2003\\n2005\\n2007\\n2009\\n−30\\n−20\\n−10\\n0\\n10\\n20\\n30\\n20\\n30\\n40\\n50\\n60\\n70\\n80\\n−50\\n−40\\n−30\\n−20\\n−10\\n0\\n10\\n20\\n−30\\n−20\\n−10\\n0\\n10\\n20\\n30\\n40\\n<HS\\nHS\\n<Coll\\nColl\\n>Coll\\nf1(year)\\nf2(age)\\nf3(education)\\nyear\\nage\\neducation\\nFIGURE 7.11. For the Wage data, plots of the relationship between each feature\\nand the response, wage, in the fitted model (7.16). Each plot displays the fitted\\nfunction and pointwise standard errors. The first two functions are natural splines\\nin year and age, with four and five degrees of freedom, respectively. The third\\nfunction is a step function, fit to the qualitative variable education.\\nexamine GAMs for a quantitative response in Section 7.7.1, and then for a\\nqualitative response in Section 7.7.2.\\n7.7.1\\nGAMs for Regression Problems\\nA natural way to extend the multiple linear regression model\\nyi = β0 + β1xi1 + β2xi2 + · · · + βpxip + ϵi\\nin order to allow for non-linear relationships between each feature and the\\nresponse is to replace each linear component βjxij with a (smooth) non-\\nlinear function fj(xij). We would then write the model as\\nyi\\n=\\nβ0 +\\np\\n0\\nj=1\\nfj(xij) + ϵi\\n=\\nβ0 + f1(xi1) + f2(xi2) + · · · + fp(xip) + ϵi.\\n(7.15)\\nThis is an example of a GAM. It is called an additive model because we\\ncalculate a separate fj for each Xj, and then add together all of their\\ncontributions.\\nIn Sections 7.1–7.6, we discuss many methods for fitting functions to a\\nsingle variable. The beauty of GAMs is that we can use these methods\\nas building blocks for fitting an additive model. In fact, for most of the\\nmethods that we have seen so far in this chapter, this can be done fairly\\ntrivially. Take, for example, natural splines, and consider the task of fitting\\nthe model\\nwage = β0 + f1(year) + f2(age) + f3(education) + ϵ\\n(7.16)\\non the Wage data. Here year and age are quantitative variables, while the\\nvariable education is qualitative with five levels: <HS, HS, <Coll, Coll, >Coll,\\nreferring to the amount of high school or college education that an individ-\\nual has completed. We fit the first two functions using natural splines. We 7.7 Generalized Additive Models\\n307\\n2003\\n2005\\n2007\\n2009\\n−30\\n−20\\n−10\\n0\\n10\\n20\\n30\\n20\\n30\\n40\\n50\\n60\\n70\\n80\\n−50\\n−40\\n−30\\n−20\\n−10\\n0\\n10\\n20\\n−30\\n−20\\n−10\\n0\\n10\\n20\\n30\\n40\\n<HS\\nHS\\n<Coll\\nColl\\n>Coll\\nf1(year)\\nf2(age)\\nf3(education)\\nyear\\nage\\neducation\\nFIGURE 7.12. Details are as in Figure 7.11, but now f1 and f2 are smoothing\\nsplines with four and five degrees of freedom, respectively.\\nfit the third function using a separate constant for each level, via the usual\\ndummy variable approach of Section 3.3.1.\\nFigure 7.11 shows the results of fitting the model (7.16) using least\\nsquares. This is easy to do, since as discussed in Section 7.4, natural splines\\ncan be constructed using an appropriately chosen set of basis functions.\\nHence the entire model is just a big regression onto spline basis variables\\nand dummy variables, all packed into one big regression matrix.\\nFigure 7.11 can be easily interpreted. The left-hand panel indicates that\\nholding age and education fixed, wage tends to increase slightly with year;\\nthis may be due to inflation. The center panel indicates that holding\\neducation and year fixed, wage tends to be highest for intermediate val-\\nues of age, and lowest for the very young and very old. The right-hand\\npanel indicates that holding year and age fixed, wage tends to increase\\nwith education: the more educated a person is, the higher their salary, on\\naverage. All of these findings are intuitive.\\nFigure 7.12 shows a similar triple of plots, but this time f1 and f2 are\\nsmoothing splines with four and five degrees of freedom, respectively. Fit-\\nting a GAM with a smoothing spline is not quite as simple as fitting a GAM\\nwith a natural spline, since in the case of smoothing splines, least squares\\ncannot be used. However, standard software such as the Python package\\npygam can be used to fit GAMs using smoothing splines, via an approach pygam\\nknown as backfitting. This method fits a model involving multiple predic- backfitting\\ntors by repeatedly updating the fit for each predictor in turn, holding the\\nothers fixed. The beauty of this approach is that each time we update a\\nfunction, we simply apply the fitting method for that variable to a partial\\nresidual.6\\nThe fitted functions in Figures 7.11 and 7.12 look rather similar. In most\\nsituations, the differences in the GAMs obtained using smoothing splines\\nversus natural splines are small.\\n6A partial residual for X3, for example, has the form ri = yi −f1(xi1)−f2(xi2). If we\\nknow f1 and f2, then we can fit f3 by treating this residual as a response in a non-linear\\nregression on X3. 308\\n7. Moving Beyond Linearity\\nWe do not have to use splines as the building blocks for GAMs: we can\\njust as well use local regression, polynomial regression, or any combination\\nof the approaches seen earlier in this chapter in order to create a GAM.\\nGAMs are investigated in further detail in the lab at the end of this chapter.\\nPros and Cons of GAMs\\nBefore we move on, let us summarize the advantages and limitations of a\\nGAM.\\nL GAMs allow us to fit a non-linear fj to each Xj, so that we can\\nautomatically model non-linear relationships that standard linear re-\\ngression will miss. This means that we do not need to manually try\\nout many different transformations on each variable individually.\\nL The non-linear fits can potentially make more accurate predictions\\nfor the response Y .\\nL Because the model is additive, we can examine the effect of each Xj\\non Y individually while holding all of the other variables fixed.\\nL The smoothness of the function fj for the variable Xj can be sum-\\nmarized via degrees of freedom.\\nN The main limitation of GAMs is that the model is restricted to be\\nadditive. With many variables, important interactions can be missed.\\nHowever, as with linear regression, we can manually add interaction\\nterms to the GAM model by including additional predictors of the\\nform Xj × Xk. In addition we can add low-dimensional interaction\\nfunctions of the form fjk(Xj, Xk) into the model; such terms can\\nbe fit using two-dimensional smoothers such as local regression, or\\ntwo-dimensional splines (not covered here).\\nFor fully general models, we have to look for even more flexible approaches\\nsuch as random forests and boosting, described in Chapter 8. GAMs provide\\na useful compromise between linear and fully nonparametric models.\\n7.7.2\\nGAMs for Classification Problems\\nGAMs can also be used in situations where Y is qualitative. For simplicity,\\nhere we assume Y takes on values 0 or 1, and let p(X) = Pr(Y = 1|X) be\\nthe conditional probability (given the predictors) that the response equals\\none. Recall the logistic regression model (4.6):\\nlog\\n*\\np(X)\\n1 −p(X)\\n+\\n= β0 + β1X1 + β2X2 + · · · + βpXp.\\n(7.17)\\nThe left-hand side is the log of the odds of P(Y = 1|X) versus P(Y = 0|X),\\nwhich (7.17) represents as a linear function of the predictors. A natural way\\nto extend (7.17) to allow for non-linear relationships is to use the model\\nlog\\n*\\np(X)\\n1 −p(X)\\n+\\n= β0 + f1(X1) + f2(X2) + · · · + fp(Xp).\\n(7.18) 7.8 Lab: Non-Linear Modeling\\n309\\n2003\\n2005\\n2007\\n2009\\n−4\\n−2\\n0\\n2\\n4\\n20\\n30\\n40\\n50\\n60\\n70\\n80\\n−8\\n−6\\n−4\\n−2\\n0\\n2\\n−400\\n−200\\n0\\n200\\n400\\n<HS\\nHS\\n<Coll\\nColl\\n>Coll\\nf1(year)\\nf2(age)\\nf3(education)\\nyear\\nage\\neducation\\nFIGURE 7.13. For the Wage data, the logistic regression GAM given in (7.19)\\nis fit to the binary response I(wage>250). Each plot displays the fitted function\\nand pointwise standard errors. The first function is linear in year, the second\\nfunction a smoothing spline with five degrees of freedom in age, and the third a\\nstep function for education. There are very wide standard errors for the first\\nlevel <HS of education.\\nEquation 7.18 is a logistic regression GAM. It has all the same pros and\\ncons as discussed in the previous section for quantitative responses.\\nWe fit a GAM to the Wage data in order to predict the probability that\\nan individual’s income exceeds $250,000 per year. The GAM that we fit\\ntakes the form\\nlog\\n*\\np(X)\\n1 −p(X)\\n+\\n= β0 + β1 × year + f2(age) + f3(education),\\n(7.19)\\nwhere\\np(X) = Pr(wage > 250|year, age, education).\\nOnce again f2 is fit using a smoothing spline with five degrees of freedom,\\nand f3 is fit as a step function, by creating dummy variables for each of the\\nlevels of education. The resulting fit is shown in Figure 7.13. The last panel\\nlooks suspicious, with very wide confidence intervals for level <HS. In fact,\\nno response values equal one for that category: no individuals with less than\\na high school education make more than $250,000 per year. Hence we refit\\nthe GAM, excluding the individuals with less than a high school education.\\nThe resulting model is shown in Figure 7.14. As in Figures 7.11 and 7.12,\\nall three panels have similar vertical scales. This allows us to visually assess\\nthe relative contributions of each of the variables. We observe that age and\\neducation have a much larger effect than year on the probability of being\\na high earner.\\n7.8\\nLab: Non-Linear Modeling\\nIn this lab, we demonstrate some of the nonlinear models discussed in\\nthis chapter. We use the Wage data as a running example, and show that\\nmany of the complex non-linear fitting procedures discussed can easily be\\nimplemented in Python. 310\\n7. Moving Beyond Linearity\\n2003\\n2005\\n2007\\n2009\\n−4\\n−2\\n0\\n2\\n4\\n20\\n30\\n40\\n50\\n60\\n70\\n80\\n−8\\n−6\\n−4\\n−2\\n0\\n2\\n−4\\n−2\\n0\\n2\\n4\\nHS\\n<Coll\\nColl\\n>Coll\\nf1(year)\\nf2(age)\\nf3(education)\\nyear\\nage\\neducation\\nFIGURE 7.14. The same model is fit as in Figure 7.13, this time excluding the\\nobservations for which education is <HS. Now we see that increased education\\ntends to be associated with higher salaries.\\nAs usual, we start with some of our standard imports.\\nIn [1]: import numpy as np, pandas as pd\\nfrom matplotlib.pyplot import subplots\\nimport statsmodels.api as sm\\nfrom ISLP import load_data\\nfrom ISLP.models import (summarize,\\npoly,\\nModelSpec as MS)\\nfrom statsmodels.stats.anova import anova_lm\\nWe again collect the new imports needed for this lab. Many of these are\\ndeveloped specifically for the ISLP package.\\nIn [2]: from pygam import (s as s_gam,\\nl as l_gam,\\nf as f_gam,\\nLinearGAM,\\nLogisticGAM)\\nfrom ISLP.transforms import (BSpline,\\nNaturalSpline)\\nfrom ISLP.models import bs, ns\\nfrom ISLP.pygam import (approx_lam ,\\ndegrees_of_freedom ,\\nplot as plot_gam,\\nanova as anova_gam)\\n7.8.1\\nPolynomial Regression and Step Functions\\nWe start by demonstrating how Figure 7.1 can be reproduced. Let’s begin\\nby loading the data.\\nIn [3]: Wage = load_data(\\'Wage\\')\\ny = Wage[\\'wage\\']\\nage = Wage[\\'age\\'] 7.8 Lab: Non-Linear Modeling\\n311\\nThroughout most of this lab, our response is Wage[\\'wage\\'], which we\\nhave stored as y above. As in Section 3.6.6, we will use the poly() function\\nto create a model matrix that will fit a 4th degree polynomial in age.\\nIn [4]: poly_age = MS([poly(\\'age\\', degree=4)]).fit(Wage)\\nM = sm.OLS(y, poly_age.transform(Wage)).fit()\\nsummarize(M)\\nOut[4]:\\ncoef\\nstd err\\nt\\nP>|t|\\nintercept\\n111.7036\\n0.729\\n153.283\\n0.000\\npoly(age, degree=4)[0]\\n447.0679\\n39.915\\n11.201\\n0.000\\npoly(age, degree=4)[1] -478.3158\\n39.915\\n-11.983\\n0.000\\npoly(age, degree=4)[2]\\n125.5217\\n39.915\\n3.145\\n0.002\\npoly(age, degree=4)[3]\\n-77.9112\\n39.915\\n-1.952\\n0.051\\nThis polynomial is constructed using the function poly(), which cre-\\nates a special transformer Poly() (using sklearn terminology for feature transformer\\ntransformations such as PCA() seen in Section 6.5.3) which allows for easy\\nevaluation of the polynomial at new data points. Here poly() is referred\\nto as a helper function, and sets up the transformation; Poly() is the ac- helper\\ntual workhorse that computes the transformation. See also the discussion\\nof transformations on page 118.\\nIn the code above, the first line executes the fit() method using the\\ndataframe Wage. This recomputes and stores as attributes any parameters\\nneeded by Poly() on the training data, and these will be used on all sub-\\nsequent evaluations of the transform() method. For example, it is used on\\nthe second line, as well as in the plotting function developed below.\\nWe now create a grid of values for age at which we want predictions.\\nIn [5]: age_grid = np.linspace(age.min(),\\nage.max(),\\n100)\\nage_df = pd.DataFrame({\\'age\\': age_grid})\\nFinally, we wish to plot the data and add the fit from the fourth-degree\\npolynomial. As we will make several similar plots below, we first write a\\nfunction to create all the ingredients and produce the plot. Our function\\ntakes in a model specification (here a basis specified by a transform), as\\nwell as a grid of age values. The function produces a fitted curve as well\\nas 95% confidence bands. By using an argument for basis we can produce\\nand plot the results with several different transforms, such as the splines\\nwe will see shortly.\\nIn [6]: def plot_wage_fit(age_df,\\nbasis,\\ntitle):\\nX = basis.transform(Wage)\\nXnew = basis.transform(age_df)\\nM = sm.OLS(y, X).fit()\\npreds = M.get_prediction(Xnew)\\nbands = preds.conf_int(alpha=0.05)\\nfig, ax = subplots(figsize=(8,8))\\nax.scatter(age,\\ny, 312\\n7. Moving Beyond Linearity\\nfacecolor=\\'gray\\',\\nalpha=0.5)\\nfor val, ls in zip([preds.predicted_mean ,\\nbands[:,0],\\nbands[:,1]],\\n[\\'b\\',\\'r--\\',\\'r--\\']):\\nax.plot(age_df.values, val, ls, linewidth=3)\\nax.set_title(title, fontsize=20)\\nax.set_xlabel(\\'Age\\', fontsize=20)\\nax.set_ylabel(\\'Wage\\', fontsize=20);\\nreturn ax\\nWe include an argument alpha to ax.scatter() to add some transparency\\nto the points. This provides a visual indication of density. Notice the use\\nof the zip() function in the for loop above (see Section 2.3.8). We have\\nthree lines to plot, each with different colors and line types. Here zip()\\nconveniently bundles these together as iterators in the loop.7\\niterator\\nWe now plot the fit of the fourth-degree polynomial using this function.\\nIn [7]: plot_wage_fit(age_df,\\npoly_age,\\n\\'Degree-4 Polynomial\\');\\nWith polynomial regression we must decide on the degree of the polyno-\\nmial to use. Sometimes we just wing it, and decide to use second or third\\ndegree polynomials, simply to obtain a nonlinear fit. But we can make such\\na decision in a more systematic way. One way to do this is through hypoth-\\nesis tests, which we demonstrate here. We now fit a series of models ranging\\nfrom linear (degree-one) to degree-five polynomials, and look to determine\\nthe simplest model that is sufficient to explain the relationship between\\nwage and age. We use the anova_lm() function, which performs a series of\\nANOVA tests. An analysis of variance or ANOVA tests the null hypothesis analysis of\\nvariance\\nthat a model M1 is sufficient to explain the data against the alternative\\nhypothesis that a more complex model M2 is required. The determination\\nis based on an F-test. To perform the test, the models M1 and M2 must\\nbe nested: the space spanned by the predictors in M1 must be a subspace\\nof the space spanned by the predictors in M2. In this case, we fit five dif-\\nferent polynomial models and sequentially compare the simpler model to\\nthe more complex model.\\nIn [8]: models = [MS([poly(\\'age\\', degree=d)])\\nfor d in range(1, 6)]\\nXs = [model.fit_transform(Wage) for model in models]\\nanova_lm(*[sm.OLS(y, X_).fit()\\nfor X_ in Xs])\\nOut[8]:\\ndf_resid\\nssr\\ndf_diff\\nss_diff\\nF\\nPr(>F)\\n0\\n2998.0\\n5.022e+06\\n0.0\\nNaN\\nNaN\\nNaN\\n1\\n2997.0\\n4.793e+06\\n1.0\\n228786.010\\n143.593\\n2.364e-32\\n2\\n2996.0\\n4.778e+06\\n1.0\\n15755.694\\n9.889\\n1.679e-03\\n3\\n2995.0\\n4.772e+06\\n1.0\\n6070.152\\n3.810\\n5.105e-02\\n7In Python speak, an “iterator” is an object with a finite number of values, that can\\nbe iterated on, as in a loop. 7.8 Lab: Non-Linear Modeling\\n313\\n4\\n2994.0\\n4.770e+06\\n1.0\\n1282.563\\n0.805\\n3.697e-01\\nNotice the * in the anova_lm() line above. This function takes a variable\\nnumber of non-keyword arguments, in this case fitted models. When these\\nmodels are provided as a list (as is done here), it must be prefixed by *.\\nThe p-value comparing the linear models[0] to the quadratic models[1]\\nis essentially zero, indicating that a linear fit is not sufficient.8 Similarly the\\np-value comparing the quadratic models[1] to the cubic models[2] is very\\nlow (0.0017), so the quadratic fit is also insufficient. The p-value comparing\\nthe cubic and degree-four polynomials, models[2] and models[3], is approx-\\nimately 5%, while the degree-five polynomial models[4] seems unnecessary\\nbecause its p-value is 0.37. Hence, either a cubic or a quartic polynomial\\nappear to provide a reasonable fit to the data, but lower- or higher-order\\nmodels are not justified.\\nIn this case, instead of using the anova() function, we could have obtained\\nthese p-values more succinctly by exploiting the fact that poly() creates\\northogonal polynomials.\\nIn [9]: summarize(M)\\nOut[9]:\\ncoef\\nstd err\\nt\\nP>|t|\\nintercept\\n111.7036\\n0.729\\n153.283\\n0.000\\npoly(age, degree=4)[0]\\n447.0679\\n39.915\\n11.201\\n0.000\\npoly(age, degree=4)[1] -478.3158\\n39.915\\n-11.983\\n0.000\\npoly(age, degree=4)[2]\\n125.5217\\n39.915\\n3.145\\n0.002\\npoly(age, degree=4)[3]\\n-77.9112\\n39.915\\n-1.952\\n0.051\\nNotice that the p-values are the same, and in fact the square of the\\nt-statistics are equal to the F-statistics from the anova_lm() function; for\\nexample:\\nIn [10]: (-11.983)**2\\nOut[10]: 143.59228\\nHowever, the ANOVA method works whether or not we used orthogonal\\npolynomials, provided the models are nested. For example, we can use\\nanova_lm() to compare the following three models, which all have a linear\\nterm in education and a polynomial in age of different degrees:\\nIn [11]: models = [MS([\\'education\\', poly(\\'age\\', degree=d)])\\nfor d in range(1, 4)]\\nXEs = [model.fit_transform(Wage)\\nfor model in models]\\nanova_lm(*[sm.OLS(y, X_).fit() for X_ in XEs])\\nOut[11]:\\ndf_resid\\nssr\\ndf_diff\\nss_diff\\nF\\nPr(>F)\\n0\\n2997.0\\n3.902e+06\\n0.0\\nNaN\\nNaN\\nNaN\\n1\\n2996.0\\n3.759e+06\\n1.0\\n142862.701\\n113.992\\n3.838e-26\\n2\\n2995.0\\n3.754e+06\\n1.0\\n5926.207\\n4.729\\n2.974e-02\\n8Indexing starting at zero is confusing for the polynomial degree example, since\\nmodels[1] is quadratic rather than linear! 314\\n7. Moving Beyond Linearity\\nAs an alternative to using hypothesis tests and ANOVA, we could choose\\nthe polynomial degree using cross-validation, as discussed in Chapter 5.\\nNext we consider the task of predicting whether an individual earns more\\nthan $250,000 per year. We proceed much as before, except that first we\\ncreate the appropriate response vector, and then apply the glm() function\\nusing the binomial family in order to fit a polynomial logistic regression\\nmodel.\\nIn [12]: X = poly_age.transform(Wage)\\nhigh_earn = Wage[\\'high_earn\\'] = y > 250 # shorthand\\nglm = sm.GLM(y > 250,\\nX,\\nfamily=sm.families.Binomial())\\nB = glm.fit()\\nsummarize(B)\\nOut[12]:\\ncoef\\nstd err\\nz\\nP>|z|\\nintercept\\n-4.3012\\n0.345 -12.457\\n0.000\\npoly(age, degree=4)[0]\\n71.9642\\n26.133\\n2.754\\n0.006\\npoly(age, degree=4)[1] -85.7729\\n35.929\\n-2.387\\n0.017\\npoly(age, degree=4)[2]\\n34.1626\\n19.697\\n1.734\\n0.083\\npoly(age, degree=4)[3] -47.4008\\n24.105\\n-1.966\\n0.049\\nOnce again, we make predictions using the get_prediction() method.\\nIn [13]: newX = poly_age.transform(age_df)\\npreds = B.get_prediction(newX)\\nbands = preds.conf_int(alpha=0.05)\\nWe now plot the estimated relationship.\\nIn [14]: fig, ax = subplots(figsize=(8,8))\\nrng = np.random.default_rng(0)\\nax.scatter(age +\\n0.2 * rng.uniform(size=y.shape[0]),\\nnp.where(high_earn, 0.198, 0.002),\\nfc=\\'gray\\',\\nmarker=\\'|\\')\\nfor val, ls in zip([preds.predicted_mean ,\\nbands[:,0],\\nbands[:,1]],\\n[\\'b\\',\\'r--\\',\\'r--\\']):\\nax.plot(age_df.values, val, ls, linewidth=3)\\nax.set_title(\\'Degree -4 Polynomial\\', fontsize=20)\\nax.set_xlabel(\\'Age\\', fontsize=20)\\nax.set_ylim([0,0.2])\\nax.set_ylabel(\\'P(Wage > 250)\\', fontsize=20);\\nWe have drawn the age values corresponding to the observations with wage\\nvalues above 250 as gray marks on the top of the plot, and those with\\nwage values below 250 are shown as gray marks on the bottom of the plot.\\nWe added a small amount of noise to jitter the age values a bit so that\\nobservations with the same age value do not cover each other up. This type\\nof plot is often called a rug plot.\\nrug plot\\nIn order to fit a step function, as discussed in Section 7.2, we first use\\nthe pd.qcut() function to discretize age based on quantiles. Then we use pd.qcut() 7.8 Lab: Non-Linear Modeling\\n315\\npd.get_dummies() to create the columns of the model matrix for this cate- pd.get_\\ndummies()\\ngorical variable. Note that this function will include all columns for a given\\ncategorical, rather than the usual approach which drops one of the levels.\\nIn [15]: cut_age = pd.qcut(age, 4)\\nsummarize(sm.OLS(y, pd.get_dummies(cut_age)).fit())\\nOut[15]:\\ncoef\\nstd err\\nt\\nP>|t|\\n(17.999, 33.75]\\n94.1584\\n1.478\\n63.692\\n0.0\\n(33.75, 42.0]\\n116.6608\\n1.470\\n79.385\\n0.0\\n(42.0, 51.0]\\n119.1887\\n1.416\\n84.147\\n0.0\\n(51.0, 80.0]\\n116.5717\\n1.559\\n74.751\\n0.0\\nHere pd.qcut() automatically picked the cutpoints based on the quan-\\ntiles 25%, 50% and 75%, which results in four regions. We could also have\\nspecified our own quantiles directly instead of the argument 4. For cuts\\nnot based on quantiles we would use the pd.cut() function. The function pd.cut()\\npd.qcut() (and pd.cut()) returns an ordered categorical variable. The re-\\ngression model then creates a set of dummy variables for use in the regres-\\nsion. Since age is the only variable in the model, the value $94,158.40 is the\\naverage salary for those under 33.75 years of age, and the other coefficients\\nare the average salary for those in the other age groups. We can produce\\npredictions and plots just as we did in the case of the polynomial fit.\\n7.8.2\\nSplines\\nIn order to fit regression splines, we use transforms from the ISLP package.\\nThe actual spline evaluation functions are in the scipy.interpolate pack-\\nage; we have simply wrapped them as transforms similar to Poly() and\\nPCA().\\nIn Section 7.4, we saw that regression splines can be fit by constructing\\nan appropriate matrix of basis functions. The BSpline() function generates BSpline()\\nthe entire matrix of basis functions for splines with the specified set of\\nknots. By default, the B-splines produced are cubic. To change the degree,\\nuse the argument degree.\\nIn [16]: bs_ = BSpline(internal_knots=[25,40,60], intercept=True).fit(age)\\nbs_age = bs_.transform(age)\\nbs_age.shape\\nOut[16]: (3000, 7)\\nThis results in a seven-column matrix, which is what is expected for a cubic-\\nspline basis with 3 interior knots. We can form this same matrix using the\\nbs() object, which facilitates adding this to a model-matrix builder (as in\\npoly() versus its workhorse Poly()) described in Section 7.8.1.\\nWe now fit a cubic spline model to the Wage data.\\nIn [17]: bs_age = MS([bs(\\'age\\', internal_knots=[25,40,60])])\\nXbs = bs_age.fit_transform(Wage)\\nM = sm.OLS(y, Xbs).fit()\\nsummarize(M) 316\\n7. Moving Beyond Linearity\\nOut[17]:\\ncoef\\nstd err\\n...\\nintercept\\n60.494\\n9.460\\n...\\nbs(age, internal_knots=[25, 40, 60])[0]\\n3.980\\n12.538\\n...\\nbs(age, internal_knots=[25, 40, 60])[1]\\n44.631\\n9.626\\n...\\nbs(age, internal_knots=[25, 40, 60])[2]\\n62.839\\n10.755\\n...\\nbs(age, internal_knots=[25, 40, 60])[3]\\n55.991\\n10.706\\n...\\nbs(age, internal_knots=[25, 40, 60])[4]\\n50.688\\n14.402\\n...\\nbs(age, internal_knots=[25, 40, 60])[5]\\n16.606\\n19.126\\n...\\nThe column names are a little cumbersome, and have caused us to trun-\\ncate the printed summary. They can be set on construction using the name\\nargument as follows.\\nIn [18]: bs_age = MS([bs(\\'age\\',\\ninternal_knots=[25,40,60],\\nname=\\'bs(age)\\')])\\nXbs = bs_age.fit_transform(Wage)\\nM = sm.OLS(y, Xbs).fit()\\nsummarize(M)\\nOut[18]:\\ncoef std err\\nt\\nP>|t|\\nintercept\\n60.494\\n9.460\\n6.394\\n0.000\\nbs(age, knots)[0]\\n3.981\\n12.538\\n0.317\\n0.751\\nbs(age, knots)[1]\\n44.631\\n9.626\\n4.636\\n0.000\\nbs(age, knots)[2]\\n62.839\\n10.755\\n5.843\\n0.000\\nbs(age, knots)[3]\\n55.991\\n10.706\\n5.230\\n0.000\\nbs(age, knots)[4]\\n50.688\\n14.402\\n3.520\\n0.000\\nbs(age, knots)[5]\\n16.606\\n19.126\\n0.868\\n0.385\\nNotice that there are 6 spline coefficients rather than 7. This is because, by\\ndefault, bs() assumes intercept=False, since we typically have an overall\\nintercept in the model. So it generates the spline basis with the given knots,\\nand then discards one of the basis functions to account for the intercept.\\nWe could also use the df (degrees of freedom) option to specify the com-\\nplexity of the spline. We see above that with 3 knots, the spline basis has\\n6 columns or degrees of freedom. When we specify df=6 rather than the\\nactual knots, bs() will produce a spline with 3 knots chosen at uniform\\nquantiles of the training data. We can see these chosen knots most easily\\nusing Bspline() directly:\\nIn [19]: BSpline(df=6).fit(age).internal_knots_\\nOut[19]: array([33.75, 42.0, 51.0])\\nWhen asking for six degrees of freedom, the transform chooses knots at\\nages 33.75, 42.0, and 51.0, which correspond to the 25th, 50th, and 75th\\npercentiles of age.\\nWhen using B-splines we need not limit ourselves to cubic polynomials\\n(i.e. degree=3). For instance, using degree=0 results in piecewise constant\\nfunctions, as in our example with pd.qcut() above.\\nIn [20]: bs_age0 = MS([bs(\\'age\\',\\ndf=3,\\ndegree=0)]).fit(Wage)\\nXbs0 = bs_age0.transform(Wage)\\nsummarize(sm.OLS(y, Xbs0).fit()) 7.8 Lab: Non-Linear Modeling\\n317\\nOut[20]:\\ncoef\\nstd err\\nt\\nP>|t|\\nintercept\\n94.158\\n1.478\\n63.687\\n0.0\\nbs(age, df=3, degree=0)[0]\\n22.349\\n2.152\\n10.388\\n0.0\\nbs(age, df=3, degree=0)[1]\\n24.808\\n2.044\\n12.137\\n0.0\\nbs(age, df=3, degree=0)[2]\\n22.781\\n2.087\\n10.917\\n0.0\\nThis fit should be compared with cell [15] where we use qcut() to create\\nfour bins by cutting at the 25%, 50% and 75% quantiles of age. Since we\\nspecified df=3 for degree-zero splines here, there will also be knots at the\\nsame three quantiles. Although the coefficients appear different, we see that\\nthis is a result of the different coding. For example, the first coefficient is\\nidentical in both cases, and is the mean response in the first bin. For the\\nsecond coefficient, we have 94.158 + 22.349 = 116.507 ≈116.611, the latter\\nbeing the mean in the second bin in cell [15]. Here the intercept is coded by\\na column of ones, so the second, third and fourth coefficients are increments\\nfor those bins. Why is the sum not exactly the same? It turns out that the\\nqcut() uses ≤, while bs() uses < when deciding bin membership.\\nIn order to fit a natural spline, we use the NaturalSpline() transform Natural\\nSpline()\\nwith the corresponding helper ns(). Here we fit a natural spline with five\\ndegrees of freedom (excluding the intercept) and plot the results.\\nIn [21]: ns_age = MS([ns(\\'age\\', df=5)]).fit(Wage)\\nM_ns = sm.OLS(y, ns_age.transform(Wage)).fit()\\nsummarize(M_ns)\\nOut[21]:\\ncoef\\nstd err\\nt\\nP>|t|\\nintercept\\n60.475\\n4.708\\n12.844\\n0.000\\nns(age, df=5)[0]\\n61.527\\n4.709\\n13.065\\n0.000\\nns(age, df=5)[1]\\n55.691\\n5.717\\n9.741\\n0.000\\nns(age, df=5)[2]\\n46.818\\n4.948\\n9.463\\n0.000\\nns(age, df=5)[3]\\n83.204\\n11.918\\n6.982\\n0.000\\nns(age, df=5)[4]\\n6.877\\n9.484\\n0.725\\n0.468\\nWe now plot the natural spline using our plotting function.\\nIn [22]: plot_wage_fit(age_df,\\nns_age,\\n\\'Natural spline, df=5\\');\\n7.8.3\\nSmoothing Splines and GAMs\\nA smoothing spline is a special case of a GAM with squared-error loss\\nand a single feature. To fit GAMs in Python we will use the pygam package pygam\\nwhich can be installed via pip install pygam. The estimator LinearGAM() LinearGAM()\\nuses squared-error loss. The GAM is specified by associating each column\\nof a model matrix with a particular smoothing operation: s for smoothing\\nspline; l for linear, and f for factor or categorical variables. The argument 0\\npassed to s below indicates that this smoother will apply to the first column\\nof a feature matrix. Below, we pass it a matrix with a single column: X_age.\\nThe argument lam is the penalty parameter λ as discussed in Section 7.5.2.\\nIn [23]: X_age = np.asarray(age).reshape((-1,1))\\ngam = LinearGAM(s_gam(0, lam=0.6))\\ngam.fit(X_age, y) 318\\n7. Moving Beyond Linearity\\nOut[23]: LinearGAM(callbacks=[Deviance(), Diffs()], fit_intercept=True,\\nmax_iter=100, scale=None, terms=s(0) + intercept, tol=0.0001,\\nverbose=False)\\nThe pygam library generally expects a matrix of features so we reshape age\\nto be a matrix (a two-dimensional array) instead of a vector (i.e. a one-\\ndimensional array). The -1 in the call to the reshape() method tells numpy\\nto impute the size of that dimension based on the remaining entries of the\\nshape tuple.\\nLet’s investigate how the fit changes with the smoothing parameter lam.\\nThe function np.logspace() is similar to np.linspace() but spaces points np.logspace()\\nevenly on the log-scale. Below we vary lam from 10−2 to 106.\\nIn [24]: fig, ax = subplots(figsize=(8,8))\\nax.scatter(age, y, facecolor=\\'gray\\', alpha=0.5)\\nfor lam in np.logspace(-2, 6, 5):\\ngam = LinearGAM(s_gam(0, lam=lam)).fit(X_age, y)\\nax.plot(age_grid,\\ngam.predict(age_grid),\\nlabel=\\'{:.1e}\\'.format(lam),\\nlinewidth=3)\\nax.set_xlabel(\\'Age\\', fontsize=20)\\nax.set_ylabel(\\'Wage\\', fontsize=20);\\nax.legend(title=\\'$\\\\lambda$\\');\\nThe pygam package can perform a search for an optimal smoothing pa-\\nrameter.\\nIn [25]: gam_opt = gam.gridsearch(X_age, y)\\nax.plot(age_grid,\\ngam_opt.predict(age_grid),\\nlabel=\\'Grid search\\',\\nlinewidth=4)\\nax.legend()\\nfig\\nAlternatively, we can fix the degrees of freedom of the smoothing spline\\nusing a function included in the ISLP.pygam package. Below we find a value\\nof λ that gives us roughly four degrees of freedom. We note here that these\\ndegrees of freedom include the unpenalized intercept and linear term of the\\nsmoothing spline, hence there are at least two degrees of freedom.\\nIn [26]: age_term = gam.terms[0]\\nlam_4 = approx_lam(X_age, age_term, 4)\\nage_term.lam = lam_4\\ndegrees_of_freedom(X_age, age_term)\\nOut[26]: 4.000000100004728\\nLet’s vary the degrees of freedom in a similar plot to above. We choose the\\ndegrees of freedom as the desired degrees of freedom plus one to account\\nfor the fact that these smoothing splines always have an intercept term.\\nHence, a value of one for df is just a linear fit.\\nIn [27]: fig, ax = subplots(figsize=(8,8))\\nax.scatter(X_age,\\ny, 7.8 Lab: Non-Linear Modeling\\n319\\nfacecolor=\\'gray\\',\\nalpha=0.3)\\nfor df in [1,3,4,8,15]:\\nlam = approx_lam(X_age, age_term, df+1)\\nage_term.lam = lam\\ngam.fit(X_age, y)\\nax.plot(age_grid,\\ngam.predict(age_grid),\\nlabel=\\'{:d}\\'.format(df),\\nlinewidth=4)\\nax.set_xlabel(\\'Age\\', fontsize=20)\\nax.set_ylabel(\\'Wage\\', fontsize=20);\\nax.legend(title=\\'Degrees of freedom\\');\\nAdditive Models with Several Terms\\nThe strength of generalized additive models lies in their ability to fit mul-\\ntivariate regression models with more flexibility than linear models. We\\ndemonstrate two approaches: the first in a more manual fashion using nat-\\nural splines and piecewise constant functions, and the second using the\\npygam package and smoothing splines.\\nWe now fit a GAM by hand to predict wage using natural spline functions\\nof year and age, treating education as a qualitative predictor, as in (7.16).\\nSince this is just a big linear regression model using an appropriate choice\\nof basis functions, we can simply do this using the sm.OLS() function.\\nWe will build the model matrix in a more manual fashion here, since we\\nwish to access the pieces separately when constructing partial dependence\\nplots.\\nIn [28]: ns_age = NaturalSpline(df=4).fit(age)\\nns_year = NaturalSpline(df=5).fit(Wage[\\'year\\'])\\nXs = [ns_age.transform(age),\\nns_year.transform(Wage[\\'year\\']),\\npd.get_dummies(Wage[\\'education\\']).values]\\nX_bh = np.hstack(Xs)\\ngam_bh = sm.OLS(y, X_bh).fit()\\nHere the function NaturalSpline() is the workhorse supporting the ns()\\nhelper function. We chose to use all columns of the indicator matrix for\\nthe categorical variable education, making an intercept redundant. Finally,\\nwe stacked the three component matrices horizontally to form the model\\nmatrix X_bh.\\nWe now show how to construct partial dependence plots for each of the\\nterms in our rudimentary GAM. We can do this by hand, given grids for\\nage and year. We simply predict with new X matrices, fixing all but one\\nof the features at a time.\\nIn [29]: age_grid = np.linspace(age.min(),\\nage.max(),\\n100)\\nX_age_bh = X_bh.copy()[:100]\\nX_age_bh[:] = X_bh[:].mean(0)[None,:]\\nX_age_bh[:,:4] = ns_age.transform(age_grid)\\npreds = gam_bh.get_prediction(X_age_bh)\\nbounds_age = preds.conf_int(alpha=0.05) 320\\n7. Moving Beyond Linearity\\npartial_age = preds.predicted_mean\\ncenter = partial_age.mean()\\npartial_age -= center\\nbounds_age -= center\\nfig, ax = subplots(figsize=(8,8))\\nax.plot(age_grid, partial_age , \\'b\\', linewidth=3)\\nax.plot(age_grid, bounds_age[:,0], \\'r--\\', linewidth=3)\\nax.plot(age_grid, bounds_age[:,1], \\'r--\\', linewidth=3)\\nax.set_xlabel(\\'Age\\')\\nax.set_ylabel(\\'Effect on wage\\')\\nax.set_title(\\'Partial dependence of age on wage\\', fontsize=20);\\nLet’s explain in some detail what we did above. The idea is to create a new\\nprediction matrix, where all but the columns belonging to age are constant\\n(and set to their training-data means). The four columns for age are filled\\nin with the natural spline basis evaluated at the 100 values in age_grid.\\n1. We made a grid of length 100 in age, and created a matrix X_age_bh\\nwith 100 rows and the same number of columns as X_bh.\\n2. We replaced every row of this matrix with the column means of the\\noriginal.\\n3. We then replace just the first four columns representing age with the\\nnatural spline basis computed at the values in age_grid.\\nThe remaining steps should by now be familiar.\\nWe also look at the effect of year on wage; the process is the same.\\nIn [30]: year_grid = np.linspace(2003, 2009, 100)\\nyear_grid = np.linspace(Wage[\\'year\\'].min(),\\nWage[\\'year\\'].max(),\\n100)\\nX_year_bh = X_bh.copy()[:100]\\nX_year_bh[:] = X_bh[:].mean(0)[None,:]\\nX_year_bh[:,4:9] = ns_year.transform(year_grid)\\npreds = gam_bh.get_prediction(X_year_bh)\\nbounds_year = preds.conf_int(alpha=0.05)\\npartial_year = preds.predicted_mean\\ncenter = partial_year.mean()\\npartial_year -= center\\nbounds_year -= center\\nfig, ax = subplots(figsize=(8,8))\\nax.plot(year_grid, partial_year , \\'b\\', linewidth=3)\\nax.plot(year_grid, bounds_year[:,0], \\'r--\\', linewidth=3)\\nax.plot(year_grid, bounds_year[:,1], \\'r--\\', linewidth=3)\\nax.set_xlabel(\\'Year\\')\\nax.set_ylabel(\\'Effect on wage\\')\\nax.set_title(\\'Partial dependence of year on wage\\', fontsize=20);\\nWe now fit the model (7.16) using smoothing splines rather than nat-\\nural splines. All of the terms in (7.16) are fit simultaneously, taking each\\nother into account to explain the response. The pygam package only works\\nwith matrices, so we must convert the categorical series education to its\\narray representation, which can be found with the cat.codes attribute of\\neducation. As year only has 7 unique values, we use only seven basis func-\\ntions for it. 7.8 Lab: Non-Linear Modeling\\n321\\nIn [31]: gam_full = LinearGAM(s_gam(0) +\\ns_gam(1, n_splines=7) +\\nf_gam(2, lam=0))\\nXgam = np.column_stack([age,\\nWage[\\'year\\'],\\nWage[\\'education\\'].cat.codes])\\ngam_full = gam_full.fit(Xgam, y)\\nThe two s_gam() terms result in smoothing spline fits, and use a default\\nvalue for λ (lam=0.6), which is somewhat arbitrary. For the categorical term\\neducation, specified using a f_gam() term, we specify lam=0 to avoid any\\nshrinkage. We produce the partial dependence plot in age to see the effect\\nof these choices.\\nThe values for the plot are generated by the pygam package. We provide\\na plot_gam() function for partial-dependence plots in ISLP.pygam, which plot_gam()\\nmakes this job easier than in our last example with natural splines.\\nIn [32]: fig, ax = subplots(figsize=(8,8))\\nplot_gam(gam_full, 0, ax=ax)\\nax.set_xlabel(\\'Age\\')\\nax.set_ylabel(\\'Effect on wage\\')\\nax.set_title(\\'Partial dependence of age on wage - default lam=0.6\\',\\nfontsize=20);\\nWe see that the function is somewhat wiggly. It is more natural to specify\\nthe df than a value for lam. We refit a GAM using four degrees of freedom\\neach for age and year. Recall that the addition of one below takes into\\naccount the intercept of the smoothing spline.\\nIn [33]: age_term = gam_full.terms[0]\\nage_term.lam = approx_lam(Xgam, age_term, df=4+1)\\nyear_term = gam_full.terms[1]\\nyear_term.lam = approx_lam(Xgam, year_term, df=4+1)\\ngam_full = gam_full.fit(Xgam, y)\\nNote that updating age_term.lam above updates it in gam_full.terms[0] as\\nwell! Likewise for year_term.lam.\\nRepeating the plot for age, we see that it is much smoother. We also\\nproduce the plot for year.\\nIn [34]: fig, ax = subplots(figsize=(8,8))\\nplot_gam(gam_full,\\n1,\\nax=ax)\\nax.set_xlabel(\\'Year\\')\\nax.set_ylabel(\\'Effect on wage\\')\\nax.set_title(\\'Partial dependence of year on wage\\', fontsize=20)\\nFinally we plot education, which is categorical. The partial dependence\\nplot is different, and more suitable for the set of fitted constants for each\\nlevel of this variable.\\nIn [35]: fig, ax = subplots(figsize=(8, 8))\\nax = plot_gam(gam_full, 2)\\nax.set_xlabel(\\'Education\\')\\nax.set_ylabel(\\'Effect on wage\\') 322\\n7. Moving Beyond Linearity\\nax.set_title(\\'Partial dependence of wage on education\\',\\nfontsize=20);\\nax.set_xticklabels(Wage[\\'education\\'].cat.categories, fontsize=8);\\nANOVA Tests for Additive Models\\nIn all of our models, the function of year looks rather linear. We can perform\\na series of ANOVA tests in order to determine which of these three models\\nis best: a GAM that excludes year (M1), a GAM that uses a linear function\\nof year (M2), or a GAM that uses a spline function of year (M3).\\nIn [36]: gam_0 = LinearGAM(age_term + f_gam(2, lam=0))\\ngam_0.fit(Xgam, y)\\ngam_linear = LinearGAM(age_term +\\nl_gam(1, lam=0) +\\nf_gam(2, lam=0))\\ngam_linear.fit(Xgam, y)\\nOut[36]: LinearGAM(callbacks=[Deviance(), Diffs()], fit_intercept=True,\\nmax_iter=100, scale=None, terms=s(0) + l(1) + f(2) + intercept,\\ntol=0.0001, verbose=False)\\nNotice our use of age_term in the expressions above. We do this because\\nearlier we set the value for lam in this term to achieve four degrees of\\nfreedom.\\nTo directly assess the effect of year we run an ANOVA on the three\\nmodels fit above.\\nIn [37]: anova_gam(gam_0, gam_linear, gam_full)\\nOut[37]:\\ndeviance\\ndf deviance_diff\\ndf_diff\\nF\\npvalue\\n0\\n3714362.366\\n2991.004\\nNaN\\nNaN\\nNaN\\nNaN\\n1\\n3696745.823\\n2990.005\\n17616.543\\n0.999\\n14.265\\n0.002\\n2\\n3693142.930\\n2987.007\\n3602.894\\n2.998\\n0.972\\n0.436\\nWe find that there is compelling evidence that a GAM with a linear function\\nin year is better than a GAM that does not include year at all (p-value=\\n0.002). However, there is no evidence that a non-linear function of year\\nis needed (p-value=0.435). In other words, based on the results of this\\nANOVA, M2 is preferred.\\nWe can repeat the same process for age as well. We see there is very clear\\nevidence that a non-linear term is required for age.\\nIn [38]: gam_0 = LinearGAM(year_term +\\nf_gam(2, lam=0))\\ngam_linear = LinearGAM(l_gam(0, lam=0) +\\nyear_term +\\nf_gam(2, lam=0))\\ngam_0.fit(Xgam, y)\\ngam_linear.fit(Xgam, y)\\nanova_gam(gam_0, gam_linear, gam_full) 7.8 Lab: Non-Linear Modeling\\n323\\nOut[38]:\\ndeviance\\ndf deviance_diff\\ndf_diff\\nF\\npvalue\\n0\\n3975443.045\\n2991.001\\nNaN\\nNaN\\nNaN\\nNaN\\n1\\n3850246.908\\n2990.001\\n125196.137\\n1.000\\n101.270\\n0.000\\n2\\n3693142.930\\n2987.007\\n157103.978\\n2.993\\n42.448\\n0.000\\nThere is a (verbose) summary() method for the GAM fit. (We do not\\nreproduce it here.)\\nIn [39]: gam_full.summary()\\nWe can make predictions from gam objects, just like from lm objects, using\\nthe predict() method for the class gam. Here we make predictions on the\\ntraining set.\\nIn [40]: Yhat = gam_full.predict(Xgam)\\nIn order to fit a logistic regression GAM, we use LogisticGAM() from LogisticGAM()\\npygam.\\nIn [41]: gam_logit = LogisticGAM(age_term +\\nl_gam(1, lam=0) +\\nf_gam(2, lam=0))\\ngam_logit.fit(Xgam, high_earn)\\nOut[41]: LogisticGAM(callbacks=[Deviance(), Diffs(), Accuracy()],\\nfit_intercept=True, max_iter=100,\\nterms=s(0) + l(1) + f(2) + intercept, tol=0.0001, verbose=False)\\nIn [42]: fig, ax = subplots(figsize=(8, 8))\\nax = plot_gam(gam_logit, 2)\\nax.set_xlabel(\\'Education\\')\\nax.set_ylabel(\\'Effect on wage\\')\\nax.set_title(\\'Partial dependence of wage on education\\',\\nfontsize=20);\\nax.set_xticklabels(Wage[\\'education\\'].cat.categories, fontsize=8);\\nThe model seems to be very flat, with especially high error bars for the\\nfirst category. Let’s look at the data a bit more closely.\\nIn [43]: pd.crosstab(Wage[\\'high_earn\\'], Wage[\\'education\\'])\\nWe see that there are no high earners in the first category of education,\\nmeaning that the model will have a hard time fitting. We will fit a logistic\\nregression GAM excluding all observations falling into this category. This\\nprovides more sensible results.\\nTo do so, we could subset the model matrix, though this will not remove\\nthe column from Xgam. While we can deduce which column corresponds to\\nthis feature, for reproducibility’s sake we reform the model matrix on this\\nsmaller subset.\\nIn [44]: only_hs = Wage[\\'education\\'] == \\'1. < HS Grad\\'\\nWage_ = Wage.loc[∼only_hs]\\nXgam_ = np.column_stack([Wage_[\\'age\\'],\\nWage_[\\'year\\'],\\nWage_[\\'education\\'].cat.codes -1])\\nhigh_earn_ = Wage_[\\'high_earn\\'] 324\\n7. Moving Beyond Linearity\\nIn the second-to-last line above, we subtract one from the codes of the\\ncategory, due to a bug in pygam. It just relabels the education values and\\nhence has no effect on the fit.\\nWe now fit the model.\\nIn [45]: gam_logit_ = LogisticGAM(age_term +\\nyear_term +\\nf_gam(2, lam=0))\\ngam_logit_.fit(Xgam_, high_earn_)\\nOut[45]: LogisticGAM(callbacks=[Deviance(), Diffs(), Accuracy()],\\nfit_intercept=True, max_iter=100,\\nterms=s(0) + s(1) + f(2) + intercept, tol=0.0001, verbose=False)\\nLet’s look at the effect of education, year and age on high earner status\\nnow that we’ve removed those observations.\\nIn [46]: fig, ax = subplots(figsize=(8, 8))\\nax = plot_gam(gam_logit_, 2)\\nax.set_xlabel(\\'Education\\')\\nax.set_ylabel(\\'Effect on wage\\')\\nax.set_title(\\'Partial dependence of high earner status on education\\n\\', fontsize=20);\\nax.set_xticklabels(Wage[\\'education\\'].cat.categories[1:],\\nfontsize=8);\\nIn [47]: fig, ax = subplots(figsize=(8, 8))\\nax = plot_gam(gam_logit_, 1)\\nax.set_xlabel(\\'Year\\')\\nax.set_ylabel(\\'Effect on wage\\')\\nax.set_title(\\'Partial dependence of high earner status on year\\',\\nfontsize=20);\\nIn [48]: fig, ax = subplots(figsize=(8, 8))\\nax = plot_gam(gam_logit_, 0)\\nax.set_xlabel(\\'Age\\')\\nax.set_ylabel(\\'Effect on wage\\')\\nax.set_title(\\'Partial dependence of high earner status on age\\',\\nfontsize=20);\\n7.8.4\\nLocal Regression\\nWe illustrate the use of local regression using the lowess() function from lowess()\\nsm.nonparametric. Some implementations of GAMs allow terms to be local\\nregression operators; this is not the case in pygam.\\nHere we fit local linear regression models using spans of 0.2 and 0.5;\\nthat is, each neighborhood consists of 20% or 50% of the observations. As\\nexpected, using a span of 0.5 is smoother than 0.2.\\nIn [49]: lowess = sm.nonparametric.lowess\\nfig, ax = subplots(figsize=(8,8))\\nax.scatter(age, y, facecolor=\\'gray\\', alpha=0.5)\\nfor span in [0.2, 0.5]:\\nfitted = lowess(y, 7.9 Exercises\\n325\\nage,\\nfrac=span,\\nxvals=age_grid)\\nax.plot(age_grid,\\nfitted,\\nlabel=\\'{:.1f}\\'.format(span),\\nlinewidth=4)\\nax.set_xlabel(\\'Age\\', fontsize=20)\\nax.set_ylabel(\\'Wage\\', fontsize=20);\\nax.legend(title=\\'span\\', fontsize=15);\\n7.9\\nExercises\\nConceptual\\n1. It was mentioned in this chapter that a cubic regression spline with\\none knot at ξ can be obtained using a basis of the form x, x2, x3,\\n(x −ξ)3\\n+, where (x −ξ)3\\n+ = (x −ξ)3 if x > ξ and equals 0 otherwise.\\nWe will now show that a function of the form\\nf(x) = β0 + β1x + β2x2 + β3x3 + β4(x −ξ)3\\n+\\nis indeed a cubic regression spline, regardless of the values of β0, β1, β2,\\nβ3, β4.\\n(a) Find a cubic polynomial\\nf1(x) = a1 + b1x + c1x2 + d1x3\\nsuch that f(x) = f1(x) for all x ≤ξ. Express a1, b1, c1, d1 in\\nterms of β0, β1, β2, β3, β4.\\n(b) Find a cubic polynomial\\nf2(x) = a2 + b2x + c2x2 + d2x3\\nsuch that f(x) = f2(x) for all x > ξ. Express a2, b2, c2, d2 in\\nterms of β0, β1, β2, β3, β4. We have now established that f(x) is\\na piecewise polynomial.\\n(c) Show that f1(ξ) = f2(ξ). That is, f(x) is continuous at ξ.\\n(d) Show that f ′\\n1(ξ) = f ′\\n2(ξ). That is, f ′(x) is continuous at ξ.\\n(e) Show that f ′′\\n1 (ξ) = f ′′\\n2 (ξ). That is, f ′′(x) is continuous at ξ.\\nTherefore, f(x) is indeed a cubic spline.\\nHint: Parts (d) and (e) of this problem require knowledge of single-\\nvariable calculus. As a reminder, given a cubic polynomial\\nf1(x) = a1 + b1x + c1x2 + d1x3,\\nthe first derivative takes the form\\nf ′\\n1(x) = b1 + 2c1x + 3d1x2 326\\n7. Moving Beyond Linearity\\nand the second derivative takes the form\\nf ′′\\n1 (x) = 2c1 + 6d1x.\\n2. Suppose that a curve ˆg is computed to smoothly fit a set of n points\\nusing the following formula:\\nˆg = arg min\\ng\\n> n\\n0\\ni=1\\n(yi −g(xi))2 + λ\\nL 7\\ng(m)(x)\\n82\\ndx\\n?\\n,\\nwhere g(m) represents the mth derivative of g (and g(0) = g). Provide\\nexample sketches of ˆg in each of the following scenarios.\\n(a) λ = ∞, m = 0.\\n(b) λ = ∞, m = 1.\\n(c) λ = ∞, m = 2.\\n(d) λ = ∞, m = 3.\\n(e) λ = 0, m = 3.\\n3. Suppose we fit a curve with basis functions b1(X) = X, b2(X) =\\n(X −1)2I(X ≥1). (Note that I(X ≥1) equals 1 for X ≥1 and 0\\notherwise.) We fit the linear regression model\\nY = β0 + β1b1(X) + β2b2(X) + ϵ,\\nand obtain coefficient estimates ˆβ0 = 1, ˆβ1 = 1, ˆβ2 = −2. Sketch the\\nestimated curve between X = −2 and X = 2. Note the intercepts,\\nslopes, and other relevant information.\\n4. Suppose we fit a curve with basis functions b1(X) = I(0 ≤X ≤2) −\\n(X −1)I(1 ≤X ≤2), b2(X) = (X −3)I(3 ≤X ≤4)+I(4 < X ≤5).\\nWe fit the linear regression model\\nY = β0 + β1b1(X) + β2b2(X) + ϵ,\\nand obtain coefficient estimates ˆβ0 = 1, ˆβ1 = 1, ˆβ2 = 3. Sketch the\\nestimated curve between X = −2 and X = 6. Note the intercepts,\\nslopes, and other relevant information.\\n5. Consider two curves, ˆg1 and ˆg2, defined by\\nˆg1 = arg min\\ng\\n> n\\n0\\ni=1\\n(yi −g(xi))2 + λ\\nL 7\\ng(3)(x)\\n82\\ndx\\n?\\n,\\nˆg2 = arg min\\ng\\n> n\\n0\\ni=1\\n(yi −g(xi))2 + λ\\nL 7\\ng(4)(x)\\n82\\ndx\\n?\\n,\\nwhere g(m) represents the mth derivative of g.\\n(a) As λ →∞, will ˆg1 or ˆg2 have the smaller training RSS?\\n(b) As λ →∞, will ˆg1 or ˆg2 have the smaller test RSS?\\n(c) For λ = 0, will ˆg1 or ˆg2 have the smaller training and test RSS? 7.9 Exercises\\n327\\nApplied\\n6. In this exercise, you will further analyze the Wage data set considered\\nthroughout this chapter.\\n(a) Perform polynomial regression to predict wage using age. Use\\ncross-validation to select the optimal degree d for the polyno-\\nmial. What degree was chosen, and how does this compare to\\nthe results of hypothesis testing using ANOVA? Make a plot of\\nthe resulting polynomial fit to the data.\\n(b) Fit a step function to predict wage using age, and perform cross-\\nvalidation to choose the optimal number of cuts. Make a plot of\\nthe fit obtained.\\n7. The Wage data set contains a number of other features not explored\\nin this chapter, such as marital status (maritl), job class (jobclass),\\nand others. Explore the relationships between some of these other\\npredictors and wage, and use non-linear fitting techniques in order to\\nfit flexible models to the data. Create plots of the results obtained,\\nand write a summary of your findings.\\n8. Fit some of the non-linear models investigated in this chapter to the\\nAuto data set. Is there evidence for non-linear relationships in this\\ndata set? Create some informative plots to justify your answer.\\n9. This question uses the variables dis (the weighted mean of distances\\nto five Boston employment centers) and nox (nitrogen oxides concen-\\ntration in parts per 10 million) from the Boston data. We will treat\\ndis as the predictor and nox as the response.\\n(a) Use the poly() function from the ISLP.models module to fit a\\ncubic polynomial regression to predict nox using dis. Report the\\nregression output, and plot the resulting data and polynomial\\nfits.\\n(b) Plot the polynomial fits for a range of different polynomial\\ndegrees (say, from 1 to 10), and report the associated residual\\nsum of squares.\\n(c) Perform cross-validation or another approach to select the opti-\\nmal degree for the polynomial, and explain your results.\\n(d) Use the bs() function from the ISLP.models module to fit a re-\\ngression spline to predict nox using dis. Report the output for\\nthe fit using four degrees of freedom. How did you choose the\\nknots? Plot the resulting fit.\\n(e) Now fit a regression spline for a range of degrees of freedom, and\\nplot the resulting fits and report the resulting RSS. Describe the\\nresults obtained.\\n(f) Perform cross-validation or another approach in order to select\\nthe best degrees of freedom for a regression spline on this data.\\nDescribe your results. 328\\n7. Moving Beyond Linearity\\n10. This question relates to the College data set.\\n(a) Split the data into a training set and a test set. Using out-of-state\\ntuition as the response and the other variables as the predictors,\\nperform forward stepwise selection on the training set in order\\nto identify a satisfactory model that uses just a subset of the\\npredictors.\\n(b) Fit a GAM on the training data, using out-of-state tuition as\\nthe response and the features selected in the previous step as\\nthe predictors. Plot the results, and explain your findings.\\n(c) Evaluate the model obtained on the test set, and explain the\\nresults obtained.\\n(d) For which variables, if any, is there evidence of a non-linear\\nrelationship with the response?\\n11. In Section 7.7, it was mentioned that GAMs are generally fit using\\na backfitting approach. The idea behind backfitting is actually quite\\nsimple. We will now explore backfitting in the context of multiple\\nlinear regression.\\nSuppose that we would like to perform multiple linear regression, but\\nwe do not have software to do so. Instead, we only have software\\nto perform simple linear regression. Therefore, we take the following\\niterative approach: we repeatedly hold all but one coefficient esti-\\nmate fixed at its current value, and update only that coefficient\\nestimate using a simple linear regression. The process is continued un-\\ntil convergence—that is, until the coefficient estimates stop changing.\\nWe now try this out on a toy example.\\n(a) Generate a response Y and two predictors X1 and X2, with\\nn = 100.\\n(b) Write a function simple_reg() that takes two arguments outcome\\nand feature, fits a simple linear regression model with this out-\\ncome and feature, and returns the estimated intercept and slope.\\n(c) Initialize beta1 to take on a value of your choice. It does not\\nmatter what value you choose.\\n(d) Keeping beta1 fixed, use your function simple_reg() to fit the\\nmodel:\\nY −beta1 · X1 = β0 + β2X2 + ϵ.\\nStore the resulting values as beta0 and beta2.\\n(e) Keeping beta2 fixed, fit the model\\nY −beta2 · X2 = β0 + β1X1 + ϵ.\\nStore the result as beta0 and beta1 (overwriting their previous\\nvalues).\\n(f) Write a for loop to repeat (c) and (d) 1,000 times. Report the\\nestimates of beta0, beta1, and beta2 at each iteration of the for\\nloop. Create a plot in which each of these values is displayed,\\nwith beta0, beta1, and beta2. 7.9 Exercises\\n329\\n(g) Compare your answer in (e) to the results of simply perform-\\ning multiple linear regression to predict Y using X1 and X2.\\nUse axline() method to overlay those multiple linear regression\\ncoefficient estimates on the plot obtained in (e).\\n(h) On this data set, how many backfitting iterations were required\\nin order to obtain a “good” approximation to the multiple re-\\ngression coefficient estimates?\\n12. This problem is a continuation of the previous exercise. In a toy\\nexample with p = 100, show that one can approximate the multiple\\nlinear regression coefficient estimates by repeatedly performing simple\\nlinear regression in a backfitting procedure. How many backfitting\\niterations are required in order to obtain a “good” approximation to\\nthe multiple regression coefficient estimates? Create a plot to justify\\nyour answer. 8\\nTree-Based Methods\\nIn this chapter, we describe tree-based methods for regression and classifi-\\ncation. These involve stratifying or segmenting the predictor space into a\\nnumber of simple regions. In order to make a prediction for a given ob-\\nservation, we typically use the mean or the mode response value for the\\ntraining observations in the region to which it belongs. Since the set of\\nsplitting rules used to segment the predictor space can be summarized in\\na tree, these types of approaches are known as decision tree methods.\\ndecision tree\\nTree-based methods are simple and useful for interpretation. However,\\nthey typically are not competitive with the best supervised learning ap-\\nproaches, such as those seen in Chapters 6 and 7, in terms of prediction\\naccuracy. Hence in this chapter we also introduce bagging, random forests,\\nboosting, and Bayesian additive regression trees. Each of these approaches\\ninvolves producing multiple trees which are then combined to yield a single\\nconsensus prediction. We will see that combining a large number of trees\\ncan often result in dramatic improvements in prediction accuracy, at the\\nexpense of some loss in interpretation.\\n8.1\\nThe Basics of Decision Trees\\nDecision trees can be applied to both regression and classification problems.\\nWe first consider regression problems, and then move on to classification.\\n8.1.1\\nRegression Trees\\nIn order to motivate regression trees, we begin with a simple example.\\nregression\\ntree\\n© Springer Nature Switzerland AG 2023 \\nG. James et al., An Introduction to Statistical Learning, Springer Texts in Statistics, \\nhttps://doi.org/10.1007/978-3-031-38747-0_8 \\n331 332\\n8. Tree-Based Methods\\n|\\nYears < 4.5\\nHits < 117.5\\n5.11\\n6.00\\n6.74\\nFIGURE 8.1. For the Hitters data, a regression tree for predicting the log\\nsalary of a baseball player, based on the number of years that he has played in\\nthe major leagues and the number of hits that he made in the previous year. At a\\ngiven internal node, the label (of the form Xj < tk) indicates the left-hand branch\\nemanating from that split, and the right-hand branch corresponds to Xj ≥tk.\\nFor instance, the split at the top of the tree results in two large branches. The\\nleft-hand branch corresponds to Years<4.5, and the right-hand branch corresponds\\nto Years>=4.5. The tree has two internal nodes and three terminal nodes, or\\nleaves. The number in each leaf is the mean of the response for the observations\\nthat fall there.\\nPredicting Baseball Players’ Salaries Using Regression Trees\\nWe use the Hitters data set to predict a baseball player’s Salary based on\\nYears (the number of years that he has played in the major leagues) and\\nHits (the number of hits that he made in the previous year). We first remove\\nobservations that are missing Salary values, and log-transform Salary so\\nthat its distribution has more of a typical bell-shape. (Recall that Salary\\nis measured in thousands of dollars.)\\nFigure 8.1 shows a regression tree fit to this data. It consists of a series\\nof splitting rules, starting at the top of the tree. The top split assigns\\nobservations having Years<4.5 to the left branch.1 The predicted salary\\nfor these players is given by the mean response value for the players in\\nthe data set with Years<4.5. For such players, the mean log salary is 5.107,\\nand so we make a prediction of e5.107 thousands of dollars, i.e. $165,174, for\\nthese players. Players with Years>=4.5 are assigned to the right branch, and\\nthen that group is further subdivided by Hits. Overall, the tree stratifies\\nor segments the players into three regions of predictor space: players who\\nhave played for four or fewer years, players who have played for five or more\\nyears and who made fewer than 118 hits last year, and players who have\\nplayed for five or more years and who made at least 118 hits last year. These\\nthree regions can be written as R1 ={X | Years<4.5}, R2 ={X | Years>=4.5,\\nHits<117.5}, and R3 ={X | Years>=4.5, Hits>=117.5}. Figure 8.2 illustrates\\n1Both Years and Hits are integers in these data; the function used to fit this tree\\nlabels the splits at the midpoint between two adjacent values. 8.1 The Basics of Decision Trees\\n333\\nYears\\nHits\\n1\\n117.5\\n238\\n1\\n4.5\\n24\\nR1\\nR3\\nR2\\nFIGURE 8.2. The three-region partition for the Hitters data set from the\\nregression tree illustrated in Figure 8.1.\\nthe regions as a function of Years and Hits. The predicted salaries for these\\nthree groups are $1,000×e5.107 =$165,174, $1,000×e5.999 =$402,834, and\\n$1,000×e6.740 =$845,346 respectively.\\nIn keeping with the tree analogy, the regions R1, R2, and R3 are known as\\nterminal nodes or leaves of the tree. As is the case for Figure 8.1, decision terminal\\nnode\\nleaf\\ntrees are typically drawn upside down, in the sense that the leaves are at\\nthe bottom of the tree. The points along the tree where the predictor space\\nis split are referred to as internal nodes.\\nIn Figure 8.1, the two internal internal\\nnode\\nnodes are indicated by the text Years<4.5 and Hits<117.5. We refer to the\\nsegments of the trees that connect the nodes as branches.\\nbranch\\nWe might interpret the regression tree displayed in Figure 8.1 as follows:\\nYears is the most important factor in determining Salary, and players with\\nless experience earn lower salaries than more experienced players. Given\\nthat a player is less experienced, the number of hits that he made in the\\nprevious year seems to play little role in his salary. But among players who\\nhave been in the major leagues for five or more years, the number of hits\\nmade in the previous year does affect salary, and players who made more\\nhits last year tend to have higher salaries. The regression tree shown in\\nFigure 8.1 is likely an over-simplification of the true relationship between\\nHits, Years, and Salary. However, it has advantages over other types of\\nregression models (such as those seen in Chapters 3 and 6): it is easier to\\ninterpret, and has a nice graphical representation.\\nPrediction via Stratification of the Feature Space\\nWe now discuss the process of building a regression tree. Roughly speaking,\\nthere are two steps.\\n1. We divide the predictor space — that is, the set of possible values\\nfor X1, X2, . . . , Xp — into J distinct and non-overlapping regions,\\nR1, R2, . . . , RJ. 334\\n8. Tree-Based Methods\\n2. For every observation that falls into the region Rj, we make the same\\nprediction, which is simply the mean of the response values for the\\ntraining observations in Rj.\\nFor instance, suppose that in Step 1 we obtain two regions, R1 and R2,\\nand that the response mean of the training observations in the first region\\nis 10, while the response mean of the training observations in the second\\nregion is 20. Then for a given observation X = x, if x ∈R1 we will predict\\na value of 10, and if x ∈R2 we will predict a value of 20.\\nWe now elaborate on Step 1 above. How do we construct the regions\\nR1, . . . , RJ? In theory, the regions could have any shape. However, we\\nchoose to divide the predictor space into high-dimensional rectangles, or\\nboxes, for simplicity and for ease of interpretation of the resulting predic-\\ntive model. The goal is to find boxes R1, . . . , RJ that minimize the RSS,\\ngiven by\\nJ\\n0\\nj=1\\n0\\ni∈Rj\\n(yi −ˆyRj)2,\\n(8.1)\\nwhere ˆyRj is the mean response for the training observations within the\\njth box. Unfortunately, it is computationally infeasible to consider every\\npossible partition of the feature space into J boxes. For this reason, we take\\na top-down, greedy approach that is known as recursive binary splitting. The recursive\\nbinary\\nsplitting\\napproach is top-down because it begins at the top of the tree (at which point\\nall observations belong to a single region) and then successively splits the\\npredictor space; each split is indicated via two new branches further down\\non the tree. It is greedy because at each step of the tree-building process,\\nthe best split is made at that particular step, rather than looking ahead\\nand picking a split that will lead to a better tree in some future step.\\nIn order to perform recursive binary splitting, we first select the pre-\\ndictor Xj and the cutpoint s such that splitting the predictor space into\\nthe regions {X|Xj < s} and {X|Xj ≥s} leads to the greatest possible\\nreduction in RSS. (The notation {X|Xj < s} means the region of predictor\\nspace in which Xj takes on a value less than s.) That is, we consider all\\npredictors X1, . . . , Xp, and all possible values of the cutpoint s for each of\\nthe predictors, and then choose the predictor and cutpoint such that the\\nresulting tree has the lowest RSS. In greater detail, for any j and s, we\\ndefine the pair of half-planes\\nR1(j, s) = {X|Xj < s} and R2(j, s) = {X|Xj ≥s},\\n(8.2)\\nand we seek the value of j and s that minimize the equation\\n0\\ni: xi∈R1(j,s)\\n(yi −ˆyR1)2 +\\n0\\ni: xi∈R2(j,s)\\n(yi −ˆyR2)2,\\n(8.3)\\nwhere ˆyR1 is the mean response for the training observations in R1(j, s),\\nand ˆyR2 is the mean response for the training observations in R2(j, s).\\nFinding the values of j and s that minimize (8.3) can be done quite quickly,\\nespecially when the number of features p is not too large.\\nNext, we repeat the process, looking for the best predictor and best\\ncutpoint in order to split the data further so as to minimize the RSS within 8.1 The Basics of Decision Trees\\n335\\n|\\nt1\\nt2\\nt3\\nt4\\nR1\\nR1\\nR2\\nR2\\nR3\\nR3\\nR4\\nR4\\nR5\\nR5\\nX1\\nX1\\nX1\\nX2\\nX2\\nX2\\nX1 ≤t1\\nX2 ≤t2\\nX1 ≤t3\\nX2 ≤t4\\nFIGURE 8.3. Top Left: A partition of two-dimensional feature space that could\\nnot result from recursive binary splitting. Top Right: The output of recursive\\nbinary splitting on a two-dimensional example. Bottom Left: A tree corresponding\\nto the partition in the top right panel. Bottom Right: A perspective plot of the\\nprediction surface corresponding to that tree.\\neach of the resulting regions. However, this time, instead of splitting the\\nentire predictor space, we split one of the two previously identified regions.\\nWe now have three regions. Again, we look to split one of these three regions\\nfurther, so as to minimize the RSS. The process continues until a stopping\\ncriterion is reached; for instance, we may continue until no region contains\\nmore than five observations.\\nOnce the regions R1, . . . , RJ have been created, we predict the response\\nfor a given test observation using the mean of the training observations in\\nthe region to which that test observation belongs.\\nA five-region example of this approach is shown in Figure 8.3.\\nTree Pruning\\nThe process described above may produce good predictions on the training\\nset, but is likely to overfit the data, leading to poor test set performance.\\nThis is because the resulting tree might be too complex. A smaller tree 336\\n8. Tree-Based Methods\\nwith fewer splits (that is, fewer regions R1, . . . , RJ) might lead to lower\\nvariance and better interpretation at the cost of a little bias. One possible\\nalternative to the process described above is to build the tree only so long\\nas the decrease in the RSS due to each split exceeds some (high) threshold.\\nThis strategy will result in smaller trees, but is too short-sighted since a\\nseemingly worthless split early on in the tree might be followed by a very\\ngood split—that is, a split that leads to a large reduction in RSS later on.\\nTherefore, a better strategy is to grow a very large tree T0, and then\\nprune it back in order to obtain a subtree. How do we determine the best prune\\nsubtree\\nway to prune the tree? Intuitively, our goal is to select a subtree that\\nleads to the lowest test error rate. Given a subtree, we can estimate its\\ntest error using cross-validation or the validation set approach. However,\\nestimating the cross-validation error for every possible subtree would be too\\ncumbersome, since there is an extremely large number of possible subtrees.\\nInstead, we need a way to select a small set of subtrees for consideration.\\nCost complexity pruning—also known as weakest link pruning—gives us cost\\ncomplexity\\npruning\\nweakest link\\npruning\\na way to do just this. Rather than considering every possible subtree, we\\nconsider a sequence of trees indexed by a nonnegative tuning parameter α.\\nFor each value of α there corresponds a subtree T ⊂T0 such that\\n|T |\\n0\\nm=1\\n0\\ni: xi∈Rm\\n(yi −ˆyRm)2 + α|T|\\n(8.4)\\nis as small as possible. Here |T| indicates the number of terminal nodes\\nof the tree T, Rm is the rectangle (i.e. the subset of predictor space) cor-\\nresponding to the mth terminal node, and ˆyRm is the predicted response\\nassociated with Rm—that is, the mean of the training observations in Rm.\\nThe tuning parameter α controls a trade-off between the subtree’s com-\\nplexity and its fit to the training data. When α = 0, then the subtree T\\nwill simply equal T0, because then (8.4) just measures the training error.\\nHowever, as α increases, there is a price to pay for having a tree with\\nmany terminal nodes, and so the quantity (8.4) will tend to be minimized\\nfor a smaller subtree. Equation 8.4 is reminiscent of the lasso (6.7) from\\nChapter 6, in which a similar formulation was used in order to control the\\ncomplexity of a linear model.\\nIt turns out that as we increase α from zero in (8.4), branches get pruned\\nfrom the tree in a nested and predictable fashion, so obtaining the whole\\nsequence of subtrees as a function of α is easy. We can select a value of\\nα using a validation set or using cross-validation. We then return to the\\nfull data set and obtain the subtree corresponding to α. This process is\\nsummarized in Algorithm 8.1.\\nFigures 8.4 and 8.5 display the results of fitting and pruning a regression\\ntree on the Hitters data, using nine of the features. First, we randomly\\ndivided the data set in half, yielding 132 observations in the training set\\nand 131 observations in the test set. We then built a large regression tree\\non the training data and varied α in (8.4) in order to create subtrees with\\ndifferent numbers of terminal nodes. Finally, we performed six-fold cross-\\nvalidation in order to estimate the cross-validated MSE of the trees as 8.1 The Basics of Decision Trees\\n337\\nAlgorithm 8.1 Building a Regression Tree\\n1. Use recursive binary splitting to grow a large tree on the training\\ndata, stopping only when each terminal node has fewer than some\\nminimum number of observations.\\n2. Apply cost complexity pruning to the large tree in order to obtain a\\nsequence of best subtrees, as a function of α.\\n3. Use K-fold cross-validation to choose α. That is, divide the training\\nobservations into K folds. For each k = 1, . . . , K:\\n(a) Repeat Steps 1 and 2 on all but the kth fold of the training data.\\n(b) Evaluate the mean squared prediction error on the data in the\\nleft-out kth fold, as a function of α.\\nAverage the results for each value of α, and pick α to minimize the\\naverage error.\\n4. Return the subtree from Step 2 that corresponds to the chosen value\\nof α.\\na function of α. (We chose to perform six-fold cross-validation because\\n132 is an exact multiple of six.) The unpruned regression tree is shown\\nin Figure 8.4. The green curve in Figure 8.5 shows the CV error as a\\nfunction of the number of leaves,2 while the orange curve indicates the\\ntest error. Also shown are standard error bars around the estimated errors.\\nFor reference, the training error curve is shown in black. The CV error\\nis a reasonable approximation of the test error: the CV error takes on its\\nminimum for a three-node tree, while the test error also dips down at the\\nthree-node tree (though it takes on its lowest value at the ten-node tree).\\nThe pruned tree containing three terminal nodes is shown in Figure 8.1.\\n8.1.2\\nClassification Trees\\nA classification tree is very similar to a regression tree, except that it is classification\\ntree\\nused to predict a qualitative response rather than a quantitative one. Re-\\ncall that for a regression tree, the predicted response for an observation is\\ngiven by the mean response of the training observations that belong to the\\nsame terminal node. In contrast, for a classification tree, we predict that\\neach observation belongs to the most commonly occurring class of training\\nobservations in the region to which it belongs. In interpreting the results of\\na classification tree, we are often interested not only in the class prediction\\ncorresponding to a particular terminal node region, but also in the class\\nproportions among the training observations that fall into that region.\\nThe task of growing a classification tree is quite similar to the task of\\ngrowing a regression tree. Just as in the regression setting, we use recursive\\n2Although CV error is computed as a function of α, it is convenient to display the\\nresult as a function of |T|, the number of leaves; this is based on the relationship between\\nα and |T| in the original tree grown to all the training data. 338\\n8. Tree-Based Methods\\n|\\nYears < 4.5\\nRBI < 60.5\\nPutouts < 82\\nYears < 3.5\\nYears < 3.5\\nHits < 117.5\\nWalks < 43.5\\nRuns < 47.5\\nWalks < 52.5\\nRBI < 80.5\\nYears < 6.5\\n5.487\\n4.622\\n5.183\\n5.394\\n6.189\\n6.015\\n5.571\\n6.407\\n6.549\\n6.459\\n7.007\\n7.289\\nFIGURE 8.4. Regression tree analysis for the Hitters data. The unpruned tree\\nthat results from top-down greedy splitting on the training data is shown.\\nbinary splitting to grow a classification tree. However, in the classification\\nsetting, RSS cannot be used as a criterion for making the binary splits.\\nA natural alternative to RSS is the classification error rate. Since we plan classification\\nerror rate\\nto assign an observation in a given region to the most commonly occurring\\nclass of training observations in that region, the classification error rate is\\nsimply the fraction of the training observations in that region that do not\\nbelong to the most common class:\\nE = 1 −max\\nk (ˆpmk).\\n(8.5)\\nHere ˆpmk represents the proportion of training observations in the mth\\nregion that are from the kth class. However, it turns out that classification\\nerror is not sufficiently sensitive for tree-growing, and in practice two other\\nmeasures are preferable.\\nThe Gini index is defined by\\nGini index\\nG =\\nK\\n0\\nk=1\\nˆpmk(1 −ˆpmk),\\n(8.6)\\na measure of total variance across the K classes. It is not hard to see\\nthat the Gini index takes on a small value if all of the ˆpmk’s are close to\\nzero or one. For this reason the Gini index is referred to as a measure of 8.1 The Basics of Decision Trees\\n339\\n2\\n4\\n6\\n8\\n10\\n0.0\\n0.2\\n0.4\\n0.6\\n0.8\\n1.0\\nTree Size\\nMean Squared Error\\nTraining\\nCross−Validation\\nTest\\nFIGURE 8.5.\\nRegression tree analysis for the Hitters data. The training,\\ncross-validation, and test MSE are shown as a function of the number of terminal\\nnodes in the pruned tree. Standard error bands are displayed. The minimum\\ncross-validation error occurs at a tree size of three.\\nnode purity—a small value indicates that a node contains predominantly\\nobservations from a single class.\\nAn alternative to the Gini index is entropy, given by\\nentropy\\nD = −\\nK\\n0\\nk=1\\nˆpmk log ˆpmk.\\n(8.7)\\nSince 0 ≤ˆpmk ≤1, it follows that 0 ≤−ˆpmk log ˆpmk. One can show that\\nthe entropy will take on a value near zero if the ˆpmk’s are all near zero or\\nnear one. Therefore, like the Gini index, the entropy will take on a small\\nvalue if the mth node is pure. In fact, it turns out that the Gini index and\\nthe entropy are quite similar numerically.\\nWhen building a classification tree, either the Gini index or the entropy\\nare typically used to evaluate the quality of a particular split, since these\\ntwo approaches are more sensitive to node purity than is the classification\\nerror rate. Any of these three approaches might be used when pruning the\\ntree, but the classification error rate is preferable if prediction accuracy of\\nthe final pruned tree is the goal.\\nFigure 8.6 shows an example on the Heart data set. These data con-\\ntain a binary outcome HD for 303 patients who presented with chest pain.\\nAn outcome value of Yes indicates the presence of heart disease based on\\nan angiographic test, while No means no heart disease. There are 13 predic-\\ntors including Age, Sex, Chol (a cholesterol measurement), and other heart\\nand lung function measurements. Cross-validation results in a tree with six\\nterminal nodes.\\nIn our discussion thus far, we have assumed that the predictor vari-\\nables take on continuous values. However, decision trees can be constructed\\neven in the presence of qualitative predictor variables. For instance, in the\\nHeart data, some of the predictors, such as Sex, Thal (Thallium stress test), 340\\n8. Tree-Based Methods\\n|\\nThal:a\\nCa < 0.5\\nMaxHR < 161.5\\nRestBP < 157\\nChol < 244\\nMaxHR < 156\\nMaxHR < 145.5\\nChestPain:bc\\nChol < 244\\nSex < 0.5\\nCa < 0.5\\nSlope < 1.5\\nAge < 52\\nThal:b\\nChestPain:a\\nOldpeak < 1.1\\nRestECG < 1\\nNo\\nYes\\nNo\\nNo\\nYes\\nNo\\nNo\\nNo\\nNo\\nYes\\nYes\\nNo\\nNo\\nNo\\nYes\\nYes\\nYes\\nYes\\n5\\n10\\n15\\n0.0\\n0.1\\n0.2\\n0.3\\n0.4\\n0.5\\n0.6\\nTree Size\\nError\\nTraining\\nCross−Validation\\nTest\\n|\\nThal:a\\nCa < 0.5\\nMaxHR < 161.5\\nChestPain:bc\\nCa < 0.5\\nNo\\nNo\\nNo\\nYes\\nYes\\nYes\\nFIGURE 8.6. Heart data. Top: The unpruned tree. Bottom Left: Cross-valida-\\ntion error, training, and test error, for different sizes of the pruned tree. Bottom\\nRight: The pruned tree corresponding to the minimal cross-validation error.\\nand ChestPain, are qualitative. Therefore, a split on one of these variables\\namounts to assigning some of the qualitative values to one branch and\\nassigning the remaining to the other branch. In Figure 8.6, some of the in-\\nternal nodes correspond to splitting qualitative variables. For instance, the\\ntop internal node corresponds to splitting Thal. The text Thal:a indicates\\nthat the left-hand branch coming out of that node consists of observations\\nwith the first value of the Thal variable (normal), and the right-hand node\\nconsists of the remaining observations (fixed or reversible defects). The text\\nChestPain:bc two splits down the tree on the left indicates that the left-hand\\nbranch coming out of that node consists of observations with the second\\nand third values of the ChestPain variable, where the possible values are\\ntypical angina, atypical angina, non-anginal pain, and asymptomatic.\\nFigure 8.6 has a surprising characteristic: some of the splits yield two\\nterminal nodes that have the same predicted value. For instance, consider\\nthe split RestECG<1 near the bottom right of the unpruned tree. Regardless\\nof the value of RestECG, a response value of Yes is predicted for those ob- 8.1 The Basics of Decision Trees\\n341\\nservations. Why, then, is the split performed at all? The split is performed\\nbecause it leads to increased node purity. That is, all 9 of the observations\\ncorresponding to the right-hand leaf have a response value of Yes, whereas\\n7/11 of those corresponding to the left-hand leaf have a response value of\\nYes. Why is node purity important? Suppose that we have a test obser-\\nvation that belongs to the region given by that right-hand leaf. Then we\\ncan be pretty certain that its response value is Yes. In contrast, if a test\\nobservation belongs to the region given by the left-hand leaf, then its re-\\nsponse value is probably Yes, but we are much less certain. Even though\\nthe split RestECG<1 does not reduce the classification error, it improves the\\nGini index and the entropy, which are more sensitive to node purity.\\n8.1.3\\nTrees Versus Linear Models\\nRegression and classification trees have a very different flavor from the more\\nclassical approaches for regression and classification presented in Chapters 3\\nand 4. In particular, linear regression assumes a model of the form\\nf(X) = β0 +\\np\\n0\\nj=1\\nXjβj,\\n(8.8)\\nwhereas regression trees assume a model of the form\\nf(X) =\\nM\\n0\\nm=1\\ncm · 1(X∈Rm)\\n(8.9)\\nwhere R1, . . . , RM represent a partition of feature space, as in Figure 8.3.\\nWhich model is better? It depends on the problem at hand. If the re-\\nlationship between the features and the response is well approximated by\\na linear model as in (8.8), then an approach such as linear regression will\\nlikely work well, and will outperform a method such as a regression tree\\nthat does not exploit this linear structure. If instead there is a highly non-\\nlinear and complex relationship between the features and the response as\\nindicated by model (8.9), then decision trees may outperform classical ap-\\nproaches. An illustrative example is displayed in Figure 8.7. The relative\\nperformances of tree-based and classical approaches can be assessed by es-\\ntimating the test error, using either cross-validation or the validation set\\napproach (Chapter 5).\\nOf course, other considerations beyond simply test error may come into\\nplay in selecting a statistical learning method; for instance, in certain set-\\ntings, prediction using a tree may be preferred for the sake of interpretabil-\\nity and visualization.\\n8.1.4\\nAdvantages and Disadvantages of Trees\\nDecision trees for regression and classification have a number of advantages\\nover the more classical approaches seen in Chapters 3 and 4:\\nL Trees are very easy to explain to people. In fact, they are even easier\\nto explain than linear regression! 342\\n8. Tree-Based Methods\\n−2\\n−1\\n0\\n1\\n2\\n−2\\n−1\\n0\\n1\\n2\\nX1\\nX2\\n−2\\n−1\\n0\\n1\\n2\\n−2\\n−1\\n0\\n1\\n2\\nX1\\nX2\\n−2\\n−1\\n0\\n1\\n2\\n−2\\n−1\\n0\\n1\\n2\\nX1\\nX2\\n−2\\n−1\\n0\\n1\\n2\\n−2\\n−1\\n0\\n1\\n2\\nX1\\nX2\\nFIGURE 8.7. Top Row: A two-dimensional classification example in which the\\ntrue decision boundary is linear, and is indicated by the shaded regions. A classical\\napproach that assumes a linear boundary (left) will outperform a decision tree\\nthat performs splits parallel to the axes (right). Bottom Row: Here the true de-\\ncision boundary is non-linear. Here a linear model is unable to capture the true\\ndecision boundary (left), whereas a decision tree is successful (right).\\nL Some people believe that decision trees more closely mirror human\\ndecision-making than do the regression and classification approaches\\nseen in previous chapters.\\nL Trees can be displayed graphically, and are easily interpreted even by\\na non-expert (especially if they are small).\\nL Trees can easily handle qualitative predictors without the need to\\ncreate dummy variables.\\nM Unfortunately, trees generally do not have the same level of predictive\\naccuracy as some of the other regression and classification approaches\\nseen in this book.\\nM Additionally, trees can be very non-robust. In other words, a small\\nchange in the data can cause a large change in the final estimated\\ntree.\\nHowever, by aggregating many decision trees, using methods like bagging,\\nrandom forests, and boosting, the predictive performance of trees can be\\nsubstantially improved. We introduce these concepts in the next section. 8.2 Bagging, Random Forests, Boosting, and Bayesian Additive Regression Trees\\n343\\n8.2\\nBagging, Random Forests, Boosting, and\\nBayesian Additive Regression Trees\\nAn ensemble method is an approach that combines many simple “building ensemble\\nblock” models in order to obtain a single and potentially very powerful\\nmodel. These simple building block models are sometimes known as weak\\nlearners, since they may lead to mediocre predictions on their own.\\nweak\\nlearners\\nWe will now discuss bagging, random forests, boosting, and Bayesian\\nadditive regression trees. These are ensemble methods for which the simple\\nbuilding block is a regression or a classification tree.\\n8.2.1\\nBagging\\nThe bootstrap, introduced in Chapter 5, is an extremely powerful idea. It is\\nused in many situations in which it is hard or even impossible to directly\\ncompute the standard deviation of a quantity of interest. We see here that\\nthe bootstrap can be used in a completely different context, in order to\\nimprove statistical learning methods such as decision trees.\\nThe decision trees discussed in Section 8.1 suffer from high variance.\\nThis means that if we split the training data into two parts at random,\\nand fit a decision tree to both halves, the results that we get could be\\nquite different. In contrast, a procedure with low variance will yield similar\\nresults if applied repeatedly to distinct data sets; linear regression tends\\nto have low variance, if the ratio of n to p is moderately large. Bootstrap\\naggregation, or bagging, is a general-purpose procedure for reducing the bagging\\nvariance of a statistical learning method; we introduce it here because it is\\nparticularly useful and frequently used in the context of decision trees.\\nRecall that given a set of n independent observations Z1, . . . , Zn, each\\nwith variance σ2, the variance of the mean ¯Z of the observations is given\\nby σ2/n. In other words, averaging a set of observations reduces variance.\\nHence a natural way to reduce the variance and increase the test set ac-\\ncuracy of a statistical learning method is to take many training sets from\\nthe population, build a separate prediction model using each training set,\\nand average the resulting predictions. In other words, we could calculate\\nˆf 1(x), ˆf 2(x), . . . , ˆf B(x) using B separate training sets, and average them\\nin order to obtain a single low-variance statistical learning model, given by\\nˆfavg(x) = 1\\nB\\nB\\n0\\nb=1\\nˆf b(x).\\nOf course, this is not practical because we generally do not have access\\nto multiple training sets. Instead, we can bootstrap, by taking repeated\\nsamples from the (single) training data set. In this approach we generate\\nB different bootstrapped training data sets. We then train our method on\\nthe bth bootstrapped training set in order to get ˆf ∗b(x), and finally average\\nall the predictions, to obtain\\nˆfbag(x) = 1\\nB\\nB\\n0\\nb=1\\nˆf ∗b(x). 344\\n8. Tree-Based Methods\\n0\\n50\\n100\\n150\\n200\\n250\\n300\\n0.10\\n0.15\\n0.20\\n0.25\\n0.30\\nNumber of Trees\\nError\\nTest: Bagging\\nTest: RandomForest\\nOOB: Bagging\\nOOB: RandomForest\\nFIGURE 8.8. Bagging and random forest results for the Heart data. The test\\nerror (black and orange) is shown as a function of B, the number of bootstrapped\\ntraining sets used. Random forests were applied with m = √p. The dashed line\\nindicates the test error resulting from a single classification tree. The green and\\nblue traces show the OOB error, which in this case is — by chance — considerably\\nlower.\\nThis is called bagging.\\nWhile bagging can improve predictions for many regression methods,\\nit is particularly useful for decision trees. To apply bagging to regression\\ntrees, we simply construct B regression trees using B bootstrapped training\\nsets, and average the resulting predictions. These trees are grown deep,\\nand are not pruned. Hence each individual tree has high variance, but\\nlow bias. Averaging these B trees reduces the variance. Bagging has been\\ndemonstrated to give impressive improvements in accuracy by combining\\ntogether hundreds or even thousands of trees into a single procedure.\\nThus far, we have described the bagging procedure in the regression\\ncontext, to predict a quantitative outcome Y . How can bagging be extended\\nto a classification problem where Y is qualitative? In that situation, there\\nare a few possible approaches, but the simplest is as follows. For a given test\\nobservation, we can record the class predicted by each of the B trees, and\\ntake a majority vote: the overall prediction is the most commonly occurring majority\\nvote\\nclass among the B predictions.\\nFigure 8.8 shows the results from bagging trees on the Heart data. The\\ntest error rate is shown as a function of B, the number of trees constructed\\nusing bootstrapped training data sets. We see that the bagging test error\\nrate is slightly lower in this case than the test error rate obtained from a\\nsingle tree. The number of trees B is not a critical parameter with bagging;\\nusing a very large value of B will not lead to overfitting. In practice we 8.2 Bagging, Random Forests, Boosting, and Bayesian Additive Regression Trees\\n345\\nuse a value of B sufficiently large that the error has settled down. Using\\nB = 100 is sufficient to achieve good performance in this example.\\nOut-of-Bag Error Estimation\\nIt turns out that there is a very straightforward way to estimate the test\\nerror of a bagged model, without the need to perform cross-validation or\\nthe validation set approach. Recall that the key to bagging is that trees are\\nrepeatedly fit to bootstrapped subsets of the observations. One can show\\nthat on average, each bagged tree makes use of around two-thirds of the\\nobservations.3 The remaining one-third of the observations not used to fit a\\ngiven bagged tree are referred to as the out-of-bag (OOB) observations. We out-of-bag\\ncan predict the response for the ith observation using each of the trees in\\nwhich that observation was OOB. This will yield around B/3 predictions\\nfor the ith observation. In order to obtain a single prediction for the ith\\nobservation, we can average these predicted responses (if regression is the\\ngoal) or can take a majority vote (if classification is the goal). This leads\\nto a single OOB prediction for the ith observation. An OOB prediction\\ncan be obtained in this way for each of the n observations, from which the\\noverall OOB MSE (for a regression problem) or classification error (for a\\nclassification problem) can be computed. The resulting OOB error is a valid\\nestimate of the test error for the bagged model, since the response for each\\nobservation is predicted using only the trees that were not fit using that\\nobservation. Figure 8.8 displays the OOB error on the Heart data. It can\\nbe shown that with B sufficiently large, OOB error is virtually equivalent\\nto leave-one-out cross-validation error. The OOB approach for estimating\\nthe test error is particularly convenient when performing bagging on large\\ndata sets for which cross-validation would be computationally onerous.\\nVariable Importance Measures\\nAs we have discussed, bagging typically results in improved accuracy over\\nprediction using a single tree. Unfortunately, however, it can be difficult to\\ninterpret the resulting model. Recall that one of the advantages of decision\\ntrees is the attractive and easily interpreted diagram that results, such as\\nthe one displayed in Figure 8.1. However, when we bag a large number of\\ntrees, it is no longer possible to represent the resulting statistical learning\\nprocedure using a single tree, and it is no longer clear which variables\\nare most important to the procedure. Thus, bagging improves prediction\\naccuracy at the expense of interpretability.\\nAlthough the collection of bagged trees is much more difficult to interpret\\nthan a single tree, one can obtain an overall summary of the importance of\\neach predictor using the RSS (for bagging regression trees) or the Gini index\\n(for bagging classification trees). In the case of bagging regression trees, we\\ncan record the total amount that the RSS (8.1) is decreased due to splits\\nover a given predictor, averaged over all B trees. A large value indicates\\nan important predictor. Similarly, in the context of bagging classification\\n3This relates to Exercise 2 of Chapter 5. 346\\n8. Tree-Based Methods\\nThal\\nCa\\nChestPain\\nOldpeak\\nMaxHR\\nRestBP\\nAge\\nChol\\nSlope\\nSex\\nExAng\\nRestECG\\nFbs\\n0\\n20\\n40\\n60\\n80\\n100\\nVariable Importance\\nFIGURE 8.9. A variable importance plot for the Heart data. Variable impor-\\ntance is computed using the mean decrease in Gini index, and expressed relative\\nto the maximum.\\ntrees, we can add up the total amount that the Gini index (8.6) is decreased\\nby splits over a given predictor, averaged over all B trees.\\nA graphical representation of the variable importances in the Heart data variable\\nimportance\\nis shown in Figure 8.9. We see the mean decrease in Gini index for each vari-\\nable, relative to the largest. The variables with the largest mean decrease\\nin Gini index are Thal, Ca, and ChestPain.\\n8.2.2\\nRandom Forests\\nRandom forests provide an improvement over bagged trees by way of a random\\nforest\\nsmall tweak that decorrelates the trees. As in bagging, we build a number\\nof decision trees on bootstrapped training samples. But when building these\\ndecision trees, each time a split in a tree is considered, a random sample of\\nm predictors is chosen as split candidates from the full set of p predictors.\\nThe split is allowed to use only one of those m predictors. A fresh sample of\\nm predictors is taken at each split, and typically we choose m ≈√p—that\\nis, the number of predictors considered at each split is approximately equal\\nto the square root of the total number of predictors (4 out of the 13 for the\\nHeart data).\\nIn other words, in building a random forest, at each split in the tree,\\nthe algorithm is not even allowed to consider a majority of the available\\npredictors. This may sound crazy, but it has a clever rationale. Suppose\\nthat there is one very strong predictor in the data set, along with a num-\\nber of other moderately strong predictors. Then in the collection of bagged\\ntrees, most or all of the trees will use this strong predictor in the top split.\\nConsequently, all of the bagged trees will look quite similar to each other. 8.2 Bagging, Random Forests, Boosting, and Bayesian Additive Regression Trees\\n347\\nHence the predictions from the bagged trees will be highly correlated. Un-\\nfortunately, averaging many highly correlated quantities does not lead to\\nas large of a reduction in variance as averaging many uncorrelated quan-\\ntities. In particular, this means that bagging will not lead to a substantial\\nreduction in variance over a single tree in this setting.\\nRandom forests overcome this problem by forcing each split to consider\\nonly a subset of the predictors. Therefore, on average (p −m)/p of the\\nsplits will not even consider the strong predictor, and so other predictors\\nwill have more of a chance. We can think of this process as decorrelating\\nthe trees, thereby making the average of the resulting trees less variable\\nand hence more reliable.\\nThe main difference between bagging and random forests is the choice\\nof predictor subset size m. For instance, if a random forest is built using\\nm = p, then this amounts simply to bagging. On the Heart data, random\\nforests using m = √p leads to a reduction in both test error and OOB error\\nover bagging (Figure 8.8).\\nUsing a small value of m in building a random forest will typically be\\nhelpful when we have a large number of correlated predictors. We applied\\nrandom forests to a high-dimensional biological data set consisting of ex-\\npression measurements of 4,718 genes measured on tissue samples from 349\\npatients. There are around 20,000 genes in humans, and individual genes\\nhave different levels of activity, or expression, in particular cells, tissues,\\nand biological conditions. In this data set, each of the patient samples has\\na qualitative label with 15 different levels: either normal or 1 of 14 different\\ntypes of cancer. Our goal was to use random forests to predict cancer type\\nbased on the 500 genes that have the largest variance in the training set.\\nWe randomly divided the observations into a training and a test set, and\\napplied random forests to the training set for three different values of the\\nnumber of splitting variables m. The results are shown in Figure 8.10. The\\nerror rate of a single tree is 45.7 %, and the null rate is 75.4 %.4 We see that\\nusing 400 trees is sufficient to give good performance, and that the choice\\nm = √p gave a small improvement in test error over bagging (m = p) in\\nthis example. As with bagging, random forests will not overfit if we increase\\nB, so in practice we use a value of B sufficiently large for the error rate to\\nhave settled down.\\n8.2.3\\nBoosting\\nWe now discuss boosting, yet another approach for improving the predic- boosting\\ntions resulting from a decision tree. Like bagging, boosting is a general\\napproach that can be applied to many statistical learning methods for re-\\ngression or classification. Here we restrict our discussion of boosting to the\\ncontext of decision trees.\\nRecall that bagging involves creating multiple copies of the original train-\\ning data set using the bootstrap, fitting a separate decision tree to each\\ncopy, and then combining all of the trees in order to create a single predic-\\n4The null rate results from simply classifying each observation to the dominant class\\noverall, which is in this case the normal class. 348\\n8. Tree-Based Methods\\n0\\n100\\n200\\n300\\n400\\n500\\n0.2\\n0.3\\n0.4\\n0.5\\nNumber of Trees\\nTest Classification Error\\nm=p\\nm=p/2\\nm= p\\nFIGURE 8.10. Results from random forests for the 15-class gene expression\\ndata set with p = 500 predictors. The test error is displayed as a function of\\nthe number of trees. Each colored line corresponds to a different value of m, the\\nnumber of predictors available for splitting at each interior tree node. Random\\nforests (m < p) lead to a slight improvement over bagging (m = p). A single\\nclassification tree has an error rate of 45.7 %.\\ntive model. Notably, each tree is built on a bootstrap data set, independent\\nof the other trees. Boosting works in a similar way, except that the trees are\\ngrown sequentially: each tree is grown using information from previously\\ngrown trees. Boosting does not involve bootstrap sampling; instead each\\ntree is fit on a modified version of the original data set.\\nConsider first the regression setting. Like bagging, boosting involves com-\\nbining a large number of decision trees, ˆf 1, . . . , ˆf B. Boosting is described\\nin Algorithm 8.2.\\nWhat is the idea behind this procedure? Unlike fitting a single large deci-\\nsion tree to the data, which amounts to fitting the data hard and potentially\\noverfitting, the boosting approach instead learns slowly. Given the current\\nmodel, we fit a decision tree to the residuals from the model. That is, we\\nfit a tree using the current residuals, rather than the outcome Y , as the re-\\nsponse. We then add this new decision tree into the fitted function in order\\nto update the residuals. Each of these trees can be rather small, with just\\na few terminal nodes, determined by the parameter d in the algorithm. By\\nfitting small trees to the residuals, we slowly improve ˆf in areas where it\\ndoes not perform well. The shrinkage parameter λ slows the process down\\neven further, allowing more and different shaped trees to attack the resid-\\nuals. In general, statistical learning approaches that learn slowly tend to\\nperform well. Note that in boosting, unlike in bagging, the construction of\\neach tree depends strongly on the trees that have already been grown.\\nWe have just described the process of boosting regression trees. Boosting\\nclassification trees proceeds in a similar but slightly more complex way, and\\nthe details are omitted here. 8.2 Bagging, Random Forests, Boosting, and Bayesian Additive Regression Trees\\n349\\nAlgorithm 8.2 Boosting for Regression Trees\\n1. Set ˆf(x) = 0 and ri = yi for all i in the training set.\\n2. For b = 1, 2, . . . , B, repeat:\\n(a) Fit a tree ˆf b with d splits (d + 1 terminal nodes) to the training\\ndata (X, r).\\n(b) Update ˆf by adding in a shrunken version of the new tree:\\nˆf(x) ←ˆf(x) + λ ˆf b(x).\\n(8.10)\\n(c) Update the residuals,\\nri ←ri −λ ˆf b(xi).\\n(8.11)\\n3. Output the boosted model,\\nˆf(x) =\\nB\\n0\\nb=1\\nλ ˆf b(x).\\n(8.12)\\nBoosting has three tuning parameters:\\n1. The number of trees B. Unlike bagging and random forests, boosting\\ncan overfit if B is too large, although this overfitting tends to occur\\nslowly if at all. We use cross-validation to select B.\\n2. The shrinkage parameter λ, a small positive number. This controls\\nthe rate at which boosting learns. Typical values are 0.01 or 0.001, and\\nthe right choice can depend on the problem. Very small λ can require\\nusing a very large value of B in order to achieve good performance.\\n3. The number d of splits in each tree, which controls the complexity\\nof the boosted ensemble. Often d = 1 works well, in which case each\\ntree is a stump, consisting of a single split. In this case, the boosted stump\\nensemble is fitting an additive model, since each term involves only a\\nsingle variable. More generally d is the interaction depth, and controls interaction\\ndepth\\nthe interaction order of the boosted model, since d splits can involve\\nat most d variables.\\nIn Figure 8.11, we applied boosting to the 15-class cancer gene expression\\ndata set, in order to develop a classifier that can distinguish the normal\\nclass from the 14 cancer classes. We display the test error as a function of\\nthe total number of trees and the interaction depth d. We see that simple\\nstumps with an interaction depth of one perform well if enough of them\\nare included. This model outperforms the depth-two model, and both out-\\nperform a random forest. This highlights one difference between boosting\\nand random forests: in boosting, because the growth of a particular tree\\ntakes into account the other trees that have already been grown, smaller 350\\n8. Tree-Based Methods\\n0\\n1000\\n2000\\n3000\\n4000\\n5000\\n0.05\\n0.10\\n0.15\\n0.20\\n0.25\\nNumber of Trees\\nTest Classification Error\\nBoosting: depth=1\\nBoosting: depth=2\\nRandomForest: m= p\\nFIGURE 8.11. Results from performing boosting and random forests on the\\n15-class gene expression data set in order to predict cancer versus normal. The test\\nerror is displayed as a function of the number of trees. For the two boosted models,\\nλ = 0.01. Depth-1 trees slightly outperform depth-2 trees, and both outperform\\nthe random forest, although the standard errors are around 0.02, making none of\\nthese differences significant. The test error rate for a single tree is 24 %.\\ntrees are typically sufficient. Using smaller trees can aid in interpretability\\nas well; for instance, using stumps leads to an additive model.\\n8.2.4\\nBayesian Additive Regression Trees\\nFinally, we discuss Bayesian additive regression trees (BART), another en- Bayesian\\nadditive\\nregression\\ntrees\\nsemble method that uses decision trees as its building blocks. For simplicity,\\nwe present BART for regression (as opposed to classification).\\nRecall that bagging and random forests make predictions from an aver-\\nage of regression trees, each of which is built using a random sample of data\\nand/or predictors. Each tree is built separately from the others. By con-\\ntrast, boosting uses a weighted sum of trees, each of which is constructed\\nby fitting a tree to the residual of the current fit. Thus, each new tree at-\\ntempts to capture signal that is not yet accounted for by the current set\\nof trees. BART is related to both approaches: each tree is constructed in\\na random manner as in bagging and random forests, and each tree tries to\\ncapture signal not yet accounted for by the current model, as in boosting.\\nThe main novelty in BART is the way in which new trees are generated.\\nBefore we introduce the BART algorithm, we define some notation. We\\nlet K denote the number of regression trees, and B the number of iterations\\nfor which the BART algorithm will be run. The notation ˆf b\\nk(x) represents\\nthe prediction at x for the kth regression tree used in the bth iteration. At\\nthe end of each iteration, the K trees from that iteration will be summed,\\ni.e. ˆf b(x) = )K\\nk=1 ˆf b\\nk(x) for b = 1, . . . , B.\\nIn the first iteration of the BART algorithm, all trees are initialized to\\nhave a single root node, with ˆf 1\\nk(x) =\\n1\\nnK\\n)n\\ni=1 yi, the mean of the response 8.2 Bagging, Random Forests, Boosting, and Bayesian Additive Regression Trees\\n351\\n(a): ˆf b−1\\nk\\n(X)\\n(b): Possibility #1 for ˆf b\\nk(X)\\n|\\nX < 169.17\\nX < 114.305\\nX < 140.35\\n−0.5031\\n 0.2667\\n−0.2470\\n 0.4079\\n|\\nX < 169.17\\nX < 114.305\\nX < 140.35\\n−0.5110\\n 0.2693\\n−0.2649\\n 0.4221\\n(c): Possibility #2 for ˆf b\\nk(X)\\n(d): Possibility #3 for ˆf b\\nk(X)\\n|\\nX < 169.17\\n−0.1218\\n 0.4079\\n|\\nX < 169.17\\nX < 114.305\\nX < 106.755\\nX < 140.35\\n−0.05089\\n−1.03100\\n 0.26670\\n−0.24700\\n 0.40790\\nFIGURE 8.12. A schematic of perturbed trees from the BART algorithm. (a):\\nThe kth tree at the (b −1)st iteration, ˆf b−1\\nk\\n(X), is displayed. Panels (b)–(d)\\ndisplay three of many possibilities for ˆf b\\nk(X), given the form of ˆf b−1\\nk\\n(X). (b): One\\npossibility is that ˆf b\\nk(X) has the same structure as ˆf b−1\\nk\\n(X), but with different\\npredictions at the terminal nodes. (c): Another possibility is that ˆf b\\nk(X) results\\nfrom pruning ˆf b−1\\nk\\n(X). (d): Alternatively, ˆf b\\nk(X) may have more terminal nodes\\nthan ˆf b−1\\nk\\n(X).\\nvalues divided by the total number of trees. Thus, ˆf 1(x) = )K\\nk=1 ˆf 1\\nk(x) =\\n1\\nn\\n)n\\ni=1 yi.\\nIn subsequent iterations, BART updates each of the K trees, one at a\\ntime. In the bth iteration, to update the kth tree, we subtract from each\\nresponse value the predictions from all but the kth tree, in order to obtain\\na partial residual\\nri = yi −\\n0\\nk′<k\\nˆf b\\nk′(xi) −\\n0\\nk′>k\\nˆf b−1\\nk′\\n(xi)\\nfor the ith observation, i = 1, . . . , n. Rather than fitting a fresh tree to this\\npartial residual, BART randomly chooses a perturbation to the tree from\\nthe previous iteration ( ˆf b−1\\nk\\n) from a set of possible perturbations, favoring\\nones that improve the fit to the partial residual. There are two components\\nto this perturbation:\\n1. We may change the structure of the tree by adding or pruning branches.\\n2. We may change the prediction in each terminal node of the tree.\\nFigure 8.12 illustrates examples of possible perturbations to a tree.\\nThe output of BART is a collection of prediction models,\\nˆf b(x) =\\nK\\n0\\nk=1\\nˆf b\\nk(x), for b = 1, 2, . . . , B. 352\\n8. Tree-Based Methods\\nAlgorithm 8.3 Bayesian Additive Regression Trees\\n1. Let ˆf 1\\n1 (x) = ˆf 1\\n2 (x) = · · · = ˆf 1\\nK(x) =\\n1\\nnK\\n)n\\ni=1 yi.\\n2. Compute ˆf 1(x) = )K\\nk=1 ˆf 1\\nk(x) = 1\\nn\\n)n\\ni=1 yi.\\n3. For b = 2, . . . , B:\\n(a) For k = 1, 2, . . . , K:\\ni. For i = 1, . . . , n, compute the current partial residual\\nri = yi −\\n0\\nk′<k\\nˆf b\\nk′(xi) −\\n0\\nk′>k\\nˆf b−1\\nk′\\n(xi).\\nii. Fit a new tree, ˆf b\\nk(x), to ri, by randomly perturbing the\\nkth tree from the previous iteration, ˆf b−1\\nk\\n(x). Perturbations\\nthat improve the fit are favored.\\n(b) Compute ˆf b(x) = )K\\nk=1 ˆf b\\nk(x).\\n4. Compute the mean after L burn-in samples,\\nˆf(x) =\\n1\\nB −L\\nB\\n0\\nb=L+1\\nˆf b(x).\\nWe typically throw away the first few of these prediction models, since\\nmodels obtained in the earlier iterations — known as the burn-in period burn-in\\n— tend not to provide very good results. We can let L denote the num-\\nber of burn-in iterations; for instance, we might take L = 200. Then, to\\nobtain a single prediction, we simply take the average after the burn-in\\niterations, ˆf(x) =\\n1\\nB−L\\n)B\\nb=L+1 ˆf b(x). However, it is also possible to com-\\npute quantities other than the average: for instance, the percentiles of\\nˆf L+1(x), . . . , ˆf B(x) provide a measure of uncertainty in the final predic-\\ntion. The overall BART procedure is summarized in Algorithm 8.3.\\nA key element of the BART approach is that in Step 3(a)ii., we do not fit\\na fresh tree to the current partial residual: instead, we try to improve the fit\\nto the current partial residual by slightly modifying the tree obtained in the\\nprevious iteration (see Figure 8.12). Roughly speaking, this guards against\\noverfitting since it limits how “hard” we fit the data in each iteration.\\nFurthermore, the individual trees are typically quite small. We limit the\\ntree size in order to avoid overfitting the data, which would be more likely\\nto occur if we grew very large trees.\\nFigure 8.13 shows the result of applying BART to the Heart data, using\\nK = 200 trees, as the number of iterations is increased to 10, 000. During\\nthe initial iterations, the test and training errors jump around a bit. After\\nthis initial burn-in period, the error rates settle down. We note that there\\nis only a small difference between the training error and the test error,\\nindicating that the tree perturbation process largely avoids overfitting. 8.2 Bagging, Random Forests, Boosting, and Bayesian Additive Regression Trees\\n353\\n5\\n10\\n50\\n100\\n500\\n5000\\n0.0\\n0.1\\n0.2\\n0.3\\n0.4\\n0.5\\nNumber of Iterations\\nError\\nBART Training Error\\nBART Test Error\\nBoosting Training Error\\nBoosting Test Error\\nFIGURE 8.13. BART and boosting results for the Heart data. Both training\\nand test errors are displayed. After a burn-in period of 100 iterations (shown in\\ngray), the error rates for BART settle down. Boosting begins to overfit after a\\nfew hundred iterations.\\nThe training and test errors for boosting are also displayed in Figure 8.13.\\nWe see that the test error for boosting approaches that of BART, but then\\nbegins to increase as the number of iterations increases. Furthermore, the\\ntraining error for boosting decreases as the number of iterations increases,\\nindicating that boosting has overfit the data.\\nThough the details are outside of the scope of this book, it turns out\\nthat the BART method can be viewed as a Bayesian approach to fitting an\\nensemble of trees: each time we randomly perturb a tree in order to fit the\\nresiduals, we are in fact drawing a new tree from a posterior distribution.\\n(Of course, this Bayesian connection is the motivation for BART’s name.)\\nFurthermore, Algorithm 8.3 can be viewed as a Markov chain Monte Carlo Markov\\nchain Monte\\nCarlo\\nalgorithm for fitting the BART model.\\nWhen we apply BART, we must select the number of trees K, the number\\nof iterations B, and the number of burn-in iterations L. We typically choose\\nlarge values for B and K, and a moderate value for L: for instance, K = 200,\\nB = 1,000, and L = 100 is a reasonable choice. BART has been shown to\\nhave very impressive out-of-box performance — that is, it performs well\\nwith minimal tuning.\\n8.2.5\\nSummary of Tree Ensemble Methods\\nTrees are an attractive choice of weak learner for an ensemble method\\nfor a number of reasons, including their flexibility and ability to handle 354\\n8. Tree-Based Methods\\npredictors of mixed types (i.e. qualitative as well as quantitative). We have\\nnow seen four approaches for fitting an ensemble of trees: bagging, random\\nforests, boosting, and BART.\\n• In bagging, the trees are grown independently on random samples of\\nthe observations. Consequently, the trees tend to be quite similar to\\neach other. Thus, bagging can get caught in local optima and can fail\\nto thoroughly explore the model space.\\n• In random forests, the trees are once again grown independently on\\nrandom samples of the observations. However, each split on each tree\\nis performed using a random subset of the features, thereby decorre-\\nlating the trees, and leading to a more thorough exploration of model\\nspace relative to bagging.\\n• In boosting, we only use the original data, and do not draw any ran-\\ndom samples. The trees are grown successively, using a “slow” learn-\\ning approach: each new tree is fit to the signal that is left over from\\nthe earlier trees, and shrunken down before it is used.\\n• In BART, we once again only make use of the original data, and we\\ngrow the trees successively. However, each tree is perturbed in order\\nto avoid local minima and achieve a more thorough exploration of\\nthe model space.\\n8.3\\nLab: Tree-Based Methods\\nWe import some of our usual libraries at this top level.\\nIn [1]: import numpy as np\\nimport pandas as pd\\nfrom matplotlib.pyplot import subplots\\nfrom statsmodels.datasets import get_rdataset\\nimport sklearn.model_selection as skm\\nfrom ISLP import load_data, confusion_table\\nfrom ISLP.models import ModelSpec as MS\\nWe also collect the new imports needed for this lab.\\nIn [2]: from sklearn.tree import (DecisionTreeClassifier as DTC,\\nDecisionTreeRegressor as DTR,\\nplot_tree,\\nexport_text)\\nfrom sklearn.metrics import (accuracy_score ,\\nlog_loss)\\nfrom sklearn.ensemble import \\\\\\n(RandomForestRegressor as RF,\\nGradientBoostingRegressor as GBR)\\nfrom ISLP.bart import BART 8.3 Lab: Tree-Based Methods\\n355\\n8.3.1\\nFitting Classification Trees\\nWe first use classification trees to analyze the Carseats data set. In these\\ndata, Sales is a continuous variable, and so we begin by recoding it as a\\nbinary variable. We use the where() function to create a variable, called where()\\nHigh, which takes on a value of Yes if the Sales variable exceeds 8, and\\ntakes on a value of No otherwise.\\nIn [3]: Carseats = load_data(\\'Carseats\\')\\nHigh = np.where(Carseats.Sales > 8,\\n\"Yes\",\\n\"No\")\\nWe now use DecisionTreeClassifier() to fit a classification tree in order DecisionTree\\nClassifier()\\nto predict High using all variables but Sales. To do so, we must form a\\nmodel matrix as we did when fitting regression models.\\nIn [4]: model = MS(Carseats.columns.drop(\\'Sales\\'), intercept=False)\\nD = model.fit_transform(Carseats)\\nfeature_names = list(D.columns)\\nX = np.asarray(D)\\nWe have converted D from a data frame to an array X, which is needed in\\nsome of the analysis below. We also need the feature_names for annotating\\nour plots later.\\nThere are several options needed to specify the classifier, such as max_depth\\n(how deep to grow the tree), min_samples_split (minimum number of ob-\\nservations in a node to be eligible for splitting) and criterion (whether to\\nuse Gini or cross-entropy as the split criterion). We also set random_state\\nfor reproducibility; ties in the split criterion are broken at random.\\nIn [5]: clf = DTC(criterion=\\'entropy\\',\\nmax_depth=3,\\nrandom_state=0)\\nclf.fit(X, High)\\nOut[5]: DecisionTreeClassifier(criterion=\\'entropy\\', max_depth=3)\\nIn our discussion of qualitative features in Section 3.3, we noted that for\\na linear regression model such a feature could be represented by including a\\nmatrix of dummy variables (one-hot-encoding) in the model matrix, using\\nthe formula notation of statsmodels. As mentioned in Section 8.1, there is\\na more natural way to handle qualitative features when building a decision\\ntree, that does not require such dummy variables; each split amounts to\\npartitioning the levels into two groups. However, the sklearn implementa-\\ntion of decision trees does not take advantage of this approach; instead it\\nsimply treats the one-hot-encoded levels as separate variables.\\nIn [6]: accuracy_score(High, clf.predict(X))\\nOut[6]: 0.7275\\nWith only the default arguments, the training error rate is 21%. For clas-\\nsification trees, we can access the value of the deviance using log_loss(),\\nlog_loss() 356\\n8. Tree-Based Methods\\n−2\\n0\\nm\\n0\\nk\\nnmk log ˆpmk,\\nwhere nmk is the number of observations in the mth terminal node that\\nbelong to the kth class.\\nIn [7]: resid_dev = np.sum(log_loss(High, clf.predict_proba(X)))\\nresid_dev\\nOut[7]: 0.4711\\nThis is closely related to the entropy, defined in (8.7). A small deviance\\nindicates a tree that provides a good fit to the (training) data.\\nOne of the most attractive properties of trees is that they can be graphi-\\ncally displayed. Here we use the plot() function to display the tree structure\\n(not shown here).\\nIn [8]: ax = subplots(figsize=(12,12))[1]\\nplot_tree(clf,\\nfeature_names=feature_names ,\\nax=ax);\\nThe most important indicator of Sales appears to be ShelveLoc.\\nWe can see a text representation of the tree using export_text(), which export_text()\\ndisplays the split criterion (e.g. Price <= 92.5) for each branch. For leaf\\nnodes it shows the overall prediction (Yes or No). We can also see the number\\nof observations in that leaf that take on values of Yes and No by specifying\\nshow_weights=True.\\nIn [9]: print(export_text(clf,\\nfeature_names=feature_names ,\\nshow_weights=True))\\nOut[9]: |--- ShelveLoc[Good] <= 0.50\\n|\\n|--- Price <= 92.50\\n|\\n|\\n|--- Income <= 57.00\\n|\\n|\\n|\\n|--- weights: [7.00, 3.00] class: No\\n|\\n|\\n|--- Income >\\n57.00\\n|\\n|\\n|\\n|--- weights: [7.00, 29.00] class: Yes\\n|\\n|--- Price >\\n92.50\\n|\\n|\\n|--- Advertising <= 13.50\\n|\\n|\\n|\\n|--- weights: [183.00, 41.00] class: No\\n|\\n|\\n|--- Advertising >\\n13.50\\n|\\n|\\n|\\n|--- weights: [20.00, 25.00] class: Yes\\n|--- ShelveLoc[Good] >\\n0.50\\n|\\n|--- Price <= 135.00\\n|\\n|\\n|--- US[Yes] <= 0.50\\n|\\n|\\n|\\n|--- weights: [6.00, 11.00] class: Yes\\n|\\n|\\n|--- US[Yes] >\\n0.50\\n|\\n|\\n|\\n|--- weights: [2.00, 49.00] class: Yes\\n|\\n|--- Price >\\n135.00\\n|\\n|\\n|--- Income <= 46.00\\n|\\n|\\n|\\n|--- weights: [6.00, 0.00] class: No\\n|\\n|\\n|--- Income >\\n46.00\\n|\\n|\\n|\\n|--- weights: [5.00, 6.00] class: Yes 8.3 Lab: Tree-Based Methods\\n357\\nIn order to properly evaluate the performance of a classification tree on\\nthese data, we must estimate the test error rather than simply computing\\nthe training error. We split the observations into a training set and a test\\nset, build the tree using the training set, and evaluate its performance\\non the test data. This pattern is similar to that in Chapter 6, with the\\nlinear models replaced here by decision trees — the code for validation is\\nalmost identical. This approach leads to correct predictions for 68.5% of\\nthe locations in the test data set.\\nIn [10]: validation = skm.ShuffleSplit(n_splits=1,\\ntest_size=200,\\nrandom_state=0)\\nresults = skm.cross_validate(clf,\\nD,\\nHigh,\\ncv=validation)\\nresults[\\'test_score\\']\\nOut[10]: array([0.685])\\nNext, we consider whether pruning the tree might lead to improved clas-\\nsification performance. We first split the data into a training and test set.\\nWe will use cross-validation to prune the tree on the training set, and then\\nevaluate the performance of the pruned tree on the test set.\\nIn [11]: (X_train,\\nX_test,\\nHigh_train,\\nHigh_test) = skm.train_test_split(X,\\nHigh,\\ntest_size=0.5,\\nrandom_state=0)\\nWe first refit the full tree on the training set; here we do not set a max_depth\\nparameter, since we will learn that through cross-validation.\\nIn [12]: clf = DTC(criterion=\\'entropy\\', random_state=0)\\nclf.fit(X_train, High_train)\\naccuracy_score(High_test, clf.predict(X_test))\\nOut[12]: 0.735\\nNext we use the cost_complexity_pruning_path() method of clf to extract cost_\\ncomplexity_\\npruning_\\npath()\\ncost-complexity values.\\nIn [13]: ccp_path = clf.cost_complexity_pruning_path(X_train, High_train)\\nkfold = skm.KFold(10,\\nrandom_state=1,\\nshuffle=True)\\nThis yields a set of impurities and α values from which we can extract an\\noptimal one by cross-validation.\\nIn [14]: grid = skm.GridSearchCV(clf,\\n{\\'ccp_alpha\\': ccp_path.ccp_alphas},\\nrefit=True, 358\\n8. Tree-Based Methods\\ncv=kfold,\\nscoring=\\'accuracy\\')\\ngrid.fit(X_train, High_train)\\ngrid.best_score_\\nOut[14]: 0.685\\nLet’s take a look at the pruned true.\\nIn [15]: ax = subplots(figsize=(12, 12))[1]\\nbest_ = grid.best_estimator_\\nplot_tree(best_,\\nfeature_names=feature_names ,\\nax=ax);\\nThis is quite a bushy tree. We could count the leaves, or query best_ instead.\\nIn [16]: best_.tree_.n_leaves\\nOut[16]: 30\\nThe tree with 30 terminal nodes results in the lowest cross-validation error\\nrate, with an accuracy of 68.5%. How well does this pruned tree perform\\non the test data set? Once again, we apply the predict() function.\\nIn [17]: print(accuracy_score(High_test,\\nbest_.predict(X_test)))\\nconfusion = confusion_table(best_.predict(X_test),\\nHigh_test)\\nconfusion\\nOut[17]: 0.72\\nTruth\\nNo\\nYes\\nPredicted\\nNo\\n108\\n61\\nYes\\n10\\n21\\nNow 72.0% of the test observations are correctly classified, which is\\nslightly worse than the error for the full tree (with 35 leaves). So cross-\\nvalidation has not helped us much here; it only pruned off 5 leaves, at\\na cost of a slightly worse error. These results would change if we were to\\nchange the random number seeds above; even though cross-validation gives\\nan unbiased approach to model selection, it does have variance.\\n8.3.2\\nFitting Regression Trees\\nHere we fit a regression tree to the Boston data set. The steps are similar\\nto those for classification trees.\\nIn [18]: Boston = load_data(\"Boston\")\\nmodel = MS(Boston.columns.drop(\\'medv\\'), intercept=False)\\nD = model.fit_transform(Boston)\\nfeature_names = list(D.columns)\\nX = np.asarray(D) 8.3 Lab: Tree-Based Methods\\n359\\nFirst, we split the data into training and test sets, and fit the tree to the\\ntraining data. Here we use 30% of the data for the test set.\\nIn [19]: (X_train,\\nX_test,\\ny_train,\\ny_test) = skm.train_test_split(X,\\nBoston[\\'medv\\'],\\ntest_size=0.3,\\nrandom_state=0)\\nHaving formed our training and test data sets, we fit the regression tree.\\nIn [20]: reg = DTR(max_depth=3)\\nreg.fit(X_train, y_train)\\nax = subplots(figsize=(12,12))[1]\\nplot_tree(reg,\\nfeature_names=feature_names ,\\nax=ax);\\nThe variable lstat measures the percentage of individuals with lower\\nsocioeconomic status. The tree indicates that lower values of lstat corre-\\nspond to more expensive houses. The tree predicts a median house price\\nof $12,042 for small-sized homes (rm < 6.8), in suburbs in which residents\\nhave low socioeconomic status (lstat > 14.4) and the crime-rate is mod-\\nerate (crim > 5.8).\\nNow we use the cross-validation function to see whether pruning the tree\\nwill improve performance.\\nIn [21]: ccp_path = reg.cost_complexity_pruning_path(X_train, y_train)\\nkfold = skm.KFold(5,\\nshuffle=True,\\nrandom_state=10)\\ngrid = skm.GridSearchCV(reg,\\n{\\'ccp_alpha\\': ccp_path.ccp_alphas},\\nrefit=True,\\ncv=kfold,\\nscoring=\\'neg_mean_squared_error\\')\\nG = grid.fit(X_train, y_train)\\nIn keeping with the cross-validation results, we use the pruned tree to\\nmake predictions on the test set.\\nIn [22]: best_ = grid.best_estimator_\\nnp.mean((y_test - best_.predict(X_test))**2)\\nOut[22]: 28.07\\nIn other words, the test set MSE associated with the regression tree is\\n28.07. The square root of the MSE is therefore around 5.30, indicating that\\nthis model leads to test predictions that are within around $5300 of the\\ntrue median home value for the suburb.\\nLet’s plot the best tree to see how interpretable it is.\\nIn [23]: ax = subplots(figsize=(12,12))[1]\\nplot_tree(G.best_estimator_ ,\\nfeature_names=feature_names ,\\nax=ax); 360\\n8. Tree-Based Methods\\n8.3.3\\nBagging and Random Forests\\nHere we apply bagging and random forests to the Boston data, using the\\nRandomForestRegressor() from the sklearn.ensemble package. Recall that RandomForest\\nRegressor()\\nsklearn.\\nensemble\\nbagging is simply a special case of a random forest with m = p. Therefore,\\nthe RandomForestRegressor() function can be used to perform both bagging\\nand random forests. We start with bagging.\\nIn [24]: bag_boston = RF(max_features=X_train.shape[1], random_state=0)\\nbag_boston.fit(X_train, y_train)\\nOut[24]: RandomForestRegressor(max_features=12, random_state=0)\\nThe argument max_features indicates that all 12 predictors should be\\nconsidered for each split of the tree — in other words, that bagging should\\nbe done. How well does this bagged model perform on the test set?\\nIn [25]: ax = subplots(figsize=(8,8))[1]\\ny_hat_bag = bag_boston.predict(X_test)\\nax.scatter(y_hat_bag, y_test)\\nnp.mean((y_test - y_hat_bag)**2)\\nOut[25]: 14.63\\nThe test set MSE associated with the bagged regression tree is 14.63, about\\nhalf that obtained using an optimally-pruned single tree. We could change\\nthe number of trees grown from the default of 100 by using the n_estimators\\nargument:\\nIn [26]: bag_boston = RF(max_features=X_train.shape[1],\\nn_estimators=500,\\nrandom_state=0).fit(X_train, y_train)\\ny_hat_bag = bag_boston.predict(X_test)\\nnp.mean((y_test - y_hat_bag)**2)\\nOut[26]: 14.61\\nThere is not much change. Bagging and random forests cannot overfit by\\nincreasing the number of trees, but can underfit if the number is too small.\\nGrowing a random forest proceeds in exactly the same way, except\\nthat we use a smaller value of the max_features argument. By default,\\nRandomForestRegressor() uses p variables when building a random forest of\\nregression trees (i.e. it defaults to bagging), and RandomForestClassifier()\\nuses √p variables when building a random forest of classification trees. Here\\nwe use max_features=6.\\nIn [27]: RF_boston = RF(max_features=6,\\nrandom_state=0).fit(X_train, y_train)\\ny_hat_RF = RF_boston.predict(X_test)\\nnp.mean((y_test - y_hat_RF)**2)\\nOut[27]: 20.04\\nThe test set MSE is 20.04; this indicates that random forests did some-\\nwhat worse than bagging in this case. Extracting the feature_importances_\\nvalues from the fitted model, we can view the importance of each variable. 8.3 Lab: Tree-Based Methods\\n361\\nIn [28]: feature_imp = pd.DataFrame(\\n{\\'importance\\':RF_boston.feature_importances_},\\nindex=feature_names)\\nfeature_imp.sort_values(by=\\'importance\\', ascending=False)\\nOut[28]:\\nimportance\\nlstat\\n0.368683\\nrm\\n0.333842\\nptratio\\n0.057306\\nindus\\n0.053303\\ncrim\\n0.052426\\ndis\\n0.042493\\nnox\\n0.034410\\nage\\n0.024327\\ntax\\n0.022368\\nrad\\n0.005048\\nzn\\n0.003238\\nchas\\n0.002557\\nThis is a relative measure of the total decrease in node impurity that results\\nfrom splits over that variable, averaged over all trees (this was plotted in\\nFigure 8.9 for a model fit to the Heart data).\\nThe results indicate that across all of the trees considered in the random\\nforest, the wealth level of the community (lstat) and the house size (rm)\\nare by far the two most important variables.\\n8.3.4\\nBoosting\\nHere we use GradientBoostingRegressor() from sklearn.ensemble to fit Gradient\\nBoosting\\nRegressor()\\nboosted regression trees to the Boston data set. For classification we would\\nuse GradientBoostingClassifier(). The argument n_estimators=5000 indi-\\nGradient\\nBoosting\\nClassifier()\\ncates that we want 5000 trees, and the option max_depth=3 limits the depth\\nof each tree. The argument learning_rate is the λ mentioned earlier in the\\ndescription of boosting.\\nIn [29]: boost_boston = GBR(n_estimators=5000,\\nlearning_rate=0.001,\\nmax_depth=3,\\nrandom_state=0)\\nboost_boston.fit(X_train, y_train)\\nWe can see how the training error decreases with the train_score_ at-\\ntribute. To get an idea of how the test error decreases we can use the\\nstaged_predict() method to get the predicted values along the path.\\nIn [30]: test_error = np.zeros_like(boost_boston.train_score_)\\nfor idx, y_ in enumerate(boost_boston.staged_predict(X_test)):\\ntest_error[idx] = np.mean((y_test - y_)**2)\\nplot_idx = np.arange(boost_boston.train_score_.shape[0])\\nax = subplots(figsize=(8,8))[1]\\nax.plot(plot_idx,\\nboost_boston.train_score_ ,\\n\\'b\\',\\nlabel=\\'Training\\') 362\\n8. Tree-Based Methods\\nax.plot(plot_idx,\\ntest_error,\\n\\'r\\',\\nlabel=\\'Test\\')\\nax.legend();\\nWe now use the boosted model to predict medv on the test set:\\nIn [31]: y_hat_boost = boost_boston.predict(X_test);\\nnp.mean((y_test - y_hat_boost)**2)\\nOut[31]: 14.48\\nThe test MSE obtained is 14.48, similar to the test MSE for bagging. If we\\nwant to, we can perform boosting with a different value of the shrinkage\\nparameter λ in (8.10). The default value is 0.001, but this is easily modified.\\nHere we take λ = 0.2.\\nIn [32]: boost_boston = GBR(n_estimators=5000,\\nlearning_rate=0.2,\\nmax_depth=3,\\nrandom_state=0)\\nboost_boston.fit(X_train,\\ny_train)\\ny_hat_boost = boost_boston.predict(X_test);\\nnp.mean((y_test - y_hat_boost)**2)\\nOut[32]: 14.50\\nIn this case, using λ = 0.2 leads to a almost the same test MSE as when\\nusing λ = 0.001.\\n8.3.5\\nBayesian Additive Regression Trees\\nIn this section we demonstrate a Python implementation of BART found in\\nthe ISLP.bart package. We fit a model to the Boston housing data set. This\\nBART() estimator is designed for quantitative outcome variables, though BART()\\nother implementations are available for fitting logistic and probit models\\nto categorical outcomes.\\nIn [33]: bart_boston = BART(random_state=0, burnin=5, ndraw=15)\\nbart_boston.fit(X_train, y_train)\\nOut[33]: BART(burnin=5, ndraw=15, random_state=0)\\nOn this data set, with this split into test and training, we see that the\\ntest error of BART is similar to that of random forest.\\nIn [34]: yhat_test = bart_boston.predict(X_test.astype(np.float32))\\nnp.mean((y_test - yhat_test)**2)\\nOut[34]: 20.92\\nWe can check how many times each variable appeared in the collection\\nof trees. This gives a summary similar to the variable importance plot for\\nboosting and random forests. 8.4 Exercises\\n363\\nIn [35]: var_inclusion = pd.Series(bart_boston.variable_inclusion_.mean(0),\\nindex=D.columns)\\nvar_inclusion\\nOut[35]:\\ncrim\\n25.333333\\nzn\\n27.000000\\nindus\\n21.266667\\nchas\\n20.466667\\nnox\\n25.400000\\nrm\\n32.400000\\nage\\n26.133333\\ndis\\n25.666667\\nrad\\n24.666667\\ntax\\n23.933333\\nptratio\\n25.000000\\nlstat\\n31.866667\\ndtype: float64\\n8.4\\nExercises\\nConceptual\\n1. Draw an example (of your own invention) of a partition of two-\\ndimensional feature space that could result from recursive binary\\nsplitting. Your example should contain at least six regions. Draw a\\ndecision tree corresponding to this partition. Be sure to label all as-\\npects of your figures, including the regions R1, R2, . . ., the cutpoints\\nt1, t2, . . ., and so forth.\\nHint: Your result should look something like Figures 8.1 and 8.2.\\n2. It is mentioned in Section 8.2.3 that boosting using depth-one trees\\n(or stumps) leads to an additive model: that is, a model of the form\\nf(X) =\\np\\n0\\nj=1\\nfj(Xj).\\nExplain why this is the case. You can begin with (8.12) in\\nAlgorithm 8.2.\\n3. Consider the Gini index, classification error, and entropy in a simple\\nclassification setting with two classes. Create a single plot that dis-\\nplays each of these quantities as a function of ˆpm1. The x-axis should\\ndisplay ˆpm1, ranging from 0 to 1, and the y-axis should display the\\nvalue of the Gini index, classification error, and entropy.\\nHint: In a setting with two classes, ˆpm1 = 1 −ˆpm2. You could make\\nthis plot by hand, but it will be much easier to make in R.\\n4. This question relates to the plots in Figure 8.14. 364\\n8. Tree-Based Methods\\n| \\nX2 < 1 \\nX1 < 1 \\nX1 < 0 \\nX2 < 2 \\n-1.80 \\n-1.06 \\n 0.21 \\n 0.63 \\n 2.49 \\n \\n \\n5 \\n15 \\n10 \\n0 \\n3 \\n0 \\n1 \\nX2 \\nX1 \\n0 \\n1 \\nFIGURE 8.14. Left: A partition of the predictor space corresponding to Exer-\\ncise 4a. Right: A tree corresponding to Exercise 4b.\\n(a) Sketch the tree corresponding to the partition of the predictor\\nspace illustrated in the left-hand panel of Figure 8.14. The num-\\nbers inside the boxes indicate the mean of Y within each region.\\n(b) Create a diagram similar to the left-hand panel of Figure 8.14,\\nusing the tree illustrated in the right-hand panel of the same\\nfigure. You should divide up the predictor space into the correct\\nregions, and indicate the mean for each region.\\n5. Suppose we produce ten bootstrapped samples from a data set\\ncontaining red and green classes. We then apply a classification tree\\nto each bootstrapped sample and, for a specific value of X, produce\\n10 estimates of P(Class is Red|X):\\n0.1, 0.15, 0.2, 0.2, 0.55, 0.6, 0.6, 0.65, 0.7, and 0.75.\\nThere are two common ways to combine these results together into a\\nsingle class prediction. One is the majority vote approach discussed in\\nthis chapter. The second approach is to classify based on the average\\nprobability. In this example, what is the final classification under each\\nof these two approaches?\\n6. Provide a detailed explanation of the algorithm that is used to fit a\\nregression tree.\\nApplied\\n7. In Section 8.3.3, we applied random forests to the Boston data using\\nmax_features = 6 and using n_estimators = 100 and n_estimators =\\n500. Create a plot displaying the test error resulting from random\\nforests on this data set for a more comprehensive range of values\\nfor max_features and n_estimators. You can model your plot after\\nFigure 8.10. Describe the results obtained.\\n8. In the lab, a classification tree was applied to the Carseats data set af-\\nter converting Sales into a qualitative response variable. Now we will\\nseek to predict Sales using regression trees and related approaches,\\ntreating the response as a quantitative variable. 8.4 Exercises\\n365\\n(a) Split the data set into a training set and a test set.\\n(b) Fit a regression tree to the training set. Plot the tree, and inter-\\npret the results. What test MSE do you obtain?\\n(c) Use cross-validation in order to determine the optimal level of\\ntree complexity. Does pruning the tree improve the test MSE?\\n(d) Use the bagging approach in order to analyze this data. What\\ntest MSE do you obtain? Use the feature_importance_ values to\\ndetermine which variables are most important.\\n(e) Use random forests to analyze this data. What test MSE do\\nyou obtain? Use the feature_importance_ values to determine\\nwhich variables are most important. Describe the effect of m, the\\nnumber of variables considered at each split, on the error rate\\nobtained.\\n(f) Now analyze the data using BART, and report your results.\\n9. This problem involves the OJ data set which is part of the ISLP\\npackage.\\n(a) Create a training set containing a random sample of 800 obser-\\nvations, and a test set containing the remaining observations.\\n(b) Fit a tree to the training data, with Purchase as the response\\nand the other variables as predictors. What is the training error\\nrate?\\n(c) Create a plot of the tree, and interpret the results. How many\\nterminal nodes does the tree have?\\n(d) Use the export_tree() function to produce a text summary of\\nthe fitted tree. Pick one of the terminal nodes, and interpret the\\ninformation displayed.\\n(e) Predict the response on the test data, and produce a confusion\\nmatrix comparing the test labels to the predicted test labels.\\nWhat is the test error rate?\\n(f) Use cross-validation on the training set in order to determine\\nthe optimal tree size.\\n(g) Produce a plot with tree size on the x-axis and cross-validated\\nclassification error rate on the y-axis.\\n(h) Which tree size corresponds to the lowest cross-validated classi-\\nfication error rate?\\n(i) Produce a pruned tree corresponding to the optimal tree size\\nobtained using cross-validation. If cross-validation does not lead\\nto selection of a pruned tree, then create a pruned tree with five\\nterminal nodes.\\n(j) Compare the training error rates between the pruned and un-\\npruned trees. Which is higher?\\n(k) Compare the test error rates between the pruned and unpruned\\ntrees. Which is higher? 366\\n8. Tree-Based Methods\\n10. We now use boosting to predict Salary in the Hitters data set.\\n(a) Remove the observations for whom the salary information is\\nunknown, and then log-transform the salaries.\\n(b) Create a training set consisting of the first 200 observations, and\\na test set consisting of the remaining observations.\\n(c) Perform boosting on the training set with 1,000 trees for a range\\nof values of the shrinkage parameter λ. Produce a plot with\\ndifferent shrinkage values on the x-axis and the corresponding\\ntraining set MSE on the y-axis.\\n(d) Produce a plot with different shrinkage values on the x-axis and\\nthe corresponding test set MSE on the y-axis.\\n(e) Compare the test MSE of boosting to the test MSE that results\\nfrom applying two of the regression approaches seen in\\nChapters 3 and 6.\\n(f) Which variables appear to be the most important predictors in\\nthe boosted model?\\n(g) Now apply bagging to the training set. What is the test set MSE\\nfor this approach?\\n11. This question uses the Caravan data set.\\n(a) Create a training set consisting of the first 1,000 observations,\\nand a test set consisting of the remaining observations.\\n(b) Fit a boosting model to the training set with Purchase as the\\nresponse and the other variables as predictors. Use 1,000 trees,\\nand a shrinkage value of 0.01. Which predictors appear to be\\nthe most important?\\n(c) Use the boosting model to predict the response on the test data.\\nPredict that a person will make a purchase if the estimated prob-\\nability of purchase is greater than 20 %. Form a confusion ma-\\ntrix. What fraction of the people predicted to make a purchase\\ndo in fact make one? How does this compare with the results\\nobtained from applying KNN or logistic regression to this data\\nset?\\n12. Apply boosting, bagging, random forests, and BART to a data set\\nof your choice. Be sure to fit the models on a training set and to\\nevaluate their performance on a test set. How accurate are the results\\ncompared to simple methods like linear or logistic regression? Which\\nof these approaches yields the best performance? 9\\nSupport Vector Machines\\nIn this chapter, we discuss the support vector machine (SVM), an approach\\nfor classification that was developed in the computer science community in\\nthe 1990s and that has grown in popularity since then. SVMs have been\\nshown to perform well in a variety of settings, and are often considered one\\nof the best “out of the box” classifiers.\\nThe support vector machine is a generalization of a simple and intu-\\nitive classifier called the maximal margin classifier, which we introduce in\\nSection 9.1. Though it is elegant and simple, we will see that this classifier\\nunfortunately cannot be applied to most data sets, since it requires that\\nthe classes be separable by a linear boundary. In Section 9.2, we introduce\\nthe support vector classifier, an extension of the maximal margin classifier\\nthat can be applied in a broader range of cases. Section 9.3 introduces the\\nsupport vector machine, which is a further extension of the support vec-\\ntor classifier in order to accommodate non-linear class boundaries. Support\\nvector machines are intended for the binary classification setting in which\\nthere are two classes; in Section 9.4 we discuss extensions of support vector\\nmachines to the case of more than two classes. In Section 9.5 we discuss\\nthe close connections between support vector machines and other statistical\\nmethods such as logistic regression.\\nPeople often loosely refer to the maximal margin classifier, the support\\nvector classifier, and the support vector machine as “support vector\\nmachines”. To avoid confusion, we will carefully distinguish between these\\nthree notions in this chapter.\\n9.1\\nMaximal Margin Classifier\\nIn this section, we define a hyperplane and introduce the concept of an\\noptimal separating hyperplane.\\n© Springer Nature Switzerland AG 2023 \\nG. James et al., An Introduction to Statistical Learning, Springer Texts in Statistics, \\nhttps://doi.org/10.1007/978-3-031-38747-0_9 \\n367 368\\n9. Support Vector Machines\\n9.1.1\\nWhat Is a Hyperplane?\\nIn a p-dimensional space, a hyperplane is a flat affine subspace of hyperplane\\ndimension p −1.1 For instance, in two dimensions, a hyperplane is a flat\\none-dimensional subspace—in other words, a line. In three dimensions, a\\nhyperplane is a flat two-dimensional subspace—that is, a plane. In p > 3\\ndimensions, it can be hard to visualize a hyperplane, but the notion of a\\n(p −1)-dimensional flat subspace still applies.\\nThe mathematical definition of a hyperplane is quite simple. In two di-\\nmensions, a hyperplane is defined by the equation\\nβ0 + β1X1 + β2X2 = 0\\n(9.1)\\nfor parameters β0, β1, and β2. When we say that (9.1) “defines” the hyper-\\nplane, we mean that any X = (X1, X2)T for which (9.1) holds is a point\\non the hyperplane. Note that (9.1) is simply the equation of a line, since\\nindeed in two dimensions a hyperplane is a line.\\nEquation 9.1 can be easily extended to the p-dimensional setting:\\nβ0 + β1X1 + β2X2 + · · · + βpXp = 0\\n(9.2)\\ndefines a p-dimensional hyperplane, again in the sense that if a point X =\\n(X1, X2, . . . , Xp)T in p-dimensional space (i.e. a vector of length p) satisfies\\n(9.2), then X lies on the hyperplane.\\nNow, suppose that X does not satisfy (9.2); rather,\\nβ0 + β1X1 + β2X2 + · · · + βpXp > 0.\\n(9.3)\\nThen this tells us that X lies to one side of the hyperplane. On the other\\nhand, if\\nβ0 + β1X1 + β2X2 + · · · + βpXp < 0,\\n(9.4)\\nthen X lies on the other side of the hyperplane. So we can think of the\\nhyperplane as dividing p-dimensional space into two halves. One can easily\\ndetermine on which side of the hyperplane a point lies by simply calculating\\nthe sign of the left-hand side of (9.2). A hyperplane in two-dimensional\\nspace is shown in Figure 9.1.\\n9.1.2\\nClassification Using a Separating Hyperplane\\nNow suppose that we have an n × p data matrix X that consists of n\\ntraining observations in p-dimensional space,\\nx1 =\\n\\uf8eb\\n\\uf8ec\\n\\uf8ed\\nx11\\n...\\nx1p\\n\\uf8f6\\n\\uf8f7\\n\\uf8f8, . . . , xn =\\n\\uf8eb\\n\\uf8ec\\n\\uf8ed\\nxn1\\n...\\nxnp\\n\\uf8f6\\n\\uf8f7\\n\\uf8f8,\\n(9.5)\\nand that these observations fall into two classes—that is, y1, . . . , yn ∈\\n{−1, 1} where −1 represents one class and 1 the other class. We also have a\\n1The word affine indicates that the subspace need not pass through the origin. 9.1 Maximal Margin Classifier\\n369\\n−1.5\\n−1.0\\n−0.5\\n0.0\\n0.5\\n1.0\\n1.5\\n−1.5\\n−1.0\\n−0.5\\n0.0\\n0.5\\n1.0\\n1.5\\nX1\\nX2\\nFIGURE 9.1. The hyperplane 1 + 2X1 + 3X2 = 0 is shown. The blue region is\\nthe set of points for which 1 + 2X1 + 3X2 > 0, and the purple region is the set of\\npoints for which 1 + 2X1 + 3X2 < 0.\\ntest observation, a p-vector of observed features x∗=\\n\\'x∗\\n1\\n. . .\\nx∗\\np\\n(T . Our\\ngoal is to develop a classifier based on the training data that will correctly\\nclassify the test observation using its feature measurements. We have seen\\na number of approaches for this task, such as linear discriminant analysis\\nand logistic regression in Chapter 4, and classification trees, bagging, and\\nboosting in Chapter 8. We will now see a new approach that is based upon\\nthe concept of a separating hyperplane.\\nseparating\\nhyperplane\\nSuppose that it is possible to construct a hyperplane that separates the\\ntraining observations perfectly according to their class labels. Examples\\nof three such separating hyperplanes are shown in the left-hand panel of\\nFigure 9.2. We can label the observations from the blue class as yi = 1 and\\nthose from the purple class as yi = −1. Then a separating hyperplane has\\nthe property that\\nβ0 + β1xi1 + β2xi2 + · · · + βpxip > 0 if yi = 1,\\n(9.6)\\nand\\nβ0 + β1xi1 + β2xi2 + · · · + βpxip < 0 if yi = −1.\\n(9.7)\\nEquivalently, a separating hyperplane has the property that\\nyi(β0 + β1xi1 + β2xi2 + · · · + βpxip) > 0\\n(9.8)\\nfor all i = 1, . . . , n.\\nIf a separating hyperplane exists, we can use it to construct a very natural\\nclassifier: a test observation is assigned a class depending on which side of\\nthe hyperplane it is located. The right-hand panel of Figure 9.2 shows\\nan example of such a classifier. That is, we classify the test observation x∗\\nbased on the sign of f(x∗) = β0+β1x∗\\n1+β2x∗\\n2+· · ·+βpx∗\\np. If f(x∗) is positive,\\nthen we assign the test observation to class 1, and if f(x∗) is negative, then\\nwe assign it to class −1. We can also make use of the magnitude of f(x∗). If 370\\n9. Support Vector Machines\\n−1\\n0\\n1\\n2\\n3\\n−1\\n0\\n1\\n2\\n3\\n−1\\n0\\n1\\n2\\n3\\n−1\\n0\\n1\\n2\\n3\\nX1\\nX1\\nX2\\nX2\\nFIGURE 9.2. Left: There are two classes of observations, shown in blue and\\nin purple, each of which has measurements on two variables. Three separating\\nhyperplanes, out of many possible, are shown in black. Right: A separating hy-\\nperplane is shown in black. The blue and purple grid indicates the decision rule\\nmade by a classifier based on this separating hyperplane: a test observation that\\nfalls in the blue portion of the grid will be assigned to the blue class, and a test\\nobservation that falls into the purple portion of the grid will be assigned to the\\npurple class.\\nf(x∗) is far from zero, then this means that x∗lies far from the hyperplane,\\nand so we can be confident about our class assignment for x∗. On the other\\nhand, if f(x∗) is close to zero, then x∗is located near the hyperplane, and so\\nwe are less certain about the class assignment for x∗. Not surprisingly, and\\nas we see in Figure 9.2, a classifier that is based on a separating hyperplane\\nleads to a linear decision boundary.\\n9.1.3\\nThe Maximal Margin Classifier\\nIn general, if our data can be perfectly separated using a hyperplane, then\\nthere will in fact exist an infinite number of such hyperplanes. This is\\nbecause a given separating hyperplane can usually be shifted a tiny bit up or\\ndown, or rotated, without coming into contact with any of the observations.\\nThree possible separating hyperplanes are shown in the left-hand panel\\nof Figure 9.2. In order to construct a classifier based upon a separating\\nhyperplane, we must have a reasonable way to decide which of the infinite\\npossible separating hyperplanes to use.\\nA natural choice is the maximal margin hyperplane (also known as the maximal\\nmargin\\nhyperplane\\noptimal separating hyperplane), which is the separating hyperplane that\\noptimal\\nseparating\\nhyperplane\\nis farthest from the training observations. That is, we can compute the\\n(perpendicular) distance from each training observation to a given separat-\\ning hyperplane; the smallest such distance is the minimal distance from the\\nobservations to the hyperplane, and is known as the margin. The maximal\\nmargin\\nmargin hyperplane is the separating hyperplane for which the margin is\\nlargest—that is, it is the hyperplane that has the farthest minimum dis-\\ntance to the training observations. We can then classify a test observation\\nbased on which side of the maximal margin hyperplane it lies. This is known 9.1 Maximal Margin Classifier\\n371\\n−1\\n0\\n1\\n2\\n3\\n−1\\n0\\n1\\n2\\n3\\nX1\\nX2\\nFIGURE 9.3.\\nThere are two classes of observations, shown in blue and in\\npurple. The maximal margin hyperplane is shown as a solid line. The margin\\nis the distance from the solid line to either of the dashed lines. The two blue\\npoints and the purple point that lie on the dashed lines are the support vectors,\\nand the distance from those points to the hyperplane is indicated by arrows. The\\npurple and blue grid indicates the decision rule made by a classifier based on this\\nseparating hyperplane.\\nas the maximal margin classifier. We hope that a classifier that has a large maximal\\nmargin\\nclassifier\\nmargin on the training data will also have a large margin on the test data,\\nand hence will classify the test observations correctly. Although the maxi-\\nmal margin classifier is often successful, it can also lead to overfitting when\\np is large.\\nIf β0, β1, . . . , βp are the coefficients of the maximal margin hyperplane,\\nthen the maximal margin classifier classifies the test observation x∗based\\non the sign of f(x∗) = β0 + β1x∗\\n1 + β2x∗\\n2 + · · · + βpx∗\\np.\\nFigure 9.3 shows the maximal margin hyperplane on the data set of\\nFigure 9.2. Comparing the right-hand panel of Figure 9.2 to Figure 9.3,\\nwe see that the maximal margin hyperplane shown in Figure 9.3 does in-\\ndeed result in a greater minimal distance between the observations and the\\nseparating hyperplane—that is, a larger margin. In a sense, the maximal\\nmargin hyperplane represents the mid-line of the widest “slab” that we can\\ninsert between the two classes.\\nExamining Figure 9.3, we see that three training observations are equidis-\\ntant from the maximal margin hyperplane and lie along the dashed lines\\nindicating the width of the margin. These three observations are known as\\nsupport vectors, since they are vectors in p-dimensional space (in Figure 9.3, support\\nvector\\np = 2) and they “support” the maximal margin hyperplane in the sense\\nthat if these points were moved slightly then the maximal margin hyper-\\nplane would move as well. Interestingly, the maximal margin hyperplane\\ndepends directly on the support vectors, but not on the other observations:\\na movement to any of the other observations would not affect the separating\\nhyperplane, provided that the observation’s movement does not cause it to 372\\n9. Support Vector Machines\\ncross the boundary set by the margin. The fact that the maximal margin\\nhyperplane depends directly on only a small subset of the observations is\\nan important property that will arise later in this chapter when we discuss\\nthe support vector classifier and support vector machines.\\n9.1.4\\nConstruction of the Maximal Margin Classifier\\nWe now consider the task of constructing the maximal margin hyperplane\\nbased on a set of n training observations x1, . . . , xn ∈Rp and associated\\nclass labels y1, . . . , yn ∈{−1, 1}. Briefly, the maximal margin hyperplane\\nis the solution to the optimization problem\\nmaximize\\nβ0,β1,...,βp,M M\\n(9.9)\\nsubject to\\np\\n0\\nj=1\\nβ2\\nj = 1,\\n(9.10)\\nyi(β0 + β1xi1 + β2xi2 + · · · + βpxip) ≥M ∀i = 1, . . . , n. (9.11)\\nThis optimization problem (9.9)–(9.11) is actually simpler than it looks.\\nFirst of all, the constraint in (9.11) that\\nyi(β0 + β1xi1 + β2xi2 + · · · + βpxip) ≥M ∀i = 1, . . . , n\\nguarantees that each observation will be on the correct side of the hyper-\\nplane, provided that M is positive. (Actually, for each observation to be\\non the correct side of the hyperplane we would simply need yi(β0 +β1xi1 +\\nβ2xi2+· · ·+βpxip) > 0, so the constraint in (9.11) in fact requires that each\\nobservation be on the correct side of the hyperplane, with some cushion,\\nprovided that M is positive.)\\nSecond, note that (9.10) is not really a constraint on the hyperplane, since\\nif β0 + β1xi1 + β2xi2 + · · · + βpxip = 0 defines a hyperplane, then so does\\nk(β0 + β1xi1 + β2xi2 + · · · + βpxip) = 0 for any k ̸= 0. However, (9.10) adds\\nmeaning to (9.11); one can show that with this constraint the perpendicular\\ndistance from the ith observation to the hyperplane is given by\\nyi(β0 + β1xi1 + β2xi2 + · · · + βpxip).\\nTherefore, the constraints (9.10) and (9.11) ensure that each observation\\nis on the correct side of the hyperplane and at least a distance M from the\\nhyperplane. Hence, M represents the margin of our hyperplane, and the\\noptimization problem chooses β0, β1, . . . , βp to maximize M. This is exactly\\nthe definition of the maximal margin hyperplane! The problem (9.9)–(9.11)\\ncan be solved efficiently, but details of this optimization are outside of the\\nscope of this book.\\n9.1.5\\nThe Non-separable Case\\nThe maximal margin classifier is a very natural way to perform classifi-\\ncation, if a separating hyperplane exists. However, as we have hinted, in\\nmany cases no separating hyperplane exists, and so there is no maximal 9.2 Support Vector Classifiers\\n373\\n0\\n1\\n2\\n3\\n−1.0\\n−0.5\\n0.0\\n0.5\\n1.0\\n1.5\\n2.0\\nX1\\nX2\\nFIGURE 9.4.\\nThere are two classes of observations, shown in blue and in\\npurple. In this case, the two classes are not separable by a hyperplane, and so the\\nmaximal margin classifier cannot be used.\\nmargin classifier. In this case, the optimization problem (9.9)–(9.11) has no\\nsolution with M > 0. An example is shown in Figure 9.4. In this case, we\\ncannot exactly separate the two classes. However, as we will see in the next\\nsection, we can extend the concept of a separating hyperplane in order to\\ndevelop a hyperplane that almost separates the classes, using a so-called\\nsoft margin. The generalization of the maximal margin classifier to the\\nnon-separable case is known as the support vector classifier.\\n9.2\\nSupport Vector Classifiers\\n9.2.1\\nOverview of the Support Vector Classifier\\nIn Figure 9.4, we see that observations that belong to two classes are not\\nnecessarily separable by a hyperplane. In fact, even if a separating hyper-\\nplane does exist, then there are instances in which a classifier based on\\na separating hyperplane might not be desirable. A classifier based on a\\nseparating hyperplane will necessarily perfectly classify all of the training\\nobservations; this can lead to sensitivity to individual observations. An ex-\\nample is shown in Figure 9.5. The addition of a single observation in the\\nright-hand panel of Figure 9.5 leads to a dramatic change in the maxi-\\nmal margin hyperplane. The resulting maximal margin hyperplane is not\\nsatisfactory—for one thing, it has only a tiny margin. This is problematic\\nbecause as discussed previously, the distance of an observation from the\\nhyperplane can be seen as a measure of our confidence that the obser-\\nvation was correctly classified. Moreover, the fact that the maximal mar-\\ngin hyperplane is extremely sensitive to a change in a single observation\\nsuggests that it may have overfit the training data.\\nIn this case, we might be willing to consider a classifier based on a hy-\\nperplane that does not perfectly separate the two classes, in the interest of 374\\n9. Support Vector Machines\\n−1\\n0\\n1\\n2\\n3\\n−1\\n0\\n1\\n2\\n3\\n−1\\n0\\n1\\n2\\n3\\n−1\\n0\\n1\\n2\\n3\\nX1\\nX1\\nX2\\nX2\\nFIGURE 9.5.\\nLeft: Two classes of observations are shown in blue and in\\npurple, along with the maximal margin hyperplane. Right: An additional blue\\nobservation has been added, leading to a dramatic shift in the maximal margin\\nhyperplane shown as a solid line. The dashed line indicates the maximal margin\\nhyperplane that was obtained in the absence of this additional point.\\n• Greater robustness to individual observations, and\\n• Better classification of most of the training observations.\\nThat is, it could be worthwhile to misclassify a few training observations\\nin order to do a better job in classifying the remaining observations.\\nThe support vector classifier, sometimes called a soft margin classifier, support\\nvector\\nclassifier\\nsoft margin\\nclassifier\\ndoes exactly this. Rather than seeking the largest possible margin so that\\nevery observation is not only on the correct side of the hyperplane but\\nalso on the correct side of the margin, we instead allow some observations\\nto be on the incorrect side of the margin, or even the incorrect side of\\nthe hyperplane. (The margin is soft because it can be violated by some\\nof the training observations.) An example is shown in the left-hand panel\\nof Figure 9.6. Most of the observations are on the correct side of the margin.\\nHowever, a small subset of the observations are on the wrong side of the\\nmargin.\\nAn observation can be not only on the wrong side of the margin, but also\\non the wrong side of the hyperplane. In fact, when there is no separating\\nhyperplane, such a situation is inevitable. Observations on the wrong side of\\nthe hyperplane correspond to training observations that are misclassified by\\nthe support vector classifier. The right-hand panel of Figure 9.6 illustrates\\nsuch a scenario.\\n9.2.2\\nDetails of the Support Vector Classifier\\nThe support vector classifier classifies a test observation depending on\\nwhich side of a hyperplane it lies. The hyperplane is chosen to correctly\\nseparate most of the training observations into the two classes, but may 9.2 Support Vector Classifiers\\n375\\n−0.5\\n0.0\\n0.5\\n1.0\\n1.5\\n2.0\\n2.5\\n−1\\n0\\n1\\n2\\n3\\n4\\n1\\n2\\n3\\n4\\n5\\n6\\n7\\n8\\n9\\n10\\n−0.5\\n0.0\\n0.5\\n1.0\\n1.5\\n2.0\\n2.5\\n−1\\n0\\n1\\n2\\n3\\n4\\n1\\n2\\n3\\n4\\n5\\n6\\n7\\n8\\n9\\n10\\n11\\n12\\nX1\\nX1\\nX2\\nX2\\nFIGURE 9.6. Left: A support vector classifier was fit to a small data set. The\\nhyperplane is shown as a solid line and the margins are shown as dashed lines.\\nPurple observations: Observations 3, 4, 5, and 6 are on the correct side of the\\nmargin, observation 2 is on the margin, and observation 1 is on the wrong side of\\nthe margin. Blue observations: Observations 7 and 10 are on the correct side of\\nthe margin, observation 9 is on the margin, and observation 8 is on the wrong side\\nof the margin. No observations are on the wrong side of the hyperplane. Right:\\nSame as left panel with two additional points, 11 and 12. These two observations\\nare on the wrong side of the hyperplane and the wrong side of the margin.\\nmisclassify a few observations. It is the solution to the optimization problem\\nmaximize\\nβ0,β1,...,βp,ϵ1,...,ϵn, M M\\n(9.12)\\nsubject to\\np\\n0\\nj=1\\nβ2\\nj = 1,\\n(9.13)\\nyi(β0 + β1xi1 + β2xi2 + · · · + βpxip) ≥M(1 −ϵi),\\n(9.14)\\nϵi ≥0,\\nn\\n0\\ni=1\\nϵi ≤C,\\n(9.15)\\nwhere C is a nonnegative tuning parameter. As in (9.11), M is the width\\nof the margin; we seek to make this quantity as large as possible. In (9.14),\\nϵ1, . . . , ϵn are slack variables that allow individual observations to be on slack\\nvariable\\nthe wrong side of the margin or the hyperplane; we will explain them in\\ngreater detail momentarily. Once we have solved (9.12)–(9.15), we classify\\na test observation x∗as before, by simply determining on which side of the\\nhyperplane it lies. That is, we classify the test observation based on the\\nsign of f(x∗) = β0 + β1x∗\\n1 + · · · + βpx∗\\np.\\nThe problem (9.12)–(9.15) seems complex, but insight into its behavior\\ncan be made through a series of simple observations presented below. First\\nof all, the slack variable ϵi tells us where the ith observation is located,\\nrelative to the hyperplane and relative to the margin. If ϵi = 0 then the ith\\nobservation is on the correct side of the margin, as we saw in Section 9.1.4.\\nIf ϵi > 0 then the ith observation is on the wrong side of the margin, and\\nwe say that the ith observation has violated the margin. If ϵi > 1 then it is\\non the wrong side of the hyperplane. 376\\n9. Support Vector Machines\\nWe now consider the role of the tuning parameter C. In (9.15), C bounds\\nthe sum of the ϵi’s, and so it determines the number and severity of the vio-\\nlations to the margin (and to the hyperplane) that we will tolerate. We can\\nthink of C as a budget for the amount that the margin can be violated\\nby the n observations. If C = 0 then there is no budget for violations to\\nthe margin, and it must be the case that ϵ1 = · · · = ϵn = 0, in which case\\n(9.12)–(9.15) simply amounts to the maximal margin hyperplane optimiza-\\ntion problem (9.9)–(9.11). (Of course, a maximal margin hyperplane exists\\nonly if the two classes are separable.) For C > 0 no more than C observa-\\ntions can be on the wrong side of the hyperplane, because if an observation\\nis on the wrong side of the hyperplane then ϵi > 1, and (9.15) requires\\nthat )n\\ni=1 ϵi ≤C. As the budget C increases, we become more tolerant of\\nviolations to the margin, and so the margin will widen. Conversely, as C\\ndecreases, we become less tolerant of violations to the margin and so the\\nmargin narrows. An example is shown in Figure 9.7.\\nIn practice, C is treated as a tuning parameter that is generally chosen via\\ncross-validation. As with the tuning parameters that we have seen through-\\nout this book, C controls the bias-variance trade-off of the statistical learn-\\ning technique. When C is small, we seek narrow margins that are rarely\\nviolated; this amounts to a classifier that is highly fit to the data, which\\nmay have low bias but high variance. On the other hand, when C is larger,\\nthe margin is wider and we allow more violations to it; this amounts to\\nfitting the data less hard and obtaining a classifier that is potentially more\\nbiased but may have lower variance.\\nThe optimization problem (9.12)–(9.15) has a very interesting property:\\nit turns out that only observations that either lie on the margin or that\\nviolate the margin will affect the hyperplane, and hence the classifier ob-\\ntained. In other words, an observation that lies strictly on the correct side\\nof the margin does not affect the support vector classifier! Changing the\\nposition of that observation would not change the classifier at all, provided\\nthat its position remains on the correct side of the margin. Observations\\nthat lie directly on the margin, or on the wrong side of the margin for\\ntheir class, are known as support vectors. These observations do affect the\\nsupport vector classifier.\\nThe fact that only support vectors affect the classifier is in line with our\\nprevious assertion that C controls the bias-variance trade-off of the support\\nvector classifier. When the tuning parameter C is large, then the margin is\\nwide, many observations violate the margin, and so there are many support\\nvectors. In this case, many observations are involved in determining the\\nhyperplane. The top left panel in Figure 9.7 illustrates this setting: this\\nclassifier has low variance (since many observations are support vectors)\\nbut potentially high bias. In contrast, if C is small, then there will be fewer\\nsupport vectors and hence the resulting classifier will have low bias but\\nhigh variance. The bottom right panel in Figure 9.7 illustrates this setting,\\nwith only eight support vectors.\\nThe fact that the support vector classifier’s decision rule is based only\\non a potentially small subset of the training observations (the support vec-\\ntors) means that it is quite robust to the behavior of observations that\\nare far away from the hyperplane. This property is distinct from some of 9.3 Support Vector Machines\\n377\\n−1\\n0\\n1\\n2\\n−3\\n−2\\n−1\\n0\\n1\\n2\\n3\\n−1\\n0\\n1\\n2\\n−3\\n−2\\n−1\\n0\\n1\\n2\\n3\\n−1\\n0\\n1\\n2\\n−3\\n−2\\n−1\\n0\\n1\\n2\\n3\\n−1\\n0\\n1\\n2\\n−3\\n−2\\n−1\\n0\\n1\\n2\\n3\\nX1\\nX1\\nX1\\nX1\\nX2\\nX2\\nX2\\nX2\\nFIGURE 9.7.\\nA support vector classifier was fit using four different values\\nof the tuning parameter C in (9.12)–(9.15). The largest value of C was used\\nin the top left panel, and smaller values were used in the top right, bottom left,\\nand bottom right panels. When C is large, then there is a high tolerance for\\nobservations being on the wrong side of the margin, and so the margin will be\\nlarge. As C decreases, the tolerance for observations being on the wrong side of\\nthe margin decreases, and the margin narrows.\\nthe other classification methods that we have seen in preceding chapters,\\nsuch as linear discriminant analysis. Recall that the LDA classification rule\\ndepends on the mean of all of the observations within each class, as well as\\nthe within-class covariance matrix computed using all of the observations.\\nIn contrast, logistic regression, unlike LDA, has very low sensitivity to ob-\\nservations far from the decision boundary. In fact we will see in Section 9.5\\nthat the support vector classifier and logistic regression are closely related.\\n9.3\\nSupport Vector Machines\\nWe first discuss a general mechanism for converting a linear classifier into\\none that produces non-linear decision boundaries. We then introduce the\\nsupport vector machine, which does this in an automatic way. 378\\n9. Support Vector Machines\\n−4\\n−2\\n0\\n2\\n4\\n−4\\n−2\\n0\\n2\\n4\\n−4\\n−2\\n0\\n2\\n4\\n−4\\n−2\\n0\\n2\\n4\\nX1\\nX1\\nX2\\nX2\\nFIGURE 9.8.\\nLeft: The observations fall into two classes, with a non-lin-\\near boundary between them. Right: The support vector classifier seeks a linear\\nboundary, and consequently performs very poorly.\\n9.3.1\\nClassification with Non-Linear Decision Boundaries\\nThe support vector classifier is a natural approach for classification in the\\ntwo-class setting, if the boundary between the two classes is linear. How-\\never, in practice we are sometimes faced with non-linear class boundaries.\\nFor instance, consider the data in the left-hand panel of Figure 9.8. It is\\nclear that a support vector classifier or any linear classifier will perform\\npoorly here. Indeed, the support vector classifier shown in the right-hand\\npanel of Figure 9.8 is useless here.\\nIn Chapter 7, we are faced with an analogous situation. We see there\\nthat the performance of linear regression can suffer when there is a non-\\nlinear relationship between the predictors and the outcome. In that case,\\nwe consider enlarging the feature space using functions of the predictors,\\nsuch as quadratic and cubic terms, in order to address this non-linearity.\\nIn the case of the support vector classifier, we could address the prob-\\nlem of possibly non-linear boundaries between classes in a similar way, by\\nenlarging the feature space using quadratic, cubic, and even higher-order\\npolynomial functions of the predictors. For instance, rather than fitting a\\nsupport vector classifier using p features\\nX1, X2, . . . , Xp,\\nwe could instead fit a support vector classifier using 2p features\\nX1, X2\\n1, X2, X2\\n2, . . . , Xp, X2\\np. 9.3 Support Vector Machines\\n379\\nThen (9.12)–(9.15) would become\\nmaximize\\nβ0,β11,β12,...,βp1,βp2,ϵ1,...,ϵn, M M\\n(9.16)\\nsubject to yi\\n\\uf8eb\\n\\uf8edβ0 +\\np\\n0\\nj=1\\nβj1xij +\\np\\n0\\nj=1\\nβj2x2\\nij\\n\\uf8f6\\n\\uf8f8≥M(1 −ϵi),\\nn\\n0\\ni=1\\nϵi ≤C, ϵi ≥0,\\np\\n0\\nj=1\\n2\\n0\\nk=1\\nβ2\\njk = 1.\\nWhy does this lead to a non-linear decision boundary? In the enlarged\\nfeature space, the decision boundary that results from (9.16) is in fact lin-\\near. But in the original feature space, the decision boundary is of the form\\nq(x) = 0, where q is a quadratic polynomial, and its solutions are gener-\\nally non-linear. One might additionally want to enlarge the feature space\\nwith higher-order polynomial terms, or with interaction terms of the form\\nXjXj′ for j ̸= j′. Alternatively, other functions of the predictors could\\nbe considered rather than polynomials. It is not hard to see that there\\nare many possible ways to enlarge the feature space, and that unless we\\nare careful, we could end up with a huge number of features. Then compu-\\ntations would become unmanageable. The support vector machine, which\\nwe present next, allows us to enlarge the feature space used by the support\\nvector classifier in a way that leads to efficient computations.\\n9.3.2\\nThe Support Vector Machine\\nThe support vector machine (SVM) is an extension of the support vector support\\nvector\\nmachine\\nclassifier that results from enlarging the feature space in a specific way,\\nusing kernels. We will now discuss this extension, the details of which are\\nkernel\\nsomewhat complex and beyond the scope of this book. However, the main\\nidea is described in Section 9.3.1: we may want to enlarge our feature space\\nin order to accommodate a non-linear boundary between the classes. The\\nkernel approach that we describe here is simply an efficient computational\\napproach for enacting this idea.\\nWe have not discussed exactly how the support vector classifier is com-\\nputed because the details become somewhat technical. However, it turns\\nout that the solution to the support vector classifier problem (9.12)–(9.15)\\ninvolves only the inner products of the observations (as opposed to the\\nobservations themselves). The inner product of two r-vectors a and b is\\ndefined as ⟨a, b⟩= )r\\ni=1 aibi. Thus the inner product of two observations\\nxi, xi′ is given by\\n⟨xi, xi′⟩=\\np\\n0\\nj=1\\nxijxi′j.\\n(9.17)\\nIt can be shown that\\n• The linear support vector classifier can be represented as\\nf(x) = β0 +\\nn\\n0\\ni=1\\nαi⟨x, xi⟩,\\n(9.18) 380\\n9. Support Vector Machines\\nwhere there are n parameters αi, i = 1, . . . , n, one per training\\nobservation.\\n• To estimate the parameters α1, . . . , αn and β0, all we need are the\\n\\'n\\n2\\n(\\ninner products ⟨xi, xi′⟩between all pairs of training observations.\\n(The notation\\n\\'n\\n2\\n(\\nmeans n(n −1)/2, and gives the number of pairs\\namong a set of n items.)\\nNotice that in (9.18), in order to evaluate the function f(x), we need to\\ncompute the inner product between the new point x and each of the training\\npoints xi. However, it turns out that αi is nonzero only for the support\\nvectors in the solution—that is, if a training observation is not a support\\nvector, then its αi equals zero. So if S is the collection of indices of these\\nsupport points, we can rewrite any solution function of the form (9.18) as\\nf(x) = β0 +\\n0\\ni∈S\\nαi⟨x, xi⟩,\\n(9.19)\\nwhich typically involves far fewer terms than in (9.18).2\\nTo summarize, in representing the linear classifier f(x), and in computing\\nits coefficients, all we need are inner products.\\nNow suppose that every time the inner product (9.17) appears in the\\nrepresentation (9.18), or in a calculation of the solution for the support\\nvector classifier, we replace it with a generalization of the inner product of\\nthe form\\nK(xi, xi′),\\n(9.20)\\nwhere K is some function that we will refer to as a kernel. A kernel is a kernel\\nfunction that quantifies the similarity of two observations. For instance, we\\ncould simply take\\nK(xi, xi′) =\\np\\n0\\nj=1\\nxijxi′j,\\n(9.21)\\nwhich would just give us back the support vector classifier. Equation 9.21\\nis known as a linear kernel because the support vector classifier is linear\\nin the features; the linear kernel essentially quantifies the similarity of a\\npair of observations using Pearson (standard) correlation. But one could\\ninstead choose another form for (9.20). For instance, one could replace\\nevery instance of )p\\nj=1 xijxi′j with the quantity\\nK(xi, xi′) = (1 +\\np\\n0\\nj=1\\nxijxi′j)d.\\n(9.22)\\nThis is known as a polynomial kernel of degree d, where d is a positive polynomial\\nkernel\\ninteger. Using such a kernel with d > 1, instead of the standard linear\\nkernel (9.21), in the support vector classifier algorithm leads to a much more\\nflexible decision boundary. It essentially amounts to fitting a support vector\\n2By expanding each of the inner products in (9.19), it is easy to see that f(x) is\\na linear function of the coordinates of x. Doing so also establishes the correspondence\\nbetween the αi and the original parameters βj. 9.3 Support Vector Machines\\n381\\n−4\\n−2\\n0\\n2\\n4\\n−4\\n−2\\n0\\n2\\n4\\n  \\n  \\n  \\n  \\n  \\n  \\n−4\\n−2\\n0\\n2\\n4\\n−4\\n−2\\n0\\n2\\n4\\n  \\n  \\n  \\n  \\nX1\\nX1\\nX2\\nX2\\nFIGURE 9.9. Left: An SVM with a polynomial kernel of degree 3 is applied to\\nthe non-linear data from Figure 9.8, resulting in a far more appropriate decision\\nrule. Right: An SVM with a radial kernel is applied. In this example, either kernel\\nis capable of capturing the decision boundary.\\nclassifier in a higher-dimensional space involving polynomials of degree d,\\nrather than in the original feature space. When the support vector classifier\\nis combined with a non-linear kernel such as (9.22), the resulting classifier is\\nknown as a support vector machine. Note that in this case the (non-linear)\\nfunction has the form\\nf(x) = β0 +\\n0\\ni∈S\\nαiK(x, xi).\\n(9.23)\\nThe left-hand panel of Figure 9.9 shows an example of an SVM with a\\npolynomial kernel applied to the non-linear data from Figure 9.8. The fit is\\na substantial improvement over the linear support vector classifier. When\\nd = 1, then the SVM reduces to the support vector classifier seen earlier in\\nthis chapter.\\nThe polynomial kernel shown in (9.22) is one example of a possible\\nnon-linear kernel, but alternatives abound. Another popular choice is the\\nradial kernel, which takes the form\\nradial kernel\\nK(xi, xi′) = exp(−γ\\np\\n0\\nj=1\\n(xij −xi′j)2).\\n(9.24)\\nIn (9.24), γ is a positive constant. The right-hand panel of Figure 9.9 shows\\nan example of an SVM with a radial kernel on this non-linear data; it also\\ndoes a good job in separating the two classes.\\nHow does the radial kernel (9.24) actually work? If a given test obser-\\nvation x∗= (x∗\\n1, . . . , x∗\\np)T is far from a training observation xi in terms of\\nEuclidean distance, then )p\\nj=1(x∗\\nj −xij)2 will be large, and so K(x∗, xi) =\\nexp(−γ )p\\nj=1(x∗\\nj −xij)2) will be tiny. This means that in (9.23), xi will\\nplay virtually no role in f(x∗). Recall that the predicted class label for the\\ntest observation x∗is based on the sign of f(x∗). In other words, training\\nobservations that are far from x∗will play essentially no role in the pre-\\ndicted class label for x∗. This means that the radial kernel has very local 382\\n9. Support Vector Machines\\nFalse positive rate\\nTrue positive rate\\n0.0\\n0.2\\n0.4\\n0.6\\n0.8\\n1.0\\n0.0\\n0.2\\n0.4\\n0.6\\n0.8\\n1.0\\nSupport Vector Classifier\\nLDA\\nFalse positive rate\\nTrue positive rate\\n0.0\\n0.2\\n0.4\\n0.6\\n0.8\\n1.0\\n0.0\\n0.2\\n0.4\\n0.6\\n0.8\\n1.0\\nSupport Vector Classifier\\nSVM: γ=10−3\\nSVM: γ=10−2\\nSVM: γ=10−1\\nFIGURE 9.10. ROC curves for the Heart data training set. Left: The support\\nvector classifier and LDA are compared. Right: The support vector classifier is\\ncompared to an SVM using a radial basis kernel with γ = 10−3, 10−2, and 10−1.\\nbehavior, in the sense that only nearby training observations have an effect\\non the class label of a test observation.\\nWhat is the advantage of using a kernel rather than simply enlarging\\nthe feature space using functions of the original features, as in (9.16)? One\\nadvantage is computational, and it amounts to the fact that using kernels,\\none need only compute K(xi, x′\\ni) for all\\n\\'n\\n2\\n(\\ndistinct pairs i, i′. This can be\\ndone without explicitly working in the enlarged feature space. This is im-\\nportant because in many applications of SVMs, the enlarged feature space\\nis so large that computations are intractable. For some kernels, such as the\\nradial kernel (9.24), the feature space is implicit and infinite-dimensional,\\nso we could never do the computations there anyway!\\n9.3.3\\nAn Application to the Heart Disease Data\\nIn Chapter 8 we apply decision trees and related methods to the Heart data.\\nThe aim is to use 13 predictors such as Age, Sex, and Chol in order to predict\\nwhether an individual has heart disease. We now investigate how an SVM\\ncompares to LDA on this data. After removing 6 missing observations, the\\ndata consist of 297 subjects, which we randomly split into 207 training and\\n90 test observations.\\nWe first fit LDA and the support vector classifier to the training data.\\nNote that the support vector classifier is equivalent to an SVM using a poly-\\nnomial kernel of degree d = 1. The left-hand panel of Figure 9.10 displays\\nROC curves (described in Section 4.4.2) for the training set predictions for\\nboth LDA and the support vector classifier. Both classifiers compute scores\\nof the form ˆf(X) = ˆβ0 + ˆβ1X1 + ˆβ2X2 + · · · + ˆβpXp for each observation.\\nFor any given cutoff t, we classify observations into the heart disease or\\nno heart disease categories depending on whether ˆf(X) < t or ˆf(X) ≥t.\\nThe ROC curve is obtained by forming these predictions and computing\\nthe false positive and true positive rates for a range of values of t. An opti-\\nmal classifier will hug the top left corner of the ROC plot. In this instance 9.4 SVMs with More than Two Classes\\n383\\nFalse positive rate\\nTrue positive rate\\n0.0\\n0.2\\n0.4\\n0.6\\n0.8\\n1.0\\n0.0\\n0.2\\n0.4\\n0.6\\n0.8\\n1.0\\nSupport Vector Classifier\\nLDA\\nFalse positive rate\\nTrue positive rate\\n0.0\\n0.2\\n0.4\\n0.6\\n0.8\\n1.0\\n0.0\\n0.2\\n0.4\\n0.6\\n0.8\\n1.0\\nSupport Vector Classifier\\nSVM: γ=10−3\\nSVM: γ=10−2\\nSVM: γ=10−1\\nFIGURE 9.11. ROC curves for the test set of the Heart data. Left: The support\\nvector classifier and LDA are compared. Right: The support vector classifier is\\ncompared to an SVM using a radial basis kernel with γ = 10−3, 10−2, and 10−1.\\nLDA and the support vector classifier both perform well, though there is a\\nsuggestion that the support vector classifier may be slightly superior.\\nThe right-hand panel of Figure 9.10 displays ROC curves for SVMs using\\na radial kernel, with various values of γ. As γ increases and the fit becomes\\nmore non-linear, the ROC curves improve. Using γ = 10−1 appears to give\\nan almost perfect ROC curve. However, these curves represent training\\nerror rates, which can be misleading in terms of performance on new test\\ndata. Figure 9.11 displays ROC curves computed on the 90 test observa-\\ntions. We observe some differences from the training ROC curves. In the\\nleft-hand panel of Figure 9.11, the support vector classifier appears to have\\na small advantage over LDA (although these differences are not statisti-\\ncally significant). In the right-hand panel, the SVM using γ = 10−1, which\\nshowed the best results on the training data, produces the worst estimates\\non the test data. This is once again evidence that while a more flexible\\nmethod will often produce lower training error rates, this does not neces-\\nsarily lead to improved performance on test data. The SVMs with γ = 10−2\\nand γ = 10−3 perform comparably to the support vector classifier, and all\\nthree outperform the SVM with γ = 10−1.\\n9.4\\nSVMs with More than Two Classes\\nSo far, our discussion has been limited to the case of binary classification:\\nthat is, classification in the two-class setting. How can we extend SVMs\\nto the more general case where we have some arbitrary number of classes?\\nIt turns out that the concept of separating hyperplanes upon which SVMs\\nare based does not lend itself naturally to more than two classes. Though\\na number of proposals for extending SVMs to the K-class case have been\\nmade, the two most popular are the one-versus-one and one-versus-all\\napproaches. We briefly discuss those two approaches here. 384\\n9. Support Vector Machines\\n9.4.1\\nOne-Versus-One Classification\\nSuppose that we would like to perform classification using SVMs, and there\\nare K > 2 classes. A one-versus-one or all-pairs approach constructs\\n\\'K\\n2\\n(\\none-versus-\\none\\nSVMs, each of which compares a pair of classes. For example, one such\\nSVM might compare the kth class, coded as +1, to the k′th class, coded\\nas −1. We classify a test observation using each of the\\n\\'K\\n2\\n(\\nclassifiers, and\\nwe tally the number of times that the test observation is assigned to each\\nof the K classes. The final classification is performed by assigning the test\\nobservation to the class to which it was most frequently assigned in these\\n\\'K\\n2\\n(\\npairwise classifications.\\n9.4.2\\nOne-Versus-All Classification\\nThe one-versus-all approach (also referred to as one-versus-rest) is an al- one-versus-\\nall\\none-versus-\\nrest\\nternative procedure for applying SVMs in the case of K > 2 classes. We\\nfit K SVMs, each time comparing one of the K classes to the remaining\\nK −1 classes. Let β0k, β1k, . . . , βpk denote the parameters that result from\\nfitting an SVM comparing the kth class (coded as +1) to the others (coded\\nas −1). Let x∗denote a test observation. We assign the observation to the\\nclass for which β0k +β1kx∗\\n1 +β2kx∗\\n2 +· · ·+βpkx∗\\np is largest, as this amounts\\nto a high level of confidence that the test observation belongs to the kth\\nclass rather than to any of the other classes.\\n9.5\\nRelationship to Logistic Regression\\nWhen SVMs were first introduced in the mid-1990s, they made quite a\\nsplash in the statistical and machine learning communities. This was due\\nin part to their good performance, good marketing, and also to the fact\\nthat the underlying approach seemed both novel and mysterious. The idea\\nof finding a hyperplane that separates the data as well as possible, while al-\\nlowing some violations to this separation, seemed distinctly different from\\nclassical approaches for classification, such as logistic regression and lin-\\near discriminant analysis. Moreover, the idea of using a kernel to expand\\nthe feature space in order to accommodate non-linear class boundaries ap-\\npeared to be a unique and valuable characteristic.\\nHowever, since that time, deep connections between SVMs and other\\nmore classical statistical methods have emerged. It turns out that one can\\nrewrite the criterion (9.12)–(9.15) for fitting the support vector classifier\\nf(X) = β0 + β1X1 + · · · + βpXp as\\nminimize\\nβ0,β1,...,βp\\n\\uf8f1\\n\\uf8f2\\n\\uf8f3\\nn\\n0\\ni=1\\nmax [0, 1 −yif(xi)] + λ\\np\\n0\\nj=1\\nβ2\\nj\\n\\uf8fc\\n\\uf8fd\\n\\uf8fe,\\n(9.25)\\nwhere λ is a nonnegative tuning parameter. When λ is large then β1, . . . , βp\\nare small, more violations to the margin are tolerated, and a low-variance\\nbut high-bias classifier will result. When λ is small then few violations\\nto the margin will occur; this amounts to a high-variance but low-bias 9.5 Relationship to Logistic Regression\\n385\\nclassifier. Thus, a small value of λ in (9.25) amounts to a small value of C\\nin (9.15). Note that the λ )p\\nj=1 β2\\nj term in (9.25) is the ridge penalty term\\nfrom Section 6.2.1, and plays a similar role in controlling the bias-variance\\ntrade-off for the support vector classifier.\\nNow (9.25) takes the “Loss + Penalty” form that we have seen repeatedly\\nthroughout this book:\\nminimize\\nβ0,β1,...,βp {L(X, y, β) + λP(β)} .\\n(9.26)\\nIn (9.26), L(X, y, β) is some loss function quantifying the extent to which\\nthe model, parametrized by β, fits the data (X, y), and P(β) is a penalty\\nfunction on the parameter vector β whose effect is controlled by a nonneg-\\native tuning parameter λ. For instance, ridge regression and the lasso both\\ntake this form with\\nL(X, y, β) =\\nn\\n0\\ni=1\\n\\uf8eb\\n\\uf8edyi −β0 −\\np\\n0\\nj=1\\nxijβj\\n\\uf8f6\\n\\uf8f8\\n2\\nand with P(β) = )p\\nj=1 β2\\nj for ridge regression and P(β) = )p\\nj=1 |βj| for\\nthe lasso. In the case of (9.25) the loss function instead takes the form\\nL(X, y, β) =\\nn\\n0\\ni=1\\nmax [0, 1 −yi(β0 + β1xi1 + · · · + βpxip)] .\\nThis is known as hinge loss, and is depicted in Figure 9.12. However, it hinge loss\\nturns out that the hinge loss function is closely related to the loss function\\nused in logistic regression, also shown in Figure 9.12.\\nAn interesting characteristic of the support vector classifier is that only\\nsupport vectors play a role in the classifier obtained; observations on the\\ncorrect side of the margin do not affect it. This is due to the fact that the\\nloss function shown in Figure 9.12 is exactly zero for observations for which\\nyi(β0 + β1xi1 + · · · + βpxip) ≥1; these correspond to observations that are\\non the correct side of the margin.3 In contrast, the loss function for logistic\\nregression shown in Figure 9.12 is not exactly zero anywhere. But it is very\\nsmall for observations that are far from the decision boundary. Due to the\\nsimilarities between their loss functions, logistic regression and the support\\nvector classifier often give very similar results. When the classes are well\\nseparated, SVMs tend to behave better than logistic regression; in more\\noverlapping regimes, logistic regression is often preferred.\\nWhen the support vector classifier and SVM were first introduced, it was\\nthought that the tuning parameter C in (9.15) was an unimportant “nui-\\nsance” parameter that could be set to some default value, like 1. However,\\nthe “Loss + Penalty” formulation (9.25) for the support vector classifier\\nindicates that this is not the case. The choice of tuning parameter is very\\nimportant and determines the extent to which the model underfits or over-\\nfits the data, as illustrated, for example, in Figure 9.7.\\n3With this hinge-loss + penalty representation, the margin corresponds to the value\\none, and the width of the margin is determined by ! β2\\nj . 386\\n9. Support Vector Machines\\n−6\\n−4\\n−2\\n0\\n2\\n0\\n2\\n4\\n6\\n8\\nLoss\\nSVM Loss\\nLogistic Regression Loss\\nyi(β0 + β1xi1 + . . . + βpxip)\\nFIGURE 9.12. The SVM and logistic regression loss functions are compared,\\nas a function of yi(β0 + β1xi1 + · · · + βpxip). When yi(β0 + β1xi1 + · · · + βpxip) is\\ngreater than 1, then the SVM loss is zero, since this corresponds to an observation\\nthat is on the correct side of the margin. Overall, the two loss functions have quite\\nsimilar behavior.\\nWe have established that the support vector classifier is closely related\\nto logistic regression and other preexisting statistical methods. Is the SVM\\nunique in its use of kernels to enlarge the feature space to accommodate\\nnon-linear class boundaries? The answer to this question is “no”. We could\\njust as well perform logistic regression or many of the other classification\\nmethods seen in this book using non-linear kernels; this is closely related\\nto some of the non-linear approaches seen in Chapter 7. However, for his-\\ntorical reasons, the use of non-linear kernels is much more widespread in\\nthe context of SVMs than in the context of logistic regression or other\\nmethods.\\nThough we have not addressed it here, there is in fact an extension\\nof the SVM for regression (i.e. for a quantitative rather than a qualita-\\ntive response), called support vector regression. In Chapter 3, we saw that support\\nvector\\nregression\\nleast squares regression seeks coefficients β0, β1, . . . , βp such that the sum\\nof squared residuals is as small as possible. (Recall from Chapter 3 that\\nresiduals are defined as yi −β0 −β1xi1 −· · · −βpxip.) Support vector\\nregression instead seeks coefficients that minimize a different type of loss,\\nwhere only residuals larger in absolute value than some positive constant\\ncontribute to the loss function. This is an extension of the margin used in\\nsupport vector classifiers to the regression setting. 9.6 Lab: Support Vector Machines\\n387\\n9.6\\nLab: Support Vector Machines\\nIn this lab, we use the sklearn.svm library to demonstrate the support\\nvector classifier and the support vector machine.\\nWe import some of our usual libraries.\\nIn [1]: import numpy as np\\nfrom matplotlib.pyplot import subplots, cm\\nimport sklearn.model_selection as skm\\nfrom ISLP import load_data, confusion_table\\nWe also collect the new imports needed for this lab.\\nIn [2]: from sklearn.svm import SVC\\nfrom ISLP.svm import plot as plot_svm\\nfrom sklearn.metrics import RocCurveDisplay\\nWe will use the function RocCurveDisplay.from_estimator() to produce RocCurve\\nDisplay.from_\\nestimator()\\nseveral ROC plots, using a shorthand roc_curve.\\nIn [3]: roc_curve = RocCurveDisplay.from_estimator # shorthand\\n9.6.1\\nSupport Vector Classifier\\nWe now use the SupportVectorClassifier() function (abbreviated SVC()) SupportVector\\nClassifier()\\nfrom sklearn to fit the support vector classifier for a given value of the\\nparameter C. The C argument allows us to specify the cost of a violation\\nto the margin. When the cost argument is small, then the margins will be\\nwide and many support vectors will be on the margin or will violate the\\nmargin. When the C argument is large, then the margins will be narrow and\\nthere will be few support vectors on the margin or violating the margin.\\nHere we demonstrate the use of SVC() on a two-dimensional example, so\\nthat we can plot the resulting decision boundary. We begin by generating\\nthe observations, which belong to two classes, and checking whether the\\nclasses are linearly separable.\\nIn [4]: rng = np.random.default_rng(1)\\nX = rng.standard_normal((50, 2))\\ny = np.array([-1]*25+[1]*25)\\nX[y==1] += 1\\nfig, ax = subplots(figsize=(8,8))\\nax.scatter(X[:,0],\\nX[:,1],\\nc=y,\\ncmap=cm.coolwarm);\\nThey are not. We now fit the classifier.\\nIn [5]: svm_linear = SVC(C=10, kernel=\\'linear\\')\\nsvm_linear.fit(X, y)\\nOut[5]: SVC(C=10, kernel=\\'linear\\')\\nThe support vector classifier with two features can be visualized by plot-\\nting values of its decision function. We have included a function for this in decision\\nfunction\\nthe ISLP package (inspired by a similar example in the sklearn docs). 388\\n9. Support Vector Machines\\nIn [6]: fig, ax = subplots(figsize=(8,8))\\nplot_svm(X,\\ny,\\nsvm_linear,\\nax=ax)\\nThe decision boundary between the two classes is linear (because we\\nused the argument kernel=\\'linear\\'). The support vectors are marked with\\n+ and the remaining observations are plotted as circles.\\nWhat if we instead used a smaller value of the cost parameter?\\nIn [7]: svm_linear_small = SVC(C=0.1, kernel=\\'linear\\')\\nsvm_linear_small.fit(X, y)\\nfig, ax = subplots(figsize=(8,8))\\nplot_svm(X,\\ny,\\nsvm_linear_small ,\\nax=ax)\\nWith a smaller value of the cost parameter, we obtain a larger number of\\nsupport vectors, because the margin is now wider. For linear kernels, we\\ncan extract the coefficients of the linear decision boundary as follows:\\nIn [8]: svm_linear.coef_\\nOut[8]: array([[1.173\\n, 0.7734]])\\nSince the support vector machine is an estimator in sklearn, we can use\\nthe usual machinery to tune it.\\nIn [9]: kfold = skm.KFold(5,\\nrandom_state=0,\\nshuffle=True)\\ngrid = skm.GridSearchCV(svm_linear,\\n{\\'C\\':[0.001,0.01,0.1,1,5,10,100]},\\nrefit=True,\\ncv=kfold,\\nscoring=\\'accuracy\\')\\ngrid.fit(X, y)\\ngrid.best_params_\\nOut[9]: {\\'C\\': 1}\\nWe can easily access the cross-validation errors for each of these models in\\ngrid.cv_results_. This prints out a lot of detail, so we extract the accuracy\\nresults only.\\nIn [10]: grid.cv_results_[(\\'mean_test_score\\')]\\nOut[10]: array([0.46, 0.46, 0.72, 0.74, 0.74, 0.74, 0.74])\\nWe see that C=1 results in the highest cross-validation accuracy of 0.74,\\nthough the accuracy is the same for several values of C. The classifier\\ngrid.best_estimator_ can be used to predict the class label on a set of\\ntest observations. Let’s generate a test data set. 9.6 Lab: Support Vector Machines\\n389\\nIn [11]: X_test = rng.standard_normal((20, 2))\\ny_test = np.array([-1]*10+[1]*10)\\nX_test[y_test==1] += 1\\nNow we predict the class labels of these test observations. Here we use the\\nbest model selected by cross-validation in order to make the predictions.\\nIn [12]: best_ = grid.best_estimator_\\ny_test_hat = best_.predict(X_test)\\nconfusion_table(y_test_hat , y_test)\\nOut[12]:\\nTruth\\n-1\\n1\\nPredicted\\n-1\\n8\\n4\\n1\\n2\\n6\\nThus, with this value of C, 70% of the test observations are correctly clas-\\nsified. What if we had instead used C=0.001?\\nIn [13]: svm_ = SVC(C=0.001,\\nkernel=\\'linear\\').fit(X, y)\\ny_test_hat = svm_.predict(X_test)\\nconfusion_table(y_test_hat , y_test)\\nOut[13]:\\nTruth\\n-1\\n1\\nPredicted\\n-1\\n2\\n0\\n1\\n8\\n10\\nIn this case 60% of test observations are correctly classified.\\nWe now consider a situation in which the two classes are linearly sepa-\\nrable. Then we can find an optimal separating hyperplane using the SVC()\\nestimator. We first further separate the two classes in our simulated data\\nso that they are linearly separable:\\nIn [14]: X[y==1] += 1.9;\\nfig, ax = subplots(figsize=(8,8))\\nax.scatter(X[:,0], X[:,1], c=y, cmap=cm.coolwarm);\\nNow the observations are just barely linearly separable.\\nIn [15]: svm_ = SVC(C=1e5, kernel=\\'linear\\').fit(X, y)\\ny_hat = svm_.predict(X)\\nconfusion_table(y_hat, y)\\nOut[15]:\\nTruth\\n-1\\n1\\nPredicted\\n-1\\n25\\n0\\n1\\n0\\n25\\nWe fit the support vector classifier and plot the resulting hyperplane, using\\na very large value of C so that no observations are misclassified.\\nIn [16]: fig, ax = subplots(figsize=(8,8))\\nplot_svm(X,\\ny,\\nsvm_,\\nax=ax) 390\\n9. Support Vector Machines\\nIndeed no training errors were made and only three support vectors were\\nused. In fact, the large value of C also means that these three support points\\nare on the margin, and define it. One may wonder how good the classifier\\ncould be on test data that depends on only three data points! We now try\\na smaller value of C.\\nIn [17]: svm_ = SVC(C=0.1, kernel=\\'linear\\').fit(X, y)\\ny_hat = svm_.predict(X)\\nconfusion_table(y_hat, y)\\nOut[17]:\\nTruth\\n-1\\n1\\nPredicted\\n-1\\n25\\n0\\n1\\n0\\n25\\nUsing C=0.1, we again do not misclassify any training observations, but we\\nalso obtain a much wider margin and make use of twelve support vectors.\\nThese jointly define the orientation of the decision boundary, and since\\nthere are more of them, it is more stable. It seems possible that this model\\nwill perform better on test data than the model with C=1e5 (and indeed, a\\nsimple experiment with a large test set would bear this out).\\nIn [18]: fig, ax = subplots(figsize=(8,8))\\nplot_svm(X,\\ny,\\nsvm_,\\nax=ax)\\n9.6.2\\nSupport Vector Machine\\nIn order to fit an SVM using a non-linear kernel, we once again use the\\nSVC() estimator. However, now we use a different value of the parameter\\nkernel. To fit an SVM with a polynomial kernel we use kernel=\"poly\", and\\nto fit an SVM with a radial kernel we use kernel=\"rbf\". In the former case\\nwe also use the degree argument to specify a degree for the polynomial\\nkernel (this is d in (9.22)), and in the latter case we use gamma to specify a\\nvalue of γ for the radial basis kernel (9.24).\\nWe first generate some data with a non-linear class boundary, as follows:\\nIn [19]: X = rng.standard_normal((200, 2))\\nX[:100] += 2\\nX[100:150] -= 2\\ny = np.array([1]*150+[2]*50)\\nPlotting the data makes it clear that the class boundary is indeed non-\\nlinear.\\nIn [20]: fig, ax = subplots(figsize=(8,8))\\nax.scatter(X[:,0],\\nX[:,1],\\nc=y,\\ncmap=cm.coolwarm)\\nOut[20]: <matplotlib.collections.PathCollection at 0x7faa9ba52eb0 > 9.6 Lab: Support Vector Machines\\n391\\nThe data is randomly split into training and testing groups. We then fit\\nthe training data using the SVC() estimator with a radial kernel and γ = 1:\\nIn [21]: (X_train,\\nX_test,\\ny_train,\\ny_test) = skm.train_test_split(X,\\ny,\\ntest_size=0.5,\\nrandom_state=0)\\nsvm_rbf = SVC(kernel=\"rbf\", gamma=1, C=1)\\nsvm_rbf.fit(X_train, y_train)\\nThe plot shows that the resulting SVM has a decidedly non-linear bound-\\nary.\\nIn [22]: fig, ax = subplots(figsize=(8,8))\\nplot_svm(X_train,\\ny_train,\\nsvm_rbf,\\nax=ax)\\nWe can see from the figure that there are a fair number of training errors\\nin this SVM fit. If we increase the value of C, we can reduce the number\\nof training errors. However, this comes at the price of a more irregular\\ndecision boundary that seems to be at risk of overfitting the data.\\nIn [23]: svm_rbf = SVC(kernel=\"rbf\", gamma=1, C=1e5)\\nsvm_rbf.fit(X_train, y_train)\\nfig, ax = subplots(figsize=(8,8))\\nplot_svm(X_train,\\ny_train,\\nsvm_rbf,\\nax=ax)\\nWe can perform cross-validation using skm.GridSearchCV() to select the\\nbest choice of γ and C for an SVM with a radial kernel:\\nIn [24]: kfold = skm.KFold(5,\\nrandom_state=0,\\nshuffle=True)\\ngrid = skm.GridSearchCV(svm_rbf,\\n{\\'C\\':[0.1,1,10,100,1000],\\n\\'gamma\\':[0.5,1,2,3,4]},\\nrefit=True,\\ncv=kfold,\\nscoring=\\'accuracy\\');\\ngrid.fit(X_train, y_train)\\ngrid.best_params_\\nOut[24]: {\\'C\\': 100, \\'gamma\\': 1}\\nThe best choice of parameters under five-fold CV is achieved at C=1 and\\ngamma=0.5, though several other values also achieve the same value.\\nIn [25]: best_svm = grid.best_estimator_\\nfig, ax = subplots(figsize=(8,8))\\nplot_svm(X_train, 392\\n9. Support Vector Machines\\ny_train,\\nbest_svm,\\nax=ax)\\ny_hat_test = best_svm.predict(X_test)\\nconfusion_table(y_hat_test , y_test)\\nOut[25]:\\nTruth\\n1\\n2\\nPredicted\\n1\\n69\\n6\\n2\\n6\\n19\\nWith these parameters, 12% of test observations are misclassified by this\\nSVM.\\n9.6.3\\nROC Curves\\nSVMs and support vector classifiers output class labels for each observation.\\nHowever, it is also possible to obtain fitted values for each observation,\\nwhich are the numerical scores used to obtain the class labels. For instance,\\nin the case of a support vector classifier, the fitted value for an observation\\nX = (X1, X2, . . . , Xp)T takes the form ˆβ0 + ˆβ1X1 + ˆβ2X2 +. . . + ˆβpXp. For\\nan SVM with a non-linear kernel, the equation that yields the fitted value\\nis given in (9.23). The sign of the fitted value determines on which side\\nof the decision boundary the observation lies. Therefore, the relationship\\nbetween the fitted value and the class prediction for a given observation\\nis simple: if the fitted value exceeds zero then the observation is assigned\\nto one class, and if it is less than zero then it is assigned to the other.\\nBy changing this threshold from zero to some positive value, we skew the\\nclassifications in favor of one class versus the other. By considering a range\\nof these thresholds, positive and negative, we produce the ingredients for a\\nROC plot. We can access these values by calling the decision_function() .function_\\ndecision()\\nmethod of a fitted SVM estimator.\\nThe function ROCCurveDisplay.from_estimator() (which we have abbre-\\nviated to roc_curve()) will produce a plot of a ROC curve. It takes a fitted roc_curve()\\nestimator as its first argument, followed by a model matrix X and labels y.\\nThe argument name is used in the legend, while color is used for the color\\nof the line. Results are plotted on our axis object ax.\\nIn [26]: fig, ax = subplots(figsize=(8,8))\\nroc_curve(best_svm,\\nX_train,\\ny_train,\\nname=\\'Training\\',\\ncolor=\\'r\\',\\nax=ax);\\nIn this example, the SVM appears to provide accurate predictions. By\\nincreasing γ we can produce a more flexible fit and generate further im-\\nprovements in accuracy.\\nIn [27]: svm_flex = SVC(kernel=\"rbf\",\\ngamma=50, 9.6 Lab: Support Vector Machines\\n393\\nC=1)\\nsvm_flex.fit(X_train, y_train)\\nfig, ax = subplots(figsize=(8,8))\\nroc_curve(svm_flex,\\nX_train,\\ny_train,\\nname=\\'Training $\\\\gamma=50$\\',\\ncolor=\\'r\\',\\nax=ax);\\nHowever, these ROC curves are all on the training data. We are really\\nmore interested in the level of prediction accuracy on the test data. When\\nwe compute the ROC curves on the test data, the model with γ = 0.5\\nappears to provide the most accurate results.\\nIn [28]: roc_curve(svm_flex,\\nX_test,\\ny_test,\\nname=\\'Test $\\\\gamma=50$\\',\\ncolor=\\'b\\',\\nax=ax)\\nfig;\\nLet’s look at our tuned SVM.\\nIn [29]: fig, ax = subplots(figsize=(8,8))\\nfor (X_, y_, c, name) in zip(\\n(X_train, X_test),\\n(y_train, y_test),\\n(\\'r\\', \\'b\\'),\\n(\\'CV tuned on training\\',\\n\\'CV tuned on test\\')):\\nroc_curve(best_svm,\\nX_,\\ny_,\\nname=name,\\nax=ax,\\ncolor=c)\\n9.6.4\\nSVM with Multiple Classes\\nIf the response is a factor containing more than two levels, then the SVC()\\nfunction will perform multi-class classification using either the one-versus-\\none approach (when decision_function_shape==\\'ovo\\') or one-versus-rest4\\n(when decision_function_shape==\\'ovr\\'). We explore that setting briefly\\nhere by generating a third class of observations.\\nIn [30]: rng = np.random.default_rng(123)\\nX = np.vstack([X, rng.standard_normal((50, 2))])\\ny = np.hstack([y, [0]*50])\\nX[y==0,1] += 2\\nfig, ax = subplots(figsize=(8,8))\\nax.scatter(X[:,0], X[:,1], c=y, cmap=cm.coolwarm);\\n4One-versus-rest is also known as one-versus-all. 394\\n9. Support Vector Machines\\nWe now fit an SVM to the data:\\nIn [31]: svm_rbf_3 = SVC(kernel=\"rbf\",\\nC=10,\\ngamma=1,\\ndecision_function_shape=\\'ovo\\');\\nsvm_rbf_3.fit(X, y)\\nfig, ax = subplots(figsize=(8,8))\\nplot_svm(X,\\ny,\\nsvm_rbf_3,\\nscatter_cmap=cm.tab10,\\nax=ax)\\nThe sklearn.svm library can also be used to perform support vector re-\\ngression with a numerical response using the estimator SupportVector-\\nRegression().\\nSupportVector\\nRegression()\\n9.6.5\\nApplication to Gene Expression Data\\nWe now examine the Khan data set, which consists of a number of tissue\\nsamples corresponding to four distinct types of small round blue cell tu-\\nmors. For each tissue sample, gene expression measurements are available.\\nThe data set consists of training data, xtrain and ytrain, and testing data,\\nxtest and ytest.\\nWe examine the dimension of the data:\\nIn [32]: Khan = load_data(\\'Khan\\')\\nKhan[\\'xtrain\\'].shape, Khan[\\'xtest\\'].shape\\nOut[32]: ((63, 2308), (20, 2308))\\nThis data set consists of expression measurements for 2,308 genes. The\\ntraining and test sets consist of 63 and 20 observations, respectively.\\nWe will use a support vector approach to predict cancer subtype using\\ngene expression measurements. In this data set, there is a very large number\\nof features relative to the number of observations. This suggests that we\\nshould use a linear kernel, because the additional flexibility that will result\\nfrom using a polynomial or radial kernel is unnecessary.\\nIn [33]: khan_linear = SVC(kernel=\\'linear\\', C=10)\\nkhan_linear.fit(Khan[\\'xtrain\\'], Khan[\\'ytrain\\'])\\nconfusion_table(khan_linear.predict(Khan[\\'xtrain\\']),\\nKhan[\\'ytrain\\'])\\nOut[33]:\\nTruth\\n1\\n2\\n3\\n4\\nPredicted\\n1\\n8\\n0\\n0\\n0\\n2\\n0\\n23\\n0\\n0\\n3\\n0\\n0\\n12\\n0\\n4\\n0\\n0\\n0\\n20\\nWe see that there are no training errors. In fact, this is not surprising,\\nbecause the large number of variables relative to the number of observations\\nimplies that it is easy to find hyperplanes that fully separate the classes. 9.7 Exercises\\n395\\nWe are more interested in the support vector classifier’s performance on\\nthe test observations.\\nIn [34]: confusion_table(khan_linear.predict(Khan[\\'xtest\\']),\\nKhan[\\'ytest\\'])\\nOut[34]:\\nTruth\\n1\\n2\\n3\\n4\\nPredicted\\n1\\n3\\n0\\n0\\n0\\n2\\n0\\n6\\n2\\n0\\n3\\n0\\n0\\n4\\n0\\n4\\n0\\n0\\n0\\n5\\nWe see that using C=10 yields two test set errors on these data.\\n9.7\\nExercises\\nConceptual\\n1. This problem involves hyperplanes in two dimensions.\\n(a) Sketch the hyperplane 1 + 3X1 −X2 = 0. Indicate the set of\\npoints for which 1 + 3X1 −X2 > 0, as well as the set of points\\nfor which 1 + 3X1 −X2 < 0.\\n(b) On the same plot, sketch the hyperplane −2 + X1 + 2X2 = 0.\\nIndicate the set of points for which −2 + X1 + 2X2 > 0, as well\\nas the set of points for which −2 + X1 + 2X2 < 0.\\n2. We have seen that in p = 2 dimensions, a linear decision boundary\\ntakes the form β0+β1X1+β2X2 = 0. We now investigate a non-linear\\ndecision boundary.\\n(a) Sketch the curve\\n(1 + X1)2 + (2 −X2)2 = 4.\\n(b) On your sketch, indicate the set of points for which\\n(1 + X1)2 + (2 −X2)2 > 4,\\nas well as the set of points for which\\n(1 + X1)2 + (2 −X2)2 ≤4.\\n(c) Suppose that a classifier assigns an observation to the blue class\\nif\\n(1 + X1)2 + (2 −X2)2 > 4,\\nand to the red class otherwise. To what class is the observation\\n(0, 0) classified? (−1, 1)? (2, 2)? (3, 8)?\\n(d) Argue that while the decision boundary in (c) is not linear in\\nterms of X1 and X2, it is linear in terms of X1, X2\\n1, X2, and\\nX2\\n2. 396\\n9. Support Vector Machines\\n3. Here we explore the maximal margin classifier on a toy data set.\\n(a) We are given n = 7 observations in p = 2 dimensions. For each\\nobservation, there is an associated class label.\\nObs.\\nX1\\nX2\\nY\\n1\\n3\\n4\\nRed\\n2\\n2\\n2\\nRed\\n3\\n4\\n4\\nRed\\n4\\n1\\n4\\nRed\\n5\\n2\\n1\\nBlue\\n6\\n4\\n3\\nBlue\\n7\\n4\\n1\\nBlue\\nSketch the observations.\\n(b) Sketch the optimal separating hyperplane, and provide the equa-\\ntion for this hyperplane (of the form (9.1)).\\n(c) Describe the classification rule for the maximal margin classifier.\\nIt should be something along the lines of “Classify to Red if\\nβ0 + β1X1 + β2X2 > 0, and classify to Blue otherwise.” Provide\\nthe values for β0, β1, and β2.\\n(d) On your sketch, indicate the margin for the maximal margin\\nhyperplane.\\n(e) Indicate the support vectors for the maximal margin classifier.\\n(f) Argue that a slight movement of the seventh observation would\\nnot affect the maximal margin hyperplane.\\n(g) Sketch a hyperplane that is not the optimal separating hyper-\\nplane, and provide the equation for this hyperplane.\\n(h) Draw an additional observation on the plot so that the two\\nclasses are no longer separable by a hyperplane.\\nApplied\\n4. Generate a simulated two-class data set with 100 observations and\\ntwo features in which there is a visible but non-linear separation be-\\ntween the two classes. Show that in this setting, a support vector\\nmachine with a polynomial kernel (with degree greater than 1) or a\\nradial kernel will outperform a support vector classifier on the train-\\ning data. Which technique performs best on the test data? Make\\nplots and report training and test error rates in order to back up\\nyour assertions.\\n5. We have seen that we can fit an SVM with a non-linear kernel in order\\nto perform classification using a non-linear decision boundary. We will\\nnow see that we can also obtain a non-linear decision boundary by\\nperforming logistic regression using non-linear transformations of the\\nfeatures. 9.7 Exercises\\n397\\n(a) Generate a data set with n = 500 and p = 2, such that the obser-\\nvations belong to two classes with a quadratic decision boundary\\nbetween them. For instance, you can do this as follows:\\nrng = np.random.default_rng(5)\\nx1 = rng.uniform(size=500) - 0.5\\nx2 = rng.uniform(size=500) - 0.5\\ny = x1**2 - x2**2 > 0\\n(b) Plot the observations, colored according to their class labels.\\nYour plot should display X1 on the x-axis, and X2 on the y-\\naxis.\\n(c) Fit a logistic regression model to the data, using X1 and X2 as\\npredictors.\\n(d) Apply this model to the training data in order to obtain a pre-\\ndicted class label for each training observation. Plot the ob-\\nservations, colored according to the predicted class labels. The\\ndecision boundary should be linear.\\n(e) Now fit a logistic regression model to the data using non-linear\\nfunctions of X1 and X2 as predictors (e.g. X2\\n1, X1×X2, log(X2),\\nand so forth).\\n(f) Apply this model to the training data in order to obtain a pre-\\ndicted class label for each training observation. Plot the ob-\\nservations, colored according to the predicted class labels. The\\ndecision boundary should be obviously non-linear. If it is not,\\nthen repeat (a)–(e) until you come up with an example in which\\nthe predicted class labels are obviously non-linear.\\n(g) Fit a support vector classifier to the data with X1 and X2 as\\npredictors. Obtain a class prediction for each training observa-\\ntion. Plot the observations, colored according to the predicted\\nclass labels.\\n(h) Fit a SVM using a non-linear kernel to the data. Obtain a class\\nprediction for each training observation. Plot the observations,\\ncolored according to the predicted class labels.\\n(i) Comment on your results.\\n6. At the end of Section 9.6.1, it is claimed that in the case of data that is\\njust barely linearly separable, a support vector classifier with a small\\nvalue of C that misclassifies a couple of training observations may\\nperform better on test data than one with a huge value of C that does\\nnot misclassify any training observations. You will now investigate\\nthis claim.\\n(a) Generate two-class data with p = 2 in such a way that the classes\\nare just barely linearly separable.\\n(b) Compute the cross-validation error rates for support vector\\nclassifiers with a range of C values. How many training obser-\\nvations are misclassified for each value of C considered, and how\\ndoes this relate to the cross-validation errors obtained? 398\\n9. Support Vector Machines\\n(c) Generate an appropriate test data set, and compute the test\\nerrors corresponding to each of the values of C considered. Which\\nvalue of C leads to the fewest test errors, and how does this\\ncompare to the values of C that yield the fewest training errors\\nand the fewest cross-validation errors?\\n(d) Discuss your results.\\n7. In this problem, you will use support vector approaches in order to\\npredict whether a given car gets high or low gas mileage based on the\\nAuto data set.\\n(a) Create a binary variable that takes on a 1 for cars with gas\\nmileage above the median, and a 0 for cars with gas mileage\\nbelow the median.\\n(b) Fit a support vector classifier to the data with various values of\\nC, in order to predict whether a car gets high or low gas mileage.\\nReport the cross-validation errors associated with different val-\\nues of this parameter. Comment on your results. Note you will\\nneed to fit the classifier without the gas mileage variable to pro-\\nduce sensible results.\\n(c) Now repeat (b), this time using SVMs with radial and polyno-\\nmial basis kernels, with different values of gamma and degree and\\nC. Comment on your results.\\n(d) Make some plots to back up your assertions in (b) and (c).\\nHint: In the lab, we used the plot_svm() function for fitted SVMs.\\nWhen p > 2, you can use the keyword argument features to\\ncreate plots displaying pairs of variables at a time.\\n8. This problem involves the OJ data set which is part of the ISLP\\npackage.\\n(a) Create a training set containing a random sample of 800\\nobservations,\\nand\\na\\ntest\\nset\\ncontaining\\nthe\\nremaining\\nobservations.\\n(b) Fit a support vector classifier to the training data using\\nC = 0.01, with Purchase as the response and the other variables\\nas predictors. How many support points are there?\\n(c) What are the training and test error rates?\\n(d) Use cross-validation to select an optimal C. Consider values in\\nthe range 0.01 to 10.\\n(e) Compute the training and test error rates using this new value\\nfor C.\\n(f) Repeat parts (b) through (e) using a support vector machine\\nwith a radial kernel. Use the default value for gamma.\\n(g) Repeat parts (b) through (e) using a support vector machine\\nwith a polynomial kernel. Set degree = 2.\\n(h) Overall, which approach seems to give the best results on this\\ndata? 10\\nDeep Learning\\nThis chapter covers the important topic of deep learning. At the time of deep\\nlearning\\nwriting (2020), deep learning is a very active area of research in the machine\\nlearning and artificial intelligence communities. The cornerstone of deep\\nlearning is the neural network.\\nneural\\nnetwork\\nNeural networks rose to fame in the late 1980s. There was a lot of excite-\\nment and a certain amount of hype associated with this approach, and they\\nwere the impetus for the popular Neural Information Processing Systems\\nmeetings (NeurIPS, formerly NIPS) held every year, typically in exotic\\nplaces like ski resorts. This was followed by a synthesis stage, where the\\nproperties of neural networks were analyzed by machine learners, math-\\nematicians and statisticians; algorithms were improved, and the method-\\nology stabilized. Then along came SVMs, boosting, and random forests,\\nand neural networks fell somewhat from favor. Part of the reason was that\\nneural networks required a lot of tinkering, while the new methods were\\nmore automatic. Also, on many problems the new methods outperformed\\npoorly-trained neural networks. This was the status quo for the first decade\\nin the new millennium.\\nAll the while, though, a core group of neural-network enthusiasts were\\npushing their technology harder on ever-larger computing architectures and\\ndata sets. Neural networks resurfaced after 2010 with the new name deep\\nlearning, with new architectures, additional bells and whistles, and a string\\nof success stories on some niche problems such as image and video classifi-\\ncation, speech and text modeling. Many in the field believe that the major\\nreason for these successes is the availability of ever-larger training datasets,\\nmade possible by the wide-scale use of digitization in science and industry.\\nIn this chapter we discuss the basics of neural networks and deep learn-\\ning, and then go into some of the specializations for specific problems, such\\nas convolutional neural networks (CNNs) for image classification, and re-\\ncurrent neural networks (RNNs) for time series and other sequences. We\\n© Springer Nature Switzerland AG 2023 \\nG. James et al., An Introduction to Statistical Learning, Springer Texts in Statistics, \\nhttps://doi.org/10.1007/978-3-031-38747-0_10 \\n399 400\\n10. Deep Learning\\nX1\\nX2\\nX3\\nX4\\nA1\\nA2\\nA3\\nA4\\nA5\\nf(X)\\nY\\nHidden\\nLayer\\nInput\\nLayer\\nOutput\\nLayer\\nFIGURE 10.1. Neural network with a single hidden layer. The hidden layer\\ncomputes activations Ak = hk(X) that are nonlinear transformations of linear\\ncombinations of the inputs X1, X2, . . . , Xp. Hence these Ak are not directly ob-\\nserved. The functions hk(·) are not fixed in advance, but are learned during the\\ntraining of the network. The output layer is a linear model that uses these acti-\\nvations Ak as inputs, resulting in a function f(X).\\nwill also demonstrate these models using the Python torch package, along\\nwith a number of helper packages.\\nThe material in this chapter is slightly more challenging than elsewhere\\nin this book.\\n10.1\\nSingle Layer Neural Networks\\nA neural network takes an input vector of p variables X = (X1, X2, . . . , Xp)\\nand builds a nonlinear function f(X) to predict the response Y . We have\\nbuilt nonlinear prediction models in earlier chapters, using trees, boosting\\nand generalized additive models. What distinguishes neural networks from\\nthese methods is the particular structure of the model. Figure 10.1 shows\\na simple feed-forward neural network for modeling a quantitative response feed-forward\\nneural\\nnetwork\\nusing p = 4 predictors. In the terminology of neural networks, the four fea-\\ntures X1, . . . , X4 make up the units in the input layer. The arrows indicate\\ninput layer\\nthat each of the inputs from the input layer feeds into each of the K hidden\\nunits (we get to pick K; here we chose 5). The neural network model has hidden units\\nthe form\\nf(X)\\n=\\nβ0 + )K\\nk=1 βkhk(X)\\n=\\nβ0 + )K\\nk=1 βkg(wk0 + )p\\nj=1 wkjXj).\\n(10.1)\\nIt is built up here in two steps. First the K activations Ak, k = 1, . . . , K, in activations\\nthe hidden layer are computed as functions of the input features X1, . . . , Xp,\\nAk = hk(X) = g(wk0 + )p\\nj=1 wkjXj),\\n(10.2) 10.1 Single Layer Neural Networks\\n401\\n−4\\n−2\\n0\\n2\\n4\\n0.0\\n0.2\\n0.4\\n0.6\\n0.8\\n1.0\\nz\\ng(z)\\nsigmoid\\nReLU\\nFIGURE 10.2. Activation functions. The piecewise-linear ReLU function is pop-\\nular for its efficiency and computability. We have scaled it down by a factor of\\nfive for ease of comparison.\\nwhere g(z) is a nonlinear activation function that is specified in advance. activation\\nfunction\\nWe can think of each Ak as a different transformation hk(X) of the original\\nfeatures, much like the basis functions of Chapter 7. These K activations\\nfrom the hidden layer then feed into the output layer, resulting in\\nf(X) = β0 +\\nK\\n0\\nk=1\\nβkAk,\\n(10.3)\\na linear regression model in the K = 5 activations. All the parameters\\nβ0, . . . , βK and w10, . . . , wKp need to be estimated from data. In the early\\ninstances of neural networks, the sigmoid activation function was favored,\\nsigmoid\\ng(z) =\\nez\\n1 + ez =\\n1\\n1 + e−z ,\\n(10.4)\\nwhich is the same function used in logistic regression to convert a linear\\nfunction into probabilities between zero and one (see Figure 10.2). The\\npreferred choice in modern neural networks is the ReLU (rectified linear ReLU\\nunit) activation function, which takes the form\\nrectified\\nlinear unit\\ng(z) = (z)+ =\\nK 0\\nif z < 0\\nz\\notherwise.\\n(10.5)\\nA ReLU activation can be computed and stored more efficiently than a\\nsigmoid activation. Although it thresholds at zero, because we apply it to a\\nlinear function (10.2) the constant term wk0 will shift this inflection point.\\nSo in words, the model depicted in Figure 10.1 derives five new features\\nby computing five different linear combinations of X, and then squashes\\neach through an activation function g(·) to transform it. The final model\\nis linear in these derived variables.\\nThe name neural network originally derived from thinking of these hidden\\nunits as analogous to neurons in the brain — values of the activations\\nAk = hk(X) close to one are firing, while those close to zero are silent\\n(using the sigmoid activation function).\\nThe nonlinearity in the activation function g(·) is essential, since without\\nit the model f(X) in (10.1) would collapse into a simple linear model in 402\\n10. Deep Learning\\nX1, . . . , Xp. Moreover, having a nonlinear activation function allows the\\nmodel to capture complex nonlinearities and interaction effects. Consider\\na very simple example with p = 2 input variables X = (X1, X2), and\\nK = 2 hidden units h1(X) and h2(X) with g(z) = z2. We specify the other\\nparameters as\\nβ0 = 0,\\nβ1 = 1\\n4,\\nβ2 = −1\\n4,\\nw10 = 0,\\nw11 = 1,\\nw12 =\\n1,\\nw20 = 0,\\nw21 = 1,\\nw22 = −1.\\n(10.6)\\nFrom (10.2), this means that\\nh1(X)\\n=\\n(0 + X1 + X2)2,\\nh2(X)\\n=\\n(0 + X1 −X2)2.\\n(10.7)\\nThen plugging (10.7) into (10.1), we get\\nf(X)\\n=\\n0 + 1\\n4 · (0 + X1 + X2)2 −1\\n4 · (0 + X1 −X2)2\\n=\\n1\\n4\\nN\\n(X1 + X2)2 −(X1 −X2)2O\\n=\\nX1X2.\\n(10.8)\\nSo the sum of two nonlinear transformations of linear functions can give\\nus an interaction! In practice we would not use a quadratic function for\\ng(z), since we would always get a second-degree polynomial in the original\\ncoordinates X1, . . . , Xp. The sigmoid or ReLU activations do not have such\\na limitation.\\nFitting a neural network requires estimating the unknown parameters in\\n(10.1). For a quantitative response, typically squared-error loss is used, so\\nthat the parameters are chosen to minimize\\nn\\n0\\ni=1\\n(yi −f(xi))2 .\\n(10.9)\\nDetails about how to perform this minimization are provided in Section 10.7.\\n10.2\\nMultilayer Neural Networks\\nModern neural networks typically have more than one hidden layer, and\\noften many units per layer. In theory a single hidden layer with a large\\nnumber of units has the ability to approximate most functions. However,\\nthe learning task of discovering a good solution is made much easier with\\nmultiple layers each of modest size.\\nWe will illustrate a large dense network on the famous and publicly\\navailable MNIST handwritten digit dataset.1 Figure 10.3 shows examples of\\nthese digits. The idea is to build a model to classify the images into their\\ncorrect digit class 0–9. Every image has p = 28 × 28 = 784 pixels, each\\nof which is an eight-bit grayscale value between 0 and 255 representing\\n1See LeCun, Cortes, and Burges (2010) “The MNIST database of handwritten digits”,\\navailable at http://yann.lecun.com/exdb/mnist. 10.2 Multilayer Neural Networks\\n403\\nFIGURE 10.3. Examples of handwritten digits from the MNIST corpus. Each\\ngrayscale image has 28 × 28 pixels, each of which is an eight-bit number (0–255)\\nwhich represents how dark that pixel is. The first 3, 5, and 8 are enlarged to show\\ntheir 784 individual pixel values.\\nthe relative amount of the written digit in that tiny square.2 These pixels\\nare stored in the input vector X (in, say, column order). The output is\\nthe class label, represented by a vector Y = (Y0, Y1, . . . , Y9) of 10 dummy\\nvariables, with a one in the position corresponding to the label, and zeros\\nelsewhere. In the machine learning community, this is known as one-hot\\nencoding. There are 60,000 training images, and 10,000 test images.\\none-hot\\nencoding\\nOn a historical note, digit recognition problems were the catalyst that\\naccelerated the development of neural network technology in the late 1980s\\nat AT&T Bell Laboratories and elsewhere. Pattern recognition tasks of this\\nkind are relatively simple for humans. Our visual system occupies a large\\nfraction of our brains, and good recognition is an evolutionary force for\\nsurvival. These tasks are not so simple for machines, and it has taken more\\nthan 30 years to refine the neural-network architectures to match human\\nperformance.\\nFigure 10.4 shows a multilayer network architecture that works well for\\nsolving the digit-classification task. It differs from Figure 10.1 in several\\nways:\\n• It has two hidden layers L1 (256 units) and L2 (128 units) rather\\nthan one. Later we will see a network with seven hidden layers.\\n• It has ten output variables, rather than one. In this case the ten vari-\\nables really represent a single qualitative variable and so are quite\\ndependent. (We have indexed them by the digit class 0–9 rather than\\n1–10, for clarity.) More generally, in multi-task learning one can pre- multi-task\\nlearning\\ndict different responses simultaneously with a single network; they\\nall have a say in the formation of the hidden layers.\\n• The loss function used for training the network is tailored for the\\nmulticlass classification task.\\n2In the analog-to-digital conversion process, only part of the written numeral may\\nfall in the square representing a particular pixel. 404\\n10. Deep Learning\\nX1\\nX2\\nX3\\nX4\\nX5\\nX6\\n...\\nXp\\nA(1)\\n1\\nA(1)\\n2\\nA(1)\\n3\\nA(1)\\n4\\n...\\nA(1)\\nK1\\nA(2)\\n1\\nA(2)\\n2\\nA(2)\\n3\\n...\\nA(2)\\nK2\\nf0(X)\\nY0\\nf1(X)\\nY1\\n...\\n...\\nf9(X)\\nY9\\nHidden\\nlayer L2\\nHidden\\nlayer L1\\nInput\\nlayer\\nOutput\\nlayer\\nW1\\nW2\\nB\\nFIGURE 10.4. Neural network diagram with two hidden layers and multiple\\noutputs, suitable for the MNIST handwritten-digit problem. The input layer has\\np = 784 units, the two hidden layers K1 = 256 and K2 = 128 units respectively,\\nand the output layer 10 units. Along with intercepts (referred to as biases in the\\ndeep-learning community) this network has 235,146 parameters (referred to as\\nweights).\\nThe first hidden layer is as in (10.2), with\\nA(1)\\nk\\n=\\nh(1)\\nk (X)\\n=\\ng(w(1)\\nk0 + )p\\nj=1 w(1)\\nkj Xj)\\n(10.10)\\nfor k = 1, . . . , K1. The second hidden layer treats the activations A(1)\\nk\\nof\\nthe first hidden layer as inputs and computes new activations\\nA(2)\\nℓ\\n=\\nh(2)\\nℓ(X)\\n=\\ng(w(2)\\nℓ0 + )K1\\nk=1 w(2)\\nℓk A(1)\\nk )\\n(10.11)\\nfor ℓ= 1, . . . , K2. Notice that each of the activations in the second layer\\nA(2)\\nℓ\\n= h(2)\\nℓ(X) is a function of the input vector X. This is the case because\\nwhile they are explicitly a function of the activations A(1)\\nk\\nfrom layer L1,\\nthese in turn are functions of X. This would also be the case with more\\nhidden layers. Thus, through a chain of transformations, the network is\\nable to build up fairly complex transformations of X that ultimately feed\\ninto the output layer as features.\\nWe have introduced additional superscript notation such as h(2)\\nℓ(X) and\\nw(2)\\nℓj\\nin (10.10) and (10.11) to indicate to which layer the activations and\\nweights (coefficients) belong, in this case layer 2. The notation W1 in Fig- weights 10.2 Multilayer Neural Networks\\n405\\nure 10.4 represents the entire matrix of weights that feed from the input\\nlayer to the first hidden layer L1. This matrix will have 785×256 = 200,960\\nelements; there are 785 rather than 784 because we must account for the\\nintercept or bias term.3\\nbias\\nEach element A(1)\\nk\\nfeeds to the second hidden layer L2 via the matrix of\\nweights W2 of dimension 257 × 128 = 32,896.\\nWe now get to the output layer, where we now have ten responses rather\\nthan one. The first step is to compute ten different linear models similar\\nto our single model (10.1),\\nZm\\n=\\nβm0 + )K2\\nℓ=1 βmℓh(2)\\nℓ(X)\\n=\\nβm0 + )K2\\nℓ=1 βmℓA(2)\\nℓ,\\n(10.12)\\nfor m = 0, 1, . . . , 9. The matrix B stores all 129 × 10 = 1,290 of these\\nweights.\\nIf these were all separate quantitative responses, we would simply set\\neach fm(X) = Zm and be done. However, we would like our estimates to\\nrepresent class probabilities fm(X) = Pr(Y = m|X), just like in multi-\\nnomial logistic regression in Section 4.3.5. So we use the special softmax softmax\\nactivation function (see (4.13) on page 145),\\nfm(X) = Pr(Y = m|X) =\\neZm\\n)9\\nℓ=0 eZℓ,\\n(10.13)\\nfor m = 0, 1, . . . , 9. This ensures that the 10 numbers behave like proba-\\nbilities (non-negative and sum to one). Even though the goal is to build\\na classifier, our model actually estimates a probability for each of the 10\\nclasses. The classifier then assigns the image to the class with the highest\\nprobability.\\nTo train this network, since the response is qualitative, we look for coef-\\nficient estimates that minimize the negative multinomial log-likelihood\\n−\\nn\\n0\\ni=1\\n9\\n0\\nm=0\\nyim log(fm(xi)),\\n(10.14)\\nalso known as the cross-entropy. This is a generalization of the crite- cross-\\nentropy\\nrion (4.5) for two-class logistic regression. Details on how to minimize this\\nobjective are given in Section 10.7. If the response were quantitative, we\\nwould instead minimize squared-error loss as in (10.9).\\nTable 10.1 compares the test performance of the neural network with\\ntwo simple models presented in Chapter 4 that make use of linear decision\\nboundaries: multinomial logistic regression and linear discriminant analysis.\\nThe improvement of neural networks over both of these linear methods is\\ndramatic: the network with dropout regularization achieves a test error rate\\nbelow 2% on the 10,000 test images. (We describe dropout regularization in\\nSection 10.7.3.) In Section 10.9.2 of the lab, we present the code for fitting\\nthis model, which runs in just over two minutes on a laptop computer.\\n3The use of “weights” for coefficients and “bias” for the intercepts wk0 in (10.2) is\\npopular in the machine learning community; this use of bias is not to be confused with\\nthe “bias-variance” usage elsewhere in this book. 406\\n10. Deep Learning\\nMethod\\nTest Error\\nNeural Network + Ridge Regularization\\n2.3%\\nNeural Network + Dropout Regularization\\n1.8%\\nMultinomial Logistic Regression\\n7.2%\\nLinear Discriminant Analysis\\n12.7%\\nTABLE 10.1. Test error rate on the MNIST data, for neural networks with two\\nforms of regularization, as well as multinomial logistic regression and linear dis-\\ncriminant analysis. In this example, the extra complexity of the neural network\\nleads to a marked improvement in test error.\\nFIGURE 10.5. A sample of images from the CIFAR100 database: a collection of\\nnatural images from everyday life, with 100 different classes represented.\\nAdding the number of coefficients in W1, W2 and B, we get 235,146 in\\nall, more than 33 times the number 785×9 = 7,065 needed for multinomial\\nlogistic regression. Recall that there are 60,000 images in the training set.\\nWhile this might seem like a large training set, there are almost four times\\nas many coefficients in the neural network model as there are observations in\\nthe training set! To avoid overfitting, some regularization is needed. In this\\nexample, we used two forms of regularization: ridge regularization, which\\nis similar to ridge regression from Chapter 6, and dropout regularization. dropout\\nWe discuss both forms of regularization in Section 10.7.\\n10.3\\nConvolutional Neural Networks\\nNeural networks rebounded around 2010 with big successes in image classi-\\nfication. Around that time, massive databases of labeled images were being\\naccumulated, with ever-increasing numbers of classes. Figure 10.5 shows\\n75 images drawn from the CIFAR100 database.4 This database consists of\\n60,000 images labeled according to 20 superclasses (e.g. aquatic mammals),\\nwith five classes per superclass (beaver, dolphin, otter, seal, whale). Each\\nimage has a resolution of 32 × 32 pixels, with three eight-bit numbers per\\npixel representing red, green and blue. The numbers for each image are\\norganized in a three-dimensional array called a feature map. The first two feature map\\n4See\\nChapter\\n3\\nof\\nKrizhevsky\\n(2009)\\n“Learning\\nmultiple\\nlayers\\nof\\nfea-\\ntures\\nfrom\\ntiny\\nimages”,\\navailable\\nat\\nhttps://www.cs.toronto.edu/~kriz/\\nlearning-features-2009-TR.pdf. 10.3 Convolutional Neural Networks\\n407\\nFIGURE 10.6. Schematic showing how a convolutional neural network classifies\\nan image of a tiger. The network takes in the image and identifies local features.\\nIt then combines the local features in order to create compound features, which in\\nthis example include eyes and ears. These compound features are used to output\\nthe label “tiger”.\\naxes are spatial (both are 32-dimensional), and the third is the channel channel\\naxis,5 representing the three colors. There is a designated training set of\\n50,000 images, and a test set of 10,000.\\nA special family of convolutional neural networks (CNNs) has evolved for convolutional\\nneural\\nnetworks\\nclassifying images such as these, and has shown spectacular success on a\\nwide range of problems. CNNs mimic to some degree how humans classify\\nimages, by recognizing specific features or patterns anywhere in the image\\nthat distinguish each particular object class. In this section we give a brief\\noverview of how they work.\\nFigure 10.6 illustrates the idea behind a convolutional neural network on\\na cartoon image of a tiger.6\\nThe network first identifies low-level features in the input image, such\\nas small edges, patches of color, and the like. These low-level features are\\nthen combined to form higher-level features, such as parts of ears, eyes,\\nand so on. Eventually, the presence or absence of these higher-level features\\ncontributes to the probability of any given output class.\\nHow does a convolutional neural network build up this hierarchy? It com-\\nbines two specialized types of hidden layers, called convolution layers and\\npooling layers. Convolution layers search for instances of small patterns in\\nthe image, whereas pooling layers downsample these to select a prominent\\nsubset. In order to achieve state-of-the-art results, contemporary neural-\\nnetwork architectures make use of many convolution and pooling layers.\\nWe describe convolution and pooling layers next.\\n10.3.1\\nConvolution Layers\\nA convolution layer is made up of a large number of convolution filters, each convolution\\nlayer\\nconvolution\\nfilter\\n5The term channel is taken from the signal-processing literature. Each channel is a\\ndistinct source of information.\\n6\\nThanks\\nto\\nElena\\nTuzhilina\\nfor\\nproducing\\nthe\\ndiagram\\nand\\nhttps://www.\\ncartooning4kids.com/ for permission to use the cartoon tiger. 408\\n10. Deep Learning\\nof which is a template that determines whether a particular local feature is\\npresent in an image. A convolution filter relies on a very simple operation,\\ncalled a convolution, which basically amounts to repeatedly multiplying\\nmatrix elements and then adding the results.\\nTo understand how a convolution filter works, consider a very simple\\nexample of a 4 × 3 image:\\nOriginal Image =\\n\\uf8ee\\n\\uf8ef\\uf8ef\\uf8f0\\na\\nb\\nc\\nd\\ne\\nf\\ng\\nh\\ni\\nj\\nk\\nl\\n\\uf8f9\\n\\uf8fa\\uf8fa\\uf8fb.\\nNow consider a 2 × 2 filter of the form\\nConvolution Filter =\\n3α\\nβ\\nγ\\nδ\\n4\\n.\\nWhen we convolve the image with the filter, we get the result7\\nConvolved Image =\\n\\uf8ee\\n\\uf8f0\\naα + bβ + dγ + eδ\\nbα + cβ + eγ + fδ\\ndα + eβ + gγ + hδ\\neα + fβ + hγ + iδ\\ngα + hβ + jγ + kδ\\nhα + iβ + kγ + lδ\\n\\uf8f9\\n\\uf8fb.\\nFor instance, the top-left element comes from multiplying each element in\\nthe 2 × 2 filter by the corresponding element in the top left 2 × 2 portion\\nof the image, and adding the results. The other elements are obtained in a\\nsimilar way: the convolution filter is applied to every 2×2 submatrix of the\\noriginal image in order to obtain the convolved image. If a 2 × 2 submatrix\\nof the original image resembles the convolution filter, then it will have a\\nlarge value in the convolved image; otherwise, it will have a small value.\\nThus, the convolved image highlights regions of the original image that\\nresemble the convolution filter.\\nWe have used 2 × 2 as an example; in\\ngeneral convolution filters are small ℓ1 × ℓ2 arrays, with ℓ1 and ℓ2 small\\npositive integers that are not necessarily equal.\\nFigure 10.7 illustrates the application of two convolution filters to a 192×\\n179 image of a tiger, shown on the left-hand side.8 Each convolution filter\\nis a 15 × 15 image containing mostly zeros (black), with a narrow strip\\nof ones (white) oriented either vertically or horizontally within the image.\\nWhen each filter is convolved with the image of the tiger, areas of the tiger\\nthat resemble the filter (i.e. that have either horizontal or vertical stripes or\\nedges) are given large values, and areas of the tiger that do not resemble the\\nfeature are given small values. The convolved images are displayed on the\\nright-hand side. We see that the horizontal stripe filter picks out horizontal\\nstripes and edges in the original image, whereas the vertical stripe filter\\npicks out vertical stripes and edges in the original image.\\n7The convolved image is smaller than the original image because its dimension is\\ngiven by the number of 2 × 2 submatrices in the original image. Note that 2 × 2 is the\\ndimension of the convolution filter. If we want the convolved image to have the same\\ndimension as the original image, then padding can be applied.\\n8The tiger image used in Figures 10.7–10.9 was obtained from the public domain\\nimage resource https://www.needpix.com/. 10.3 Convolutional Neural Networks\\n409\\nFIGURE 10.7. Convolution filters find local features in an image, such as edges\\nand small shapes. We begin with the image of the tiger shown on the left, and\\napply the two small convolution filters in the middle. The convolved images high-\\nlight areas in the original image where details similar to the filters are found.\\nSpecifically, the top convolved image highlights the tiger’s vertical stripes, whereas\\nthe bottom convolved image highlights the tiger’s horizontal stripes. We can think\\nof the original image as the input layer in a convolutional neural network, and\\nthe convolved images as the units in the first hidden layer.\\nWe have used a large image and two large filters in Figure 10.7 for illus-\\ntration. For the CIFAR100 database there are 32×32 color pixels per image,\\nand we use 3 × 3 convolution filters.\\nIn a convolution layer, we use a whole bank of filters to pick out a variety\\nof differently-oriented edges and shapes in the image. Using predefined\\nfilters in this way is standard practice in image processing. By contrast,\\nwith CNNs the filters are learned for the specific classification task. We can\\nthink of the filter weights as the parameters going from an input layer to a\\nhidden layer, with one hidden unit for each pixel in the convolved image.\\nThis is in fact the case, though the parameters are highly structured and\\nconstrained (see Exercise 4 for more details). They operate on localized\\npatches in the input image (so there are many structural zeros), and the\\nsame weights in a given filter are reused for all possible patches in the image\\n(so the weights are constrained).9\\nWe now give some additional details.\\n• Since the input image is in color, it has three channels represented\\nby a three-dimensional feature map (array). Each channel is a two-\\ndimensional (32 × 32) feature map — one for red, one for green, and\\none for blue. A single convolution filter will also have three channels,\\none per color, each of dimension 3×3, with potentially different filter\\nweights. The results of the three convolutions are summed to form\\na two-dimensional output feature map. Note that at this point the\\ncolor information has been used, and is not passed on to subsequent\\nlayers except through its role in the convolution.\\n9This used to be called weight sharing in the early years of neural networks. 410\\n10. Deep Learning\\n• If we use K different convolution filters at this first hidden layer,\\nwe get K two-dimensional output feature maps, which together are\\ntreated as a single three-dimensional feature map. We view each of\\nthe K output feature maps as a separate channel of information, so\\nnow we have K channels in contrast to the three color channels of\\nthe original input feature map. The three-dimensional feature map is\\njust like the activations in a hidden layer of a simple neural network,\\nexcept organized and produced in a spatially structured way.\\n• We typically apply the ReLU activation function (10.5) to the con-\\nvolved image. This step is sometimes viewed as a separate layer in\\nthe convolutional neural network, in which case it is referred to as a\\ndetector layer.\\ndetector\\nlayer\\n10.3.2\\nPooling Layers\\nA pooling layer provides a way to condense a large image into a smaller pooling\\nsummary image. While there are a number of possible ways to perform\\npooling, the max pooling operation summarizes each non-overlapping 2 × 2\\nblock of pixels in an image using the maximum value in the block. This\\nreduces the size of the image by a factor of two in each direction, and it\\nalso provides some location invariance: i.e. as long as there is a large value\\nin one of the four pixels in the block, the whole block registers as a large\\nvalue in the reduced image.\\nHere is a simple example of max pooling:\\nMax pool\\n\\uf8ee\\n\\uf8ef\\uf8ef\\uf8f0\\n1\\n2\\n5\\n3\\n3\\n0\\n1\\n2\\n2\\n1\\n3\\n4\\n1\\n1\\n2\\n0\\n\\uf8f9\\n\\uf8fa\\uf8fa\\uf8fb→\\n33\\n5\\n2\\n4\\n4\\n.\\n10.3.3\\nArchitecture of a Convolutional Neural Network\\nSo far we have defined a single convolution layer — each filter produces a\\nnew two-dimensional feature map. The number of convolution filters in a\\nconvolution layer is akin to the number of units at a particular hidden layer\\nin a fully-connected neural network of the type we saw in Section 10.2.\\nThis number also defines the number of channels in the resulting three-\\ndimensional feature map. We have also described a pooling layer, which\\nreduces the first two dimensions of each three-dimensional feature map.\\nDeep CNNs have many such layers. Figure 10.8 shows a typical architecture\\nfor a CNN for the CIFAR100 image classification task.\\nAt the input layer, we see the three-dimensional feature map of a color\\nimage, where the channel axis represents each color by a 32 × 32 two-\\ndimensional feature map of pixels. Each convolution filter produces a new\\nchannel at the first hidden layer, each of which is a 32 × 32 feature map\\n(after some padding at the edges). After this first round of convolutions, we\\nnow have a new “image”; a feature map with considerably more channels\\nthan the three color input channels (six in the figure, since we used six\\nconvolution filters). 32\\n16\\n32\\n8\\n16\\n4\\n32\\n2\\n500\\n100\\nconvolve\\nconvolve\\nconvolve\\npool\\npool\\npool\\nflatten\\n8\\nFIGURE 10.8. Architecture of a deep CNN for the CIFAR100 classification task.\\nConvolution layers are interspersed with 2 × 2 max-pool layers, which reduce the\\nsize by a factor of 2 in both dimensions.\\nThis is followed by a max-pool layer, which reduces the size of the feature\\nmap in each channel by a factor of four: two in each dimension.\\nThis convolve-then-pool sequence is now repeated for the next two layers.\\nSome details are as follows:\\n• Each subsequent convolve layer is similar to the first. It takes as input\\nthe three-dimensional feature map from the previous layer and treats\\nit like a single multi-channel image. Each convolution filter learned\\nhas as many channels as this feature map.\\n• Since the channel feature maps are reduced in size after each pool\\nlayer, we usually increase the number of filters in the next convolve\\nlayer to compensate.\\n• Sometimes we repeat several convolve layers before a pool layer. This\\neffectively increases the dimension of the filter.\\nThese operations are repeated until the pooling has reduced each channel\\nfeature map down to just a few pixels in each dimension. At this point the\\nthree-dimensional feature maps are flattened — the pixels are treated as\\nseparate units — and fed into one or more fully-connected layers before\\nreaching the output layer, which is a softmax activation for the 100 classes\\n(as in (10.13)).\\nThere are many tuning parameters to be selected in constructing such a\\nnetwork, apart from the number, nature, and sizes of each layer. Dropout\\nlearning can be used at each layer, as well as lasso or ridge regularization\\n(see Section 10.7). The details of constructing a convolutional neural net-\\nwork can seem daunting. Fortunately, terrific software is available, with\\nextensive examples and vignettes that provide guidance on sensible choices\\nfor the parameters. For the CIFAR100 official test set, the best accuracy as\\nof this writing is just above 75%, but undoubtedly this performance will\\ncontinue to improve.\\n10.3.4\\nData Augmentation\\nAn additional important trick used with image modeling is data augment- data aug-\\nmentation\\nation. Essentially, each training image is replicated many times, with each\\nreplicate randomly distorted in a natural way such that human recognition\\nis unaffected. Figure 10.9 shows some examples. Typical distortions are\\n10.3 Convolutional Neural Networks\\n411 412\\n10. Deep Learning\\nFIGURE 10.9. Data augmentation. The original image (leftmost) is distorted\\nin natural ways to produce different images with the same class label. These\\ndistortions do not fool humans, and act as a form of regularization when fitting\\nthe CNN.\\nzoom, horizontal and vertical shift, shear, small rotations, and in this case\\nhorizontal flips. At face value this is a way of increasing the training set\\nconsiderably with somewhat different examples, and thus protects against\\noverfitting. In fact we can see this as a form of regularization: we build a\\ncloud of images around each original image, all with the same label. This\\nkind of fattening of the data is similar in spirit to ridge regularization.\\nWe will see in Section 10.7.2 that the stochastic gradient descent al-\\ngorithms for fitting deep learning models repeatedly process randomly-\\nselected batches of, say, 128 training images at a time. This works hand-in-\\nglove with augmentation, because we can distort each image in the batch\\non the fly, and hence do not have to store all the new images.\\n10.3.5\\nResults Using a Pretrained Classifier\\nHere we use an industry-level pretrained classifier to predict the class of\\nsome new images. The resnet50 classifier is a convolutional neural network\\nthat was trained using the imagenet data set, which consists of millions of\\nimages that belong to an ever-growing number of categories.10 Figure 10.10\\ndemonstrates the performance of resnet50 on six photographs (private col-\\nlection of one of the authors).11 The CNN does a reasonable job classifying\\nthe hawk in the second image. If we zoom out as in the third image, it\\ngets confused and chooses the fountain rather than the hawk. In the final\\nimage a “jacamar” is a tropical bird from South and Central America with\\nsimilar coloring to the South African Cape Weaver. We give more details\\non this example in Section 10.9.4.\\nMuch of the work in fitting a CNN is in learning the convolution filters\\nat the hidden layers; these are the coefficients of a CNN. For models fit to\\nmassive corpora such as imagenet with many classes, the output of these\\nfilters can serve as features for general natural-image classification prob-\\nlems. One can use these pretrained hidden layers for new problems with\\nmuch smaller training sets (a process referred to as weight freezing), and weight\\nfreezing\\njust train the last few layers of the network, which requires much less data.\\n10For more information about resnet50, see He, Zhang, Ren, and Sun (2015) “Deep\\nresidual learning for image recognition”, https://arxiv.org/abs/1512.03385. For de-\\ntails about imagenet, see Russakovsky, Deng, et al. (2015) “ImageNet Large Scale\\nVisual Recognition Challenge”, in International Journal of Computer Vision.\\n11These resnet results can change with time, since the publicly-trained model gets\\nupdated periodically. 10.4 Document Classification\\n413\\nflamingo\\nCooper’s hawk\\nCooper’s hawk\\nflamingo\\n0.83\\nkite\\n0.60\\nfountain\\n0.35\\nspoonbill\\n0.17\\ngreat grey owl\\n0.09\\nnail\\n0.12\\nwhite stork\\n0.00\\nrobin\\n0.06\\nhook\\n0.07\\nLhasa Apso\\ncat\\nCape weaver\\nTibetan terrier\\n0.56\\nOld English sheepdog\\n0.82\\njacamar\\n0.28\\nLhasa\\n0.32\\nShih-Tzu\\n0.04\\nmacaw\\n0.12\\ncocker spaniel\\n0.03\\nPersian cat\\n0.04\\nrobin\\n0.12\\nFIGURE 10.10. Classification of six photographs using the resnet50 CNN\\ntrained on the imagenet corpus. The table below the images displays the true\\n(intended) label at the top of each panel, and the top three choices of the classifier\\n(out of 100). The numbers are the estimated probabilities for each choice. (A kite\\nis a raptor, but not a hawk.)\\nThe vignettes and book12 that accompany the keras package give more\\ndetails on such applications.\\n10.4\\nDocument Classification\\nIn this section we introduce a new type of example that has important\\napplications in industry and science: predicting attributes of documents.\\nExamples of documents include articles in medical journals, Reuters news\\nfeeds, emails, tweets, and so on. Our example will be IMDb (Internet Movie\\nDatabase) ratings — short documents where viewers have written critiques\\nof movies.13 The response in this case is the sentiment of the review, which\\nwill be positive or negative.\\n12Deep Learning with R by F. Chollet and J.J. Allaire, 2018, Manning Publications.\\n13For details, see Maas et al. (2011) “Learning word vectors for sentiment analysis”,\\nin Proceedings of the 49th Annual Meeting of the Association for Computational Lin-\\nguistics: Human Language Technologies, pages 142–150. 414\\n10. Deep Learning\\nHere is the beginning of a rather amusing negative review:\\nThis has to be one of the worst films of the 1990s. When my\\nfriends & I were watching this film (being the target audience it\\nwas aimed at) we just sat & watched the first half an hour with\\nour jaws touching the floor at how bad it really was. The rest\\nof the time, everyone else in the theater just started talking to\\neach other, leaving or generally crying into their popcorn . . .\\nEach review can be a different length, include slang or non-words, have\\nspelling errors, etc. We need to find a way to featurize such a document. featurize\\nThis is modern parlance for defining a set of predictors.\\nThe simplest and most common featurization is the bag-of-words model. bag-of-words\\nWe score each document for the presence or absence of each of the words in\\na language dictionary — in this case an English dictionary. If the dictionary\\ncontains M words, that means for each document we create a binary feature\\nvector of length M, and score a 1 for every word present, and 0 otherwise.\\nThat can be a very wide feature vector, so we limit the dictionary — in\\nthis case to the 10,000 most frequently occurring words in the training\\ncorpus of 25,000 reviews. Fortunately there are nice tools for doing this\\nautomatically. Here is the beginning of a positive review that has been\\nredacted in this way:\\n⟨START ⟩this film was just brilliant casting location scenery\\nstory direction everyone’s really suited the part they played and\\nyou could just imagine being there robert ⟨UNK ⟩is an amazing\\nactor and now the same being director ⟨UNK ⟩father came from\\nthe same scottish island as myself so i loved . . .\\nHere we can see many words have been omitted, and some unknown words\\n(UNK) have been marked as such. With this reduction the binary feature\\nvector has length 10,000, and consists mostly of 0’s and a smattering of 1’s\\nin the positions corresponding to words that are present in the document.\\nWe have a training set and test set, each with 25,000 examples, and each\\nbalanced with regard to sentiment. The resulting training feature matrix X\\nhas dimension 25,000×10,000, but only 1.3% of the binary entries are non-\\nzero. We call such a matrix sparse, because most of the values are the same\\n(zero in this case); it can be stored efficiently in sparse matrix format.14\\nsparse\\nmatrix\\nformat\\nThere are a variety of ways to account for the document length; here we\\nonly score a word as in or out of the document, but for example one could\\ninstead record the relative frequency of words. We split off a validation set\\nof size 2,000 from the 25,000 training observations (for model tuning), and\\nfit two model sequences:\\n• A lasso logistic regression using the glmnet package;\\n• A two-class neural network with two hidden layers, each with 16\\nReLU units.\\n14Rather than store the whole matrix, we can store instead the location and values for\\nthe nonzero entries. In this case, since the nonzero entries are all 1, just the locations\\nare stored. 10.4 Document Classification\\n415\\nFIGURE 10.11. Accuracy of the lasso and a two-hidden-layer neural network\\non the IMDb data. For the lasso, the x-axis displays −log(λ), while for the neural\\nnetwork it displays epochs (number of times the fitting algorithm passes through\\nthe training set). Both show a tendency to overfit, and achieve approximately the\\nsame test accuracy.\\nBoth methods produce a sequence of solutions. The lasso sequence is in-\\ndexed by the regularization parameter λ. The neural-net sequence is in-\\ndexed by the number of gradient-descent iterations used in the fitting,\\nas measured by training epochs or passes through the training set (Sec-\\ntion 10.7). Notice that the training accuracy in Figure 10.11 (black points)\\nincreases monotonically in both cases. We can use the validation error to\\npick a good solution from each sequence (blue points in the plots), which\\nwould then be used to make predictions on the test data set.\\nNote that a two-class neural network amounts to a nonlinear logistic\\nregression model. From (10.12) and (10.13) we can see that\\nlog\\n*Pr(Y = 1|X)\\nPr(Y = 0|X)\\n+\\n=\\nZ1 −Z0\\n(10.15)\\n=\\n(β10 −β00) +\\nK2\\n0\\nℓ=1\\n(β1ℓ−β0ℓ)A(2)\\nℓ.\\n(This shows the redundancy in the softmax function; for K classes we\\nreally only need to estimate K −1 sets of coefficients. See Section 4.3.5.) In\\nFigure 10.11 we show accuracy (fraction correct) rather than classification accuracy\\nerror (fraction incorrect), the former being more popular in the machine\\nlearning community. Both models achieve a test-set accuracy of about 88%.\\nThe bag-of-words model summarizes a document by the words present,\\nand ignores their context. There are at least two popular ways to take the\\ncontext into account:\\n• The bag-of-n-grams model. For example, a bag of 2-grams records bag-of-n-\\ngrams\\n●●●●●●\\n●●●\\n●●●\\n●\\n●\\n●\\n●\\n●\\n●\\n●●\\n●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●\\n4\\n6\\n8\\n10\\n12\\n0.6\\n0.7\\n0.8\\n0.9\\n1.0\\nLasso\\n−log(λ)\\nAccuracy\\n●●●●●●\\n●●●\\n●●\\n●\\n●\\n●\\n●\\n●●\\n●●●\\n●\\n●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●\\n●●●●●●\\n●●●\\n●●●\\n●\\n●\\n●\\n●\\n●●\\n●\\n●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●\\n●\\n●\\n●\\ntrain\\nvalidation\\ntest\\n●\\n●\\n●\\n●\\n●\\n●\\n●\\n●\\n●\\n●\\n●\\n●\\n●\\n●\\n●\\n●\\n●\\n●\\n●\\n●\\n5\\n10\\n15\\n20\\n0.6\\n0.7\\n0.8\\n0.9\\n1.0\\nNeural Net\\nEpochs\\nAccuracy\\n●\\n●\\n●\\n●\\n●\\n●\\n●\\n●\\n●\\n●\\n●\\n●\\n●\\n●\\n●\\n●\\n●\\n●\\n●\\n●\\n●\\n●\\n●\\n●\\n●\\n●\\n●\\n●\\n●\\n●\\n●\\n●\\n●\\n●\\n●\\n●\\n●\\n●\\n●\\n● 416\\n10. Deep Learning\\nthe consecutive co-occurrence of every distinct pair of words. “Bliss-\\nfully long” can be seen as a positive phrase in a movie review, while\\n“blissfully short” a negative.\\n• Treat the document as a sequence, taking account of all the words in\\nthe context of those that preceded and those that follow.\\nIn the next section we explore models for sequences of data, which have\\napplications in weather forecasting, speech recognition, language transla-\\ntion, and time-series prediction, to name a few. We continue with this IMDb\\nexample there.\\n10.5\\nRecurrent Neural Networks\\nMany data sources are sequential in nature, and call for special treatment\\nwhen building predictive models. Examples include:\\n• Documents such as book and movie reviews, newspaper articles, and\\ntweets. The sequence and relative positions of words in a document\\ncapture the narrative, theme and tone, and can be exploited in tasks\\nsuch as topic classification, sentiment analysis, and language transla-\\ntion.\\n• Time series of temperature, rainfall, wind speed, air quality, and so\\non. We may want to forecast the weather several days ahead, or cli-\\nmate several decades ahead.\\n• Financial time series, where we track market indices, trading volumes,\\nstock and bond prices, and exchange rates. Here prediction is often\\ndifficult, but as we will see, certain indices can be predicted with\\nreasonable accuracy.\\n• Recorded speech, musical recordings, and other sound recordings. We\\nmay want to give a text transcription of a speech, or perhaps a lan-\\nguage translation. We may want to assess the quality of a piece of\\nmusic, or assign certain attributes.\\n• Handwriting, such as doctor’s notes, and handwritten digits such as\\nzip codes. Here we want to turn the handwriting into digital text, or\\nread the digits (optical character recognition).\\nIn a recurrent neural network (RNN), the input object X is a sequence. recurrent\\nneural\\nnetwork\\nConsider a corpus of documents, such as the collection of IMDb movie re-\\nviews. Each document can be represented as a sequence of L words, so\\nX = {X1, X2, . . . , XL}, where each Xℓrepresents a word. The order of\\nthe words, and closeness of certain words in a sentence, convey semantic\\nmeaning. RNNs are designed to accommodate and take advantage of the\\nsequential nature of such input objects, much like convolutional neural net-\\nworks accommodate the spatial structure of image inputs. The output Y\\ncan also be a sequence (such as in language translation), but often is a\\nscalar, like the binary sentiment label of a movie review document. 10.5 Recurrent Neural Networks\\n417\\nA1\\nA2\\nA3\\nAL-1\\nAL\\n=\\nAℓ\\nOℓ\\nY\\nXℓ\\nO1\\nX1\\nO2\\nX2\\nO3\\nX3\\nOL-1\\nXL-1\\nOL\\nY\\nXL\\n. . .\\nW\\nU\\nB\\nW\\nB\\nW\\nB\\nW\\nB\\nW\\nB\\nW\\nB\\nU\\nU\\nU\\nU\\nFIGURE 10.12. Schematic of a simple recurrent neural network. The input is a\\nsequence of vectors {Xℓ}L\\n1 , and here the target is a single response. The network\\nprocesses the input sequence X sequentially; each Xℓfeeds into the hidden layer,\\nwhich also has as input the activation vector Aℓ−1 from the previous element in\\nthe sequence, and produces the current activation vector Aℓ. The same collections\\nof weights W, U and B are used as each element of the sequence is processed. The\\noutput layer produces a sequence of predictions Oℓfrom the current activation\\nAℓ, but typically only the last of these, OL, is of relevance. To the left of the equal\\nsign is a concise representation of the network, which is unrolled into a more\\nexplicit version on the right.\\nFigure 10.12 illustrates the structure of a very basic RNN with a sequence\\nX = {X1, X2, . . . , XL} as input, a simple output Y , and a hidden-layer\\nsequence {Aℓ}L\\n1 = {A1, A2, . . . , AL}. Each Xℓis a vector; in the document\\nexample Xℓcould represent a one-hot encoding for the ℓth word based on\\nthe language dictionary for the corpus (see the top panel in Figure 10.13\\nfor a simple example). As the sequence is processed one vector Xℓat a\\ntime, the network updates the activations Aℓin the hidden layer, taking\\nas input the vector Xℓand the activation vector Aℓ−1 from the previous\\nstep in the sequence. Each Aℓfeeds into the output layer and produces a\\nprediction Oℓfor Y . OL, the last of these, is the most relevant.\\nIn detail, suppose each vector Xℓof the input sequence has p components\\nXT\\nℓ= (Xℓ1, Xℓ2, . . . , Xℓp), and the hidden layer consists of K units AT\\nℓ=\\n(Aℓ1, Aℓ2, . . . , AℓK). As in Figure 10.4, we represent the collection of K ×\\n(p+1) shared weights wkj for the input layer by a matrix W, and similarly\\nU is a K × K matrix of the weights uks for the hidden-to-hidden layers,\\nand B is a K + 1 vector of weights βk for the output layer. Then\\nAℓk = g\\n1\\nwk0 +\\np\\n0\\nj=1\\nwkjXℓj +\\nK\\n0\\ns=1\\nuksAℓ−1,s\\n2\\n,\\n(10.16)\\nand the output Oℓis computed as\\nOℓ= β0 +\\nK\\n0\\nk=1\\nβkAℓk\\n(10.17)\\nfor a quantitative response, or with an additional sigmoid activation func-\\ntion for a binary response, for example. Here g(·) is an activation function\\nsuch as ReLU. Notice that the same weights W, U and B are used as we 418\\n10. Deep Learning\\nprocess each element in the sequence, i.e. they are not functions of ℓ. This\\nis a form of weight sharing used by RNNs, and similar to the use of filters weight\\nsharing\\nin convolutional neural networks (Section 10.3.1.) As we proceed from be-\\nginning to end, the activations Aℓaccumulate a history of what has been\\nseen before, so that the learned context can be used for prediction.\\nFor regression problems the loss function for an observation (X, Y ) is\\n(Y −OL)2,\\n(10.18)\\nwhich only references the final output OL = β0+)K\\nk=1 βkALk. Thus O1, O2,\\n. . . , OL−1 are not used. When we fit the model, each element Xℓof the input\\nsequence X contributes to OL via the chain (10.16), and hence contributes\\nindirectly to learning the shared parameters W, U and B via the loss\\n(10.18). With n input sequence/response pairs (xi, yi), the parameters are\\nfound by minimizing the sum of squares\\nn\\n0\\ni=1\\n(yi−oiL)2 =\\nn\\n0\\ni=1\\n1\\nyi−\\n\\'\\nβ0+\\nK\\n0\\nk=1\\nβkg\\n\\'\\nwk0+\\np\\n0\\nj=1\\nwkjxiLj+\\nK\\n0\\ns=1\\nuksai,L−1,s\\n((22\\n.\\n(10.19)\\nHere we use lowercase letters for the observed yi and vector sequences\\nxi = {xi1, xi2, . . . , xiL},15 as well as the derived activations.\\nSince the intermediate outputs Oℓare not used, one may well ask why\\nthey are there at all. First of all, they come for free, since they use the same\\noutput weights B needed to produce OL, and provide an evolving prediction\\nfor the output. Furthermore, for some learning tasks the response is also a\\nsequence, and so the output sequence {O1, O2, . . . , OL} is explicitly needed.\\nWhen used at full strength, recurrent neural networks can be quite com-\\nplex. We illustrate their use in two simple applications. In the first, we\\ncontinue with the IMDb sentiment analysis of the previous section, where\\nwe process the words in the reviews sequentially. In the second application,\\nwe illustrate their use in a financial time series forecasting problem.\\n10.5.1\\nSequential Models for Document Classification\\nHere we return to our classification task with the IMDb reviews. Our ap-\\nproach in Section 10.4 was to use the bag-of-words model. Here the plan\\nis to use instead the sequence of words occurring in a document to make\\npredictions about the label for the entire document.\\nWe have, however, a dimensionality problem: each word in our document\\nis represented by a one-hot-encoded vector (dummy variable) with 10,000\\nelements (one per word in the dictionary)! An approach that has become\\npopular is to represent each word in a much lower-dimensional embedding embedding\\nspace. This means that rather than representing each word by a binary\\nvector with 9,999 zeros and a single one in some position, we will represent\\nit instead by a set of m real numbers, none of which are typically zero. Here\\nm is the embedding dimension, and can be in the low 100s, or even less.\\nThis means (in our case) that we need a matrix E of dimension m×10,000,\\n15This is a sequence of vectors; each element xiℓis a p-vector. 10.5 Recurrent Neural Networks\\n419\\nthis\\nis\\none\\nof\\nthe\\nbest\\nfilms\\nactually\\nthe\\nbest\\nI\\nhave\\never\\nseen\\nthe\\nfilm\\nstarts\\none\\nfall\\nday\\nOne−hot\\nEmbed\\nFIGURE 10.13. Depiction of a sequence of 20 words representing a single doc-\\nument: one-hot encoded using a dictionary of 16 words (top panel) and embedded\\nin an m-dimensional space with m = 5 (bottom panel).\\nwhere each column is indexed by one of the 10,000 words in our dictionary,\\nand the values in that column give the m coordinates for that word in the\\nembedding space.\\nFigure 10.13 illustrates the idea (with a dictionary of 16 rather than\\n10,000, and m = 5). Where does E come from? If we have a large corpus\\nof labeled documents, we can have the neural network learn E as part\\nof the optimization. In this case E is referred to as an embedding layer, embedding\\nlayer\\nand a specialized E is learned for the task at hand. Otherwise we can\\ninsert a precomputed matrix E in the embedding layer, a process known\\nas weight freezing. Two pretrained embeddings, word2vec and GloVe, are weight\\nfreezing\\nword2vec\\nGloVe\\nwidely used.16 These are built from a very large corpus of documents by\\na variant of principal components analysis (Section 12.2). The idea is that\\nthe positions of words in the embedding space preserve semantic meaning;\\ne.g. synonyms should appear near each other.\\nSo far, so good. Each document is now represented as a sequence of m-\\nvectors that represents the sequence of words. The next step is to limit\\neach document to the last L words. Documents that are shorter than L\\nget padded with zeros upfront. So now each document is represented by a\\nseries consisting of L vectors X = {X1, X2, . . . , XL}, and each Xℓin the\\nsequence has m components.\\nWe now use the RNN structure in Figure 10.12. The training corpus\\nconsists of n separate series (documents) of length L, each of which gets\\nprocessed sequentially from left to right. In the process, a parallel series of\\nhidden activation vectors Aℓ, ℓ= 1, . . . , L is created as in (10.16) for each\\ndocument. Aℓfeeds into the output layer to produce the evolving prediction\\nOℓ. We use the final value OL to predict the response: the sentiment of the\\nreview.\\n16 word2vec is described in Mikolov, Chen, Corrado, and Dean (2013), available\\nat https://code.google.com/archive/p/word2vec. GloVe is described in Pennington,\\nSocher, and Manning (2014), available at https://nlp.stanford.edu/projects/glove. 420\\n10. Deep Learning\\nThis is a simple RNN, and has relatively few parameters. If there are K\\nhidden units, the common weight matrix W has K × (m + 1) parameters,\\nthe matrix U has K × K parameters, and B has 2(K + 1) for the two-class\\nlogistic regression as in (10.15). These are used repeatedly as we process\\nthe sequence X = {Xℓ}L\\n1 from left to right, much like we use a single\\nconvolution filter to process each patch in an image (Section 10.3.1). If the\\nembedding layer E is learned, that adds an additional m × D parameters\\n(D = 10,000 here), and is by far the biggest cost.\\nWe fit the RNN as described in Figure 10.12 and the accompaying text to\\nthe IMDb data. The model had an embedding matrix E with m = 32 (which\\nwas learned in training as opposed to precomputed), followed by a single\\nrecurrent layer with K = 32 hidden units. The model was trained with\\ndropout regularization on the 25,000 reviews in the designated training\\nset, and achieved a disappointing 76% accuracy on the IMDb test data. A\\nnetwork using the GloVe pretrained embedding matrix E performed slightly\\nworse.\\nFor ease of exposition we have presented a very simple RNN. More elab-\\norate versions use long term and short term memory (LSTM). Two tracks\\nof hidden-layer activations are maintained, so that when the activation Aℓ\\nis computed, it gets input from hidden units both further back in time,\\nand closer in time — a so-called LSTM RNN. With long sequences, this LSTM RNN\\novercomes the problem of early signals being washed out by the time they\\nget propagated through the chain to the final activation vector AL.\\nWhen we refit our model using the LSTM architecture for the hidden\\nlayer, the performance improved to 87% on the IMDb test data. This is com-\\nparable with the 88% achieved by the bag-of-words model in Section 10.4.\\nWe give details on fitting these models in Section 10.9.6.\\nDespite this added LSTM complexity, our RNN is still somewhat “entry\\nlevel”. We could probably achieve slightly better results by changing the\\nsize of the model, changing the regularization, and including additional\\nhidden layers. However, LSTM models take a long time to train, which\\nmakes exploring many architectures and parameter optimization tedious.\\nRNNs provide a rich framework for modeling data sequences, and they\\ncontinue to evolve. There have been many advances in the development\\nof RNNs — in architecture, data augmentation, and in the learning algo-\\nrithms. At the time of this writing (early 2020) the leading RNN configura-\\ntions report accuracy above 95% on the IMDb data. The details are beyond\\nthe scope of this book.17\\n10.5.2\\nTime Series Forecasting\\nFigure 10.14 shows historical trading statistics from the New York Stock\\nExchange. Shown are three daily time series covering the period December\\n3, 1962 to December 31, 1986:18\\n17An IMDb\\nleaderboard\\ncan\\nbe\\nfound\\nat\\nhttps://paperswithcode.com/sota/\\nsentiment-analysis-on-imdb.\\n18These data were assembled by LeBaron and Weigend (1998) IEEE Transactions on\\nNeural Networks, 9(1): 213–220. 10.5 Recurrent Neural Networks\\n421\\nLog(Trading Volume)\\n−1.0\\n0.0\\n0.5\\n1.0\\nDow Jones Return\\n−0.04\\n0.00\\n0.04\\n1965\\n1970\\n1975\\n1980\\n1985\\n−13\\n−11\\n−9\\n−8\\nLog(Volatility)\\nFIGURE 10.14. Historical trading statistics from the New York Stock Exchange.\\nDaily values of the normalized log trading volume, DJIA return, and log volatility\\nare shown for a 24-year period from 1962–1986. We wish to predict trading volume\\non any day, given the history on all earlier days. To the left of the red bar (January\\n2, 1980) is training data, and to the right test data.\\n• Log trading volume. This is the fraction of all outstanding shares that\\nare traded on that day, relative to a 100-day moving average of past\\nturnover, on the log scale.\\n• Dow Jones return. This is the difference between the log of the Dow\\nJones Industrial Index on consecutive trading days.\\n• Log volatility. This is based on the absolute values of daily price\\nmovements.\\nPredicting stock prices is a notoriously hard problem, but it turns out that\\npredicting trading volume based on recent past history is more manageable\\n(and is useful for planning trading strategies).\\nAn observation here consists of the measurements (vt, rt, zt) on day t, in\\nthis case the values for log_volume, DJ_return and log_volatility. There\\nare a total of T = 6,051 such triples, each of which is plotted as a time series\\nin Figure 10.14. One feature that strikes us immediately is that the day-\\nto-day observations are not independent of each other. The series exhibit\\nauto-correlation — in this case values nearby in time tend to be similar auto-\\ncorrelation\\nto each other. This distinguishes time series from other data sets we have\\nencountered, in which observations can be assumed to be independent of 422\\n10. Deep Learning\\n0\\n5\\n10\\n15\\n20\\n25\\n30\\n35\\n0.0\\n0.4\\n0.8\\nLag\\nAutocorrelation Function\\nFIGURE 10.15. The autocorrelation function for log_volume. We see that\\nnearby values are fairly strongly correlated, with correlations above 0.2 as far as\\n20 days apart.\\neach other. To be clear, consider pairs of observations (vt, vt−ℓ), a lag of ℓlag\\ndays apart. If we take all such pairs in the vt series and compute their corre-\\nlation coefficient, this gives the autocorrelation at lag ℓ. Figure 10.15 shows\\nthe autocorrelation function for all lags up to 37, and we see considerable\\ncorrelation.\\nAnother interesting characteristic of this forecasting problem is that the\\nresponse variable vt — log_volume — is also a predictor! In particular, we\\nwill use the past values of log_volume to predict values in the future.\\nRNN forecaster\\nWe wish to predict a value vt from past values vt−1, vt−2, . . ., and also to\\nmake use of past values of the other series rt−1, rt−2, . . . and zt−1, zt−2, . . ..\\nAlthough our combined data is quite a long series with 6,051 trading\\ndays, the structure of the problem is different from the previous document-\\nclassification example.\\n• We only have one series of data, not 25,000.\\n• We have an entire series of targets vt, and the inputs include past\\nvalues of this series.\\nHow do we represent this problem in terms of the structure displayed in\\nFigure 10.12? The idea is to extract many short mini-series of input se-\\nquences X = {X1, X2, . . . , XL} with a predefined length L (called the lag lag\\nin this context), and a corresponding target Y . They have the form\\nX1 =\\n\\uf8eb\\n\\uf8ed\\nvt−L\\nrt−L\\nzt−L\\n\\uf8f6\\n\\uf8f8, X2 =\\n\\uf8eb\\n\\uf8ed\\nvt−L+1\\nrt−L+1\\nzt−L+1\\n\\uf8f6\\n\\uf8f8, · · · , XL =\\n\\uf8eb\\n\\uf8ed\\nvt−1\\nrt−1\\nzt−1\\n\\uf8f6\\n\\uf8f8, and Y = vt.\\n(10.20)\\nSo here the target Y is the value of log_volume vt at a single timepoint t,\\nand the input sequence X is the series of 3-vectors {Xℓ}L\\n1 each consisting\\nof the three measurements log_volume, DJ_return and log_volatility from\\nday t −L, t −L + 1, up to t −1. Each value of t makes a separate (X, Y )\\npair, for t running from L + 1 to T. For the NYSE data we will use the past 10.5 Recurrent Neural Networks\\n423\\n1980\\n1982\\n1984\\n1986\\n−1.0\\n0.0\\n0.5\\n1.0\\nlog(Trading Volume)\\nFIGURE 10.16. RNN forecast of log_volume on the NYSE test data. The black\\nlines are the true volumes, and the superimposed orange the forecasts. The fore-\\ncasted series accounts for 42% of the variance of log_volume.\\nfive trading days to predict the next day’s trading volume. Hence, we use\\nL = 5. Since T = 6,051, we can create 6,046 such (X, Y ) pairs. Clearly L\\nis a parameter that should be chosen with care, perhaps using validation\\ndata.\\nWe fit this model with K = 12 hidden units using the 4,281 training\\nsequences derived from the data before January 2, 1980 (see Figure 10.14),\\nand then used it to forecast the 1,770 values of log_volume after this date.\\nWe achieve an R2 = 0.42 on the test data. Details are given in Sec-\\ntion 10.9.6. As a straw man,19 using yesterday’s value for log_volume as\\nthe prediction for today has R2 = 0.18. Figure 10.16 shows the forecast\\nresults. We have plotted the observed values of the daily log_volume for the\\ntest period 1980–1986 in black, and superimposed the predicted series in\\norange. The correspondence seems rather good.\\nIn forecasting the value of log_volume in the test period, we have to use\\nthe test data itself in forming the input sequences X. This may feel like\\ncheating, but in fact it is not; we are always using past data to predict the\\nfuture.\\nAutoregression\\nThe RNN we just fit has much in common with a traditional autoregression auto-\\nregression\\n(AR) linear model, which we present now for comparison. We first consider\\nthe response sequence vt alone, and construct a response vector y and a\\nmatrix M of predictors for least squares regression as follows:\\ny =\\n\\uf8ee\\n\\uf8ef\\uf8ef\\uf8ef\\uf8ef\\uf8ef\\uf8f0\\nvL+1\\nvL+2\\nvL+3\\n...\\nvT\\n\\uf8f9\\n\\uf8fa\\uf8fa\\uf8fa\\uf8fa\\uf8fa\\uf8fb\\nM =\\n\\uf8ee\\n\\uf8ef\\uf8ef\\uf8ef\\uf8ef\\uf8ef\\uf8f0\\n1\\nvL\\nvL−1\\n· · ·\\nv1\\n1\\nvL+1\\nvL\\n· · ·\\nv2\\n1\\nvL+2\\nvL+1\\n· · ·\\nv3\\n...\\n...\\n...\\n...\\n...\\n1\\nvT −1\\nvT −2\\n· · ·\\nvT −L\\n\\uf8f9\\n\\uf8fa\\uf8fa\\uf8fa\\uf8fa\\uf8fa\\uf8fb\\n.\\n(10.21)\\nM and y each have T −L rows, one per observation. We see that the\\npredictors for any given response vt on day t are the previous L values\\n19A straw man here refers to a simple and sensible prediction that can be used as a\\nbaseline for comparison. 424\\n10. Deep Learning\\nof the same series. Fitting a regression of y on M amounts to fitting the\\nmodel\\nˆvt = ˆβ0 + ˆβ1vt−1 + ˆβ2vt−2 + · · · + ˆβLvt−L,\\n(10.22)\\nand is called an order-L autoregressive model, or simply AR(L). For the\\nNYSE data we can include lagged versions of DJ_return and log_volatility,\\nrt and zt, in the predictor matrix M, resulting in 3L + 1 columns. An AR\\nmodel with L = 5 achieves a test R2 of 0.41, slightly inferior to the 0.42\\nachieved by the RNN.\\nOf course the RNN and AR models are very similar. They both use\\nthe same response Y and input sequences X of length L = 5 and dimen-\\nsion p = 3 in this case. The RNN processes this sequence from left to\\nright with the same weights W (for the input layer), while the AR model\\nsimply treats all L elements of the sequence equally as a vector of L × p\\npredictors — a process called flattening in the neural network literature. flattening\\nOf course the RNN also includes the hidden layer activations Aℓwhich\\ntransfer information along the sequence, and introduces additional nonlin-\\nearity. From (10.19) with K = 12 hidden units, we see that the RNN has\\n13 + 12 × (1 + 3 + 12) = 205 parameters, compared to the 16 for the AR(5)\\nmodel.\\nAn obvious extension of the AR model is to use the set of lagged predic-\\ntors as the input vector to an ordinary feedforward neural network (10.1),\\nand hence add more flexibility. This achieved a test R2 = 0.42, slightly\\nbetter than the linear AR, and the same as the RNN.\\nAll the models can be improved by including the variable day_of_week\\ncorresponding to the day t of the target vt (which can be learned from the\\ncalendar dates supplied with the data); trading volume is often higher on\\nMondays and Fridays. Since there are five trading days, this one-hot en-\\ncodes to five binary variables. The performance of the AR model improved\\nto R2 = 0.46 as did the RNN, and the nonlinear AR model improved to\\nR2 = 0.47.\\nWe used the most simple version of the RNN in our examples here.\\nAdditional experiments with the LSTM extension of the RNN yielded small\\nimprovements, typically of up to 1% in R2 in these examples.\\nWe give details of how we fit all three models in Section 10.9.6.\\n10.5.3\\nSummary of RNNs\\nWe have illustrated RNNs through two simple use cases, and have only\\nscratched the surface.\\nThere are many variations and enhancements of the simple RNN we\\nused for sequence modeling. One approach we did not discuss uses a one-\\ndimensional convolutional neural network, treating the sequence of vectors\\n(say words, as represented in the embedding space) as an image. The con-\\nvolution filter slides along the sequence in a one-dimensional fashion, with\\nthe potential to learn particular phrases or short subsequences relevant to\\nthe learning task.\\nOne can also have additional hidden layers in an RNN. For example,\\nwith two hidden layers, the sequence Aℓis treated as an input sequence to\\nthe next hidden layer in an obvious fashion. 10.6 When to Use Deep Learning\\n425\\nThe RNN we used scanned the document from beginning to end; alter-\\nnative bidirectional RNNs scan the sequences in both directions.\\nbidirectional\\nIn language translation the target is also a sequence of words, in a\\nlanguage different from that of the input sequence. Both the input se-\\nquence and the target sequence are represented by a structure similar to\\nFigure 10.12, and they share the hidden units. In this so-called Seq2Seq Seq2Seq\\nlearning, the hidden units are thought to capture the semantic meaning\\nof the sentences. Some of the big breakthroughs in language modeling and\\ntranslation resulted from the relatively recent improvements in such RNNs.\\nAlgorithms used to fit RNNs can be complex and computationally costly.\\nFortunately, good software protects users somewhat from these complexi-\\nties, and makes specifying and fitting these models relatively painless. Many\\nof the models that we enjoy in daily life (like Google Translate) use state-\\nof-the-art architectures developed by teams of highly skilled engineers, and\\nhave been trained using massive computational and data resources.\\n10.6\\nWhen to Use Deep Learning\\nThe performance of deep learning in this chapter has been rather impres-\\nsive. It nailed the digit classification problem, and deep CNNs have really\\nrevolutionized image classification. We see daily reports of new success sto-\\nries for deep learning. Many of these are related to image classification\\ntasks, such as machine diagnosis of mammograms or digital X-ray images,\\nophthalmology eye scans, annotations of MRI scans, and so on. Likewise\\nthere are numerous successes of RNNs in speech and language translation,\\nforecasting, and document modeling. The question that then begs an an-\\nswer is: should we discard all our older tools, and use deep learning on every\\nproblem with data? To address this question, we revisit our Hitters dataset\\nfrom Chapter 6.\\nThis is a regression problem, where the goal is to predict the Salary of\\na baseball player in 1987 using his performance statistics from 1986. After\\nremoving players with missing responses, we are left with 263 players and\\n19 variables. We randomly split the data into a training set of 176 players\\n(two thirds), and a test set of 87 players (one third). We used three methods\\nfor fitting a regression model to these data.\\n• A linear model was used to fit the training data, and make predictions\\non the test data. The model has 20 parameters.\\n• The same linear model was fit with lasso regularization. The tuning\\nparameter was selected by 10-fold cross-validation on the training\\ndata. It selected a model with 12 variables having nonzero coefficients.\\n• A neural network with one hidden layer consisting of 64 ReLU units\\nwas fit to the data. This model has 1,345 parameters.20\\n20The model was fit by stochastic gradient descent with a batch size of 32 for 1,000\\nepochs, and 10% dropout regularization. The test error performance flattened out and\\nstarted to slowly increase after 1,000 epochs. These fitting details are discussed in Sec-\\ntion 10.7. 426\\n10. Deep Learning\\nModel\\n# Parameters\\nMean Abs. Error\\nTest Set R2\\nLinear Regression\\n20\\n254.7\\n0.56\\nLasso\\n12\\n252.3\\n0.51\\nNeural Network\\n1345\\n257.4\\n0.54\\nTABLE 10.2. Prediction results on the Hitters test data for linear models fit\\nby ordinary least squares and lasso, compared to a neural network fit by stochastic\\ngradient descent with dropout regularization.\\nCoefficient\\nStd. error\\nt-statistic\\np-value\\nIntercept\\n-226.67\\n86.26\\n-2.63\\n0.0103\\nHits\\n3.06\\n1.02\\n3.00\\n0.0036\\nWalks\\n0.181\\n2.04\\n0.09\\n0.9294\\nCRuns\\n0.859\\n0.12\\n7.09\\n< 0.0001\\nPutOuts\\n0.465\\n0.13\\n3.60\\n0.0005\\nTABLE 10.3. Least squares coefficient estimates associated with the regres-\\nsion of Salary on four variables chosen by lasso on the Hitters data set. This\\nmodel achieved the best performance on the test data, with a mean absolute error\\nof 224.8. The results reported here were obtained from a regression on the test\\ndata, which was not used in fitting the lasso model.\\nTable 10.2 compares the results. We see similar performance for all three\\nmodels. We report the mean absolute error on the test data, as well as\\nthe test R2 for each method, which are all respectable (see Exercise 5).\\nWe spent a fair bit of time fiddling with the configuration parameters of\\nthe neural network to achieve these results. It is possible that if we were to\\nspend more time, and got the form and amount of regularization just right,\\nthat we might be able to match or even outperform linear regression and\\nthe lasso. But with great ease we obtained linear models that work well.\\nLinear models are much easier to present and understand than the neural\\nnetwork, which is essentially a black box. The lasso selected 12 of the 19\\nvariables in making its prediction. So in cases like this we are much better\\noff following the Occam’s razor principle: when faced with several methods Occam’s\\nrazor\\nthat give roughly equivalent performance, pick the simplest.\\nAfter a bit more exploration with the lasso model, we identified an even\\nsimpler model with four variables. We then refit the linear model with these\\nfour variables to the training data (the so-called relaxed lasso), and achieved\\na test mean absolute error of 224.8, the overall winner! It is tempting to\\npresent the summary table from this fit, so we can see coefficients and p-\\nvalues; however, since the model was selected on the training data, there\\nwould be selection bias. Instead, we refit the model on the test data, which\\nwas not used in the selection. Table 10.3 shows the results.\\nWe have a number of very powerful tools at our disposal, including neural\\nnetworks, random forests and boosting, support vector machines and gen-\\neralized additive models, to name a few. And then we have linear models,\\nand simple variants of these. When faced with new data modeling and pre-\\ndiction problems, it’s tempting to always go for the trendy new methods.\\nOften they give extremely impressive results, especially when the datasets\\nare very large and can support the fitting of high-dimensional nonlinear\\nmodels. However, if we can produce models with the simpler tools that 10.7 Fitting a Neural Network\\n427\\nperform as well, they are likely to be easier to fit and understand, and po-\\ntentially less fragile than the more complex approaches. Wherever possible,\\nit makes sense to try the simpler models as well, and then make a choice\\nbased on the performance/complexity tradeoff.\\nTypically we expect deep learning to be an attractive choice when the\\nsample size of the training set is extremely large, and when interpretability\\nof the model is not a high priority.\\n10.7\\nFitting a Neural Network\\nFitting neural networks is somewhat complex, and we give a brief overview\\nhere. The ideas generalize to much more complex networks. Readers who\\nfind this material challenging can safely skip it. Fortunately, as we see in\\nthe lab at the end of this chapter, good software is available to fit neural\\nnetwork models in a relatively automated way, without worrying about the\\ntechnical details of the model-fitting procedure.\\nWe start with the simple network depicted in Figure 10.1 in Section 10.1.\\nIn model (10.1) the parameters are β = (β0, β1, . . . , βK), as well as each of\\nthe wk = (wk0, wk1, . . . , wkp), k = 1, . . . , K. Given observations (xi, yi), i =\\n1, . . . , n, we could fit the model by solving a nonlinear least squares problem\\nminimize\\n{wk}K\\n1 , β\\n1\\n2\\nn\\n0\\ni=1\\n(yi −f(xi))2,\\n(10.23)\\nwhere\\nf(xi) = β0 +\\nK\\n0\\nk=1\\nβkg\\n1\\nwk0 +\\np\\n0\\nj=1\\nwkjxij\\n2\\n.\\n(10.24)\\nThe objective in (10.23) looks simple enough, but because of the nested\\narrangement of the parameters and the symmetry of the hidden units, it is\\nnot straightforward to minimize. The problem is nonconvex in the param-\\neters, and hence there are multiple solutions. As an example, Figure 10.17\\nshows a simple nonconvex function of a single variable θ; there are two\\nsolutions: one is a local minimum and the other is a global minimum. Fur- local\\nminimum\\nglobal\\nminimum\\nthermore, (10.1) is the very simplest of neural networks; in this chapter we\\nhave presented much more complex ones where these problems are com-\\npounded. To overcome some of these issues and to protect from overfitting,\\ntwo general strategies are employed when fitting neural networks.\\n• Slow Learning: the model is fit in a somewhat slow iterative fash-\\nion, using gradient descent. The fitting process is then stopped when gradient\\ndescent\\noverfitting is detected.\\n• Regularization: penalties are imposed on the parameters, usually lasso\\nor ridge as discussed in Section 6.2.\\nSuppose we represent all the parameters in one long vector θ. Then we\\ncan rewrite the objective in (10.23) as\\nR(θ) = 1\\n2\\nn\\n0\\ni=1\\n(yi −fθ(xi))2,\\n(10.25) 428\\n10. Deep Learning\\nFIGURE 10.17. Illustration of gradient descent for one-dimensional θ. The\\nobjective function R(θ) is not convex, and has two minima, one at θ = −0.46\\n(local), the other at θ = 1.02 (global). Starting at some value θ0 (typically ran-\\ndomly chosen), each step in θ moves downhill — against the gradient — until it\\ncannot go down any further. Here gradient descent reached the global minimum\\nin 7 steps.\\nwhere we make explicit the dependence of f on the parameters. The idea\\nof gradient descent is very simple.\\n1. Start with a guess θ0 for all the parameters in θ, and set t = 0.\\n2. Iterate until the objective (10.25) fails to decrease:\\n(a) Find a vector δ that reflects a small change in θ, such that θt+1 =\\nθt + δ reduces the objective; i.e. such that R(θt+1) < R(θt).\\n(b) Set t ←t + 1.\\nOne can visualize (Figure 10.17) standing in a mountainous terrain, and\\nthe goal is to get to the bottom through a series of steps. As long as each\\nstep goes downhill, we must eventually get to the bottom. In this case we\\nwere lucky, because with our starting guess θ0 we end up at the global\\nminimum. In general we can hope to end up at a (good) local minimum.\\n10.7.1\\nBackpropagation\\nHow do we find the directions to move θ so as to decrease the objective R(θ)\\nin (10.25)? The gradient of R(θ), evaluated at some current value θ = θm, gradient\\nis the vector of partial derivatives at that point:\\n∇R(θm) = ∂R(θ)\\n∂θ\\nVVV\\nθ=θm.\\n(10.26)\\nThe subscript θ = θm means that after computing the vector of derivatives,\\nwe evaluate it at the current guess, θm. This gives the direction in θ-space\\nin which R(θ) increases most rapidly. The idea of gradient descent is to\\nmove θ a little in the opposite direction (since we wish to go downhill):\\nθm+1 ←θm −ρ∇R(θm).\\n(10.27)\\n−1.0\\n−0.5\\n0.0\\n0.5\\n1.0\\n0\\n1\\n2\\n3\\n4\\n5\\n6\\nθ\\nR(θ)\\nθ0\\nθ1\\nθ2\\nθ7\\n●\\n●\\n●\\n●\\nR(θ0)R(θ1)\\nR(θ2)\\nR(θ7) 10.7 Fitting a Neural Network\\n429\\nFor a small enough value of the learning rate ρ, this step will decrease the learning rate\\nobjective R(θ); i.e. R(θm+1) ≤R(θm). If the gradient vector is zero, then\\nwe may have arrived at a minimum of the objective.\\nHow complicated is the calculation (10.26)? It turns out that it is quite\\nsimple here, and remains simple even for much more complex networks,\\nbecause of the chain rule of differentiation.\\nchain rule\\nSince R(θ) = )n\\ni=1 Ri(θ) = 1\\n2\\n)n\\ni=1(yi −fθ(xi))2 is a sum, its gradient\\nis also a sum over the n observations, so we will just examine one of these\\nterms,\\nRi(θ) = 1\\n2\\n1\\nyi −β0 −\\nK\\n0\\nk=1\\nβkg\\n\\'\\nwk0 +\\np\\n0\\nj=1\\nwkjxij\\n(22\\n.\\n(10.28)\\nTo simplify the expressions to follow, we write zik = wk0 + )p\\nj=1 wkjxij.\\nFirst we take the derivative with respect to βk:\\n∂Ri(θ)\\n∂βk\\n=\\n∂Ri(θ)\\n∂fθ(xi) · ∂fθ(xi)\\n∂βk\\n=\\n−(yi −fθ(xi)) · g(zik).\\n(10.29)\\nAnd now we take the derivative with respect to wkj:\\n∂Ri(θ)\\n∂wkj\\n=\\n∂Ri(θ)\\n∂fθ(xi) · ∂fθ(xi)\\n∂g(zik) · ∂g(zik)\\n∂zik\\n· ∂zik\\n∂wkj\\n=\\n−(yi −fθ(xi)) · βk · g′(zik) · xij.\\n(10.30)\\nNotice that both these expressions contain the residual yi −fθ(xi). In\\n(10.29) we see that a fraction of that residual gets attributed to each of\\nthe hidden units according to the value of g(zik). Then in (10.30) we see\\na similar attribution to input j via hidden unit k. So the act of differen-\\ntiation assigns a fraction of the residual to each of the parameters via the\\nchain rule — a process known as backpropagation in the neural network backprop-\\nagation\\nliterature. Although these calculations are straightforward, it takes careful\\nbookkeeping to keep track of all the pieces.\\n10.7.2\\nRegularization and Stochastic Gradient Descent\\nGradient descent usually takes many steps to reach a local minimum. In\\npractice, there are a number of approaches for accelerating the process.\\nAlso, when n is large, instead of summing (10.29)–(10.30) over all n ob-\\nservations, we can sample a small fraction or minibatch of them each time minibatch\\nwe compute a gradient step. This process is known as stochastic gradient\\ndescent (SGD) and is the state of the art for learning deep neural networks. stochastic\\ngradient\\ndescent\\nFortunately, there is very good software for setting up deep learning mod-\\nels, and for fitting them to data, so most of the technicalities are hidden\\nfrom the user.\\nWe now turn to the multilayer network (Figure 10.4) used in the digit\\nrecognition problem. The network has over 235,000 weights, which is around\\nfour times the number of training examples. Regularization is essential here 430\\n10. Deep Learning\\n0\\n5\\n10\\n15\\n20\\n25\\n30\\n0.1\\n0.2\\n0.3\\n0.4\\nEpochs\\nValue of Objective Function\\nTraining Set\\nValidation Set\\n0\\n5\\n10\\n15\\n20\\n25\\n30\\n0.00\\n0.02\\n0.04\\n0.06\\n0.08\\n0.10\\n0.12\\nEpochs\\nClassification Error\\nFIGURE 10.18. Evolution of training and validation errors for the MNIST neural\\nnetwork depicted in Figure 10.4, as a function of training epochs. The objective\\nrefers to the log-likelihood (10.14).\\nto avoid overfitting. The first row in Table 10.1 uses ridge regularization on\\nthe weights. This is achieved by augmenting the objective function (10.14)\\nwith a penalty term:\\nR(θ; λ) = −\\nn\\n0\\ni=1\\n9\\n0\\nm=0\\nyim log(fm(xi)) + λ\\n0\\nj\\nθ2\\nj.\\n(10.31)\\nThe parameter λ is often preset at a small value, or else it is found using the\\nvalidation-set approach of Section 5.3.1. We can also use different values of\\nλ for the groups of weights from different layers; in this case W1 and W2\\nwere penalized, while the relatively few weights B of the output layer were\\nnot penalized at all. Lasso regularization is also popular as an additional\\nform of regularization, or as an alternative to ridge.\\nFigure 10.18 shows some metrics that evolve during the training of the\\nnetwork on the MNIST data. It turns out that SGD naturally enforces its\\nown form of approximately quadratic regularization.21 Here the minibatch\\nsize was 128 observations per gradient update. The term epochs labeling the epochs\\nhorizontal axis in Figure 10.18 counts the number of times an equivalent of\\nthe full training set has been processed. For this network, 20% of the 60,000\\ntraining observations were used as a validation set in order to determine\\nwhen training should stop. So in fact 48,000 observations were used for\\ntraining, and hence there are 48,000/128 ≈375 minibatch gradient updates\\nper epoch. We see that the value of the validation objective actually starts\\nto increase by 30 epochs, so early stopping can also be used as an additional early\\nstopping\\nform of regularization.\\n21This and other properties of SGD for deep learning are the subject of much research\\nin the machine learning literature at the time of writing. 10.7 Fitting a Neural Network\\n431\\nFIGURE 10.19. Dropout Learning. Left: a fully connected network. Right: net-\\nwork with dropout in the input and hidden layer. The nodes in grey are selected\\nat random, and ignored in an instance of training.\\n10.7.3\\nDropout Learning\\nThe second row in Table 10.1 is labeled dropout. This is a relatively new dropout\\nand efficient form of regularization, similar in some respects to ridge reg-\\nularization. Inspired by random forests (Section 8.2), the idea is to ran-\\ndomly remove a fraction φ of the units in a layer when fitting the model.\\nFigure 10.19 illustrates this. This is done separately each time a training\\nobservation is processed. The surviving units stand in for those missing,\\nand their weights are scaled up by a factor of 1/(1 −φ) to compensate.\\nThis prevents nodes from becoming over-specialized, and can be seen as\\na form of regularization. In practice dropout is achieved by randomly set-\\nting the activations for the “dropped out” units to zero, while keeping the\\narchitecture intact.\\n10.7.4\\nNetwork Tuning\\nThe network in Figure 10.4 is considered to be relatively straightforward;\\nit nevertheless requires a number of choices that all have an effect on the\\nperformance:\\n• The number of hidden layers, and the number of units per layer.\\nModern thinking is that the number of units per hidden layer can\\nbe large, and overfitting can be controlled via the various forms of\\nregularization.\\n• Regularization tuning parameters. These include the dropout rate φ\\nand the strength λ of lasso and ridge regularization, and are typically\\nset separately at each layer.\\n• Details of stochastic gradient descent. These include the batch size,\\nthe number of epochs, and if used, details of data augmentation (Sec-\\ntion 10.3.4.)\\nChoices such as these can make a difference. In preparing this MNIST exam-\\nple, we achieved a respectable 1.8% misclassification error after some trial\\nand error. Finer tuning and training of a similar network can get under\\n1% error on these data, but the tinkering process can be tedious, and can\\nresult in overfitting if done carelessly. 432\\n10. Deep Learning\\n2\\n5\\n10\\n20\\n50\\n0.0\\n0.5\\n1.0\\n1.5\\n2.0\\nDegrees of Freedom\\nError\\nTraining Error\\nTest Error\\nFIGURE 10.20. Double descent phenomenon, illustrated using error plots for a\\none-dimensional natural spline example. The horizontal axis refers to the number\\nof spline basis functions on the log scale. The training error hits zero when the\\ndegrees of freedom coincides with the sample size n = 20, the “interpolation\\nthreshold”, and remains zero thereafter. The test error increases dramatically\\nat this threshold, but then descends again to a reasonable value before finally\\nincreasing again.\\n10.8\\nInterpolation and Double Descent\\nThroughout this book, we have repeatedly discussed the bias-variance trade-\\noff, first presented in Section 2.2.2. This trade-off indicates that statistical\\nlearning methods tend to perform the best, in terms of test-set error, for an\\nintermediate level of model complexity. In particular, if we plot “flexibil-\\nity” on the x-axis and error on the y-axis, then we generally expect to see\\nthat test error has a U-shape, whereas training error decreases monotoni-\\ncally. Two “typical” examples of this behavior can be seen in the right-hand\\npanel of Figure 2.9 on page 29, and in Figure 2.17 on page 39. One implica-\\ntion of the bias-variance trade-off is that it is generally not a good idea to\\ninterpolate the training data — that is, to get zero training error — since interpolate\\nthat will often result in very high test error.\\nHowever, it turns out that in certain specific settings it can be possible for\\na statistical learning method that interpolates the training data to perform\\nwell — or at least, better than a slightly less complex model that does not\\nquite interpolate the data. This phenomenon is known as double descent,\\nand is displayed in Figure 10.20. “Double descent” gets its name from the\\nfact that the test error has a U-shape before the interpolation threshold is\\nreached, and then it descends again (for a while, at least) as an increasingly\\nflexible model is fit.\\nWe now describe the set-up that resulted in Figure 10.20. We simulated\\nn = 20 observations from the model\\nY = sin(X) + ϵ,\\nwhere X ∼U[−5, 5] (uniform distribution), and ϵ ∼N(0, σ2) with σ = 0.3.\\nWe then fit a natural spline to the data, as described in Section 7.4, with d 10.8 Interpolation and Double Descent\\n433\\n−4\\n−2\\n0\\n2\\n4\\n−3\\n−2\\n−1\\n0\\n1\\n2\\n3\\n8 Degrees of Freedom\\n−4\\n−2\\n0\\n2\\n4\\n−3\\n−2\\n−1\\n0\\n1\\n2\\n3\\n20 Degrees of Freedom\\n−4\\n−2\\n0\\n2\\n4\\n−3\\n−2\\n−1\\n0\\n1\\n2\\n3\\n42 Degrees of Freedom\\n−4\\n−2\\n0\\n2\\n4\\n−3\\n−2\\n−1\\n0\\n1\\n2\\n3\\n80 Degrees of Freedom\\nFIGURE 10.21. Fitted functions ˆfd(X) (orange), true function f(X) (black)\\nand the observed 20 training data points. A different value of d (degrees of freedom)\\nis used in each panel. For d ≥20 the orange curves all interpolate the training\\npoints, and hence the training error is zero.\\ndegrees of freedom.22 Recall from Section 7.4 that fitting a natural spline\\nwith d degrees of freedom amounts to fitting a least-squares regression\\nof the response onto a set of d basis functions. The upper-left panel of\\nFigure 10.21 shows the data, the true function f(X), and ˆf8(X), the fitted\\nnatural spline with d = 8 degrees of freedom.\\nNext, we fit a natural spline with d = 20 degrees of freedom. Since n = 20,\\nthis means that n = d, and we have zero training error; in other words, we\\nhave interpolated the training data! We can see from the top-right panel of\\nFigure 10.21 that ˆf20(X) makes wild excursions, and hence the test error\\nwill be large.\\nWe now continue to fit natural splines to the data, with increasing values\\nof d. For d > 20, the least squares regression of Y onto d basis functions\\nis not unique: there are an infinite number of least squares coefficient es-\\ntimates that achieve zero error. To select among them, we choose the one\\nwith the smallest sum of squared coefficients, )d\\nj=1 ˆβ2\\nj . This is known as\\nthe minimum-norm solution.\\nThe two lower panels of Figure 10.21 show the minimum-norm natural\\nspline fits with d = 42 and d = 80 degrees of freedom. Incredibly, ˆf42(X)\\nis quite a bit less less wild than ˆf20(X), even though it makes use of more\\ndegrees of freedom. And ˆf80(X) is not much different. How can this be?\\nEssentially, ˆf20(X) is very wild because there is just a single way to interpo-\\nlate n = 20 observations using d = 20 basis functions, and that single way\\nresults in a somewhat extreme fitted function. By contrast, there are an\\n22This implies the choice of d knots, here chosen at d equi-probability quantiles of the\\ntraining data. When d > n, the quantiles are found by interpolation. 434\\n10. Deep Learning\\ninfinite number of ways to interpolate n = 20 observations using d = 42 or\\nd = 80 basis functions, and the smoothest of them — that is, the minimum\\nnorm solution — is much less wild than ˆf20(X)!\\nIn Figure 10.20, we display the training error and test error associated\\nwith ˆfd(X), for a range of values of the degrees of freedom d. We see that\\nthe training error drops to zero once d = 20 and beyond; i.e. once the\\ninterpolation threshold is reached. By contrast, the test error shows a U-\\nshape for d ≤20, grows extremely large around d = 20, and then shows a\\nsecond region of descent for d > 20. For this example the signal-to-noise\\nratio — Var(f(X))/σ2 — is 5.9, which is quite high (the data points are\\nclose to the true curve). So an estimate that interpolates the data and does\\nnot wander too far inbetween the observed data points will likely do well.\\nIn Figures 10.20 and 10.21, we have illustrated the double descent phe-\\nnomenon in a simple one-dimensional setting using natural splines. How-\\never, it turns out that the same phenomenon can arise for deep learning.\\nBasically, when we fit neural networks with a huge number of parameters,\\nwe are sometimes able to get good results with zero training error. This is\\nparticularly true in problems with high signal-to-noise ratio, such as natural\\nimage recognition and language translation, for example. This is because\\nthe techniques used to fit neural networks, including stochastic gradient\\ndescent, naturally lend themselves to selecting a “smooth” interpolating\\nmodel that has good test-set performance on these kinds of problems.\\nSome points are worth emphasizing:\\n• The double-descent phenomenon does not contradict the bias-variance\\ntrade-off, as presented in Section 2.2.2. Rather, the double-descent\\ncurve seen in the right-hand side of Figure 10.20 is a consequence of\\nthe fact that the x-axis displays the number of spline basis functions\\nused, which does not properly capture the true “flexibility” of models\\nthat interpolate the training data. Stated another way, in this exam-\\nple, the minimum-norm natural spline with d = 42 has lower variance\\nthan the natural spline with d = 20.\\n• Most of the statistical learning methods seen in this book do not exhibit\\ndouble descent. For instance, regularization approaches typically do\\nnot interpolate the training data, and thus double descent does not\\noccur. This is not a drawback of regularized methods: they can give\\ngreat results without interpolating the data!\\nIn particular, in the examples here, if we had fit the natural splines\\nusing ridge regression with an appropriately-chosen penalty rather\\nthan least squares, then we would not have seen double descent, and\\nin fact would have obtained better test error results.\\n• In Chapter 9, we saw that maximal margin classifiers and SVMs that\\nhave zero training error nonetheless often achieve very good test error.\\nThis is in part because those methods seek smooth minimum norm\\nsolutions. This is similar to the fact that the minimum-norm natural\\nspline can give good results with zero training error.\\n• The double-descent phenomenon has been used by the machine learn-\\ning community to explain the successful practice of using an over- 10.9 Lab: Deep Learning\\n435\\nparametrized neural network (many layers, and many hidden units),\\nand then fitting all the way to zero training error. However, fitting\\nto zero error is not always optimal, and whether it is advisable de-\\npends on the signal-to-noise ratio. For instance, we may use ridge\\nregularization to avoid overfitting a neural network, as in (10.31). In\\nthis case, provided that we use an appropriate choice for the tuning\\nparameter λ, we will never interpolate the training data, and thus\\nwill not see the double descent phenomenon. Nonetheless we can get\\nvery good test-set performance, likely much better than we would\\nhave achieved had we interpolated the training data. Early stopping\\nduring stochastic gradient descent can also serve as a form of regular-\\nization that prevents us from interpolating the training data, while\\nstill getting very good results on test data.\\nTo summarize: though double descent can sometimes occur in neural net-\\nworks, we typically do not want to rely on this behavior. Moreover, it\\nis important to remember that the bias-variance trade-off always holds\\n(though it is possible that test error as a function of flexibility may not\\nexhibit a U-shape, depending on how we have parametrized the notion of\\n“flexibility” on the x-axis).\\n10.9\\nLab: Deep Learning\\nIn this section we demonstrate how to fit the examples discussed in the\\ntext. We use the Python torch package, along with the pytorch_lightning torch\\npytorch_\\nlightning\\npackage which provides utilities to simplify fitting and evaluating mod-\\nels. This code can be impressively fast with certain special processors,\\nsuch as Apple’s new M1 chip. The package is well-structured, flexible, and\\nwill feel comfortable to Python users. A good companion is the site py-\\ntorch.org/tutorials. Much of our code is adapted from there, as well as the\\npytorch_lightning documentation.23\\nWe start with several standard imports that we have seen before.\\nIn [1]: import numpy as np, pandas as pd\\nfrom matplotlib.pyplot import subplots\\nfrom sklearn.linear_model import \\\\\\n(LinearRegression ,\\nLogisticRegression ,\\nLasso)\\nfrom sklearn.preprocessing import StandardScaler\\nfrom sklearn.model_selection import KFold\\nfrom sklearn.pipeline import Pipeline\\nfrom ISLP import load_data\\nfrom ISLP.models import ModelSpec as MS\\nfrom sklearn.model_selection import \\\\\\n(train_test_split ,\\nGridSearchCV)\\n23The precise URLs at the time of writing are https://pytorch.org/tutorials/\\nbeginner/basics/intro.html and https://pytorch-lightning.readthedocs.io/en/\\nlatest/. 436\\n10. Deep Learning\\nTorch-Specific Imports\\nThere are a number of imports for torch. (These are not included with\\nISLP, so must be installed separately.) First we import the main library\\nand essential tools used to specify sequentially-structured networks.\\nIn [2]: import torch\\nfrom torch import nn\\nfrom torch.optim import RMSprop\\nfrom torch.utils.data import TensorDataset\\nThere are several other helper packages for torch. For instance, the\\ntorchmetrics package has utilities to compute various metrics to evalu- torchmetrics\\nate performance when fitting a model. The torchinfo package provides a torchinfo\\nuseful summary of the layers of a model. We use the read_image() function read_image()\\nwhen loading test images in Section 10.9.4.\\nIn [3]: from torchmetrics import (MeanAbsoluteError ,\\nR2Score)\\nfrom torchinfo import summary\\nfrom torchvision.io import read_image\\nThe package pytorch_lightning is a somewhat higher-level interface to\\ntorch that simplifies the specification and fitting of models by reducing the\\namount of boilerplate code needed (compared to using torch alone).\\nIn [4]: from pytorch_lightning import Trainer\\nfrom pytorch_lightning.loggers import CSVLogger\\nIn order to reproduce results we use seed_everything(). We will also seed_\\neverything()\\ninstruct torch to use deterministic algorithms where possible.\\nIn [5]: from pytorch_lightning.utilities.seed import seed_everything\\nseed_everything(0, workers=True)\\ntorch.use_deterministic_algorithms(True, warn_only=True)\\nWe will use several datasets shipped with torchvision for our examples: torchvision\\na pretrained network for image classification, as well as some transforms\\nused for preprocessing.\\nIn [6]: from torchvision.datasets import MNIST, CIFAR100\\nfrom torchvision.models import (resnet50,\\nResNet50_Weights)\\nfrom torchvision.transforms import (Resize,\\nNormalize,\\nCenterCrop ,\\nToTensor)\\nWe have provided a few utilities in ISLP specifically for this lab. The\\nSimpleDataModule and SimpleModule are simple versions of objects used\\nin pytorch_lightning, the high-level module for fitting torch models. Al-\\nthough more advanced uses such as computing on graphical processing\\nunits (GPUs) and parallel data processing are possible in this module, we\\nwill not be focusing much on these in this lab. The ErrorTracker handles\\ncollections of targets and predictions over each mini-batch in the validation\\nor test stage, allowing computation of the metric over the entire validation\\nor test data set. 10.9 Lab: Deep Learning\\n437\\nIn [7]: from ISLP.torch import (SimpleDataModule ,\\nSimpleModule ,\\nErrorTracker ,\\nrec_num_workers)\\nIn addition we have included some helper functions to load the IMDb\\ndatabase, as well as a lookup that maps integers to particular keys in the\\ndatabase. We’ve included a slightly modified copy of the preprocessed IMDb\\ndata from keras, a separate package for fitting deep learning models. This keras\\nsaves us significant preprocessing and allows us to focus on specifying and\\nfitting the models themselves.\\nIn [8]: from ISLP.torch.imdb import (load_lookup ,\\nload_tensor ,\\nload_sparse ,\\nload_sequential)\\nFinally, we introduce some utility imports not directly related to torch.\\nThe glob() function from the glob module is used to find all files matching glob()\\nwildcard characters, which we will use in our example applying the ResNet50\\nmodel to some of our own images. The json module will be used to load a json\\nJSON file for looking up classes to identify the labels of the pictures in the\\nResNet50 example.\\nIn [9]: from glob import glob\\nimport json\\n10.9.1\\nSingle Layer Network on Hitters Data\\nWe start by fitting the models in Section 10.6 on the Hitters data.\\nIn [10]: Hitters = load_data(\\'Hitters\\').dropna()\\nn = Hitters.shape[0]\\nWe will fit two linear models (least squares and lasso) and compare their\\nperformance to that of a neural network. For this comparison we will use\\nmean absolute error on a validation dataset.\\nMAE(y, ˆy) = 1\\nn\\nn\\n0\\ni=1\\n|yi −ˆyi|.\\nWe set up the model matrix and the response.\\nIn [11]: model = MS(Hitters.columns.drop(\\'Salary\\'), intercept=False)\\nX = model.fit_transform(Hitters).to_numpy()\\nY = Hitters[\\'Salary\\'].to_numpy()\\nThe to_numpy() method above converts pandas data frames or series to to_numpy()\\nnumpy arrays. We do this because we will need to use sklearn to fit the\\nlasso model, and it requires this conversion. We also use a linear regres-\\nsion method from sklearn, rather than the method in Chapter 3 from\\nstatsmodels, to facilitate the comparisons.\\nWe now split the data into test and training, fixing the random state\\nused by sklearn to do the split. 438\\n10. Deep Learning\\nIn [12]: (X_train,\\nX_test,\\nY_train,\\nY_test) = train_test_split(X,\\nY,\\ntest_size=1/3,\\nrandom_state=1)\\nLinear Models\\nWe fit the linear model and evaluate the test error directly.\\nIn [13]: hit_lm = LinearRegression().fit(X_train, Y_train)\\nYhat_test = hit_lm.predict(X_test)\\nnp.abs(Yhat_test - Y_test).mean()\\nOut[13]: 259.7153\\nNext we fit the lasso using sklearn. We are using mean absolute error\\nto select and evaluate a model, rather than mean squared error. The spe-\\ncialized solver we used in Section 6.5.2 uses only mean squared error. So\\nhere, with a bit more work, we create a cross-validation grid and perform\\nthe cross-validation directly.\\nWe encode a pipeline with two steps: we first normalize the features\\nusing a StandardScaler() transform, and then fit the lasso without further\\nnormalization.\\nIn [14]: scaler = StandardScaler(with_mean=True, with_std=True)\\nlasso = Lasso(warm_start=True, max_iter=30000)\\nstandard_lasso = Pipeline(steps=[(\\'scaler\\', scaler),\\n(\\'lasso\\', lasso)])\\nWe need to create a grid of values for λ. As is common practice, we\\nchoose a grid of 100 values of λ, uniform on the log scale from lam_max\\ndown to 0.01*lam_max. Here lam_max is the smallest value of λ with an all-\\nzero solution. This value equals the largest absolute inner-product between\\nany predictor and the (centered) response.24\\nIn [15]: X_s = scaler.fit_transform(X_train)\\nn = X_s.shape[0]\\nlam_max = np.fabs(X_s.T.dot(Y_train - Y_train.mean())).max() / n\\nparam_grid = {\\'alpha\\': np.exp(np.linspace(0, np.log(0.01), 100))\\n* lam_max}\\nNote that we had to transform the data first, since the scale of the vari-\\nables impacts the choice of λ. We now perform cross-validation using this\\nsequence of λ values.\\nIn [16]: cv = KFold(10,\\nshuffle=True,\\nrandom_state=1)\\ngrid = GridSearchCV(lasso,\\n24The derivation of this result is beyond the scope of this book. 10.9 Lab: Deep Learning\\n439\\nparam_grid ,\\ncv=cv,\\nscoring=\\'neg_mean_absolute_error\\')\\ngrid.fit(X_train, Y_train);\\nWe extract the lasso model with best cross-validated mean absolute error,\\nand evaluate its performance on X_test and Y_test, which were not used\\nin cross-validation.\\nIn [17]: trained_lasso = grid.best_estimator_\\nYhat_test = trained_lasso.predict(X_test)\\nnp.fabs(Yhat_test - Y_test).mean()\\nOut[17]: 257.2382\\nThis is similar to the results we got for the linear model fit by least squares.\\nHowever, these results can vary a lot for different train/test splits; we en-\\ncourage the reader to try a different seed in code block 12 and rerun the\\nsubsequent code up to this point.\\nSpecifying a Network: Classes and Inheritance\\nTo fit the neural network, we first set up a model structure that describes\\nthe network. Doing so requires us to define new classes specific to the model\\nwe wish to fit. Typically this is done in pytorch by sub-classing a generic\\nrepresentation of a network, which is the approach we take here. Although\\nthis example is simple, we will go through the steps in some detail, since it\\nwill serve us well for the more complex examples to follow.\\nIn [18]: class HittersModel(nn.Module):\\ndef __init__(self, input_size):\\nsuper(HittersModel , self).__init__()\\nself.flatten = nn.Flatten()\\nself.sequential = nn.Sequential(\\nnn.Linear(input_size, 50),\\nnn.ReLU(),\\nnn.Dropout(0.4),\\nnn.Linear(50, 1))\\ndef forward(self, x):\\nx = self.flatten(x)\\nreturn torch.flatten(self.sequential(x))\\nThe class statement identifies the code chunk as a declaration for a class\\nHittersModel that inherits from the base class nn.Module. This base class is\\nubiquitous in torch and represents the mappings in the neural networks.\\nIndented beneath the class statement are the methods of this class: in\\nthis case __init__ and forward. The __init__ method is called when an\\ninstance of the class is created as in the cell below. In the methods, self\\nalways refers to an instance of the class. In the __init__ method, we have\\nattached two objects to self as attributes: flatten and sequential. These\\nare used in the forward method to describe the map that this module\\nimplements. 440\\n10. Deep Learning\\nThere is one additional line in the __init__ method, which is a call to\\nsuper(). This function allows subclasses (i.e. HittersModel) to access meth- super()\\nods of the class they inherit from. For example, the class nn.Module has its\\nown __init__ method, which is different from the HittersModel.__init__()\\nmethod we’ve written above. Using super() allows us to call the method\\nof the base class. For torch models, we will always be making this super()\\ncall as it is necessary for the model to be properly interpreted by torch.\\nThe object nn.Module has more methods than simply __init__ and forward.\\nThese methods are directly accessible to HittersModel instances because of\\nthis inheritance. One such method we will see shortly is the eval() method,\\nused to disable dropout for when we want to evaluate the model on test\\ndata.\\nIn [19]: hit_model = HittersModel(X.shape[1])\\nThe object self.sequential is a composition of four maps. The first\\nmaps the 19 features of Hitters to 50 dimensions, introducing 50 × 19 + 50\\nparameters for the weights and intercept of the map (often called the bias).\\nThis layer is then mapped to a ReLU layer followed by a 40% dropout layer,\\nand finally a linear map down to 1 dimension, again with a bias. The total\\nnumber of trainable parameters is therefore 50 × 19 + 50 + 50 + 1 = 1051.\\nThe package torchinfo provides a summary() function that neatly sum-\\nmarizes this information. We specify the size of the input and see the size\\nof each tensor as it passes through layers of the network.\\nIn [20]: summary(hit_model,\\ninput_size=X_train.shape,\\ncol_names=[\\'input_size\\',\\n\\'output_size\\',\\n\\'num_params\\'])\\nOut[20]: =====================================================================\\nLayer (type:depth-idx)\\nInput Shape\\nOutput Shape\\nParam #\\n=====================================================================\\nHittersModel\\n[175, 19]\\n[175]\\n--\\nFlatten: 1-1\\n[175, 19]\\n[175, 19]\\n--\\nSequential: 1-2\\n[175, 19]\\n[175, 1]\\n--\\nLinear: 2-1\\n[175, 19]\\n[175, 50]\\n1,000\\nReLU: 2-2\\n[175, 50]\\n[175, 50]\\n--\\nDropout: 2-3\\n[175, 50]\\n[175, 50]\\n--\\nLinear: 2-4\\n[175, 50]\\n[175, 1]\\n51\\n=====================================================================\\nTotal params: 1,051\\nTrainable params: 1,051\\nWe have truncated the end of the output slightly, here and in subsequent\\nuses.\\nWe now need to transform our training data into a form accessible to\\ntorch. The basic datatype in torch is a tensor, which is very similar to\\nan ndarray from early chapters. We also note here that torch typically\\nworks with 32-bit (single precision) rather than 64-bit (double precision)\\nfloating point numbers. We therefore convert our data to np.float32 before\\nforming the tensor. The X and Y tensors are then arranged into a Dataset Dataset 10.9 Lab: Deep Learning\\n441\\nrecognized by torch using TensorDataset().\\nTensor\\nDataset()\\nIn [21]: X_train_t = torch.tensor(X_train.astype(np.float32))\\nY_train_t = torch.tensor(Y_train.astype(np.float32))\\nhit_train = TensorDataset(X_train_t, Y_train_t)\\nWe do the same for the test data.\\nIn [22]: X_test_t = torch.tensor(X_test.astype(np.float32))\\nY_test_t = torch.tensor(Y_test.astype(np.float32))\\nhit_test = TensorDataset(X_test_t, Y_test_t)\\nFinally, this dataset is passed to a DataLoader() which ultimately passes\\ndata into our network. While this may seem like a lot of overhead, this\\nstructure is helpful for more complex tasks where data may live on different\\nmachines, or where data must be passed to a GPU. We provide a helper\\nfunction SimpleDataModule() in ISLP to make this task easier for standard SimpleData\\nModule()\\nusage. One of its arguments is num_workers, which indicates how many\\nprocesses we will use for loading the data. For small data like Hitters\\nthis will have little effect, but it does provide an advantage for the MNIST\\nand CIFAR100 examples below. The torch package will inspect the process\\nrunning and determine a maximum number of workers.25 We’ve included a\\nfunction rec_num_workers() to compute this so we know how many workers\\nmight be reasonable (here the max was 16).\\nIn [23]: max_num_workers = rec_num_workers()\\nThe general training setup in pytorch_lightning involves training, vali-\\ndation and test data. These are each represented by different data loaders.\\nDuring each epoch, we run a training step to learn the model and a vali-\\ndation step to track the error. The test data is typically used at the end of\\ntraining to evaluate the model.\\nIn this case, as we had split only into test and training, we’ll use the\\ntest data as validation data with the argument validation=hit_test. The\\nvalidation argument can be a float between 0 and 1, an integer, or a\\nDataset. If a float (respectively, integer), it is interpreted as a percentage\\n(respectively number) of the training observations to be used for validation.\\nIf it is a Dataset, it is passed directly to a data loader.\\nIn [24]: hit_dm = SimpleDataModule(hit_train,\\nhit_test,\\nbatch_size=32,\\nnum_workers=min(4, max_num_workers),\\nvalidation=hit_test)\\nNext we must provide a pytorch_lightning module that controls the\\nsteps performed during the training process. We provide methods for our\\nSimpleModule() that simply record the value of the loss function and any\\nadditional metrics at the end of each epoch. These operations are controlled\\nby the methods SimpleModule.[training/test/validation]_step(), though\\nwe will not be modifying these in our examples.\\n25This depends on the computing hardware and the number of cores available. 442\\n10. Deep Learning\\nIn [25]: hit_module = SimpleModule.regression(hit_model,\\nmetrics={\\'mae\\':MeanAbsoluteError()})\\nBy using the SimpleModule.regression() method, we indicate that we SimpleModule.\\nregression()\\nwill use squared-error loss as in (10.23). We have also asked for mean ab-\\nsolute error to be tracked as well in the metrics that are logged.\\nWe log our results via CSVLogger(), which in this case stores the results\\nin a CSV file within a directory logs/hitters. After the fitting is complete,\\nthis allows us to load the results as a pd.DataFrame() and visualize them\\nbelow. There are several ways to log the results within pytorch_lightning,\\nthough we will not cover those here in detail.\\nIn [26]: hit_logger = CSVLogger(\\'logs\\', name=\\'hitters\\')\\nFinally we are ready to train our model and log the results. We use\\nthe Trainer() object from pytorch_lightning to do this work. The argu-\\nment datamodule=hit_dm tells the trainer how training/validation/test logs\\nare produced, while the first argument hit_module specifies the network\\narchitecture as well as the training/validation/test steps. The callbacks\\nargument allows for several tasks to be carried out at various points while\\ntraining a model. Here our ErrorTracker() callback will enable us to com-\\npute validation error while training and, finally, the test error. We now fit\\nthe model for 50 epochs.\\nIn [27]: hit_trainer = Trainer(deterministic=True,\\nmax_epochs=50,\\nlog_every_n_steps=5,\\nlogger=hit_logger,\\ncallbacks=[ErrorTracker()])\\nhit_trainer.fit(hit_module, datamodule=hit_dm)\\nAt each step of SGD, the algorithm randomly selects 32 training observa-\\ntions for the computation of the gradient. Recall from Section 10.7 that an\\nepoch amounts to the number of SGD steps required to process n observa-\\ntions. Since the training set has n = 175, and we specified a batch_size of\\n32 in the construction of hit_dm, an epoch is 175/32 = 5.5 SGD steps.\\nAfter having fit the model, we can evaluate performance on our test data\\nusing the test() method of our trainer.\\nIn [28]: hit_trainer.test(hit_module, datamodule=hit_dm)\\nOut[28]: [{\\'test_loss\\': 104098.5469, \\'test_mae\\': 229.5012}]\\nThe results of the fit have been logged into a CSV file. We can find the\\nresults specific to this run in the experiment.metrics_file_path attribute\\nof our logger. Note that each time the model is fit, the logger will output\\nresults into a new subdirectory of our directory logs/hitters.\\nWe now create a plot of the MAE (mean absolute error) as a function of\\nthe number of epochs. First we retrieve the logged summaries.\\nhit_results = pd.read_csv(hit_logger.experiment.metrics_file_path)\\nSince we will produce similar plots in later examples, we write a simple\\ngeneric function to produce this plot. 10.9 Lab: Deep Learning\\n443\\nIn [29]: def summary_plot(results,\\nax,\\ncol=\\'loss\\',\\nvalid_legend=\\'Validation\\',\\ntraining_legend=\\'Training\\',\\nylabel=\\'Loss\\',\\nfontsize=20):\\nfor (column,\\ncolor,\\nlabel) in zip([f\\'train_{col}_epoch\\',\\nf\\'valid_{col}\\'],\\n[\\'black\\',\\n\\'red\\'],\\n[training_legend ,\\nvalid_legend]):\\nresults.plot(x=\\'epoch\\',\\ny=column,\\nlabel=label,\\nmarker=\\'o\\',\\ncolor=color,\\nax=ax)\\nax.set_xlabel(\\'Epoch\\')\\nax.set_ylabel(ylabel)\\nreturn ax\\nWe now set up our axes, and use our function to produce the MAE plot.\\nIn [30]: fig, ax = subplots(1, 1, figsize=(6, 6))\\nax = summary_plot(hit_results ,\\nax,\\ncol=\\'mae\\',\\nylabel=\\'MAE\\',\\nvalid_legend=\\'Validation (=Test)\\')\\nax.set_ylim([0, 400])\\nax.set_xticks(np.linspace(0, 50, 11).astype(int));\\nWe can predict directly from the final model, and evaluate its per-\\nformance on the test data. Before fitting, we call the eval() method of\\nhit_model. This tells torch to effectively consider this model to be fitted,\\nso that we can use it to predict on new data. For our model here, the biggest\\nchange is that the dropout layers will be turned off, i.e. no weights will be\\nrandomly dropped in predicting on new data.\\nIn [31]: hit_model.eval()\\npreds = hit_module(X_test_t)\\ntorch.abs(Y_test_t - preds).mean()\\nOut[31]: tensor(229.5012, grad_fn=<MeanBackward0 >)\\nCleanup\\nIn setting up our data module, we had initiated several worker processes\\nthat will remain running. We delete all references to the torch objects to\\nensure these processes will be killed. 444\\n10. Deep Learning\\nIn [32]: del(Hitters,\\nhit_model, hit_dm,\\nhit_logger,\\nhit_test, hit_train,\\nX, Y,\\nX_test, X_train,\\nY_test, Y_train,\\nX_test_t, Y_test_t,\\nhit_trainer , hit_module)\\n10.9.2\\nMultilayer Network on the MNIST Digit Data\\nThe torchvision package comes with a number of example datasets, includ-\\ning the MNIST digit data. Our first step is to retrieve the training and test\\ndata sets; the MNIST() function within torchvision.datasets is provided for MNIST()\\nthis purpose. The data will be downloaded the first time this function is\\nexecuted, and stored in the directory data/MNIST.\\nIn [33]: (mnist_train ,\\nmnist_test) = [MNIST(root=\\'data\\',\\ntrain=train,\\ndownload=True,\\ntransform=ToTensor())\\nfor train in [True, False]]\\nmnist_train\\nOut[33]: Dataset MNIST\\nNumber of datapoints: 60000\\nRoot location: data\\nSplit: Train\\nStandardTransform\\nTransform: ToTensor()\\nThere are 60,000 images in the training data and 10,000 in the test data.\\nThe images are 28 × 28, and stored as a matrix of pixels. We need to\\ntransform each one into a vector.\\nNeural networks are somewhat sensitive to the scale of the inputs, much\\nas ridge and lasso regularization are affected by scaling. Here the inputs\\nare eight-bit grayscale values between 0 and 255, so we rescale to the unit\\ninterval.26 This transformation, along with some reordering of the axes, is\\nperformed by the ToTensor() transform from the torchvision.transforms\\npackage.\\nAs in our Hitters example, we form a data module from the training and\\ntest datasets, setting aside 20% of the training images for validation.\\nIn [34]: mnist_dm = SimpleDataModule(mnist_train ,\\nmnist_test ,\\nvalidation=0.2,\\nnum_workers=max_num_workers ,\\nbatch_size=256)\\n26Note: eight bits means 28, which equals 256. Since the convention is to start at 0,\\nthe possible values range from 0 to 255. 10.9 Lab: Deep Learning\\n445\\nLet’s take a look at the data that will get fed into our network. We loop\\nthrough the first few chunks of the test dataset, breaking after 2 batches:\\nIn [35]: for idx, (X_ ,Y_) in enumerate(mnist_dm.train_dataloader()):\\nprint(\\'X: \\', X_.shape)\\nprint(\\'Y: \\', Y_.shape)\\nif idx >= 1:\\nbreak\\nX:\\ntorch.Size([256, 1, 28, 28])\\nY:\\ntorch.Size([256])\\nX:\\ntorch.Size([256, 1, 28, 28])\\nY:\\ntorch.Size([256])\\nWe see that the X for each batch consists of 256 images of size 1x28x28.\\nHere the 1 indicates a single channel (greyscale). For RGB images such as\\nCIFAR100 below, we will see that the 1 in the size will be replaced by 3 for\\nthe three RGB channels.\\nNow we are ready to specify our neural network.\\nIn [36]: class MNISTModel(nn.Module):\\ndef __init__(self):\\nsuper(MNISTModel, self).__init__()\\nself.layer1 = nn.Sequential(\\nnn.Flatten(),\\nnn.Linear(28*28, 256),\\nnn.ReLU(),\\nnn.Dropout(0.4))\\nself.layer2 = nn.Sequential(\\nnn.Linear(256, 128),\\nnn.ReLU(),\\nnn.Dropout(0.3))\\nself._forward = nn.Sequential(\\nself.layer1,\\nself.layer2,\\nnn.Linear(128, 10))\\ndef forward(self, x):\\nreturn self._forward(x)\\nWe see that in the first layer, each 1x28x28 image is flattened, then\\nmapped to 256 dimensions where we apply a ReLU activation with 40%\\ndropout. A second layer maps the first layer’s output down to 128 di-\\nmensions, applying a ReLU activation with 30% dropout. Finally, the 128\\ndimensions are mapped down to 10, the number of classes in the MNIST\\ndata.\\nIn [37]: mnist_model = MNISTModel()\\nWe can check that the model produces output of expected size based on\\nour existing batch X_ above.\\nIn [38]: mnist_model(X_).size()\\nOut[38]: torch.Size([256, 10])\\nLet’s take a look at the summary of the model. Instead of an input_size\\nwe can pass a tensor of correct shape. In this case, we pass through the\\nfinal batched X_ from above. 446\\n10. Deep Learning\\nIn [39]: summary(mnist_model ,\\ninput_data=X_,\\ncol_names=[\\'input_size\\',\\n\\'output_size\\',\\n\\'num_params\\'])\\nOut[39]: =====================================================================\\nLayer (type:depth-idx)\\nInput Shape\\nOutput Shape\\nParam #\\n=====================================================================\\nMNISTModel\\n[256, 1, 28, 28]\\n[256, 10]\\n--\\nSequential: 1-1\\n[256, 1, 28, 28]\\n[256, 10]\\n--\\nSequential: 2-1\\n[256, 1, 28, 28]\\n[256, 256]\\n--\\nFlatten: 3-1\\n[256, 1, 28, 28]\\n[256, 784]\\n--\\nLinear: 3-2\\n[256, 784]\\n[256, 256]\\n200,960\\nReLU: 3-3\\n[256, 256]\\n[256, 256]\\n--\\nDropout: 3-4\\n[256, 256]\\n[256, 256]\\n--\\nSequential: 2-2\\n[256, 256]\\n[256, 128]\\n--\\nLinear: 3-5\\n[256, 256]\\n[256, 128]\\n32,896\\nReLU: 3-6\\n[256, 128]\\n[256, 128]\\n--\\nDropout: 3-7\\n[256, 128]\\n[256, 128]\\n--\\nLinear: 2-3\\n[256, 128]\\n[256, 10]\\n1,290\\n=====================================================================\\nTotal params: 235,146\\nTrainable params: 235,146\\nHaving set up both the model and the data module, fitting this model is\\nnow almost identical to the Hitters example. In contrast to our regression\\nmodel, here we will use the SimpleModule.classification() method which SimpleModule.\\nclassifi-\\ncation()\\nuses the cross-entropy loss function instead of mean squared error.\\nIn [40]: mnist_module = SimpleModule.classification(mnist_model)\\nmnist_logger = CSVLogger(\\'logs\\', name=\\'MNIST\\')\\nNow we are ready to go. The final step is to supply training data, and\\nfit the model.\\nIn [41]: mnist_trainer = Trainer(deterministic=True,\\nmax_epochs=30,\\nlogger=mnist_logger ,\\ncallbacks=[ErrorTracker()])\\nmnist_trainer.fit(mnist_module ,\\ndatamodule=mnist_dm)\\nWe have suppressed the output here, which is a progress report on the\\nfitting of the model, grouped by epoch. This is very useful, since on large\\ndatasets fitting can take time. Fitting this model took 245 seconds on a\\nMacBook Pro with an Apple M1 Pro chip with 10 cores and 16 GB of\\nRAM. Here we specified a validation split of 20%, so training is actually\\nperformed on 80% of the 60,000 observations in the training set. This is an\\nalternative to actually supplying validation data, like we did for the Hitters\\ndata. SGD uses batches of 256 observations in computing the gradient, and\\ndoing the arithmetic, we see that an epoch corresponds to 188 gradient\\nsteps.\\nSimpleModule.classification() includes an accuracy metric by default.\\nOther classification metrics can be added from torchmetrics. We will use\\nour summary_plot() function to display accuracy across epochs. 10.9 Lab: Deep Learning\\n447\\nIn [42]: mnist_results = pd.read_csv(mnist_logger.experiment.\\nmetrics_file_path)\\nfig, ax = subplots(1, 1, figsize=(6, 6))\\nsummary_plot(mnist_results ,\\nax,\\ncol=\\'accuracy\\',\\nylabel=\\'Accuracy\\')\\nax.set_ylim([0.5, 1])\\nax.set_ylabel(\\'Accuracy\\')\\nax.set_xticks(np.linspace(0, 30, 7).astype(int));\\nOnce again we evaluate the accuracy using the test() method of our\\ntrainer. This model achieves 97% accuracy on the test data.\\nIn [43]: mnist_trainer.test(mnist_module ,\\ndatamodule=mnist_dm)\\nOut[43]: [{\\'test_loss\\': 0.1471, \\'test_accuracy\\': 0.9681}]\\nTable 10.1 also reports the error rates resulting from LDA (Chapter 4) and\\nmulticlass logistic regression. For LDA we refer the reader to Section 4.7.3.\\nAlthough we could use the sklearn function LogisticRegression() to fit\\nmulticlass logistic regression, we are set up here to fit such a model with\\ntorch. We just have an input layer and an output layer, and omit the hidden\\nlayers!\\nIn [44]: class MNIST_MLR(nn.Module):\\ndef __init__(self):\\nsuper(MNIST_MLR, self).__init__()\\nself.linear = nn.Sequential(nn.Flatten(),\\nnn.Linear(784, 10))\\ndef forward(self, x):\\nreturn self.linear(x)\\nmlr_model = MNIST_MLR()\\nmlr_module = SimpleModule.classification(mlr_model)\\nmlr_logger = CSVLogger(\\'logs\\', name=\\'MNIST_MLR\\')\\nIn [45]: mlr_trainer = Trainer(deterministic=True,\\nmax_epochs=30,\\ncallbacks=[ErrorTracker()])\\nmlr_trainer.fit(mlr_module, datamodule=mnist_dm)\\nWe fit the model just as before and compute the test results.\\nIn [46]: mlr_trainer.test(mlr_module,\\ndatamodule=mnist_dm)\\nOut[46]: [{\\'test_loss\\': 0.3187, \\'test_accuracy\\': 0.9241}]\\nThe accuracy is above 90% even for this pretty simple model.\\nAs in the Hitters example, we delete some of the objects we created\\nabove.\\nIn [47]: del(mnist_test,\\nmnist_train , 448\\n10. Deep Learning\\nmnist_model ,\\nmnist_dm,\\nmnist_trainer ,\\nmnist_module ,\\nmnist_results ,\\nmlr_model,\\nmlr_module,\\nmlr_trainer)\\n10.9.3\\nConvolutional Neural Networks\\nIn this section we fit a CNN to the CIFAR100 data, which is available in the\\ntorchvision package. It is arranged in a similar fashion as the MNIST data.\\nIn [48]: (cifar_train ,\\ncifar_test) = [CIFAR100(root=\"data\",\\ntrain=train,\\ndownload=True)\\nfor train in [True, False]]\\nIn [49]: transform = ToTensor()\\ncifar_train_X = torch.stack([transform(x) for x in\\ncifar_train.data])\\ncifar_test_X = torch.stack([transform(x) for x in\\ncifar_test.data])\\ncifar_train = TensorDataset(cifar_train_X ,\\ntorch.tensor(cifar_train.targets))\\ncifar_test = TensorDataset(cifar_test_X ,\\ntorch.tensor(cifar_test.targets))\\nThe CIFAR100 dataset consists of 50,000 training images, each represented\\nby a three-dimensional tensor: each three-color image is represented as a\\nset of three channels, each of which consists of 32 × 32 eight-bit pixels. We\\nstandardize as we did for the digits, but keep the array structure. This is\\naccomplished with the ToTensor() transform.\\nCreating the data module is similar to the MNIST example.\\nIn [50]: cifar_dm = SimpleDataModule(cifar_train ,\\ncifar_test ,\\nvalidation=0.2,\\nnum_workers=max_num_workers ,\\nbatch_size=128)\\nWe again look at the shape of typical batches in our data loaders.\\nIn [51]: for idx, (X_ ,Y_) in enumerate(cifar_dm.train_dataloader()):\\nprint(\\'X: \\', X_.shape)\\nprint(\\'Y: \\', Y_.shape)\\nif idx >= 1:\\nbreak\\nX:\\ntorch.Size([128, 3, 32, 32])\\nY:\\ntorch.Size([128])\\nX:\\ntorch.Size([128, 3, 32, 32])\\nY:\\ntorch.Size([128]) 10.9 Lab: Deep Learning\\n449\\nBefore we start, we look at some of the training images; similar code\\nproduced Figure 10.5 on page 406. The example below also illustrates that\\nTensorDataset objects can be indexed with integers — we are choosing ran-\\ndom images from the training data by indexing cifar_train. In order to dis-\\nplay correctly, we must reorder the dimensions by a call to np.transpose().\\nIn [52]: fig, axes = subplots(5, 5, figsize=(10,10))\\nrng = np.random.default_rng(4)\\nindices = rng.choice(np.arange(len(cifar_train)), 25,\\nreplace=False).reshape((5,5))\\nfor i in range(5):\\nfor j in range(5):\\nidx = indices[i,j]\\naxes[i,j].imshow(np.transpose(cifar_train[idx][0],\\n[1,2,0]),\\ninterpolation=None)\\naxes[i,j].set_xticks([])\\naxes[i,j].set_yticks([])\\nHere the imshow() method recognizes from the shape of its argument that .imshow()\\nit is a 3-dimensional array, with the last dimension indexing the three RGB\\ncolor channels.\\nWe specify a moderately-sized CNN for demonstration purposes, simi-\\nlar in structure to Figure 10.8. We use several layers, each consisting of\\nconvolution, ReLU, and max-pooling steps. We first define a module that\\ndefines one of these layers. As in our previous examples, we overwrite the\\n__init__() and forward() methods of nn.Module. This user-defined module\\ncan now be used in ways just like nn.Linear() or nn.Dropout().\\nIn [53]: class BuildingBlock(nn.Module):\\ndef __init__(self,\\nin_channels ,\\nout_channels):\\nsuper(BuildingBlock , self).__init__()\\nself.conv = nn.Conv2d(in_channels=in_channels ,\\nout_channels=out_channels ,\\nkernel_size=(3,3),\\npadding=\\'same\\')\\nself.activation = nn.ReLU()\\nself.pool = nn.MaxPool2d(kernel_size=(2,2))\\ndef forward(self, x):\\nreturn self.pool(self.activation(self.conv(x)))\\nNotice that we used the padding = \"same\" argument to nn.Conv2d(),\\nwhich ensures that the output channels have the same dimension as the\\ninput channels. There are 32 channels in the first hidden layer, in contrast\\nto the three channels in the input layer. We use a 3 × 3 convolution fil-\\nter for each channel in all the layers. Each convolution is followed by a\\nmax-pooling layer over 2 × 2 blocks.\\nIn forming our deep learning model for the CIFAR100 data, we use several\\nof our BuildingBlock() modules sequentially. This simple example illus-\\ntrates some of the power of torch. Users can define modules of their own, 450\\n10. Deep Learning\\nwhich can be combined in other modules. Ultimately, everything is fit by\\na generic trainer.\\nIn [54]: class CIFARModel(nn.Module):\\ndef __init__(self):\\nsuper(CIFARModel, self).__init__()\\nsizes = [(3,32),\\n(32,64),\\n(64,128),\\n(128,256)]\\nself.conv = nn.Sequential(*[BuildingBlock(in_, out_)\\nfor in_, out_ in sizes])\\nself.output = nn.Sequential(nn.Dropout(0.5),\\nnn.Linear(2*2*256, 512),\\nnn.ReLU(),\\nnn.Linear(512, 100))\\ndef forward(self, x):\\nval = self.conv(x)\\nval = torch.flatten(val, start_dim=1)\\nreturn self.output(val)\\nWe build the model and look at the summary. (We had created examples\\nof X_ earlier.)\\nIn [55]: cifar_model = CIFARModel()\\nsummary(cifar_model ,\\ninput_data=X_,\\ncol_names=[\\'input_size\\',\\n\\'output_size\\',\\n\\'num_params\\'])\\nOut[55]: ======================================================================\\nLayer (type:depth-idx)\\nInput Shape\\nOutput Shape\\nParam #\\n======================================================================\\nCIFARModel\\n[128, 3, 32, 32]\\n[128, 100]\\n--\\nSequential: 1-1\\n[128, 3, 32, 32]\\n[128, 256, 2, 2]\\n--\\nBuildingBlock: 2-1\\n[128, 3, 32, 32]\\n[128, 32, 16, 16]\\n--\\nConv2d: 3-1\\n[128, 3, 32, 32]\\n[128, 32, 32, 32]\\n896\\nReLU: 3-2\\n[128, 32, 32, 32]\\n[128, 32, 32, 32]\\n--\\nMaxPool2d: 3-3\\n[128, 32, 32, 32]\\n[128, 32, 16, 16]\\n--\\nBuildingBlock: 2-2\\n[128, 32, 16, 16]\\n[128, 64, 8, 8]\\n--\\nConv2d: 3-4\\n[128, 32, 16, 16]\\n[128, 64, 16, 16]\\n18,496\\nReLU: 3-5\\n[128, 64, 16, 16]\\n[128, 64, 16, 16]\\n--\\nMaxPool2d: 3-6\\n[128, 64, 16, 16]\\n[128, 64, 8, 8]\\n--\\nBuildingBlock: 2-3\\n[128, 64, 8, 8]\\n[128, 128, 4, 4]\\n--\\nConv2d: 3-7\\n[128, 64, 8, 8]\\n[128, 128, 8, 8]\\n73,856\\nReLU: 3-8\\n[128, 128, 8, 8]\\n[128, 128, 8, 8]\\n--\\nMaxPool2d: 3-9\\n[128, 128, 8, 8]\\n[128, 128, 4, 4]\\n--\\nBuildingBlock: 2-4\\n[128, 128, 4, 4]\\n[128, 256, 2, 2]\\n--\\nConv2d: 3-10\\n[128, 128, 4, 4]\\n[128, 256, 4, 4]\\n295,168\\nReLU: 3-11\\n[128, 256, 4, 4]\\n[128, 256, 4, 4]\\n--\\nMaxPool2d: 3-12 [128, 256, 4, 4]\\n[128, 256, 2, 2]\\n--\\nSequential: 1-2\\n[128, 1024]\\n[128, 100]\\n--\\nDropout: 2-5\\n[128, 1024]\\n[128, 1024]\\n--\\nLinear: 2-6\\n[128, 1024]\\n[128, 512]\\n524,800 10.9 Lab: Deep Learning\\n451\\nReLU: 2-7\\n[128, 512]\\n[128, 512]\\n--\\nLinear: 2-8\\n[128, 512]\\n[128, 100]\\n51,300\\n======================================================================\\nTotal params: 964,516\\nTrainable params: 964,516\\nThe total number of trainable parameters is 964,516. By studying the size\\nof the parameters, we can see that the channels halve in both dimensions\\nafter each of these max-pooling operations. After the last of these we have\\na layer with 256 channels of dimension 2 × 2. These are then flattened to\\na dense layer of size 1,024; in other words, each of the 2 × 2 matrices is\\nturned into a 4-vector, and put side-by-side in one layer. This is followed\\nby a dropout regularization layer, then another dense layer of size 512, and\\nfinally, the output layer.\\nUp to now, we have been using a default optimizer in SimpleModule(). For\\nthese data, experiments show that a smaller learning rate performs better\\nthan the default 0.01. We use a custom optimizer here with a learning rate\\nof 0.001. Besides this, the logging and training follow a similar pattern\\nto our previous examples. The optimizer takes an argument params that\\ninforms the optimizer which parameters are involved in SGD (stochastic\\ngradient descent).\\nWe saw earlier that entries of a module’s parameters are tensors. In\\npassing the parameters to the optimizer we are doing more than simply\\npassing arrays; part of the structure of the graph is encoded in the tensors\\nthemselves.\\nIn [56]: cifar_optimizer = RMSprop(cifar_model.parameters(), lr=0.001)\\ncifar_module = SimpleModule.classification(cifar_model ,\\noptimizer=cifar_optimizer)\\ncifar_logger = CSVLogger(\\'logs\\', name=\\'CIFAR100\\')\\nIn [57]: cifar_trainer = Trainer(deterministic=True,\\nmax_epochs=30,\\nlogger=cifar_logger ,\\ncallbacks=[ErrorTracker()])\\ncifar_trainer.fit(cifar_module ,\\ndatamodule=cifar_dm)\\nThis model takes 10 minutes or more to run and achieves about 42%\\naccuracy on the test data. Although this is not terrible for 100-class data\\n(a random classifier gets 1% accuracy), searching the web we see results\\naround 75%. Typically it takes a lot of architecture carpentry, fiddling with\\nregularization, and time, to achieve such results.\\nLet’s take a look at the validation and training accuracy across epochs.\\nIn [58]: log_path = cifar_logger.experiment.metrics_file_path\\ncifar_results = pd.read_csv(log_path)\\nfig, ax = subplots(1, 1, figsize=(6, 6))\\nsummary_plot(cifar_results ,\\nax,\\ncol=\\'accuracy\\',\\nylabel=\\'Accuracy\\')\\nax.set_xticks(np.linspace(0, 10, 6).astype(int))\\nax.set_ylabel(\\'Accuracy\\')\\nax.set_ylim([0, 1]); 452\\n10. Deep Learning\\nFinally, we evaluate our model on our test data.\\nIn [59]: cifar_trainer.test(cifar_module ,\\ndatamodule=cifar_dm)\\nOut[59]: [{\\'test_loss\\': 2.4238 \\'test_accuracy\\': 0.4206}]\\nHardware Acceleration\\nAs deep learning has become ubiquitous in machine learning, hardware\\nmanufacturers have produced special libraries that can often speed up the\\ngradient-descent steps.\\nFor instance, Mac OS devices with the M1 chip may have the Metal pro-\\ngramming framework enabled, which can speed up the torch computations.\\nWe present an example of how to use this acceleration.\\nThe main changes are to the Trainer() call as well as to the metrics that\\nwill be evaluated on the data. These metrics must be told where the data\\nwill be located at evaluation time. This is accomplished with a call to the\\nto() method of the metrics.\\nIn [60]: try:\\nfor name, metric in cifar_module.metrics.items():\\ncifar_module.metrics[name] = metric.to(\\'mps\\')\\ncifar_trainer_mps = Trainer(accelerator=\\'mps\\',\\ndeterministic=True,\\nmax_epochs=30)\\ncifar_trainer_mps.fit(cifar_module ,\\ndatamodule=cifar_dm)\\ncifar_trainer_mps.test(cifar_module ,\\ndatamodule=cifar_dm)\\nexcept:\\npass\\nThis yields approximately two- or three-fold acceleration for each epoch.\\nWe have protected this code block using try: and except: clauses; if it\\nworks, we get the speedup, if it fails, nothing happens.\\n10.9.4\\nUsing Pretrained CNN Models\\nWe now show how to use a CNN pretrained on the imagenet database to\\nclassify natural images, and demonstrate how we produced Figure 10.10.\\nWe copied six JPEG images from a digital photo album into the direc-\\ntory book_images. These images are available from the data section of www.\\nstatlearning.com, the ISLP book website. Download book_images.zip;\\nwhen clicked it creates the book_images directory.\\nThe pretrained network we use is called resnet50; specification details\\ncan be found on the web. We will read in the images, and convert them into\\nthe array format expected by the torch software to match the specifications\\nin resnet50. The conversion involves a resize, a crop and then a predefined\\nstandardization for each of the three channels. We now read in the images\\nand preprocess them. 10.9 Lab: Deep Learning\\n453\\nIn [61]: resize = Resize((232,232))\\ncrop = CenterCrop(224)\\nnormalize = Normalize([0.485,0.456,0.406],\\n[0.229,0.224,0.225])\\nimgfiles = sorted([f for f in glob(\\'book_images/*\\')])\\nimgs = torch.stack([torch.div(crop(resize(read_image(f))), 255)\\nfor f in imgfiles])\\nimgs = normalize(imgs)\\nimgs.size()\\nOut[61]: torch.Size([6, 3, 224, 224])\\nWe now set up the trained network with the weights we read in code\\nblock 6. The model has 50 layers, with a fair bit of complexity.\\nIn [62]: resnet_model = resnet50(weights=ResNet50_Weights.DEFAULT)\\nsummary(resnet_model ,\\ninput_data=imgs,\\ncol_names=[\\'input_size\\',\\n\\'output_size\\',\\n\\'num_params\\'])\\nWe set the mode to eval() to ensure that the model is ready to predict on\\nnew data.\\nIn [63]: resnet_model.eval()\\nInspecting the output above, we see that when setting up the resnet_model,\\nthe authors defined a Bottleneck, much like our BuildingBlock module.\\nWe now feed our six images through the fitted network.\\nIn [64]: img_preds = resnet_model(imgs)\\nLet’s look at the predicted probabilities for each of the top 3 choices.\\nFirst we compute the probabilities by applying the softmax to the logits\\nin img_preds. Note that we have had to call the detach() method on the\\ntensor img_preds in order to convert it to our a more familiar ndarray.\\nIn [65]: img_probs = np.exp(np.asarray(img_preds.detach()))\\nimg_probs /= img_probs.sum(1)[:,None]\\nIn order to see the class labels, we must download the index file associated\\nwith imagenet.27\\nIn [66]: labs = json.load(open(\\'imagenet_class_index.json\\'))\\nclass_labels = pd.DataFrame([(int(k), v[1]) for k, v in\\nlabs.items()],\\ncolumns=[\\'idx\\', \\'label\\'])\\nclass_labels = class_labels.set_index(\\'idx\\')\\nclass_labels = class_labels.sort_index()\\nWe’ll now construct a data frame for each image file with the labels with\\nthe three highest probabilities as estimated by the model above.\\n27This is avalable from the book website and s3.amazonaws.com/deep-learning-\\nmodels/image-models/imagenet_class_index.json. 454\\n10. Deep Learning\\nIn [67]: for i, imgfile in enumerate(imgfiles):\\nimg_df = class_labels.copy()\\nimg_df[\\'prob\\'] = img_probs[i]\\nimg_df = img_df.sort_values(by=\\'prob\\', ascending=False)[:3]\\nprint(f\\'Image: {imgfile}\\')\\nprint(img_df.reset_index().drop(columns=[\\'idx\\']))\\nImage: book_images/Cape_Weaver.jpg\\nlabel\\nprob\\n0\\njacamar\\n0.287283\\n1\\nbee_eater\\n0.046768\\n2\\nbulbul\\n0.037507\\nImage: book_images/Flamingo.jpg\\nlabel\\nprob\\n0\\nflamingo\\n0.591761\\n1\\nspoonbill\\n0.012386\\n2\\nAmerican_egret\\n0.002105\\nImage: book_images/Hawk_Fountain.jpg\\nlabel\\nprob\\n0\\ngreat_grey_owl\\n0.287959\\n1\\nkite\\n0.039478\\n2\\nfountain\\n0.029384\\nImage: book_images/Hawk_cropped.jpg\\nlabel\\nprob\\n0\\nkite\\n0.301830\\n1\\njay\\n0.121674\\n2\\nmagpie\\n0.015513\\nImage: book_images/Lhasa_Apso.jpg\\nlabel\\nprob\\n0\\nLhasa\\n0.151143\\n1\\nShih-Tzu\\n0.129850\\n2\\nTibetan_terrier\\n0.102358\\nImage: book_images/Sleeping_Cat.jpg\\nlabel\\nprob\\n0\\ntabby\\n0.173627\\n1\\ntiger_cat\\n0.110414\\n2\\ndoormat\\n0.093447\\nWe see that the model is quite confident about Flamingo.jpg, but a little\\nless so for the other images.\\nWe end this section with our usual cleanup.\\nIn [68]: del(cifar_test,\\ncifar_train ,\\ncifar_dm,\\ncifar_module ,\\ncifar_logger ,\\ncifar_optimizer ,\\ncifar_trainer)\\n10.9.5\\nIMDB Document Classification\\nWe now implement models for sentiment classification (Section 10.4) on\\nthe IMDB dataset. As mentioned above code block 8, we are using a prepro-\\ncessed version of the IMDB dataset found in the keras package. As keras uses 10.9 Lab: Deep Learning\\n455\\ntensorflow, a different tensor and deep learning library, we have converted\\nthe data to be suitable for torch. The code used to convert from keras\\nis available in the module ISLP.torch._make_imdb. It requires some of the\\nkeras packages to run. These data use a dictionary of size 10,000.\\nWe have stored three different representations of the review data for this\\nlab:\\n• load_tensor(), a sparse tensor version usable by torch;\\n• load_sparse(), a sparse matrix version usable by sklearn, since we\\nwill compare with a lasso fit;\\n• load_sequential(), a padded version of the original sequence repre-\\nsentation, limited to the last 500 words of each review.\\nIn [69]: (imdb_seq_train ,\\nimdb_seq_test) = load_sequential(root=\\'data/IMDB\\')\\npadded_sample = np.asarray(imdb_seq_train.tensors[0][0])\\nsample_review = padded_sample[padded_sample > 0][:12]\\nsample_review[:12]\\nOut[69]: array([\\n1,\\n14,\\n22,\\n16,\\n43,\\n530,\\n973, 1622, 1385,\\n65,\\n458, 4468], dtype=int32)\\nThe datasets imdb_seq_train and imdb_seq_test are both instances of the\\nclass TensorDataset. The tensors used to construct them can be found in\\nthe tensors attribute, with the first tensor the features X and the second\\nthe outcome Y. We have taken the first row of features and stored it as\\npadded_sample. In the preprocessing used to form these data, sequences\\nwere padded with 0s in the beginning if they were not long enough, hence\\nwe remove this padding by restricting to entries where padded_sample > 0.\\nWe then provide the first 12 words of the sample review.\\nWe can find these words in the lookup dictionary from the ISLP.torch.imdb\\nmodule.\\nIn [70]: lookup = load_lookup(root=\\'data/IMDB\\')\\n\\' \\'.join(lookup[i] for i in sample_review)\\nOut[70]: \"<START> this film was just brilliant casting location scenery\\nstory direction everyone\\'s\"\\nFor our first model, we have created a binary feature for each of the\\n10,000 possible words in the dataset, with an entry of one in the i, j entry\\nif word j appears in review i. As most reviews are quite short, such a feature\\nmatrix has over 98% zeros. These data are accessed using load_tensor()\\nfrom the ISLP library.\\nIn [71]: max_num_workers=10\\n(imdb_train ,\\nimdb_test) = load_tensor(root=\\'data/IMDB\\')\\nimdb_dm = SimpleDataModule(imdb_train,\\nimdb_test,\\nvalidation=2000,\\nnum_workers=min(6, max_num_workers),\\nbatch_size=512) 456\\n10. Deep Learning\\nWe’ll use a two-layer model for our first model.\\nIn [72]: class IMDBModel(nn.Module):\\ndef __init__(self, input_size):\\nsuper(IMDBModel, self).__init__()\\nself.dense1 = nn.Linear(input_size, 16)\\nself.activation = nn.ReLU()\\nself.dense2 = nn.Linear(16, 16)\\nself.output = nn.Linear(16, 1)\\ndef forward(self, x):\\nval = x\\nfor _map in [self.dense1,\\nself.activation,\\nself.dense2,\\nself.activation,\\nself.output]:\\nval = _map(val)\\nreturn torch.flatten(val)\\nWe now instantiate our model and look at a summary (not shown).\\nIn [73]: imdb_model = IMDBModel(imdb_test.tensors[0].size()[1])\\nsummary(imdb_model,\\ninput_size=imdb_test.tensors[0].size(),\\ncol_names=[\\'input_size\\',\\n\\'output_size\\',\\n\\'num_params\\'])\\nWe’ll again use a smaller learning rate for these data, hence we pass an\\noptimizer to the SimpleModule. Since the reviews are classified into positive\\nor negative sentiment, we use SimpleModule.binary_classification().28\\nIn [74]: imdb_optimizer = RMSprop(imdb_model.parameters(), lr=0.001)\\nimdb_module = SimpleModule.binary_classification(\\nimdb_model ,\\noptimizer=imdb_optimizer)\\nHaving loaded the datasets into a data module and created a SimpleModule,\\nthe remaining steps are familiar.\\nIn [75]: imdb_logger = CSVLogger(\\'logs\\', name=\\'IMDB\\')\\nimdb_trainer = Trainer(deterministic=True,\\nmax_epochs=30,\\nlogger=imdb_logger ,\\ncallbacks=[ErrorTracker()])\\nimdb_trainer.fit(imdb_module ,\\ndatamodule=imdb_dm)\\nEvaluating the test error yields roughly 86% accuracy.\\nIn [76]: test_results = imdb_trainer.test(imdb_module , datamodule=imdb_dm)\\ntest_results\\n28Our use of binary_classification() instead of classification() is due to\\nsome subtlety in how torchmetrics.Accuracy() works, as well as the data type of\\nthe targets. 10.9 Lab: Deep Learning\\n457\\nOut[76]: [{\\'test_loss\\': 1.0863, \\'test_accuracy\\': 0.8550}]\\nComparison to Lasso\\nWe now fit a lasso logistic regression model using LogisticRegression()\\nfrom sklearn. Since sklearn does not recognize the sparse tensors of torch,\\nwe use a sparse matrix that is recognized by sklearn.\\nIn [77]: ((X_train, Y_train),\\n(X_valid, Y_valid),\\n(X_test, Y_test)) = load_sparse(validation=2000,\\nrandom_state=0,\\nroot=\\'data/IMDB\\')\\nSimilar to what we did in Section 10.9.1, we construct a series of 50\\nvalues for the lasso reguralization parameter λ.\\nIn [78]: lam_max = np.abs(X_train.T * (Y_train - Y_train.mean())).max()\\nlam_val = lam_max * np.exp(np.linspace(np.log(1),\\nnp.log(1e-4), 50))\\nWith LogisticRegression() the regularization parameter C is specified as\\nthe inverse of λ. There are several solvers for logistic regression; here we\\nuse liblinear which works well with the sparse input format.\\nIn [79]: logit = LogisticRegression(penalty=\\'l1\\',\\nC=1/lam_max,\\nsolver=\\'liblinear\\',\\nwarm_start=True,\\nfit_intercept=True)\\nThe path of 50 values takes approximately 40 seconds to run.\\nIn [80]: coefs = []\\nintercepts = []\\nfor l in lam_val:\\nlogit.C = 1/l\\nlogit.fit(X_train, Y_train)\\ncoefs.append(logit.coef_.copy())\\nintercepts.append(logit.intercept_)\\nThe coefficient and intercepts have an extraneous dimension which can\\nbe removed by the np.squeeze() function.\\nIn [81]: coefs = np.squeeze(coefs)\\nintercepts = np.squeeze(intercepts)\\nWe’ll now make a plot to compare our neural network results with the\\nlasso.\\nIn [82]: %%capture\\nfig, axes = subplots(1, 2, figsize=(16, 8), sharey=True)\\nfor ((X_, Y_),\\ndata_,\\ncolor) in zip([(X_train, Y_train),\\n(X_valid, Y_valid),\\n(X_test, Y_test)], 458\\n10. Deep Learning\\n[\\'Training\\', \\'Validation\\', \\'Test\\'],\\n[\\'black\\', \\'red\\', \\'blue\\']):\\nlinpred_ = X_ * coefs.T + intercepts[None,:]\\nlabel_ = np.array(linpred_ > 0)\\naccuracy_ = np.array([np.mean(Y_ == l) for l in label_.T])\\naxes[0].plot(-np.log(lam_val / X_train.shape[0]),\\naccuracy_,\\n\\'.--\\',\\ncolor=color,\\nmarkersize=13,\\nlinewidth=2,\\nlabel=data_)\\naxes[0].legend()\\naxes[0].set_xlabel(r\\'$-\\\\log(\\\\lambda)$\\', fontsize=20)\\naxes[0].set_ylabel(\\'Accuracy\\', fontsize=20)\\nNotice the use of %%capture, which suppresses the displaying of the partially %%capture\\ncompleted figure. This is useful when making a complex figure, since the\\nsteps can be spread across two or more cells. We now add a plot of the lasso\\naccuracy, and display the composed figure by simply entering its name at\\nthe end of the cell.\\nIn [83]: imdb_results = pd.read_csv(imdb_logger.experiment.metrics_file_path)\\nsummary_plot(imdb_results ,\\naxes[1],\\ncol=\\'accuracy\\',\\nylabel=\\'Accuracy\\')\\naxes[1].set_xticks(np.linspace(0, 30, 7).astype(int))\\naxes[1].set_ylabel(\\'Accuracy\\', fontsize=20)\\naxes[1].set_xlabel(\\'Epoch\\', fontsize=20)\\naxes[1].set_ylim([0.5, 1]);\\naxes[1].axhline(test_results[0][\\'test_accuracy\\'],\\ncolor=\\'blue\\',\\nlinestyle=\\'--\\',\\nlinewidth=3)\\nfig\\nFrom the graphs we see that the accuracy of the lasso logistic regression\\npeaks at about 0.88, as it does for the neural network.\\nOnce again, we end with a cleanup.\\nIn [84]: del(imdb_model,\\nimdb_trainer ,\\nimdb_logger ,\\nimdb_dm,\\nimdb_train,\\nimdb_test)\\n10.9.6\\nRecurrent Neural Networks\\nIn this lab we fit the models illustrated in Section 10.5.\\nSequential Models for Document Classification\\nHere we fit a simple LSTM RNN for sentiment prediction to the IMDb\\nmovie-review data, as discussed in Section 10.5.1. For an RNN we use 10.9 Lab: Deep Learning\\n459\\nthe sequence of words in a document, taking their order into account. We\\nloaded the preprocessed data at the beginning of Section 10.9.5. A script\\nthat details the preprocessing can be found in the ISLP library. Notably,\\nsince more than 90% of the documents had fewer than 500 words, we set\\nthe document length to 500. For longer documents, we used the last 500\\nwords, and for shorter documents, we padded the front with blanks.\\nIn [85]: imdb_seq_dm = SimpleDataModule(imdb_seq_train ,\\nimdb_seq_test ,\\nvalidation=2000,\\nbatch_size=300,\\nnum_workers=min(6, max_num_workers)\\n)\\nThe first layer of the RNN is an embedding layer of size 32, which will\\nbe learned during training. This layer one-hot encodes each document as a\\nmatrix of dimension 500 × 10, 003, and then maps these 10, 003 dimensions\\ndown to 32.29 Since each word is represented by an integer, this is effectively\\nachieved by the creation of an embedding matrix of size 10, 003 × 32; each\\nof the 500 integers in the document are then mapped to the appropriate\\n32 real numbers by indexing the appropriate rows of this matrix.\\nThe second layer is an LSTM with 32 units, and the output layer is a\\nsingle logit for the binary classification task. In the last line of the forward()\\nmethod below, we take the last 32-dimensional output of the LSTM and\\nmap it to our response.\\nIn [86]: class LSTMModel(nn.Module):\\ndef __init__(self, input_size):\\nsuper(LSTMModel, self).__init__()\\nself.embedding = nn.Embedding(input_size, 32)\\nself.lstm = nn.LSTM(input_size=32,\\nhidden_size=32,\\nbatch_first=True)\\nself.dense = nn.Linear(32, 1)\\ndef forward(self, x):\\nval, (h_n, c_n) = self.lstm(self.embedding(x))\\nreturn torch.flatten(self.dense(val[:,-1]))\\nWe instantiate and take a look at the summary of the model, using the\\nfirst 10 documents in the corpus.\\nIn [87]: lstm_model = LSTMModel(X_test.shape[-1])\\nsummary(lstm_model,\\ninput_data=imdb_seq_train.tensors[0][:10],\\ncol_names=[\\'input_size\\',\\n\\'output_size\\',\\n\\'num_params\\'])\\nOut[87]: ====================================================================\\nLayer (type:depth-idx)\\nInput Shape\\nOutput Shape\\nParam #\\n====================================================================\\nLSTMModel\\n[10, 500]\\n[10]\\n--\\n29The extra 3 dimensions correspond to commonly occurring non-word entries in the\\nreviews. 460\\n10. Deep Learning\\nEmbedding: 1-1\\n[10, 500]\\n[10, 500, 32]\\n320,096\\nLSTM: 1-2\\n[10, 500, 32]\\n[10, 500, 32]\\n8,448\\nLinear: 1-3\\n[10, 32]\\n[10, 1]\\n33\\n====================================================================\\nTotal params: 328,577\\nTrainable params: 328,577\\nThe 10,003 is suppressed in the summary, but we see it in the parameter\\ncount, since 10, 003 × 32 = 320, 096.\\nIn [88]: lstm_module = SimpleModule.binary_classification(lstm_model)\\nlstm_logger = CSVLogger(\\'logs\\', name=\\'IMDB_LSTM\\')\\nIn [89]: lstm_trainer = Trainer(deterministic=True,\\nmax_epochs=20,\\nlogger=lstm_logger ,\\ncallbacks=[ErrorTracker()])\\nlstm_trainer.fit(lstm_module ,\\ndatamodule=imdb_seq_dm)\\nThe rest is now similar to other networks we have fit. We track the test\\nperformance as the network is fit, and see that it attains 85% accuracy.\\nIn [90]: lstm_trainer.test(lstm_module , datamodule=imdb_seq_dm)\\nOut[90]: [{\\'test_loss\\': 0.8178, \\'test_accuracy\\': 0.8476}]\\nWe once again show the learning progress, followed by cleanup.\\nIn [91]: lstm_results = pd.read_csv(lstm_logger.experiment.metrics_file_path)\\nfig, ax = subplots(1, 1, figsize=(6, 6))\\nsummary_plot(lstm_results ,\\nax,\\ncol=\\'accuracy\\',\\nylabel=\\'Accuracy\\')\\nax.set_xticks(np.linspace(0, 20, 5).astype(int))\\nax.set_ylabel(\\'Accuracy\\')\\nax.set_ylim([0.5, 1])\\nIn [92]: del(lstm_model,\\nlstm_trainer ,\\nlstm_logger ,\\nimdb_seq_dm ,\\nimdb_seq_train ,\\nimdb_seq_test)\\nTime Series Prediction\\nWe now show how to fit the models in Section 10.5.2 for time series pre-\\ndiction. We first load and standardize the data.\\nIn [93]: NYSE = load_data(\\'NYSE\\')\\ncols = [\\'DJ_return\\', \\'log_volume\\', \\'log_volatility\\']\\nX = pd.DataFrame(StandardScaler(\\nwith_mean=True,\\nwith_std=True).fit_transform(NYSE[cols]),\\ncolumns=NYSE[cols].columns,\\nindex=NYSE.index) 10.9 Lab: Deep Learning\\n461\\nNext we set up the lagged versions of the data, dropping any rows with\\nmissing values using the dropna() method.\\nIn [94]: for lag in range(1, 6):\\nfor col in cols:\\nnewcol = np.zeros(X.shape[0]) * np.nan\\nnewcol[lag:] = X[col].values[:-lag]\\nX.insert(len(X.columns), \"{0}_{1}\".format(col, lag), newcol)\\nX.insert(len(X.columns), \\'train\\', NYSE[\\'train\\'])\\nX = X.dropna()\\nFinally, we extract the response, training indicator, and drop the current\\nday’s DJ_return and log_volatility to predict only from previous day’s\\ndata.\\nIn [95]: Y, train = X[\\'log_volume\\'], X[\\'train\\']\\nX = X.drop(columns=[\\'train\\'] + cols)\\nX.columns\\nOut[95]: Index([\\'DJ_return_1\\', \\'log_volume_1\\', \\'log_volatility_1\\',\\n\\'DJ_return_2\\', \\'log_volume_2\\', \\'log_volatility_2\\',\\n\\'DJ_return_3\\', \\'log_volume_3\\', \\'log_volatility_3\\',\\n\\'DJ_return_4\\', \\'log_volume_4\\', \\'log_volatility_4\\',\\n\\'DJ_return_5\\', \\'log_volume_5\\', \\'log_volatility_5\\'],\\ndtype=\\'object\\')\\nWe first fit a simple linear model and compute the R2 on the test data\\nusing the score() method.\\nIn [96]: M = LinearRegression()\\nM.fit(X[train], Y[train])\\nM.score(X[∼train], Y[∼train])\\nOut[96]: 0.4129\\nWe refit this model, including the factor variable day_of_week. For a cate-\\ngorical series in pandas, we can form the indicators using the get_dummies()\\nmethod.\\nIn [97]: X_day = pd.merge(X,\\npd.get_dummies(NYSE[\\'day_of_week\\']),\\non=\\'date\\')\\nNote that we do not have to reinstantiate the linear regression model as its\\nfit() method accepts a design matrix and a response directly.\\nIn [98]: M.fit(X_day[train], Y[train])\\nM.score(X_day[∼train], Y[∼train])\\nOut[98]: 0.4595\\nThis model achieves an R2 of about 46%.\\nTo fit the RNN, we must reshape the data, as it will expect 5 lagged\\nversions of each feature as indicated by the input_shape argument to the\\nlayer nn.RNN() below. We first ensure the columns of our data frame are\\nsuch that a reshaped matrix will have the variables correctly lagged. We\\nuse the reindex() method to do this. 462\\n10. Deep Learning\\nFor an input shape (5,3), each row represents a lagged version of the\\nthree variables. The nn.RNN() layer also expects the first row of each obser-\\nvation to be earliest in time, so we must reverse the current order. Hence\\nwe loop over range(5,0,-1) below, which is an example of using a slice()\\nto index iterable objects. The general notation is start:end:step.\\nIn [99]: ordered_cols = []\\nfor lag in range(5,0,-1):\\nfor col in cols:\\nordered_cols.append(\\'{0}_{1}\\'.format(col, lag))\\nX = X.reindex(columns=ordered_cols)\\nX.columns\\nOut[99]: Index([\\'DJ_return_5\\', \\'log_volume_5\\', \\'log_volatility_5\\',\\n\\'DJ_return_4\\', \\'log_volume_4\\', \\'log_volatility_4\\',\\n\\'DJ_return_3\\', \\'log_volume_3\\', \\'log_volatility_3\\',\\n\\'DJ_return_2\\', \\'log_volume_2\\', \\'log_volatility_2\\',\\n\\'DJ_return_1\\', \\'log_volume_1\\', \\'log_volatility_1\\'],\\ndtype=\\'object\\')\\nWe now reshape the data.\\nIn [100]: X_rnn = X.to_numpy().reshape((-1,5,3))\\nX_rnn.shape\\nOut[100]: (6046, 5, 3)\\nBy specifying the first size as -1, numpy.reshape() deduces its size based on\\nthe remaining arguments.\\nNow we are ready to proceed with the RNN, which uses 12 hidden units,\\nand 10% dropout. After passing through the RNN, we extract the final\\ntime point as val[:,-1] in forward() below. This gets passed through a\\n10% dropout and then flattened through a linear layer.\\nIn [101]: class NYSEModel(nn.Module):\\ndef __init__(self):\\nsuper(NYSEModel, self).__init__()\\nself.rnn = nn.RNN(3,\\n12,\\nbatch_first=True)\\nself.dense = nn.Linear(12, 1)\\nself.dropout = nn.Dropout(0.1)\\ndef forward(self, x):\\nval, h_n = self.rnn(x)\\nval = self.dense(self.dropout(val[:,-1]))\\nreturn torch.flatten(val)\\nnyse_model = NYSEModel()\\nWe fit the model in a similar fashion to previous networks. We supply\\nthe fit function with test data as validation data, so that when we monitor\\nits progress and plot the history function we can see the progress on the\\ntest data. Of course we should not use this as a basis for early stopping,\\nsince then the test performance would be biased.\\nWe form the training dataset similar to our Hitters example. 10.9 Lab: Deep Learning\\n463\\nIn [102]: datasets = []\\nfor mask in [train, ∼train]:\\nX_rnn_t = torch.tensor(X_rnn[mask].astype(np.float32))\\nY_t = torch.tensor(Y[mask].astype(np.float32))\\ndatasets.append(TensorDataset(X_rnn_t, Y_t))\\nnyse_train, nyse_test = datasets\\nFollowing our usual pattern, we inspect the summary.\\nIn [103]: summary(nyse_model,\\ninput_data=X_rnn_t,\\ncol_names=[\\'input_size\\',\\n\\'output_size\\',\\n\\'num_params\\'])\\nOut[103]: ====================================================================\\nLayer (type:depth-idx)\\nInput Shape\\nOutput Shape\\nParam #\\n====================================================================\\nNYSEModel\\n[1770, 5, 3]\\n[1770]\\n--\\nRNN: 1-1\\n[1770, 5, 3]\\n[1770, 5, 12]\\n204\\nDropout: 1-2\\n[1770, 12]\\n[1770, 12]\\n--\\nLinear: 1-3\\n[1770, 12]\\n[1770, 1]\\n13\\n====================================================================\\nTotal params: 217\\nTrainable params: 217\\nWe again put the two datasets into a data module, with a batch size of 64.\\nIn [104]: nyse_dm = SimpleDataModule(nyse_train,\\nnyse_test,\\nnum_workers=min(4, max_num_workers),\\nvalidation=nyse_test,\\nbatch_size=64)\\nWe run some data through our model to be sure the sizes match up cor-\\nrectly.\\nIn [105]: for idx, (x, y) in enumerate(nyse_dm.train_dataloader()):\\nout = nyse_model(x)\\nprint(y.size(), out.size())\\nif idx >= 2:\\nbreak\\ntorch.Size([64]) torch.Size([64])\\ntorch.Size([64]) torch.Size([64])\\ntorch.Size([64]) torch.Size([64])\\nWe follow our previous example for setting up a trainer for a regression\\nproblem, requesting the R2 metric to be be computed at each epoch.\\nIn [106]: nyse_optimizer = RMSprop(nyse_model.parameters(),\\nlr=0.001)\\nnyse_module = SimpleModule.regression(nyse_model,\\noptimizer=nyse_optimizer ,\\nmetrics={\\'r2\\':R2Score()})\\nFitting the model should by now be familiar. The results on the test data\\nare very similar to the linear AR model. 464\\n10. Deep Learning\\nIn [107]: nyse_trainer = Trainer(deterministic=True,\\nmax_epochs=200,\\ncallbacks=[ErrorTracker()])\\nnyse_trainer.fit(nyse_module ,\\ndatamodule=nyse_dm)\\nnyse_trainer.test(nyse_module ,\\ndatamodule=nyse_dm)\\nOut[107]: [{\\'test_loss\\': 0.6141, \\'test_r2\\': 0.4172}]\\nWe could also fit a model without the nn.RNN() layer by just using a\\nnn.Flatten() layer instead. This would be a nonlinear AR model. If in\\naddition we excluded the hidden layer, this would be equivalent to our\\nearlier linear AR model.\\nInstead we will fit a nonlinear AR model using the feature set X_day that\\nincludes the day_of_week indicators. To do so, we must first create our test\\nand training datasets and a corresponding data module. This may seem a\\nlittle burdensome, but is part of the general pipeline for torch.\\nIn [108]: datasets = []\\nfor mask in [train, ∼train]:\\nX_day_t = torch.tensor(\\nnp.asarray(X_day[mask]).astype(np.float32))\\nY_t = torch.tensor(np.asarray(Y[mask]).astype(np.float32))\\ndatasets.append(TensorDataset(X_day_t, Y_t))\\nday_train, day_test = datasets\\nCreating a data module follows a familiar pattern.\\nIn [109]: day_dm = SimpleDataModule(day_train,\\nday_test,\\nnum_workers=min(4, max_num_workers),\\nvalidation=day_test,\\nbatch_size=64)\\nWe build a NonLinearARModel() that takes as input the 20 features and\\na hidden layer with 32 units. The remaining steps are familiar.\\nIn [110]: class NonLinearARModel(nn.Module):\\ndef __init__(self):\\nsuper(NonLinearARModel , self).__init__()\\nself._forward = nn.Sequential(nn.Flatten(),\\nnn.Linear(20, 32),\\nnn.ReLU(),\\nnn.Dropout(0.5),\\nnn.Linear(32, 1))\\ndef forward(self, x):\\nreturn torch.flatten(self._forward(x))\\nIn [111]: nl_model = NonLinearARModel()\\nnl_optimizer = RMSprop(nl_model.parameters(),\\nlr=0.001)\\nnl_module = SimpleModule.regression(nl_model,\\noptimizer=nl_optimizer ,\\nmetrics={\\'r2\\':R2Score()}) 10.10 Exercises\\n465\\nWe continue with the usual training steps, fit the model, and evaluate\\nthe test error. We see the test R2 is a slight improvement over the linear\\nAR model that also includes day_of_week.\\nIn [112]: nl_trainer = Trainer(deterministic=True,\\nmax_epochs=20,\\ncallbacks=[ErrorTracker()])\\nnl_trainer.fit(nl_module, datamodule=day_dm)\\nnl_trainer.test(nl_module, datamodule=day_dm)\\nOut[112]: [{\\'test_loss\\': 0.5625, \\'test_r2\\': 0.4662}]\\n10.10\\nExercises\\nConceptual\\n1. Consider a neural network with two hidden layers: p = 4 input units,\\n2 units in the first hidden layer, 3 units in the second hidden layer,\\nand a single output.\\n(a) Draw a picture of the network, similar to Figures 10.1 or 10.4.\\n(b) Write out an expression for f(X), assuming ReLU activation\\nfunctions. Be as explicit as you can!\\n(c) Now plug in some values for the coefficients and write out the\\nvalue of f(X).\\n(d) How many parameters are there?\\n2. Consider the softmax function in (10.13) (see also (4.13) on page 145)\\nfor modeling multinomial probabilities.\\n(a) In (10.13), show that if we add a constant c to each of the zℓ,\\nthen the probability is unchanged.\\n(b) In (4.13), show that if we add constants cj, j = 0, 1, . . . , p, to\\neach of the corresponding coefficients for each of the classes, then\\nthe predictions at any new point x are unchanged.\\nThis shows that the softmax function is over-parametrized. However, over-\\nparametrized\\nregularization and SGD typically constrain the solutions so that this\\nis not a problem.\\n3. Show that the negative multinomial log-likelihood (10.14) is equiva-\\nlent to the negative log of the likelihood expression (4.5) when there\\nare M = 2 classes.\\n4. Consider a CNN that takes in 32 × 32 grayscale images and has a\\nsingle convolution layer with three 5 × 5 convolution filters (without\\nboundary padding).\\n(a) Draw a sketch of the input and first hidden layer similar to\\nFigure 10.8. 466\\n10. Deep Learning\\n(b) How many parameters are in this model?\\n(c) Explain how this model can be thought of as an ordinary feed-\\nforward neural network with the individual pixels as inputs, and\\nwith constraints on the weights in the hidden units. What are\\nthe constraints?\\n(d) If there were no constraints, then how many weights would there\\nbe in the ordinary feed-forward neural network in (c)?\\n5. In Table 10.2 on page 426, we see that the ordering of the three\\nmethods with respect to mean absolute error is different from the\\nordering with respect to test set R2. How can this be?\\nApplied\\n6. Consider the simple function R(β) = sin(β) + β/10.\\n(a) Draw a graph of this function over the range β ∈[−6, 6].\\n(b) What is the derivative of this function?\\n(c) Given β0 = 2.3, run gradient descent to find a local minimum\\nof R(β) using a learning rate of ρ = 0.1. Show each of β0, β1, . . .\\nin your plot, as well as the final answer.\\n(d) Repeat with β0 = 1.4.\\n7. Fit a neural network to the Default data. Use a single hidden layer\\nwith 10 units, and dropout regularization. Have a look at Labs 10.9.1–\\n10.9.2 for guidance. Compare the classification performance of your\\nmodel with that of linear logistic regression.\\n8. From your collection of personal photographs, pick 10 images of an-\\nimals (such as dogs, cats, birds, farm animals, etc.). If the subject\\ndoes not occupy a reasonable part of the image, then crop the image.\\nNow use a pretrained image classification CNN as in Lab 10.9.4 to\\npredict the class of each of your images, and report the probabilities\\nfor the top five predicted classes for each image.\\n9. Fit a lag-5 autoregressive model to the NYSE data, as described in\\nthe text and Lab 10.9.6. Refit the model with a 12-level factor repre-\\nsenting the month. Does this factor improve the performance of the\\nmodel?\\n10. In Section 10.9.6, we showed how to fit a linear AR model to the\\nNYSE data using the LinearRegression() function. However, we also\\nmentioned that we can “flatten” the short sequences produced for\\nthe RNN model in order to fit a linear AR model. Use this latter\\napproach to fit a linear AR model to the NYSE data. Compare the test\\nR2 of this linear AR model to that of the linear AR model that we fit\\nin the lab. What are the advantages/disadvantages of each approach?\\n11. Repeat the previous exercise, but now fit a nonlinear AR model by\\n“flattening” the short sequences produced for the RNN model. 10.10 Exercises\\n467\\n12. Consider the RNN fit to the NYSE data in Section 10.9.6. Modify the\\ncode to allow inclusion of the variable day_of_week, and fit the RNN.\\nCompute the test R2.\\n13. Repeat the analysis of Lab 10.9.5 on the IMDb data using a similarly\\nstructured neural network. We used 16 hidden units at each of two\\nhidden layers. Explore the effect of increasing this to 32 and 64 units\\nper layer, with and without 30% dropout regularization. 11\\nSurvival Analysis and Censored Data\\nIn this chapter, we will consider the topics of survival analysis and censored survival\\nanalysis\\ndata. These arise in the analysis of a unique kind of outcome variable: the\\ncensored\\ndata\\ntime until an event occurs.\\nFor example, suppose that we have conducted a five-year medical study,\\nin which patients have been treated for cancer. We would like to fit a model\\nto predict patient survival time, using features such as baseline health mea-\\nsurements or type of treatment. At first pass, this may sound like a regres-\\nsion problem of the kind discussed in Chapter 3. But there is an important\\ncomplication: hopefully some or many of the patients have survived until\\nthe end of the study. Such a patient’s survival time is said to be censored: we\\nknow that it is at least five years, but we do not know its true value. We do\\nnot want to discard this subset of surviving patients, as the fact that they\\nsurvived at least five years amounts to valuable information. However, it is\\nnot clear how to make use of this information using the techniques covered\\nthus far in this textbook.\\nThough the phrase “survival analysis” evokes a medical study, the ap-\\nplications of survival analysis extend far beyond medicine. For example,\\nconsider a company that wishes to model churn, the process by which cus-\\ntomers cancel subscription to a service. The company might collect data on\\ncustomers over some time period, in order to model each customer’s time\\nto cancellation as a function of demographics or other predictors. However,\\npresumably not all customers will have canceled their subscription by the\\nend of this time period; for such customers, the time to cancellation is\\ncensored.\\nIn fact, survival analysis is relevant even in application areas that are\\nunrelated to time. For instance, suppose we wish to model a person’s weight\\nas a function of some covariates, using a dataset with measurements for a\\nlarge number of people. Unfortunately, the scale used to weigh those people\\nis unable to report weights above a certain number. Then, any weights that\\n© Springer Nature Switzerland AG 2023 \\nG. James et al., An Introduction to Statistical Learning, Springer Texts in Statistics, \\nhttps://doi.org/10.1007/978-3-031-38747-0_11 \\n469 470\\n11. Survival Analysis and Censored Data\\nexceed that number are censored. The survival analysis methods presented\\nin this chapter could be used to analyze this dataset.\\nSurvival analysis is a very well-studied topic within statistics, due to its\\ncritical importance in a variety of applications, both in and out of medicine.\\nHowever, it has received relatively little attention in the machine learning\\ncommunity.\\n11.1\\nSurvival and Censoring Times\\nFor each individual, we suppose that there is a true survival time, T, as well survival time\\nas a true censoring time, C. (The survival time is also known as the failure censoring\\ntime\\ntime or the event time.) The survival time represents the time at which the\\nfailure time\\nevent time\\nevent of interest occurs: for instance, the time at which the patient dies,\\nor the customer cancels his or her subscription. By contrast, the censoring\\ntime is the time at which censoring occurs: for example, the time at which\\nthe patient drops out of the study or the study ends.\\nWe observe either the survival time T or else the censoring time C.\\nSpecifically, we observe the random variable\\nY = min(T, C).\\n(11.1)\\nIn other words, if the event occurs before censoring (i.e. T < C) then we\\nobserve the true survival time T; however, if censoring occurs before the\\nevent (T > C) then we observe the censoring time. We also observe a status\\nindicator,\\nδ =\\n=\\n1\\nif T ≤C\\n0\\nif T > C.\\nThus, δ = 1 if we observe the true survival time, and δ = 0 if we instead\\nobserve the censoring time.\\nNow, suppose we observe n (Y, δ) pairs, which we denote as (y1, δ1), . . . ,\\n(yn, δn). Figure 11.1 displays an example from a (fictitious) medical study\\nin which we observe n = 4 patients for a 365-day follow-up period. For\\npatients 1 and 3, we observe the time to event (such as death or disease\\nrelapse) T = ti. Patient 2 was alive when the study ended, and patient 4\\ndropped out of the study, or was “lost to follow-up”; for these patients we\\nobserve C = ci. Therefore, y1 = t1, y3 = t3, y2 = c2, y4 = c4, δ1 = δ3 = 1,\\nand δ2 = δ4 = 0.\\n11.2\\nA Closer Look at Censoring\\nIn order to analyze survival data, we need to make some assumptions about\\nwhy censoring has occurred. For instance, suppose that a number of patients\\ndrop out of a cancer study early because they are very sick. An analysis that\\ndoes not take into consideration the reason why the patients dropped out\\nwill likely overestimate the true average survival time. Similarly, suppose\\nthat males who are very sick are more likely to drop out of the study than 11.2 A Closer Look at Censoring\\n471\\n0\\n100\\n200\\n300\\n1\\n2\\n3\\n4\\nTime in Days\\nPatient\\nFIGURE 11.1. Illustration of censored survival data. For patients 1 and 3, the\\nevent was observed. Patient 2 was alive when the study ended. Patient 4 dropped\\nout of the study.\\nfemales who are very sick. Then a comparison of male and female survival\\ntimes may wrongly suggest that males survive longer than females.\\nIn general, we need to assume that the censoring mechanism is indepen-\\ndent: conditional on the features, the event time T is independent of the\\ncensoring time C. The two examples above violate the assumption of inde-\\npendent censoring. Typically, it is not possible to determine from the data\\nitself whether the censoring mechanism is independent. Instead, one has to\\ncarefully consider the data collection process in order to determine whether\\nindependent censoring is a reasonable assumption. In the remainder of this\\nchapter, we will assume that the censoring mechanism is independent.1\\nIn this chapter, we focus on right censoring, which occurs when T ≥Y ,\\ni.e. the true event time T is at least as large as the observed time Y .\\n(Notice that T ≥Y is a consequence of (11.1). Right censoring derives its\\nname from the fact that time is typically displayed from left to right, as in\\nFigure 11.1.) However, other types of censoring are possible. For instance,\\nin left censoring, the true event time T is less than or equal to the observed\\ntime Y . For example, in a study of pregnancy duration, suppose that we\\nsurvey patients 250 days after conception, when some have already had\\ntheir babies. Then we know that for those patients, pregnancy duration is\\nless than 250 days. More generally, interval censoring refers to the setting\\nin which we do not know the exact event time, but we know that it falls\\nin some interval. For instance, this setting arises if we survey patients once\\nper week in order to determine whether the event has occurred. While left\\ncensoring and interval censoring can be accommodated using variants of\\nthe ideas presented in this chapter, in what follows we focus specifically on\\nright censoring.\\n1The assumption of independent censoring can be relaxed somewhat using the notion\\nof non-informative censoring; however, the definition of non-informative censoring is too\\ntechnical for this book. 472\\n11. Survival Analysis and Censored Data\\n11.3\\nThe Kaplan–Meier Survival Curve\\nThe survival curve, or survival function, is defined as\\nsurvival\\ncurve\\nsurvival\\nfunction\\nS(t) = Pr(T > t).\\n(11.2)\\nThis decreasing function quantifies the probability of surviving past time\\nt. For example, suppose that a company is interested in modeling customer\\nchurn. Let T represent the time that a customer cancels a subscription to\\nthe company’s service. Then S(t) represents the probability that a customer\\ncancels later than time t. The larger the value of S(t), the less likely that\\nthe customer will cancel before time t.\\nIn this section, we will consider the task of estimating the survival\\ncurve. Our investigation is motivated by the BrainCancer dataset, which\\ncontains the survival times for patients with primary brain tumors un-\\ndergoing treatment with stereotactic radiation methods.2 The predictors\\nare gtv (gross tumor volume, in cubic centimeters); sex (male or female);\\ndiagnosis (meningioma, LG glioma, HG glioma, or other); loc (the tumor\\nlocation: either infratentorial or supratentorial); ki (Karnofsky index); and\\nstereo (stereotactic method: either stereotactic radiosurgery or fraction-\\nated stereotactic radiotherapy, abbreviated as SRS and SRT, respectively).\\nOnly 53 of the 88 patients were still alive at the end of the study.\\nNow, we consider the task of estimating the survival curve (11.2) for\\nthese data. To estimate S(20) = Pr(T > 20), the probability that a patient\\nsurvives for at least t = 20 months, it is tempting to simply compute the\\nproportion of patients who are known to have survived past 20 months, i.e.\\nthe proportion of patients for whom Y > 20. This turns out to be 48/88,\\nor approximately 55%. However, this does not seem quite right, since Y\\nand T represent different quantities. In particular, 17 of the 40 patients\\nwho did not survive to 20 months were actually censored, and this analysis\\nimplicitly assumes that T < 20 for all of those censored patients; of course,\\nwe do not know whether that is true.\\nAlternatively, to estimate S(20), we could consider computing the pro-\\nportion of patients for whom Y > 20, out of the 71 patients who were not\\ncensored by time t = 20; this comes out to 48/71, or approximately 68%.\\nHowever, this is not quite right either, since it amounts to completely ig-\\nnoring the patients who were censored before time t = 20, even though the\\ntime at which they are censored is potentially informative. For instance, a\\npatient who was censored at time t = 19.9 likely would have survived past\\nt = 20 had he or she not been censored.\\nWe have seen that estimating S(t) is complicated by the presence of\\ncensoring. We now present an approach to overcome these challenges. We\\nlet d1 < d2 < · · · < dK denote the K unique death times among the non-\\ncensored patients, and we let qk denote the number of patients who died\\nat time dk. For k = 1, . . . , K, we let rk denote the number of patients alive\\n2This dataset is described in the following paper: Selingerová et al. (2016) Survival\\nof patients with primary brain tumors: Comparison of two statistical approaches. PLoS\\nOne, 11(2):e0148733. 11.3 The Kaplan–Meier Survival Curve\\n473\\nand in the study just before dk; these are the at risk patients. The set of\\npatients that are at risk at a given time are referred to as the risk set.\\nrisk set\\nBy the law of total probability,3\\nPr(T > dk) = Pr(T > dk|T > dk−1) Pr(T > dk−1)\\n+ Pr(T > dk|T ≤dk−1) Pr(T ≤dk−1).\\nThe fact that dk−1 < dk implies that Pr(T > dk|T ≤dk−1) = 0 (it is\\nimpossible for a patient to survive past time dk if he or she did not survive\\nuntil an earlier time dk−1). Therefore,\\nS(dk) = Pr(T > dk) = Pr(T > dk|T > dk−1) Pr(T > dk−1).\\nPlugging in (11.2) again, we see that\\nS(dk) = Pr(T > dk|T > dk−1)S(dk−1).\\nThis implies that\\nS(dk) = Pr(T > dk|T > dk−1) × · · · × Pr(T > d2|T > d1) Pr(T > d1).\\nWe now must simply plug in estimates of each of the terms on the right-\\nhand side of the previous equation. It is natural to use the estimator\\n6\\nPr(T > dj|T > dj−1) = (rj −qj)/rj,\\nwhich is the fraction of the risk set at time dj who survived past time dj.\\nThis leads to the Kaplan–Meier estimator of the survival curve:\\nKaplan–\\nMeier\\nestimator\\nWS(dk) =\\nk\\nE\\nj=1\\n*rj −qj\\nrj\\n+\\n.\\n(11.3)\\nFor times t between dk and dk+1, we set WS(t) = WS(dk). Consequently, the\\nKaplan–Meier survival curve has a step-like shape.\\nThe Kaplan–Meier survival curve for the BrainCancer data is displayed\\nin Figure 11.2. Each point in the solid step-like curve shows the estimated\\nprobability of surviving past the time indicated on the horizontal axis. The\\nestimated probability of survival past 20 months is 71%, which is quite a\\nbit higher than the naive estimates of 55% and 68% presented earlier.\\nThe sequential construction of the Kaplan–Meier estimator — starting\\nat time zero and mapping out the observed events as they unfold in time —\\nis fundamental to many of the key techniques in survival analysis. These\\ninclude the log-rank test of Section 11.4, and Cox’s proportional hazard\\nmodel of Section 11.5.2.\\n3The law of total probability states that for any two events A and B, Pr(A) =\\nPr(A|B) Pr(B) + Pr(A|Bc) Pr(Bc), where Bc is the complement of the event B, i.e. it\\nis the event that B does not hold. 474\\n11. Survival Analysis and Censored Data\\n0\\n20\\n40\\n60\\n80\\n0.0\\n0.2\\n0.4\\n0.6\\n0.8\\n1.0\\nMonths\\nEstimated Probability of Survival\\nFIGURE 11.2. For the BrainCancer data, we display the Kaplan–Meier survival\\ncurve (solid curve), along with standard error bands (dashed curves).\\n0\\n20\\n40\\n60\\n80\\n0.0\\n0.2\\n0.4\\n0.6\\n0.8\\n1.0\\nMonths\\nEstimated Probability of Survival\\nFemale\\nMale\\nFIGURE 11.3. For the BrainCancer data, Kaplan–Meier survival curves for\\nmales and females are displayed.\\n11.4\\nThe Log-Rank Test\\nWe now continue our analysis of the BrainCancer data introduced in Sec-\\ntion 11.3. We wish to compare the survival of males to that of females.\\nFigure 11.3 shows the Kaplan–Meier survival curves for the two groups.\\nFemales seem to fare a little better up to about 50 months, but then the\\ntwo curves both level off to about 50%. How can we carry out a formal test\\nof equality of the two survival curves?\\nAt first glance, a two-sample t-test seems like an obvious choice: we could\\ntest whether the mean survival time among the females equals the mean\\nsurvival time among the males. But the presence of censoring again creates\\na complication. To overcome this challenge, we will conduct a log-rank test,4\\nlog-rank test 11.4 The Log-Rank Test\\n475\\nGroup 1\\nGroup 2\\nTotal\\nDied\\nq1k\\nq2k\\nqk\\nSurvived\\nr1k −q1k\\nr2k −q2k\\nrk −qk\\nTotal\\nr1k\\nr2k\\nrk\\nTABLE 11.1. Among the set of patients at risk at time dk, the number of patients\\nwho died and survived in each of two groups is reported.\\nwhich examines how the events in each group unfold sequentially in time.\\nRecall from Section 11.3 that d1 < d2 < · · · < dK are the unique death\\ntimes among the non-censored patients, rk is the number of patients at\\nrisk at time dk, and qk is the number of patients who died at time dk. We\\nfurther define r1k and r2k to be the number of patients in groups 1 and 2,\\nrespectively, who are at risk at time dk. Similarly, we define q1k and q2k to\\nbe the number of patients in groups 1 and 2, respectively, who died at time\\ndk. Note that r1k + r2k = rk and q1k + q2k = qk.\\nAt each death time dk, we construct a 2 × 2 table of counts of the form\\nshown in Table 11.1. Note that if the death times are unique (i.e. no two\\nindividuals die at the same time), then one of q1k and q2k equals one, and\\nthe other equals zero.\\nThe main idea behind the log-rank test statistic is as follows. In order\\nto test H0 : E(X) = µ for some random variable X, one approach is to\\nconstruct a test statistic of the form\\nW =\\nX −µ\\n5\\nVar(X)\\n.\\n(11.4)\\nTo construct the log-rank test statistic, we compute a quantity that takes\\nexactly the form (11.4), with X = )K\\nk=1 q1k, where q1k is given in the top\\nleft of Table 11.1.\\nIn greater detail, if there is no difference in survival between the two\\ngroups, and conditioning on the row and column totals in Table 11.1, the\\nexpected value of q1k is\\nµk = r1k\\nrk\\nqk.\\n(11.5)\\nSo the expected value of X = )K\\nk=1 q1k is µ = )K\\nk=1\\nr1k\\nrk qk. Furthermore,\\nit can be shown5 that the variance of q1k is\\nVar (q1k) = qk(r1k/rk)(1 −r1k/rk)(rk −qk)\\nrk −1\\n.\\n(11.6)\\nThough q11, . . . , q1K may be correlated, we nonetheless estimate\\nVar\\n> K\\n0\\nk=1\\nq1k\\n?\\n≈\\nK\\n0\\nk=1\\nVar (q1k) =\\nK\\n0\\nk=1\\nqk(r1k/rk)(1 −r1k/rk)(rk −qk)\\nrk −1\\n.\\n(11.7)\\n4The log-rank test is also known as the Mantel–Haenszel test\\nor Cochran–Mantel–\\nHaenszel test.\\n5For details, see Exercise 7 at the end of this chapter. 476\\n11. Survival Analysis and Censored Data\\nTherefore, to compute the log-rank test statistic, we simply proceed as\\nin (11.4), with X = )K\\nk=1 q1k, making use of (11.5) and (11.7). That is, we\\ncalculate\\nW =\\n)K\\nk=1 (q1k −µk)\\nG)K\\nk=1 Var (q1k)\\n=\\n)K\\nk=1\\n1\\nq1k −qk\\nrk r1k\\n2\\nG)K\\nk=1\\nqk(r1k/rk)(1−r1k/rk)(rk−qk)\\nrk−1\\n.\\n(11.8)\\nWhen the sample size is large, the log-rank test statistic W has ap-\\nproximately a standard normal distribution; this can be used to compute\\na p-value for the null hypothesis that there is no difference between the\\nsurvival curves in the two groups.6\\nComparing the survival times of females and males on the BrainCancer\\ndata gives a log-rank test statistic of W = 1.2, which corresponds to a two-\\nsided p-value of 0.2 using the theoretical null distribution, and a p-value\\nof 0.25 using the permutation null distribution with 1,000 permutations.\\nThus, we cannot reject the null hypothesis of no difference in survival curves\\nbetween females and males.\\nThe log-rank test is closely related to Cox’s proportional hazards model,\\nwhich we discuss in Section 11.5.2.\\n11.5\\nRegression Models With a Survival Response\\nWe now consider the task of fitting a regression model to survival data.\\nAs in Section 11.1, the observations are of the form (Y, δ), where Y =\\nmin(T, C) is the (possibly censored) survival time, and δ is an indicator\\nvariable that equals 1 if T ≤C. Furthermore, X ∈Rp is a vector of p\\nfeatures. We wish to predict the true survival time T.\\nSince the observed quantity Y is positive and may have a long right\\ntail, we might be tempted to fit a linear regression of log(Y ) on X. But\\nas the reader will surely guess, censoring again creates a problem since\\nwe are actually interested in predicting T and not Y . To overcome this\\ndifficulty, we instead make use of a sequential construction, similar to the\\nconstructions of the Kaplan–Meier survival curve in Section 11.3 and the\\nlog-rank test in Section 11.4.\\n11.5.1\\nThe Hazard Function\\nThe hazard function or hazard rate — also known as the force of mortality hazard\\nfunction\\n— is formally defined as\\nh(t) = lim\\n∆t→0\\nPr(t < T ≤t + ∆t|T > t)\\n∆t\\n,\\n(11.9)\\n6Alternatively, we can estimate the p-value via permutations, using ideas that will\\nbe presented in Section 13.5. The permutation distribution is obtained by randomly\\nswapping the labels for the observations in the two groups. 11.5 Regression Models With a Survival Response\\n477\\nwhere T is the (unobserved) survival time. It is the death rate in the instant\\nafter time t, given survival past that time.7 In (11.9), we take the limit as\\n∆t approaches zero, so we can think of ∆t as being an extremely tiny\\nnumber. Thus, more informally, (11.9) implies that\\nh(t) ≈Pr(t < T ≤t + ∆t|T > t)\\n∆t\\nfor some arbitrarily small ∆t.\\nWhy should we care about the hazard function? First of all, it is closely\\nrelated to the survival curve (11.2), as we will see next. Second, it turns out\\nthat a key approach for modeling survival data as a function of covariates\\nrelies heavily on the hazard function; we will introduce this approach —\\nCox’s proportional hazards model — in Section 11.5.2.\\nWe now consider the hazard function h(t) in a bit more detail. Recall\\nthat for two events A and B, the probability of A given B can be expressed\\nas Pr(A | B) = Pr(A ∩B)/ Pr(B), i.e. the probability that A and B both\\noccur divided by the probability that B occurs. Furthermore, recall from\\n(11.2) that S(t) = Pr(T > t). Thus,\\nh(t)\\n=\\nlim\\n∆t→0\\nPr ((t < T ≤t + ∆t) ∩(T > t)) /∆t\\nPr(T > t)\\n=\\nlim\\n∆t→0\\nPr(t < T ≤t + ∆t)/∆t\\nPr(T > t)\\n=\\nf(t)\\nS(t),\\n(11.10)\\nwhere\\nf(t) = lim\\n∆t→0\\nPr(t < T ≤t + ∆t)\\n∆t\\n(11.11)\\nis the probability density function associated with T, i.e. it is the instanta- probability\\ndensity\\nfunction\\nneous rate of death at time t. The second equality in (11.10) made use of\\nthe fact that if t < T ≤t + ∆t, then it must be the case that T > t.\\nEquation 11.10 implies a relationship between the hazard function h(t),\\nthe survival function S(t), and the probability density function f(t). In\\nfact, these are three equivalent ways8 of describing the distribution of T.\\nThe likelihood associated with the ith observation is\\nLi =\\n=\\nf(yi)\\nif the ith observation is not censored\\nS(yi)\\nif the ith observation is censored\\n= f(yi)δiS(yi)1−δi.\\n(11.12)\\nThe intuition behind (11.12) is as follows: if Y = yi and the ith observation\\nis not censored, then the likelihood is the probability of dying in a tiny in-\\nterval around time yi. If the ith observation is censored, then the likelihood\\n7Due to the ∆t in the denominator of (11.9), the hazard function is a rate of death,\\nrather than a probability of death. However, higher values of h(t) directly correspond\\nto a higher probability of death, just as higher values of a probability density function\\ncorrespond to more likely outcomes for a random variable. In fact, h(t) is the probability\\ndensity function for T conditional on T > t.\\n8See Exercise 8. 478\\n11. Survival Analysis and Censored Data\\nis the probability of surviving at least until time yi. Assuming that the n\\nobservations are independent, the likelihood for the data takes the form\\nL =\\nn\\nE\\ni=1\\nf(yi)δiS(yi)1−δi =\\nn\\nE\\ni=1\\nh(yi)δiS(yi),\\n(11.13)\\nwhere the second equality follows from (11.10).\\nWe now consider the task of modeling the survival times. If we assume ex-\\nponential survival, i.e. that the probability density function of the survival\\ntime T takes the form f(t) = λ exp(−λt), then estimating the parameter λ\\nby maximizing the likelihood in (11.13) is straightforward.9 Alternatively,\\nwe could assume that the survival times are drawn from a more flexible\\nfamily of distributions, such as the Gamma or Weibull family. Another\\npossibility is to model the survival times non-parametrically, as was done\\nin Section 11.3 using the Kaplan–Meier estimator.\\nHowever, what we would really like to do is model the survival time as\\na function of the covariates. To do this, it is convenient to work directly\\nwith the hazard function, instead of the probability density function.10\\nOne possible approach is to assume a functional form for the hazard func-\\ntion h(t|xi), such as h(t|xi) = exp\\n1\\nβ0 + )p\\nj=1 βjxij\\n2\\n, where the exponent\\nfunction guarantees that the hazard function is non-negative. Note that\\nthe exponential hazard function is special, in that it does not vary with\\ntime.11 Given h(t|xi), we could calculate S(t|xi). Plugging these equations\\ninto (11.13), we could then maximize the likelihood in order to estimate the\\nparameter β = (β0, β1, . . . , βp)T . However, this approach is quite restric-\\ntive, in the sense that it requires us to make a very stringent assumption\\non the form of the hazard function h(t|xi). In the next section, we will\\nconsider a much more flexible approach.\\n11.5.2\\nProportional Hazards\\nThe Proportional Hazards Assumption\\nThe proportional hazards assumption states that\\nproportional\\nhazards\\nassumption\\nh(t|xi) = h0(t) exp\\n\\uf8eb\\n\\uf8ed\\np\\n0\\nj=1\\nxijβj\\n\\uf8f6\\n\\uf8f8,\\n(11.14)\\nwhere h0(t) ≥0 is an unspecified function, known as the baseline hazard. baseline\\nhazard\\nIt is the hazard function for an individual with features xi1 = · · · = xip =\\n0. The name “proportional hazards” arises from the fact that the hazard\\nfunction for an individual with feature vector xi is some unknown function\\n9See Exercise 9.\\n10Given the close relationship between the hazard function h(t) and the density func-\\ntion f(t) explored in Exercise 8, posing an assumption about the form of the hazard\\nfunction is closely related to posing an assumption about the form of the density func-\\ntion, as was done in the previous paragraph.\\n11The notation h(t|xi) indicates that we are now considering the hazard function for\\nthe ith observation conditional on the values of the covariates, xi. 11.5 Regression Models With a Survival Response\\n479\\n0.5\\n1.0\\n1.5\\n2.0\\n−3\\n−2\\n−1\\n0\\nTime\\nLog Hazard\\n0.5\\n1.0\\n1.5\\n2.0\\n0.0\\n0.2\\n0.4\\n0.6\\n0.8\\n1.0\\nTime\\nSurvival Probability\\n0.5\\n1.0\\n1.5\\n2.0\\n−4\\n−2\\n0\\n1\\n2\\nTime\\nLog Hazard\\n0.5\\n1.0\\n1.5\\n2.0\\n0.0\\n0.2\\n0.4\\n0.6\\n0.8\\n1.0\\nTime\\nSurvival Probability\\nFIGURE 11.4. Top: In a simple example with p = 1 and a binary covariate\\nxi ∈{0, 1}, the log hazard and the survival function under the model (11.14)\\nare shown (green for xi = 0 and black for xi = 1). Because of the proportional\\nhazards assumption (11.14), the log hazard functions differ by a constant, and the\\nsurvival functions do not cross. Bottom: Again we have a single binary covariate\\nxi ∈{0, 1}. However, the proportional hazards assumption (11.14) does not hold.\\nThe log hazard functions cross, as do the survival functions.\\nh0(t) times the factor exp\\n1)p\\nj=1 xijβj\\n2\\n. The quantity exp\\n1)p\\nj=1 xijβj\\n2\\nis called the relative risk for the feature vector xi = (xi1, . . . , xip)T , relative\\nto that for the feature vector xi = (0, . . . , 0)T .\\nWhat does it mean that the baseline hazard function h0(t) in (11.14) is\\nunspecified? Basically, we make no assumptions about its functional form.\\nWe allow the instantaneous probability of death at time t, given that one\\nhas survived at least until time t, to take any form. This means that the\\nhazard function is very flexible and can model a wide range of relationships\\nbetween the covariates and survival time. Our only assumption is that a\\none-unit increase in xij corresponds to an increase in h(t|xi) by a factor of\\nexp(βj).\\nAn illustration of the proportional hazards assumption (11.14) is given in\\nFigure 11.4, in a simple setting with a single binary covariate xi ∈{0, 1} (so\\nthat p = 1). In the top row, the proportional hazards assumption (11.14)\\nholds. Thus, the hazard functions of the two groups are a constant multiple\\nof each other, so that on the log scale, the gap between them is constant.\\nFurthermore, the survival curves never cross, and in fact the gap between\\nthe survival curves tends to (initially) increase over time. By contrast, in\\nthe bottom row, (11.14) does not hold. We see that the log hazard functions\\nfor the two groups cross, as do the survival curves. 480\\n11. Survival Analysis and Censored Data\\nCox’s Proportional Hazards Model\\nBecause the form of h0(t) in the proportional hazards assumption (11.14)\\nis unknown, we cannot simply plug h(t|xi) into the likelihood (11.13) and\\nthen estimate β = (β1, . . . , βp)T by maximum likelihood. The magic of\\nCox’s proportional hazards model lies in the fact that it is in fact possible Cox’s\\nproportional\\nhazards\\nmodel\\nto estimate β without having to specify the form of h0(t).\\nTo accomplish this, we make use of the same “sequential in time” logic\\nthat we used to derive the Kaplan–Meier survival curve and the log-rank\\ntest. For simplicity, assume that there are no ties among the failure, or\\ndeath, times: i.e. each failure occurs at a distinct time. Assume that δi =\\n1, i.e. the ith observation is uncensored, and thus yi is its failure time.\\nThen the hazard function for the ith observation at time yi is h(yi|xi) =\\nh0(yi) exp\\n1)p\\nj=1 xijβj\\n2\\n, and the total hazard at time yi for the at risk\\nobservations12 is\\n0\\ni′:yi′≥yi\\nh0(yi) exp\\n\\uf8eb\\n\\uf8ed\\np\\n0\\nj=1\\nxi′jβj\\n\\uf8f6\\n\\uf8f8.\\nTherefore, the probability that the ith observation is the one to fail at time\\nyi (as opposed to one of the other observations in the risk set) is\\nh0(yi) exp\\n1)p\\nj=1 xijβj\\n2\\n)\\ni′:yi′≥yi h0(yi) exp\\n1)p\\nj=1 xi′jβj\\n2 =\\nexp\\n1)p\\nj=1 xijβj\\n2\\n)\\ni′:yi′≥yi exp\\n1)p\\nj=1 xi′jβj\\n2.\\n(11.15)\\nNotice that the unspecified baseline hazard function h0(yi) cancels out of\\nthe numerator and denominator!\\nThe partial likelihood is simply the product of these probabilities over all partial\\nlikelihood\\nof the uncensored observations,\\nPL(β) =\\nE\\ni:δi=1\\nexp\\n1)p\\nj=1 xijβj\\n2\\n)\\ni′:yi′≥yi exp\\n1)p\\nj=1 xi′jβj\\n2.\\n(11.16)\\nCritically, the partial likelihood is valid regardless of the true value of h0(t),\\nmaking the model very flexible and robust.13\\nTo estimate β, we simply maximize the partial likelihood (11.16) with\\nrespect to β. As was the case for logistic regression in Chapter 4, no closed-\\nform solution is available, and so iterative algorithms are required.\\nIn addition to estimating β, we can also obtain other model outputs that\\nwe saw in the context of least squares regression in Chapter 3 and logistic\\nregression in Chapter 4. For example, we can obtain p-values corresponding\\n12Recall that the “at risk” observations at time yi are those that are still at risk of\\nfailure, i.e. those that have not yet failed or been censored before time yi.\\n13In general, the partial likelihood is used in settings where it is difficult to compute\\nthe full likelihood for all of the parameters. Instead, we compute a likelihood for just the\\nparameters of primary interest: in this case, β1, . . . , βp. It can be shown that maximizing\\n(11.16) provides good estimates for these parameters. 11.5 Regression Models With a Survival Response\\n481\\nto particular null hypotheses (e.g. H0 : βj = 0), as well as confidence\\nintervals associated with the coefficients.\\nConnection With The Log-Rank Test\\nSuppose we have just a single predictor (p = 1), which we assume to be\\nbinary, i.e. xi ∈{0, 1}. In order to determine whether there is a difference\\nbetween the survival times of the observations in the group {i : xi = 0}\\nand those in the group {i : xi = 1}, we can consider taking two possible\\napproaches:\\nApproach #1: Fit a Cox proportional hazards model, and test the\\nnull hypothesis H0 : β = 0. (Since p = 1, β is a scalar.)\\nApproach #2: Perform a log-rank test to compare the two groups, as\\nin Section 11.4.\\nWhich one should we prefer?\\nIn fact, there is a close relationship between these two approaches. In\\nparticular, when taking Approach #1, there are a number of possible ways\\nto test H0. One way is known as a score test. It turns out that in the case of\\na single binary covariate, the score test for H0 : β = 0 in Cox’s proportional\\nhazards model is exactly equal to the log-rank test. In other words, it does\\nnot matter whether we take Approach #1 or Approach #2!\\nAdditional Details\\nThe discussion of Cox’s proportional hazards model glossed over a few\\nsubtleties:\\n• There is no intercept in (11.14) nor in the equations that follow,\\nbecause an intercept can be absorbed into the baseline hazard h0(t).\\n• We have assumed that there are no tied failure times. In the case\\nof ties, the exact form of the partial likelihood (11.16) is a bit more\\ncomplicated, and a number of computational approximations must\\nbe used.\\n• (11.16) is known as the partial likelihood because it is not exactly a\\nlikelihood. That is, it does not correspond exactly to the probability\\nof the data under the assumption (11.14). However, it is a very good\\napproximation.\\n• We have focused only on estimation of the coefficients β = (β1, . . . , βp)T .\\nHowever, at times we may also wish to estimate the baseline hazard\\nh0(t), for instance so that we can estimate the survival curve S(t|x) for\\nan individual with feature vector x. The details are beyond the scope\\nof this book. Estimation of h0(t) is implemented in the lifelines\\npackage in Python, which we will see in Section 11.8. 482\\n11. Survival Analysis and Censored Data\\n11.5.3\\nExample: Brain Cancer Data\\nTable 11.2 shows the result of fitting the proportional hazards model to\\nthe BrainCancer data, which was originally described in Section 11.3. The\\ncoefficient column displays ˆβj. The results indicate, for instance, that the\\nestimated hazard for a male patient is e0.18 = 1.2 times greater than for\\na female patient: in other words, with all other features held fixed, males\\nhave a 1.2 times greater chance of dying than females, at any point in time.\\nHowever, the p-value is 0.61, which indicates that this difference between\\nmales and females is not significant.\\nAs another example, we also see that each one-unit increase in the\\nKarnofsky index corresponds to a multiplier of exp(−0.05) = 0.95 in the\\ninstantaneous chance of dying. In other words, the higher the Karnofsky\\nindex, the lower the chance of dying at any given point in time. This effect\\nis highly significant, with a p-value of 0.0027.\\nCoefficient\\nStd. error\\nz-statistic\\np-value\\nsex[Male]\\n0.18\\n0.36\\n0.51\\n0.61\\ndiagnosis[LG Glioma]\\n0.92\\n0.64\\n1.43\\n0.15\\ndiagnosis[HG Glioma]\\n2.15\\n0.45\\n4.78\\n0.00\\ndiagnosis[Other]\\n0.89\\n0.66\\n1.35\\n0.18\\nloc[Supratentorial]\\n0.44\\n0.70\\n0.63\\n0.53\\nki\\n-0.05\\n0.02\\n-3.00\\n<0.01\\ngtv\\n0.03\\n0.02\\n1.54\\n0.12\\nstereo[SRT]\\n0.18\\n0.60\\n0.30\\n0.77\\nTABLE\\n11.2. Results\\nfor\\nCox’s\\nproportional\\nhazards\\nmodel\\nfit\\nto\\nthe\\nBrainCancer data, which was first described in Section 11.3. The variable\\ndiagnosis is qualitative with four levels: meningioma, LG glioma, HG glioma,\\nor other. The variables sex, loc, and stereo are binary.\\n11.5.4\\nExample: Publication Data\\nNext, we consider the dataset Publication involving the time to publica-\\ntion of journal papers reporting the results of clinical trials funded by the\\nNational Heart, Lung, and Blood Institute.14 For 244 trials, the time in\\nmonths until publication is recorded. Of the 244 trials, only 156 were pub-\\nlished during the study period; the remaining studies were censored. The\\ncovariates include whether the trial focused on a clinical endpoint (clinend),\\nwhether the trial involved multiple centers (multi), the funding mechanism\\nwithin the National Institutes of Health (mech), trial sample size (sampsize),\\nbudget (budget), impact (impact, related to the number of citations), and\\nwhether the trial produced a positive (significant) result (posres). The last\\ncovariate is particularly interesting, as a number of studies have suggested\\nthat positive trials have a higher publication rate.\\n14This dataset is described in the following paper: Gordon et al. (2013) Publication of\\ntrials funded by the National Heart, Lung, and Blood Institute. New England Journal\\nof Medicine, 369(20):1926–1934. 11.5 Regression Models With a Survival Response\\n483\\n0\\n20\\n40\\n60\\n80\\n100\\n120\\n0.0\\n0.2\\n0.4\\n0.6\\n0.8\\n1.0\\nMonths\\nProbability of Not Being Published\\nNegative Result\\nPositive Result\\nFIGURE 11.5. Survival curves for time until publication for the Publication\\ndata described in Section 11.5.4, stratified by whether or not the study produced\\na positive result.\\nFigure 11.5 shows the Kaplan–Meier curves for the time until publication,\\nstratified by whether or not the study produced a positive result. We see\\nslight evidence that time until publication is lower for studies with a positive\\nresult. However, the log-rank test yields a very unimpressive p-value of 0.36.\\nWe now consider a more careful analysis that makes use of all of the\\navailable predictors. The results of fitting Cox’s proportional hazards model\\nusing all of the available features are shown in Table 11.3. We find that the\\nchance of publication of a study with a positive result is e0.55 = 1.74 times\\nhigher than the chance of publication of a study with a negative result\\nat any point in time, holding all other covariates fixed. The very small\\np-value associated with posres in Table 11.3 indicates that this result is\\nhighly significant. This is striking, especially in light of our earlier finding\\nthat a log-rank test comparing time to publication for studies with positive\\nversus negative results yielded a p-value of 0.36. How can we explain this\\ndiscrepancy? The answer stems from the fact that the log-rank test did not\\nconsider any other covariates, whereas the results in Table 11.3 are based\\non a Cox model using all of the available covariates. In other words, after\\nwe adjust for all of the other covariates, then whether or not the study\\nyielded a positive result is highly predictive of the time to publication.\\nIn order to gain more insight into this result, in Figure 11.6 we display\\nestimates of the survival curves associated with positive and negative re-\\nsults, adjusting for the other predictors. To produce these survival curves,\\nwe estimated the underlying baseline hazard h0(t). We also needed to se-\\nlect representative values for the other predictors; we used the mean value\\nfor each predictor, except for the categorical predictor mech, for which we\\nused the most prevalent category (R01). Adjusting for the other predictors,\\nwe now see a clear difference in the survival curves between studies with\\npositive versus negative results.\\nOther interesting insights can be gleaned from Table 11.3. For example,\\nstudies with a clinical endpoint are more likely to be published at any\\ngiven point in time than those with a non-clinical endpoint. The funding 484\\n11. Survival Analysis and Censored Data\\nCoefficient\\nStd. error\\nz-statistic\\np-value\\nposres[Yes]\\n0.55\\n0.18\\n3.02\\n0.00\\nmulti[Yes]\\n0.15\\n0.31\\n0.47\\n0.64\\nclinend[Yes]\\n0.51\\n0.27\\n1.89\\n0.06\\nmech[K01]\\n1.05\\n1.06\\n1.00\\n0.32\\nmech[K23]\\n-0.48\\n1.05\\n-0.45\\n0.65\\nmech[P01]\\n-0.31\\n0.78\\n-0.40\\n0.69\\nmech[P50]\\n0.60\\n1.06\\n0.57\\n0.57\\nmech[R01]\\n0.10\\n0.32\\n0.30\\n0.76\\nmech[R18]\\n1.05\\n1.05\\n0.99\\n0.32\\nmech[R21]\\n-0.05\\n1.06\\n-0.04\\n0.97\\nmech[R24,K24]\\n0.81\\n1.05\\n0.77\\n0.44\\nmech[R42]\\n-14.78\\n3414.38\\n-0.00\\n1.00\\nmech[R44]\\n-0.57\\n0.77\\n-0.73\\n0.46\\nmech[RC2]\\n-14.92\\n2243.60\\n-0.01\\n0.99\\nmech[U01]\\n-0.22\\n0.32\\n-0.70\\n0.48\\nmech[U54]\\n0.47\\n1.07\\n0.44\\n0.66\\nsampsize\\n0.00\\n0.00\\n0.19\\n0.85\\nbudget\\n0.00\\n0.00\\n1.67\\n0.09\\nimpact\\n0.06\\n0.01\\n8.23\\n0.00\\nTABLE\\n11.3. Results\\nfor\\nCox’s\\nproportional\\nhazards\\nmodel\\nfit\\nto\\nthe\\nPublication data, using all of the available features. The features posres, multi,\\nand clinend are binary. The feature mech is qualitative with 14 levels; it is coded\\nso that the baseline level is Contract.\\nmechanism did not appear to be significantly associated with time until\\npublication.\\n11.6\\nShrinkage for the Cox Model\\nIn this section, we illustrate that the shrinkage methods of Section 6.2\\ncan be applied to the survival data setting. In particular, motivated by\\nthe “loss+penalty” formulation of Section 6.2, we consider minimizing a\\npenalized version of the negative log partial likelihood in (11.16),\\n−log\\n\\uf8eb\\n\\uf8edE\\ni:δi=1\\nexp\\n1)p\\nj=1 xijβj\\n2\\n)\\ni′:yi′≥yi exp\\n1)p\\nj=1 xi′jβj\\n2\\n\\uf8f6\\n\\uf8f8+ λP(β),\\n(11.17)\\nwith respect to β = (β1, . . . , βp)T . We might take P(β) = )p\\nj=1 β2\\nj , which\\ncorresponds to a ridge penalty, or P(β) = )p\\nj=1 |βj|, which corresponds to\\na lasso penalty.\\nIn (11.17), λ is a non-negative tuning parameter; typically we will mini-\\nmize it over a range of values of λ. When λ = 0, then minimizing (11.17) is\\nequivalent to simply maximizing the usual Cox partial likelihood (11.16).\\nHowever, when λ > 0, then minimizing (11.17) yields a shrunken version of\\nthe coefficient estimates. When λ is large, then using a ridge penalty will\\ngive small coefficients that are not exactly equal to zero. By contrast, for a 11.6 Shrinkage for the Cox Model\\n485\\n0\\n20\\n40\\n60\\n80\\n100\\n120\\n0.0\\n0.2\\n0.4\\n0.6\\n0.8\\n1.0\\nMonths\\nProbability of Not Being Published\\nNegative Result\\nPositive Result\\nFIGURE 11.6. For the Publication data, we display survival curves for time\\nuntil publication, stratified by whether or not the study produced a positive result,\\nafter adjusting for all other covariates.\\nsufficiently large value of λ, using a lasso penalty will give some coefficients\\nthat are exactly equal to zero.\\nWe now apply the lasso-penalized Cox model to the Publication data, de-\\nscribed in Section 11.5.4. We first randomly split the 244 trials into equally-\\nsized training and test sets. The cross-validation results from the training\\nset are shown in Figure 11.7. The “partial likelihood deviance”, shown on\\nthe y-axis, is twice the cross-validated negative log partial likelihood; it\\nplays the role of the cross-validation error.15 Note the “U-shape” of the\\npartial likelihood deviance: just as we saw in previous chapters, the cross-\\nvalidation error is minimized for an intermediate level of model complexity.\\nSpecifically, this occurs when just two predictors, budget and impact, have\\nnon-zero estimated coefficients.\\nNow, how do we apply this model to the test set? This brings up an\\nimportant conceptual point: in essence, there is no simple way to compare\\npredicted survival times and true survival times on the test set. The first\\nproblem is that some of the observations are censored, and so the true sur-\\nvival times for those observations are unobserved. The second issue arises\\nfrom the fact that in the Cox model, rather than predicting a single sur-\\nvival time given a covariate vector x, we instead estimate an entire survival\\ncurve, S(t|x), as a function of t.\\nTherefore, to assess the model fit, we must take a different approach,\\nwhich involves stratifying the observations using the coefficient estimates.\\nIn particular, for each test observation, we compute the “risk” score\\nbudgeti · ˆβbudget + impacti · ˆβimpact,\\nwhere ˆβbudget and ˆβimpact are the coefficient estimates for these two features\\nfrom the training set. We then use these risk scores to categorize the obser-\\nvations based on their “risk”. For instance, the high risk group consists of\\nthe observations for which budgeti · ˆβbudget + impacti · ˆβimpact is largest; by\\n15Cross-validation for the Cox model is more involved than for linear or logistic re-\\ngression, because the objective function is not a sum over the observations. 486\\n11. Survival Analysis and Censored Data\\n2e−04\\n5e−04\\n1e−03\\n2e−03\\n5e−03\\n1e−02\\n2e−02\\n5e−02\\n1e−01\\n2e−01\\n9.0\\n9.1\\n9.2\\n9.3\\n9.4\\n9.5\\n9.6\\n9.7\\nPartial Likelihood Deviance\\n∥ˆβL\\nλ ∥1\\n!\\n| ˆβ∥1\\nFIGURE 11.7. For the Publication data described in Section 11.5.4, cross-val-\\nidation results for the lasso-penalized Cox model are shown. The y-axis displays\\nthe partial likelihood deviance, which plays the role of the cross-validation error.\\nThe x-axis displays the ℓ1 norm (that is, the sum of the absolute values) of the\\ncoefficients of the lasso-penalized Cox model with tuning parameter λ, divided by\\nthe ℓ1 norm of the coefficients of the unpenalized Cox model. The dashed line\\nindicates the minimum cross-validation error.\\n(11.14), we see that these are the observations for which the instantaneous\\nprobability of being published at any moment in time is largest. In other\\nwords, the high risk group consists of the trials that are likely to be pub-\\nlished sooner. On the Publication data, we stratify the observations into\\ntertiles of low, medium, and high risk. The resulting survival curves for\\neach of the three strata are displayed in Figure 11.8. We see that there is\\nclear separation between the three strata, and that the strata are correctly\\nordered in terms of low, medium, and high risk of publication.\\n11.7\\nAdditional Topics\\n11.7.1\\nArea Under the Curve for Survival Analysis\\nIn Chapter 4, we introduced the area under the ROC curve — often referred\\nto as the “AUC” — as a way to quantify the performance of a two-class clas-\\nsifier. Define the score for the ith observation to be the classifier’s estimate\\nof Pr(Y = 1|X = xi). It turns out that if we consider all pairs consisting of\\none observation in Class 1 and one observation in Class 2, then the AUC\\nis the fraction of pairs for which the score for the observation in Class 1\\nexceeds the score for the observation in Class 2.\\nThis suggests a way to generalize the notion of AUC to survival anal-\\nysis. We calculate an estimated risk score, ˆηi = ˆβ1xi1 + · · · + ˆβpxip, for\\ni = 1, . . . , n, using the Cox model coefficients. If ˆηi′ > ˆηi, then the model\\npredicts that the i′th observation has a larger hazard than the ith obser-\\nvation, and thus that the survival time ti will be greater than ti′. Thus, it\\nis tempting to try to generalize AUC by computing the proportion of ob-\\nservations for which ti > ti′ and ˆηi′ > ˆηi. However, things are not quite so\\neasy, because recall that we do not observe t1, . . . , tn; instead, we observe 11.7 Additional Topics\\n487\\n0\\n20\\n40\\n60\\n80\\n100\\n120\\n0.0\\n0.2\\n0.4\\n0.6\\n0.8\\n1.0\\nMonths\\nProbability of Not Being Published\\nLow Risk\\nMedium Risk\\nHigh Risk\\nFIGURE 11.8. For the Publication data introduced in Section 11.5.4, we\\ncompute tertiles of “risk” in the test set using coefficients estimated on the training\\nset. There is clear separation between the resulting survival curves.\\nthe (possibly-censored) times y1, . . . , yn, as well as the censoring indicators\\nδ1, . . . , δn.\\nTherefore, Harrell’s concordance index (or C-index) computes the pro- Harrell’s\\nconcordance\\nindex\\nportion of observation pairs for which ˆηi′ > ˆηi and yi > yi′:\\nC =\\n)\\ni,i′:yi>yi′ I(ˆηi′ > ˆηi)δi′\\n)\\ni,i′:yi>yi′ δi′\\n,\\nwhere the indicator variable I(ˆηi′ > ˆηi) equals one if ˆηi′ > ˆηi, and equals\\nzero otherwise. The numerator and denominator are multiplied by the sta-\\ntus indicator δi′, since if the i′th observation is uncensored (i.e. if δi′ = 1),\\nthen yi > yi′ implies that ti > ti′. By contrast, if δi′ = 0, then yi > yi′\\ndoes not imply that ti > ti′.\\nWe fit a Cox proportional hazards model on the training set of the\\nPublication data, and computed the C-index on the test set. This yielded\\nC = 0.733. Roughly speaking, given two random papers from the test set,\\nthe model can predict with 73.3% accuracy which will be published first.\\n11.7.2\\nChoice of Time Scale\\nIn the examples considered thus far in this chapter, it has been fairly clear\\nhow to define time. For example, in the Publication example, time zero for\\neach paper was defined to be the calendar time at the end of the study,\\nand the failure time was defined to be the number of months that elapsed\\nfrom the end of the study until the paper was published.\\nHowever, in other settings, the definitions of time zero and failure time\\nmay be more subtle. For example, when examining the association between\\nrisk factors and disease occurrence in an epidemiological study, one might\\nuse the patient’s age to define time, so that time zero is the patient’s date\\nof birth. With this choice, the association between age and survival cannot\\nbe measured; however, there is no need to adjust for age in the analysis.\\nWhen examining covariates associated with disease-free survival (i.e. the 488\\n11. Survival Analysis and Censored Data\\namount of time elapsed between treatment and disease recurrence), one\\nmight use the date of treatment as time zero.\\n11.7.3\\nTime-Dependent Covariates\\nA powerful feature of the proportional hazards model is its ability to handle\\ntime-dependent covariates, predictors whose value may change over time.\\nFor example, suppose we measure a patient’s blood pressure every week\\nover the course of a medical study. In this case, we can think of the blood\\npressure for the ith observation not as xi, but rather as xi(t) at time t.\\nBecause the partial likelihood in (11.16) is constructed sequentially in\\ntime, dealing with time-dependent covariates is straightforward. In partic-\\nular, we simply replace xij and xi′j in (11.16) with xij(yi) and xi′j(yi),\\nrespectively; these are the current values of the predictors at time yi. By\\ncontrast, time-dependent covariates would pose a much greater challenge\\nwithin the context of a traditional parametric approach, such as (11.13).\\nOne example of time-dependent covariates appears in the analysis of data\\nfrom the Stanford Heart Transplant Program. Patients in need of a heart\\ntransplant were put on a waiting list. Some patients received a transplant,\\nbut others died while still on the waiting list. The primary objective of the\\nanalysis was to determine whether a transplant was associated with longer\\npatient survival.\\nA naïve approach would use a fixed covariate to represent transplant\\nstatus: that is, xi = 1 if the ith patient ever received a transplant, and xi =\\n0 otherwise. But this approach overlooks the fact that patients had to live\\nlong enough to get a transplant, and hence, on average, healthier patients\\nreceived transplants. This problem can be solved by using a time-dependent\\ncovariate for transplant: xi(t) = 1 if the patient received a transplant by\\ntime t, and xi(t) = 0 otherwise.\\n11.7.4\\nChecking the Proportional Hazards Assumption\\nWe have seen that Cox’s proportional hazards model relies on the propor-\\ntional hazards assumption (11.14). While results from the Cox model tend\\nto be fairly robust to violations of this assumption, it is still a good idea to\\ncheck whether it holds. In the case of a qualitative feature, we can plot the\\nlog hazard function for each level of the feature. If (11.14) holds, then the\\nlog hazard functions should just differ by a constant, as seen in the top-left\\npanel of Figure 11.4. In the case of a quantitative feature, we can take a\\nsimilar approach by stratifying the feature.\\n11.7.5\\nSurvival Trees\\nIn Chapter 8, we discussed flexible and adaptive learning procedures such as\\ntrees, random forests, and boosting, which we applied in both the regression\\nand classification settings. Most of these approaches can be generalized to\\nthe survival analysis setting. For example, survival trees are a modification survival\\ntrees\\nof classification and regression trees that use a split criterion that maximizes 11.8 Lab: Survival Analysis\\n489\\nthe difference between the survival curves in the resulting daughter nodes.\\nSurvival trees can then be used to create random survival forests.\\n11.8\\nLab: Survival Analysis\\nIn this lab, we perform survival analyses on three separate data sets. In\\nSection 11.8.1 we analyze the BrainCancer data that was first described\\nin Section 11.3. In Section 11.8.2, we examine the Publication data from\\nSection 11.5.4. Finally, Section 11.8.3 explores a simulated call-center data\\nset.\\nWe begin by importing some of our libraries at this top level. This makes\\nthe code more readable, as scanning the first few lines of the notebook tell\\nus what libraries are used in this notebook.\\nIn [1]: from matplotlib.pyplot import subplots\\nimport numpy as np\\nimport pandas as pd\\nfrom ISLP.models import ModelSpec as MS\\nfrom ISLP import load_data\\nWe also collect the new imports needed for this lab.\\nIn [2]: from lifelines import \\\\\\n(KaplanMeierFitter ,\\nCoxPHFitter)\\nfrom lifelines.statistics import \\\\\\n(logrank_test ,\\nmultivariate_logrank_test)\\nfrom ISLP.survival import sim_time\\n11.8.1\\nBrain Cancer Data\\nWe begin with the BrainCancer data set, contained in the ISLP package.\\nIn [3]: BrainCancer = load_data(\\'BrainCancer\\')\\nBrainCancer.columns\\nOut[3]: Index([\\'sex\\', \\'diagnosis\\', \\'loc\\', \\'ki\\', \\'gtv\\', \\'stereo\\',\\n\\'status\\', \\'time\\'],\\ndtype=\\'object\\')\\nThe rows index the 88 patients, while the 8 columns contain the predictors\\nand outcome variables. We first briefly examine the data.\\nIn [4]: BrainCancer[\\'sex\\'].value_counts()\\nOut[4]: Female\\n45\\nMale\\n43\\nName: sex, dtype: int64\\nIn [5]: BrainCancer[\\'diagnosis\\'].value_counts() 490\\n11. Survival Analysis and Censored Data\\nOut[5]: Meningioma\\n42\\nHG glioma\\n22\\nOther\\n14\\nLG glioma\\n9\\nName: diagnosis, dtype: int64\\nIn [6]: BrainCancer[\\'status\\'].value_counts()\\nOut[6]: 0\\n53\\n1\\n35\\nName: status, dtype: int64\\nBefore beginning an analysis, it is important to know how the status\\nvariable has been coded. Most software uses the convention that a status\\nof 1 indicates an uncensored observation (often death), and a status of 0\\nindicates a censored observation. But some scientists might use the opposite\\ncoding. For the BrainCancer data set 35 patients died before the end of the\\nstudy, so we are using the conventional coding.\\nTo begin the analysis, we re-create the Kaplan-Meier survival curve\\nshown in Figure 11.2. The main package we will use for survival analy-\\nsis is lifelines. The variable time corresponds to yi, the time to the ith lifelines\\nevent (either censoring or death). The first argument to km.fit is the event\\ntime, and the second argument is the censoring variable, with a 1 indicat-\\ning an observed failure time. The plot() method produces a survival curve .plot()\\nwith pointwise confidence intervals. By default, these are 90% confidence\\nintervals, but this can be changed by setting the alpha argument to one\\nminus the desired confidence level.\\nIn [7]: fig, ax = subplots(figsize=(8,8))\\nkm = KaplanMeierFitter()\\nkm_brain = km.fit(BrainCancer[\\'time\\'], BrainCancer[\\'status\\'])\\nkm_brain.plot(label=\\'Kaplan Meier estimate\\', ax=ax)\\nNext we create Kaplan-Meier survival curves that are stratified by sex, in\\norder to reproduce Figure 11.3. We do this using the groupby() method of .groupby()\\na dataframe. This method returns a generator that can be iterated over in\\nthe for loop. In this case, the items in the for loop are 2-tuples representing\\nthe groups: the first entry is the value of the grouping column sex while\\nthe second value is the dataframe consisting of all rows in the dataframe\\nmatching that value of sex. We will want to use this data below in the log-\\nrank test, hence we store this information in the dictionary by_sex. Finally,\\nwe have also used the notion of string interpolation to automatically label string\\ninterpolation\\nthe different lines in the plot. String interpolation is a powerful technique\\nto format strings — Python has many ways to facilitate such operations.\\nIn [8]: fig, ax = subplots(figsize=(8,8))\\nby_sex = {}\\nfor sex, df in BrainCancer.groupby(\\'sex\\'):\\nby_sex[sex] = df\\nkm_sex = km.fit(df[\\'time\\'], df[\\'status\\'])\\nkm_sex.plot(label=\\'Sex=%s\\' % sex, ax=ax)\\nAs discussed in Section 11.4, we can perform a log-rank test to compare\\nthe survival of males to females. We use the logrank_test() function from logrank_\\ntest() 11.8 Lab: Survival Analysis\\n491\\nthe lifelines.statistics module. The first two arguments are the event\\ntimes, with the second denoting the corresponding (optional) censoring\\nindicators.\\nIn [9]: logrank_test(by_sex[\\'Male\\'][\\'time\\'],\\nby_sex[\\'Female\\'][\\'time\\'],\\nby_sex[\\'Male\\'][\\'status\\'],\\nby_sex[\\'Female\\'][\\'status\\'])\\nOut[9]:\\nt_0\\n-1\\nnull_distribution\\nchi squared\\ndegrees_of_freedom\\n1\\ntest_name\\nlogrank_test\\ntest_statistic\\np\\n-log2(p)\\n1.44 0.23\\n2.12\\nThe resulting p-value is 0.23, indicating no evidence of a difference in\\nsurvival between the two sexes.\\nNext, we use the CoxPHFitter() estimator from lifelines to fit Cox CoxPHFitter()\\nproportional hazards models. To begin, we consider a model that uses sex\\nas the only predictor.\\nIn [10]: coxph = CoxPHFitter # shorthand\\nsex_df = BrainCancer[[\\'time\\', \\'status\\', \\'sex\\']]\\nmodel_df = MS([\\'time\\', \\'status\\', \\'sex\\'],\\nintercept=False).fit_transform(sex_df)\\ncox_fit = coxph().fit(model_df,\\n\\'time\\',\\n\\'status\\')\\ncox_fit.summary[[\\'coef\\', \\'se(coef)\\', \\'p\\']]\\nOut[10]:\\ncoef\\nse(coef)\\np\\ncovariate\\nsex[Male]\\n0.407667\\n0.342004\\n0.233263\\nThe first argument to fit should be a data frame containing at least the\\nevent time (the second argument time in this case), as well as an op-\\ntional censoring variable (the argument status in this case). Note also that\\nthe Cox model does not include an intercept, which is why we used the\\nintercept=False argument to ModelSpec above. The summary() method de-\\nlivers many columns; we chose to abbreviate its output here. It is possible\\nto obtain the likelihood ratio test comparing this model to the one with no\\nfeatures as follows:\\nIn [11]: cox_fit.log_likelihood_ratio_test()\\nOut[11]:\\nnull_distribution\\nchi squared\\ndegrees_freedom\\n1\\ntest_name\\nlog-likelihood ratio test\\ntest_statistic\\np\\n-log2(p)\\n1.44 0.23\\n2.12 492\\n11. Survival Analysis and Censored Data\\nRegardless of which test we use, we see that there is no clear evidence for\\na difference in survival between males and females. As we learned in this\\nchapter, the score test from the Cox model is exactly equal to the log rank\\ntest statistic!\\nNow we fit a model that makes use of additional predictors. We first note\\nthat one of our diagnosis values is missing, hence we drop that observation\\nbefore continuing.\\nIn [12]: cleaned = BrainCancer.dropna()\\nall_MS = MS(cleaned.columns, intercept=False)\\nall_df = all_MS.fit_transform(cleaned)\\nfit_all = coxph().fit(all_df,\\n\\'time\\',\\n\\'status\\')\\nfit_all.summary[[\\'coef\\', \\'se(coef)\\', \\'p\\']]\\nOut[12]:\\ncoef\\nse(coef)\\np\\ncovariate\\nsex[Male]\\n0.183748\\n0.360358\\n0.610119\\ndiagnosis[LG glioma]\\n-1.239541\\n0.579557\\n0.032454\\ndiagnosis[Meningioma]\\n-2.154566\\n0.450524\\n0.000002\\ndiagnosis[Other]\\n-1.268870\\n0.617672\\n0.039949\\nloc[Supratentorial]\\n0.441195\\n0.703669\\n0.530664\\nki\\n-0.054955\\n0.018314\\n0.002693\\ngtv\\n0.034293\\n0.022333\\n0.124660\\nstereo[SRT]\\n0.177778\\n0.601578\\n0.767597\\nThe diagnosis variable has been coded so that the baseline corresponds to\\nHG glioma. The results indicate that the risk associated with HG glioma\\nis more than eight times (i.e. e2.15 = 8.62) the risk associated with menin-\\ngioma. In other words, after adjusting for the other predictors, patients\\nwith HG glioma have much worse survival compared to those with menin-\\ngioma. In addition, larger values of the Karnofsky index, ki, are associated\\nwith lower risk, i.e. longer survival.\\nFinally, we plot estimated survival curves for each diagnosis category,\\nadjusting for the other predictors. To make these plots, we set the values of\\nthe other predictors equal to the mean for quantitative variables and equal\\nto the mode for categorical. To do this, we use the apply() method across\\nrows (i.e. axis=0) with a function representative that checks if a column\\nis categorical or not.\\nIn [13]: levels = cleaned[\\'diagnosis\\'].unique()\\ndef representative(series):\\nif hasattr(series.dtype, \\'categories\\'):\\nreturn pd.Series.mode(series)\\nelse:\\nreturn series.mean()\\nmodal_data = cleaned.apply(representative , axis=0)\\nWe make four copies of the column means and assign the diagnosis\\ncolumn to be the four different diagnoses.\\nIn [14]: modal_df = pd.DataFrame(\\n[modal_data.iloc[0] for _ in range(len(levels))])\\nmodal_df[\\'diagnosis\\'] = levels\\nmodal_df 11.8 Lab: Survival Analysis\\n493\\nOut[14]:\\nsex\\ndiagnosis\\nloc\\nki\\ngtv\\nstereo\\n...\\nFemale\\nMeningioma\\nSupratentorial\\n80.920\\n8.687\\nSRT\\n...\\nFemale\\nHG glioma\\nSupratentorial\\n80.920\\n8.687\\nSRT\\n...\\nFemale\\nLG glioma\\nSupratentorial\\n80.920\\n8.687\\nSRT\\n...\\nFemale\\nOther\\nSupratentorial\\n80.920\\n8.687\\nSRT\\n...\\nWe then construct the model matrix based on the model specification\\nall_MS used to fit the model, and name the rows according to the levels of\\ndiagnosis.\\nIn [15]: modal_X = all_MS.transform(modal_df)\\nmodal_X.index = levels\\nmodal_X\\nWe can use the predict_survival_function() method to obtain the esti- .predict_\\nsurvival_\\nfunction()\\nmated survival function.\\nIn [16]: predicted_survival = fit_all.predict_survival_function(modal_X)\\npredicted_survival\\nOut[16]:\\nMeningioma\\nHG glioma\\nLG glioma\\nOther\\n0.070\\n0.998\\n0.982\\n0.995\\n0.995\\n1.180\\n0.998\\n0.982\\n0.995\\n0.995\\n1.410\\n0.996\\n0.963\\n0.989\\n0.990\\n1.540\\n0.996\\n0.963\\n0.989\\n0.990\\n...\\n...\\n...\\n...\\n...\\n67.380\\n0.689\\n0.040\\n0.394\\n0.405\\n73.740\\n0.689\\n0.040\\n0.394\\n0.405\\n78.750\\n0.689\\n0.040\\n0.394\\n0.405\\n82.560\\n0.689\\n0.040\\n0.394\\n0.405\\n85 rows × 4 columns\\nThis returns a data frame, whose plot methods yields the different survival\\ncurves. To avoid clutter in the plots, we do not display confidence intervals.\\nIn [17]: fig, ax = subplots(figsize=(8, 8))\\npredicted_survival.plot(ax=ax);\\n11.8.2\\nPublication Data\\nThe Publication data presented in Section 11.5.4 can be found in the\\nISLP package. We first reproduce Figure 11.5 by plotting the Kaplan-Meier\\ncurves stratified on the posres variable, which records whether the study\\nhad a positive or negative result.\\nIn [18]: fig, ax = subplots(figsize=(8,8))\\nPublication = load_data(\\'Publication\\')\\nby_result = {}\\nfor result, df in Publication.groupby(\\'posres\\'):\\nby_result[result] = df\\nkm_result = km.fit(df[\\'time\\'], df[\\'status\\'])\\nkm_result.plot(label=\\'Result=%d\\' % result, ax=ax)\\nAs discussed previously, the p-values from fitting Cox’s proportional haz-\\nards model to the posres variable are quite large, providing no evidence\\nof a difference in time-to-publication between studies with positive versus\\nnegative results. 494\\n11. Survival Analysis and Censored Data\\nIn [19]: posres_df = MS([\\'posres\\',\\n\\'time\\',\\n\\'status\\'],\\nintercept=False).fit_transform(Publication)\\nposres_fit = coxph().fit(posres_df,\\n\\'time\\',\\n\\'status\\')\\nposres_fit.summary[[\\'coef\\', \\'se(coef)\\', \\'p\\']]\\nOut[19]:\\ncoef\\nse(coef)\\np\\ncovariate\\nposres\\n0.148076\\n0.161625\\n0.359578\\nHowever, the results change dramatically when we include other predic-\\ntors in the model. Here we exclude the funding mechanism variable.\\nIn [20]: model = MS(Publication.columns.drop(\\'mech\\'),\\nintercept=False)\\ncoxph().fit(model.fit_transform(Publication),\\n\\'time\\',\\n\\'status\\').summary[[\\'coef\\', \\'se(coef)\\', \\'p\\']]\\nOut[20]:\\ncoef\\nse(coef)\\np\\ncovariate\\nposres\\n0.570774\\n0.175960\\n1.179606e-03\\nmulti\\n-0.040863\\n0.251194\\n8.707727e-01\\nclinend\\n0.546180\\n0.262001\\n3.710099e-02\\nsampsize\\n0.000005\\n0.000015\\n7.506978e-01\\nbudget\\n0.004386\\n0.002464\\n7.511276e-02\\nimpact\\n0.058318\\n0.006676\\n2.426779e-18\\nWe see that there are a number of statistically significant variables, in-\\ncluding whether the trial focused on a clinical endpoint, the impact of the\\nstudy, and whether the study had positive or negative results.\\n11.8.3\\nCall Center Data\\nIn this section, we will simulate survival data using the relationship between\\ncumulative hazard and the survival function explored in Exercise 8. Our\\nsimulated data will represent the observed wait times (in seconds) for 2,000\\ncustomers who have phoned a call center. In this context, censoring occurs\\nif a customer hangs up before his or her call is answered.\\nThere are three covariates: Operators (the number of call center operators\\navailable at the time of the call, which can range from 5 to 15), Center\\n(either A, B, or C), and Time of day (Morning, Afternoon, or Evening). We\\ngenerate data for these covariates so that all possibilities are equally likely:\\nfor instance, morning, afternoon and evening calls are equally likely, and\\nany number of operators from 5 to 15 is equally likely.\\nIn [21]: rng = np.random.default_rng(10)\\nN = 2000\\nOperators = rng.choice(np.arange(5, 16),\\nN,\\nreplace=True) 11.8 Lab: Survival Analysis\\n495\\nCenter = rng.choice([\\'A\\', \\'B\\', \\'C\\'],\\nN,\\nreplace=True)\\nTime = rng.choice([\\'Morn.\\', \\'After.\\', \\'Even.\\'],\\nN,\\nreplace=True)\\nD = pd.DataFrame({\\'Operators\\': Operators,\\n\\'Center\\': pd.Categorical(Center),\\n\\'Time\\': pd.Categorical(Time)})\\nWe then build a model matrix (omitting the intercept)\\nIn [22]: model = MS([\\'Operators\\',\\n\\'Center\\',\\n\\'Time\\'],\\nintercept=False)\\nX = model.fit_transform(D)\\nIt is worthwhile to take a peek at the model matrix X, so that we can be\\nsure that we understand how the variables have been coded. By default,\\nthe levels of categorical variables are sorted and, as usual, the first column\\nof the one-hot encoding of the variable is dropped.\\nIn [23]: X[:5]\\nOut[23]:\\nOperators\\nCenter[B]\\nCenter[C]\\nTime[Even.]\\nTime[Morn.]\\n0\\n13\\n0.0\\n1.0\\n0.0\\n0.0\\n1\\n15\\n0.0\\n0.0\\n1.0\\n0.0\\n2\\n7\\n1.0\\n0.0\\n0.0\\n1.0\\n3\\n7\\n0.0\\n1.0\\n0.0\\n1.0\\n4\\n13\\n0.0\\n1.0\\n1.0\\n0.0\\nNext, we specify the coefficients and the hazard function.\\nIn [24]: true_beta = np.array([0.04, -0.3, 0, 0.2, -0.2])\\ntrue_linpred = X.dot(true_beta)\\nhazard = lambda t: 1e-5 * t\\nHere, we have set the coefficient associated with Operators to equal 0.04;\\nin other words, each additional operator leads to a e0.04 = 1.041-fold in-\\ncrease in the “risk” that the call will be answered, given the Center and\\nTime covariates. This makes sense: the greater the number of operators at\\nhand, the shorter the wait time! The coefficient associated with Center ==\\nB is −0.3, and Center == A is treated as the baseline. This means that the\\nrisk of a call being answered at Center B is 0.74 times the risk that it will\\nbe answered at Center A; in other words, the wait times are a bit longer\\nat Center B.\\nRecall from Section 2.3.7 the use of lambda for creating short functions\\non the fly. We use the function sim_time() from the ISLP.survival pack- sim_time()\\nage. This function uses the relationship between the survival function and\\ncumulative hazard S(t) = exp(−H(t)) and the specific form of the cumu-\\nlative hazard function in the Cox model to simulate data based on values\\nof the linear predictor true_linpred and the cumulative hazard. We need\\nto provide the cumulative hazard function, which we do here.\\nIn [25]: cum_hazard = lambda t: 1e-5 * t**2 / 2 496\\n11. Survival Analysis and Censored Data\\nWe are now ready to generate data under the Cox proportional hazards\\nmodel. We truncate the maximum time to 1000 seconds to keep simulated\\nwait times reasonable. The function sim_time() takes a linear predictor, a\\ncumulative hazard function and a random number generator.\\nIn [26]: W = np.array([sim_time(l, cum_hazard, rng)\\nfor l in true_linpred])\\nD[\\'Wait time\\'] = np.clip(W, 0, 1000)\\nWe now simulate our censoring variable, for which we assume 90% of\\ncalls were answered (Failed==1) before the customer hung up (Failed==0).\\nIn [27]: D[\\'Failed\\'] = rng.choice([1, 0],\\nN,\\np=[0.9, 0.1])\\nD[:5]\\nOut[27]:\\nOperators Center\\nTime\\nWait time\\nFailed\\n0\\n13\\nC\\nAfter.\\n525.064979\\n1\\n1\\n15\\nA\\nEven.\\n254.677835\\n1\\n2\\n7\\nB\\nMorn.\\n487.739224\\n1\\n3\\n7\\nC\\nMorn.\\n308.580292\\n1\\n4\\n13\\nC\\nEven.\\n154.174608\\n1\\nIn [28]: D[\\'Failed\\'].mean()\\nOut[28]: 0.8985\\nWe now plot Kaplan-Meier survival curves. First, we stratify by Center.\\nIn [29]: fig, ax = subplots(figsize=(8,8))\\nby_center = {}\\nfor center, df in D.groupby(\\'Center\\'):\\nby_center[center] = df\\nkm_center = km.fit(df[\\'Wait time\\'], df[\\'Failed\\'])\\nkm_center.plot(label=\\'Center=%s\\' % center, ax=ax)\\nax.set_title(\"Probability of Still Being on Hold\")\\nNext, we stratify by Time.\\nIn [30]: fig, ax = subplots(figsize=(8,8))\\nby_time = {}\\nfor time, df in D.groupby(\\'Time\\'):\\nby_time[time] = df\\nkm_time = km.fit(df[\\'Wait time\\'], df[\\'Failed\\'])\\nkm_time.plot(label=\\'Time=%s\\' % time, ax=ax)\\nax.set_title(\"Probability of Still Being on Hold\")\\nIt seems that calls at Call Center B take longer to be answered than calls\\nat Centers A and C. Similarly, it appears that wait times are longest in the\\nmorning and shortest in the evening hours. We can use a log-rank test to\\ndetermine whether these differences are statistically significant using the\\nfunction multivariate_logrank_test().\\nIn [31]: multivariate_logrank_test(D[\\'Wait time\\'],\\nD[\\'Center\\'],\\nD[\\'Failed\\']) 11.8 Lab: Survival Analysis\\n497\\nOut[31]:\\nt_0\\n-1\\nnull_distribution\\nchi squared\\ndegrees_of_freedom\\n2\\ntest_name\\nmultivariate_logrank_test\\ntest_statistic\\np\\n-log2(p)\\n20.30 <0.005\\n14.65\\nNext, we consider the effect of Time.\\nIn [32]: multivariate_logrank_test(D[\\'Wait time\\'],\\nD[\\'Time\\'],\\nD[\\'Failed\\'])\\nOut[32]:\\nt_0\\n-1\\nnull_distribution\\nchi squared\\ndegrees_of_freedom\\n2\\ntest_name\\nmultivariate_logrank_test\\ntest_statistic\\np\\n-log2(p)\\n49.90 <0.005\\n35.99\\nAs in the case of a categorical variable with 2 levels, these results are\\nsimilar to the likelihood ratio test from the Cox proportional hazards model.\\nFirst, we look at the results for Center.\\nIn [33]: X = MS([\\'Wait time\\',\\n\\'Failed\\',\\n\\'Center\\'],\\nintercept=False).fit_transform(D)\\nF = coxph().fit(X, \\'Wait time\\', \\'Failed\\')\\nF.log_likelihood_ratio_test()\\nOut[33]:\\nnull_distribution\\nchi squared\\ndegrees_freedom\\n2\\ntest_name\\nlog-likelihood ratio test\\ntest_statistic\\np\\n-log2(p)\\n20.58 <0.005\\n14.85\\nNext, we look at the results for Time.\\nIn [34]: X = MS([\\'Wait time\\',\\n\\'Failed\\',\\n\\'Time\\'],\\nintercept=False).fit_transform(D)\\nF = coxph().fit(X, \\'Wait time\\', \\'Failed\\')\\nF.log_likelihood_ratio_test()\\nOut[34]:\\nnull_distribution\\nchi squared\\ndegrees_freedom\\n2\\ntest_name\\nlog-likelihood ratio test\\ntest_statistic\\np\\n-log2(p)\\n48.12 <0.005\\n34.71\\nWe find that differences between centers are highly significant, as are\\ndifferences between times of day.\\nFinally, we fit Cox’s proportional hazards model to the data. 498\\n11. Survival Analysis and Censored Data\\nIn [35]: X = MS(D.columns,\\nintercept=False).fit_transform(D)\\nfit_queuing = coxph().fit(\\nX,\\n\\'Wait time\\',\\n\\'Failed\\')\\nfit_queuing.summary[[\\'coef\\', \\'se(coef)\\', \\'p\\']]\\nOut[35]:\\ncoef\\nse(coef)\\np\\ncovariate\\nOperators\\n0.043934\\n0.007520\\n5.143677e-09\\nCenter[B]\\n-0.236059\\n0.058113\\n4.864734e-05\\nCenter[C]\\n0.012231\\n0.057518\\n8.316083e-01\\nTime[Even.]\\n0.268845\\n0.057797\\n3.294914e-06\\nTime[Morn.]\\n-0.148215\\n0.057334\\n9.734378e-03\\nThe p-values for Center B and evening time are very small. It is also\\nclear that the hazard — that is, the instantaneous risk that a call will be\\nanswered — increases with the number of operators. Since we generated\\nthe data ourselves, we know that the true coefficients for Operators, Center\\n= B, Center = C, Time = Even. and Time = Morn. are 0.04, −0.3, 0, 0.2, and\\n−0.2, respectively. The coefficient estimates from the fitted Cox model are\\nfairly accurate.\\n11.9\\nExercises\\nConceptual\\n1. For each example, state whether or not the censoring mechanism is\\nindependent. Justify your answer.\\n(a) In a study of disease relapse, due to a careless research scientist,\\nall patients whose phone numbers begin with the number “2”\\nare lost to follow up.\\n(b) In a study of longevity, a formatting error causes all patient ages\\nthat exceed 99 years to be lost (i.e. we know that those patients\\nare more than 99 years old, but we do not know their exact\\nages).\\n(c) Hospital A conducts a study of longevity. However, very sick\\npatients tend to be transferred to Hospital B, and are lost to\\nfollow up.\\n(d) In a study of unemployment duration, the people who find work\\nearlier are less motivated to stay in touch with study investiga-\\ntors, and therefore are more likely to be lost to follow up.\\n(e) In a study of pregnancy duration, women who deliver their ba-\\nbies pre-term are more likely to do so away from their usual\\nhospital, and thus are more likely to be censored, relative to\\nwomen who deliver full-term babies. 11.9 Exercises\\n499\\n(f) A researcher wishes to model the number of years of education\\nof the residents of a small town. Residents who enroll in college\\nout of town are more likely to be lost to follow up, and are\\nalso more likely to attend graduate school, relative to those who\\nattend college in town.\\n(g) Researchers conduct a study of disease-free survival (i.e. time\\nuntil disease relapse following treatment). Patients who have\\nnot relapsed within five years are considered to be cured, and\\nthus their survival time is censored at five years.\\n(h) We wish to model the failure time for some electrical component.\\nThis component can be manufactured in Iowa or in Pittsburgh,\\nwith no difference in quality. The Iowa factory opened five years\\nago, and so components manufactured in Iowa are censored at\\nfive years. The Pittsburgh factory opened two years ago, so those\\ncomponents are censored at two years.\\n(i) We wish to model the failure time of an electrical component\\nmade in two different factories, one of which opened before the\\nother. We have reason to believe that the components manufac-\\ntured in the factory that opened earlier are of higher quality.\\n2. We conduct a study with n = 4 participants who have just purchased\\ncell phones, in order to model the time until phone replacement. The\\nfirst participant replaces her phone after 1.2 years. The second par-\\nticipant still has not replaced her phone at the end of the two-year\\nstudy period. The third participant changes her phone number and is\\nlost to follow up (but has not yet replaced her phone) 1.5 years into\\nthe study. The fourth participant replaces her phone after 0.2 years.\\nFor each of the four participants (i = 1, . . . , 4), answer the following\\nquestions using the notation introduced in Section 11.1:\\n(a) Is the participant’s cell phone replacement time censored?\\n(b) Is the value of ci known, and if so, then what is it?\\n(c) Is the value of ti known, and if so, then what is it?\\n(d) Is the value of yi known, and if so, then what is it?\\n(e) Is the value of δi known, and if so, then what is it?\\n3. For the example in Exercise 2, report the values of K, d1, . . . , dK,\\nr1, . . . , rK, and q1, . . . , qK, where this notation was defined in Sec-\\ntion 11.3.\\n4. This problem makes use of the Kaplan-Meier survival curve displayed\\nin Figure 11.9. The raw data that went into plotting this survival\\ncurve is given in Table 11.4. The covariate column of that table is\\nnot needed for this problem.\\n(a) What is the estimated probability of survival past 50 days? 500\\n11. Survival Analysis and Censored Data\\nObservation (Y )\\nCensoring Indicator (δ)\\nCovariate (X)\\n26.5\\n1\\n0.1\\n37.2\\n1\\n11\\n57.3\\n1\\n-0.3\\n90.8\\n0\\n2.8\\n20.2\\n0\\n1.8\\n89.8\\n0\\n0.4\\nTABLE 11.4. Data used in Exercise 4.\\n(b) Write out an analytical expression for the estimated survival\\nfunction. For instance, your answer might be something along\\nthe lines of\\nWS(t) =\\n\\uf8f1\\n\\uf8f4\\n\\uf8f2\\n\\uf8f4\\n\\uf8f3\\n0.8\\nif t < 31\\n0.5\\nif 31 ≤t < 77\\n0.22\\nif 77 ≤t.\\n(The previous equation is for illustration only: it is not the cor-\\nrect answer!)\\n5. Sketch the survival function given by the equation\\nWS(t) =\\n\\uf8f1\\n\\uf8f4\\n\\uf8f2\\n\\uf8f4\\n\\uf8f3\\n0.8\\nif t < 31\\n0.5\\nif 31 ≤t < 77\\n0.22\\nif 77 ≤t.\\nYour answer should look something like Figure 11.9.\\n0\\n20\\n40\\n60\\n80\\n0.0\\n0.2\\n0.4\\n0.6\\n0.8\\n1.0\\nTime in Days\\nEstimated Probability of Survival\\nFIGURE 11.9. A Kaplan-Meier survival curve used in Exercise 4.\\n6. This problem makes use of the data displayed in Figure 11.1. In\\ncompleting this problem, you can refer to the observation times as\\ny1, . . . , y4. The ordering of these observation times can be seen from\\nFigure 11.1; their exact values are not required.\\n(a) Report the values of δ1, . . . , δ4, K, d1, . . . , dK, r1, . . . , rK, and\\nq1, . . . , qK. The relevant notation is defined in Sections 11.1 and\\n11.3. 11.9 Exercises\\n501\\n(b) Sketch the Kaplan-Meier survival curve corresponding to this\\ndata set. (You do not need to use any software to do this — you\\ncan sketch it by hand using the results obtained in (a).)\\n(c) Based on the survival curve estimated in (b), what is the proba-\\nbility that the event occurs within 200 days? What is the prob-\\nability that the event does not occur within 310 days?\\n(d) Write out an expression for the estimated survival curve from\\n(b).\\n7. In this problem, we will derive (11.5) and (11.6), which are needed\\nfor the construction of the log-rank test statistic (11.8). Recall the\\nnotation in Table 11.1.\\n(a) Assume that there is no difference between the survival functions\\nof the two groups. Then we can think of q1k as the number of\\nfailures if we draw r1k observations, without replacement, from\\na risk set of rk observations that contains a total of qk failures.\\nArgue that q1k follows a hypergeometric distribution. Write the hyper-\\ngeometric\\ndistribution\\nparameters of this distribution in terms of r1k, rk, and qk.\\n(b) Given your previous answer, and the properties of the hyper-\\ngeometric distribution, what are the mean and variance of q1k?\\nCompare your answer to (11.5) and (11.6).\\n8. Recall that the survival function S(t), the hazard function h(t), and\\nthe density function f(t) are defined in (11.2), (11.9), and (11.11),\\nrespectively. Furthermore, define F(t) = 1 −S(t). Show that the\\nfollowing relationships hold:\\nf(t)\\n=\\ndF(t)/dt\\nS(t)\\n=\\nexp\\n*\\n−\\nL t\\n0\\nh(u)du\\n+\\n.\\n9. In this exercise, we will explore the consequences of assuming that\\nthe survival times follow an exponential distribution.\\n(a) Suppose that a survival time follows an Exp(λ) distribution,\\nso that its density function is f(t) = λ exp(−λt). Using the\\nrelationships provided in Exercise 8, show that S(t) = exp(−λt).\\n(b) Now suppose that each of n independent survival times follows\\nan Exp(λ) distribution. Write out an expression for the likeli-\\nhood function (11.13).\\n(c) Show that the maximum likelihood estimator for λ is\\nˆλ =\\nn\\n0\\ni=1\\nδi/\\nn\\n0\\ni=1\\nyi.\\n(d) Use your answer to (c) to derive an estimator of the mean sur-\\nvival time.\\nHint: For (d), recall that the mean of an Exp(λ) random variable is\\n1/λ. 502\\n11. Survival Analysis and Censored Data\\nApplied\\n10. This exercise focuses on the brain tumor data, which is included in\\nthe ISLP library.\\n(a) Plot the Kaplan-Meier survival curve with ±1 standard error\\nbands, using the KaplanMeierFitter() estimator in the lifelines\\npackage.\\n(b) Draw a bootstrap sample of size n = 88 from the pairs (yi, δi),\\nand compute the resulting Kaplan-Meier survival curve. Repeat\\nthis process B = 200 times. Use the results to obtain an estimate\\nof the standard error of the Kaplan-Meier survival curve at each\\ntimepoint. Compare this to the standard errors obtained in (a).\\n(c) Fit a Cox proportional hazards model that uses all of the pre-\\ndictors to predict survival. Summarize the main findings.\\n(d) Stratify the data by the value of ki. (Since only one observation\\nhas ki==40, you can group that observation together with the ob-\\nservations that have ki==60.) Plot Kaplan-Meier survival curves\\nfor each of the five strata, adjusted for the other predictors.\\n11. This exercise makes use of the data in Table 11.4.\\n(a) Create two groups of observations. In Group 1, X < 2, whereas\\nin Group 2, X ≥2. Plot the Kaplan-Meier survival curves corre-\\nsponding to the two groups. Be sure to label the curves so that\\nit is clear which curve corresponds to which group. By eye, does\\nthere appear to be a difference between the two groups’ survival\\ncurves?\\n(b) Fit Cox’s proportional hazards model, using the group indicator\\nas a covariate. What is the estimated coefficient? Write a sen-\\ntence providing the interpretation of this coefficient, in terms\\nof the hazard or the instantaneous probability of the event. Is\\nthere evidence that the true coefficient value is non-zero?\\n(c) Recall from Section 11.5.2 that in the case of a single binary\\ncovariate, the log-rank test statistic should be identical to the\\nscore statistic for the Cox model. Conduct a log-rank test to de-\\ntermine whether there is a difference between the survival curves\\nfor the two groups. How does the p-value for the log-rank test\\nstatistic compare to the p-value for the score statistic for the\\nCox model from (b)? 12\\nUnsupervised Learning\\nMost\\nof\\nthis\\nbook\\nconcerns\\nsupervised\\nlearning\\nmethods\\nsuch\\nas\\nregression and classification. In the supervised learning setting, we typically\\nhave access to a set of p features X1, X2, . . . , Xp, measured on n obser-\\nvations, and a response Y also measured on those same n observations.\\nThe goal is then to predict Y using X1, X2, . . . , Xp.\\nThis chapter will instead focus on unsupervised learning, a set of sta-\\ntistical tools intended for the setting in which we have only a set of fea-\\ntures X1, X2, . . . , Xp measured on n observations. We are not interested\\nin prediction, because we do not have an associated response variable Y .\\nRather, the goal is to discover interesting things about the measurements\\non X1, X2, . . . , Xp. Is there an informative way to visualize the data? Can\\nwe discover subgroups among the variables or among the observations?\\nUnsupervised learning refers to a diverse set of techniques for answering\\nquestions such as these. In this chapter, we will focus on two particu-\\nlar types of unsupervised learning: principal components analysis, a tool\\nused for data visualization or data pre-processing before supervised tech-\\nniques are applied, and clustering, a broad class of methods for discovering\\nunknown subgroups in data.\\n12.1\\nThe Challenge of Unsupervised Learning\\nSupervised learning is a well-understood area. In fact, if you have read\\nthe preceding chapters in this book, then you should by now have a good\\ngrasp of supervised learning. For instance, if you are asked to predict a\\nbinary outcome from a data set, you have a very well developed set of tools\\nat your disposal (such as logistic regression, linear discriminant analysis,\\nclassification trees, support vector machines, and more) as well as a clear\\n© Springer Nature Switzerland AG 2023 \\nG. James et al., An Introduction to Statistical Learning, Springer Texts in Statistics, \\nhttps://doi.org/10.1007/978-3-031-38747-0_12 \\n503 504\\n12. Unsupervised Learning\\nunderstanding of how to assess the quality of the results obtained (using\\ncross-validation, validation on an independent test set, and so forth).\\nIn contrast, unsupervised learning is often much more challenging. The\\nexercise tends to be more subjective, and there is no simple goal for the\\nanalysis, such as prediction of a response. Unsupervised learning is often\\nperformed as part of an exploratory data analysis. Furthermore, it can be exploratory\\ndata\\nanalysis\\nhard to assess the results obtained from unsupervised learning methods,\\nsince there is no universally accepted mechanism for performing cross-\\nvalidation or validating results on an independent data set. The reason\\nfor this difference is simple. If we fit a predictive model using a supervised\\nlearning technique, then it is possible to check our work by seeing how\\nwell our model predicts the response Y on observations not used in fitting\\nthe model. However, in unsupervised learning, there is no way to check our\\nwork because we don’t know the true answer—the problem is unsupervised.\\nTechniques for unsupervised learning are of growing importance in a\\nnumber of fields. A cancer researcher might assay gene expression levels in\\n100 patients with breast cancer. He or she might then look for subgroups\\namong the breast cancer samples, or among the genes, in order to obtain\\na better understanding of the disease. An online shopping site might try\\nto identify groups of shoppers with similar browsing and purchase histo-\\nries, as well as items that are of particular interest to the shoppers within\\neach group. Then an individual shopper can be preferentially shown the\\nitems in which he or she is particularly likely to be interested, based on\\nthe purchase histories of similar shoppers. A search engine might choose\\nwhich search results to display to a particular individual based on the click\\nhistories of other individuals with similar search patterns. These statistical\\nlearning tasks, and many more, can be performed via unsupervised learning\\ntechniques.\\n12.2\\nPrincipal Components Analysis\\nPrincipal components are discussed in Section 6.3.1 in the context of\\nprincipal components regression. When faced with a large set of corre-\\nlated variables, principal components allow us to summarize this set with\\na smaller number of representative variables that collectively explain most\\nof the variability in the original set. The principal component directions\\nare presented in Section 6.3.1 as directions in feature space along which\\nthe original data are highly variable. These directions also define lines and\\nsubspaces that are as close as possible to the data cloud. To perform\\nprincipal components regression, we simply use principal components as\\npredictors in a regression model in place of the original larger set of vari-\\nables.\\nPrincipal components analysis (PCA) refers to the process by which prin- principal\\ncomponents\\nanalysis\\ncipal components are computed, and the subsequent use of these compo-\\nnents in understanding the data. PCA is an unsupervised approach, since\\nit involves only a set of features X1, X2, . . . , Xp, and no associated response\\nY . Apart from producing derived variables for use in supervised learning\\nproblems, PCA also serves as a tool for data visualization (visualization of 12.2 Principal Components Analysis\\n505\\nthe observations or visualization of the variables). It can also be used as a\\ntool for data imputation — that is, for filling in missing values in a data\\nmatrix.\\nWe now discuss PCA in greater detail, focusing on the use of PCA as\\na tool for unsupervised data exploration, in keeping with the topic of this\\nchapter.\\n12.2.1\\nWhat Are Principal Components?\\nSuppose that we wish to visualize n observations with measurements on a\\nset of p features, X1, X2, . . . , Xp, as part of an exploratory data analysis.\\nWe could do this by examining two-dimensional scatterplots of the data,\\neach of which contains the n observations’ measurements on two of the\\nfeatures. However, there are\\n\\'p\\n2\\n(\\n= p(p−1)/2 such scatterplots; for example,\\nwith p = 10 there are 45 plots! If p is large, then it will certainly not be\\npossible to look at all of them; moreover, most likely none of them will\\nbe informative since they each contain just a small fraction of the total\\ninformation present in the data set. Clearly, a better method is required to\\nvisualize the n observations when p is large. In particular, we would like to\\nfind a low-dimensional representation of the data that captures as much of\\nthe information as possible. For instance, if we can obtain a two-dimensional\\nrepresentation of the data that captures most of the information, then we\\ncan plot the observations in this low-dimensional space.\\nPCA provides a tool to do just this. It finds a low-dimensional represen-\\ntation of a data set that contains as much as possible of the variation. The\\nidea is that each of the n observations lives in p-dimensional space, but not\\nall of these dimensions are equally interesting. PCA seeks a small number\\nof dimensions that are as interesting as possible, where the concept of in-\\nteresting is measured by the amount that the observations vary along each\\ndimension. Each of the dimensions found by PCA is a linear combination\\nof the p features. We now explain the manner in which these dimensions,\\nor principal components, are found.\\nThe first principal component of a set of features X1, X2, . . . , Xp is the\\nnormalized linear combination of the features\\nZ1 = φ11X1 + φ21X2 + · · · + φp1Xp\\n(12.1)\\nthat has the largest variance. By normalized, we mean that )p\\nj=1 φ2\\nj1 = 1.\\nWe refer to the elements φ11, . . . , φp1 as the loadings of the first principal loading\\ncomponent; together, the loadings make up the principal component load-\\ning vector, φ1 = (φ11 φ21 . . . φp1)T . We constrain the loadings so that\\ntheir sum of squares is equal to one, since otherwise setting these elements\\nto be arbitrarily large in absolute value could result in an arbitrarily large\\nvariance.\\nGiven an n × p data set X, how do we compute the first principal com-\\nponent? Since we are only interested in variance, we assume that each of\\nthe variables in X has been centered to have mean zero (that is, the col-\\numn means of X are zero). We then look for the linear combination of the\\nsample feature values of the form\\nzi1 = φ11xi1 + φ21xi2 + · · · + φp1xip\\n(12.2) 506\\n12. Unsupervised Learning\\nthat has largest sample variance, subject to the constraint that )p\\nj=1 φ2\\nj1=1.\\nIn other words, the first principal component loading vector solves the op-\\ntimization problem\\nmaximize\\nφ11,...,φp1\\n\\uf8f1\\n\\uf8f4\\n\\uf8f2\\n\\uf8f4\\n\\uf8f3\\n1\\nn\\nn\\n0\\ni=1\\n\\uf8eb\\n\\uf8ed\\np\\n0\\nj=1\\nφj1xij\\n\\uf8f6\\n\\uf8f8\\n2\\uf8fc\\n\\uf8f4\\n\\uf8fd\\n\\uf8f4\\n\\uf8fe\\nsubject to\\np\\n0\\nj=1\\nφ2\\nj1 = 1.\\n(12.3)\\nFrom (12.2) we can write the objective in (12.3) as\\n1\\nn\\n)n\\ni=1 z2\\ni1. Since\\n1\\nn\\n)n\\ni=1 xij = 0, the average of the z11, . . . , zn1 will be zero as well. Hence\\nthe objective that we are maximizing in (12.3) is just the sample variance of\\nthe n values of zi1. We refer to z11, . . . , zn1 as the scores of the first princi- score\\npal component. Problem (12.3) can be solved via an eigen decomposition, eigen decom-\\nposition\\na standard technique in linear algebra, but the details are outside of the\\nscope of this book.1\\nThere is a nice geometric interpretation of the first principal component.\\nThe loading vector φ1 with elements φ11, φ21, . . . , φp1 defines a direction in\\nfeature space along which the data vary the most. If we project the n data\\npoints x1, . . . , xn onto this direction, the projected values are the princi-\\npal component scores z11, . . . , zn1 themselves. For instance, Figure 6.14 on\\npage 254 displays the first principal component loading vector (green solid\\nline) on an advertising data set. In these data, there are only two features,\\nand so the observations as well as the first principal component loading\\nvector can be easily displayed. As can be seen from (6.19), in that data set\\nφ11 = 0.839 and φ21 = 0.544.\\nAfter the first principal component Z1 of the features has been deter-\\nmined, we can find the second principal component Z2. The second princi-\\npal component is the linear combination of X1, . . . , Xp that has maximal\\nvariance out of all linear combinations that are uncorrelated with Z1. The\\nsecond principal component scores z12, z22, . . . , zn2 take the form\\nzi2 = φ12xi1 + φ22xi2 + · · · + φp2xip,\\n(12.4)\\nwhere φ2 is the second principal component loading vector, with elements\\nφ12, φ22, . . . , φp2. It turns out that constraining Z2 to be uncorrelated with\\nZ1 is equivalent to constraining the direction φ2 to be orthogonal (perpen-\\ndicular) to the direction φ1. In the example in Figure 6.14, the observations\\nlie in two-dimensional space (since p = 2), and so once we have found φ1,\\nthere is only one possibility for φ2, which is shown as a blue dashed line.\\n(From Section 6.3.1, we know that φ12 = 0.544 and φ22 = −0.839.) But in\\na larger data set with p > 2 variables, there are multiple distinct principal\\ncomponents, and they are defined in a similar manner. To find φ2, we solve\\na problem similar to (12.3) with φ2 replacing φ1, and with the additional\\nconstraint that φ2 is orthogonal to φ1.2\\n1As an alternative to the eigen decomposition, a related technique called the singular\\nvalue decomposition can be used. This will be explored in the lab at the end of this\\nchapter.\\n2On a technical note, the principal component directions φ1, φ2, φ3, . . . are given\\nby the ordered sequence of eigenvectors of the matrix XT X, and the variances of the\\ncomponents are the eigenvalues. There are at most min(n −1, p) principal components. 12.2 Principal Components Analysis\\n507\\n−3\\n−2\\n−1\\n0\\n1\\n2\\n3\\n−3\\n−2\\n−1\\n0\\n1\\n2\\n3\\nFirst Principal Component\\nSecond Principal Component\\nAlabama\\nAlaska\\nArizona\\nArkansas\\nCalifornia\\nColorado\\nConnecticut\\nDelaware\\nFlorida\\nGeorgia\\nHawaii\\nIdaho\\nIllinois\\nIndiana\\nIowa\\nKansas\\nKentucky\\nLouisiana\\nMaine\\nMaryland\\nMassachusetts\\nMichigan\\nMinnesota\\nMississippi\\nMissouri\\nMontana\\nNebraska\\nNevada\\nNew Hampshire\\nNew Jersey\\nNew Mexico\\nNew York\\nNorth Carolina\\nrth Dakota\\nOhio\\nOklahoma\\nOregon\\nPennsylvania\\nRhode Island\\nSouth Carolina\\nSouth Dakota\\nTennessee\\nTexas\\nUtah\\nVermont\\nVirginia\\nWashington\\nWest Virginia\\nWisconsin\\nWyoming\\n−0.5\\n0.0\\n0.5\\n−0.5\\n0.0\\n0.5\\nMurder\\nAssault\\nUrbanPop\\nRape\\nFIGURE 12.1. The first two principal components for the USArrests data. The\\nblue state names represent the scores for the first two principal components. The\\norange arrows indicate the first two principal component loading vectors (with axes\\non the top and right). For example, the loading for Rape on the first component\\nis 0.54, and its loading on the second principal component 0.17 (the word Rape\\nis centered at the point (0.54, 0.17)). This figure is known as a biplot, because it\\ndisplays both the principal component scores and the principal component loadings.\\nOnce we have computed the principal components, we can plot them\\nagainst each other in order to produce low-dimensional views of the data.\\nFor instance, we can plot the score vector Z1 against Z2, Z1 against Z3,\\nZ2 against Z3, and so forth. Geometrically, this amounts to projecting the\\noriginal data down onto the subspace spanned by φ1, φ2, and φ3, and\\nplotting the projected points.\\nWe illustrate the use of PCA on the USArrests data set. For each of the\\n50 states in the United States, the data set contains the number of arrests\\nper 100, 000 residents for each of three crimes: Assault, Murder, and Rape.\\nWe also record UrbanPop (the percent of the population in each state living\\nin urban areas). The principal component score vectors have length n = 50,\\nand the principal component loading vectors have length p = 4. PCA was\\nperformed after standardizing each variable to have mean zero and standard 508\\n12. Unsupervised Learning\\nPC1\\nPC2\\nMurder\\n0.5358995\\n−0.4181809\\nAssault\\n0.5831836\\n−0.1879856\\nUrbanPop\\n0.2781909\\n0.8728062\\nRape\\n0.5434321\\n0.1673186\\nTABLE 12.1. The principal component loading vectors, φ1 and φ2, for the\\nUSArrests data. These are also displayed in Figure 12.1.\\ndeviation one. Figure 12.1 plots the first two principal components of these\\ndata. The figure represents both the principal component scores and the\\nloading vectors in a single biplot display. The loadings are also given in biplot\\nTable 12.2.1.\\nIn Figure 12.1, we see that the first loading vector places approximately\\nequal weight on Assault, Murder, and Rape, but with much less weight on\\nUrbanPop. Hence this component roughly corresponds to a measure of overall\\nrates of serious crimes. The second loading vector places most of its weight\\non UrbanPop and much less weight on the other three features. Hence, this\\ncomponent roughly corresponds to the level of urbanization of the state.\\nOverall, we see that the crime-related variables (Murder, Assault, and Rape)\\nare located close to each other, and that the UrbanPop variable is far from the\\nother three. This indicates that the crime-related variables are correlated\\nwith each other—states with high murder rates tend to have high assault\\nand rape rates—and that the UrbanPop variable is less correlated with the\\nother three.\\nWe can examine differences between the states via the two principal com-\\nponent score vectors shown in Figure 12.1. Our discussion of the loading\\nvectors suggests that states with large positive scores on the first compo-\\nnent, such as California, Nevada and Florida, have high crime rates, while\\nstates like North Dakota, with negative scores on the first component, have\\nlow crime rates. California also has a high score on the second component,\\nindicating a high level of urbanization, while the opposite is true for states\\nlike Mississippi. States close to zero on both components, such as Indiana,\\nhave approximately average levels of both crime and urbanization.\\n12.2.2\\nAnother Interpretation of Principal Components\\nThe first two principal component loading vectors in a simulated three-\\ndimensional data set are shown in the left-hand panel of Figure 12.2; these\\ntwo loading vectors span a plane along which the observations have the\\nhighest variance.\\nIn the previous section, we describe the principal component loading vec-\\ntors as the directions in feature space along which the data vary the most,\\nand the principal component scores as projections along these directions.\\nHowever, an alternative interpretation of principal components can also be 12.2 Principal Components Analysis\\n509\\nFirst principal component\\nSecond principal component\\n−1.0\\n−0.5\\n0.0\\n0.5\\n1.0\\n−1.0\\n−0.5\\n0.0\\n0.5\\n1.0\\n•\\n• •\\n•\\n•\\n•\\n•\\n•\\n•\\n•\\n•\\n•\\n•\\n•\\n•\\n•\\n•\\n•\\n•\\n•\\n•\\n•\\n•\\n• •\\n••\\n•\\n•\\n•\\n•\\n•\\n•\\n•\\n•\\n•\\n•\\n•\\n•\\n•\\n•\\n• •\\n•\\n•\\n•\\n•\\n•\\n•\\n•\\n•\\n•\\n•\\n•\\n• •\\n•\\n•\\n•\\n•\\n•\\n•\\n•\\n• •\\n•\\n•\\n•\\n•\\n•\\n•\\n•\\n•\\n•\\n•\\n•\\n•\\n•\\n•\\n•\\n•\\n•\\n•\\n•\\n•\\n•\\n•\\n•\\n•\\n•\\nFIGURE 12.2. Ninety observations simulated in three dimensions. The obser-\\nvations are displayed in color for ease of visualization. Left: the first two principal\\ncomponent directions span the plane that best fits the data. The plane is positioned\\nto minimize the sum of squared distances to each point. Right: the first two prin-\\ncipal component score vectors give the coordinates of the projection of the 90\\nobservations onto the plane.\\nuseful: principal components provide low-dimensional linear surfaces that\\nare closest to the observations. We expand upon that interpretation here.3\\nThe first principal component loading vector has a very special property:\\nit is the line in p-dimensional space that is closest to the n observations\\n(using average squared Euclidean distance as a measure of closeness). This\\ninterpretation can be seen in the left-hand panel of Figure 6.15; the dashed\\nlines indicate the distance between each observation and the line defined\\nby the first principal component loading vector. The appeal of this inter-\\npretation is clear: we seek a single dimension of the data that lies as close\\nas possible to all of the data points, since such a line will likely provide a\\ngood summary of the data.\\nThe notion of principal components as the dimensions that are clos-\\nest to the n observations extends beyond just the first principal com-\\nponent. For instance, the first two principal components of a data set\\nspan the plane that is closest to the n observations, in terms of average\\nsquared Euclidean distance. An example is shown in the left-hand panel\\nof Figure 12.2. The first three principal components of a data set span\\nthe three-dimensional hyperplane that is closest to the n observations, and\\nso forth.\\nUsing this interpretation, together the first M principal component score\\nvectors and the first M principal component loading vectors provide the\\nbest M-dimensional approximation (in terms of Euclidean distance) to\\n3In this section, we continue to assume that each column of the data matrix X has\\nbeen centered to have mean zero—that is, the column mean has been subtracted from\\neach column. 510\\n12. Unsupervised Learning\\nthe ith observation xij. This representation can be written as\\nxij ≈\\nM\\n0\\nm=1\\nzimφjm.\\n(12.5)\\nWe can state this more formally by writing down an optimization prob-\\nlem. Suppose the data matrix X is column-centered. Out of all approxima-\\ntions of the form xij ≈)M\\nm=1 aimbjm, we could ask for the one with the\\nsmallest residual sum of squares:\\nminimize\\nA∈Rn×M,B∈Rp×M\\n\\uf8f1\\n\\uf8f2\\n\\uf8f3\\np\\n0\\nj=1\\nn\\n0\\ni=1\\n>\\nxij −\\nM\\n0\\nm=1\\naimbjm\\n?2\\uf8fc\\n\\uf8fd\\n\\uf8fe.\\n(12.6)\\nHere, A is an n × M matrix whose (i, m) element is aim, and B is a p × M\\nelement whose (j, m) element is bjm.\\nIt can be shown that for any value of M, the columns of the matrices\\nˆA and ˆB that solve (12.6) are in fact the first M principal components\\nscore and loading vectors. In other words, if ˆA and ˆB solve (12.6), then\\nˆaim = zim and ˆbjm = φjm.4 This means that the smallest possible value of\\nthe objective in (12.6) is\\np\\n0\\nj=1\\nn\\n0\\ni=1\\n>\\nxij −\\nM\\n0\\nm=1\\nzimφjm\\n?2\\n.\\n(12.7)\\nIn summary, together the M principal component score vectors and M\\nprincipal component loading vectors can give a good approximation to the\\ndata when M is sufficiently large. When M = min(n −1, p), then the\\nrepresentation is exact: xij = )M\\nm=1 zimφjm.\\n12.2.3\\nThe Proportion of Variance Explained\\nIn Figure 12.2, we performed PCA on a three-dimensional data set (left-\\nhand panel) and projected the data onto the first two principal component\\nloading vectors in order to obtain a two-dimensional view of the data (i.e.\\nthe principal component score vectors; right-hand panel). We see that this\\ntwo-dimensional representation of the three-dimensional data does success-\\nfully capture the major pattern in the data: the orange, green, and cyan\\nobservations that are near each other in three-dimensional space remain\\nnearby in the two-dimensional representation. Similarly, we have seen on\\nthe USArrests data set that we can summarize the 50 observations and 4\\nvariables using just the first two principal component score vectors and the\\nfirst two principal component loading vectors.\\nWe can now ask a natural question: how much of the information in\\na given data set is lost by projecting the observations onto the first few\\nprincipal components? That is, how much of the variance in the data is not\\ncontained in the first few principal components? More generally, we are\\ninterested in knowing the proportion of variance explained (PVE) by each proportion\\nof variance\\nexplained\\n4Technically, the solution to (12.6) is not unique. Thus, it is more precise to state\\nthat any solution to (12.6) can be easily transformed to yield the principal components. 12.2 Principal Components Analysis\\n511\\nprincipal component. The total variance present in a data set (assuming\\nthat the variables have been centered to have mean zero) is defined as\\np\\n0\\nj=1\\nVar(Xj) =\\np\\n0\\nj=1\\n1\\nn\\nn\\n0\\ni=1\\nx2\\nij,\\n(12.8)\\nand the variance explained by the mth principal component is\\n1\\nn\\nn\\n0\\ni=1\\nz2\\nim = 1\\nn\\nn\\n0\\ni=1\\n\\uf8eb\\n\\uf8ed\\np\\n0\\nj=1\\nφjmxij\\n\\uf8f6\\n\\uf8f8\\n2\\n.\\n(12.9)\\nTherefore, the PVE of the mth principal component is given by\\n)n\\ni=1 z2\\nim\\n)p\\nj=1\\n)n\\ni=1 x2\\nij\\n=\\n)n\\ni=1\\n1)p\\nj=1 φjmxij\\n22\\n)p\\nj=1\\n)n\\ni=1 x2\\nij\\n.\\n(12.10)\\nThe PVE of each principal component is a positive quantity. In order to\\ncompute the cumulative PVE of the first M principal components, we can\\nsimply sum (12.10) over each of the first M PVEs. In total, there are\\nmin(n −1, p) principal components, and their PVEs sum to one.\\nIn Section 12.2.2, we showed that the first M principal component load-\\ning and score vectors can be interpreted as the best M-dimensional approx-\\nimation to the data, in terms of residual sum of squares. It turns out that\\nthe variance of the data can be decomposed into the variance of the first M\\nprincipal components plus the mean squared error of this M-dimensional\\napproximation, as follows:\\np\\n\\'\\nj=1\\n1\\nn\\nn\\n\\'\\ni=1\\nx2\\nij\\n(\\n)*\\n+\\nVar. of data\\n=\\nM\\n\\'\\nm=1\\n1\\nn\\nn\\n\\'\\ni=1\\nz2\\nim\\n(\\n)*\\n+\\nVar. of first M PCs\\n+ 1\\nn\\np\\n\\'\\nj=1\\nn\\n\\'\\ni=1\\n,\\nxij −\\nM\\n\\'\\nm=1\\nzimφjm\\n-2\\n(\\n)*\\n+\\nMSE of M-dimensional approximation\\n(12.11)\\nThe three terms in this decomposition are discussed in (12.8), (12.9), and\\n(12.7), respectively. Since the first term is fixed, we see that by maximizing\\nthe variance of the first M principal components, we minimize the mean\\nsquared error of the M-dimensional approximation, and vice versa. This ex-\\nplains why principal components can be equivalently viewed as minimizing\\nthe approximation error (as in Section 12.2.2) or maximizing the variance\\n(as in Section 12.2.1).\\nMoreover, we can use (12.11) to see that the PVE defined in (12.10)\\nequals\\n1 −\\n)p\\nj=1\\n)n\\ni=1\\n1\\nxij −)M\\nm=1 zimφjm\\n22\\n)p\\nj=1\\n)n\\ni=1 x2\\nij\\n= 1 −RSS\\nTSS,\\nwhere TSS represents the total sum of squared elements of X, and RSS\\nrepresents the residual sum of squares of the M-dimensional approxima-\\ntion given by the principal components. Recalling the definition of R2 from\\n(3.17), this means that we can interpret the PVE as the R2 of the approx-\\nimation for X given by the first M principal components. 512\\n12. Unsupervised Learning\\n1.0\\n1.5\\n2.0\\n2.5\\n3.0\\n3.5\\n4.0\\n0.0\\n0.2\\n0.4\\n0.6\\n0.8\\n1.0\\nPrincipal Component\\nProp. Variance Explained\\n1.0\\n1.5\\n2.0\\n2.5\\n3.0\\n3.5\\n4.0\\n0.0\\n0.2\\n0.4\\n0.6\\n0.8\\n1.0\\nPrincipal Component\\nCumulative Prop. Variance Explained\\nFIGURE 12.3. Left: a scree plot depicting the proportion of variance explained\\nby each of the four principal components in the USArrests data. Right: the cu-\\nmulative proportion of variance explained by the four principal components in the\\nUSArrests data.\\nIn the USArrests data, the first principal component explains 62.0 % of\\nthe variance in the data, and the next principal component explains 24.7 %\\nof the variance. Together, the first two principal components explain almost\\n87 % of the variance in the data, and the last two principal components\\nexplain only 13 % of the variance. This means that Figure 12.1 provides a\\npretty accurate summary of the data using just two dimensions. The PVE\\nof each principal component, as well as the cumulative PVE, is shown\\nin Figure 12.3. The left-hand panel is known as a scree plot, and will be scree plot\\ndiscussed later in this chapter.\\n12.2.4\\nMore on PCA\\nScaling the Variables\\nWe have already mentioned that before PCA is performed, the variables\\nshould be centered to have mean zero. Furthermore, the results obtained\\nwhen we perform PCA will also depend on whether the variables have\\nbeen individually scaled (each multiplied by a different constant). This is in\\ncontrast to some other supervised and unsupervised learning techniques,\\nsuch as linear regression, in which scaling the variables has no effect. (In\\nlinear regression, multiplying a variable by a factor of c will simply lead to\\nmultiplication of the corresponding coefficient estimate by a factor of 1/c,\\nand thus will have no substantive effect on the model obtained.)\\nFor instance, Figure 12.1 was obtained after scaling each of the variables\\nto have standard deviation one. This is reproduced in the left-hand plot in\\nFigure 12.4. Why does it matter that we scaled the variables? In these data,\\nthe variables are measured in different units; Murder, Rape, and Assault are\\nreported as the number of occurrences per 100, 000 people, and UrbanPop is\\nthe percentage of the state’s population that lives in an urban area. These\\nfour variables have variances of 18.97, 87.73, 6945.16, and 209.5, respec-\\ntively. Consequently, if we perform PCA on the unscaled variables, then 12.2 Principal Components Analysis\\n513\\n−3\\n−2\\n−1\\n0\\n1\\n2\\n3\\n−3\\n−2\\n−1\\n0\\n1\\n2\\n3\\nFirst Principal Component\\nSecond Principal Component\\n*\\n*\\n*\\n*\\n*\\n*\\n*\\n*\\n*\\n*\\n*\\n*\\n*\\n*\\n*\\n*\\n*\\n*\\n*\\n*\\n*\\n*\\n*\\n*\\n*\\n*\\n*\\n*\\n*\\n*\\n*\\n*\\n*\\n*\\n*\\n* *\\n*\\n*\\n*\\n*\\n*\\n*\\n*\\n*\\n*\\n*\\n*\\n*\\n*\\n−0.5\\n0.0\\n0.5\\n−0.5\\n0.0\\n0.5\\nMurder\\nAssault\\nUrbanPop\\nRape\\nScaled\\n−100\\n−50\\n0\\n50\\n100\\n150\\n−100\\n−50\\n0\\n50\\n100\\n150\\nFirst Principal Component\\nSecond Principal Component\\n*\\n*\\n*\\n*\\n*\\n*\\n*\\n*\\n*\\n*\\n*\\n*\\n*\\n*\\n*\\n*\\n*\\n*\\n*\\n*\\n*\\n*\\n*\\n*\\n*\\n*\\n*\\n*\\n*\\n*\\n*\\n*\\n*\\n*\\n*\\n* *\\n*\\n*\\n*\\n*\\n*\\n*\\n*\\n*\\n*\\n*\\n*\\n*\\n*\\n−0.5\\n0.0\\n0.5\\n1.0\\n−0.5\\n0.0\\n0.5\\n1.0\\nMurder\\nAssau\\nUrbanPop\\nRape\\nUnscaled\\nFIGURE 12.4. Two principal component biplots for the USArrests data. Left:\\nthe same as Figure 12.1, with the variables scaled to have unit standard deviations.\\nRight: principal components using unscaled data. Assault has by far the largest\\nloading on the first principal component because it has the highest variance among\\nthe four variables. In general, scaling the variables to have standard deviation one\\nis recommended.\\nthe first principal component loading vector will have a very large loading\\nfor Assault, since that variable has by far the highest variance. The right-\\nhand plot in Figure 12.4 displays the first two principal components for the\\nUSArrests data set, without scaling the variables to have standard devia-\\ntion one. As predicted, the first principal component loading vector places\\nalmost all of its weight on Assault, while the second principal component\\nloading vector places almost all of its weight on UrbanPop. Comparing this\\nto the left-hand plot, we see that scaling does indeed have a substantial\\neffect on the results obtained.\\nHowever, this result is simply a consequence of the scales on which the\\nvariables were measured. For instance, if Assault were measured in units\\nof the number of occurrences per 100 people (rather than number of oc-\\ncurrences per 100, 000 people), then this would amount to dividing all of\\nthe elements of that variable by 1, 000. Then the variance of the variable\\nwould be tiny, and so the first principal component loading vector would\\nhave a very small value for that variable. Because it is undesirable for the\\nprincipal components obtained to depend on an arbitrary choice of scaling,\\nwe typically scale each variable to have standard deviation one before we\\nperform PCA.\\nIn certain settings, however, the variables may be measured in the same\\nunits. In this case, we might not wish to scale the variables to have stan-\\ndard deviation one before performing PCA. For instance, suppose that the\\nvariables in a given data set correspond to expression levels for p genes.\\nThen since expression is measured in the same “units” for each gene, we\\nmight choose not to scale the genes to each have standard deviation one. 514\\n12. Unsupervised Learning\\nUniqueness of the Principal Components\\nWhile in theory the principal components need not be unique, in almost all\\npractical settings they are (up to sign flips). This means that two different\\nsoftware packages will yield the same principal component loading vectors,\\nalthough the signs of those loading vectors may differ. The signs may differ\\nbecause each principal component loading vector specifies a direction in p-\\ndimensional space: flipping the sign has no effect as the direction does not\\nchange. (Consider Figure 6.14—the principal component loading vector is\\na line that extends in either direction, and flipping its sign would have no\\neffect.) Similarly, the score vectors are unique up to a sign flip, since the\\nvariance of Z is the same as the variance of −Z. It is worth noting that\\nwhen we use (12.5) to approximate xij we multiply zim by φjm. Hence, if\\nthe sign is flipped on both the loading and score vectors, the final product\\nof the two quantities is unchanged.\\nDeciding How Many Principal Components to Use\\nIn general, an n × p data matrix X has min(n −1, p) distinct principal\\ncomponents. However, we usually are not interested in all of them; rather,\\nwe would like to use just the first few principal components in order to\\nvisualize or interpret the data. In fact, we would like to use the smallest\\nnumber of principal components required to get a good understanding of the\\ndata. How many principal components are needed? Unfortunately, there is\\nno single (or simple!) answer to this question.\\nWe typically decide on the number of principal components required\\nto visualize the data by examining a scree plot, such as the one shown\\nin the left-hand panel of Figure 12.3. We choose the smallest number of\\nprincipal components that are required in order to explain a sizable amount\\nof the variation in the data. This is done by eyeballing the scree plot, and\\nlooking for a point at which the proportion of variance explained by each\\nsubsequent principal component drops off. This drop is often referred to\\nas an elbow in the scree plot. For instance, by inspection of Figure 12.3,\\none might conclude that a fair amount of variance is explained by the first\\ntwo principal components, and that there is an elbow after the second\\ncomponent. After all, the third principal component explains less than ten\\npercent of the variance in the data, and the fourth principal component\\nexplains less than half that and so is essentially worthless.\\nHowever, this type of visual analysis is inherently ad hoc. Unfortunately,\\nthere is no well-accepted objective way to decide how many principal com-\\nponents are enough. In fact, the question of how many principal compo-\\nnents are enough is inherently ill-defined, and will depend on the specific\\narea of application and the specific data set. In practice, we tend to look\\nat the first few principal components in order to find interesting patterns\\nin the data. If no interesting patterns are found in the first few principal\\ncomponents, then further principal components are unlikely to be of inter-\\nest. Conversely, if the first few principal components are interesting, then\\nwe typically continue to look at subsequent principal components until no\\nfurther interesting patterns are found. This is admittedly a subjective ap- 12.3 Missing Values and Matrix Completion\\n515\\nproach, and is reflective of the fact that PCA is generally used as a tool for\\nexploratory data analysis.\\nOn the other hand, if we compute principal components for use in a\\nsupervised analysis, such as the principal components regression presented\\nin Section 6.3.1, then there is a simple and objective way to determine how\\nmany principal components to use: we can treat the number of principal\\ncomponent score vectors to be used in the regression as a tuning parameter\\nto be selected via cross-validation or a related approach. The comparative\\nsimplicity of selecting the number of principal components for a supervised\\nanalysis is one manifestation of the fact that supervised analyses tend to\\nbe more clearly defined and more objectively evaluated than unsupervised\\nanalyses.\\n12.2.5\\nOther Uses for Principal Components\\nWe saw in Section 6.3.1 that we can perform regression using the principal\\ncomponent score vectors as features. In fact, many statistical techniques,\\nsuch as regression, classification, and clustering, can be easily adapted to\\nuse the n × M matrix whose columns are the first M ≪p principal com-\\nponent score vectors, rather than using the full n × p data matrix. This\\ncan lead to less noisy results, since it is often the case that the signal (as\\nopposed to the noise) in a data set is concentrated in its first few principal\\ncomponents.\\n12.3\\nMissing Values and Matrix Completion\\nOften datasets have missing values, which can be a nuisance. For example,\\nsuppose that we wish to analyze the USArrests data, and discover that 20\\nof the 200 values have been randomly corrupted and marked as missing.\\nUnfortunately, the statistical learning methods that we have seen in this\\nbook cannot handle missing values. How should we proceed?\\nWe could remove the rows that contain missing observations and per-\\nform our data analysis on the complete rows. But this seems wasteful, and\\ndepending on the fraction missing, unrealistic. Alternatively, if xij is miss-\\ning, then we could replace it by the mean of the jth column (using the\\nnon-missing entries to compute the mean). Although this is a common and\\nconvenient strategy, often we can do better by exploiting the correlation\\nbetween the variables.\\nIn this section we show how principal components can be used to impute impute\\nimputation\\nthe missing values, through a process known as matrix completion. The\\nmatrix\\ncompletion\\ncompleted matrix can then be used in a statistical learning method, such\\nas linear regression or LDA.\\nThis approach for imputing missing data is appropriate if the missingness\\nis random. For example, it is suitable if a patient’s weight is missing because missing at\\nrandom\\nthe battery of the electronic scale was flat at the time of his exam. By\\ncontrast, if the weight is missing because the patient was too heavy to\\nclimb on the scale, then this is not missing at random; the missingness is 516\\n12. Unsupervised Learning\\ninformative, and the approach described here for handling missing data is\\nnot suitable.\\nSometimes data is missing by necessity. For example, if we form a matrix\\nof the ratings (on a scale from 1 to 5) that n customers have given to the\\nentire Netflix catalog of p movies, then most of the matrix will be missing,\\nsince no customer will have seen and rated more than a tiny fraction of the\\ncatalog. If we can impute the missing values well, then we will have an idea\\nof what each customer will think of movies they have not yet seen. Hence\\nmatrix completion can be used to power recommender systems.\\nrecommender\\nsystems\\nPrincipal Components with Missing Values\\nIn Section 12.2.2, we showed that the first M principal component score\\nand loading vectors provide the “best” approximation to the data matrix\\nX, in the sense of (12.6). Suppose that some of the observations xij are\\nmissing. We now show how one can both impute the missing values and\\nsolve the principal component problem at the same time. We return to a\\nmodified form of the optimization problem (12.6),\\nminimize\\nA∈Rn×M,B∈Rp×M\\n\\uf8f1\\n\\uf8f2\\n\\uf8f3\\n0\\n(i,j)∈O\\n>\\nxij −\\nM\\n0\\nm=1\\naimbjm\\n?2\\uf8fc\\n\\uf8fd\\n\\uf8fe,\\n(12.12)\\nwhere O is the set of all observed pairs of indices (i, j), a subset of the\\npossible n × p pairs.\\nOnce we solve this problem:\\n• we can estimate a missing observation xij using ˆxij = )M\\nm=1 ˆaimˆbjm,\\nwhere ˆaim and ˆbjm are the (i, m) and (j, m) elements, respectively,\\nof the matrices ˆA and ˆB that solve (12.12); and\\n• we can (approximately) recover the M principal component scores\\nand loadings, as we did when the data were complete.\\nIt turns out that solving (12.12) exactly is difficult, unlike in the case of\\ncomplete data: the eigen decomposition no longer applies. But the sim-\\nple iterative approach in Algorithm 12.1, which is demonstrated in Sec-\\ntion 12.5.2, typically provides a good solution.56\\nWe illustrate Algorithm 12.1 on the USArrests data. There are p = 4\\nvariables and n = 50 observations (states). We first standardized the data\\nso each variable has mean zero and standard deviation one. We then ran-\\ndomly selected 20 of the 50 states, and then for each of these we randomly\\nset one of the four variables to be missing. Thus, 10% of the elements of the\\ndata matrix were missing. We applied Algorithm 12.1 with M = 1 principal\\ncomponent. Figure 12.5 shows that the recovery of the missing elements\\n5This algorithm is referred to as “Hard-Impute” in Mazumder, Hastie, and Tibshi-\\nrani (2010) “Spectral regularization algorithms for learning large incomplete matrices”,\\npublished in Journal of Machine Learning Research, pages 2287–2322.\\n6Each iteration of Step 2 of this algorithm decreases the objective (12.14). However,\\nthe algorithm is not guaranteed to achieve the global optimum of (12.12). 12.3 Missing Values and Matrix Completion\\n517\\nAlgorithm 12.1 Iterative Algorithm for Matrix Completion\\n1. Create a complete data matrix ˜X of dimension n × p of which the\\n(i, j) element equals\\n˜xij =\\nK xij\\nif (i, j) ∈O\\n¯xj\\nif (i, j) /∈O,\\nwhere ¯xj is the average of the observed values for the jth variable in\\nthe incomplete data matrix X. Here, O indexes the observations that\\nare observed in X.\\n2. Repeat steps (a)–(c) until the objective (12.14) fails to decrease:\\n(a) Solve\\nminimize\\nA∈Rn×M,B∈Rp×M\\n\\uf8f1\\n\\uf8f2\\n\\uf8f3\\np\\n0\\nj=1\\nn\\n0\\ni=1\\n>\\n˜xij −\\nM\\n0\\nm=1\\naimbjm\\n?2\\uf8fc\\n\\uf8fd\\n\\uf8fe\\n(12.13)\\nby computing the principal components of ˜X.\\n(b) For each element (i, j) /∈O, set ˜xij ←)M\\nm=1 ˆaimˆbjm.\\n(c) Compute the objective\\n0\\n(i,j)∈O\\n>\\nxij −\\nM\\n0\\nm=1\\nˆaimˆbjm\\n?2\\n.\\n(12.14)\\n3. Return the estimated missing entries ˜xij, (i, j) /∈O.\\nis pretty accurate. Over 100 random runs of this experiment, the average\\ncorrelation between the true and imputed values of the missing elements\\nis 0.63, with a standard deviation of 0.11. Is this good performance? To\\nanswer this question, we can compare this correlation to what we would\\nhave gotten if we had estimated these 20 values using the complete data\\n— that is, if we had simply computed ˆxij = zi1φj1, where zi1 and φj1 are\\nelements of the first principal component score and loading vectors of the\\ncomplete data.7 Using the complete data in this way results in an average\\ncorrelation of 0.79 between the true and estimated values for these 20 el-\\nements, with a standard deviation of 0.08. Thus, our imputation method\\ndoes worse than the method that uses all of the data (0.63 ± 0.11 versus\\n0.79 ± 0.08), but its performance is still pretty good. (And of course, the\\nmethod that uses all of the data cannot be applied in a real-world setting\\nwith missing data.)\\nFigure 12.6 further indicates that Algorithm 12.1 performs fairly well on\\nthis dataset.\\n7This is an unattainable gold standard, in the sense that with missing data, we of\\ncourse cannot compute the principal components of the complete data. 518\\n12. Unsupervised Learning\\nFIGURE 12.5. Missing value imputation on the USArrests data. Twenty values\\n(10% of the total number of matrix elements) were artificially set to be missing,\\nand then imputed via Algorithm 12.1 with M = 1. The figure displays the true\\nvalue xij and the imputed value ˆxij for all twenty missing values. For each of the\\ntwenty missing values, the color indicates the variable, and the label indicates the\\nstate. The correlation between the true and imputed values is around 0.63.\\nWe close with a few observations:\\n• The USArrests data has only four variables, which is on the low end\\nfor methods like Algorithm 12.1 to work well. For this reason, for this\\ndemonstration we randomly set at most one variable per state to be\\nmissing, and only used M = 1 principal component.\\n• In general, in order to apply Algorithm 12.1, we must select M, the\\nnumber of principal components to use for the imputation. One ap-\\nproach is to randomly leave out a few additional elements from the\\nmatrix, and select M based on how well those known values are re-\\ncovered. This is closely related to the validation-set approach seen in\\nChapter 5.\\nRecommender Systems\\nDigital streaming services like Netflix and Amazon use data about the con-\\ntent that a customer has viewed in the past, as well as data from other\\ncustomers, to suggest other content for the customer. As a concrete ex-\\nample, some years back, Netflix had customers rate each movie that they\\nhad seen with a score from 1–5. This resulted in a very big n × p matrix\\nfor which the (i, j) element is the rating given by the ith customer to the\\n−1.5\\n−1.0\\n−0.5\\n0.0\\n0.5\\n1.0\\n1.5\\n−1.5\\n−1.0\\n−0.5\\n0.0\\n0.5\\n1.0\\n1.5\\nOriginal Value\\nImputed Value\\n●\\n●\\n●\\n●\\n●\\n●\\n●\\n●\\n●\\n●\\n●\\n●\\n●\\n●\\n●\\n●\\n●\\n●\\nOR\\nWA\\nTN\\nPA\\nMO\\nID\\nMD\\nVA\\nAL\\nTX\\nWY\\nMN\\nAK\\nUT\\nGA\\nMT\\nMA\\nNY\\n●\\n●\\n●\\n●\\nMurder\\nAssault\\nUrbanPop\\nRape 12.3 Missing Values and Matrix Completion\\n519\\nFIGURE 12.6. As described in the text, in each of 100 trials, we left out 20\\nelements of the USArrests dataset. In each trial, we applied Algorithm 12.1 with\\nM = 1 to impute the missing elements and compute the principal components.\\nLeft: For each of the 50 states, the imputed first principal component scores\\n(averaged over 100 trials, and displayed with a standard deviation bar) are plotted\\nagainst the first principal component scores computed using all the data. Right:\\nThe imputed principal component loadings (averaged over 100 trials, and displayed\\nwith a standard deviation bar) are plotted against the true principal component\\nloadings.\\njth movie. One specific early example of this matrix had n = 480,189 cus-\\ntomers and p = 17,770 movies. However, on average each customer had seen\\naround 200 movies, so 99% of the matrix had missing elements. Table 12.2\\nillustrates the setup.\\nIn order to suggest a movie that a particular customer might like, Netflix\\nneeded a way to impute the missing values of this data matrix. The key idea\\nis as follows: the set of movies that the ith customer has seen will overlap\\nwith those that other customers have seen. Furthermore, some of those\\nother customers will have similar movie preferences to the ith customer.\\nThus, it should be possible to use similar customers’ ratings of movies that\\nthe ith customer has not seen to predict whether the ith customer will like\\nthose movies.\\nMore concretely, by applying Algorithm 12.1, we can predict the ith cus-\\ntomer’s rating for the jth movie using ˆxij = )M\\nm=1 ˆaimˆbjm. Furthermore,\\nwe can interpret the M components in terms of “cliques” and “genres”:\\n• ˆaim represents the strength with which the ith user belongs to the\\nmth clique, where a clique is a group of customers that enjoys movies\\nof the mth genre;\\n• ˆbjm represents the strength with which the jth movie belongs to the\\nmth genre.\\nExamples of genres include Romance, Western, and Action.\\nPrincipal component models similar to Algorithm 12.1 are at the heart\\nof many recommender systems. Although the data matrices involved are\\n−3\\n−2\\n−1\\n0\\n1\\n2\\n3\\n−3\\n−2\\n−1\\n0\\n1\\n2\\n3\\nTrue First Principal Component\\nImputed First Principal Component\\n●●\\n●\\n●●●●●●●●●●●●●●●●\\n●\\n●●●●●●●●●●●●●●●●●\\n●●●\\n●●\\n●●●●●●\\n●●\\n2\\n4\\n6\\n8\\n10\\n12\\n2\\n4\\n6\\n8\\n10\\n12\\nTrue PC Variances\\nImputed PC Variances\\n●\\n●\\n●\\n● 520\\n12. Unsupervised Learning\\nJerry Maguire\\nOceans\\nRoad to Perdition\\nA Fortunate Man\\nCatch Me If You Can\\nDriving Miss Daisy\\nThe Two Popes\\nThe Laundromat\\nCode 8\\nThe Social Network\\n· · ·\\nCustomer 1\\n•\\n•\\n•\\n•\\n4\\n•\\n•\\n•\\n•\\n•\\n· · ·\\nCustomer 2\\n•\\n•\\n3\\n•\\n•\\n•\\n3\\n•\\n•\\n3\\n· · ·\\nCustomer 3\\n•\\n2\\n•\\n4\\n•\\n•\\n•\\n•\\n2\\n•\\n· · ·\\nCustomer 4\\n3\\n•\\n•\\n•\\n•\\n•\\n•\\n•\\n•\\n•\\n· · ·\\nCustomer 5\\n5\\n1\\n•\\n•\\n4\\n•\\n•\\n•\\n•\\n•\\n· · ·\\nCustomer 6\\n•\\n•\\n•\\n•\\n•\\n2\\n4\\n•\\n•\\n•\\n· · ·\\nCustomer 7\\n•\\n•\\n5\\n•\\n•\\n•\\n•\\n3\\n•\\n•\\n· · ·\\nCustomer 8\\n•\\n•\\n•\\n•\\n•\\n•\\n•\\n•\\n•\\n•\\n· · ·\\nCustomer 9\\n3\\n•\\n•\\n•\\n5\\n•\\n•\\n1\\n•\\n•\\n· · ·\\n...\\n...\\n...\\n...\\n...\\n...\\n...\\n...\\n...\\n...\\n...\\n...\\nTABLE 12.2. Excerpt of the Netflix movie rating data. The movies are rated\\nfrom 1 (worst) to 5 (best). The symbol • represents a missing value: a movie that\\nwas not rated by the corresponding customer.\\ntypically massive, algorithms have been developed that can exploit the high\\nlevel of missingness in order to perform efficient computations.\\n12.4\\nClustering Methods\\nClustering refers to a very broad set of techniques for finding subgroups, or clustering\\nclusters, in a data set. When we cluster the observations of a data set, we\\nseek to partition them into distinct groups so that the observations within\\neach group are quite similar to each other, while observations in different\\ngroups are quite different from each other. Of course, to make this concrete,\\nwe must define what it means for two or more observations to be similar\\nor different. Indeed, this is often a domain-specific consideration that must\\nbe made based on knowledge of the data being studied.\\nFor instance, suppose that we have a set of n observations, each with p\\nfeatures. The n observations could correspond to tissue samples for patients\\nwith breast cancer, and the p features could correspond to measurements\\ncollected for each tissue sample; these could be clinical measurements, such\\nas tumor stage or grade, or they could be gene expression measurements.\\nWe may have a reason to believe that there is some heterogeneity among\\nthe n tissue samples; for instance, perhaps there are a few different un-\\nknown subtypes of breast cancer. Clustering could be used to find these\\nsubgroups. This is an unsupervised problem because we are trying to dis-\\ncover structure—in this case, distinct clusters—on the basis of a data set.\\nThe goal in supervised problems, on the other hand, is to try to predict\\nsome outcome vector such as survival time or response to drug treatment.\\nBoth clustering and PCA seek to simplify the data via a small number\\nof summaries, but their mechanisms are different: 12.4 Clustering Methods\\n521\\n• PCA looks to find a low-dimensional representation of the observa-\\ntions that explain a good fraction of the variance;\\n• Clustering looks to find homogeneous subgroups among the observa-\\ntions.\\nAnother application of clustering arises in marketing. We may have ac-\\ncess to a large number of measurements (e.g. median household income,\\noccupation, distance from nearest urban area, and so forth) for a large\\nnumber of people. Our goal is to perform market segmentation by identify-\\ning subgroups of people who might be more receptive to a particular form\\nof advertising, or more likely to purchase a particular product. The task of\\nperforming market segmentation amounts to clustering the people in the\\ndata set.\\nSince clustering is popular in many fields, there exist a great num-\\nber of clustering methods. In this section we focus on perhaps the two\\nbest-known clustering approaches: K-means clustering and hierarchical K-means\\nclustering\\nclustering. In K-means clustering, we seek to partition the observations\\nhierarchical\\nclustering\\ninto a pre-specified number of clusters. On the other hand, in hierarchical\\nclustering, we do not know in advance how many clusters we want; in fact,\\nwe end up with a tree-like visual representation of the observations, called\\na dendrogram, that allows us to view at once the clusterings obtained for dendrogram\\neach possible number of clusters, from 1 to n. There are advantages and\\ndisadvantages to each of these clustering approaches, which we highlight in\\nthis chapter.\\nIn general, we can cluster observations on the basis of the features in\\norder to identify subgroups among the observations, or we can cluster fea-\\ntures on the basis of the observations in order to discover subgroups among\\nthe features. In what follows, for simplicity we will discuss clustering obser-\\nvations on the basis of the features, though the converse can be performed\\nby simply transposing the data matrix.\\n12.4.1\\nK-Means Clustering\\nK-means clustering is a simple and elegant approach for partitioning a\\ndata set into K distinct, non-overlapping clusters. To perform K-means\\nclustering, we must first specify the desired number of clusters K; then the\\nK-means algorithm will assign each observation to exactly one of the K\\nclusters. Figure 12.7 shows the results obtained from performing K-means\\nclustering on a simulated example consisting of 150 observations in two\\ndimensions, using three different values of K.\\nThe K-means clustering procedure results from a simple and intuitive\\nmathematical problem. We begin by defining some notation. Let C1, . . . , CK\\ndenote sets containing the indices of the observations in each cluster. These\\nsets satisfy two properties:\\n1. C1 ∪C2 ∪· · · ∪CK = {1, . . . , n}. In other words, each observation\\nbelongs to at least one of the K clusters.\\n2. Ck ∩Ck′ = ∅for all k ̸= k′. In other words, the clusters are non-\\noverlapping: no observation belongs to more than one cluster. 522\\n12. Unsupervised Learning\\nK=2\\nK=3\\nK=4\\nFIGURE 12.7. A simulated data set with 150 observations in two-dimensional\\nspace. Panels show the results of applying K-means clustering with different values\\nof K, the number of clusters. The color of each observation indicates the cluster\\nto which it was assigned using the K-means clustering algorithm. Note that there\\nis no ordering of the clusters, so the cluster coloring is arbitrary. These cluster\\nlabels were not used in clustering; instead, they are the outputs of the clustering\\nprocedure.\\nFor instance, if the ith observation is in the kth cluster, then i ∈Ck. The\\nidea behind K-means clustering is that a good clustering is one for which the\\nwithin-cluster variation is as small as possible. The within-cluster variation\\nfor cluster Ck is a measure W(Ck) of the amount by which the observations\\nwithin a cluster differ from each other. Hence we want to solve the problem\\nminimize\\nC1,...,CK\\n= K\\n0\\nk=1\\nW(Ck)\\nX\\n.\\n(12.15)\\nIn words, this formula says that we want to partition the observations into\\nK clusters such that the total within-cluster variation, summed over all K\\nclusters, is as small as possible.\\nSolving (12.15) seems like a reasonable idea, but in order to make it\\nactionable we need to define the within-cluster variation. There are many\\npossible ways to define this concept, but by far the most common choice\\ninvolves squared Euclidean distance. That is, we define\\nW(Ck) =\\n1\\n|Ck|\\n0\\ni,i′∈Ck\\np\\n0\\nj=1\\n(xij −xi′j)2,\\n(12.16)\\nwhere |Ck| denotes the number of observations in the kth cluster. In other\\nwords, the within-cluster variation for the kth cluster is the sum of all of\\nthe pairwise squared Euclidean distances between the observations in the\\nkth cluster, divided by the total number of observations in the kth cluster.\\nCombining (12.15) and (12.16) gives the optimization problem that defines 12.4 Clustering Methods\\n523\\nK-means clustering,\\nminimize\\nC1,...,CK\\n\\uf8f1\\n\\uf8f2\\n\\uf8f3\\nK\\n0\\nk=1\\n1\\n|Ck|\\n0\\ni,i′∈Ck\\np\\n0\\nj=1\\n(xij −xi′j)2\\n\\uf8fc\\n\\uf8fd\\n\\uf8fe.\\n(12.17)\\nNow, we would like to find an algorithm to solve (12.17)—that is, a\\nmethod to partition the observations into K clusters such that the objective\\nof (12.17) is minimized. This is in fact a very difficult problem to solve\\nprecisely, since there are almost Kn ways to partition n observations into K\\nclusters. This is a huge number unless K and n are tiny! Fortunately, a very\\nsimple algorithm can be shown to provide a local optimum—a pretty good\\nsolution—to the K-means optimization problem (12.17). This approach is\\nlaid out in Algorithm 12.2.\\nAlgorithm 12.2 K-Means Clustering\\n1. Randomly assign a number, from 1 to K, to each of the observations.\\nThese serve as initial cluster assignments for the observations.\\n2. Iterate until the cluster assignments stop changing:\\n(a) For each of the K clusters, compute the cluster centroid. The\\nkth cluster centroid is the vector of the p feature means for the\\nobservations in the kth cluster.\\n(b) Assign each observation to the cluster whose centroid is closest\\n(where closest is defined using Euclidean distance).\\nAlgorithm 12.2 is guaranteed to decrease the value of the objective (12.17)\\nat each step. To understand why, the following identity is illuminating:\\n1\\n|Ck|\\n0\\ni,i′∈Ck\\np\\n0\\nj=1\\n(xij −xi′j)2 = 2\\n0\\ni∈Ck\\np\\n0\\nj=1\\n(xij −¯xkj)2,\\n(12.18)\\nwhere ¯xkj =\\n1\\n|Ck|\\n)\\ni∈Ck xij is the mean for feature j in cluster Ck.\\nIn Step 2(a) the cluster means for each feature are the constants that\\nminimize the sum-of-squared deviations, and in Step 2(b), reallocating the\\nobservations can only improve (12.18). This means that as the algorithm\\nis run, the clustering obtained will continually improve until the result no\\nlonger changes; the objective of (12.17) will never increase. When the result\\nno longer changes, a local optimum has been reached. Figure 12.8 shows\\nthe progression of the algorithm on the toy example from Figure 12.7.\\nK-means clustering derives its name from the fact that in Step 2(a), the\\ncluster centroids are computed as the mean of the observations assigned to\\neach cluster.\\nBecause the K-means algorithm finds a local rather than a global opti-\\nmum, the results obtained will depend on the initial (random) cluster as-\\nsignment of each observation in Step 1 of Algorithm 12.2. For this reason,\\nit is important to run the algorithm multiple times from different random 524\\n12. Unsupervised Learning\\nData\\nStep 1\\nIteration 1, Step 2a\\nIteration 1, Step 2b\\nIteration 2, Step 2a\\nFinal Results\\nFIGURE 12.8.\\nThe progress of the K-means algorithm on the example of\\nFigure 12.7 with K=3. Top left: the observations are shown. Top center: in\\nStep 1 of the algorithm, each observation is randomly assigned to a cluster. Top\\nright: in Step 2(a), the cluster centroids are computed. These are shown as large\\ncolored disks. Initially the centroids are almost completely overlapping because\\nthe initial cluster assignments were chosen at random. Bottom left: in Step 2(b),\\neach observation is assigned to the nearest centroid. Bottom center: Step 2(a) is\\nonce again performed, leading to new cluster centroids. Bottom right: the results\\nobtained after ten iterations.\\ninitial configurations. Then one selects the best solution, i.e. that for which\\nthe objective (12.17) is smallest. Figure 12.9 shows the local optima ob-\\ntained by running K-means clustering six times using six different initial\\ncluster assignments, using the toy data from Figure 12.7. In this case, the\\nbest clustering is the one with an objective value of 235.8.\\nAs we have seen, to perform K-means clustering, we must decide how\\nmany clusters we expect in the data. The problem of selecting K is far from\\nsimple. This issue, along with other practical considerations that arise in\\nperforming K-means clustering, is addressed in Section 12.4.3. 12.4 Clustering Methods\\n525\\n320.9\\n235.8\\n235.8\\n235.8\\n235.8\\n310.9\\nFIGURE 12.9.\\nK-means clustering performed six times on the data from\\nFigure 12.7 with K = 3, each time with a different random assignment of the\\nobservations in Step 1 of the K-means algorithm. Above each plot is the value\\nof the objective (12.17). Three different local optima were obtained, one of which\\nresulted in a smaller value of the objective and provides better separation between\\nthe clusters. Those labeled in red all achieved the same best solution, with an\\nobjective value of 235.8.\\n12.4.2\\nHierarchical Clustering\\nOne potential disadvantage of K-means clustering is that it requires us to\\npre-specify the number of clusters K. Hierarchical clustering is an alter-\\nnative approach which does not require that we commit to a particular\\nchoice of K. Hierarchical clustering has an added advantage over K-means\\nclustering in that it results in an attractive tree-based representation of the\\nobservations, called a dendrogram.\\nIn this section, we describe bottom-up or agglomerative\\nclustering.\\nbottom-up\\nagglomerative\\nThis is the most common type of hierarchical clustering, and refers to\\nthe fact that a dendrogram (generally depicted as an upside-down tree; see\\nFigure 12.11) is built starting from the leaves and combining clusters up to\\nthe trunk. We will begin with a discussion of how to interpret a dendrogram 526\\n12. Unsupervised Learning\\n−6\\n−4\\n−2\\n0\\n2\\n−2\\n0\\n2\\n4\\nX1\\nX2\\nFIGURE 12.10.\\nForty-five observations generated in two-dimensional space.\\nIn reality there are three distinct classes, shown in separate colors. However, we\\nwill treat these class labels as unknown and will seek to cluster the observations\\nin order to discover the classes from the data.\\nand then discuss how hierarchical clustering is actually performed—that is,\\nhow the dendrogram is built.\\nInterpreting a Dendrogram\\nWe begin with the simulated data set shown in Figure 12.10, consisting of\\n45 observations in two-dimensional space. The data were generated from a\\nthree-class model; the true class labels for each observation are shown in\\ndistinct colors. However, suppose that the data were observed without the\\nclass labels, and that we wanted to perform hierarchical clustering of the\\ndata. Hierarchical clustering (with complete linkage, to be discussed later)\\nyields the result shown in the left-hand panel of Figure 12.11. How can we\\ninterpret this dendrogram?\\nIn the left-hand panel of Figure 12.11, each leaf of the dendrogram rep-\\nresents one of the 45 observations in Figure 12.10. However, as we move\\nup the tree, some leaves begin to fuse into branches. These correspond to\\nobservations that are similar to each other. As we move higher up the tree,\\nbranches themselves fuse, either with leaves or other branches. The earlier\\n(lower in the tree) fusions occur, the more similar the groups of observa-\\ntions are to each other. On the other hand, observations that fuse later\\n(near the top of the tree) can be quite different. In fact, this statement\\ncan be made precise: for any two observations, we can look for the point in\\nthe tree where branches containing those two observations are first fused.\\nThe height of this fusion, as measured on the vertical axis, indicates how\\ndifferent the two observations are. Thus, observations that fuse at the very\\nbottom of the tree are quite similar to each other, whereas observations\\nthat fuse close to the top of the tree will tend to be quite different.\\nThis highlights a very important point in interpreting dendrograms that\\nis often misunderstood. Consider the left-hand panel of Figure 12.12, which\\nshows a simple dendrogram obtained from hierarchically clustering nine 12.4 Clustering Methods\\n527\\n0\\n2\\n4\\n6\\n8\\n10\\n0\\n2\\n4\\n6\\n8\\n10\\n0\\n2\\n4\\n6\\n8\\n10\\nFIGURE 12.11. Left: dendrogram obtained from hierarchically clustering the\\ndata from Figure 12.10 with complete linkage and Euclidean distance. Center: the\\ndendrogram from the left-hand panel, cut at a height of nine (indicated by the\\ndashed line). This cut results in two distinct clusters, shown in different colors.\\nRight: the dendrogram from the left-hand panel, now cut at a height of five. This\\ncut results in three distinct clusters, shown in different colors. Note that the colors\\nwere not used in clustering, but are simply used for display purposes in this figure.\\nobservations. One can see that observations 5 and 7 are quite similar to\\neach other, since they fuse at the lowest point on the dendrogram. Obser-\\nvations 1 and 6 are also quite similar to each other. However, it is tempting\\nbut incorrect to conclude from the figure that observations 9 and 2 are\\nquite similar to each other on the basis that they are located near each\\nother on the dendrogram. In fact, based on the information contained in\\nthe dendrogram, observation 9 is no more similar to observation 2 than it\\nis to observations 8, 5, and 7. (This can be seen from the right-hand panel\\nof Figure 12.12, in which the raw data are displayed.) To put it mathe-\\nmatically, there are 2n−1 possible reorderings of the dendrogram, where n\\nis the number of leaves. This is because at each of the n −1 points where\\nfusions occur, the positions of the two fused branches could be swapped\\nwithout affecting the meaning of the dendrogram. Therefore, we cannot\\ndraw conclusions about the similarity of two observations based on their\\nproximity along the horizontal axis. Rather, we draw conclusions about\\nthe similarity of two observations based on the location on the vertical axis\\nwhere branches containing those two observations first are fused.\\nNow that we understand how to interpret the left-hand panel of Fig-\\nure 12.11, we can move on to the issue of identifying clusters on the basis\\nof a dendrogram. In order to do this, we make a horizontal cut across the\\ndendrogram, as shown in the center and right-hand panels of Figure 12.11.\\nThe distinct sets of observations beneath the cut can be interpreted as clus-\\nters. In the center panel of Figure 12.11, cutting the dendrogram at a height\\nof nine results in two clusters, shown in distinct colors. In the right-hand\\npanel, cutting the dendrogram at a height of five results in three clusters.\\nFurther cuts can be made as one descends the dendrogram in order to ob-\\ntain any number of clusters, between 1 (corresponding to no cut) and n 528\\n12. Unsupervised Learning\\n3\\n4\\n1\\n6\\n9\\n2\\n8\\n5\\n7\\n0.0\\n0.5\\n1.0\\n1.5\\n2.0\\n2.5\\n3.0\\n1\\n2\\n3\\n4\\n5\\n6\\n7\\n8\\n9\\n−1.5\\n−1.0\\n−0.5\\n0.0\\n0.5\\n1.0\\n−1.5\\n−1.0\\n−0.5\\n0.0\\n0.5\\nX1\\nX2\\nFIGURE 12.12.\\nAn illustration of how to properly interpret a dendrogram\\nwith nine observations in two-dimensional space. Left: a dendrogram generated\\nusing Euclidean distance and complete linkage. Observations 5 and 7 are quite\\nsimilar to each other, as are observations 1 and 6. However, observation 9 is no\\nmore similar to observation 2 than it is to observations 8, 5, and 7, even though\\nobservations 9 and 2 are close together in terms of horizontal distance. This is\\nbecause observations 2, 8, 5, and 7 all fuse with observation 9 at the same height,\\napproximately 1.8. Right: the raw data used to generate the dendrogram can be\\nused to confirm that indeed, observation 9 is no more similar to observation 2\\nthan it is to observations 8, 5, and 7.\\n(corresponding to a cut at height 0, so that each observation is in its own\\ncluster). In other words, the height of the cut to the dendrogram serves\\nthe same role as the K in K-means clustering: it controls the number of\\nclusters obtained.\\nFigure 12.11 therefore highlights a very attractive aspect of hierarchical\\nclustering: one single dendrogram can be used to obtain any number of\\nclusters. In practice, people often look at the dendrogram and select by eye\\na sensible number of clusters, based on the heights of the fusion and the\\nnumber of clusters desired. In the case of Figure 12.11, one might choose\\nto select either two or three clusters. However, often the choice of where to\\ncut the dendrogram is not so clear.\\nThe term hierarchical refers to the fact that clusters obtained by cutting\\nthe dendrogram at a given height are necessarily nested within the clusters\\nobtained by cutting the dendrogram at any greater height. However, on\\nan arbitrary data set, this assumption of hierarchical structure might be\\nunrealistic. For instance, suppose that our observations correspond to a\\ngroup of men and women, evenly split among Americans, Japanese, and\\nFrench. We can imagine a scenario in which the best division into two\\ngroups might split these people by gender, and the best division into three\\ngroups might split them by nationality. In this case, the true clusters are\\nnot nested, in the sense that the best division into three groups does not\\nresult from taking the best division into two groups and splitting up one\\nof those groups. Consequently, this situation could not be well-represented\\nby hierarchical clustering. Due to situations such as this one, hierarchical\\nclustering can sometimes yield worse (i.e. less accurate) results than K-\\nmeans clustering for a given number of clusters. 12.4 Clustering Methods\\n529\\nAlgorithm 12.3 Hierarchical Clustering\\n1. Begin with n observations and a measure (such as Euclidean dis-\\ntance) of all the\\n\\'n\\n2\\n(\\n= n(n −1)/2 pairwise dissimilarities. Treat each\\nobservation as its own cluster.\\n2. For i = n, n −1, . . . , 2:\\n(a) Examine all pairwise inter-cluster dissimilarities among the i\\nclusters and identify the pair of clusters that are least dissimilar\\n(that is, most similar). Fuse these two clusters. The dissimilarity\\nbetween these two clusters indicates the height in the dendro-\\ngram at which the fusion should be placed.\\n(b) Compute the new pairwise inter-cluster dissimilarities among\\nthe i −1 remaining clusters.\\nThe Hierarchical Clustering Algorithm\\nThe hierarchical clustering dendrogram is obtained via an extremely simple\\nalgorithm. We begin by defining some sort of dissimilarity measure between\\neach pair of observations. Most often, Euclidean distance is used; we will\\ndiscuss the choice of dissimilarity measure later in this chapter. The algo-\\nrithm proceeds iteratively. Starting out at the bottom of the dendrogram,\\neach of the n observations is treated as its own cluster. The two clusters\\nthat are most similar to each other are then fused so that there now are\\nn−1 clusters. Next the two clusters that are most similar to each other are\\nfused again, so that there now are n −2 clusters. The algorithm proceeds\\nin this fashion until all of the observations belong to one single cluster, and\\nthe dendrogram is complete. Figure 12.13 depicts the first few steps of the\\nalgorithm, for the data from Figure 12.12. To summarize, the hierarchical\\nclustering algorithm is given in Algorithm 12.3.\\nThis algorithm seems simple enough, but one issue has not been ad-\\ndressed. Consider the bottom right panel in Figure 12.13. How did we\\ndetermine that the cluster {5, 7} should be fused with the cluster {8}?\\nWe have a concept of the dissimilarity between pairs of observations, but\\nhow do we define the dissimilarity between two clusters if one or both of\\nthe clusters contains multiple observations? The concept of dissimilarity\\nbetween a pair of observations needs to be extended to a pair of groups\\nof observations. This extension is achieved by developing the notion of\\nlinkage, which defines the dissimilarity between two groups of observa- linkage\\ntions. The four most common types of linkage—complete, average, single,\\nand centroid—are briefly described in Table 12.3. Average, complete, and\\nsingle linkage are most popular among statisticians. Average and complete\\nlinkage are generally preferred over single linkage, as they tend to yield\\nmore balanced dendrograms. Centroid linkage is often used in genomics,\\nbut suffers from a major drawback in that an inversion can occur, whereby inversion\\ntwo clusters are fused at a height below either of the individual clusters in\\nthe dendrogram. This can lead to difficulties in visualization as well as in in-\\nterpretation of the dendrogram. The dissimilarities computed in Step 2(b) 530\\n12. Unsupervised Learning\\nLinkage\\nDescription\\nComplete\\nMaximal intercluster dissimilarity. Compute all pairwise\\ndissimilarities between the observations in cluster A and the\\nobservations in cluster B, and record the largest of these dis-\\nsimilarities.\\nSingle\\nMinimal intercluster dissimilarity. Compute all pairwise dis-\\nsimilarities between the observations in cluster A and the\\nobservations in cluster B, and record the smallest of these\\ndissimilarities. Single linkage can result in extended, trailing\\nclusters in which single observations are fused one-at-a-time.\\nAverage\\nMean intercluster dissimilarity. Compute all pairwise dis-\\nsimilarities between the observations in cluster A and the\\nobservations in cluster B, and record the average of these\\ndissimilarities.\\nCentroid\\nDissimilarity between the centroid for cluster A (a mean\\nvector of length p) and the centroid for cluster B. Centroid\\nlinkage can result in undesirable inversions.\\nTABLE 12.3. A summary of the four most commonly-used types of linkage in\\nhierarchical clustering.\\nof the hierarchical clustering algorithm will depend on the type of linkage\\nused, as well as on the choice of dissimilarity measure. Hence, the resulting\\ndendrogram typically depends quite strongly on the type of linkage used,\\nas is shown in Figure 12.14.\\nChoice of Dissimilarity Measure\\nThus far, the examples in this chapter have used Euclidean distance as the\\ndissimilarity measure. But sometimes other dissimilarity measures might\\nbe preferred. For example, correlation-based distance considers two obser-\\nvations to be similar if their features are highly correlated, even though the\\nobserved values may be far apart in terms of Euclidean distance. This is\\nan unusual use of correlation, which is normally computed between vari-\\nables; here it is computed between the observation profiles for each pair\\nof observations. Figure 12.15 illustrates the difference between Euclidean\\nand correlation-based distance. Correlation-based distance focuses on the\\nshapes of observation profiles rather than their magnitudes.\\nThe choice of dissimilarity measure is very important, as it has a strong\\neffect on the resulting dendrogram. In general, careful attention should be\\npaid to the type of data being clustered and the scientific question at hand.\\nThese considerations should determine what type of dissimilarity measure\\nis used for hierarchical clustering.\\nFor instance, consider an online retailer interested in clustering shoppers\\nbased on their past shopping histories. The goal is to identify subgroups\\nof similar shoppers, so that shoppers within each subgroup can be shown\\nitems and advertisements that are particularly likely to interest them. Sup-\\npose the data takes the form of a matrix where the rows are the shoppers\\nand the columns are the items available for purchase; the elements of the\\ndata matrix indicate the number of times a given shopper has purchased a 12.4 Clustering Methods\\n531\\n1\\n2\\n3\\n4\\n5\\n6\\n7\\n8\\n9\\n−1.5\\n−1.0\\n−0.5\\n0.0\\n0.5\\n1.0\\n−1.5\\n−1.0\\n−0.5\\n0.0\\n0.5\\n1\\n2\\n3\\n4\\n5\\n6\\n7\\n8\\n9\\n−1.5\\n−1.0\\n−0.5\\n0.0\\n0.5\\n1.0\\n−1.5\\n−1.0\\n−0.5\\n0.0\\n0.5\\n1\\n2\\n3\\n4\\n5\\n6\\n7\\n8\\n9\\n−1.5\\n−1.0\\n−0.5\\n0.0\\n0.5\\n1.0\\n−1.5\\n−1.0\\n−0.5\\n0.0\\n0.5\\n1\\n2\\n3\\n4\\n5\\n6\\n7\\n8\\n9\\n−1.5\\n−1.0\\n−0.5\\n0.0\\n0.5\\n1.0\\n−1.5\\n−1.0\\n−0.5\\n0.0\\n0.5\\nX1\\nX1\\nX1\\nX1\\nX2\\nX2\\nX2\\nX2\\nFIGURE 12.13.\\nAn illustration of the first few steps of the hierarchical\\nclustering algorithm, using the data from Figure 12.12, with complete linkage\\nand Euclidean distance. Top Left: initially, there are nine distinct clusters,\\n{1}, {2}, . . . , {9}. Top Right: the two clusters that are closest together, {5} and\\n{7}, are fused into a single cluster. Bottom Left: the two clusters that are closest\\ntogether, {6} and {1}, are fused into a single cluster. Bottom Right: the two clus-\\nters that are closest together using complete linkage, {8} and the cluster {5, 7},\\nare fused into a single cluster.\\ngiven item (i.e. a 0 if the shopper has never purchased this item, a 1 if the\\nshopper has purchased it once, etc.) What type of dissimilarity measure\\nshould be used to cluster the shoppers? If Euclidean distance is used, then\\nshoppers who have bought very few items overall (i.e. infrequent users of\\nthe online shopping site) will be clustered together. This may not be desir-\\nable. On the other hand, if correlation-based distance is used, then shoppers\\nwith similar preferences (e.g. shoppers who have bought items A and B but\\nnever items C or D) will be clustered together, even if some shoppers with\\nthese preferences are higher-volume shoppers than others. Therefore, for\\nthis application, correlation-based distance may be a better choice.\\nIn addition to carefully selecting the dissimilarity measure used, one must\\nalso consider whether or not the variables should be scaled to have stan-\\ndard deviation one before the dissimilarity between the observations is com-\\nputed. To illustrate this point, we continue with the online shopping ex- 532\\n12. Unsupervised Learning\\nAverage Linkage\\nComplete Linkage\\nSingle Linkage\\nFIGURE 12.14. Average, complete, and single linkage applied to an example\\ndata set. Average and complete linkage tend to yield more balanced clusters.\\nample just described. Some items may be purchased more frequently than\\nothers; for instance, a shopper might buy ten pairs of socks a year, but a\\ncomputer very rarely. High-frequency purchases like socks therefore tend\\nto have a much larger effect on the inter-shopper dissimilarities, and hence\\non the clustering ultimately obtained, than rare purchases like computers.\\nThis may not be desirable. If the variables are scaled to have standard de-\\nviation one before the inter-observation dissimilarities are computed, then\\neach variable will in effect be given equal importance in the hierarchical\\nclustering performed. We might also want to scale the variables to have\\nstandard deviation one if they are measured on different scales; otherwise,\\nthe choice of units (e.g. centimeters versus kilometers) for a particular vari-\\nable will greatly affect the dissimilarity measure obtained. It should come\\nas no surprise that whether or not it is a good decision to scale the variables\\nbefore computing the dissimilarity measure depends on the application at\\nhand. An example is shown in Figure 12.16. We note that the issue of\\nwhether or not to scale the variables before performing clustering applies\\nto K-means clustering as well.\\n12.4.3\\nPractical Issues in Clustering\\nClustering can be a very useful tool for data analysis in the unsupervised\\nsetting. However, there are a number of issues that arise in performing\\nclustering. We describe some of these issues here.\\nSmall Decisions with Big Consequences\\nIn order to perform clustering, some decisions must be made. 12.4 Clustering Methods\\n533\\n5\\n10\\n15\\n20\\n0\\n5\\n10\\n15\\n20\\nVariable Index\\nObservation 1\\nObservation 2\\nObservation 3\\n1\\n2\\n3\\nFIGURE 12.15.\\nThree observations with measurements on 20 variables are\\nshown. Observations 1 and 3 have similar values for each variable and so there\\nis a small Euclidean distance between them. But they are very weakly correlated,\\nso they have a large correlation-based distance. On the other hand, observations\\n1 and 2 have quite different values for each variable, and so there is a large\\nEuclidean distance between them. But they are highly correlated, so there is a\\nsmall correlation-based distance between them.\\n• Should the observations or features first be standardized in some way?\\nFor instance, maybe the variables should be scaled to have standard\\ndeviation one.\\n• In the case of hierarchical clustering,\\n– What dissimilarity measure should be used?\\n– What type of linkage should be used?\\n– Where should we cut the dendrogram in order to obtain clusters?\\n• In the case of K-means clustering, how many clusters should we look\\nfor in the data?\\nEach of these decisions can have a strong impact on the results obtained.\\nIn practice, we try several different choices, and look for the one with\\nthe most useful or interpretable solution. With these methods, there is no\\nsingle right answer—any solution that exposes some interesting aspects of\\nthe data should be considered.\\nValidating the Clusters Obtained\\nAny time clustering is performed on a data set we will find clusters. But we\\nreally want to know whether the clusters that have been found represent\\ntrue subgroups in the data, or whether they are simply a result of clustering\\nthe noise. For instance, if we were to obtain an independent set of observa-\\ntions, then would those observations also display the same set of clusters?\\nThis is a hard question to answer. There exist a number of techniques for\\nassigning a p-value to a cluster in order to assess whether there is more 534\\n12. Unsupervised Learning\\nSocks\\nComputers\\n0\\n2\\n4\\n6\\n8\\n10\\nSocks\\nComputers\\n0.0\\n0.2\\n0.4\\n0.6\\n0.8\\n1.0\\n1.2\\nSocks\\nComputers\\n0\\n500\\n1000\\n1500\\nFIGURE 12.16. An eclectic online retailer sells two items: socks and computers.\\nLeft: the number of pairs of socks, and computers, purchased by eight online shop-\\npers is displayed. Each shopper is shown in a different color. If inter-observation\\ndissimilarities are computed using Euclidean distance on the raw variables, then\\nthe number of socks purchased by an individual will drive the dissimilarities ob-\\ntained, and the number of computers purchased will have little effect. This might\\nbe undesirable, since (1) computers are more expensive than socks and so the\\nonline retailer may be more interested in encouraging shoppers to buy computers\\nthan socks, and (2) a large difference in the number of socks purchased by two\\nshoppers may be less informative about the shoppers’ overall shopping preferences\\nthan a small difference in the number of computers purchased. Center: the same\\ndata are shown, after scaling each variable by its standard deviation. Now the\\ntwo products will have a comparable effect on the inter-observation dissimilarities\\nobtained. Right: the same data are displayed, but now the y-axis represents the\\nnumber of dollars spent by each online shopper on socks and on computers. Since\\ncomputers are much more expensive than socks, now computer purchase history\\nwill drive the inter-observation dissimilarities obtained.\\nevidence for the cluster than one would expect due to chance. However,\\nthere has been no consensus on a single best approach. More details can\\nbe found in ESL.8\\nOther Considerations in Clustering\\nBoth K-means and hierarchical clustering will assign each observation to\\na cluster. However, sometimes this might not be appropriate. For instance,\\nsuppose that most of the observations truly belong to a small number of\\n(unknown) subgroups, and a small subset of the observations are quite\\ndifferent from each other and from all other observations. Then since K-\\nmeans and hierarchical clustering force every observation into a cluster, the\\nclusters found may be heavily distorted due to the presence of outliers that\\ndo not belong to any cluster. Mixture models are an attractive approach\\nfor accommodating the presence of such outliers. These amount to a soft\\nversion of K-means clustering, and are described in ESL.\\n8ESL: The Elements of Statistical Learning by Hastie, Tibshirani and Friedman. 12.5 Lab: Unsupervised Learning\\n535\\nIn addition, clustering methods generally are not very robust to pertur-\\nbations to the data. For instance, suppose that we cluster n observations,\\nand then cluster the observations again after removing a subset of the n\\nobservations at random. One would hope that the two sets of clusters ob-\\ntained would be quite similar, but often this is not the case!\\nA Tempered Approach to Interpreting the Results of Clustering\\nWe have described some of the issues associated with clustering. However,\\nclustering can be a very useful and valid statistical tool if used properly. We\\nmentioned that small decisions in how clustering is performed, such as how\\nthe data are standardized and what type of linkage is used, can have a large\\neffect on the results. Therefore, we recommend performing clustering with\\ndifferent choices of these parameters, and looking at the full set of results\\nin order to see what patterns consistently emerge. Since clustering can be\\nnon-robust, we recommend clustering subsets of the data in order to get a\\nsense of the robustness of the clusters obtained. Most importantly, we must\\nbe careful about how the results of a clustering analysis are reported. These\\nresults should not be taken as the absolute truth about a data set. Rather,\\nthey should constitute a starting point for the development of a scientific\\nhypothesis and further study, preferably on an independent data set.\\n12.5\\nLab: Unsupervised Learning\\nIn this lab we demonstrate PCA and clustering on several datasets. As in\\nother labs, we import some of our libraries at this top level. This makes\\nthe code more readable, as scanning the first few lines of the notebook tell\\nus what libraries are used in this notebook.\\nIn [1]: import numpy as np\\nimport pandas as pd\\nimport matplotlib.pyplot as plt\\nfrom statsmodels.datasets import get_rdataset\\nfrom sklearn.decomposition import PCA\\nfrom sklearn.preprocessing import StandardScaler\\nfrom ISLP import load_data\\nWe also collect the new imports needed for this lab.\\nIn [2]: from sklearn.cluster import \\\\\\n(KMeans,\\nAgglomerativeClustering)\\nfrom scipy.cluster.hierarchy import \\\\\\n(dendrogram ,\\ncut_tree)\\nfrom ISLP.cluster import compute_linkage\\n12.5.1\\nPrincipal Components Analysis\\nIn this lab, we perform PCA on USArrests, a data set in the R computing\\nenvironment. We retrieve the data using get_rdataset(), which can fetch get_\\nrdataset() 536\\n12. Unsupervised Learning\\ndata from many standard R packages.\\nThe rows of the data set contain the 50 states, in alphabetical order.\\nIn [3]: USArrests = get_rdataset(\\'USArrests\\').data\\nUSArrests\\nOut[3]:\\nMurder\\nAssault\\nUrbanPop\\nRape\\nAlabama\\n13.2\\n236\\n58\\n21.2\\nAlaska\\n10.0\\n263\\n48\\n44.5\\nArizona\\n8.1\\n294\\n80\\n31.0\\n...\\n...\\n...\\n...\\n...\\nWisconsin\\n2.6\\n53\\n66\\n10.8\\nWyoming\\n6.8\\n161\\n60\\n15.6\\nThe columns of the data set contain the four variables.\\nIn [4]: USArrests.columns\\nOut[4]: Index([\\'Murder\\', \\'Assault\\', \\'UrbanPop\\', \\'Rape\\'],\\ndtype=\\'object\\')\\nWe first briefly examine the data. We notice that the variables have\\nvastly different means.\\nIn [5]: USArrests.mean()\\nOut[5]: Murder\\n7.788\\nAssault\\n170.760\\nUrbanPop\\n65.540\\nRape\\n21.232\\ndtype: float64\\nDataframes have several useful methods for computing column-wise sum-\\nmaries. We can also examine the variance of the four variables using the\\nvar() method.\\nIn [6]: USArrests.var()\\nOut[6]: Murder\\n18.970465\\nAssault\\n6945.165714\\nUrbanPop\\n209.518776\\nRape\\n87.729159\\ndtype: float64\\nNot surprisingly, the variables also have vastly different variances. The\\nUrbanPop variable measures the percentage of the population in each state\\nliving in an urban area, which is not a comparable number to the number of\\nrapes in each state per 100,000 individuals. PCA looks for derived variables\\nthat account for most of the variance in the data set. If we do not scale the\\nvariables before performing PCA, then the principal components would\\nmostly be driven by the Assault variable, since it has by far the largest\\nvariance. So if the variables are measured in different units or vary widely\\nin scale, it is recommended to standardize the variables to have standard\\ndeviation one before performing PCA. Typically we set the means to zero\\nas well. 12.5 Lab: Unsupervised Learning\\n537\\nThis scaling can be done via the StandardScaler() transform imported\\nabove. We first fit the scaler, which computes the necessary means and\\nstandard deviations and then apply it to our data using the transform\\nmethod. As before, we combine these steps using the fit_transform()\\nmethod.\\nIn [7]: scaler = StandardScaler(with_std=True,\\nwith_mean=True)\\nUSArrests_scaled = scaler.fit_transform(USArrests)\\nHaving scaled the data, we can then perform principal components analysis\\nusing the PCA() transform from the sklearn.decomposition package.\\nPCA()\\nIn [8]: pcaUS = PCA()\\n(By default, the PCA() transform centers the variables to have mean zero\\nthough it does not scale them.) The transform pcaUS can be used to find\\nthe PCA scores returned by fit(). Once the fit method has been called,\\nthe pcaUS object also contains a number of useful quantities.\\nIn [9]: pcaUS.fit(USArrests_scaled)\\nAfter fitting, the mean_ attribute corresponds to the means of the vari-\\nables. In this case, since we centered and scaled the data with scaler() the\\nmeans will all be 0.\\nIn [10]: pcaUS.mean_\\nOut[10]: array([-0.,\\n0., -0.,\\n0.])\\nThe scores can be computed using the transform() method of pcaUS after\\nit has been fit.\\nIn [11]: scores = pcaUS.transform(USArrests_scaled)\\nWe will plot these scores a bit further down. The components_ attribute\\nprovides the principal component loadings: each row of pcaUS.components_\\ncontains the corresponding principal component loading vector.\\nIn [12]: pcaUS.components_\\nOut[12]: array([[ 0.53589947,\\n0.58318363,\\n0.27819087,\\n0.54343209],\\n[ 0.41818087,\\n0.1879856 , -0.87280619, -0.16731864],\\n[-0.34123273, -0.26814843, -0.37801579,\\n0.81777791],\\n[ 0.6492278 , -0.74340748,\\n0.13387773,\\n0.08902432]])\\nThe biplot is a common visualization method used with PCA. It is not\\nbuilt in as a standard part of sklearn, though there are python packages\\nthat do produce such plots. Here we make a simple biplot manually.\\nIn [13]: i, j = 0, 1 # which components\\nfig, ax = plt.subplots(1, 1, figsize=(8, 8))\\nax.scatter(scores[:,0], scores[:,1])\\nax.set_xlabel(\\'PC%d\\' % (i+1))\\nax.set_ylabel(\\'PC%d\\' % (j+1))\\nfor k in range(pcaUS.components_.shape[1]): 538\\n12. Unsupervised Learning\\nax.arrow(0, 0, pcaUS.components_[i,k], pcaUS.components_[j,k])\\nax.text(pcaUS.components_[i,k],\\npcaUS.components_[j,k],\\nUSArrests.columns[k])\\nNotice that this figure is a reflection of Figure 12.1 through the y-axis.\\nRecall that the principal components are only unique up to a sign change,\\nso we can reproduce that figure by flipping the signs of the second set of\\nscores and loadings. We also increase the length of the arrows to emphasize\\nthe loadings.\\nIn [14]: scale_arrow = s_ = 2\\nscores[:,1] *= -1\\npcaUS.components_[1] *= -1 # flip the y-axis\\nfig, ax = plt.subplots(1, 1, figsize=(8, 8))\\nax.scatter(scores[:,0], scores[:,1])\\nax.set_xlabel(\\'PC%d\\' % (i+1))\\nax.set_ylabel(\\'PC%d\\' % (j+1))\\nfor k in range(pcaUS.components_.shape[1]):\\nax.arrow(0, 0, s_*pcaUS.components_[i,k], s_*pcaUS.components_[\\nj,k])\\nax.text(s_*pcaUS.components_[i,k],\\ns_*pcaUS.components_[j,k],\\nUSArrests.columns[k])\\nThe standard deviations of the principal component scores are as follows:\\nIn [15]: scores.std(0, ddof=1)\\nOut[15]: array([1.5909, 1.0050, 0.6032, 0.4207])\\nThe variance of each score can be extracted directly from the pcaUS object\\nvia the explained_variance_ attribute.\\nIn [16]: pcaUS.explained_variance_\\nOut[16]: array([2.5309, 1.01\\n, 0.3638, 0.177 ])\\nThe proportion of variance explained by each principal component (PVE)\\nis stored as explained_variance_ratio_:\\nIn [17]: pcaUS.explained_variance_ratio_\\nOut[17]: array([0.6201, 0.2474, 0.0891, 0.0434])\\nWe see that the first principal component explains 62.0% of the variance\\nin the data, the next principal component explains 24.7% of the variance,\\nand so forth. We can plot the PVE explained by each component, as well\\nas the cumulative PVE. We first plot the proportion of variance explained.\\nIn [18]: %%capture\\nfig, axes = plt.subplots(1, 2, figsize=(15, 6))\\nticks = np.arange(pcaUS.n_components_)+1\\nax = axes[0]\\nax.plot(ticks,\\npcaUS.explained_variance_ratio_ ,\\nmarker=\\'o\\') 12.5 Lab: Unsupervised Learning\\n539\\nax.set_xlabel(\\'Principal Component\\');\\nax.set_ylabel(\\'Proportion of Variance Explained\\')\\nax.set_ylim([0,1])\\nax.set_xticks(ticks)\\nNotice the use of %%capture, which suppresses the displaying of the partially\\ncompleted figure.\\nIn [19]: ax = axes[1]\\nax.plot(ticks,\\npcaUS.explained_variance_ratio_.cumsum(),\\nmarker=\\'o\\')\\nax.set_xlabel(\\'Principal Component\\')\\nax.set_ylabel(\\'Cumulative Proportion of Variance Explained\\')\\nax.set_ylim([0, 1])\\nax.set_xticks(ticks)\\nfig\\nThe result is similar to that shown in Figure 12.3. Note that the method\\ncumsum() computes the cumulative sum of the elements of a numeric vector. cumsum()\\nFor instance:\\nIn [20]: a = np.array([1,2,8,-3])\\nnp.cumsum(a)\\nOut[20]: array([ 1,\\n3, 11,\\n8])\\n12.5.2\\nMatrix Completion\\nWe now re-create the analysis carried out on the USArrests data in Sec-\\ntion 12.3.\\nWe saw in Section 12.2.2 that solving the optimization problem (12.6) on\\na centered data matrix X is equivalent to computing the first M principal\\ncomponents of the data. We use our scaled and centered USArrests data as\\nX below. The singular value decomposition (SVD) is a general algorithm singular\\nvalue de-\\ncomposition\\nsvd()\\nfor solving (12.6).\\nIn [21]: X = USArrests_scaled\\nU, D, V = np.linalg.svd(X, full_matrices=False)\\nU.shape, D.shape, V.shape\\nOut[21]: ((50, 4), (4,), (4, 4))\\nThe np.linalg.svd() function returns three components, U, D and V. The np.linalg.\\nsvd()\\nmatrix V is equivalent to the loading matrix from principal components (up\\nto an unimportant sign flip). Using the full_matrices=False option ensures\\nthat for a tall matrix the shape of U is the same as the shape of X.\\nIn [22]: V\\nOut[22]: array([[-0.53589947, -0.58318363, -0.27819087, -0.54343209],\\n[ 0.41818087,\\n0.1879856 , -0.87280619, -0.16731864],\\n[-0.34123273, -0.26814843, -0.37801579,\\n0.81777791],\\n[ 0.6492278 , -0.74340748,\\n0.13387773,\\n0.08902432]]) 540\\n12. Unsupervised Learning\\nIn [23]: pcaUS.components_\\nOut[23]: array([[ 0.53589947,\\n0.58318363,\\n0.27819087,\\n0.54343209],\\n[ 0.41818087,\\n0.1879856 , -0.87280619, -0.16731864],\\n[-0.34123273, -0.26814843, -0.37801579,\\n0.81777791],\\n[ 0.6492278 , -0.74340748,\\n0.13387773,\\n0.08902432]])\\nThe matrix U corresponds to a standardized version of the PCA score matrix\\n(each column standardized to have sum-of-squares one). If we multiply each\\ncolumn of U by the corresponding element of D, we recover the PCA scores\\nexactly (up to a meaningless sign flip).\\nIn [24]: (U * D[None,:])[:3]\\nOut[24]: array([[-0.9856,\\n1.1334, -0.4443,\\n0.1563],\\n[-1.9501,\\n1.0732,\\n2.04\\n, -0.4386],\\n[-1.7632, -0.746 ,\\n0.0548, -0.8347]])\\nIn [25]: scores[:3]\\nOut[25]: array([[ 0.9856, -1.1334, -0.4443,\\n0.1563],\\n[ 1.9501, -1.0732,\\n2.04\\n, -0.4386],\\n[ 1.7632,\\n0.746 ,\\n0.0548, -0.8347]])\\nWhile it would be possible to carry out this lab using the PCA() estimator,\\nhere we use the np.linalg.svd() function in order to illustrate its use.\\nWe now omit 20 entries in the 50 × 4 data matrix at random. We do\\nso by first selecting 20 rows (states) at random, and then selecting one of\\nthe four entries in each row at random. This ensures that every row has at\\nleast three observed values.\\nIn [26]: n_omit = 20\\nnp.random.seed(15)\\nr_idx = np.random.choice(np.arange(X.shape[0]),\\nn_omit,\\nreplace=False)\\nc_idx = np.random.choice(np.arange(X.shape[1]),\\nn_omit,\\nreplace=True)\\nXna = X.copy()\\nXna[r_idx, c_idx] = np.nan\\nHere the array r_idx contains 20 integers from 0 to 49; this represents\\nthe states (rows of X) that are selected to contain missing values. And c_idx\\ncontains 20 integers from 0 to 3, representing the features (columns in X)\\nthat contain the missing values for each of the selected states.\\nWe now write some code to implement Algorithm 12.1. We first write a\\nfunction that takes in a matrix, and returns an approximation to the matrix\\nusing the svd() function. This will be needed in Step 2 of Algorithm 12.1.\\nIn [27]: def low_rank(X, M=1):\\nU, D, V = np.linalg.svd(X)\\nL = U[:,:M] * D[None,:M]\\nreturn L.dot(V[:M]) 12.5 Lab: Unsupervised Learning\\n541\\nTo conduct Step 1 of the algorithm, we initialize Xhat — this is ˜X in\\nAlgorithm 12.1 — by replacing the missing values with the column means\\nof the non-missing entries. These are stored in Xbar below after running\\nnp.nanmean() over the row axis. We make a copy so that when we assign np.nanmean()\\nvalues to Xhat below we do not also overwrite the values in Xna.\\nIn [28]: Xhat = Xna.copy()\\nXbar = np.nanmean(Xhat, axis=0)\\nXhat[r_idx, c_idx] = Xbar[c_idx]\\nBefore we begin Step 2, we set ourselves up to measure the progress of\\nour iterations:\\nIn [29]: thresh = 1e-7\\nrel_err = 1\\ncount = 0\\nismiss = np.isnan(Xna)\\nmssold = np.mean(Xhat[∼ismiss]**2)\\nmss0 = np.mean(Xna[∼ismiss]**2)\\nHere ismiss is a logical matrix with the same dimensions as Xna; a given\\nelement is True if the corresponding matrix element is missing. The notation\\n∼ismiss negates this boolean vector. This is useful because it allows us to\\naccess both the missing and non-missing entries. We store the mean of the\\nsquared non-missing elements in mss0. We store the mean squared error\\nof the non-missing elements of the old version of Xhat in mssold (which\\ncurrently agrees with mss0). We plan to store the mean squared error of\\nthe non-missing elements of the current version of Xhat in mss, and will then\\niterate Step 2 of Algorithm 12.1 until the relative error, defined as (mssold\\n- mss) / mss0, falls below thresh = 1e-7. 9\\nIn Step 2(a) of Algorithm 12.1, we approximate Xhat using low_rank();\\nwe call this Xapp. In Step 2(b), we use Xapp to update the estimates for\\nelements in Xhat that are missing in Xna. Finally, in Step 2(c), we compute\\nthe relative error. These three steps are contained in the following while\\nloop:\\nIn [30]: while rel_err > thresh:\\ncount += 1\\n# Step 2(a)\\nXapp = low_rank(Xhat, M=1)\\n# Step 2(b)\\nXhat[ismiss] = Xapp[ismiss]\\n# Step 2(c)\\nmss = np.mean(((Xna - Xapp)[∼ismiss])**2)\\nrel_err = (mssold - mss) / mss0\\nmssold = mss\\nprint(\"Iteration: {0}, MSS:{1:.3f}, Rel.Err {2:.2e}\"\\n.format(count, mss, rel_err))\\n9Algorithm 12.1 tells us to iterate Step 2 until (12.14) is no longer decreasing. Deter-\\nmining whether (12.14) is decreasing requires us only to keep track of mssold - mss.\\nHowever, in practice, we keep track of (mssold - mss) / mss0 instead: this makes\\nit so that the number of iterations required for Algorithm 12.1 to converge does not\\ndepend on whether we multiplied the raw data X by a constant factor. 542\\n12. Unsupervised Learning\\nIteration: 1, MSS:0.395, Rel.Err 5.99e-01\\nIteration: 2, MSS:0.382, Rel.Err 1.33e-02\\nIteration: 3, MSS:0.381, Rel.Err 1.44e-03\\nIteration: 4, MSS:0.381, Rel.Err 1.79e-04\\nIteration: 5, MSS:0.381, Rel.Err 2.58e-05\\nIteration: 6, MSS:0.381, Rel.Err 4.22e-06\\nIteration: 7, MSS:0.381, Rel.Err 7.65e-07\\nIteration: 8, MSS:0.381, Rel.Err 1.48e-07\\nIteration: 9, MSS:0.381, Rel.Err 2.95e-08\\nWe see that after eight iterations, the relative error has fallen below\\nthresh = 1e-7, and so the algorithm terminates. When this happens, the\\nmean squared error of the non-missing elements equals 0.381.\\nFinally, we compute the correlation between the 20 imputed values and\\nthe actual values:\\nIn [31]: np.corrcoef(Xapp[ismiss], X[ismiss])[0,1]\\nOut[31]: 0.711\\nIn this lab, we implemented Algorithm 12.1 ourselves for didactic pur-\\nposes. However, a reader who wishes to apply matrix completion to their\\ndata might look to more specialized Python implementations.\\n12.5.3\\nClustering\\nK-Means Clustering\\nThe estimator sklearn.cluster.KMeans() performs K-means clustering in Kmeans()\\nPython. We begin with a simple simulated example in which there truly are\\ntwo clusters in the data: the first 25 observations have a mean shift relative\\nto the next 25 observations.\\nIn [32]: np.random.seed(0);\\nX = np.random.standard_normal((50,2));\\nX[:25,0] += 3;\\nX[:25,1] -= 4;\\nWe now perform K-means clustering with K = 2.\\nIn [33]: kmeans = KMeans(n_clusters=2,\\nrandom_state=2,\\nn_init=20).fit(X)\\nWe specify random_state to make the results reproducible. The cluster as-\\nsignments of the 50 observations are contained in kmeans.labels_.\\nIn [34]: kmeans.labels_\\nOut[34]: array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,\\n1, 1, 0, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,\\n0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0], dtype=int32)\\nThe K-means clustering perfectly separated the observations into two clus-\\nters even though we did not supply any group information to KMeans(). We\\ncan plot the data, with each observation colored according to its cluster\\nassignment. 12.5 Lab: Unsupervised Learning\\n543\\nIn [35]: fig, ax = plt.subplots(1, 1, figsize=(8,8))\\nax.scatter(X[:,0], X[:,1], c=kmeans.labels_)\\nax.set_title(\"K-Means Clustering Results with K=2\");\\nHere the observations can be easily plotted because they are two-dimensio-\\nnal. If there were more than two variables then we could instead perform\\nPCA and plot the first two principal component score vectors to represent\\nthe clusters.\\nIn this example, we knew that there really were two clusters because\\nwe generated the data. However, for real data, we do not know the true\\nnumber of clusters, nor whether they exist in any precise way. We could\\ninstead have performed K-means clustering on this example with K = 3.\\nIn [36]: kmeans = KMeans(n_clusters=3,\\nrandom_state=3,\\nn_init=20).fit(X)\\nfig, ax = plt.subplots(figsize=(8,8))\\nax.scatter(X[:,0], X[:,1], c=kmeans.labels_)\\nax.set_title(\"K-Means Clustering Results with K=3\");\\nWhen K = 3, K-means clustering splits up the two clusters. We have used\\nthe n_init argument to run the K-means with 20 initial cluster assignments\\n(the default is 10). If a value of n_init greater than one is used, then K-\\nmeans clustering will be performed using multiple random assignments in\\nStep 1 of Algorithm 12.2, and the KMeans() function will report only the\\nbest results. Here we compare using n_init=1 to n_init=20.\\nIn [37]: kmeans1 = KMeans(n_clusters=3,\\nrandom_state=3,\\nn_init=1).fit(X)\\nkmeans20 = KMeans(n_clusters=3,\\nrandom_state=3,\\nn_init=20).fit(X);\\nkmeans1.inertia_, kmeans20.inertia_\\nOut[37]: (78.06, 75.04)\\nNote that kmeans.inertia_ is the total within-cluster sum of squares, which\\nwe seek to minimize by performing K-means clustering (12.17).\\nWe strongly recommend always running K-means clustering with a large\\nvalue of n_init, such as 20 or 50, since otherwise an undesirable local\\noptimum may be obtained.\\nWhen performing K-means clustering, in addition to using multiple ini-\\ntial cluster assignments, it is also important to set a random seed using the\\nrandom_state argument to KMeans(). This way, the initial cluster assign-\\nments in Step 1 can be replicated, and the K-means output will be fully\\nreproducible.\\nHierarchical Clustering\\nThe AgglomerativeClustering() class from the sklearn.clustering pack- Agglomerative\\nClustering()\\nage implements hierarchical clustering. As its name is long, we use the\\nshort hand HClust for hierarchical clustering. Note that this will not change 544\\n12. Unsupervised Learning\\nthe return type when using this method, so instances will still be of class\\nAgglomerativeClustering. In the following example we use the data from\\nthe previous lab to plot the hierarchical clustering dendrogram using com-\\nplete, single, and average linkage clustering with Euclidean distance as the\\ndissimilarity measure. We begin by clustering observations using complete\\nlinkage.\\nIn [38]: HClust = AgglomerativeClustering\\nhc_comp = HClust(distance_threshold=0,\\nn_clusters=None,\\nlinkage=\\'complete\\')\\nhc_comp.fit(X)\\nThis computes the entire dendrogram. We could just as easily perform\\nhierarchical clustering with average or single linkage instead:\\nIn [39]: hc_avg = HClust(distance_threshold=0,\\nn_clusters=None,\\nlinkage=\\'average\\');\\nhc_avg.fit(X)\\nhc_sing = HClust(distance_threshold=0,\\nn_clusters=None,\\nlinkage=\\'single\\');\\nhc_sing.fit(X);\\nTo use a precomputed distance matrix, we provide an additional argu-\\nment metric=\"precomputed\". In the code below, the first four lines computes\\nthe 50 × 50 pairwise-distance matrix.\\nIn [40]: D = np.zeros((X.shape[0], X.shape[0]));\\nfor i in range(X.shape[0]):\\nx_ = np.multiply.outer(np.ones(X.shape[0]), X[i])\\nD[i] = np.sqrt(np.sum((X - x_)**2, 1));\\nhc_sing_pre = HClust(distance_threshold=0,\\nn_clusters=None,\\nmetric=\\'precomputed\\',\\nlinkage=\\'single\\')\\nhc_sing_pre.fit(D)\\nWe use dendrogram() from scipy.cluster.hierarchy to plot the dendro- dendrogram()\\ngram. However, dendrogram() expects a so-called linkage-matrix representa-\\ntion of the clustering, which is not provided by AgglomerativeClustering(),\\nbut can be computed. The function compute_linkage() in the ISLP.cluster compute_\\nlinkage()\\nISLP.cluster\\npackage is provided for this purpose.\\nWe can now plot the dendrograms. The numbers at the bottom of the\\nplot identify each observation. The dendrogram() function has a default\\nmethod to color different branches of the tree that suggests a pre-defined\\ncut of the tree at a particular depth. We prefer to overwrite this default by\\nsetting this threshold to be infinite. Since we want this behavior for many\\ndendrograms, we store these values in a dictionary cargs and pass this as\\nkeyword arguments using the notation **cargs.\\nIn [41]: cargs = {\\'color_threshold\\':-np.inf,\\n\\'above_threshold_color\\':\\'black\\'}\\nlinkage_comp = compute_linkage(hc_comp)\\nfig, ax = plt.subplots(1, 1, figsize=(8, 8)) 12.5 Lab: Unsupervised Learning\\n545\\ndendrogram(linkage_comp ,\\nax=ax,\\n**cargs);\\nWe may want to color branches of the tree above and below a cut-\\nthreshold differently. This can be achieved by changing the color_threshold.\\nLet’s cut the tree at a height of 4, coloring links that merge above 4 in black.\\nIn [42]: fig, ax = plt.subplots(1, 1, figsize=(8, 8))\\ndendrogram(linkage_comp ,\\nax=ax,\\ncolor_threshold=4,\\nabove_threshold_color=\\'black\\');\\nTo determine the cluster labels for each observation associated with a\\ngiven cut of the dendrogram, we can use the cut_tree() function from cut_tree()\\nscipy.cluster.hierarchy:\\nIn [43]: cut_tree(linkage_comp , n_clusters=4).T\\nOut[43]: array([[0, 1, 0, 0, 1, 1, 0, 1, 0, 0, 2, 0, 0, 0, 1, 1, 0, 0, 1,\\n0, 0, 2, 0, 2, 2, 3, 2, 3, 3, 3, 3, 2, 3, 3, 3, 3, 2, 3,\\n3, 3, 3, 2, 3, 3, 3, 3, 3, 3, 3, 3]])\\nThis can also be achieved by providing an argument n_clusters to HClust();\\nhowever each cut would require recomputing the clustering. Similarly, trees\\nmay be cut by distance threshold with an argument of distance_threshold\\nto HClust() or height to cut_tree().\\nIn [44]: cut_tree(linkage_comp , height=5)\\nOut[44]: array([[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0,\\n0, 0, 1, 0, 1, 1, 2, 1, 2, 2, 2, 2, 1, 2, 2, 2, 2, 1, 2,\\n2, 2, 2, 1, 2, 2, 2, 2, 2, 2, 2, 2]])\\nTo scale the variables before performing hierarchical clustering of the\\nobservations, we use StandardScaler() as in our PCA example:\\nIn [45]: scaler = StandardScaler()\\nX_scale = scaler.fit_transform(X)\\nhc_comp_scale = HClust(distance_threshold=0,\\nn_clusters=None,\\nlinkage=\\'complete\\').fit(X_scale)\\nlinkage_comp_scale = compute_linkage(hc_comp_scale)\\nfig, ax = plt.subplots(1, 1, figsize=(8, 8))\\ndendrogram(linkage_comp_scale , ax=ax, **cargs)\\nax.set_title(\"Hierarchical Clustering with Scaled Features\");\\nCorrelation-based distances between observations can be used for clus-\\ntering. The correlation between two observations measures the similarity\\nof their feature values.10 With n observations, the n × n correlation matrix\\n10Suppose each observation has p features, each a single numerical value. We measure\\nthe similarity of two such observations by computing the correlation of these p pairs of\\nnumbers. 546\\n12. Unsupervised Learning\\ncan then be used as a similarity (or affinity) matrix, i.e. so that one minus\\nthe correlation matrix is the dissimilarity matrix used for clustering.\\nNote that using correlation only makes sense for data with at least three\\nfeatures since the absolute correlation between any two observations with\\nmeasurements on two features is always one. Hence, we will cluster a three-\\ndimensional data set.\\nIn [46]: X = np.random.standard_normal((30, 3))\\ncorD = 1 - np.corrcoef(X)\\nhc_cor = HClust(linkage=\\'complete\\',\\ndistance_threshold=0,\\nn_clusters=None,\\nmetric=\\'precomputed\\')\\nhc_cor.fit(corD)\\nlinkage_cor = compute_linkage(hc_cor)\\nfig, ax = plt.subplots(1, 1, figsize=(8, 8))\\ndendrogram(linkage_cor , ax=ax, **cargs)\\nax.set_title(\"Complete Linkage with Correlation -Based Dissimilarity\\n\");\\n12.5.4\\nNCI60 Data Example\\nUnsupervised techniques are often used in the analysis of genomic data.\\nIn particular, PCA and hierarchical clustering are popular tools. We illus-\\ntrate these techniques on the NCI60 cancer cell line microarray data, which\\nconsists of 6830 gene expression measurements on 64 cancer cell lines.\\nIn [47]: NCI60 = load_data(\\'NCI60\\')\\nnci_labs = NCI60[\\'labels\\']\\nnci_data = NCI60[\\'data\\']\\nEach cell line is labeled with a cancer type. We do not make use of the\\ncancer types in performing PCA and clustering, as these are unsupervised\\ntechniques. But after performing PCA and clustering, we will check to\\nsee the extent to which these cancer types agree with the results of these\\nunsupervised techniques.\\nThe data has 64 rows and 6830 columns.\\nIn [48]: nci_data.shape\\nOut[48]: (64, 6830)\\nWe begin by examining the cancer types for the cell lines.\\nIn [49]: nci_labs.value_counts()\\nOut[49]: label\\nNSCLC\\n9\\nRENAL\\n9\\nMELANOMA\\n8\\nBREAST\\n7\\nCOLON\\n7\\nLEUKEMIA\\n6\\nOVARIAN\\n6 12.5 Lab: Unsupervised Learning\\n547\\nCNS\\n5\\nPROSTATE\\n2\\nK562A-repro\\n1\\nK562B-repro\\n1\\nMCF7A-repro\\n1\\nMCF7D-repro\\n1\\nUNKNOWN\\n1\\ndtype: int64\\nPCA on the NCI60 Data\\nWe first perform PCA on the data after scaling the variables (genes) to have\\nstandard deviation one, although here one could reasonably argue that it\\nis better not to scale the genes as they are measured in the same units.\\nIn [50]: scaler = StandardScaler()\\nnci_scaled = scaler.fit_transform(nci_data)\\nnci_pca = PCA()\\nnci_scores = nci_pca.fit_transform(nci_scaled)\\nWe now plot the first few principal component score vectors, in order to\\nvisualize the data. The observations (cell lines) corresponding to a given\\ncancer type will be plotted in the same color, so that we can see to what\\nextent the observations within a cancer type are similar to each other.\\nIn [51]: cancer_types = list(np.unique(nci_labs))\\nnci_groups = np.array([cancer_types.index(lab)\\nfor lab in nci_labs.values])\\nfig, axes = plt.subplots(1, 2, figsize=(15,6))\\nax = axes[0]\\nax.scatter(nci_scores[:,0],\\nnci_scores[:,1],\\nc=nci_groups,\\nmarker=\\'o\\',\\ns=50)\\nax.set_xlabel(\\'PC1\\'); ax.set_ylabel(\\'PC2\\')\\nax = axes[1]\\nax.scatter(nci_scores[:,0],\\nnci_scores[:,2],\\nc=nci_groups,\\nmarker=\\'o\\',\\ns=50)\\nax.set_xlabel(\\'PC1\\'); ax.set_ylabel(\\'PC3\\');\\nThe resulting plots are shown in Figure 12.17. On the whole, cell lines\\ncorresponding to a single cancer type do tend to have similar values on the\\nfirst few principal component score vectors. This indicates that cell lines\\nfrom the same cancer type tend to have pretty similar gene expression\\nlevels.\\nWe can also plot the percent variance explained by the principal compo-\\nnents as well as the cumulative percent variance explained. This is similar\\nto the plots we made earlier for the USArrests data.\\nIn [52]: fig, axes = plt.subplots(1, 2, figsize=(15,6))\\nax = axes[0]\\nticks = np.arange(nci_pca.n_components_)+1 548\\n12. Unsupervised Learning\\n−40\\n−20\\n0\\n20\\n40\\n60\\n−60\\n−40\\n−20\\n0\\n20\\n−40\\n−20\\n0\\n20\\n40\\n60\\n−40\\n−20\\n0\\n20\\n40\\nZ1\\nZ1\\nZ2\\nZ3\\nFIGURE 12.17. Projections of the NCI60 cancer cell lines onto the first three\\nprincipal components (in other words, the scores for the first three principal com-\\nponents). On the whole, observations belonging to a single cancer type tend to\\nlie near each other in this low-dimensional space. It would not have been possible\\nto visualize the data without using a dimension reduction method such as PCA,\\nsince based on the full data set there are\\n.6,830\\n2\\n/\\npossible scatterplots, none of\\nwhich would have been particularly informative.\\nax.plot(ticks,\\nnci_pca.explained_variance_ratio_ ,\\nmarker=\\'o\\')\\nax.set_xlabel(\\'Principal Component\\');\\nax.set_ylabel(\\'PVE\\')\\nax = axes[1]\\nax.plot(ticks,\\nnci_pca.explained_variance_ratio_.cumsum(),\\nmarker=\\'o\\');\\nax.set_xlabel(\\'Principal Component\\')\\nax.set_ylabel(\\'Cumulative PVE\\');\\nThe resulting plots are shown in Figure 12.18.\\nWe see that together, the first seven principal components explain around\\n40% of the variance in the data. This is not a huge amount of the variance.\\nHowever, looking at the scree plot, we see that while each of the first seven\\nprincipal components explain a substantial amount of variance, there is a\\nmarked decrease in the variance explained by further principal components.\\nThat is, there is an elbow in the plot after approximately the seventh princi-\\npal component. This suggests that there may be little benefit to examining\\nmore than seven or so principal components (though even examining seven\\nprincipal components may be difficult).\\nClustering the Observations of the NCI60 Data\\nWe now perform hierarchical clustering of the cell lines in the NCI60 data\\nusing complete, single, and average linkage. Once again, the goal is to find\\nout whether or not the observations cluster into distinct types of cancer.\\nEuclidean distance is used as the dissimilarity measure. We first write a\\nshort function to produce the three dendrograms. 12.5 Lab: Unsupervised Learning\\n549\\n0\\n10\\n20\\n30\\n40\\n50\\n60\\n0\\n2\\n4\\n6\\n8\\n10\\nPrincipal Component\\nPVE\\n0\\n10\\n20\\n30\\n40\\n50\\n60\\n20\\n40\\n60\\n80\\n100\\nPrincipal Component\\nCumulative PVE\\nFIGURE 12.18. The PVE of the principal components of the NCI60 cancer cell\\nline microarray data set. Left: the PVE of each principal component is shown.\\nRight: the cumulative PVE of the principal components is shown. Together, all\\nprincipal components explain 100,% of the variance.\\nIn [53]: def plot_nci(linkage, ax, cut=-np.inf):\\ncargs = {\\'above_threshold_color\\':\\'black\\',\\n\\'color_threshold\\':cut}\\nhc = HClust(n_clusters=None,\\ndistance_threshold=0,\\nlinkage=linkage.lower()).fit(nci_scaled)\\nlinkage_ = compute_linkage(hc)\\ndendrogram(linkage_,\\nax=ax,\\nlabels=np.asarray(nci_labs),\\nleaf_font_size=10,\\n**cargs)\\nax.set_title(\\'%s Linkage\\' % linkage)\\nreturn hc\\nLet’s plot our results.\\nIn [54]: fig, axes = plt.subplots(3, 1, figsize=(15,30))\\nax = axes[0]; hc_comp = plot_nci(\\'Complete\\', ax)\\nax = axes[1]; hc_avg = plot_nci(\\'Average\\', ax)\\nax = axes[2]; hc_sing = plot_nci(\\'Single\\', ax)\\nThe results are shown in Figure 12.19. We see that the choice of linkage\\ncertainly does affect the results obtained. Typically, single linkage will tend\\nto yield trailing clusters: very large clusters onto which individual observa-\\ntions attach one-by-one. On the other hand, complete and average linkage\\ntend to yield more balanced, attractive clusters. For this reason, complete\\nand average linkage are generally preferred to single linkage. Clearly cell\\nlines within a single cancer type do tend to cluster together, although the\\nclustering is not perfect. We will use complete linkage hierarchical cluster-\\ning for the analysis that follows. 550\\n12. Unsupervised Learning\\nBREAST\\nBREAST\\nCNS\\nCNS\\nRENAL\\nBREAST\\nNSCLC\\nRENAL\\nMELANOMA\\nOVARIAN\\nOVARIAN\\nNSCLC\\nOVARIAN\\nCOLON\\nCOLON\\nOVARIAN\\nPROSTATE\\nNSCLC\\nNSCLC\\nNSCLC\\nPROSTATE\\nNSCLC\\nMELANOMA\\nRENAL\\nRENAL\\nRENAL\\nOVARIAN\\nUNKNOWN\\nOVARIAN\\nNSCLC\\nCNS\\nCNS\\nCNS\\nNSCLC\\nRENAL\\nRENAL\\nRENAL\\nRENAL\\nNSCLC\\nMELANOMA\\nMELANOMA\\nMELANOMA\\nMELANOMA\\nMELANOMA\\nMELANOMA\\nBREAST\\nBREAST\\nCOLON\\nCOLON\\nCOLON\\nCOLON\\nCOLON\\nBREAST\\nMCF7A−repro\\nBREAST\\nMCF7D−repro\\nLEUKEMIA\\nLEUKEMIA\\nLEUKEMIA\\nLEUKEMIA\\nK562B−repro\\nK562A−repro\\nLEUKEMIA\\nLEUKEMIA\\n40\\n80\\n120\\n160\\nComplete Linkage\\nLEUKEMIA\\nLEUKEMIA\\nLEUKEMIA\\nLEUKEMIA\\nLEUKEMIA\\nLEUKEMIA\\nK562B−repro\\nK562A−repro\\nRENAL\\nNSCLC\\nBREAST\\nNSCLC\\nBREAST\\nMCF7A−repro\\nBREAST\\nMCF7D−repro\\nCOLON\\nCOLON\\nCOLON\\nRENAL\\nMELANOMA\\nMELANOMA\\nBREAST\\nBREAST\\nMELANOMA\\nMELANOMA\\nMELANOMA\\nMELANOMA\\nMELANOMA\\nOVARIAN\\nOVARIAN\\nNSCLC\\nOVARIAN\\nUNKNOWN\\nOVARIAN\\nNSCLC\\nMELANOMA\\nCNS\\nCNS\\nCNS\\nRENAL\\nRENAL\\nRENAL\\nRENAL\\nRENAL\\nRENAL\\nRENAL\\nPROSTATE\\nNSCLC\\nNSCLC\\nNSCLC\\nNSCLC\\nOVARIAN\\nPROSTATE\\nNSCLC\\nCOLON\\nCOLON\\nOVARIAN\\nCOLON\\nCOLON\\nCNS\\nCNS\\nBREAST\\nBREAST\\n40\\n60\\n80\\n100\\n120\\nAverage Linkage\\nLEUKEMIA\\nRENAL\\nBREAST\\nLEUKEMIA\\nLEUKEMIA\\nCNS\\nLEUKEMIA\\nLEUKEMIA\\nK562B−repro\\nK562A−repro\\nNSCLC\\nLEUKEMIA\\nOVARIAN\\nNSCLC\\nCNS\\nBREAST\\nNSCLC\\nOVARIAN\\nCOLON\\nBREAST\\nMELANOMA\\nRENAL\\nMELANOMA\\nBREAST\\nBREAST\\nMELANOMA\\nMELANOMA\\nMELANOMA\\nMELANOMA\\nMELANOMA\\nBREAST\\nOVARIAN\\nCOLON\\nMCF7A−repro\\nBREAST\\nMCF7D−repro\\nUNKNOWN\\nOVARIAN\\nNSCLC\\nNSCLC\\nPROSTATE\\nMELANOMA\\nCOLON\\nOVARIAN\\nNSCLC\\nRENAL\\nCOLON\\nPROSTATE\\nCOLON\\nOVARIAN\\nCOLON\\nCOLON\\nNSCLC\\nNSCLC\\nRENAL\\nNSCLC\\nRENAL\\nRENAL\\nRENAL\\nRENAL\\nRENAL\\nCNS\\nCNS\\nCNS\\n40\\n60\\n80\\n100\\nSingle Linkage\\nFIGURE 12.19.\\nThe NCI60 cancer cell line microarray data, clustered with\\naverage, complete, and single linkage, and using Euclidean distance as the dissim-\\nilarity measure. Complete and average linkage tend to yield evenly sized clusters\\nwhereas single linkage tends to yield extended clusters to which single leaves are\\nfused one by one. 12.5 Lab: Unsupervised Learning\\n551\\nWe can cut the dendrogram at the height that will yield a particular\\nnumber of clusters, say four:\\nIn [55]: linkage_comp = compute_linkage(hc_comp)\\ncomp_cut = cut_tree(linkage_comp , n_clusters=4).reshape(-1)\\npd.crosstab(nci_labs[\\'label\\'],\\npd.Series(comp_cut.reshape(-1), name=\\'Complete\\'))\\nThere are some clear patterns. All the leukemia cell lines fall in one\\ncluster, while the breast cancer cell lines are spread out over three different\\nclusters.\\nWe can plot a cut on the dendrogram that produces these four clusters:\\nIn [56]: fig, ax = plt.subplots(figsize=(10,10))\\nplot_nci(\\'Complete\\', ax, cut=140)\\nax.axhline(140, c=\\'r\\', linewidth=4);\\nThe axhline() function draws a horizontal line line on top of any existing\\nset of axes. The argument 140 plots a horizontal line at height 140 on the\\ndendrogram; this is a height that results in four distinct clusters. It is easy\\nto verify that the resulting clusters are the same as the ones we obtained\\nin comp_cut.\\nWe claimed earlier in Section 12.4.2 that K-means clustering and hier-\\narchical clustering with the dendrogram cut to obtain the same number\\nof clusters can yield very different results. How do these NCI60 hierarchical\\nclustering results compare to what we get if we perform K-means clustering\\nwith K = 4?\\nIn [57]: nci_kmeans = KMeans(n_clusters=4,\\nrandom_state=0,\\nn_init=20).fit(nci_scaled)\\npd.crosstab(pd.Series(comp_cut, name=\\'HClust\\'),\\npd.Series(nci_kmeans.labels_, name=\\'K-means\\'))\\nOut[57]:\\nK-means\\n0\\n1\\n2\\n3\\nHClust\\n0\\n28\\n3\\n9\\n0\\n1\\n7\\n0\\n0\\n0\\n2\\n0\\n0\\n0\\n8\\n3\\n0\\n9\\n0\\n0\\nWe see that the four clusters obtained using hierarchical clustering and\\nK-means clustering are somewhat different. First we note that the labels\\nin the two clusterings are arbitrary. That is, swapping the identifier of the\\ncluster does not change the clustering. We see here Cluster 3 in K-means\\nclustering is identical to cluster 2 in hierarchical clustering. However, the\\nother clusters differ: for instance, cluster 0 in K-means clustering contains\\na portion of the observations assigned to cluster 0 by hierarchical cluster-\\ning, as well as all of the observations assigned to cluster 1 by hierarchical\\nclustering.\\nRather than performing hierarchical clustering on the entire data ma-\\ntrix, we can also perform hierarchical clustering on the first few principal\\ncomponent score vectors, regarding these first few components as a less\\nnoisy version of the data. 552\\n12. Unsupervised Learning\\nIn [58]: hc_pca = HClust(n_clusters=None,\\ndistance_threshold=0,\\nlinkage=\\'complete\\'\\n).fit(nci_scores[:,:5])\\nlinkage_pca = compute_linkage(hc_pca)\\nfig, ax = plt.subplots(figsize=(8,8))\\ndendrogram(linkage_pca ,\\nlabels=np.asarray(nci_labs),\\nleaf_font_size=10,\\nax=ax,\\n**cargs)\\nax.set_title(\"Hier. Clust. on First Five Score Vectors\")\\npca_labels = pd.Series(cut_tree(linkage_pca ,\\nn_clusters=4).reshape(-1),\\nname=\\'Complete-PCA\\')\\npd.crosstab(nci_labs[\\'label\\'], pca_labels)\\n12.6\\nExercises\\nConceptual\\n1. This problem involves the K-means clustering algorithm.\\n(a) Prove (12.18).\\n(b) On the basis of this identity, argue that the K-means clustering\\nalgorithm (Algorithm 12.2) decreases the objective (12.17) at\\neach iteration.\\n2. Suppose that we have four observations, for which we compute a\\ndissimilarity matrix, given by\\n\\uf8ee\\n\\uf8ef\\uf8ef\\uf8f0\\n0.3\\n0.4\\n0.7\\n0.3\\n0.5\\n0.8\\n0.4\\n0.5\\n0.45\\n0.7\\n0.8\\n0.45\\n\\uf8f9\\n\\uf8fa\\uf8fa\\uf8fb.\\nFor instance, the dissimilarity between the first and second obser-\\nvations is 0.3, and the dissimilarity between the second and fourth\\nobservations is 0.8.\\n(a) On the basis of this dissimilarity matrix, sketch the dendrogram\\nthat results from hierarchically clustering these four observa-\\ntions using complete linkage. Be sure to indicate on the plot the\\nheight at which each fusion occurs, as well as the observations\\ncorresponding to each leaf in the dendrogram.\\n(b) Repeat (a), this time using single linkage clustering.\\n(c) Suppose that we cut the dendrogram obtained in (a) such that\\ntwo clusters result. Which observations are in each cluster?\\n(d) Suppose that we cut the dendrogram obtained in (b) such that\\ntwo clusters result. Which observations are in each cluster? 12.6 Exercises\\n553\\n(e) It is mentioned in this chapter that at each fusion in the den-\\ndrogram, the position of the two clusters being fused can be\\nswapped without changing the meaning of the dendrogram. Draw\\na dendrogram that is equivalent to the dendrogram in (a), for\\nwhich two or more of the leaves are repositioned, but for which\\nthe meaning of the dendrogram is the same.\\n3. In this problem, you will perform K-means clustering manually, with\\nK = 2, on a small example with n = 6 observations and p = 2\\nfeatures. The observations are as follows.\\nObs.\\nX1\\nX2\\n1\\n1\\n4\\n2\\n1\\n3\\n3\\n0\\n4\\n4\\n5\\n1\\n5\\n6\\n2\\n6\\n4\\n0\\n(a) Plot the observations.\\n(b) Randomly assign a cluster label to each observation. You can use\\nthe np.random.choice() function to do this. Report the cluster\\nlabels for each observation.\\n(c) Compute the centroid for each cluster.\\n(d) Assign each observation to the centroid to which it is closest, in\\nterms of Euclidean distance. Report the cluster labels for each\\nobservation.\\n(e) Repeat (c) and (d) until the answers obtained stop changing.\\n(f) In your plot from (a), color the observations according to the\\ncluster labels obtained.\\n4. Suppose that for a particular data set, we perform hierarchical clus-\\ntering using single linkage and using complete linkage. We obtain two\\ndendrograms.\\n(a) At a certain point on the single linkage dendrogram, the clus-\\nters {1, 2, 3} and {4, 5} fuse. On the complete linkage dendro-\\ngram, the clusters {1, 2, 3} and {4, 5} also fuse at a certain point.\\nWhich fusion will occur higher on the tree, or will they fuse at\\nthe same height, or is there not enough information to tell?\\n(b) At a certain point on the single linkage dendrogram, the clusters\\n{5} and {6} fuse. On the complete linkage dendrogram, the clus-\\nters {5} and {6} also fuse at a certain point. Which fusion will\\noccur higher on the tree, or will they fuse at the same height, or\\nis there not enough information to tell?\\n5. In words, describe the results that you would expect if you performed\\nK-means clustering of the eight shoppers in Figure 12.16, on the\\nbasis of their sock and computer purchases, with K = 2. Give three\\nanswers, one for each of the variable scalings displayed. Explain. 554\\n12. Unsupervised Learning\\n6. We saw in Section 12.2.2 that the principal component loading and\\nscore vectors provide an approximation to a matrix, in the sense of\\n(12.5). Specifically, the principal component score and loading vectors\\nsolve the optimization problem given in (12.6).\\nNow, suppose that the M principal component score vectors zim, m =\\n1, . . . , M, are known. Using (12.6), explain that each of the first M\\nprincipal component loading vectors φjm, m = 1, . . . , M, can be ob-\\ntained by performing p separate least squares linear regressions. In\\neach regression, the principal component score vectors are the pre-\\ndictors, and one of the features of the data matrix is the response.\\nApplied\\n7. In this chapter, we mentioned the use of correlation-based distance\\nand Euclidean distance as dissimilarity measures for hierarchical clus-\\ntering. It turns out that these two measures are almost equivalent: if\\neach observation has been centered to have mean zero and standard\\ndeviation one, and if we let rij denote the correlation between the ith\\nand jth observations, then the quantity 1 −rij is proportional to the\\nsquared Euclidean distance between the ith and jth observations.\\nOn the USArrests data, show that this proportionality holds.\\nHint:\\nThe\\nEuclidean\\ndistance\\ncan\\nbe\\ncalculated\\nusing\\nthe\\npairwise_distances() function from the sklearn.metrics module, and pairwise_\\ndistances()\\ncorrelations can be calculated using the np.corrcoef() function.\\n8. In Section 12.2.3, a formula for calculating PVE was given in Equa-\\ntion 12.10. We also saw that the PVE can be obtained using the\\nexplained_variance_ratio_ attribute of a fitted PCA() estimator.\\nOn the USArrests data, calculate PVE in two ways:\\n(a) Using the explained_variance_ratio_ output of the fitted PCA()\\nestimator, as was done in Section 12.2.3.\\n(b) By applying Equation 12.10 directly. The loadings are stored\\nas the components_ attribute of the fitted PCA() estimator. Use\\nthose loadings in Equation 12.10 to obtain the PVE.\\nThese two approaches should give the same results.\\nHint: You will only obtain the same results in (a) and (b) if the same\\ndata is used in both cases. For instance, if in (a) you performed PCA()\\nusing centered and scaled variables, then you must center and scale\\nthe variables before applying Equation 12.10 in (b).\\n9. Consider the USArrests data. We will now perform hierarchical clus-\\ntering on the states.\\n(a) Using\\nhierarchical\\nclustering\\nwith\\ncomplete\\nlinkage\\nand\\nEuclidean distance, cluster the states.\\n(b) Cut the dendrogram at a height that results in three distinct\\nclusters. Which states belong to which clusters? 12.6 Exercises\\n555\\n(c) Hierarchically cluster the states using complete linkage and Eu-\\nclidean distance, after scaling the variables to have standard de-\\nviation one.\\n(d) What effect does scaling the variables have on the hierarchical\\nclustering obtained? In your opinion, should the variables be\\nscaled before the inter-observation dissimilarities are computed?\\nProvide a justification for your answer.\\n10. In this problem, you will generate simulated data, and then perform\\nPCA and K-means clustering on the data.\\n(a) Generate a simulated data set with 20 observations in each of\\nthree classes (i.e. 60 observations total), and 50 variables.\\nHint: There are a number of functions in Python that you can\\nuse to generate data. One example is the normal() method of\\nthe random() function in numpy; the uniform() method is another\\noption. Be sure to add a mean shift to the observations in each\\nclass so that there are three distinct classes.\\n(b) Perform PCA on the 60 observations and plot the first two prin-\\ncipal component score vectors. Use a different color to indicate\\nthe observations in each of the three classes. If the three classes\\nappear separated in this plot, then continue on to part (c). If\\nnot, then return to part (a) and modify the simulation so that\\nthere is greater separation between the three classes. Do not\\ncontinue to part (c) until the three classes show at least some\\nseparation in the first two principal component score vectors.\\n(c) Perform K-means clustering of the observations with K = 3.\\nHow well do the clusters that you obtained in K-means cluster-\\ning compare to the true class labels?\\nHint: You can use the pd.crosstab() function in Python to com-\\npare the true class labels to the class labels obtained by cluster-\\ning. Be careful how you interpret the results: K-means clustering\\nwill arbitrarily number the clusters, so you cannot simply check\\nwhether the true class labels and clustering labels are the same.\\n(d) Perform K-means clustering with K = 2. Describe your results.\\n(e) Now perform K-means clustering with K = 4, and describe your\\nresults.\\n(f) Now perform K-means clustering with K = 3 on the first two\\nprincipal component score vectors, rather than on the raw data.\\nThat is, perform K-means clustering on the 60 × 2 matrix of\\nwhich the first column is the first principal component score\\nvector, and the second column is the second principal component\\nscore vector. Comment on the results.\\n(g) Using the StandardScaler() estimator, perform K-means clus-\\ntering with K = 3 on the data after scaling each variable to have\\nstandard deviation one. How do these results compare to those\\nobtained in (b)? Explain. 556\\n12. Unsupervised Learning\\n11. Write a Python function to perform matrix completion as in Algo-\\nrithm 12.1, and as outlined in Section 12.5.2. In each iteration, the\\nfunction should keep track of the relative error, as well as the itera-\\ntion count. Iterations should continue until the relative error is small\\nenough or until some maximum number of iterations is reached (set a\\ndefault value for this maximum number). Furthermore, there should\\nbe an option to print out the progress in each iteration.\\nTest your function on the Boston data. First, standardize the features\\nto\\nhave\\nmean\\nzero\\nand\\nstandard\\ndeviation\\none\\nusing\\nthe\\nStandardScaler() function. Run an experiment where you randomly\\nleave out an increasing (and nested) number of observations from 5%\\nto 30%, in steps of 5%. Apply Algorithm 12.1 with M = 1, 2, . . . , 8.\\nDisplay the approximation error as a function of the fraction of ob-\\nservations that are missing, and the value of M, averaged over 10\\nrepetitions of the experiment.\\n12. In Section 12.5.2, Algorithm 12.1 was implemented using the\\nsvd() function from the np.linalg module. However, given the con-\\nnection between the svd() function and the PCA() estimator high-\\nlighted in the lab, we could have instead implemented the algorithm\\nusing PCA().\\nWrite a function to implement Algorithm 12.1 that makes use of PCA()\\nrather than svd().\\n13. On the book website, www.statlearning.com, there is a gene expres-\\nsion data set (Ch12Ex13.csv) that consists of 40 tissue samples with\\nmeasurements on 1,000 genes. The first 20 samples are from healthy\\npatients, while the second 20 are from a diseased group.\\n(a) Load in the data using pd.read_csv(). You will need to select\\nheader = None.\\n(b) Apply hierarchical clustering to the samples using correlation-\\nbased distance, and plot the dendrogram. Do the genes separate\\nthe samples into the two groups? Do your results depend on the\\ntype of linkage used?\\n(c) Your collaborator wants to know which genes differ the most\\nacross the two groups. Suggest a way to answer this question,\\nand apply it here. 13\\nMultiple Testing\\nThus far, this textbook has mostly focused on estimation and its close\\ncousin, prediction. In this chapter, we instead focus on hypothesis testing,\\nwhich is key to conducting inference. We remind the reader that inference\\nwas briefly discussed in Chapter 2.\\nWhile Section 13.1 provides a brief review of null hypotheses, p-values,\\ntest statistics, and other key ideas in hypothesis testing, this chapter as-\\nsumes that the reader has had previous exposure to these topics. In par-\\nticular, we will not focus on why or how to conduct a hypothesis test — a\\ntopic on which entire books can be (and have been) written! Instead, we\\nwill assume that the reader is interested in testing some particular set of\\nnull hypotheses, and has a specific plan in mind for how to conduct the\\ntests and obtain p-values.\\nMuch of the emphasis in classical statistics focuses on testing a single null\\nhypothesis, such as H0: the expected blood pressure of mice in the control\\ngroup equals the expected blood pressure of mice in the treatment group. Of\\ncourse, we would probably like to discover that there is a difference between\\nthe mean blood pressure in the two groups. But for reasons that will become\\nclear, we construct a null hypothesis corresponding to no difference.\\nIn contemporary settings, we are often faced with huge amounts of\\ndata, and consequently may wish to test a great many null hypotheses.\\nFor instance, rather than simply testing H0, we might want to test m\\nnull hypotheses, H01, . . . , H0m, where H0j: the expected value of the jth\\nbiomarker among mice in the control group equals the expected value of the\\njth biomarker among mice in the treatment group. When conducting mul-\\ntiple testing, we need to be very careful about how we interpret the results,\\nin order to avoid erroneously rejecting far too many null hypotheses.\\nThis chapter discusses classical as well as more contemporary ways to\\nconduct multiple testing in a big-data setting. In Section 13.2, we highlight\\nthe challenges associated with multiple testing. Classical solutions to these\\n© Springer Nature Switzerland AG 2023 \\nG. James et al., An Introduction to Statistical Learning, Springer Texts in Statistics, \\nhttps://doi.org/10.1007/978-3-031-38747-0_13 \\n557 558\\n13. Multiple Testing\\nchallenges are presented in Section 13.3, and more contemporary solutions\\nin Sections 13.4 and 13.5.\\nIn particular, Section 13.4 focuses on the false discovery rate. The no-\\ntion of the false discovery rate dates back to the 1990s. It quickly rose in\\npopularity in the early 2000s, when large-scale data sets began to come out\\nof genomics. These datasets were unique not only because of their large\\nsize,1 but also because they were typically collected for exploratory pur-\\nposes: researchers collected these datasets in order to test a huge number\\nof null hypotheses, rather than just a very small number of pre-specified\\nnull hypotheses. Today, of course, huge datasets are collected without a\\npre-specified null hypothesis across virtually all fields. As we will see, the\\nfalse discovery rate is perfectly-suited for this modern-day reality.\\nThis chapter naturally centers upon the classical statistical technique of\\np-values, used to quantify the results of hypothesis tests. At the time of\\nwriting of this book (2020), p-values have recently been the topic of exten-\\nsive commentary in the social science research community, to the extent\\nthat some social science journals have gone so far as to ban the use of\\np-values altogether! We will simply comment that when properly under-\\nstood and applied, p-values provide a powerful tool for drawing inferential\\nconclusions from our data.\\n13.1\\nA Quick Review of Hypothesis Testing\\nHypothesis tests provide a rigorous statistical framework for answering\\nsimple “yes-or-no” questions about data, such as the following:\\n1. Is the true coefficient βj in a linear regression of Y onto X1, . . . , Xp\\nequal to zero?2\\n2. Is there a difference in the expected blood pressure of laboratory mice\\nin the control group and laboratory mice in the treatment group?3\\nIn Section 13.1.1, we briefly review the steps involved in hypothesis test-\\ning. Section 13.1.2 discusses the different types of mistakes, or errors, that\\ncan occur in hypothesis testing.\\n13.1.1\\nTesting a Hypothesis\\nConducting a hypothesis test typically proceeds in four steps. First, we de-\\nfine the null and alternative hypotheses. Next, we construct a test statistic\\nthat summarizes the strength of evidence against the null hypothesis. We\\nthen compute a p-value that quantifies the probability of having obtained\\n1Microarray data was viewed as “big data” at the time, although by today’s standards,\\nthis label seems quaint: a microarray dataset can be (and typically was) stored in a\\nMicrosoft Excel spreadsheet!\\n2This hypothesis test was discussed on page 76 of Chapter 3.\\n3The “treatment group” refers to the set of mice that receive an experimental treat-\\nment, and the “control group” refers to those that do not. 13.1 A Quick Review of Hypothesis Testing\\n559\\na comparable or more extreme value of the test statistic under the null\\nhypothesis. Finally, based on the p-value, we decide whether to reject the\\nnull hypothesis. We now briefly discuss each of these steps in turn.\\nStep 1: Define the Null and Alternative Hypotheses\\nIn hypothesis testing, we divide the world into two possibilities: the null\\nhypothesis and the alternative hypothesis. The null hypothesis, denoted H0, null\\nhypothesis\\nalternative\\nhypothesis\\nis the default state of belief about the world.4 For instance, null hypotheses\\nassociated with the two questions posed earlier in this chapter are as follows:\\n1. The true coefficient βj in a linear regression of Y onto X1, . . . , Xp\\nequals zero.\\n2. There is no difference between the expected blood pressure of mice\\nin the control and treatment groups.\\nThe null hypothesis is boring by construction: it may well be true, but we\\nmight hope that our data will tell us otherwise.\\nThe alternative hypothesis, denoted Ha, represents something different\\nand unexpected: for instance, that there is a difference between the ex-\\npected blood pressure of the mice in the two groups. Typically, the alter-\\nnative hypothesis simply posits that the null hypothesis does not hold: if\\nthe null hypothesis states that there is no difference between A and B, then\\nthe alternative hypothesis states that there is a difference between A and\\nB.\\nIt is important to note that the treatment of H0 and Ha is asymmetric.\\nH0 is treated as the default state of the world, and we focus on using data\\nto reject H0. If we reject H0, then this provides evidence in favor of Ha. We\\ncan think of rejecting H0 as making a discovery about our data: namely, we\\nare discovering that H0 does not hold! By contrast, if we fail to reject H0,\\nthen our findings are more nebulous: we will not know whether we failed\\nto reject H0 because our sample size was too small (in which case testing\\nH0 again on a larger or higher-quality dataset might lead to rejection), or\\nwhether we failed to reject H0 because H0 really holds.\\nStep 2: Construct the Test Statistic\\nNext, we wish to use our data in order to find evidence for or against\\nthe null hypothesis. In order to do this, we must compute a test statistic, test statistic\\ndenoted T, which summarizes the extent to which our data are consistent\\nwith H0. The way in which we construct T depends on the nature of the\\nnull hypothesis that we are testing.\\nTo make things concrete, let xt\\n1, . . . , xt\\nnt denote the blood pressure mea-\\nsurements for the nt mice in the treatment group, and let xc\\n1, . . . , xc\\nnc denote\\nthe blood pressure measurements for the nc mice in the control group, and\\nµt = E(Xt), µc = E(Xc). To test H0 : µt = µc, we make use of a two-sample\\nt-statistic,5 defined as\\ntwo-sample\\nt-statistic\\n4H0 is pronounced “H naught” or “H zero”. 560\\n13. Multiple Testing\\n−4\\n−2\\n0\\n2\\n4\\n0.0\\n0.1\\n0.2\\n0.3\\n0.4\\nValue of Test Statistic\\nProbability Density Function\\nT=2.33\\nFIGURE 13.1. The density function for the N(0, 1) distribution, with the ver-\\ntical line indicating a value of 2.33. 1% of the area under the curve falls to the\\nright of the vertical line, so there is only a 2% chance of observing a N(0, 1) value\\nthat is greater than 2.33 or less than −2.33. Therefore, if a test statistic has a\\nN(0, 1) null distribution, then an observed test statistic of T = 2.33 leads to a\\np-value of 0.02.\\nT =\\nˆµt −ˆµc\\ns\\nG\\n1\\nnt +\\n1\\nnc\\n(13.1)\\nwhere ˆµt =\\n1\\nnt\\n)nt\\ni=1 xt\\ni, ˆµc =\\n1\\nnc\\n)nc\\ni=1 xc\\ni, and\\ns =\\n@\\n(nt −1)s2\\nt + (nc −1)s2c\\nnt + nc −2\\n(13.2)\\nis an estimator of the pooled standard deviation of the two samples.6 Here,\\ns2\\nt and s2\\nc are unbiased estimators of the variance of the blood pressure in\\nthe treatment and control groups, respectively. A large (absolute) value of\\nT provides evidence against H0 : µt = µc, and hence evidence in support\\nof Ha : µt ̸= µc.\\nStep 3: Compute the p-Value\\nIn the previous section, we noted that a large (absolute) value of a two-\\nsample t-statistic provides evidence against H0. This begs the question: how\\nlarge is large? In other words, how much evidence against H0 is provided\\nby a given value of the test statistic?\\nThe notion of a p-value provides us with a way to formalize as well as p-value\\nanswer this question. The p-value is defined as the probability of observing\\na test statistic equal to or more extreme than the observed statistic, under\\nthe assumption that H0 is in fact true. Therefore, a small p-value provides\\nevidence against H0.\\n5The t-statistic derives its name from the fact that, under H0, it follows a t-\\ndistribution.\\n6Note that (13.2) assumes that the control and treatment groups have equal variance.\\nWithout this assumption, (13.2) would take a slightly different form. 13.1 A Quick Review of Hypothesis Testing\\n561\\nTo make this concrete, suppose that T = 2.33 for the test statistic in\\n(13.1). Then, we can ask: what is the probability of having observed such\\na large value of T, if indeed H0 holds? It turns out that under H0, the\\ndistribution of T in (13.1) follows approximately a N(0, 1) distribution7 —\\nthat is, a normal distribution with mean 0 and variance 1. This distribution\\nis displayed in Figure 13.1. We see that the vast majority — 98% — of the\\nN(0, 1) distribution falls between −2.33 and 2.33. This means that under\\nH0, we would expect to see such a large value of |T| only 2% of the time.\\nTherefore, the p-value corresponding to T = 2.33 is 0.02.\\nThe distribution of the test statistic under H0 (also known as the test\\nstatistic’s null distribution) will depend on the details of what type of null\\ndistribution\\nnull hypothesis is being tested, and what type of test statistic is used. In\\ngeneral, most commonly-used test statistics follow a well-known statistical\\ndistribution under the null hypothesis — such as a normal distribution,\\na t-distribution, a χ2-distribution, or an F-distribution — provided that\\nthe sample size is sufficiently large and that some other assumptions hold.\\nTypically, the R function that is used to compute a test statistic will make\\nuse of this null distribution in order to output a p-value. In Section 13.5,\\nwe will see an approach to estimate the null distribution of a test statistic\\nusing re-sampling; in many contemporary settings, this is a very attractive\\noption, as it exploits the availability of fast computers in order to avoid\\nhaving to make potentially problematic assumptions about the data.\\nThe p-value is perhaps one of the most used and abused notions in all of\\nstatistics. In particular, it is sometimes said that the p-value is the probabil-\\nity that H0 holds, i.e., that the null hypothesis is true. This is not correct!\\nThe one and only correct interpretation of the p-value is as the fraction\\nof the time that we would expect to see such an extreme value of the test\\nstatistic8 if we repeated the experiment many many times, provided H0\\nholds.\\nIn Step 2 we computed a test statistic, and noted that a large (absolute)\\nvalue of the test statistic provides evidence against H0. In Step 3 the test\\nstatistic was converted to a p-value, with small p-values providing evidence\\nagainst H0. What, then, did we accomplish by converting the test statistic\\nfrom Step 2 into a p-value in Step 3? To answer this question, suppose\\na data analyst conducts a statistical test, and reports a test statistic of\\nT = 17.3. Does this provide strong evidence against H0? It’s impossible\\nto know, without more information: in particular, we would need to know\\n7More precisely, assuming that the observations are drawn from a normal distribution,\\nthen T follows a t-distribution with nt + nc −2 degrees of freedom. Provided that nt +\\nnc −2 is larger than around 40, this is very well-approximated by a N(0, 1) distribution.\\nIn Section 13.5, we will see an alternative and often more attractive way to approximate\\nthe null distribution of T, which avoids making stringent assumptions about the data.\\n8A one-sided p-value is the probability of seeing such an extreme value of the test\\nstatistic; e.g. the probability of seeing a test statistic greater than or equal to T = 2.33.\\nA two-sided p-value is the probability of seeing such an extreme value of the absolute\\ntest statistic; e.g. the probability of seeing a test statistic greater than or equal to 2.33\\nor less than or equal to −2.33. The default recommendation is to report a two-sided\\np-value rather than a one-sided p-value, unless there is a clear and compelling reason\\nthat only one direction of the test statistic is of scientific interest. 562\\n13. Multiple Testing\\nTruth\\nH0\\nHa\\nDecision\\nReject H0\\nType I Error\\nCorrect\\nDo Not Reject H0\\nCorrect\\nType II Error\\nTABLE 13.1. A summary of the possible scenarios associated with testing the\\nnull hypothesis H0. Type I errors are also known as false positives, and Type II\\nerrors as false negatives.\\nwhat value of the test statistic should be expected, under H0. This is exactly\\nwhat a p-value gives us. In other words, a p-value allows us to transform\\nour test statistic, which is measured on some arbitrary and uninterpretable\\nscale, into a number between 0 and 1 that can be more easily interpreted.\\nStep 4: Decide Whether to Reject the Null Hypothesis\\nOnce we have computed a p-value corresponding to H0, it remains for\\nus to decide whether or not to reject H0. (We do not usually talk about\\n“accepting” H0: instead, we talk about “failing to reject” H0.) A small p-\\nvalue indicates that such a large value of the test statistic is unlikely to\\noccur under H0, and thereby provides evidence against H0. If the p-value\\nis sufficiently small, then we will want to reject H0 (and, therefore, make\\na “discovery”). But how small is small enough to reject H0?\\nIt turns out that the answer to this question is very much in the eyes\\nof the beholder, or more specifically, the data analyst. The smaller the p-\\nvalue, the stronger the evidence against H0. In some fields, it is typical to\\nreject H0 if the p-value is below 0.05; this means that, if H0 holds, we would\\nexpect to see such a small p-value no more than 5% of the time.9 However,\\nin other fields, a much higher burden of proof is required: for example, in\\nsome areas of physics, it is typical to reject H0 only if the p-value is below\\n10−9!\\nIn the example displayed in Figure 13.1, if we use a threshold of 0.05 as\\nour cut-off for rejecting the null hypothesis, then we will reject the null. By\\ncontrast, if we use a threshold of 0.01, then we will fail to reject the null.\\nThese ideas are formalized in the next section.\\n13.1.2\\nType I and Type II Errors\\nIf the null hypothesis holds, then we say that it is a true null hypothesis; true null\\nhypothesis\\notherwise, it is a false null hypothesis. For instance, if we test H0 : µt = µc\\nfalse null\\nhypothesis\\nas in Section 13.1.1, and there is indeed no difference in the population\\nmean blood pressure for mice in the treatment group and mice in the\\ncontrol group, then H0 is true; otherwise, it is false. Of course, we do not\\nknow a priori whether H0 is true or whether it is false: this is why we need\\nto conduct a hypothesis test!\\n9Though a threshold of 0.05 to reject H0 is ubiquitous in some areas of science, we\\nadvise against blind adherence to this arbitrary choice. Furthermore, a data analyst\\nshould typically report the p-value itself, rather than just whether or not it exceeds a\\nspecified threshold value. 13.2 The Challenge of Multiple Testing\\n563\\nTable 13.1 summarizes the possible scenarios associated with testing the\\nnull hypothesis H0.10 Once the hypothesis test is performed, the row of the\\ntable is known (based on whether or not we have rejected H0); however, it\\nis impossible for us to know which column we are in. If we reject H0 when\\nH0 is false (i.e., when Ha is true), or if we do not reject H0 when it is true,\\nthen we arrived at the correct result. However, if we erroneously reject H0\\nwhen H0 is in fact true, then we have committed a Type I error. The Type I Type I error\\nerror rate is defined as the probability of making a Type I error given that Type I error\\nrate\\nH0 holds, i.e., the probability of incorrectly rejecting H0. Alternatively, if\\nwe do not reject H0 when H0 is in fact false, then we have committed a\\nType II error. The power of the hypothesis test is defined as the probability Type II\\nerror\\npower\\nof not making a Type II error given that Ha holds, i.e., the probability of\\ncorrectly rejecting H0.\\nIdeally we would like both the Type I and Type II error rates to be small.\\nBut in practice, this is hard to achieve! There typically is a trade-off: we\\ncan make the Type I error small by only rejecting H0 if we are quite sure\\nthat it doesn’t hold; however, this will result in an increase in the Type II\\nerror. Alternatively, we can make the Type II error small by rejecting H0\\nin the presence of even modest evidence that it does not hold, but this will\\ncause the Type I error to be large. In practice, we typically view Type I\\nerrors as more “serious” than Type II errors, because the former involves\\ndeclaring a scientific finding that is not correct. Hence, when we perform\\nhypothesis testing, we typically require a low Type I error rate — e.g.,\\nat most α = 0.05 — while trying to make the Type II error small (or,\\nequivalently, the power large).\\nIt turns out that there is a direct correspondence between the p-value\\nthreshold that causes us to reject H0, and the Type I error rate. By only\\nrejecting H0 when the p-value is below α, we ensure that the Type I error\\nrate will be less than or equal to α.\\n13.2\\nThe Challenge of Multiple Testing\\nIn the previous section, we saw that rejecting H0 if the p-value is below\\n(say) 0.01 provides us with a simple way to control the Type I error for H0\\nat level 0.01: if H0 is true, then there is no more than a 1% probability that\\nwe will reject it. But now suppose that we wish to test m null hypotheses,\\nH01, . . . , H0m. Will it do to simply reject all null hypotheses for which the\\ncorresponding p-value falls below (say) 0.01? Stated another way, if we\\nreject all null hypotheses for which the p-value falls below 0.01, then how\\nmany Type I errors should we expect to make?\\nAs a first step towards answering this question, consider a stockbroker\\nwho wishes to drum up new clients by convincing them of her trading\\n10There are parallels between Table 13.1 and Table 4.6, which has to do with the\\noutput of a binary classifier. In particular, recall from Table 4.6 that a false positive\\nresults from predicting a positive (non-null) label when the true label is in fact negative\\n(null). This is closely related to a Type I error, which results from rejecting the null\\nhypothesis when in fact the null hypothesis holds. 564\\n13. Multiple Testing\\nacumen. She tells 1,024 (1,024 = 210) potential new clients that she can\\ncorrectly predict whether Apple’s stock price will increase or decrease for 10\\ndays running. There are 210 possibilities for how Apple’s stock price might\\nchange over the course of these 10 days. Therefore, she emails each client\\none of these 210 possibilities. The vast majority of her potential clients\\nwill find that the stockbroker’s predictions are no better than chance (and\\nmany will find them to be even worse than chance). But a broken clock is\\nright twice a day, and one of her potential clients will be really impressed\\nto find that her predictions were correct for all 10 of the days! And so the\\nstockbroker gains a new client.\\nWhat happened here? Does the stockbroker have any actual insight into\\nwhether Apple’s stock price will increase or decrease? No. How, then, did\\nshe manage to predict Apple’s stock price perfectly for 10 days running?\\nThe answer is that she made a lot of guesses, and one of them happened\\nto be exactly right.\\nHow does this relate to multiple testing? Suppose that we flip 1,024 fair\\ncoins11 ten times each. Then we would expect (on average) one coin to\\ncome up all tails. (There’s a 1/210 = 1/1,024 chance that any single coin\\nwill come up all tails. So if we flip 1,024 coins, then we expect one coin to\\ncome up all tails, on average.) If one of our coins comes up all tails, then\\nwe might therefore conclude that this particular coin is not fair. In fact, a\\nstandard hypothesis test for the null hypothesis that this particular coin\\nis fair would lead to a p-value below 0.002!12 But it would be incorrect to\\nconclude that the coin is not fair: in fact, the null hypothesis holds, and we\\njust happen to have gotten ten tails in a row by chance.\\nThese examples illustrate the main challenge of multiple testing: when multiple\\ntesting\\ntesting a huge number of null hypotheses, we are bound to get some very\\nsmall p-values by chance. If we make a decision about whether to reject each\\nnull hypothesis without accounting for the fact that we have performed a\\nvery large number of tests, then we may end up rejecting a great number\\nof true null hypotheses — that is, making a large number of Type I errors.\\nHow severe is the problem? Recall from the previous section that if we\\nreject a single null hypothesis, H0, if its p-value is less than, say, α = 0.01,\\nthen there is a 1% chance of making a false rejection if H0 is in fact true.\\nNow what if we test m null hypotheses, H01, . . . , H0m, all of which are true?\\nThere’s a 1% chance of rejecting any individual null hypothesis; therefore,\\nwe expect to falsely reject approximately 0.01 × m null hypotheses. If m =\\n10,000, then that means that we expect to falsely reject 100 null hypotheses\\nby chance! That is a lot of Type I errors.\\nThe crux of the issue is as follows: rejecting a null hypothesis if the p-value\\nis below α controls the probability of falsely rejecting that null hypothesis\\nat level α. However, if we do this for m null hypotheses, then the chance of\\nfalsely rejecting at least one of the m null hypotheses is quite a bit higher!\\n11A fair coin is one that has an equal chance of landing heads or tails.\\n12Recall that the p-value is the probability of observing data at least this extreme,\\nunder the null hypothesis. If the coin is fair, then the probability of observing at least\\nten tails is (1/2)10 = 1/1,024 < 0.001. The p-value is therefore 2/1,024 < 0.002, since\\nthis is the probability of observing ten heads or ten tails. 13.3 The Family-Wise Error Rate\\n565\\nH0 is True\\nH0 is False\\nTotal\\nReject H0\\nV\\nS\\nR\\nDo Not Reject H0\\nU\\nW\\nm −R\\nTotal\\nm0\\nm −m0\\nm\\nTABLE 13.2. A summary of the results of testing m null hypotheses. A given\\nnull hypothesis is either true or false, and a test of that null hypothesis can either\\nreject or fail to reject it. In practice, the individual values of V , S, U, and W are\\nunknown. However, we do have access to V + S = R and U + W = m −R, which\\nare the numbers of null hypotheses rejected and not rejected, respectively.\\nWe will investigate this issue in greater detail, and pose a solution to it, in\\nSection 13.3.\\n13.3\\nThe Family-Wise Error Rate\\nIn the following sections, we will discuss testing multiple hypotheses while\\ncontrolling the probability of making at least one Type I error.\\n13.3.1\\nWhat is the Family-Wise Error Rate?\\nRecall that the Type I error rate is the probability of rejecting H0 if H0 is\\ntrue. The family-wise error rate (FWER) generalizes this notion to the set- family-wise\\nerror rate\\nting of m null hypotheses, H01, . . . , H0m, and is defined as the probability\\nof making at least one Type I error. To state this idea more formally, con-\\nsider Table 13.2, which summarizes the possible outcomes when performing\\nm hypothesis tests. Here, V represents the number of Type I errors (also\\nknown as false positives or false discoveries), S the number of true posi-\\ntives, U the number of true negatives, and W the number of Type II errors\\n(also known as false negatives). Then the family-wise error rate is given by\\nFWER = Pr(V ≥1).\\n(13.3)\\nA strategy of rejecting any null hypothesis for which the p-value is below\\nα (i.e. controlling the Type I error for each null hypothesis at level α) leads\\nto a FWER of\\nFWER(α)\\n=\\n1 −Pr(V = 0)\\n=\\n1 −Pr(do not falsely reject any null hypotheses)\\n=\\n1 −Pr\\n1Ym\\nj=1 {do not falsely reject H0j}\\n2\\n.\\n(13.4)\\nRecall from basic probability that if two events A and B are independent,\\nthen Pr(A∩B) = Pr(A) Pr(B). Therefore, if we make the additional rather\\nstrong assumptions that the m tests are independent and that all m null\\nhypotheses are true, then\\nFWER(α) = 1 −\\nm\\nE\\nj=1\\n(1 −α) = 1 −(1 −α)m.\\n(13.5) 566\\n13. Multiple Testing\\nFIGURE 13.2. The family-wise error rate, as a function of the number of\\nhypotheses tested (displayed on the log scale), for three values of α: α = 0.05\\n(orange), α = 0.01 (blue), and α = 0.001 (purple). The dashed line indicates\\n0.05. For example, in order to control the FWER at 0.05 when testing m = 50\\nnull hypotheses, we must control the Type I error for each null hypothesis at level\\nα = 0.001.\\nHence, if we test only one null hypothesis, then FWER(α) = 1−(1−α)1 =\\nα, so the Type I error rate and the FWER are equal. However, if we perform\\nm = 100 independent tests, then FWER(α) = 1 −(1 −α)100. For instance,\\ntaking α = 0.05 leads to a FWER of 1 −(1 −0.05)100 = 0.994. In other\\nwords, we are virtually guaranteed to make at least one Type I error!\\nFigure 13.2 displays (13.5) for various values of m, the number of hy-\\npotheses, and α, the Type I error. We see that setting α = 0.05 results in\\na high FWER even for moderate m. With α = 0.01, we can test no more\\nthan five null hypotheses before the FWER exceeds 0.05. Only for very\\nsmall values, such as α = 0.001, do we manage to ensure a small FWER,\\nat least for moderately-sized m.\\nWe now briefly return to the example in Section 13.1.1, in which we\\nconsider testing a single null hypothesis of the form H0 : µt = µc using a\\ntwo-sample t-statistic. Recall from Figure 13.1 that in order to guarantee\\nthat the Type I error does not exceed 0.02, we decide whether or not to\\nreject H0 using a cutpoint of 2.33 (i.e. we reject H0 if |T| ≥2.33). Now,\\nwhat if we wish to test 10 null hypotheses using two-sample t-statistics,\\ninstead of just one? We will see in Section 13.3.2 that we can guarantee\\nthat the FWER does not exceed 0.02 by rejecting only null hypotheses\\nfor which the p-value falls below 0.002. This corresponds to a much more\\nstringent cutpoint of 3.09 (i.e. we should reject H0j only if its test statistic\\n|Tj| ≥3.09, for j = 1, . . . , 10). In other words, controlling the FWER at\\nlevel α amounts to a much higher bar, in terms of evidence required to\\nreject any given null hypothesis, than simply controlling the Type I error\\nfor each null hypothesis at level α.\\n1\\n2\\n5\\n10\\n20\\n50\\n100\\n200\\n500\\n0.0\\n0.2\\n0.4\\n0.6\\n0.8\\n1.0\\nNumber of Hypotheses\\nFamily−Wise Error Rate\\nα = 0.05\\nα = 0.01\\nα = 0.001 13.3 The Family-Wise Error Rate\\n567\\nManager\\nMean, ¯x\\nStandard Deviation, s\\nt-statistic\\np-value\\nOne\\n3.0\\n7.4\\n2.86\\n0.006\\nTwo\\n-0.1\\n6.9\\n-0.10\\n0.918\\nThree\\n2.8\\n7.5\\n2.62\\n0.012\\nFour\\n0.5\\n6.7\\n0.53\\n0.601\\nFive\\n0.3\\n6.8\\n0.31\\n0.756\\nTABLE 13.3. The first two columns correspond to the sample mean and sample\\nstandard deviation of the percentage excess return, over n = 50 months, for the\\nfirst five managers in the Fund dataset. The last two columns provide the t-statistic\\n(√n · ¯\\nX/S) and associated p-value for testing H0j : µj = 0, the null hypothesis\\nthat the (population) mean return for the jth hedge fund manager equals zero.\\n13.3.2\\nApproaches to Control the Family-Wise Error Rate\\nIn this section, we briefly survey some approaches to control the FWER.\\nWe will illustrate these approaches on the Fund dataset, which records the\\nmonthly percentage excess returns for 2,000 fund managers over n = 50\\nmonths.13 Table 13.3 provides relevant summary statistics for the first five\\nmanagers.\\nWe first present the Bonferroni method and Holm’s step-down proce-\\ndure, which are very general-purpose approaches for controlling the FWER\\nthat can be applied whenever m p-values have been computed, regardless\\nof the form of the null hypotheses, the choice of test statistics, or the\\n(in)dependence of the p-values. We then briefly discuss Tukey’s method\\nand Scheffé’s method in order to illustrate the fact that, in certain sit-\\nuations, more specialized approaches for controlling the FWER may be\\npreferable.\\nThe Bonferroni Method\\nAs in the previous section, suppose we wish to test H01, . . . , H0m. Let Aj\\ndenote the event that we make a Type I error for the jth null hypothesis,\\nfor j = 1, . . . , m. Then\\nFWER = Pr(falsely reject at least one null hypothesis)\\n= Pr(∪m\\nj=1Aj)\\n≤\\nm\\n0\\nj=1\\nPr(Aj).\\n(13.6)\\nIn (13.6), the inequality results from the fact that for any two events A\\nand B, Pr(A ∪B) ≤Pr(A) + Pr(B), regardless of whether A and B are\\nindependent. The Bonferroni method, or Bonferroni correction, sets the\\nthreshold for rejecting each hypothesis test to α/m, so that Pr(Aj) ≤α/m.\\nEquation 13.6 implies that\\nFWER(α/m) ≤m × α\\nm = α,\\n13Excess returns correspond to the additional return the fund manager achieves beyond\\nthe market’s overall return. So if the market increases by 5% during a given period and\\nthe fund manager achieves a 7% return, their excess return would be 7% −5% = 2%. 568\\n13. Multiple Testing\\nso this procedure controls the FWER at level α. For instance, in order to\\ncontrol the FWER at level 0.1 while testing m = 100 null hypotheses, the\\nBonferroni procedure requires us to control the Type I error for each null\\nhypothesis at level 0.1/100 = 0.001, i.e. to reject all null hypotheses for\\nwhich the p-value is below 0.001.\\nWe now consider the Fund dataset in Table 13.3. If we control the Type\\nI error at level α = 0.05 for each fund manager separately, then we will\\nconclude that the first and third managers have significantly non-zero ex-\\ncess returns; in other words, we will reject H01 : µ1 = 0 and H03 : µ3 = 0.\\nHowever, as discussed in previous sections, this procedure does not account\\nfor the fact that we have tested multiple hypotheses, and therefore it will\\nlead to a FWER greater than 0.05. If we instead wish to control the FWER\\nat level 0.05, then, using a Bonferroni correction, we must control the Type\\nI error for each individual manager at level α/m = 0.05/5 = 0.01. Conse-\\nquently, we will reject the null hypothesis only for the first manager, since\\nthe p-values for all other managers exceed 0.01. The Bonferroni correction\\ngives us peace of mind that we have not falsely rejected too many null\\nhypotheses, but for a price: we reject few null hypotheses, and thus will\\ntypically make quite a few Type II errors.\\nThe Bonferroni correction is by far the best-known and most commonly-\\nused multiplicity correction in all of statistics. Its ubiquity is due in large\\npart to the fact that it is very easy to understand and simple to implement,\\nand also from the fact that it successfully controls Type I error regardless\\nof whether the m hypothesis tests are independent. However, as we will see,\\nit is typically neither the most powerful nor the best approach for multiple\\ntesting correction. In particular, the Bonferroni correction can be quite\\nconservative, in the sense that the true FWER is often quite a bit lower\\nthan the nominal (or target) FWER; this results from the inequality in\\n(13.6). By contrast, a less conservative procedure might allow us to control\\nthe FWER while rejecting more null hypotheses, and therefore making\\nfewer Type II errors.\\nHolm’s Step-Down Procedure\\nHolm’s method, also known as Holm’s step-down procedure or the Holm– Holm’s\\nmethod\\nBonferroni method, is an alternative to the Bonferroni procedure. Holm’s\\nmethod controls the FWER, but it is less conservative than Bonferroni, in\\nthe sense that it will reject more null hypotheses, typically resulting in fewer\\nType II errors and hence greater power. The procedure is summarized in\\nAlgorithm 13.1. The proof that this method controls the FWER is similar\\nto, but slightly more complicated than, the argument in (13.6) that the\\nBonferroni method controls the FWER. It is worth noting that in Holm’s\\nprocedure, the threshold that we use to reject each null hypothesis — p(L)\\nin Step 5 — actually depends on the values of all m of the p-values. (See the\\ndefinition of L in (13.7).) This is in contrast to the Bonferroni procedure,\\nin which to control the FWER at level α, we reject any null hypotheses for\\nwhich the p-value is below α/m, regardless of the other p-values. Holm’s\\nmethod makes no independence assumptions about the m hypothesis tests,\\nand is uniformly more powerful than the Bonferroni method — it will 13.3 The Family-Wise Error Rate\\n569\\nAlgorithm 13.1 Holm’s Step-Down Procedure to Control the FWER\\n1. Specify α, the level at which to control the FWER.\\n2. Compute p-values, p1, . . . , pm, for the m null hypotheses\\nH01, . . . , H0m.\\n3. Order the m p-values so that p(1) ≤p(2) ≤· · · ≤p(m).\\n4. Define\\nL = min\\nK\\nj : p(j) >\\nα\\nm + 1 −j\\nZ\\n.\\n(13.7)\\n5. Reject all null hypotheses H0j for which pj < p(L).\\nalways reject at least as many null hypotheses as Bonferroni — and so it\\nshould always be preferred.\\nWe now consider applying Holm’s method to the first five fund managers\\nin the Fund dataset in Table 13.3, while controlling the FWER at level 0.05.\\nThe ordered p-values are p(1) = 0.006, p(2) = 0.012, p(3) = 0.601, p(4) =\\n0.756 and p(5) = 0.918. The Holm procedure rejects the first two null\\nhypotheses, because p(1) = 0.006 < 0.05/(5 + 1 −1) = 0.01 and p(2) =\\n0.012 < 0.05/(5 + 1 −2) = 0.0125, but p(3) = 0.601 > 0.05/(5 + 1 −3) =\\n0.0167, which implies that L = 3. We note that, in this setting, Holm is\\nmore powerful than Bonferroni: the former rejects the null hypotheses for\\nthe first and third managers, whereas the latter rejects the null hypothesis\\nonly for the first manager.\\nFigure 13.3 provides an illustration of the Bonferroni and Holm methods\\non three simulated data sets in a setting involving m = 10 hypothesis tests,\\nof which m0 = 2 of the null hypotheses are true. Each panel displays the\\nten corresponding p-values, ordered from smallest to largest, and plotted\\non a log scale. The eight red points represent the false null hypotheses, and\\nthe two black points represent the true null hypotheses. We wish to control\\nthe FWER at level 0.05. The Bonferroni procedure requires us to reject all\\nnull hypotheses for which the p-value is below 0.005; this is represented by\\nthe black horizontal line. The Holm procedure requires us to reject all null\\nhypotheses that fall below the blue line. The blue line always lies above the\\nblack line, so Holm will always reject more tests than Bonferroni; the region\\nbetween the two lines corresponds to the hypotheses that are only rejected\\nby Holm. In the left-hand panel, both Bonferroni and Holm successfully\\nreject seven of the eight false null hypotheses. In the center panel, Holm\\nsuccessfully rejects all eight of the false null hypotheses, while Bonferroni\\nfails to reject one. In the right-hand panel, Bonferroni only rejects three of\\nthe false null hypotheses, while Holm rejects all eight. Neither Bonferroni\\nnor Holm makes any Type I errors in these examples.\\nTwo Special Cases: Tukey’s Method and Scheffé’s Method\\nBonferroni’s method and Holm’s method can be used in virtually any set-\\nting in which we wish to control the FWER for m null hypotheses: they 570\\n13. Multiple Testing\\n2\\n4\\n6\\n8\\n10\\n1e−05\\n1e−04\\n1e−03\\n1e−02\\n1e−01\\nOrdering of p−values\\np−values (log scale)\\n2\\n4\\n6\\n8\\n10\\n1e−07\\n1e−05\\n1e−03\\n1e−01\\nOrdering of p−values\\np−values (log scale)\\n2\\n4\\n6\\n8\\n10\\n1e−05\\n1e−04\\n1e−03\\n1e−02\\n1e−01\\nOrdering of p−values\\np−values (log scale)\\nFIGURE 13.3. Each panel displays, for a separate simulation, the sorted\\np-values for tests of m = 10 null hypotheses. The p-values corresponding to\\nthe m0 = 2 true null hypotheses are displayed in black, and the rest are in red.\\nWhen controlling the FWER at level 0.05, the Bonferroni procedure rejects all\\nnull hypotheses that fall below the black line, and the Holm procedure rejects all\\nnull hypotheses that fall below the blue line. The region between the blue and black\\nlines indicates null hypotheses that are rejected using the Holm procedure but not\\nusing the Bonferroni procedure. In the center panel, the Holm procedure rejects\\none more null hypothesis than the Bonferroni procedure. In the right-hand panel,\\nit rejects five more null hypotheses.\\nmake no assumptions about the nature of the null hypotheses, the type\\nof test statistic used, or the (in)dependence of the p-values. However, in\\ncertain very specific settings, we can achieve higher power by controlling\\nthe FWER using approaches that are more tailored to the task at hand.\\nTukey’s method and Scheffé’s method provide two such examples.\\nTable 13.3 indicates that for the Fund dataset, Managers One and Two\\nhave the greatest difference in their sample mean returns. This finding\\nmight motivate us to test the null hypothesis H0 : µ1 = µ2, where µj is the\\n(population) mean return for the jth fund manager. A two-sample t-test\\n(13.1) for H0 yields a p-value of 0.0349, suggesting modest evidence against\\nH0. However, this p-value is misleading, since we decided to compare the\\naverage returns of Managers One and Two only after having examined the\\nreturns for all five managers; this essentially amounts to having performed\\nm = 5 × (5 −1)/2 = 10 hypothesis tests, and selecting the one with the\\nsmallest p-value. This suggests that in order to control the FWER at level\\n0.05, we should make a Bonferroni correction for m = 10 hypothesis tests,\\nand therefore should only reject a null hypothesis for which the p-value\\nis below 0.005. If we do this, then we will be unable to reject the null\\nhypothesis that Managers One and Two have identical performance.\\nHowever, in this setting, a Bonferroni correction is actually a bit too\\nstringent, since it fails to consider the fact that the m = 10 hypothesis\\ntests are all somewhat related: for instance, Managers Two and Five have\\nsimilar mean returns, as do Managers Two and Four; this guarantees that\\nthe mean returns of Managers Four and Five are similar. Stated another\\nway, the m p-values for the m pairwise comparisons are not independent.\\nTherefore, it should be possible to control the FWER in a way that is 13.3 The Family-Wise Error Rate\\n571\\n2\\n4\\n6\\n8\\n10\\n12\\n14\\n1e−04\\n1e−03\\n1e−02\\n1e−01\\n1e+00\\nOrdering of p−values\\np−values (log scale)\\n2\\n4\\n6\\n8\\n10\\n12\\n14\\n1e−06\\n1e−04\\n1e−02\\n1e+00\\nOrdering of p−values\\np−values (log scale)\\n2\\n4\\n6\\n8\\n10\\n12\\n14\\n1e−04\\n1e−03\\n1e−02\\n1e−01\\n1e+00\\nOrdering of p−values\\np−values (log scale)\\nFIGURE 13.4. Each panel displays, for a separate simulation, the sorted\\np-values for tests of m = 15 hypotheses, corresponding to pairwise tests for the\\nequality of G = 6 means. The m0 = 10 true null hypotheses are displayed in black,\\nand the rest are in red. When controlling the FWER at level 0.05, the Bonferroni\\nprocedure rejects all null hypotheses that fall below the black line, whereas Tukey\\nrejects all those that fall below the blue line. Thus, Tukey’s method has slightly\\nhigher power than Bonferroni’s method. Controlling the Type I error without\\nadjusting for multiple testing involves rejecting all those that fall below the green\\nline.\\nless conservative. This is exactly the idea behind Tukey’s method: when Tukey’s\\nmethod\\nperforming m = G(G −1)/2 pairwise comparisons of G means, it allows\\nus to control the FWER at level α while rejecting all null hypotheses for\\nwhich the p-value falls below αT , for some αT > α/m.\\nFigure 13.4 illustrates Tukey’s method on three simulated data sets in a\\nsetting with G = 6 means, with µ1 = µ2 = µ3 = µ4 = µ5 ̸= µ6. Therefore,\\nof the m = G(G −1)/2 = 15 null hypotheses of the form H0 : µj = µk,\\nten are true and five are false. In each panel, the true null hypotheses\\nare displayed in black, and the false ones are in red. The horizontal lines\\nindicate that Tukey’s method always results in at least as many rejections\\nas Bonferroni’s method. In the left-hand panel, Tukey correctly rejects two\\nmore null hypotheses than Bonferroni.\\nNow, suppose that we once again examine the data in Table 13.3, and no-\\ntice that Managers One and Three have higher mean returns than Managers\\nTwo, Four, and Five. This might motivate us to test the null hypothesis\\nH0 : 1\\n2 (µ1 + µ3) = 1\\n3 (µ2 + µ4 + µ5) .\\n(13.8)\\n(Recall that µj is the population mean return for the jth hedge fund man-\\nager.) It turns out that we could test (13.8) using a variant of the two-\\nsample t-test presented in (13.1), leading to a p-value of 0.004. This sug-\\ngests strong evidence of a difference between Managers One and Three\\ncompared to Managers Two, Four, and Five. However, there is a problem:\\nwe decided to test the null hypothesis in (13.8) only after peeking at the\\ndata in Table 13.3. In a sense, this means that we have conducted multiple\\ntesting. In this setting, using Bonferroni to control the FWER at level α 572\\n13. Multiple Testing\\nwould require a p-value threshold of α/m, for an extremely large value of\\nm14.\\nScheffé’s method is designed for exactly this setting. It allows us to com- Scheffé’s\\nmethod\\npute a value αS such that rejecting the null hypothesis H0 in (13.8) if the\\np-value is below αS will control the Type I error at level α. It turns out that\\nfor the Fund example, in order to control the Type I error at level α = 0.05,\\nwe must set αS = 0.002. Therefore, we are unable to reject H0 in (13.8), de-\\nspite the apparently very small p-value of 0.004. An important advantage of\\nScheffé’s method is that we can use this same threshold of αS = 0.002 in or-\\nder to perform a pairwise comparison of any split of the managers into two\\ngroups: for instance, we could also test H0 : 1\\n3 (µ1 + µ2 + µ3) = 1\\n2 (µ4 + µ5)\\nand H0 : 1\\n4 (µ1 + µ2 + µ3 + µ4) = µ5 using the same threshold of 0.002,\\nwithout needing to further adjust for multiple testing.\\nTo summarize, Holm’s procedure and Bonferroni’s procedure are very\\ngeneral approaches for multiple testing correction that can be applied un-\\nder all circumstances. However, in certain special cases, more powerful pro-\\ncedures for multiple testing correction may be available, in order to control\\nthe FWER while achieving higher power (i.e. committing fewer Type II\\nerrors) than would be possible using Holm or Bonferroni. In this section,\\nwe have illustrated two such examples.\\n13.3.3\\nTrade-Off Between the FWER and Power\\nIn general, there is a trade-off between the FWER threshold that we choose,\\nand our power to reject the null hypotheses. Recall that power is defined\\nas the number of false null hypotheses that we reject divided by the total\\nnumber of false null hypotheses, i.e. S/(m −m0) using the notation of Ta-\\nble 13.2. Figure 13.5 illustrates the results of a simulation setting involving\\nm null hypotheses, of which 90% are true and the remaining 10% are false;\\npower is displayed as a function of the FWER. In this particular simulation\\nsetting, when m = 10, a FWER of 0.05 corresponds to power of approxi-\\nmately 60%. However, as m increases, the power decreases. With m = 500,\\nthe power is below 0.2 at a FWER of 0.05, so that we successfully reject\\nonly 20% of the false null hypotheses.\\nFigure 13.5 indicates that it is reasonable to control the FWER when m\\ntakes on a small value, like 5 or 10. However, for m = 100 or m = 1,000,\\nattempting to control the FWER will make it almost impossible to reject\\nany of the false null hypotheses. In other words, the power will be extremely\\nlow.\\nWhy is this the case? Recall that, using the notation in Table 13.2, the\\nFWER is defined as Pr(V ≥1) (13.3). In other other words, controlling the\\nFWER at level α guarantees that the data analyst is very unlikely (with\\nprobability no more than α) to reject any true null hypotheses, i.e. to have\\nany false positives. In order to make good on this guarantee when m is\\nlarge, the data analyst may be forced to reject very few null hypotheses, or\\nperhaps even none at all (since if R = 0 then also V = 0; see Table 13.2).\\n14In fact, calculating the “correct” value of m is quite technical, and outside the scope\\nof this book. 13.4 The False Discovery Rate\\n573\\n0.0\\n0.2\\n0.4\\n0.6\\n0.8\\n1.0\\n0.0\\n0.2\\n0.4\\n0.6\\n0.8\\n1.0\\nFamily−Wise Error Rate\\nPower\\nm = 10\\nm = 100\\nm = 500\\nFIGURE 13.5. In a simulation setting in which 90% of the m null hypotheses are\\ntrue, we display the power (the fraction of false null hypotheses that we successfully\\nreject) as a function of the family-wise error rate. The curves correspond to\\nm = 10 (orange), m = 100 (blue), and m = 500 (purple). As the value of m\\nincreases, the power decreases. The vertical dashed line indicates a FWER of\\n0.05.\\nThis is scientifically uninteresting, and typically results in very low power,\\nas in Figure 13.5.\\nIn practice, when m is large, we may be willing to tolerate a few false\\npositives, in the interest of making more discoveries, i.e. more rejections of\\nthe null hypothesis. This is the motivation behind the false discovery rate,\\nwhich we present next.\\n13.4\\nThe False Discovery Rate\\n13.4.1\\nIntuition for the False Discovery Rate\\nAs we just discussed, when m is large, then trying to prevent any false\\npositives (as in FWER control) is simply too stringent. Instead, we might\\ntry to make sure that the ratio of false positives (V ) to total positives (V +\\nS = R) is sufficiently low, so that most of the rejected null hypotheses are\\nnot false positives. The ratio V/R is known as the false discovery proportion false\\ndiscovery\\nproportion\\n(FDP).\\nIt might be tempting to ask the data analyst to control the FDP: to\\nmake sure that no more than, say, 20% of the rejected null hypotheses are\\nfalse positives. However, in practice, controlling the FDP is an impossible\\ntask for the data analyst, since she has no way to be certain, on any par-\\nticular dataset, which hypotheses are true and which are false. This is very\\nsimilar to the fact that the data analyst can control the FWER, i.e. she\\ncan guarantee that Pr(V ≥1) ≤α for any pre-specified α, but she cannot\\nguarantee that V = 0 on any particular dataset (short of failing to reject\\nany null hypotheses, i.e. setting R = 0). 574\\n13. Multiple Testing\\nTherefore, we instead control the false discovery rate (FDR)15, defined false\\ndiscovery\\nrate\\nas\\nFDR = E(FDP) = E(V/R).\\n(13.9)\\nWhen we control the FDR at (say) level q = 20%, we are rejecting as many\\nnull hypotheses as possible while guaranteeing that no more than 20% of\\nthose rejected null hypotheses are false positives, on average.\\nIn the definition of the FDR in (13.9), the expectation is taken over the\\npopulation from which the data are generated. For instance, suppose we\\ncontrol the FDR for m null hypotheses at q = 0.2. This means that if we\\nrepeat this experiment a huge number of times, and each time control the\\nFDR at q = 0.2, then we should expect that, on average, 20% of the rejected\\nnull hypotheses will be false positives. On a given dataset, the fraction of\\nfalse positives among the rejected hypotheses may be greater than or less\\nthan 20%.\\nThus far, we have motivated the use of the FDR from a pragmatic per-\\nspective, by arguing that when m is large, controlling the FWER is simply\\ntoo stringent, and will not lead to “enough” discoveries. An additional mo-\\ntivation for the use of the FDR is that it aligns well with the way that data\\nare often collected in contemporary applications. As datasets continue to\\ngrow in size across a variety of fields, it is increasingly common to conduct a\\nhuge number of hypothesis tests for exploratory, rather than confirmatory,\\npurposes. For instance, a genomic researcher might sequence the genomes\\nof individuals with and without some particular medical condition, and\\nthen, for each of 20,000 genes, test whether sequence variants in that gene\\nare associated with the medical condition of interest. This amounts to per-\\nforming m = 20,000 hypothesis tests. The analysis is exploratory in nature,\\nin the sense that the researcher does not have any particular hypothesis\\nin mind; instead she wishes to see whether there is modest evidence for\\nthe association between each gene and the disease, with a plan to further\\ninvestigate any genes for which there is such evidence. She is likely willing\\nto tolerate some number of false positives in the set of genes that she will\\ninvestigate further; thus, the FWER is not an appropriate choice. How-\\never, some correction for multiple testing is required: it would not be a\\ngood idea for her to simply investigate all genes with p-values less than\\n(say) 0.05, since we would expect 1,000 genes to have such small p-values\\nsimply by chance, even if no genes are associated with the disease (since\\n0.05 × 20,000 = 1,000). Controlling the FDR for her exploratory analysis\\nat 20% guarantees that — on average — no more than 20% of the genes\\nthat she investigates further are false positives.\\nIt is worth noting that unlike p-values, for which a threshold of 0.05\\nis typically viewed as the minimum standard of evidence for a “positive”\\nresult, and a threshold of 0.01 or even 0.001 is viewed as much more com-\\npelling, there is no standard accepted threshold for FDR control. Instead,\\nthe choice of FDR threshold is typically context-dependent, or even dataset-\\ndependent. For instance, the genomic researcher in the previous example\\nmight seek to control the FDR at a threshold of 10% if the planned follow-\\n15If R = 0, then we replace the ratio V/R with 0, to avoid computing 0/0. Formally,\\nFDR = E(V/R|R > 0) Pr(R > 0). 13.4 The False Discovery Rate\\n575\\nup analysis is time-consuming or expensive. Alternatively, a much larger\\nthreshold of 30% might be suitable if she plans an inexpensive follow-up\\nanalysis.\\n13.4.2\\nThe Benjamini–Hochberg Procedure\\nWe now focus on the task of controlling the FDR: that is, deciding which\\nnull hypotheses to reject while guaranteeing that the FDR, E(V/R), is less\\nthan or equal to some pre-specified value q. In order to do this, we need\\nsome way to connect the p-values, p1, . . . , pm, from the m null hypotheses\\nto the desired FDR value, q. It turns out that a very simple procedure,\\noutlined in Algorithm 13.2, can be used to control the FDR.\\nAlgorithm 13.2 Benjamini–Hochberg Procedure to Control the FDR\\n1. Specify q, the level at which to control the FDR.\\n2. Compute p-values, p1, . . . , pm, for the m null hypotheses\\nH01, . . . , H0m.\\n3. Order the m p-values so that p(1) ≤p(2) ≤· · · ≤p(m).\\n4. Define\\nL = max{j : p(j) < qj/m}.\\n(13.10)\\n5. Reject all null hypotheses H0j for which pj ≤p(L).\\nAlgorithm 13.2 is known as the Benjamini–Hochberg procedure. The crux Benjamini–\\nHochberg\\nprocedure\\nof this procedure lies in (13.10). For example, consider again the first five\\nmanagers in the Fund dataset, presented in Table 13.3. (In this example,\\nm = 5, although typically we control the FDR in settings involving a much\\ngreater number of null hypotheses.) We see that p(1) = 0.006 < 0.05 × 1/5,\\np(2) = 0.012 < 0.05 × 2/5, p(3) = 0.601 > 0.05 × 3/5, p(4) = 0.756 >\\n0.05 × 4/5, and p(5) = 0.918 > 0.05 × 5/5. Therefore, to control the FDR\\nat 5%, we reject the null hypotheses that the first and third fund managers\\nperform no better than chance.\\nAs long as the m p-values are independent or only mildly dependent,\\nthen the Benjamini–Hochberg procedure guarantees16 that\\nFDR ≤q.\\nIn other words, this procedure ensures that, on average, no more than a\\nfraction q of the rejected null hypotheses are false positives. Remarkably,\\nthis holds regardless of how many null hypotheses are true, and regardless\\nof the distribution of the p-values for the null hypotheses that are false.\\nTherefore, the Benjamini–Hochberg procedure gives us a very easy way to\\ndetermine, given a set of m p-values, which null hypotheses to reject in\\norder to control the FDR at any pre-specified level q.\\n16However, the proof is well beyond the scope of this book. 576\\n13. Multiple Testing\\nα = 0.05\\nα = 0.1\\nα = 0.3\\n1\\n5\\n50\\n500\\n1e−05\\n1e−03\\n1e−01\\nIndex\\nP−Value\\n1\\n5\\n50\\n500\\n1e−05\\n1e−03\\n1e−01\\nIndex\\nP−Value\\n1\\n5\\n50\\n500\\n1e−05\\n1e−03\\n1e−01\\nIndex\\nP−Value\\nFIGURE 13.6. Each panel displays the same set of m = 2,000 ordered p-values\\nfor the Fund data. The green lines indicate the p-value thresholds corresponding\\nto FWER control, via the Bonferroni procedure, at levels α = 0.05 (left), α = 0.1\\n(center), and α = 0.3 (right). The orange lines indicate the p-value thresholds\\ncorresponding to FDR control, via Benjamini–Hochberg, at levels q = 0.05 (left),\\nq = 0.1 (center), and q = 0.3 (right). When the FDR is controlled at level q = 0.1,\\n146 null hypotheses are rejected (center); the corresponding p-values are shown\\nin blue. When the FDR is controlled at level q = 0.3, 279 null hypotheses are\\nrejected (right); the corresponding p-values are shown in blue.\\nThere is a fundamental difference between the Bonferroni procedure of\\nSection 13.3.2 and the Benjamini–Hochberg procedure. In the Bonferroni\\nprocedure, in order to control the FWER for m null hypotheses at level\\nα, we must simply reject null hypotheses for which the p-value is below\\nα/m. This threshold of α/m does not depend on anything about the data\\n(beyond the value of m), and certainly does not depend on the p-values\\nthemselves. By contrast, the rejection threshold used in the Benjamini–\\nHochberg procedure is more complicated: we reject all null hypotheses for\\nwhich the p-value is less than or equal to the Lth smallest p-value, where\\nL is itself a function of all m p-values, as in (13.10). Therefore, when con-\\nducting the Benjamini–Hochberg procedure, we cannot plan out in advance\\nwhat threshold we will use to reject p-values; we need to first see our data.\\nFor instance, in the abstract, there is no way to know whether we will reject\\na null hypothesis corresponding to a p-value of 0.01 when using an FDR\\nthreshold of 0.1 with m = 100; the answer depends on the values of the\\nother m −1 p-values. This property of the Benjamini–Hochberg procedure\\nis shared by the Holm procedure, which also involves a data-dependent\\np-value threshold.\\nFigure 13.6 displays the results of applying the Bonferroni and Benjamini–\\nHochberg procedures on the Fund data set, using the full set of m = 2,000\\nfund managers, of which the first five were displayed in Table 13.3. When\\nthe FWER is controlled at level 0.3 using Bonferroni, only one null hypoth-\\nesis is rejected; that is, we can conclude only that a single fund manager is\\nbeating the market. This is despite the fact that a substantial portion of 13.5 A Re-Sampling Approach to p-Values and False Discovery Rates\\n577\\nthe m = 2,000 fund managers appear to have beaten the market without\\nperforming correction for multiple testing — for instance, 13 of them have\\np-values below 0.001. By contrast, when the FDR is controlled at level 0.3,\\nwe can conclude that 279 fund managers are beating the market: we expect\\nthat no more than around 279×0.3 = 83.7 of these fund managers had good\\nperformance only due to chance. Thus, we see that FDR control is much\\nmilder — and more powerful — than FWER control, in the sense that it\\nallows us to reject many more null hypotheses, with a cost of substantially\\nmore false positives.\\nThe Benjamini–Hochberg procedure has been around since the mid-\\n1990s. While a great many papers have been published since then proposing\\nalternative approaches for FDR control that can perform better in partic-\\nular scenarios, the Benjamini–Hochberg procedure remains a very useful\\nand widely-applicable approach.\\n13.5\\nA Re-Sampling Approach to p-Values and\\nFalse Discovery Rates\\nThus far, the discussion in this chapter has assumed that we are interested\\nin testing a particular null hypothesis H0 using a test statistic T, which\\nhas some known (or assumed) distribution under H0, such as a normal\\ndistribution, a t-distribution, a χ2-distribution, or an F-distribution. This\\nis referred to as the theoretical null distribution. We typically rely upon theoretical\\nnull\\ndistribution\\nthe availability of a theoretical null distribution in order to obtain a p-\\nvalue associated with our test statistic. Indeed, for most of the types of\\nnull hypotheses that we might be interested in testing, a theoretical null\\ndistribution is available, provided that we are willing to make stringent\\nassumptions about our data.\\nHowever, if our null hypothesis H0 or test statistic T is somewhat un-\\nusual, then it may be the case that no theoretical null distribution is avail-\\nable. Alternatively, even if a theoretical null distribution exists, then we\\nmay be wary of relying upon it, perhaps because some assumption that is\\nrequired for it to hold is violated. For instance, maybe the sample size is\\ntoo small.\\nIn this section, we present a framework for performing inference in this\\nsetting, which exploits the availability of fast computers in order to approx-\\nimate the null distribution of T, and thereby to obtain a p-value. While this\\nframework is very general, it must be carefully instantiated for a specific\\nproblem of interest. Therefore, in what follows, we consider a specific ex-\\nample in which we wish to test whether the means of two random variables\\nare equal, using a two-sample t-test.\\nThe discussion in this section is more challenging than the preceding\\nsections in this chapter, and can be safely skipped by a reader who is\\ncontent to use the theoretical null distribution to compute p-values for his\\nor her test statistics. 578\\n13. Multiple Testing\\n13.5.1\\nA Re-Sampling Approach to the p-Value\\nWe return to the example of Section 13.1.1, in which we wish to test whether\\nthe mean of a random variable X equals the mean of a random variable Y ,\\ni.e. H0 : E(X) = E(Y ), against the alternative Ha : E(X) ̸= E(Y ). Given\\nnX independent observations from X and nY independent observations\\nfrom Y , the two-sample t-statistic takes the form\\nT =\\nˆµX −ˆµY\\ns\\nG\\n1\\nnX +\\n1\\nnY\\n(13.11)\\nwhere ˆµX =\\n1\\nnX\\n)nX\\ni=1 xi, ˆµY =\\n1\\nnY\\n)nY\\ni=1 yi, s =\\nG\\n(nX−1)s2\\nX+(nY −1)s2\\nY\\nnX+nY −2\\n,\\nand s2\\nX and s2\\nY are unbiased estimators of the variances in the two groups.\\nA large (absolute) value of T provides evidence against H0.\\nIf nX and nY are large, then T in (13.11) approximately follows a N(0, 1)\\ndistribution. But if nX and nY are small, then in the absence of a strong\\nassumption about the distribution of X and Y , we do not know the the-\\noretical null distribution of T.17 In this case, it turns out that we can\\napproximate the null distribution of T using a re-sampling approach, or re-sampling\\nmore specifically, a permutation approach.\\npermutation\\nTo do this, we conduct a thought experiment. If H0 holds, so that E(X) =\\nE(Y ), and we make the stronger assumption that the distributions of X\\nand Y are the same, then the distribution of T is invariant under swapping\\nobservations of X with observations of Y . That is, if we randomly swap\\nsome of the observations in X with the observations in Y , then the test\\nstatistic T in (13.11) computed based on this swapped data has the same\\ndistribution as T based on the original data. This is true only if H0 holds,\\nand the distributions of X and Y are the same.\\nThis suggests that in order to approximate the null distribution of T,\\nwe can take the following approach. We randomly permute the nX + nY\\nobservations B times, for some large value of B, and each time we compute\\n(13.11). We let T ∗1, . . . , T ∗B denote the values of (13.11) on the permuted\\ndata. These can be viewed as an approximation of the null distribution\\nof T under H0. Recall that by definition, a p-value is the probability of\\nobserving a test statistic at least this extreme under H0. Therefore, to\\ncompute a p-value for T, we can simply compute\\np-value =\\n)B\\nb=1 1(|T ∗b|≥|T |)\\nB\\n,\\n(13.12)\\nthe fraction of permuted datasets for which the value of the test statistic\\nis at least as extreme as the value observed on the original data. This\\nprocedure is summarized in Algorithm 13.3.\\n17If we assume that X and Y are normally distributed, then T in (13.11) follows a\\nt-distribution with nX + nY −2 degrees of freedom under H0. However, in practice, the\\ndistribution of random variables is rarely known, and so it can be preferable to perform\\na re-sampling approach instead of making strong and unjustified assumptions. If the\\nresults of the re-sampling approach disagree with the results of assuming a theoretical\\nnull distribution, then the results of the re-sampling approach are more trustworthy. 13.5 A Re-Sampling Approach to p-Values and False Discovery Rates\\n579\\nAlgorithm 13.3 Re-Sampling p-Value for a Two-Sample t-Test\\n1. Compute T, defined in (13.11), on the original data x1, . . . , xnX and\\ny1, . . . , ynY .\\n2. For b = 1, . . . , B, where B is a large number (e.g. B = 10,000):\\n(a) Permute the nX + nY observations at random. Call the first nX\\npermuted observations x∗\\n1, . . . , x∗\\nnX, and call the remaining nY\\nobservations y∗\\n1, . . . , y∗\\nnY .\\n(b) Compute (13.11) on the permuted data x∗\\n1, . . . , x∗\\nnX\\nand\\ny∗\\n1, . . . , y∗\\nnY , and call the result T ∗b.\\n3. The p-value is given by\\n!B\\nb=1 1(|T ∗b|≥|T |)\\nB\\n.\\nWe try out this procedure on the Khan dataset, which consists of expres-\\nsion measurements for 2,308 genes in four sub-types of small round blood\\ncell tumors, a type of cancer typically seen in children. This dataset is part\\nof the ISLR2 package. We restrict our attention to the two sub-types for\\nwhich the most observations are available: rhabdomyosarcoma (nX = 29)\\nand Burkitt’s lymphoma (nY = 25).\\nA two-sample t-test for the null hypothesis that the 11th gene’s mean\\nexpression values are equal in the two groups yields T = −2.09. Using\\nthe theoretical null distribution, which is a t52 distribution (since nX +\\nnY −2 = 52), we obtain a p-value of 0.041. (Note that a t52 distribution\\nis virtually indistinguishable from a N(0, 1) distribution.) If we instead\\napply Algorithm 13.3 with B = 10,000, then we obtain a p-value of 0.042.\\nFigure 13.7 displays the theoretical null distribution, the re-sampling null\\ndistribution, and the actual value of the test statistic (T = −2.09) for this\\ngene. In this example, we see very little difference between the p-values\\nobtained using the theoretical null distribution and the re-sampling null\\ndistribution.\\nBy contrast, Figure 13.8 shows an analogous set of results for the 877th\\ngene. In this case, there is a substantial difference between the theoretical\\nand re-sampling null distributions, which results in a difference between\\ntheir p-values.\\nIn general, in settings with a smaller sample size or a more skewed data\\ndistribution (so that the theoretical null distribution is less accurate), the\\ndifference between the re-sampling and theoretical p-values will tend to\\nbe more pronounced. In fact, the substantial difference between the re-\\nsampling and theoretical null distributions in Figure 13.8 is due to the\\nfact that a single observation in the 877th gene is very far from the other\\nobservations, leading to a very skewed distribution.\\n13.5.2\\nA Re-Sampling Approach to the False Discovery Rate\\nNow, suppose that we wish to control the FDR for m null hypotheses,\\nH01, . . . , H0m, in a setting in which either no theoretical null distribution\\nis available, or else we simply prefer to avoid the use of a theoretical null 580\\n13. Multiple Testing\\nNull Distribution of Test Statistic for 11th Gene\\n−4\\n−2\\n0\\n2\\n4\\nT=−2.0936\\nFIGURE 13.7. The 11th gene in the Khan dataset has a test statistic of\\nT = −2.09. Its theoretical and re-sampling null distributions are almost identical.\\nThe theoretical p-value equals 0.041 and the re-sampling p-value equals 0.042.\\nFIGURE 13.8. The 877th gene in the Khan dataset has a test statistic of\\nT = −0.57. Its theoretical and re-sampling null distributions are quite different.\\nThe theoretical p-value equals 0.571, and the re-sampling p-value equals 0.673.\\ndistribution. As in Section 13.5.1, we make use of a two-sample t-statistic for\\neach hypothesis, leading to the test statistics T1, . . . , Tm. We could simply\\ncompute a p-value for each of the m null hypotheses, as in Section 13.5.1,\\nand then apply the Benjamini–Hochberg procedure of Section 13.4.2 to\\nthese p-values. However, it turns out that we can do this in a more direct\\nway, without even needing to compute p-values.\\nRecall from Section 13.4 that the FDR is defined as E(V/R), using the\\nnotation in Table 13.2. In order to estimate the FDR via re-sampling, we\\nfirst make the following approximation:\\nFDR = E\\n*V\\nR\\n+\\n≈E(V )\\nR\\n.\\n(13.13)\\nNow suppose we reject any null hypothesis for which the test statistic\\nexceeds c in absolute value. Then computing R in the denominator on the\\nright-hand side of (13.13) is straightforward: R = )m\\nj=1 1(|Tj|≥c).\\nNull Distribution of Test Statistic for 877th Gene\\n−4\\n−2\\n0\\n2\\n4\\nT=−0.5696 13.5 A Re-Sampling Approach to p-Values and False Discovery Rates\\n581\\nHowever, the numerator E(V ) on the right-hand side of (13.13) is more\\nchallenging. This is the expected number of false positives associated with\\nrejecting any null hypothesis for which the test statistic exceeds c in abso-\\nlute value. At the risk of stating the obvious, estimating V is challenging\\nbecause we do not know which of H01, . . . , H0m are really true, and so we\\ndo not know which rejected hypotheses are false positives. To overcome this\\nproblem, we take a re-sampling approach, in which we simulate data under\\nH01, . . . , H0m, and then compute the resulting test statistics. The number\\nof re-sampled test statistics that exceed c provides an estimate of V .\\nIn greater detail, in the case of a two-sample t-statistic (13.11) for each\\nof the null hypotheses H01, . . . , H0m, we can estimate E(V ) as follows. Let\\nx(j)\\n1 , . . . , x(j)\\nnX and y(j)\\n1 , . . . , y(j)\\nnY denote the data associated with the jth\\nnull hypothesis, j = 1, . . . , m. We permute these nX + nY observations at\\nrandom, and then compute the t-statistic on the permuted data. For this\\npermuted data, we know that all of the null hypotheses H01, . . . , H0m hold;\\ntherefore, the number of permuted t-statistics that exceed the threshold c in\\nabsolute value provides an estimate for E(V ). This estimate can be further\\nimproved by repeating the permutation process B times, for a large value\\nof B, and averaging the results.\\nAlgorithm 13.4 details this procedure.18 It provides what is known as a\\nplug-in estimate of the FDR, because the approximation in (13.13) allows us\\nto estimate the FDR by plugging R into the denominator and an estimate\\nfor E(V ) into the numerator.\\nWe apply the re-sampling approach to the FDR from Algorithm 13.4,\\nas well as the Benjamini–Hochberg approach from Algorithm 13.2 using\\ntheoretical p-values, to the m = 2,308 genes in the Khan dataset. Results are\\nshown in Figure 13.9. We see that for a given number of rejected hypotheses,\\nthe estimated FDRs are almost identical for the two methods.\\nWe began this section by noting that in order to control the FDR for m\\nhypothesis tests using a re-sampling approach, we could simply compute m\\nre-sampling p-values as in Section 13.5.1, and then apply the Benjamini–\\nHochberg procedure of Section 13.4.2 to these p-values. It turns out that if\\nwe define the jth re-sampling p-value as\\npj =\\n)m\\nj′=1\\n)B\\nb=1 1(|T ∗b\\nj′ |≥|Tj|)\\nBm\\n(13.14)\\nfor j = 1, . . . , m, instead of as in (13.12), then applying the Benjamini–\\nHochberg procedure to these re-sampled p-values is exactly equivalent to\\nAlgorithm 13.4. Note that (13.14) is an alternative to (13.12) that pools\\nthe information across all m hypothesis tests in approximating the null\\ndistribution.\\n13.5.3\\nWhen Are Re-Sampling Approaches Useful?\\nIn Sections 13.5.1 and 13.5.2, we considered testing null hypotheses of the\\nform H0 : E(X) = E(Y ) using a two-sample t-statistic (13.11), for which we\\n18To implement Algorithm 13.4 efficiently, the same set of permutations in Step 2(b)i.\\nshould be used for all m null hypotheses. 582\\n13. Multiple Testing\\nAlgorithm 13.4 Plug-In FDR for a Two-Sample T-Test\\n1. Select a threshold c, where c > 0.\\n2. For j = 1, . . . , m:\\n(a) Compute T (j), the two-sample t-statistic (13.11) for the null\\nhypothesis H0j on the basis of the original data, x(j)\\n1 , . . . , x(j)\\nnX\\nand y(j)\\n1 , . . . , y(j)\\nnY .\\n(b) For b = 1, . . . , B, where B is a large number (e.g. B = 10,000):\\ni. Permute the nX +nY observations at random. Call the first\\nnX observations x∗(j)\\n1\\n, . . . , x∗(j)\\nnX , and call the remaining ob-\\nservations y∗(j)\\n1\\n, . . . , y∗(j)\\nnY .\\nii. Compute (13.11) on the permuted data x∗(j)\\n1\\n, . . . , x∗(j)\\nnX and\\ny∗(j)\\n1\\n, . . . , y∗(j)\\nnY , and call the result T (j),∗b.\\n3. Compute R = )m\\nj=1 1(|T (j)|≥c).\\n4. Compute WV =\\n!B\\nb=1\\n!m\\nj=1 1(|T (j),∗b|≥c)\\nB\\n.\\n5. The estimated FDR associated with the threshold c is WV /R.\\napproximated the null distribution via a re-sampling approach. We saw that\\nusing the re-sampling approach gave us substantially different results from\\nusing the theoretical p-value approach in Figure 13.8, but not in Figure 13.7.\\nIn general, there are two settings in which a re-sampling approach is\\nparticularly useful:\\n1. Perhaps no theoretical null distribution is available. This may be the\\ncase if you are testing an unusual null hypothesis H0, or using an\\nunsual test statistic T.\\n2. Perhaps a theoretical null distribution is available, but the assump-\\ntions required for its validity do not hold. For instance, the two-\\nsample t-statistic in (13.11) follows a tnX+nY −2 distribution only if\\nthe observations are normally distributed. Furthermore, it follows a\\nN(0, 1) distribution only if nX and nY are quite large. If the data are\\nnon-normal and nX and nY are small, then p-values that make use\\nof the theoretical null distribution will not be valid (i.e. they will not\\nproperly control the Type I error).\\nIn general, if you can come up with a way to re-sample or permute\\nyour observations in order to generate data that follow the null distribu-\\ntion, then you can compute p-values or estimate the FDR using variants\\nof Algorithms 13.3 and 13.4. In many real-world settings, this provides a\\npowerful tool for hypothesis testing when no out-of-box hypothesis tests are\\navailable, or when the key assumptions underlying those out-of-box tests\\nare violated. 13.6 Lab: Multiple Testing\\n583\\n0\\n500\\n1000\\n1500\\n2000\\n0.0\\n0.2\\n0.4\\n0.6\\n0.8\\n1.0\\nNumber of Rejections\\nFalse Discovery Rate\\nFIGURE 13.9. For j = 1, . . . , m = 2,308, we tested the null hypothesis that for\\nthe jth gene in the Khan dataset, the mean expression in Burkitt’s lymphoma equals\\nthe mean expression in rhabdomyosarcoma. For each value of k from 1 to 2,308, the\\ny-axis displays the estimated FDR associated with rejecting the null hypotheses\\ncorresponding to the k smallest p-values. The orange dashed curve shows the\\nFDR obtained using the Benjamini–Hochberg procedure, whereas the blue solid\\ncurve shows the FDR obtained using the re-sampling approach of Algorithm 13.4,\\nwith B = 10,000. There is very little difference between the two FDR estimates.\\nAccording to either estimate, rejecting the null hypothesis for the 500 genes with\\nthe smallest p-values corresponds to an FDR of around 17.7%.\\n13.6\\nLab: Multiple Testing\\nWe include our usual imports seen in earlier labs.\\nIn [1]: import numpy as np\\nimport pandas as pd\\nimport matplotlib.pyplot as plt\\nimport statsmodels.api as sm\\nfrom ISLP import load_data\\nWe also collect the new imports needed for this lab.\\nIn [2]: from scipy.stats import \\\\\\n(ttest_1samp ,\\nttest_rel,\\nttest_ind,\\nt as t_dbn)\\nfrom statsmodels.stats.multicomp import \\\\\\npairwise_tukeyhsd\\nfrom statsmodels.stats.multitest import \\\\\\nmultipletests as mult_test\\n13.6.1\\nReview of Hypothesis Tests\\nWe begin by performing some one-sample t-tests. First we create 100 vari-\\nables, each consisting of 10 observations. The first 50 variables have mean\\n0.5 and variance 1, while the others have mean 0 and variance 1.\\nIn [3]: rng = np.random.default_rng(12)\\nX = rng.standard_normal((10, 100))\\ntrue_mean = np.array([0.5]*50 + [0]*50)\\nX += true_mean[None,:] 584\\n13. Multiple Testing\\nTo begin, we use ttest_1samp() from the scipy.stats module to test H0 : ttest_1samp()\\nµ1 = 0, the null hypothesis that the first variable has mean zero.\\nIn [4]: result = ttest_1samp(X[:,0], 0)\\nresult.pvalue\\nOut[4]: 0.931\\nThe p-value comes out to 0.931, which is not low enough to reject the null\\nhypothesis at level α = 0.05. In this case, µ1 = 0.5, so the null hypothesis\\nis false. Therefore, we have made a Type II error by failing to reject the\\nnull hypothesis when the null hypothesis is false.\\nWe now test H0,j : µj = 0 for j = 1, . . . , 100. We compute the 100 p-\\nvalues, and then construct a vector recording whether the jth p-value is\\nless than or equal to 0.05, in which case we reject H0j, or greater than 0.05,\\nin which case we do not reject H0j, for j = 1, . . . , 100.\\nIn [5]: p_values = np.empty(100)\\nfor i in range(100):\\np_values[i] = ttest_1samp(X[:,i], 0).pvalue\\ndecision = pd.cut(p_values,\\n[0, 0.05, 1],\\nlabels=[\\'Reject H0\\',\\n\\'Do not reject H0\\'])\\ntruth = pd.Categorical(true_mean == 0,\\ncategories=[True, False],\\nordered=True)\\nSince this is a simulated data set, we can create a 2 × 2 table similar to\\nTable 13.2.\\nIn [6]: pd.crosstab(decision,\\ntruth,\\nrownames=[\\'Decision\\'],\\ncolnames=[\\'H0\\'])\\nOut[6]:\\nH0\\nTrue\\nFalse\\nDecision\\nReject H0\\n5\\n15\\nDo not reject H0\\n45\\n35\\nTherefore, at level α = 0.05, we reject 15 of the 50 false null hypotheses,\\nand we incorrectly reject 5 of the true null hypotheses. Using the notation\\nfrom Section 13.3, we have V = 5, S = 15, U = 45 and W = 35. We have\\nset α = 0.05, which means that we expect to reject around 5% of the true\\nnull hypotheses. This is in line with the 2 × 2 table above, which indicates\\nthat we rejected V = 5 of the 50 true null hypotheses.\\nIn the simulation above, for the false null hypotheses, the ratio of the\\nmean to the standard deviation was only 0.5/1 = 0.5. This amounts to\\nquite a weak signal, and it resulted in a high number of Type II errors.\\nLet’s instead simulate data with a stronger signal, so that the ratio of the\\nmean to the standard deviation for the false null hypotheses equals 1. We\\nmake only 10 Type II errors. 13.6 Lab: Multiple Testing\\n585\\nIn [7]: true_mean = np.array([1]*50 + [0]*50)\\nX = rng.standard_normal((10, 100))\\nX += true_mean[None,:]\\nfor i in range(100):\\np_values[i] = ttest_1samp(X[:,i], 0).pvalue\\ndecision = pd.cut(p_values,\\n[0, 0.05, 1],\\nlabels=[\\'Reject H0\\',\\n\\'Do not reject H0\\'])\\ntruth = pd.Categorical(true_mean == 0,\\ncategories=[True, False],\\nordered=True)\\npd.crosstab(decision,\\ntruth,\\nrownames=[\\'Decision\\'],\\ncolnames=[\\'H0\\'])\\nOut[7]:\\nH0\\nTrue\\nFalse\\nDecision\\nReject H0\\n2\\n40\\nDo not reject H0\\n48\\n10\\n13.6.2\\nFamily-Wise Error Rate\\nRecall from (13.5) that if the null hypothesis is true for each of m inde-\\npendent hypothesis tests, then the FWER is equal to 1 −(1 −α)m. We\\ncan use this expression to compute the FWER for m = 1, . . . , 500 and\\nα = 0.05, 0.01, and 0.001. We plot the FWER for these values of α in order\\nto reproduce Figure 13.2.\\nIn [8]: m = np.linspace(1, 501)\\nfig, ax = plt.subplots()\\n[ax.plot(m,\\n1 - (1 - alpha)**m,\\nlabel=r\\'$\\\\alpha=%s$\\' % str(alpha))\\nfor alpha in [0.05, 0.01, 0.001]]\\nax.set_xscale(\\'log\\')\\nax.set_xlabel(\\'Number of Hypotheses\\')\\nax.set_ylabel(\\'Family-Wise Error Rate\\')\\nax.legend()\\nax.axhline(0.05, c=\\'k\\', ls=\\'--\\');\\nAs discussed previously, even for moderate values of m such as 50, the\\nFWER exceeds 0.05 unless α is set to a very low value, such as 0.001. Of\\ncourse, the problem with setting α to such a low value is that we are likely\\nto make a number of Type II errors: in other words, our power is very low.\\nWe now conduct a one-sample t-test for each of the first five managers\\nin the Fund dataset, in order to test the null hypothesis that the jth fund\\nmanager’s mean return equals zero, H0,j : µj = 0.\\nIn [9]: Fund = load_data(\\'Fund\\')\\nfund_mini = Fund.iloc[:,:5]\\nfund_mini_pvals = np.empty(5)\\nfor i in range(5): 586\\n13. Multiple Testing\\nfund_mini_pvals[i] = ttest_1samp(fund_mini.iloc[:,i], 0).pvalue\\nfund_mini_pvals\\nOut[9]: array([0.006, 0.918, 0.012, 0.601, 0.756])\\nThe p-values are low for Managers One and Three, and high for the other\\nthree managers. However, we cannot simply reject H0,1 and H0,3, since\\nthis would fail to account for the multiple testing that we have performed.\\nInstead, we will conduct Bonferroni’s method and Holm’s method to control\\nthe FWER.\\nTo do this, we use the multipletests() function from the statsmodels multiple-\\ntests()\\nmodule (abbreviated to mult_test()). Given the p-values, for methods like\\nHolm and Bonferroni the function outputs adjusted p-values, which can be adjusted\\np-values\\nthought of as a new set of p-values that have been corrected for multiple\\ntesting. If the adjusted p-value for a given hypothesis is less than or equal\\nto α, then that hypothesis can be rejected while maintaining a FWER of\\nno more than α. In other words, for such methods, the adjusted p-values\\nresulting from the multipletests() function can simply be compared to\\nthe desired FWER in order to determine whether or not to reject each\\nhypothesis. We will later see that we can use the same function to control\\nFDR as well.\\nThe mult_test() function takes p-values and a method argument, as well\\nas an optional alpha argument. It returns the decisions (reject below) as\\nwell as the adjusted p-values (bonf).\\nIn [10]: reject, bonf = mult_test(fund_mini_pvals , method = \"bonferroni\")[:2]\\nreject\\nOut[10]: array([ True, False, False, False, False])\\nThe p-values bonf are simply the fund_mini_pvalues multiplied by 5 and\\ntruncated to be less than or equal to 1.\\nIn [11]: bonf, np.minimum(fund_mini_pvals * 5, 1)\\nOut[11]: (array([0.03, 1.\\n, 0.06, 1.\\n, 1.\\n]),\\narray([0.03, 1.\\n, 0.06, 1.\\n, 1.\\n]))\\nTherefore, using Bonferroni’s method, we are able to reject the null hy-\\npothesis only for Manager One while controlling FWER at 0.05.\\nBy contrast, using Holm’s method, the adjusted p-values indicate that\\nwe can reject the null hypotheses for Managers One and Three at a FWER\\nof 0.05.\\nIn [12]: mult_test(fund_mini_pvals , method = \"holm\", alpha=0.05)[:2]\\nOut[12]: (array([ True, False,\\nTrue, False, False]),\\narray([0.03, 1.\\n, 0.05, 1.\\n, 1.\\n]))\\nAs discussed previously, Manager One seems to perform particularly well,\\nwhereas Manager Two has poor performance. 13.6 Lab: Multiple Testing\\n587\\nIn [13]: fund_mini.mean()\\nOut[13]: Manager1\\n3.0\\nManager2\\n-0.1\\nManager3\\n2.8\\nManager4\\n0.5\\nManager5\\n0.3\\ndtype: float64\\nIs there evidence of a meaningful difference in performance between these\\ntwo managers? We can check this by performing a paired t-test using the paired t-test\\nttest_rel() function from scipy.stats:\\nttest_rel()\\nIn [14]: ttest_rel(fund_mini[\\'Manager1\\'],\\nfund_mini[\\'Manager2\\']).pvalue\\nOut[14]: 0.038\\nThe test results in a p-value of 0.038, suggesting a statistically significant\\ndifference.\\nHowever, we decided to perform this test only after examining the data\\nand noting that Managers One and Two had the highest and lowest mean\\nperformances. In a sense, this means that we have implicitly performed\\n\\'5\\n2\\n(\\n= 5(5 −1)/2 = 10 hypothesis tests, rather than just one, as discussed\\nin Section 13.3.2. Hence, we use the pairwise_tukeyhsd() function from pairwise_\\ntukeyhsd()\\nstatsmodels.stats.multicomp to apply Tukey’s method in order to adjust\\nfor multiple testing. This function takes as input a fitted ANOVA regres- ANOVA\\nsion model, which is essentially just a linear regression in which all of the\\npredictors are qualitative. In this case, the response consists of the monthly\\nexcess returns achieved by each manager, and the predictor indicates the\\nmanager to which each return corresponds.\\nIn [15]: returns = np.hstack([fund_mini.iloc[:,i] for i in range(5)])\\nmanagers = np.hstack([[i+1]*50 for i in range(5)])\\ntukey = pairwise_tukeyhsd(returns, managers)\\nprint(tukey.summary())\\nMultiple Comparison of Means - Tukey HSD, FWER=0.05\\n===================================================\\ngroup1 group2 meandiff p-adj\\nlower\\nupper\\nreject\\n---------------------------------------------------\\n1\\n2\\n-3.1 0.1862 -6.9865 0.7865\\nFalse\\n1\\n3\\n-0.2 0.9999 -4.0865 3.6865\\nFalse\\n1\\n4\\n-2.5 0.3948 -6.3865 1.3865\\nFalse\\n1\\n5\\n-2.7 0.3152 -6.5865 1.1865\\nFalse\\n2\\n3\\n2.9 0.2453 -0.9865 6.7865\\nFalse\\n2\\n4\\n0.6 0.9932 -3.2865 4.4865\\nFalse\\n2\\n5\\n0.4 0.9986 -3.4865 4.2865\\nFalse\\n3\\n4\\n-2.3\\n0.482 -6.1865 1.5865\\nFalse\\n3\\n5\\n-2.5 0.3948 -6.3865 1.3865\\nFalse\\n4\\n5\\n-0.2 0.9999 -4.0865 3.6865\\nFalse\\n---------------------------------------------------\\nThe pairwise_tukeyhsd() function provides confidence intervals for the\\ndifference between each pair of managers (lower and upper), as well as a 588\\n13. Multiple Testing\\nFIGURE 13.10. 95% confidence intervals for each manager on the Fund data,\\nusing Tukey’s method to adjust for multiple testing. All of the confidence intervals\\noverlap, so none of the differences among managers are statistically significant\\nwhen controlling FWER at level 0.05.\\np-value. All of these quantities have been adjusted for multiple testing.\\nNotice that the p-value for the difference between Managers One and Two\\nhas increased from 0.038 to 0.186, so there is no longer clear evidence of\\na difference between the managers’ performances. We can plot the confi-\\ndence intervals for the pairwise comparisons using the plot_simultaneous()\\nmethod of tukey. Any pair of intervals that don’t overlap indicates a sig-\\nnificant difference at the nominal level of 0.05. In this case, no differences\\nare considered significant as reported in the table above.\\nIn [16]: fig, ax = plt.subplots(figsize=(8,8))\\ntukey.plot_simultaneous(ax=ax);\\nThe result can be seen19 in Figure 13.10.\\n13.6.3\\nFalse Discovery Rate\\nNow we perform hypothesis tests for all 2,000 fund managers in the Fund\\ndataset. We perform a one-sample t-test of H0,j : µj = 0, which states that\\nthe jth fund manager’s mean return is zero.\\nIn [17]: fund_pvalues = np.empty(2000)\\nfor i, manager in enumerate(Fund.columns):\\nfund_pvalues[i] = ttest_1samp(Fund[manager], 0).pvalue\\nThere are far too many managers to consider trying to control the\\nFWER. Instead, we focus on controlling the FDR: that is, the expected\\nfraction of rejected null hypotheses that are actually false positives. The\\n19Traditionally this plot shows intervals for each paired difference. With many groups\\nit is more convenient and equivalent to display one interval per group, as is done here.\\nBy “differencing” all pairs of intervals displayed here you recover the traditional plot. 13.6 Lab: Multiple Testing\\n589\\nmultipletests() function (abbreviated mult_test()) can be used to carry\\nout the Benjamini–Hochberg procedure.\\nIn [18]: fund_qvalues = mult_test(fund_pvalues , method = \"fdr_bh\")[1]\\nfund_qvalues[:10]\\nOut[18]: array([0.09, 0.99, 0.12, 0.92, 0.96, 0.08, 0.08, 0.08, 0.08,\\n0.08])\\nThe q-values output by the Benjamini–Hochberg procedure can be inter- q-values\\npreted as the smallest FDR threshold at which we would reject a particular\\nnull hypothesis. For instance, a q-value of 0.1 indicates that we can reject\\nthe corresponding null hypothesis at an FDR of 10% or greater, but that\\nwe cannot reject the null hypothesis at an FDR below 10%.\\nIf we control the FDR at 10%, then for how many of the fund managers\\ncan we reject H0,j : µj = 0?\\nIn [19]: (fund_qvalues <= 0.1).sum()\\nOut[19]: 146\\nWe find that 146 of the 2,000 fund managers have a q-value below 0.1;\\ntherefore, we are able to conclude that 146 of the fund managers beat the\\nmarket at an FDR of 10%. Only about 15 (10% of 146) of these fund\\nmanagers are likely to be false discoveries.\\nBy contrast, if we had instead used Bonferroni’s method to control the\\nFWER at level α = 0.1, then we would have failed to reject any null\\nhypotheses!\\nIn [20]: (fund_pvalues <= 0.1 / 2000).sum()\\nOut[20]: 0\\nFigure 13.6 displays the ordered p-values, p(1) ≤p(2) ≤· · · ≤p(2000), for\\nthe Fund dataset, as well as the threshold for rejection by the Benjamini–\\nHochberg procedure. Recall that the Benjamini–Hochberg procedure iden-\\ntifies the largest p-value such that p(j) < qj/m, and rejects all hypotheses\\nfor which the p-value is less than or equal to p(j). In the code below, we im-\\nplement the Benjamini–Hochberg procedure ourselves, in order to illustrate\\nhow it works. We first order the p-values. We then identify all p-values that\\nsatisfy p(j) < qj/m (sorted_set_). Finally, selected_ is a boolean array\\nindicating which p-values are less than or equal to the largest p-value in\\nsorted_[sorted_set_]. Therefore, selected_ indexes the p-values rejected\\nby the Benjamini–Hochberg procedure.\\nIn [21]: sorted_ = np.sort(fund_pvalues)\\nm = fund_pvalues.shape[0]\\nq = 0.1\\nsorted_set_ = np.where(sorted_ < q * np.linspace(1, m, m) / m)[0]\\nif sorted_set_.shape[0] > 0:\\nselected_ = fund_pvalues < sorted_[sorted_set_].max()\\nsorted_set_ = np.arange(sorted_set_.max())\\nelse:\\nselected_ = []\\nsorted_set_ = [] 590\\n13. Multiple Testing\\nWe now reproduce the middle panel of Figure 13.6.\\nIn [22]: fig, ax = plt.subplots()\\nax.scatter(np.arange(0, sorted_.shape[0]) + 1,\\nsorted_, s=10)\\nax.set_yscale(\\'log\\')\\nax.set_xscale(\\'log\\')\\nax.set_ylabel(\\'P-Value\\')\\nax.set_xlabel(\\'Index\\')\\nax.scatter(sorted_set_+1, sorted_[sorted_set_], c=\\'r\\', s=20)\\nax.axline((0, 0), (1,q/m), c=\\'k\\', ls=\\'--\\', linewidth=3);\\n13.6.4\\nA Re-Sampling Approach\\nHere, we implement the re-sampling approach to hypothesis testing using\\nthe Khan dataset, which we investigated in Section 13.5. First, we merge\\nthe training and testing data, which results in observations on 83 patients\\nfor 2,308 genes.\\nIn [23]: Khan = load_data(\\'Khan\\')\\nD = pd.concat([Khan[\\'xtrain\\'], Khan[\\'xtest\\']])\\nD[\\'Y\\'] = pd.concat([Khan[\\'ytrain\\'], Khan[\\'ytest\\']])\\nD[\\'Y\\'].value_counts()\\nOut[23]: 2\\n29\\n4\\n25\\n3\\n18\\n1\\n11\\nName: Y, dtype: int64\\nThere are four classes of cancer. For each gene, we compare the mean ex-\\npression in the second class (rhabdomyosarcoma) to the mean expression in\\nthe fourth class (Burkitt’s lymphoma). Performing a standard two-sample\\nt-test using ttest_ind() from scipy.stats on the 11th gene produces a ttest_ind()\\ntest-statistic of -2.09 and an associated p-value of 0.0412, suggesting mod-\\nest evidence of a difference in mean expression levels between the two cancer\\ntypes.\\nIn [24]: D2 = D[lambda df:df[\\'Y\\'] == 2]\\nD4 = D[lambda df:df[\\'Y\\'] == 4]\\ngene_11 = \\'G0011\\'\\nobservedT, pvalue = ttest_ind(D2[gene_11],\\nD4[gene_11],\\nequal_var=True)\\nobservedT, pvalue\\nOut[24]: (-2.094, 0.041)\\nHowever, this p-value relies on the assumption that under the null hy-\\npothesis of no difference between the two groups, the test statistic follows\\na t-distribution with 29 + 25 −2 = 52 degrees of freedom. Instead of us-\\ning this theoretical null distribution, we can randomly split the 54 patients\\ninto two groups of 29 and 25, and compute a new test statistic. Under the\\nnull hypothesis of no difference between the groups, this new test statis-\\ntic should have the same distribution as our original one. Repeating this 13.6 Lab: Multiple Testing\\n591\\nprocess 10,000 times allows us to approximate the null distribution of the\\ntest statistic. We compute the fraction of the time that our observed test\\nstatistic exceeds the test statistics obtained via re-sampling.\\nIn [25]: B = 10000\\nTnull = np.empty(B)\\nD_ = np.hstack([D2[gene_11], D4[gene_11]])\\nn_ = D2[gene_11].shape[0]\\nD_null = D_.copy()\\nfor b in range(B):\\nrng.shuffle(D_null)\\nttest_ = ttest_ind(D_null[:n_],\\nD_null[n_:],\\nequal_var=True)\\nTnull[b] = ttest_.statistic\\n(np.abs(Tnull) > np.abs(observedT)).mean()\\nOut[25]: 0.0398\\nThis fraction, 0.0398, is our re-sampling-based p-value. It is almost identi-\\ncal to the p-value of 0.0412 obtained using the theoretical null distribution.\\nWe can plot a histogram of the re-sampling-based test statistics in order\\nto reproduce Figure 13.7.\\nIn [26]: fig, ax = plt.subplots(figsize=(8,8))\\nax.hist(Tnull,\\nbins=100,\\ndensity=True,\\nfacecolor=\\'y\\',\\nlabel=\\'Null\\')\\nxval = np.linspace(-4.2, 4.2, 1001)\\nax.plot(xval,\\nt_dbn.pdf(xval, D_.shape[0]-2),\\nc=\\'r\\')\\nax.axvline(observedT,\\nc=\\'b\\',\\nlabel=\\'Observed\\')\\nax.legend()\\nax.set_xlabel(\"Null Distribution of Test Statistic\");\\nThe re-sampling-based null distribution is almost identical to the theoret-\\nical null distribution, which is displayed in red.\\nFinally, we implement the plug-in re-sampling FDR approach outlined\\nin Algorithm 13.4. Depending on the speed of your computer, calculating\\nthe FDR for all 2,308 genes in the Khan dataset may take a while. Hence,\\nwe will illustrate the approach on a random subset of 100 genes. For each\\ngene, we first compute the observed test statistic, and then produce 10,000\\nre-sampled test statistics. This may take a few minutes to run. If you are\\nin a rush, then you could set B equal to a smaller value (e.g. B=500).\\nIn [27]: m, B = 100, 10000\\nidx = rng.choice(Khan[\\'xtest\\'].columns, m, replace=False)\\nT_vals = np.empty(m)\\nTnull_vals = np.empty((m, B))\\nfor j in range(m):\\ncol = idx[j] 592\\n13. Multiple Testing\\nT_vals[j] = ttest_ind(D2[col],\\nD4[col],\\nequal_var=True).statistic\\nD_ = np.hstack([D2[col], D4[col]])\\nD_null = D_.copy()\\nfor b in range(B):\\nrng.shuffle(D_null)\\nttest_ = ttest_ind(D_null[:n_],\\nD_null[n_:],\\nequal_var=True)\\nTnull_vals[j,b] = ttest_.statistic\\nNext, we compute the number of rejected null hypotheses R, the esti-\\nmated number of false positives WV , and the estimated FDR, for a range\\nof threshold values c in Algorithm 13.4. The threshold values are chosen\\nusing the absolute values of the test statistics from the 100 genes.\\nIn [28]: cutoffs = np.sort(np.abs(T_vals))\\nFDRs, Rs, Vs = np.empty((3, m))\\nfor j in range(m):\\nR = np.sum(np.abs(T_vals) >= cutoffs[j])\\nV = np.sum(np.abs(Tnull_vals) >= cutoffs[j]) / B\\nRs[j] = R\\nVs[j] = V\\nFDRs[j] = V / R\\nNow, for any given FDR, we can find the genes that will be rejected.\\nFor example, with FDR controlled at 0.1, we reject 15 of the 100 null\\nhypotheses. On average, we would expect about one or two of these genes\\n(i.e. 10% of 15) to be false discoveries. At an FDR of 0.2, we can reject\\nthe null hypothesis for 28 genes, of which we expect around six to be false\\ndiscoveries.\\nThe variable idx stores which genes were included in our 100 randomly-\\nselected genes. Let’s look at the genes whose estimated FDR is less than\\n0.1.\\nIn [29]: sorted(idx[np.abs(T_vals) >= cutoffs[FDRs < 0.1].min()])\\nAt an FDR threshold of 0.2, more genes are selected, at the cost of having\\na higher expected proportion of false discoveries.\\nIn [30]: sorted(idx[np.abs(T_vals) >= cutoffs[FDRs < 0.2].min()])\\nThe next line generates Figure 13.11, which is similar to Figure 13.9,\\nexcept that it is based on only a subset of the genes.\\nIn [31]: fig, ax = plt.subplots()\\nax.plot(Rs, FDRs, \\'b\\', linewidth=3)\\nax.set_xlabel(\"Number of Rejections\")\\nax.set_ylabel(\"False Discovery Rate\"); 13.7 Exercises\\n593\\nFIGURE 13.11. The estimated false discovery rate versus the number of rejected\\nnull hypotheses, for 100 genes randomly selected from the Khan dataset.\\n13.7\\nExercises\\nConceptual\\n1. Suppose we test m null hypotheses, all of which are true. We control\\nthe Type I error for each null hypothesis at level α. For each sub-\\nproblem, justify your answer.\\n(a) In total, how many Type I errors do we expect to make?\\n(b) Suppose that the m tests that we perform are independent.\\nWhat is the family-wise error rate associated with these m tests?\\nHint: If two events A and B are independent, then Pr(A ∩B) =\\nPr(A) Pr(B).\\n(c) Suppose that m = 2, and that the p-values for the two tests are\\npositively correlated, so that if one is small then the other will\\ntend to be small as well, and if one is large then the other will\\ntend to be large. How does the family-wise error rate associated\\nwith these m = 2 tests qualitatively compare to the answer in\\n(b) with m = 2?\\nHint: First, suppose that the two p-values are perfectly correlated.\\n(d) Suppose again that m = 2, but that now the p-values for the\\ntwo tests are negatively correlated, so that if one is large then\\nthe other will tend to be small. How does the family-wise error\\nrate associated with these m = 2 tests qualitatively compare to\\nthe answer in (b) with m = 2?\\nHint: First, suppose that whenever one p-value is less than α,\\nthen the other will be greater than α. In other words, we can\\nnever reject both null hypotheses. 594\\n13. Multiple Testing\\n2. Suppose that we test m hypotheses, and control the Type I error for\\neach hypothesis at level α. Assume that all m p-values are indepen-\\ndent, and that all null hypotheses are true.\\n(a) Let the random variable Aj equal 1 if the jth null hypothesis is\\nrejected, and 0 otherwise. What is the distribution of Aj?\\n(b) What is the distribution of )m\\nj=1 Aj?\\n(c) What is the standard deviation of the number of Type I errors\\nthat we will make?\\n3. Suppose we test m null hypotheses, and control the Type I error for\\nthe jth null hypothesis at level αj, for j = 1, . . . , m. Argue that the\\nfamily-wise error rate is no greater than )m\\nj=1 αj.\\nNull Hypothesis\\np-value\\nH01\\n0.0011\\nH02\\n0.031\\nH03\\n0.017\\nH04\\n0.32\\nH05\\n0.11\\nH06\\n0.90\\nH07\\n0.07\\nH08\\n0.006\\nH09\\n0.004\\nH10\\n0.0009\\nTABLE 13.4. p-values for Exercise 4.\\n4. Suppose we test m = 10 hypotheses, and obtain the p-values shown\\nin Table 13.4.\\n(a) Suppose that we wish to control the Type I error for each null\\nhypothesis at level α = 0.05. Which null hypotheses will we\\nreject?\\n(b) Now suppose that we wish to control the FWER at level α =\\n0.05. Which null hypotheses will we reject? Justify your answer.\\n(c) Now suppose that we wish to control the FDR at level q = 0.05.\\nWhich null hypotheses will we reject? Justify your answer.\\n(d) Now suppose that we wish to control the FDR at level q = 0.2.\\nWhich null hypotheses will we reject? Justify your answer.\\n(e) Of the null hypotheses rejected at FDR level q = 0.2, approxi-\\nmately how many are false positives? Justify your answer.\\n5. For this problem, you will make up p-values that lead to a certain\\nnumber of rejections using the Bonferroni and Holm procedures.\\n(a) Give an example of five p-values (i.e. five numbers between 0 and\\n1 which, for the purpose of this problem, we will interpret as p-\\nvalues) for which both Bonferroni’s method and Holm’s method 13.7 Exercises\\n595\\nreject exactly one null hypothesis when controlling the FWER\\nat level 0.1.\\n(b) Now give an example of five p-values for which Bonferroni re-\\njects one null hypothesis and Holm rejects more than one null\\nhypothesis at level 0.1.\\n6. For each of the three panels in Figure 13.3, answer the following\\nquestions:\\n(a) How many false positives, false negatives, true positives, true\\nnegatives, Type I errors, and Type II errors result from applying\\nthe Bonferroni procedure to control the FWER at level α =\\n0.05?\\n(b) How many false positives, false negatives, true positives, true\\nnegatives, Type I errors, and Type II errors result from applying\\nthe Holm procedure to control the FWER at level α = 0.05?\\n(c) What is the false discovery proportion associated with using the\\nBonferroni procedure to control the FWER at level α = 0.05?\\n(d) What is the false discovery proportion associated with using the\\nHolm procedure to control the FWER at level α = 0.05?\\n(e) How would the answers to (a) and (c) change if we instead used\\nthe Bonferroni procedure to control the FWER at level α =\\n0.001?\\nApplied\\n7. This problem makes use of the Carseats dataset in the ISLP package.\\n(a) For each quantitative variable in the dataset besides Sales, fit\\na linear model to predict Sales using that quantitative variable.\\nReport the p-values associated with the coefficients for the vari-\\nables. That is, for each model of the form Y = β0 + β1X + ϵ,\\nreport the p-value associated with the coefficient β1. Here, Y\\nrepresents Sales and X represents one of the other quantitative\\nvariables.\\n(b) Suppose we control the Type I error at level α = 0.05 for the\\np-values obtained in (a). Which null hypotheses do we reject?\\n(c) Now suppose we control the FWER at level 0.05 for the p-values.\\nWhich null hypotheses do we reject?\\n(d) Finally, suppose we control the FDR at level 0.2 for the p-values.\\nWhich null hypotheses do we reject?\\n8. In this problem, we will simulate data from m = 100 fund managers.\\nrng = np.random.default_rng(1)\\nn, m = 20, 100\\nX = rng.normal(size=(n, m)) 596\\n13. Multiple Testing\\nThese data represent each fund manager’s percentage returns for each\\nof n = 20 months. We wish to test the null hypothesis that each\\nfund manager’s percentage returns have population mean equal to\\nzero. Notice that we simulated the data in such a way that each fund\\nmanager’s percentage returns do have population mean zero; in other\\nwords, all m null hypotheses are true.\\n(a) Conduct a one-sample t-test for each fund manager, and plot a\\nhistogram of the p-values obtained.\\n(b) If we control Type I error for each null hypothesis at level α =\\n0.05, then how many null hypotheses do we reject?\\n(c) If we control the FWER at level 0.05, then how many null hy-\\npotheses do we reject?\\n(d) If we control the FDR at level 0.05, then how many null hy-\\npotheses do we reject?\\n(e) Now suppose we “cherry-pick” the 10 fund managers who per-\\nform the best in our data. If we control the FWER for just these\\n10 fund managers at level 0.05, then how many null hypothe-\\nses do we reject? If we control the FDR for just these 10 fund\\nmanagers at level 0.05, then how many null hypotheses do we\\nreject?\\n(f) Explain why the analysis in (e) is misleading.\\nHint: The standard approaches for controlling the FWER and\\nFDR assume that all tested null hypotheses are adjusted for mul-\\ntiplicity, and that no “cherry-picking” of the smallest p-values\\nhas occurred. What goes wrong if we cherry-pick? Index\\naccuracy, 415\\nactivation, 400\\nactivation function, 401\\nadditive, 11, 94–98, 110–111\\nadditivity, 305, 306\\nadjusted R2, 87, 231, 232, 236–\\n238\\nAdvertising data set, 15, 16, 19,\\n69, 71–73, 77, 78, 80, 82,\\n83, 85, 87–90, 95, 96, 109–\\n111\\nagglomerative clustering, 525\\nAkaike information criterion, 87,\\n231, 232, 236–238\\nalternative hypothesis, 76, 559\\nanalysis of variance, 312\\nANOVA, 587\\narea under the curve, 155, 486–\\n487\\nargument, 40\\narray, 42\\nattribute, 42\\nAUC, 155\\nAuto data set, 12, 66, 98–101, 129,\\n197, 202–207, 327, 398\\nauto-correlation, 421\\nautoregression, 423\\naxes, 48\\nbackfitting, 307, 328\\nbackpropagation, 429\\nbackward stepwise selection, 87,\\n234–235\\nbag-of-n-grams, 415\\nbag-of-words, 414\\nbagging, 11, 24, 331, 343–346, 354,\\n360–361\\nBART, 343, 350, 353, 354, 362–\\n363\\nbaseline, 93, 145, 161\\nbasis function, 293–294, 296\\nBayes\\nclassifier, 35–37, 147\\ndecision boundary, 148\\nerror, 35–37\\nBayes’ theorem, 146, 250\\nBayesian, 250–251, 353\\nBayesian additive regression trees,\\n331, 343, 350, 350, 353,\\n354, 362–363\\nBayesian information criterion, 87,\\n231, 232, 236–238\\nBenjamini–Hochberg procedure, 575–\\n577\\nBernoulli distribution, 172\\nbest subset selection, 231, 246\\nbias, 31–34, 74, 90, 159, 405\\nbias-variance\\ndecomposition, 32\\n© Springer Nature Switzerland AG 2023 \\nG. James et al., An Introduction to Statistical Learning, Springer Texts in Statistics, \\nhttps://doi.org/10.1007/978-3-031-38747-0 \\n597 598\\nIndex\\ntrade-off, 31–34, 38, 111–112,\\n157, 159, 163, 164, 242,\\n254, 263, 266, 301, 336,\\n376, 385\\nbidirectional, 425\\nBikeshare data set, 12, 167–172\\nbinary, 27, 138\\nbiplot, 507, 508\\nBonferroni method, 575–577, 585\\nBoolean, 53, 176\\nboosting, 11, 24, 331, 343, 347–\\n350, 354, 361–362\\nbootstrap, 11, 201, 212–214, 343\\nBoston data set, 12, 67, 117, 122,\\n133, 199, 227, 287, 327,\\n364, 556\\nbottom-up clustering, 525\\nboxplot, 62\\nBrainCancer data set, 12, 472–\\n474, 476, 482\\nbranch, 333\\nburn-in, 352\\nC-index, 487\\nCaravan data set, 12, 184, 366\\nCarseats data set, 12, 126, 130,\\n364\\ncategorical, 2, 27\\ncensored data, 469–502\\ncensoring\\nindependent, 471\\ninterval, 471\\nleft, 471\\nmechanism, 471\\nnon-informative, 471\\nright, 471\\ntime, 470\\nchain rule, 429\\nchannel, 407\\nCIFAR100 data set, 406, 409–411,\\n448, 449\\nclassification, 2, 11, 27, 34–39, 135–\\n199, 367–382\\nerror rate, 338\\ntree, 337–341, 355–358\\nclassifier, 135\\ncluster analysis, 25–26\\nclustering, 4, 25–26, 520–535\\nagglomerative, 525\\nbottom-up, 525\\nhierarchical, 521, 525–535\\nK-means, 11, 521–524\\nCochran–Mantel–Haenszel test, 475\\ncoefficient, 71\\nCollege data set, 12, 65, 286, 328\\ncollinearity, 106–110\\nconcatenation, 41\\nconditional probability, 35\\nconfidence interval, 75–76, 90, 110,\\n292\\nconfounding, 144\\nconfusion matrix, 153, 176\\ncontinuous, 2\\ncontour, 246\\ncontour plot, 50\\ncontrast, 94\\nconvenience function, 53\\nconvolution filter, 407\\nconvolution layer, 407\\nconvolutional neural network, 406–\\n413\\ncorrelation, 79, 82–83, 530\\ncount data, 167, 170\\nCox’s proportional hazards model,\\n480, 483–486\\nCp, 87, 231, 232, 236–238\\nCredit data set, 12, 91, 92, 94,\\n97, 98, 106–109\\ncross-entropy, 405\\ncross-validation, 11, 31, 34, 201–\\n211, 231, 252, 270\\nk-fold, 206–209\\nleave-one-out, 204–206\\ncurse of dimensionality, 115, 193,\\n266\\ndata augmentation, 411\\ndata frame, 55\\nData sets\\nAdvertising, 15, 16, 19, 69,\\n71–73, 77, 78, 80, 82, 83,\\n85, 87–90, 95, 96, 109–\\n111\\nAuto, 12, 66, 98–101, 129, 197,\\n202–207, 327, 398\\nBikeshare, 12, 167–172 Index\\n599\\nBoston, 12, 67, 117, 122, 133,\\n199, 227, 287, 327, 364,\\n556\\nBrainCancer, 12, 472–474, 476,\\n482\\nCaravan, 12, 184, 366\\nCarseats, 12, 126, 130, 364\\nCIFAR100, 406, 409–411, 448,\\n449\\nCollege, 12, 65, 286, 328\\nCredit, 12, 91, 92, 94, 97, 98,\\n106–109\\nDefault, 12, 136–139, 141–\\n144, 152–156, 160, 161,\\n225, 226, 466\\nFund, 12, 567–570, 572, 575,\\n576, 585, 588, 589\\nHeart, 339, 340, 344–347, 352,\\n353, 382, 383\\nHitters, 12, 332, 333, 336,\\n338, 339, 366, 425, 426,\\n437, 446\\nIMDb, 413, 415, 416, 418, 420,\\n437, 458, 467\\nIncome, 16–18, 21–23\\nKhan, 12, 579–581, 583, 590,\\n593\\nMNIST, 402–404, 406, 430, 431,\\n441, 444, 445, 448\\nNCI60, 4, 5, 12, 546, 548–550\\nNYSE, 12, 422–424, 466, 467\\nOJ, 12, 365, 398\\nPortfolio, 12\\nPublication, 12, 482–487\\nSmarket, 2, 3, 12, 173, 184,\\n196\\nUSArrests, 12, 507, 508, 510,\\n512, 513, 515, 516, 518,\\n519\\nWage, 1, 2, 8, 9, 12, 290, 291,\\n293, 295, 297–300, 302–\\n306, 309, 315, 327\\nWeekly, 12, 196, 226\\ndata type, 42\\ndecision function, 387\\ndecision tree, 11, 331–342\\ndeep learning, 399\\nDefault data set, 12, 136–139, 141–\\n144, 152–156, 160, 161,\\n225, 226, 466\\ndegrees of freedom, 30, 266, 295,\\n296, 301\\ndendrogram, 521, 525–530\\ndensity function, 146\\ndependent variable, 15\\nderivative, 296, 300\\ndetector layer, 410\\ndeviance, 232\\ndictionary, 66\\ndimension reduction, 230, 253–262\\ndiscriminant function, 149\\ndiscriminant method, 146–161\\ndissimilarity, 530–532\\ndistance\\ncorrelation-based, 530–532, 554\\nEuclidean, 509, 522, 523, 529–\\n532\\ndouble descent, 431–435\\ndouble-exponential distribution, 251\\ndropout, 406, 431\\ndummy variable, 91–94, 138, 142,\\n292\\nearly stopping, 430\\neffective degrees of freedom, 301\\neigen decomposition, 506, 516\\nelbow, 548\\nembedding, 418\\nembedding layer, 419\\nensemble, 343–354\\nentropy, 337–339, 363\\nepochs, 430\\nerror\\nirreducible, 17, 30\\nrate, 34\\nreducible, 17\\nterm, 16\\nEuclidean distance, 509, 522, 523,\\n529–532, 554\\nevent time, 470\\nexception, 45\\nexpected value, 18\\nexploratory data analysis, 504\\nexponential, 173\\nexponential family, 173\\nF-statistic, 84 600\\nIndex\\nfactor, 92\\nfactorial, 170\\nfailure time, 470\\nfalse\\ndiscovery proportion, 155, 573\\ndiscovery rate, 558, 573–577,\\n579–582\\nnegative, 155, 562\\npositive, 155, 562, 563\\npositive rate, 155, 156, 382\\nfamily-wise error rate, 565–573, 577\\nfeature, 15\\nfeature map, 406\\nfeature selection, 230\\nfeaturize, 414\\nfeed-forward neural network, 400\\nfigure, 48\\nfit, 21\\nfitted value, 101\\nflattening, 424\\nflexible, 21\\nfloating point, 43\\nforward stepwise selection, 86, 87,\\n233–234, 268\\nfunction, 40\\nFund data set, 12, 567–570, 572,\\n575, 576, 585, 588, 589\\nGamma, 173\\nGaussian (normal) distribution, 146,\\n147, 150, 172, 561\\ngeneralized additive model, 5, 24,\\n162, 289, 290, 305–309,\\n319\\ngeneralized linear model, 5, 135,\\n167–174, 217\\ngenerative model, 146–161\\nGini index, 337–339, 345, 346, 363\\nglobal minimum, 427\\ngradient, 428\\ngradient descent, 427\\nHarrell’s concordance index, 487\\nhazard function, 476–478\\nbaseline, 478\\nhazard rate, 476\\nHeart data set, 339, 340, 344–347,\\n352, 353, 382, 383\\nheatmap, 50\\nhelper, 311\\nheteroscedasticity, 103, 168\\nhidden layer, 400\\nhidden units, 400\\nhierarchical clustering, 525–530\\ndendrogram, 525–528\\ninversion, 529\\nlinkage, 529–530\\nhierarchical principle, 96\\nhigh-dimensional, 86, 234, 263\\nhinge loss, 385\\nHitters data set, 12, 332, 333,\\n336, 338, 339, 366, 425,\\n426, 437, 446\\nhold-out set, 202\\nHolm’s method, 568, 576, 585\\nhypergeometric distribution, 501\\nhyperparameter, 187\\nhyperplane, 367–372\\nhypothesis test, 76–77, 84, 103,\\n558–583\\nIMDb data set, 413, 415, 416, 418,\\n420, 437, 458, 467\\nimputation, 515\\nIncome data set, 16–18, 21–23\\nincrement, 60\\nindependent variable, 15\\nindexable, 186\\nindicator function, 292\\ninference, 17, 18\\ninner product, 379, 380\\ninput layer, 400\\ninput variable, 15\\nintegral, 301\\ninteraction, 70, 89, 95–98, 110–\\n111, 308\\nintercept, 71, 72\\ninterpolate, 432\\ninterpretability, 229\\ninversion, 529\\nirreducible error, 17, 36, 90, 110\\niterator, 312\\njoint distribution, 158\\nK-means clustering, 11, 521–524\\nK-nearest neighbors, 135, 164–167\\nclassifier, 11, 36–37\\nregression, 111–115 Index\\n601\\nKaplan–Meier survival curve, 472–\\n474, 483\\nkernel, 379–382, 384, 394\\nlinear, 380\\nnon-linear, 377–382\\npolynomial, 380, 382\\nradial, 381–383, 390\\nkernel density estimator, 159\\nkeyword, 46\\nKhan data set, 12, 579–581, 583,\\n590, 593\\nknot, 290, 294, 296–299\\nℓ1 norm, 244\\nℓ2 norm, 242\\nlag, 422\\nLaplace distribution, 251\\nlasso, 11, 24, 244–251, 265–266,\\n336, 385, 484\\nleaf, 333, 526\\nlearning rate, 429\\nleast squares, 5, 21, 71–72, 140,\\n141, 229\\nline, 73\\nweighted, 103\\nlevel, 92\\nleverage, 104–106\\nlikelihood function, 141\\nlinear, 2, 69–115\\nlinear combination, 128, 230, 253,\\n505\\nlinear discriminant analysis, 5, 11,\\n135, 138, 147–155, 164–\\n167, 377, 382\\nlinear kernel, 380\\nlinear model, 20, 69–115\\nlinear regression, 5, 11, 69–115,\\n172–173\\nmultiple, 80–90\\nsimple, 70–80\\nlink function, 172, 173\\nlinkage, 529–530, 548\\naverage, 529–530\\ncentroid, 529–530\\ncomplete, 526, 529–530\\nsingle, 529–530\\nlist, 41\\nlist comprehension, 123\\nlocal minimum, 427\\nlocal regression, 290\\nlog odds, 145\\nlog-rank test, 474–476, 483\\nlogistic function, 139\\nlogistic regression, 5, 11, 25, 135,\\n138–144, 164–167, 172–\\n173, 308–309, 377, 384–\\n385\\nmultinomial, 145, 163\\nmultiple, 142–144\\nlogit, 140\\nloss function, 300, 385\\nlow-dimensional, 262\\nLSTM RNN, 420\\nmain effects, 96\\nmajority vote, 344\\nMallow’s Cp, 87, 231, 232, 236–\\n238\\nMantel–Haenszel test, 475\\nmargin, 370, 385\\nmarginal distribution, 158\\nMarkov chain Monte Carlo, 353\\nmatrix completion, 515\\nmatrix multiplication, 10\\nmaximal margin\\nclassifier, 367–372\\nhyperplane, 370\\nmaximum likelihood, 139–141, 143,\\n170\\nmean squared error, 28\\nmesh, 53\\nmethod, 43\\nminibatch, 429\\nmisclassification error, 35\\nmissing at random, 515\\nmissing data, 56, 515–520\\nmixed selection, 87\\nMNIST data set, 402–404, 406, 430,\\n431, 441, 444, 445, 448\\nmodel assessment, 201\\nmodel selection, 201\\nmodule, 42\\nmulticollinearity, 108, 266\\nmultinomial logistic regression, 145,\\n163\\nmultiple testing, 557–583\\nmulti-task learning, 403\\nmultivariate Gaussian, 150 602\\nIndex\\nmultivariate normal, 150\\nnaive Bayes, 135, 158–161, 164–\\n167\\nnamespace, 116\\nnatural spline, 297, 298, 301, 317\\nNCI60 data set, 4, 5, 12, 546, 548–\\n550\\nnegative binomial, 173\\nnegative predictive value, 155, 156\\nneural network, 5, 399\\nnode\\ninternal, 333\\npurity, 337–339\\nterminal, 333\\nnoise, 21, 252\\nnon-linear, 2, 11, 289–329\\ndecision boundary, 377–382\\nkernel, 377–382\\nnon-parametric, 20, 22–23, 111–\\n115, 193\\nnormal (Gaussian) distribution, 146,\\n147, 150, 172, 476, 561\\nnotebook, 40\\nnull, 152\\ndistribution, 561, 578\\nhypothesis, 76, 559\\nmodel, 87, 231, 245\\nnull rate, 186\\nNYSE data set, 12, 422–424, 466,\\n467\\nOccam’s razor, 426\\nodds, 140, 145, 195\\nOJ data set, 12, 365, 398\\none-hot encoding, 92, 126, 403\\none-standard-error rule, 240\\none-versus-all, 384\\none-versus-one, 384\\none-versus-rest, 384\\noptimal separating hyperplane, 370\\noptimism of training error, 30\\nordered categorical variable, 315\\northogonal, 257, 506\\nbasis, 125\\nout-of-bag, 345\\noutlier, 103–104\\noutput variable, 15\\nover-parametrized, 465\\noverdispersion, 172\\noverfitting, 21, 23, 25, 30–31, 88,\\n152, 233, 371\\np-value, 77, 82, 560–562, 578–579\\nadjusted, 586\\npackage, 42\\nparameter, 71\\nparametric, 20–22, 111–115\\npartial least squares, 254, 260–262,\\n282\\npartial likelihood, 480\\npath algorithm, 249\\npermutation, 578\\npermutation approach, 577–582\\nperpendicular, 257\\nPoisson distribution, 169, 172\\nPoisson regression, 135, 167–173\\npolynomial\\nkernel, 380, 382\\nregression, 98–99, 289–292, 294–\\n295\\npooling, 410\\npopulation regression line, 73\\nPortfolio data set, 12\\npositive predictive value, 155, 156\\nposterior\\ndistribution, 251\\nmode, 251\\nprobability, 147\\npower, 108, 155, 563\\nprecision, 155\\nprediction, 17\\ninterval, 90, 110\\npredictor, 15\\nprincipal components, 505\\nanalysis, 11, 254–260, 504–515\\nloading vector, 505, 506\\nmissing values, 515–520\\nproportion of variance explained,\\n510–515, 547\\nregression, 11, 254–260, 280–\\n282, 504, 515\\nscore vector, 506\\nscree plot, 514–515\\nprior\\ndistribution, 251\\nprobability, 146\\nprobability density function, 477,\\n478 Index\\n603\\nprojection, 230\\nproportional hazards assumption,\\n478\\npruning, 336\\ncost complexity, 336\\nweakest link, 336\\nPublication data set, 12, 482–\\n487\\nPython objects and functions\\n%%capture, 458\\niloc[], 58\\nloc[], 57\\nAgglomerativeClustering(),\\n543\\nanova(), 313\\nanova_lm(), 125, 129, 312,\\n313\\naxhline(), 122, 551\\naxline(), 121, 129, 329\\nBART(), 362\\nbiplot, 537\\nboot_SE(), 223\\nboxplot(), 62, 66\\nbs(), 315, 327\\nBSpline(), 315\\nclone(), 222\\ncolumns.drop(), 122\\ncompute_linkage(), 544\\nconfusion_table(), 176\\ncontour(), 50\\ncorr(), 129, 174\\ncost_complexity_pruning_path(),\\n357\\nCoxPHFitter(), 491\\ncross_val_predict(), 270\\ncross_validate(), 218, 219,\\n226\\ncumsum(), 539\\ncut_tree(), 545\\ndata.frame(), 227\\nDataset, 440\\ndecision_function(), 392\\nDecisionTreeClassifier(),\\n354, 355\\nDecisionTreeRegressor(), 354\\ndef, 121\\ndendrogram(), 544\\ndescribe(), 62, 66\\ndir(), 116\\ndrop(), 179\\ndropna(), 56, 268, 461\\nDTC(), see DecisionTreeClassifier()\\nDTR(), see DecisionTreeRegressor()\\ndtype, 43\\nElasticNetCV(), 279\\nenumerate(), 217\\nexport_text(), 356\\nexport_tree(), 365\\nfit(), 118, 181, 218\\nfit_transform(), 119\\nfor, 59\\nGaussianNB(), 182\\nGBR(), see GradientBoosting-\\nRegressor()\\nget_dummies(), 461\\nget_influence(), 121\\nget_prediction(), 120, 314\\nget_rdataset(), 535\\nglm(), 313\\nglob(), 437\\nGradientBoostingClassifier(),\\n361\\nGradientBoostingRegressor(),\\n354, 361\\nGridSearchCV(), 276\\ngroupby(), 490\\nhist(), 62\\niloc[], 58, 59\\nimport, 42\\nimshow(), 50, 449\\nISLP.bart, 362\\nISLP.cluster, 544\\njson, 437\\nKaplanMeierFitter(), 502\\nkeras, 437\\nKFold(), 219\\nKMeans(), 542, 543\\nKmeans(), 542\\nKNeighborsClassifier(), 183\\nlambda, 58\\nLDA(), see LinearDiscriminant-\\nAnalysis()\\nlegend(), 132\\nlifelines, 490\\nLinearDiscriminantAnalysis(),\\n174, 179\\nLinearGAM(), 317\\nLinearRegression(), 280 604\\nIndex\\nload_data(), 117\\nloc[], 58, 59, 177\\nlog_loss(), 355\\nLogisticGAM(), 323\\nlogrank_test(), 490\\nlowess(), 324\\nmatplotlib, 48\\nmax(), 66\\nmean(), 48\\nmedian(), 197\\nmin(), 66\\nMNIST(), 444\\nModelSpec(), 116–118, 122,\\n124, 267\\nMS(), see ModelSpec()\\nmult_test(), see multipletests()\\nmultipletests(), 586\\nmultipletests(), 583, 589\\nmultivariate_logrank_test(),\\n496\\nNaturalSpline(), 317, 319\\nndim, 42\\nnn.RNN(), 461\\nnormal(), 132, 286, 555\\nnp, see numpy\\nnp.all(), 54, 180\\nnp.allclose(), 190\\nnp.any(), 54\\nnp.arange(), 51\\nnp.argmax(), 122\\nnp.array(), 42\\nnp.concatenate(), 133\\nnp.corrcoef(), 46, 554\\nnp.empty(), 224\\nnp.isnan(), 268\\nnp.ix_(), 53\\nnp.linalg.svd(), 539\\nnp.linspace(), 50\\nnp.logspace(), 318\\nnp.mean(), 47, 176\\nnp.nan, 60\\nnp.nanmean(), 541\\nnp.percentile(), 228\\nnp.power(), 219\\nnp.random.choice(), 553\\nnp.random.default_rng(), 46,\\n47\\nnp.random.normal(), 45\\nnp.sqrt(), 45\\nnp.squeeze(), 457\\nnp.std(), 47\\nnp.sum(), 43\\nnp.var(), 47\\nnp.where(), 180\\nns(), 317\\nnumpy, 42, 555\\nos.chdir(), 55\\nouter(), 219\\npairwise_distances(), 554\\npairwise_tukeyhsd(), 587\\npandas, 55\\nparams, 175\\npartial(), 222, 269\\nPCA(), 280, 537, 540, 554\\npd, see pandas\\npd.crosstab(), 555\\npd.cut(), 315\\npd.get_dummies(), 314\\npd.plotting.scatter_matrix(),\\n62\\npd.qcut(), 314, 315\\npd.read_csv(), 55, 556\\npd.Series(), 62\\nPipeline(), 275\\nplot(), 48, 61, 356, 490\\nplot.scatter(), 120\\nplot_gam(), 321\\nplot_svm(), 398\\nPLSRegression(), 282\\npoly(), 125, 313, 327\\npredict(), 175, 178, 181, 216,\\n218, 323, 358\\npredict_survival_function(),\\n493\\nprint(), 40\\npvalues, 175\\npygam, 307, 317\\npytorch_lightning, 435\\nQDA(), see QuadraticDiscriminant-\\nAnalysis()\\nQuadraticDiscriminantAnalysis(),\\n174, 181\\nrandom(), 555\\nRandomForestRegressor(), 354,\\n360\\nread_image(), 436\\nreindex(), 461\\nreshape(), 43 Index\\n605\\nreturn, 198\\nRF(), see RandomForestRegressor()\\nrng, see np.random.default_rng()\\nrng.choice(), 60\\nrng.standard_normal(), 60\\nroc_curve(), 392\\nRocCurveDisplay.from_estimator(),\\n387\\nsavefig(), 50\\nscatter(), 49, 61\\nscipy.interpolate, 315\\nscore(), 218, 461\\nseed_everything(), 436\\nset_index(), 57\\nset_title(), 49\\nset_xlabel(), 49\\nset_xscale(), 198\\nset_ylabel(), 49\\nset_yscale(), 198\\nshape, 43\\nShuffleSplit(), 219\\nsim_time(), 495\\nSimpleDataModule(), 441\\nSimpleModule.classification(),\\n446\\nSimpleModule.regression(),\\n442\\nskl, see sklearn.linear_model\\nskl.ElasticNet(), 273, 277\\nskl.ElasticNet.path, 274\\nskl.ElasticNet.path(), 273\\nsklearn, 118, 181\\nsklearn.ensemble, 360\\nsklearn.linear_model, 267\\nsklearn.model_selection, 267\\nsklearn_selected(), 269\\nsklearn_selection_path(),\\n270\\nsklearn_sm(), 218\\nskm, see sklearn.model_selection\\nskm.cross_val_predict(), 271\\nskm.KFold(), 271\\nskm.ShuffleSplit(), 272\\nslice(), 51, 462\\nsm, see statsmodels\\nsm.GLM(), 174, 192, 226\\nsm.Logit(), 174\\nsm.OLS(), 118, 129, 174, 319\\nStandardScaler(), 185, 438,\\n537, 555\\nstatsmodels, 116, 173\\nstd(), 186\\nStepwise(), 269\\nstr.contains(), 59\\nsubplots(), 48\\nsum(), 43, 268\\nsummarize(), 118, 129, 223,\\n226\\nsummary(), 119, 322, 587\\nsuper(), 440\\nSupportVectorClassifier(),\\n387, 389–391, 393\\nSupportVectorRegression(),\\n394\\nSVC(), see SupportVector-\\nClassifier()\\nsvd(), 539\\nSVR(), see SupportVector-\\nRegression()\\nTensorDataset(), 441\\nto_numpy(), 437\\ntorch, 435\\ntorchinfo, 436\\ntorchmetrics, 436\\ntorchvision, 436\\nToTensor(), 444\\ntrain_test_split(), 186, 216\\ntransform(), 118, 119\\nttest_1samp(), 584\\nttest_ind(), 590\\nttest_rel(), 587\\ntuple, 43\\nuniform(), 555\\nvalue_counts(), 66\\nvar(), 536\\nvariance_inflation_factor(),\\n116, 124\\nVIF(), see variance_inflation-\\n_factor()\\nwhere(), 355\\nzip(), 60, 312\\nq-values, 589\\nquadratic, 98\\nquadratic discriminant analysis, 4,\\n135, 156–157, 164–167 606\\nIndex\\nqualitative, 2, 27, 91, 135, 167,\\n202\\nvariable, 91–94\\nquantitative, 2, 27, 91, 135, 167,\\n202\\nradial kernel, 381, 383, 390\\nrandom forest, 11, 331, 343, 346–\\n347, 354, 360–361\\nrandom seed, 46\\nre-sampling, 577–582\\nrecall, 155\\nreceiver operating characteristic (ROC),\\n154, 382–383\\nrecommender systems, 516\\nrectified linear unit, 401\\nrecurrent neural network, 416–427\\nrecursive binary splitting, 334, 337,\\n338\\nreducible error, 17, 90\\nregression, 2, 11, 27\\nlocal, 289, 290, 304–305\\npiecewise polynomial, 294–295\\npolynomial, 289–292, 299\\nspline, 289, 294\\ntree, 331–337, 358–360\\nregularization, 230, 240, 406, 484–\\n486\\nReLU, 401\\nresampling, 201–214\\nresidual, 71, 81\\nplot, 100\\nstandard error, 75, 77–78, 88–\\n89, 109\\nstudentized, 104\\nsum of squares, 71, 79, 81\\nresiduals, 263, 348\\nresponse, 15\\nridge regression, 11, 240–244, 385,\\n484\\nrisk set, 473\\nrobust, 374, 376, 535\\nROC curve, 154, 382–383, 486–\\n487\\nR2, 77–80, 88, 109, 238\\nrug plot, 314\\nscale equivariant, 242\\nScheffé’s method, 572\\nscree plot, 512, 514–515\\nelbow, 514\\nsemi-supervised learning, 27\\nsensitivity, 153, 155, 156\\nseparating hyperplane, 367–372\\nSeq2Seq, 425\\nsequence, 41\\nshrinkage, 230, 240, 484–486\\npenalty, 240\\nsigmoid, 401\\nsignal, 252\\nsignature, 45\\nsingular value decomposition, 539\\nslack variable, 375\\nslice, 51\\nslope, 71, 72\\nSmarket data set, 2, 3, 12, 173,\\n184, 196\\nsmoother, 308\\nsmoothing spline, 290, 300–303\\nsoft margin classifier, 372–374\\nsoft-thresholding, 250\\nsoftmax, 145, 405\\nsparse, 244, 252\\nsparse matrix format, 414\\nsparsity, 244\\nspecificity, 153, 155, 156\\nspline, 289, 294–303\\ncubic, 296\\nlinear, 296\\nnatural, 297, 301\\nregression, 289, 294–299\\nsmoothing, 30, 290, 300–303\\nthin-plate, 22\\nstandard error, 75, 101\\nstandardize, 185\\nstatistical model, 1\\nstep function, 111, 289, 292–293\\nstepwise model selection, 11, 231,\\n233\\nstochastic gradient descent, 429\\nstring, 41\\nstring interpolation, 490\\nstump, 349\\nsubset selection, 230–240\\nsubtree, 336\\nsupervised learning, 25–27, 261\\nsupport vector, 371, 376, 385\\nclassifier, 367, 372–377\\nmachine, 5, 11, 24, 377–386 Index\\n607\\nregression, 386\\nsurvival\\nanalysis, 469–502\\ncurve, 472, 483\\nfunction, 472\\ntime, 470\\nsynergy, 70, 89, 95–98, 110–111\\nsystematic, 16\\nt-distribution, 77, 165\\nt-statistic, 76\\nt-test\\none-sample, 583, 584, 588\\npaired, 587\\ntwo-sample, 559, 570, 571, 577–\\n581, 584, 590\\ntest\\nerror, 35, 37, 176\\nMSE, 28–32\\nobservations, 28\\nset, 30\\nstatistic, 559\\ntheoretical null distribution, 577\\ntime series, 101\\ntotal sum of squares, 79\\ntracking, 102\\ntrain, 21\\ntraining\\ndata, 20\\nerror, 35, 37, 176\\nMSE, 28–31\\ntransformer, 311\\ntree, 331–342\\ntree-based method, 331\\ntrue negative, 155\\ntrue positive, 155\\ntrue positive rate, 155, 156, 382\\ntruncated power basis, 296\\nTukey’s method, 571, 585, 587\\ntuning parameter, 187, 240, 484\\ntwo-sample t-test, 474\\nType I error, 155, 562–565\\nType I error rate, 563\\nType II error, 155, 563, 568, 584\\nunsupervised learning, 25–27, 255,\\n260, 503–552\\nUSArrests data set, 12, 507, 508,\\n510, 512, 513, 515, 516,\\n518, 519\\nvalidation set, 202\\napproach, 202–204\\nvariable, 15\\ndependent, 15\\ndummy, 91–94, 97–98\\nimportance, 346, 360\\nindependent, 15\\nindicator, 35\\ninput, 15\\noutput, 15\\nqualitative, 91–94, 97–98\\nselection, 86, 230, 244\\nvariance, 18, 31–34, 159\\ninflation factor, 108–110, 123\\nvarying coefficient model, 305\\nWage data set, 1, 2, 8, 9, 12, 290,\\n291, 293, 295, 297–300,\\n302–306, 309, 315, 327\\nweak learner, 343\\nweakest link pruning, 336\\nWeekly data set, 12, 196, 226\\nweight freezing, 412, 419\\nweight sharing, 418\\nweighted least squares, 103, 304\\nweights, 404\\nwith replacement, 214\\nwithin class covariance, 150\\nwrapper, 217')"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "text"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "text_splitter = RecursiveCharacterTextSplitter(chunk_size=4000, chunk_overlap=0, add_start_index=True)\n",
    "chunks = text_splitter.split_documents([text])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "vector_store = Chroma(collection_name='islp', embedding_function=OpenAIEmbeddings(model='text-embedding-3-small'), persist_directory='./chromadb', create_collection_if_not_exists=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Documents already exist in vector store\n"
     ]
    }
   ],
   "source": [
    "existing_docs = vector_store.similarity_search(\"\", k=1) \n",
    "\n",
    "if not existing_docs:\n",
    "    uuids = [str(uuid4()) for _ in range(len(chunks))]\n",
    "    vector_store.add_documents(documents=chunks, ids=uuids)\n",
    "    print(\"Documents added to vector store\")\n",
    "else:\n",
    "    print(\"Documents already exist in vector store\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "retriever = vector_store.as_retriever(\n",
    "    search_type=\"mmr\",\n",
    "    search_kwargs={\"k\": 5, \"fetch_k\": 10, \"lambda_mult\": 0.25}\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "prompt = '''\n",
    "You are a formatting and organization assistant.\n",
    "\n",
    "Your job is to take the raw information retrieved by a RAG system (provided below) and process it to create a clear, well-structured, and logically ordered context. This context will be used by another model to answer a user query, so you must not answer the query yourself.\n",
    "\n",
    "Instructions:\n",
    "- Organize the information into sections or bullet points.\n",
    "- Remove duplicates and irrelevant or conflicting data.\n",
    "- Preserve technical or factual accuracy.\n",
    "- Do not fabricate or infer missing information.\n",
    "- Make the result easy for another model to read and use as direct context.\n",
    "- Ensure all the information from the retrieved chunks is present in the \n",
    "\n",
    "Below is the raw retrieved data:\n",
    "---\n",
    "{retrieved_chunks_here}\n",
    "---\n",
    "\n",
    "Return only the cleaned and structured context below.\n",
    "\n",
    "Context:\n",
    "'''\n",
    "\n",
    "prompt = ChatPromptTemplate.from_messages(\n",
    "    [\n",
    "        (\"system\", prompt),\n",
    "    ]\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "retreived_chunks = retriever.invoke(\"What is PCA?\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "prompt = prompt.format(retrieved_chunks_here=retreived_chunks)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "response = client.chat.completions.create(\n",
    "    model=\"deepseek-chat\",\n",
    "    messages=[\n",
    "        {\"role\": \"system\", \"content\": prompt}    \n",
    "        ],\n",
    "    stream=False\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "### **Principal Components Analysis (PCA) Overview**  \n",
      "- **Definition**: PCA is an unsupervised learning technique used to summarize a large set of correlated variables into a smaller number of representative variables (principal components) that explain most of the variability in the original data.  \n",
      "- **Purpose**:  \n",
      "  - Data visualization (observations or variables).  \n",
      "  - Data imputation (filling missing values).  \n",
      "  - Derived variables for supervised learning.  \n",
      "\n",
      "### **Key Concepts in PCA**  \n",
      "1. **Principal Components**:  \n",
      "   - Directions in feature space where the original data show high variability.  \n",
      "   - Represent lines/subspaces closest to the data cloud.  \n",
      "   - Computed via eigenvalue decomposition of the covariance matrix.  \n",
      "\n",
      "2. **Low-Dimensional Representation**:  \n",
      "   - PCA finds a low-dimensional approximation of the data that retains maximal variance.  \n",
      "   - Useful for visualizing high-dimensional data (e.g., reducing 3D data to 2D).  \n",
      "\n",
      "3. **Standardization**:  \n",
      "   - Variables should be standardized (mean = 0, SD = 1) if measured in different units or scales.  \n",
      "   - Prevents high-variance variables from dominating the principal components.  \n",
      "\n",
      "### **Mathematical Formulation**  \n",
      "- **Objective**: Minimize the residual sum of squares for the approximation:  \n",
      "  \\[\n",
      "  \\text{minimize} \\sum_{i=1}^n \\sum_{j=1}^p \\left( x_{ij} - \\sum_{m=1}^M a_{im} b_{jm} \\right)^2\n",
      "  \\]  \n",
      "  where:  \n",
      "  - \\(a_{im}\\) = scores (projections of observations onto components).  \n",
      "  - \\(b_{jm}\\) = loadings (weights of original variables in components).  \n",
      "\n",
      "- **Variance Decomposition**:  \n",
      "  Total variance = Variance explained by first \\(M\\) PCs + Mean squared error of approximation.  \n",
      "\n",
      "### **Proportion of Variance Explained (PVE)**  \n",
      "- **Formula**:  \n",
      "  \\[\n",
      "  \\text{PVE}_m = \\frac{\\sum_{i=1}^n z_{im}^2}{\\sum_{j=1}^p \\sum_{i=1}^n x_{ij}^2}\n",
      "  \\]  \n",
      "  where \\(z_{im}\\) is the \\(m\\)-th principal component score.  \n",
      "- **Interpretation**:  \n",
      "  - PVE quantifies how much variance each PC captures.  \n",
      "  - Cumulative PVE helps decide the number of components to retain (e.g., 95% variance explained).  \n",
      "\n",
      "### **Practical Implementation**  \n",
      "1. **Steps**:  \n",
      "   - Center and scale the data (e.g., using `StandardScaler`).  \n",
      "   - Compute PCA (e.g., via `sklearn.decomposition.PCA`).  \n",
      "   - Extract scores, loadings, and explained variance.  \n",
      "\n",
      "2. **Outputs**:  \n",
      "   - **Scores**: Low-dimensional representation of observations.  \n",
      "   - **Loadings**: Contribution of original variables to each PC.  \n",
      "   - **Explained Variance**: PVE for each component.  \n",
      "\n",
      "3. **Visualization**:  \n",
      "   - **Biplot**: Combines scores and loadings to show relationships between observations and variables.  \n",
      "\n",
      "### **Comparison with Other Methods**  \n",
      "- **PCR (Principal Components Regression)**: Uses PCA for dimensionality reduction before regression.  \n",
      "- **PLS (Partial Least Squares)**: Supervised alternative to PCA; considers the response variable when deriving components.  \n",
      "\n",
      "### **Example Applications**  \n",
      "- **Gene Expression Data**: Identify subgroups of genes or samples.  \n",
      "- **E-commerce**: Cluster shoppers based on browsing/purchase history.  \n",
      "- **Search Engines**: Personalize results using click histories of similar users.  \n",
      "\n",
      "### **Code Example (Python)**  \n",
      "```python\n",
      "from sklearn.decomposition import PCA\n",
      "from sklearn.preprocessing import StandardScaler\n",
      "\n",
      "# Standardize data\n",
      "scaler = StandardScaler()\n",
      "X_scaled = scaler.fit_transform(X)\n",
      "\n",
      "# Perform PCA\n",
      "pca = PCA()\n",
      "scores = pca.fit_transform(X_scaled)\n",
      "\n",
      "# Access results\n",
      "loadings = pca.components_\n",
      "explained_variance = pca.explained_variance_ratio_\n",
      "```  \n",
      "\n",
      "This structured context provides a comprehensive yet concise overview of PCA, its mathematical foundations, practical implementation, and applications.\n"
     ]
    }
   ],
   "source": [
    "print(response.choices[0].message.content)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "xade-ai",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
